{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "<center>\n",
    "    <h1 id = \"top-of-app\"> \n",
    "        <div style=\"font-size:3rem;font-weight:500\"> <img src=\"{{ url_for('static', filename='logo.png') }}\" style=\"height:45px;padding:0 5px;display:inline\"/> Gene Set Library Synopsis Appyter</div>\n",
    "    </h1>\n",
    "    <br>\n",
    "    <div style=\"font-size:2rem;font-weight:500\">An appyter for the analysis and visualization of gene set libraries</div>\n",
    "</center>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# Gene Set Library Synopsis\n",
    "This Appyter processes, analyzes, and visualizes a collection of gene sets, also known as a gene set library. \n",
    "\n",
    "First it will generate summary statistics describing the size of the library and its component gene sets, as well as how well studied the genes and gene sets are. \n",
    "\n",
    "Then the Appyter will use text vectorization (TF-IDF) and dimensionality reduction (UMAP) to visualize the library as a scatterplot.\n",
    "\n",
    "To assess gene set similarity, pairwise Jaccard Indexes will be calculated, and this statistic will serve as the basis for a heatmap. The Appyter will also produce a set of figures focusing on the gene sets with the highest overlap.\n",
    "\n",
    "Finally, the Appyter will present additional plots describing the composition of your library, including bar graphs of most frequent and most studied genes, a scatterplot of gene sets arranged by size and publication count, and a scatterplot of the library among all Enrichr libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%% Imports\n",
    "import appyter\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import base64\n",
    "import math\n",
    "import seaborn as sns\n",
    "import fastcluster\n",
    "\n",
    "import matplotlib.pyplot as plt; plt.rcdefaults()\n",
    "import matplotlib.colors as colors\n",
    "%matplotlib inline\n",
    "\n",
    "import IPython\n",
    "from IPython.display import HTML, display, FileLink, Markdown, IFrame\n",
    "\n",
    "import urllib\n",
    "\n",
    "import itertools\n",
    "from itertools import chain\n",
    "from scipy.spatial.distance import squareform, pdist, jaccard\n",
    "from scipy.cluster.hierarchy import linkage\n",
    "\n",
    "from bokeh.io import output_notebook, export_svg\n",
    "from bokeh.io.export import get_screenshot_as_png\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool, CustomJS, ColumnDataSource, Span, ranges, LabelSet, BasicTicker, ColorBar, LinearColorMapper, PrintfTickFormatter\n",
    "from bokeh.layouts import layout, row, column, gridplot\n",
    "from bokeh.palettes import all_palettes, linear_palette, Turbo256, Spectral6\n",
    "from bokeh.transform import factor_cmap, transform\n",
    "\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_samples, silhouette_score\n",
    "\n",
    "import umap.umap_ as umap\n",
    "from sklearn.decomposition import NMF\n",
    "\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Notebook display functions \n",
    "def figure_legend(label,title=\"\",content=\"\"):\n",
    "    if len(title)>0:\n",
    "        display(HTML(f\"<div><b>{label}</b>: <i>{title}</i>. {content} </div>\"))\n",
    "    else:\n",
    "        display(HTML(f\"<div><b>{label}</b>: {content} </div>\"))\n",
    "\n",
    "# Figure and table counters\n",
    "fig_count = 1\n",
    "table_count = 1\n",
    "\n",
    "# Output a table of genes and counts.\n",
    "def create_download_link(df, title, filename):  \n",
    "    csv = df.to_csv(index = False)\n",
    "    b64 = base64.b64encode(csv.encode())\n",
    "    payload = b64.decode()\n",
    "    html = '<a download=\"{filename}\" href=\"data:text/csv;base64,{payload}\" target=\"_blank\">{title}</a>'\n",
    "    html = html.format(payload=payload, title=title, filename=filename)\n",
    "    return HTML(html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "{% do SectionField(\n",
    "    name='geneEntrySection', \n",
    "    title='1. Submit a Gene Set Library',\n",
    "    subtitle='Upload a GMT file containing a gene set library (gene names must be in gene symbol format) or select an existing Enrichr library. You may also choose to use the default gene library provided.',\n",
    "    img='data-upload-icon.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='barGraphSection', \n",
    "    title='2. Bar Graphs and Histograms',\n",
    "    subtitle='Bar graphs and histograms describing your library will be generated. Choose parameters to customize these visualizations',\n",
    "    img='bar-icon.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='simSection', \n",
    "    title='3. Gene Set Similarity',\n",
    "    subtitle='Similarity between gene sets will be assessed using the Jaccard Index. You may choose how this information is displayed.',\n",
    "    img='set-similarity.png'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Inputting libraries and settings\n",
    "{% set library_kind = TabField(\n",
    "    name = 'library_kind',\n",
    "    label = 'Library',\n",
    "    default = 'Upload a library',\n",
    "    description = '',\n",
    "    choices = {\n",
    "        'Upload a library': [\n",
    "            FileField(\n",
    "                name = 'library_filename',\n",
    "                label = 'Gene Library File (.gmt or .txt)',\n",
    "                default = 'CellMarker_Augmented_2021.gmt', \n",
    "                examples = {'CellMarker_Augmented_2021.gmt': url_for('static', filename = 'CellMarker_Augmented_2021.gmt'),\n",
    "                            'CellMarker_Augmented_2021.txt': url_for('static', filename = 'CellMarker_Augmented_2021.txt')}, \n",
    "                description = 'GMT is a tab-delimited file format that describes sets. Visit https://bit.ly/35crtXQ for more information and http://www.molmine.com/magma/fileformats.html to create your own.',\n",
    "                section = 'geneEntrySection'),\n",
    "            MultiCheckboxField(\n",
    "                name='species',\n",
    "                label='Species',\n",
    "                description='Which species are represented by your gene set library?',\n",
    "                default=['human'],\n",
    "                section='geneEntrySection',\n",
    "                choices=[\n",
    "                    'human',\n",
    "                    'mouse',\n",
    "                    'other',\n",
    "                ],\n",
    "            )\n",
    "        ],\n",
    "        \n",
    "        'Select a library from Enrichr': [\n",
    "            ChoiceField(\n",
    "                name = 'enrichr_library',\n",
    "                description = 'Select one Enrichr library whose genes will be counted',\n",
    "                label = 'Enrichr Library',\n",
    "                default = 'CellMarker_Augmented_2021',\n",
    "                section = 'geneEntrySection',\n",
    "                choices = [\n",
    "                    'ARCHS4_Cell-lines',\n",
    "                    'ARCHS4_IDG_Coexp',\n",
    "                    'ARCHS4_Kinases_Coexp',\n",
    "                    'ARCHS4_TFs_Coexp',\n",
    "                    'ARCHS4_Tissues',\n",
    "                    'Achilles_fitness_decrease',\n",
    "                    'Achilles_fitness_increase',\n",
    "                    'Aging_Perturbations_from_GEO_down',\n",
    "                    'Aging_Perturbations_from_GEO_up',\n",
    "                    'Allen_Brain_Atlas_10x_scRNA_2021',\n",
    "                    'Allen_Brain_Atlas_down',\n",
    "                    'Allen_Brain_Atlas_up',\n",
    "                    'Azimuth_Cell_Types_2021',\n",
    "                    'BioCarta_2013',\n",
    "                    'BioCarta_2015',\n",
    "                    'BioCarta_2016',\n",
    "                    'BioPlanet_2019',\n",
    "                    'BioPlex_2017',\n",
    "                    'CCLE_Proteomics_2020',\n",
    "                    'CORUM',\n",
    "                    'COVID-19_Related_Gene_Sets',\n",
    "                    'Cancer_Cell_Line_Encyclopedia',\n",
    "                    'CellMarker_Augmented_2021',\n",
    "                    'ChEA_2013',\n",
    "                    'ChEA_2015',\n",
    "                    'ChEA_2016',\n",
    "                    'Chromosome_Location',           \n",
    "                    'Chromosome_Location_hg19',\n",
    "                    'ClinVar_2019',\n",
    "                    'dbGaP',\n",
    "                    'DSigDB',\n",
    "                    'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "                    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "                    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "                    'Descartes_Cell_Types_and_Tissue_2021',\n",
    "                    'DisGeNET',\n",
    "                    'Disease_Perturbations_from_GEO_down',\n",
    "                    'Disease_Perturbations_from_GEO_up',\n",
    "                    'Disease_Signatures_from_GEO_down_2014',\n",
    "                    'Disease_Signatures_from_GEO_up_2014',\n",
    "                    'DrugMatrix',\n",
    "                    'Drug_Perturbations_from_GEO_2014',\n",
    "                    'Drug_Perturbations_from_GEO_down',\n",
    "                    'Drug_Perturbations_from_GEO_up',\n",
    "                    'ENCODE_Histone_Modifications_2013',\n",
    "                    'ENCODE_Histone_Modifications_2015',\n",
    "                    'ENCODE_TF_ChIP-seq_2014',\n",
    "                    'ENCODE_TF_ChIP-seq_2015',\n",
    "                    'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "                    'ESCAPE',\n",
    "                    'Elsevier_Pathway_Collection',\n",
    "                    'Enrichr_Libraries_Most_Popular_Genes',\n",
    "                    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "                    'Enrichr_Users_Contributed_Lists_2020',\n",
    "                    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "                    'GO_Biological_Process_2013',\n",
    "                    'GO_Biological_Process_2015',\n",
    "                    'GO_Biological_Process_2017',\n",
    "                    'GO_Biological_Process_2017b',\n",
    "                    'GO_Biological_Process_2018',\n",
    "                    'GO_Cellular_Component_2013',\n",
    "                    'GO_Cellular_Component_2015',\n",
    "                    'GO_Cellular_Component_2017',\n",
    "                    'GO_Cellular_Component_2017b',\n",
    "                    'GO_Cellular_Component_2018',\n",
    "                    'GO_Molecular_Function_2013',\n",
    "                    'GO_Molecular_Function_2015',\n",
    "                    'GO_Molecular_Function_2017',\n",
    "                    'GO_Molecular_Function_2017b',\n",
    "                    'GO_Molecular_Function_2018',\n",
    "                    'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "                    'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "                    'GWAS_Catalog_2019',\n",
    "                    'GeneSigDB',\n",
    "                    'Gene_Perturbations_from_GEO_down',\n",
    "                    'Gene_Perturbations_from_GEO_up',\n",
    "                    'Genes_Associated_with_NIH_Grants',\n",
    "                    'Genome_Browser_PWMs',\n",
    "                    'HMDB_Metabolites',\n",
    "                    'HMS_LINCS_KinomeScan',\n",
    "                    'HomoloGene',\n",
    "                    'HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression',\n",
    "                    'HumanCyc_2015',\n",
    "                    'HumanCyc_2016',\n",
    "                    'Human_Gene_Atlas',\n",
    "                    'Human_Phenotype_Ontology',\n",
    "                    'huMAP',\n",
    "                    'InterPro_Domains_2019',\n",
    "                    'Jensen_COMPARTMENTS',\n",
    "                    'Jensen_DISEASES',\n",
    "                    'Jensen_TISSUES',\n",
    "                    'KEA_2013',\n",
    "                    'KEA_2015',\n",
    "                    'KEGG_2013',\n",
    "                    'KEGG_2015',\n",
    "                    'KEGG_2016',\n",
    "                    'KEGG_2019_Human',\n",
    "                    'KEGG_2019_Mouse',\n",
    "                    'Kinase_Perturbations_from_GEO_down',\n",
    "                    'Kinase_Perturbations_from_GEO_up',\n",
    "                    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "                    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "                    'LINCS_L1000_Chem_Pert_down',\n",
    "                    'LINCS_L1000_Chem_Pert_up',\n",
    "                    'LINCS_L1000_Ligand_Perturbations_down',\n",
    "                    'LINCS_L1000_Ligand_Perturbations_up',\n",
    "                    'Ligand_Perturbations_from_GEO_down',\n",
    "                    'Ligand_Perturbations_from_GEO_up',\n",
    "                    'lncHUB_lncRNA_Co-Expression',\n",
    "                    'MCF7_Perturbations_from_GEO_down',\n",
    "                    'MCF7_Perturbations_from_GEO_up',\n",
    "                    'MGI_Mammalian_Phenotype_2013',\n",
    "                    'MGI_Mammalian_Phenotype_2017',\n",
    "                    'MGI_Mammalian_Phenotype_Level_3',\n",
    "                    'MGI_Mammalian_Phenotype_Level_4',\n",
    "                    'MGI_Mammalian_Phenotype_Level_4_2019',\n",
    "                    'MSigDB_Computational',\n",
    "                    'MSigDB_Hallmark_2020',\n",
    "                    'MSigDB_Oncogenic_Signatures',\n",
    "                    'Microbe_Perturbations_from_GEO_down',\n",
    "                    'Microbe_Perturbations_from_GEO_up',\n",
    "                    'miRTarBase_2017',\n",
    "                    'Mouse_Gene_Atlas',\n",
    "                    'NCI-60_Cancer_Cell_Lines',\n",
    "                    'NCI-Nature_2015',\n",
    "                    'NCI-Nature_2016',\n",
    "                    'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "                    'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "                    'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "                    'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "                    'NURSA_Human_Endogenous_Complexome',\n",
    "                    'OMIM_Disease',\n",
    "                    'OMIM_Expanded',\n",
    "                    'Old_CMAP_down',\n",
    "                    'Old_CMAP_up',\n",
    "                    'PanglaoDB_Augmented_2021',\n",
    "                    'PPI_Hub_Proteins',\n",
    "                    'Panther_2015',\n",
    "                    'Panther_2016',\n",
    "                    'Pfam_Domains_2019',\n",
    "                    'Pfam_InterPro_Domains',\n",
    "                    'PheWeb_2019',\n",
    "                    'Phosphatase_Substrates_from_DEPOD',\n",
    "                    'ProteomicsDB_2020',\n",
    "                    'RNAseq_Automatic_GEO_Signatures_Human_Down',\n",
    "                    'RNAseq_Automatic_GEO_Signatures_Human_Up',\n",
    "                    'RNAseq_Automatic_GEO_Signatures_Mouse_Down',\n",
    "                    'RNAseq_Automatic_GEO_Signatures_Mouse_Up',\n",
    "                    'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "                    'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "                    'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "                    'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "                    'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "                    'Reactome_2013',\n",
    "                    'Reactome_2015',\n",
    "                    'Reactome_2016',\n",
    "                    'SILAC_Phosphoproteomics',\n",
    "                    'SubCell_BarCode',\n",
    "                    'SysMyo_Muscle_Gene_Sets',\n",
    "                    'TF-LOF_Expression_from_GEO',\n",
    "                    'TF_Perturbations_Followed_by_Expression',\n",
    "                    'TG_GATES_2020',\n",
    "                    'TRANSFAC_and_JASPAR_PWMs',\n",
    "                    'TRRUST_Transcription_Factors_2019',\n",
    "                    'Table_Mining_of_CRISPR_Studies',\n",
    "                    'TargetScan_microRNA',\n",
    "                    'TargetScan_microRNA_2017',\n",
    "                    'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
    "                    'Tissue_Protein_Expression_from_ProteomicsDB',\n",
    "                    'Transcription_Factor_PPIs',\n",
    "                    'UK_Biobank_GWAS_v1',\n",
    "                    'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "                    'VirusMINT',\n",
    "                    'Virus_Perturbations_from_GEO_down',\n",
    "                    'Virus_Perturbations_from_GEO_up',\n",
    "                    'WikiPathways_2013',\n",
    "                    'WikiPathways_2015',\n",
    "                    'WikiPathways_2016',\n",
    "                    'WikiPathways_2019_Human',\n",
    "                    'WikiPathways_2019_Mouse'\n",
    "                ]\n",
    "            )\n",
    "        ],\n",
    "    },\n",
    "    section = 'geneEntrySection',\n",
    ") %}\n",
    "\n",
    "\n",
    "# Choose the orientation of the graph: horizontal or vertical bars\n",
    "orient_bar = \"{{ ChoiceField(name = 'orient_bar', label = 'Orientation', choices = ['Horizontal', 'Vertical'], default = 'Horizontal', description = 'Choose whether your bar graphs will be displayed horizontally or vertically', section = 'barGraphSection') }}\"\n",
    "\n",
    "# Choose color of bars\n",
    "color_bar = \"{{ ChoiceField(name = 'color_bar', label = 'Bar Color', choices = ['Black', 'Blue', 'Red', 'Green', 'Grey', 'Orange', 'Purple', 'Yellow', 'Pink'], default = 'Black', section = 'barGraphSection') }}\"\n",
    "\n",
    "# Choose whether gene counts are displayed on bar graph\n",
    "counts_bar = {{ BoolField(name = 'counts_bar', label = 'Show Counts?', default = 'true', description = 'Choose \\'Yes\\' to label the bars with their lengths.', section = 'barGraphSection') }}\n",
    "\n",
    "# Choose number of genes in bar graph\n",
    "num_bar_genes = {{ IntField(\n",
    "    name='num_bar_genes', \n",
    "    label='Top Genes', \n",
    "    min=2, \n",
    "    max=20, \n",
    "    default=20, \n",
    "    description='The number of genes that will be included in figures describing top genes (ex: most frequent, most published)', \n",
    "    section='barGraphSection'\n",
    ")}}\n",
    "\n",
    "# Choose number of genes in bar graph\n",
    "jac_cutoff = {{ FloatField(\n",
    "    name='jac_cutoff', \n",
    "    label='Jaccard Index High Threshold', \n",
    "    min=0.10, \n",
    "    max=0.99, \n",
    "    default=0.40, \n",
    "    description='The Jaccard Index will be calculated to measure similarity between sets in your library (0 = no shared genes, 1 = identical). Choose a threshold for what is considered a high Jaccard Index.', \n",
    "    section='simSection'\n",
    ")}}\n",
    "         \n",
    "# Choose which visualizations are generated for most similar sets\n",
    "jac_interactive = {{ BoolField(name = 'jac_interactive', label = 'Interactive Heatmap of Most Similar Sets?', default = 'true', description = 'Choose \\'Yes\\' to generate an interactive heatmap of the gene sets with similarity greater than the threshold you set above.', section = 'simSection') }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Color for Bar plot\n",
    "color_conversion = {\n",
    "    'Black': 'black',\n",
    "    'Blue': 'lightskyblue',\n",
    "    'Red': 'tomato',\n",
    "    'Green': 'mediumspringgreen',\n",
    "    'Grey': 'lightgrey',\n",
    "    'Orange': 'orange',\n",
    "    'Purple': 'plum',\n",
    "    'Yellow': 'yellow',\n",
    "    'Pink': 'lightpink'\n",
    "}\n",
    "\n",
    "bar_color = color_conversion[color_bar]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%- if library_kind.raw_value == 'Upload a library' %}\n",
    "library_kind = \"Upload a library\"\n",
    "library_filename = {{ library_kind.value[0] }}\n",
    "library_name = library_filename.replace(\"_\", \" \").replace(\".txt\", \"\").replace(\".gmt\", \"\")\n",
    "species = {{library_kind.value[1]}}\n",
    "\n",
    "{%- else %}\n",
    "library_kind = \"Select a library from Enrichr\"\n",
    "library_filename = \"{{ library_kind.value[0] }}\"\n",
    "library_name = \"{{ library_kind.value[0] }}\"\n",
    "species = ['human']\n",
    "if 'MOUSE' in library_name.upper().split(\"_\"):\n",
    "    species = ['mouse']\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download library from the Enrichr site using its file name\n",
    "def download_library(library_filename):\n",
    "    with open(f\"{library_filename}\", \"w\") as fw:\n",
    "        with urllib.request.urlopen(f'https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName={library_filename}') as f:\n",
    "            for line in f.readlines():\n",
    "                fw.write(line.decode('utf-8'))\n",
    "                fw.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load library \n",
    "def remove_comma(gene):\n",
    "    try:\n",
    "        comma = gene.index(',')\n",
    "        return gene[0:comma]\n",
    "    except:\n",
    "        return gene\n",
    "\n",
    "def load(library_filename, hasNA):\n",
    "    if library_kind == \"Select a library from Enrichr\":\n",
    "        download_library(library_filename)\n",
    "    library_data, library_genes, hasNA = load_library(library_filename, hasNA)\n",
    "    # to upper case\n",
    "    return library_data, library_genes, hasNA\n",
    "\n",
    "# Returns a dictionary (library_data) where the values are all the elements\n",
    "def load_library(library_filename, hasNA):\n",
    "    library_data = dict()\n",
    "    with open(library_filename, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "        library_genes = [''] * len(lines)\n",
    "        i = 0\n",
    "        for line in lines:\n",
    "            splited = line.strip().split(\"\\t\")\n",
    "            elements = pd.Series(splited[2:]).dropna()\n",
    "            if len(elements) > 0:\n",
    "                # to upper case\n",
    "                allxs = [x.upper() for x in elements]\n",
    "                allxs = pd.Series(allxs).apply(lambda x: remove_comma(x)).to_list()\n",
    "                \n",
    "                if 'NA' in allxs:\n",
    "                    allxs.remove('NA')\n",
    "                    hasNA = True\n",
    "                library_data[splited[0]] = allxs\n",
    "                library_genes[i] = (' ').join(allxs)\n",
    "            i = i + 1    \n",
    "    return library_data, library_genes, hasNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Method for gene novelty\n",
    "def gene_novelty_label(pub_count):\n",
    "    if pub_count <= 7:\n",
    "        return 'highly understudied'\n",
    "    if pub_count <= 25:\n",
    "        return 'understudied'\n",
    "    if pub_count <= 87:\n",
    "        return 'studied'\n",
    "    return 'well studied'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create geneRIF dictionary and novelty mapping dictionaries\n",
    "generif_df = pd.read_csv(\"https://appyters.maayanlab.cloud/storage/Gene_Set_Library_Synopsis/generif.tsv\", delimiter=\"\\t\", header=None)\n",
    "generif_df = generif_df.rename(columns={0:'Species',1:'Number',2:'Gene',3:'PMID',4:'Date'})\n",
    "generif_genes = generif_df['Gene']\n",
    "generif_s_genes = generif_genes.squeeze().str.upper()\n",
    "generif_counts = generif_s_genes.value_counts()\n",
    "generif_dict = generif_counts.to_dict()\n",
    "\n",
    "novel_map_dict = {\"highly understudied\": 3, \"understudied\": 2, \"studied\": 1, \"well studied\": 0}\n",
    "novel_map_dict_rev = {3: \"highly understudied\", 2: \"understudied\", 1: \"studied\", 0: \"well studied\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Load library, create genes list, set list, gene size list\n",
    "hasNA = False\n",
    "library_data, library_genes, hasNA = load(library_filename, hasNA)\n",
    "if library_kind == \"Select a library from Enrichr\":\n",
    "    library_name = library_name.replace(\"_\", \" \")\n",
    "vals = list(library_data.values())\n",
    "keys = list(library_data.keys())\n",
    "\n",
    "all_genes = list(chain(*vals))\n",
    "all_genes_unique = list(np.unique(np.array(all_genes)))\n",
    "all_genes_unique = np.array(all_genes_unique)\n",
    "all_sets = list(library_data.keys())\n",
    "\n",
    "gs_sizes = [0]*len(vals)\n",
    "for i in range(0, len(vals)): \n",
    "    gs_sizes[i] = len(vals[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make dataframes of gene sets and their genes in 1) list form, 2) string form\n",
    "library_data_onemap = dict()\n",
    "library_data_onemap_str = dict()\n",
    "for i in range(len(vals)):\n",
    "    library_data_onemap[keys[i]] = [vals[i]]\n",
    "    library_data_onemap_str[keys[i]] = (\" \").join(vals[i])\n",
    "\n",
    "library_data_onemap = pd.DataFrame(data=library_data_onemap).transpose()\n",
    "library_data_onemap = library_data_onemap.rename(columns= {0:'Genes'})\n",
    "library_data_onemap_str = pd.DataFrame(data={0:library_data_onemap_str})\n",
    "library_data_onemap_str = library_data_onemap_str.rename(columns= {0:'Genes'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# 1. Unmapped Gene Names\n",
    "This Appyter checks whether your gene set library contains unmapped gene names in _-DEC, _-MAR, and _-SEP formats. These conversions frequently occur when gene names are loaded into Excel. For example, either MARC1 or MARCH1 will automatically become '1-MAR'. Read this article for more information: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1044-7. This section also checks for genes labeled NA, which means Not Available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def month_sorter(month):\n",
    "    return month[-3]\n",
    "\n",
    "def date_sorter(month):\n",
    "    dash = month.index('-')\n",
    "    return int(month[:dash])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for unmapped genes and display them \n",
    "month_genes = all_genes_unique.copy()\n",
    "month_genes.sort()\n",
    "first = -1\n",
    "last = -1\n",
    "for i in range(len(month_genes)):\n",
    "    if len(month_genes[i]) > 4:\n",
    "        substr = month_genes[i][-4:]\n",
    "        if(substr == '-DEC' or substr == '-MAR' or substr == '-SEP'):\n",
    "            if first == -1:\n",
    "                first = i\n",
    "            last = i+1\n",
    "        else:\n",
    "            if first != -1:\n",
    "                break\n",
    "\n",
    "month_genes = month_genes[first:last]\n",
    "month_genes = sorted(month_genes,key=lambda x: (month_sorter(x), date_sorter(x)))\n",
    "if hasNA:\n",
    "    month_genes.append('NA')\n",
    "month_genes = pd.DataFrame(data=month_genes, columns=['Gene Name'])\n",
    "\n",
    "# Display if unmapped genes \n",
    "if len(month_genes) > 0:\n",
    "    month_genes_filename = 'unmapped_gene_names_' + library_name\n",
    "    found_genes_text = '' + str(len(month_genes)) + ' unmapped gene names found.'\n",
    "    display(Markdown(found_genes_text))\n",
    "    display(HTML(month_genes.to_html(index=False)))\n",
    "    figure_legend(f\"Table {table_count}\", content=f\"Unmapped gene names in {library_name}\")\n",
    "    display(create_download_link(month_genes, \"Download this table as a CSV\", month_genes_filename))\n",
    "    table_count = table_count + 1\n",
    "else:\n",
    "    print(\"No unmapped gene names found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# 2. Descriptive Statistics\n",
    "The Appyter will present descriptive statistics for your library such as: total genes, total gene sets, average genes per set, and frequency of each gene. Results will be displayed in downloadable tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the number of each gene \n",
    "count_frame = pd.Series(all_genes).value_counts().sort_index().reset_index().reset_index(drop=True)\n",
    "count_frame.columns = ['Gene', 'Count']\n",
    "count_frame.dropna()\n",
    "count_frame.sort_values(by=['Count'], inplace=True, ascending=False)\n",
    "count_frame = count_frame.reset_index(drop=True)\n",
    "# Drop skipped rows \n",
    "mask = count_frame['Gene'].str.len() > 0\n",
    "count_frame = count_frame.loc[mask]\n",
    "count_frame = count_frame[~count_frame['Gene'].isin(['NA'])]\n",
    "\n",
    "count_frame['Publications'] = count_frame['Gene'].map(generif_dict).replace(np.nan, 0)\n",
    "count_frame['Publications'] = count_frame['Publications'].astype(int)\n",
    "count_frame['Novelty'] = count_frame['Publications'].apply(lambda x: gene_novelty_label(x))\n",
    "pubhist_dat = list(count_frame['Publications'].replace(0,np.nan).dropna())\n",
    "\n",
    "# Make a copy to sort by publications\n",
    "count_frame2 = count_frame.copy(deep=True)\n",
    "count_frame2.sort_values(by=['Publications'], inplace=True, ascending=False)\n",
    "top_genes = count_frame.iloc[0:num_bar_genes]\n",
    "top_pub_genes = count_frame2.iloc[0:num_bar_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate novelty statistic for the library as a whole\n",
    "count_frame2['Novelty Num'] = count_frame2['Novelty'].map(novel_map_dict)\n",
    "novelty_weighted_sum = count_frame2['Count'] * count_frame2['Novelty Num']\n",
    "novelty_weighted_sum = novelty_weighted_sum.sum()\n",
    "count_sum = count_frame2['Count'].sum()\n",
    "\n",
    "library_nov_exact = round(novelty_weighted_sum/count_sum, 1)\n",
    "library_nov = math.floor(round(novelty_weighted_sum/count_sum, 1))\n",
    "library_nov_term = novel_map_dict_rev[library_nov]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make table describing gene sets\n",
    "geneset_df = library_data_onemap.copy(deep=True)\n",
    "geneset_df = geneset_df.reset_index()\n",
    "geneset_df = geneset_df.rename(columns={'index':'Gene Set'})\n",
    "geneset_df['Size'] = [0] * len(all_sets)\n",
    "geneset_df['Mean Publications/Gene'] = [0] * len(all_sets)\n",
    "geneset_df['Novelty'] = [''] * len(all_sets)\n",
    "\n",
    "for i in range(0,len(geneset_df['Gene Set'])):\n",
    "    genes = geneset_df.iloc[i,1]\n",
    "    temp = count_frame2[count_frame2['Gene'].isin(genes)]\n",
    "    tot_pubs = sum(temp['Publications'])\n",
    "    av_pubs = np.mean(temp['Publications'])\n",
    "    tot_genes = len(genes)\n",
    "    novelty_num = sum(temp['Novelty Num']/tot_genes)\n",
    "    novelty = library_nov_term = novel_map_dict_rev[math.floor(novelty_num)]\n",
    "    \n",
    "    geneset_df['Size'][i] = tot_genes\n",
    "    geneset_df['Mean Publications/Gene'][i] = av_pubs\n",
    "    geneset_df['Novelty'][i] = novelty\n",
    "\n",
    "geneset_df = geneset_df.sort_values(by='Size', ascending=False).drop(columns=['Genes']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Descriptive statistics summary table\n",
    "# Totals\n",
    "unique_genes = len(count_frame['Gene'])\n",
    "unique_sets = len(library_data.keys())\n",
    "avg_genes = round(np.mean(gs_sizes), 2)\n",
    "median_genes = round(np.median(gs_sizes), 2)\n",
    "\n",
    "median_publications = round(np.median(count_frame['Publications']), 2)\n",
    "avg_publications = round(np.mean(count_frame['Publications']), 2)\n",
    "\n",
    "# Novelty counts and percentages\n",
    "novelty_counts = count_frame['Novelty'].value_counts()\n",
    "novelty_counts_gs = geneset_df['Novelty'].value_counts()\n",
    "\n",
    "tot_pub = np.sum(count_frame['Publications'])\n",
    "genes_nov_dict = {'highly understudied': 0, 'understudied': 0, 'studied': 0, 'well studied':0}\n",
    "gs_nov_dict = {'highly understudied': 0, 'understudied': 0, 'studied': 0, 'well studied':0}\n",
    "genes_nov_pcnt_dict = {'highly understudied': '', 'understudied': '', 'studied': '', 'well studied':''}\n",
    "gs_nov_pcnt_dict = {'highly understudied': '', 'understudied': '', 'studied': '', 'well studied':''}\n",
    "\n",
    "# Reassign counts of highly understudied, understudied, etc. only if at least one gene of that type exists\n",
    "# Calculate percentages of each novelty \n",
    "for key in genes_nov_dict.keys():\n",
    "    if key in novelty_counts:\n",
    "        genes_nov_dict[key] = novelty_counts[key]\n",
    "    if key in novelty_counts_gs:\n",
    "        gs_nov_dict[key] = novelty_counts_gs[key]\n",
    "    \n",
    "    genes_nov_pcnt_dict[key] = str(round(genes_nov_dict[key]/unique_genes * 100, 2)) + \"%\"\n",
    "    gs_nov_pcnt_dict[key] = str(round(gs_nov_dict[key]/unique_sets * 100, 2)) + \"%\"\n",
    "\n",
    "## Load stats for all human and mouse genes\n",
    "all_human_genes = pd.read_csv(\"https://appyters.maayanlab.cloud/storage/Gene_Set_Library_Synopsis/all_human_genes_df.csv\", header=0)\n",
    "all_mouse_genes = pd.read_csv(\"https://appyters.maayanlab.cloud/storage/Gene_Set_Library_Synopsis/all_mouse_genes_df.csv\", header=0)\n",
    "\n",
    "# Change table display based on library composition\n",
    "other = False\n",
    "specname = 'Human'\n",
    "all_species_genes=all_human_genes\n",
    "if 'human' not in species:\n",
    "    if 'mouse' in species:\n",
    "        all_species_genes = all_mouse_genes\n",
    "        specname = 'Mouse'\n",
    "    if species==['other']:\n",
    "        other = True\n",
    "spec_col_name = 'All ' + specname + ' Genes'\n",
    "    \n",
    "## Make and display tables\n",
    "lib_vs_spec_title_col = [\"Total Genes\", \"Total Publications\", \"Highly Understudied Genes\", \"Understudied Genes\", \"Studied Genes\", \"Well Studied Genes\"]\n",
    "lib_col = [unique_genes, tot_pub, genes_nov_dict['highly understudied'], genes_nov_dict['understudied'], genes_nov_dict['studied'], genes_nov_dict['well studied']]\n",
    "lib_col_pcnt = ['', '', genes_nov_pcnt_dict['highly understudied'], genes_nov_pcnt_dict['understudied'], genes_nov_pcnt_dict['studied'], genes_nov_pcnt_dict['well studied']]\n",
    "\n",
    "lib_vs_spec_df = pd.DataFrame(data = {'': lib_vs_spec_title_col, f\"{library_name}\": lib_col, ' ': lib_col_pcnt})\n",
    "lib_col = [avg_publications, median_publications]\n",
    "lib_vs_spec_mm_df = pd.DataFrame(data = {'': [\"Mean Publications/Gene\", \"Median Publications/Gene\"], f\"{library_name}\": lib_col})\n",
    "                                          \n",
    "if not other:\n",
    "    spec_col = [all_species_genes['Genes'][0], all_species_genes['Publications'][0], all_species_genes['Highly Understudied'][0], all_species_genes['Understudied'][0], all_species_genes['Studied'][0], all_species_genes['Well Studied'][0]]\n",
    "    spec_col_pcnt = ['', '', str(all_species_genes['Highly Understudied'][1]) + \"%\", str(all_species_genes['Understudied'][1]) + \"%\", str(all_species_genes['Studied'][1]) + \"%\", str(all_species_genes['Well Studied'][1]) + \"%\"]\n",
    "    lib_col = [unique_genes, tot_pub, genes_nov_dict['highly understudied'], genes_nov_dict['understudied'], genes_nov_dict['studied'], genes_nov_dict['well studied']]\n",
    "    lib_col_pcnt = ['', '', genes_nov_pcnt_dict['highly understudied'], genes_nov_pcnt_dict['understudied'], genes_nov_pcnt_dict['studied'], genes_nov_pcnt_dict['well studied']]\n",
    "    lib_vs_spec_df = pd.DataFrame(data = {'': lib_vs_spec_title_col, f\"{library_name}\": lib_col, ' ': lib_col_pcnt, spec_col_name: spec_col, '  ': spec_col_pcnt})\n",
    "    lib_vs_spec_df[spec_col_name] = lib_vs_spec_df[spec_col_name].astype(int)\n",
    "\n",
    "    lib_col = [avg_publications, median_publications]\n",
    "    spec_col = [all_species_genes['Mean Publications'][0], all_species_genes['Median Publications'][0]]\n",
    "    lib_vs_spec_mm_df = pd.DataFrame(data = {'': [\"Mean Publications/Gene\", \"Median Publications/Gene\"], f\"{library_name}\": lib_col, spec_col_name: spec_col})\n",
    "\n",
    "genestat_title_col = [\"Gene Sets\", \"Highly Understudied Sets\", \"Understudied Sets\", \"Studied Sets\", \"Well Studied Sets\"]\n",
    "genestat_col = [unique_sets,  gs_nov_dict['highly understudied'], gs_nov_dict['understudied'], gs_nov_dict['studied'], gs_nov_dict['well studied']]\n",
    "genestat_col_pcnt = ['', gs_nov_pcnt_dict['highly understudied'], gs_nov_pcnt_dict['understudied'], gs_nov_pcnt_dict['studied'], gs_nov_pcnt_dict['well studied']]\n",
    "genestat_df = pd.DataFrame(data = {'': genestat_title_col, 'Total': genestat_col, ' ':genestat_col_pcnt})\n",
    "\n",
    "genestat_title_col = [\"Mean\", \"Median\"]\n",
    "genestat_col = [avg_genes, median_genes]\n",
    "genestat_mm_df = pd.DataFrame(data = {'': genestat_title_col, 'Genes / Set': genestat_col})\n",
    "\n",
    "display(HTML(lib_vs_spec_df.to_html(index=False)))\n",
    "display(HTML(lib_vs_spec_mm_df.to_html(index=False)))\n",
    "if other:\n",
    "    figure_legend(f\"Tables {table_count}A, {table_count}B\", title=f\"{library_name} Summary Statistics\", content=\"Descriptive statistics for your library. Novelty ratings are based on Geneshot. Highly understudied genes are associated with 0-7 PubMed IDs (PMIDs); understudied genes with 8-25 PMIDs; studied genes with 26-87 PMIDs; and well studied genes with 88+ PMIDs.\")\n",
    "else:\n",
    "    figure_legend(f\"Tables {table_count}A, {table_count}B\", title=f\"{library_name} vs. All {specname} Genes\", content=f\"Descriptive statistics comparing your gene set library to the set of all {specname} genes in GeneRIF. Novelty ratings are based on Geneshot. Highly understudied genes are associated with 0-7 PubMed IDs (PMIDs); understudied genes with 8-25 PMIDs; studied genes with 26-87 PMIDs; and well studied genes with 88+ PMIDs.\")\n",
    "\n",
    "table_count = table_count + 1\n",
    "display(HTML(genestat_df.to_html(index=False)))\n",
    "display(HTML(genestat_mm_df.to_html(index=False)))\n",
    "figure_legend(f\"Tables {table_count}A, {table_count}B\", title=f\"Summary Statistics for Gene Sets in {library_name} Library\", content=\"Statistics describing the composition of gene sets within your library. Novelty ratings were calculated by giving each gene in each set a numerical novelty score (based on its rating of highly understudied, understudied, studied, or well studied), taking a weighted average of those scores, and translating the result back into a term that describes the entire gene set.\")\n",
    "table_count = table_count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_bok_hist(title, hist, edges, dat, xaxis_lab, yaxis_lab, tooltips, fill_color,xtype='auto', ytype='auto', xtail=5, ytail=5):\n",
    "    yrange = 0\n",
    "    if ytype=='log':\n",
    "        hist, edges = np.histogram(dat, density=False, bins=10)\n",
    "        ycap = math.ceil(math.log10(max(pubhist_dat)))\n",
    "        yrange=(10**0, 10**ycap)\n",
    "        \n",
    "    ordered_hist = sorted(hist)\n",
    "    ordered_edges = sorted(edges)\n",
    "    \n",
    "    if ytype=='auto':\n",
    "        yrange=[ordered_hist[0], ordered_hist[-1]+ytail]\n",
    "    \n",
    "    p = figure(title=title, tooltips=tooltips, background_fill_color=\"#fafafa\", toolbar_location=\"below\",x_axis_type = xtype,y_axis_type = ytype,\n",
    "              x_range=[ordered_edges[0], ordered_edges[-1]+xtail], y_range=yrange)\n",
    "    p.quad(top=hist, bottom=0, left=edges[:-1], right=edges[1:],\n",
    "           fill_color=fill_color, line_color=\"white\", hover_alpha=0.7)\n",
    "    \n",
    "    p.y_range.start = 0\n",
    "    p.xaxis.axis_label = xaxis_lab\n",
    "    p.yaxis.axis_label = yaxis_lab\n",
    "    p.grid.grid_line_color=\"white\"\n",
    "    p.title.align = 'center'\n",
    "    return p\n",
    "\n",
    "def get_bin_params(dat_max, dat_min = 0, bins=40):\n",
    "    rng = dat_max - dat_min\n",
    "    needed = bins - rng%bins\n",
    "    new_rng = rng + needed\n",
    "    maxval = dat_min + new_rng\n",
    "    \n",
    "    if needed==bins:\n",
    "        maxval = dat_max\n",
    "        new_rng = rng\n",
    "    \n",
    "    interval = new_rng/bins\n",
    "    return maxval, interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Print out a table of the count_frame dataframe \n",
    "counts_filename = library_name.replace(\" \", \"_\") + \"_gene_counts.csv\"\n",
    "\n",
    "display(HTML(count_frame[0:num_bar_genes].to_html(index=False)))\n",
    "figure_legend(f\"Table {table_count}\", title=f\"Gene count results for {library_name}\", content=\"This table displays the counts (number of appearances throughout the entire library), number of publication associations (PMIDs), and novelty ratings of each gene. The full chart is also available for download. A weighted average statistic has been used to assign a novelty rating to your library as a whole, representing how well-studied the library is.\")\n",
    "table_count = table_count + 1\n",
    "display(create_download_link(count_frame, \"Download this table as a CSV\", counts_filename))\n",
    "print(f\"Based on the novelty and within-library frequencies of each gene, your gene set library is \\033[1m{library_nov_term}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display table of gene set statistics\n",
    "geneset_df_filename = library_name.replace(\" \", \"_\") + \"_gene_set_statistics.csv\"\n",
    "display(HTML(geneset_df.iloc[0:10,:].to_html(index=False)))\n",
    "figure_legend(f\"Table {table_count}\", f\"Gene Set Statistics in {library_name} Library\", \"Size (number of genes), average publications per gene, and novelty (calculated using a weighted average across all genes in the set) for each gene set in your library.\")\n",
    "display(create_download_link(geneset_df, \"Download this table as a CSV\", geneset_df_filename))\n",
    "table_count = table_count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# 3. Scatterplot Visualization\n",
    "In this section, the gene sets in your library will be converted into numerical vectors using TF-IDF, transformed into two dimensions using UMAP, and visualized as a scatterplot. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = library_data_onemap_str.reset_index().rename(columns={'index':'Name'})\n",
    "\n",
    "gene_list = df['Genes']\n",
    "\n",
    "try:\n",
    "    tfidf_vectorizer = TfidfVectorizer(\n",
    "        min_df = 3,\n",
    "        max_df = 0.005,\n",
    "        max_features = 100000,\n",
    "        ngram_range=(1, 1)\n",
    "    )\n",
    "    tfidf = tfidf_vectorizer.fit_transform(gene_list)\n",
    "except:\n",
    "    factor = 0.005\n",
    "    while factor*unique_sets < 3:\n",
    "        factor = factor + .005\n",
    "    \n",
    "    tfidf_vectorizer = TfidfVectorizer(\n",
    "        min_df = 3,\n",
    "        max_df = factor*unique_sets,\n",
    "        max_features = 100000,\n",
    "        ngram_range=(1, 1)\n",
    "    )\n",
    "    tfidf = tfidf_vectorizer.fit_transform(gene_list)\n",
    "\n",
    "reducer = umap.UMAP()\n",
    "reducer.fit(tfidf)\n",
    "embedding = pd.DataFrame(reducer.transform(tfidf), columns=['x','y'])\n",
    "embedding = pd.concat([embedding, df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare dimensionality-reduced matrix for clutstering\n",
    "mapped_df = embedding.copy(deep=True)\n",
    "mapped_df = mapped_df.set_index('Name')\n",
    "mapped_df = mapped_df.drop(columns=['Genes'])\n",
    "mapped_df = mapped_df.rename_axis(\"Gene Set\").reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot clustered gene sets\n",
    "xlabel = 'UMAP Dimension 1'\n",
    "ylabel = 'UMAP Dimension 2'\n",
    "\n",
    "source2 = ColumnDataSource(\n",
    "        data=dict(\n",
    "            x = mapped_df.x,\n",
    "            y = mapped_df.y,\n",
    "            alpha = [0.7] * mapped_df.shape[0],\n",
    "            size = [7] * mapped_df.shape[0],\n",
    "            gene_set = mapped_df['Gene Set'],\n",
    "        )\n",
    "    )\n",
    "\n",
    "hover_emb = HoverTool(names=[\"df\"], tooltips=\"\"\"\n",
    "    <div style=\"margin: 10\">\n",
    "        <div style=\"margin: 0 auto; width:300px;\">\n",
    "            <span style=\"font-size: 12px; font-weight: bold;\">Gene Set:</span>\n",
    "            <span style=\"font-size: 12px\">@gene_set</span>\n",
    "        <div style=\"margin: 0 auto; width:300px;\">\n",
    "            <span style=\"font-size: 12px; font-weight: bold;\">Coordinates:</span>\n",
    "            <span style=\"font-size: 12px\">(@x,@y)</span>\n",
    "        </div>\n",
    "    </div>\n",
    "    \"\"\")\n",
    "tools_emb = [hover_emb, 'pan', 'wheel_zoom', 'reset', 'save']\n",
    "title_emb = 'Gene Sets in ' + library_name + ' Library'\n",
    "plot_emb = figure(plot_width=1000, plot_height=700, tools=tools_emb, title=title_emb, x_axis_label=xlabel, y_axis_label=ylabel)\n",
    "plot_emb.circle('x', 'y', size='size', \n",
    "                alpha='alpha', line_alpha=0, line_width=0.01, source=source2, name=\"df\", \n",
    "                fill_color='grey')\n",
    "plot_emb.xaxis.axis_label_text_font_style = 'normal'\n",
    "plot_emb.xaxis.axis_label_text_font_size = '18px'\n",
    "plot_emb.yaxis.axis_label_text_font_size = '18px'\n",
    "plot_emb.yaxis.axis_label_text_font_style = 'normal'\n",
    "plot_emb.title.align = 'center'\n",
    "plot_emb.title.text_font_size = '18px'\n",
    "\n",
    "show(plot_emb)\n",
    "figure_legend(f\"Fig. {fig_count}\", f\"Scatterplot of Gene Sets in {library_name} Library\", \"Gene sets plotted by their UMAP dimensions.\")\n",
    "fig_count = fig_count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# 4. Set Similarity\n",
    "In this section, the Appyter will compute the pairwise Jaccard index for every pair of gene sets in your library as a measure of set similarity. An index closer to 1 means the gene sets have most genes in common. An index closer to 0 indicates the gene sets share very few terms. The Jaccard indices will serve as the basis for a heatmap. Additional visualizations will be generated for the most similar sets (those with Jaccard indices above the user-specified threshold)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Put all sets and genes into dataframe where each thing is a list\n",
    "def jaccard_list(u,v):\n",
    "    setu = set(chain(*u))\n",
    "    setv = set(chain(*v))\n",
    "    return len(setu.intersection(setv)) / len(setu.union(setv))\n",
    "\n",
    "res = pdist(library_data_onemap[['Genes']], jaccard_list)\n",
    "distance = pd.DataFrame(squareform(res), index=library_data_onemap.index, columns= library_data_onemap.index)\n",
    "\n",
    "# Check whether any sets have Jaccard Index > 0; if not, skip all Jaccard plots\n",
    "jac_zero_tester = pd.Series(res)\n",
    "jac_zero_tester = jac_zero_tester.replace(0, np.nan)\n",
    "jac_zero_tester = jac_zero_tester.dropna()\n",
    "jac_zero = len(jac_zero_tester)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter \"distance\" to put NA where col# <= row#\n",
    "mask = np.zeros_like(distance, dtype=bool)\n",
    "mask[np.triu_indices_from(mask)] = True\n",
    "masked_dist = distance.mask(mask, '').transpose()\n",
    "masked_dist_filename = library_name.replace(\" \", \"_\") + \"_jaccard_matrix.csv\"\n",
    "\n",
    "display(HTML(masked_dist.iloc[0:10,0:10].to_html(index=True)))\n",
    "figure_legend(f\"Table {table_count}\", f\"Jaccard Index pairings for {library_name}\", \"Upper triangle of a pairwise Jaccard matrix comparing each set in the library to each other set.\")\n",
    "display(create_download_link(masked_dist.reset_index(), \"Download this table as a CSV\", masked_dist_filename))\n",
    "table_count = table_count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if jac_zero > 0:\n",
    "    maxval, interval = get_bin_params(max(res)*1000)\n",
    "    maxval = maxval/1000\n",
    "    interval = interval/1000\n",
    "    jac_hist_bin_arr = np.arange(0, maxval, interval)\n",
    "\n",
    "    #jac_hist_bin_arr = np.arange(0, max(res)+.015, .01)\n",
    "    hist, edges = np.histogram(res, density=False, bins=jac_hist_bin_arr)\n",
    "    title = f\"Jaccard Indices for {library_name}\"\n",
    "\n",
    "    tooltips = [\n",
    "        (\"range\", \"@left{0.00}\" + \"-\" + \"@right{0.00}\"),\n",
    "        (\"pairs\", \"@top\")\n",
    "        ]\n",
    "\n",
    "    xaxis_lab = 'Jaccard Index'\n",
    "    yaxis_lab = 'Gene Set Pairs'\n",
    "\n",
    "    jac_hist = make_bok_hist(title, hist, edges, res, xaxis_lab, yaxis_lab, tooltips, bar_color, xtail=max(edges)/15, ytail=max(hist)/15)\n",
    "else:\n",
    "    print(\"All pairwise Jaccard Indices in your library are equal to 0. This means that no gene set in your library shares a single term (gene) with another gene set. The remaining tables and visualizations in this section of the analysis will not be generated.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_bok_heatmap(df, title):\n",
    "    colors = ['#ffffb2', '#fed976', '#feb24c', '#fd8d3c', '#f03b20', '#bd0026']\n",
    "    mapper = LinearColorMapper(palette=colors, low=df.jaccard.min(), high=round(df.jaccard.max(), 1), nan_color=\"white\")\n",
    "\n",
    "    rng = pd.unique(df['set1'])\n",
    "    pwidth = 1000* int(math.ceil((len(rng)/(jac_cutoff*100))))\n",
    "    pheight = 700* int(math.ceil((len(rng)/(jac_cutoff*100))))\n",
    "    if pwidth == 0:\n",
    "        pwidth = 1000\n",
    "    if pheight == 0:\n",
    "        pheight = 500\n",
    "        \n",
    "    source = ColumnDataSource(df)\n",
    "    \n",
    "    p = figure(title=title,\n",
    "               x_range=rng, y_range=list(reversed(rng)),\n",
    "               x_axis_location=\"above\", plot_width=pwidth, plot_height=pheight,\n",
    "               toolbar_location='below',\n",
    "               tooltips=[\n",
    "                   ('Set 1', '@set1'),\n",
    "                   ('Set 2', '@set2'),\n",
    "                   ('Jaccard Index', '@jaccard')\n",
    "               ])\n",
    "\n",
    "    p.grid.grid_line_color = None\n",
    "    p.axis.axis_line_color = None\n",
    "    p.axis.major_tick_line_color = None\n",
    "    p.axis.major_label_text_font_size = \"7px\"\n",
    "    if len(df['set1']) < 15:\n",
    "        p.axis.major_label_text_font_size = \"10px\"\n",
    "    p.axis.major_label_standoff = 0\n",
    "    p.xaxis.major_label_orientation = np.pi/2\n",
    "    p.title.align = 'center'\n",
    "\n",
    "    p.rect(x = 'set2', y = 'set1', width=1, height=1,\n",
    "           source=source,\n",
    "           fill_color=transform('jaccard', mapper),\n",
    "           line_color=None)\n",
    "\n",
    "    color_bar = ColorBar(color_mapper=mapper, major_label_text_font_size=\"12px\",\n",
    "                         ticker=BasicTicker(desired_num_ticks=len(colors)),\n",
    "                         formatter=PrintfTickFormatter(format=\"%.2f\"),\n",
    "                         label_standoff=6, border_line_color=None)\n",
    "    p.add_layout(color_bar, 'right')\n",
    "    return(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if jac_zero > 0:\n",
    "    res_cut = res[res>jac_cutoff]\n",
    "    if len(res_cut) > 0:\n",
    "        datmin = math.floor(jac_cutoff * 100)*10\n",
    "        datmax = math.ceil(max(res_cut)*100)*10\n",
    "        maxval, interval = get_bin_params(datmax, dat_min=datmin)\n",
    "        datmin = datmin/1000\n",
    "        maxval = maxval/1000\n",
    "        interval = interval/1000\n",
    "        #jac_hist_cut_bin_arr = np.arange(math.floor(jac_cutoff * 100)/100.0, max(res_cut)+.01, .005)\n",
    "        jac_hist_cut_bin_arr = np.arange(datmin, maxval, interval)\n",
    "        hist_cut, edges_cut = np.histogram(res_cut, density=False, bins=jac_hist_cut_bin_arr)\n",
    "        title_cut = f\"Jaccard Indices > {jac_cutoff} for {library_name}\"\n",
    "        tooltips_cut = [\n",
    "            (\"range\", \"@left{0.000}\" + \"-\" + \"@right{0.000}\"),\n",
    "            (\"pairs\", \"@top\")\n",
    "            ]\n",
    "        jac_hist_cut = make_bok_hist(title_cut, hist_cut, edges_cut, res_cut, xaxis_lab, yaxis_lab, tooltips_cut, bar_color, xtail=max(edges_cut)/15, ytail=max(hist_cut)/15)\n",
    "    else:\n",
    "        print(f\"There are no gene set pairs with a Jaccard Index greater than {jac_cutoff} in your library. The corresponding tables and visualizations will not be generated.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If user wants interactive Jaccard heatmap for high indices, create one \n",
    "if jac_interactive and jac_zero > 0:\n",
    "    dist_cut_indexes = np.where(distance > jac_cutoff)\n",
    "    dist_cut_rows_cols = np.unique(np.array(list(chain(*list(dist_cut_indexes)))))\n",
    "    dist_cut = distance.iloc[dist_cut_rows_cols, dist_cut_rows_cols]\n",
    "\n",
    "    mask2 = np.zeros_like(dist_cut, dtype=bool)\n",
    "    mask2[np.triu_indices_from(mask2)] = True\n",
    "    dist_cut_masked = dist_cut.mask(mask2, np.nan).transpose()\n",
    "\n",
    "    dist_set1 = np.array(np.repeat(dist_cut_masked.index, len(dist_cut_masked.index)))\n",
    "    sep = \",,\"\n",
    "    dist_set2 = (sep.join(list(dist_cut_masked.index))+\",,\")*len(list(dist_cut_masked.index))\n",
    "    dist_set2 = (dist_set2.split(\",,\"))[0:-1]\n",
    "    dist_vals = list(chain(*list(dist_cut_masked.values)))\n",
    "\n",
    "    dist_heat_df = pd.DataFrame(data={'set1': dist_set1, 'set2': dist_set2, 'jaccard': dist_vals})\n",
    "\n",
    "    dist_heat_title = f\"Jaccard Indices for {library_name}\"\n",
    "\n",
    "    # Make smaller heatmap\n",
    "    jac_heat_cut = make_bok_heatmap(dist_heat_df, dist_heat_title)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show Jaccard histograms\n",
    "if jac_zero > 0:\n",
    "    if len(res_cut) > 0:\n",
    "        show(row(jac_hist, jac_hist_cut))\n",
    "        figure_legend(f\"Fig. {fig_count} and {fig_count +1}\",\"Jaccard Indices Histograms\", content=f\"The histogram on the left displays the full range of Jaccard Indices for your library. The histogram on the right displays only those indices greater than {jac_cutoff}.\")\n",
    "        fig_count = fig_count + 2\n",
    "    else:\n",
    "        show(jac_hist)\n",
    "        figure_legend(f\"Fig. {fig_count}\",\"Jaccard Indices Histogram\", content=f\"This histogram displays the full range of Jaccard Indices for your library. There were no indices greater than {jac_cutoff}.\")\n",
    "        fig_count = fig_count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Table of highest Jaccard Indices\n",
    "if jac_zero > 0:\n",
    "    dist_heat_df_disp = dist_heat_df[dist_heat_df['jaccard'] > jac_cutoff].reset_index(drop=True)\n",
    "    dist_heat_df_disp = dist_heat_df_disp.rename(columns={\"set1\": \"Gene Set 1\", \"set2\": \"Gene Set 2\", \"jaccard\": \"Jaccard Index\"})\n",
    "    \n",
    "    display(HTML(dist_heat_df_disp.sort_values(by=\"Jaccard Index\").iloc[0:10,:].to_html(index=False)))\n",
    "    figure_legend(f\"Table {table_count}\",f\"Jaccard Indices > {jac_cutoff}\", content=f\"This table displays all gene set pairings with a Jaccard Index greater than {jac_cutoff}.\")\n",
    "    high_jac_filename = library_name.replace(\" \", \"_\") + \"_high_jaccard_indices.csv\"\n",
    "    display(create_download_link(dist_heat_df_disp.sort_values(by=\"Jaccard Index\"), \"Download this table as a CSV\", high_jac_filename))\n",
    "    table_count = table_count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get row indices in order and offer in table\n",
    "def sns_heatmap_to_df(g):\n",
    "    rows = g.dendrogram_row.reordered_ind\n",
    "    cmap_df = distance.reset_index().rename(columns={'index': 'gene set'}).reset_index().rename(columns={'index': 'original index'})\n",
    "    cmap_df_dict = dict(zip(cmap_df['original index'], cmap_df['gene set']))\n",
    "    cmap_df = pd.DataFrame(data={'original index': rows})\n",
    "    cmap_df['gene set'] = cmap_df['original index'].map(cmap_df_dict)\n",
    "    cmap_df_ret = pd.DataFrame(data={'gene set': cmap_df['gene set'], \n",
    "                                     'original index':cmap_df['original index'],\n",
    "                                    'new index':cmap_df.index})\n",
    "    return cmap_df_ret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clustered heatmap of full library- only if selected, or if dendrogram is not possible \n",
    "sns_clust = None\n",
    "full_heat_possible = True\n",
    "\n",
    "if jac_zero > 0 and unique_sets > 0:\n",
    "    try: \n",
    "        sns_clust = sns.clustermap(distance, cmap=\"Reds\", figsize=(13,13))\n",
    "    \n",
    "    except:\n",
    "        print(\"Unable to generate heatmap. Try a smaller library.\")\n",
    "        full_heat_possible = False \n",
    "    \n",
    "    if full_heat_possible:\n",
    "        sns_clust.ax_row_dendrogram.set_visible(False)\n",
    "        sns_clust.ax_col_dendrogram.set_visible(False)\n",
    "        sns_clust.ax_cbar.set_position((0, 0, .03, .4))\n",
    "        figure_legend(f\"Fig. {fig_count}\", \"Heatmap\", f\"This heatmap displays the Jaccard Indices of all gene sets in your library.\")\n",
    "        fig_count = fig_count + 1\n",
    "\n",
    "        sns_clust_filename = library_name.replace(\" \", \"_\") + \"_jaccard_heatmap.png\"\n",
    "        plt.savefig(sns_clust_filename, bbox_inches = 'tight')\n",
    "        display(FileLink(sns_clust_filename, result_html_prefix = str('Download png' + ': ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if full_heat_possible:\n",
    "    cmap_df = sns_heatmap_to_df(sns_clust)\n",
    "    cmap_filename = library_name.replace(\" \", \"_\") + \"_jaccard_heatmap_reordered_gene_sets.csv\"\n",
    "    display(HTML(cmap_df.head().to_html(index=False)))\n",
    "    figure_legend(f\"Table {table_count}\", f\"Reordered gene sets in heatmap of {library_name}\", \"This table lists your gene sets in the order in which they appear in the heatmap. The full table is available for download.\")\n",
    "    display(create_download_link(cmap_df, \"Download this table as a CSV\", cmap_filename))\n",
    "    table_count = table_count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If interactive heatmap is possible, display it. Otherwise, create a static heatmap and report the new indices.\n",
    "jac_static_heat=True\n",
    "if jac_interactive and jac_zero > 0 and len(res_cut) > 0:\n",
    "    if len(dist_cut_rows_cols) < 300:\n",
    "        show(jac_heat_cut)\n",
    "        figure_legend(f\"Fig. {fig_count}\", \"High Jaccard Indices Heatmap\", f\"This heatmap includes all gene sets with at least one Jaccard Index greater than {jac_cutoff} in comparison with another set.\")\n",
    "        fig_count = fig_count + 1\n",
    "        jac_static_heat=False\n",
    "    else:\n",
    "        print(\"There are too many sets to generate an interactive figure. A static heatmap will be generated instead. To see an interactive heatmap of the highest Jaccard Indices, try selecting a higher threshold value.\")\n",
    "        \n",
    "if jac_static_heat:        \n",
    "    cmap_cut = sns.clustermap(dist_cut, cmap=\"Reds\")\n",
    "    cmap_cut.ax_row_dendrogram.set_visible(False)\n",
    "    cmap_cut.ax_col_dendrogram.set_visible(False)\n",
    "    cmap_cut.ax_cbar.set_position((0.8, 0, .03, .4))\n",
    "    figure_legend(f\"Fig. {fig_count}\", \"High Jaccard Indices Heatmap\", f\"This heatmap includes all gene sets with at least one Jaccard Index greater than {jac_cutoff} in comparison with another set.\")\n",
    "    fig_count = fig_count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# 5. Visualization of Novelty and Size Distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read table of all Enrichr library novelties and append the user's library\n",
    "all_enrichr_novs = pd.read_csv(\"https://appyters.maayanlab.cloud/storage/Gene_Set_Library_Synopsis/enrichr_library_novelties.csv\", header=0)\n",
    "all_enrichr_novs = all_enrichr_novs.sort_values(by='Stat', ascending=False)\n",
    "my_lib_nov = pd.DataFrame(data={'Library':[library_name], 'Novelty':[library_nov_term], 'Stat':[library_nov_exact], 'Genes': [len(all_genes_unique)]})\n",
    "mouse_libs = ['KEGG 2019 Mouse',\n",
    " 'Mouse Gene Atlas',\n",
    " 'RNAseq Automatic GEO Signatures Mouse Down',\n",
    " 'RNAseq Automatic GEO Signatures Mouse Up',\n",
    " 'WikiPathways 2019 Mouse']\n",
    "all_enrichr_novs_mouse = all_enrichr_novs[all_enrichr_novs['Library'].isin(mouse_libs)]\n",
    "all_enrichr_novs_mix = all_enrichr_novs[~all_enrichr_novs['Library'].isin(mouse_libs)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display gene set size and novelty distribution as a scatterplot\n",
    "novelties = ['well studied', 'studied', 'understudied', 'highly understudied']\n",
    "xlabel = 'Set Size (Genes)'\n",
    "ylabel = 'Mean Publications Per Gene'\n",
    "my_gs_stat_df = pd.DataFrame(data={'Gene Set':['Average of all gene sets'], 'Novelty':[library_nov_term], 'Size':[avg_genes], 'Publications/Gene': [int(round(np.mean(geneset_df['Mean Publications/Gene'],0)))]})\n",
    "geneset_df['Novelty Num'] = geneset_df['Novelty'].map(novel_map_dict)\n",
    "geneset_df = geneset_df.sort_values(by='Novelty Num', ascending=False).reset_index(drop=True).drop(columns=['Novelty Num'])\n",
    "\n",
    "source1 = ColumnDataSource(\n",
    "        data=dict(\n",
    "            x = geneset_df.Size,\n",
    "            y = geneset_df['Mean Publications/Gene'],\n",
    "            alpha = [0.9] * geneset_df.shape[0],\n",
    "            size = [7] * geneset_df.shape[0],\n",
    "            novelty = geneset_df.Novelty,\n",
    "            geneset = geneset_df['Gene Set'],\n",
    "        )\n",
    "    )\n",
    "\n",
    "source2 = ColumnDataSource(\n",
    "        data=dict(\n",
    "            x = my_gs_stat_df.Size,\n",
    "            y = my_gs_stat_df['Publications/Gene'],\n",
    "            alpha = [0.7] * my_gs_stat_df.shape[0],\n",
    "            size = [10] * my_gs_stat_df.shape[0],\n",
    "            novelty = my_gs_stat_df.Novelty,\n",
    "            geneset = my_gs_stat_df['Gene Set']\n",
    "        )\n",
    "    )\n",
    "#print(embedding.shape[0])\n",
    "\n",
    "hover_gs_nov = HoverTool(names=[\"df\"], tooltips=\"\"\"\n",
    "    <div style=\"margin: 10\">\n",
    "        <div style=\"margin: 0 auto; width:300px;\">\n",
    "            <span style=\"font-size: 12px; font-weight: bold;\">Gene Set:</span>\n",
    "            <span style=\"font-size: 12px\">@geneset</span>\n",
    "        <div style=\"margin: 0 auto; width:300px;\">\n",
    "            <span style=\"font-size: 12px; font-weight: bold;\">Novelty:</span>\n",
    "            <span style=\"font-size: 12px\">@novelty</span>\n",
    "        <div style=\"margin: 0 auto; width:300px;\">\n",
    "            <span style=\"font-size: 12px; font-weight: bold;\">Set Size (Genes):</span>\n",
    "            <span style=\"font-size: 12px\">@x</span>\n",
    "         <div style=\"margin: 0 auto; width:300px;\">\n",
    "            <span style=\"font-size: 12px; font-weight: bold;\">Mean Publications Per Gene:</span>\n",
    "            <span style=\"font-size: 12px\">@y</span>\n",
    "        </div>\n",
    "    </div>\n",
    "    \"\"\")\n",
    "tools_gs_nov = [hover_gs_nov, 'pan', 'wheel_zoom', 'reset', 'save']\n",
    "title_gs_nov = f\"Novelties and Sizes of Gene Sets Within {library_name} Library\"\n",
    "plot_gs_nov = figure(plot_width=700, plot_height=700, tools=tools_gs_nov, title=title_gs_nov, x_axis_label=xlabel, y_axis_label=ylabel, x_axis_type='log', y_axis_type='log')\n",
    "plot_gs_nov.circle('x', 'y', size='size', \n",
    "                alpha='alpha', line_alpha=0, line_width=0.01, source=source1, name=\"df\", \n",
    "                fill_color=factor_cmap('novelty', palette=Spectral6, factors=novelties),\n",
    "               legend_field='novelty')\n",
    "plot_gs_nov.circle('x', 'y', size='size', \n",
    "                alpha='alpha', line_alpha=0, line_width=0.01, source=source2, name=\"df\", \n",
    "                fill_color='red')\n",
    "plot_gs_nov.xaxis.axis_label_text_font_style = 'normal'\n",
    "plot_gs_nov.yaxis.axis_label_text_font_style = 'normal'\n",
    "plot_gs_nov.title.align = 'center'\n",
    "plot_gs_nov.legend.location = \"bottom_right\"\n",
    "plot_gs_nov.xaxis.axis_label_text_font_size = '18px'\n",
    "plot_gs_nov.yaxis.axis_label_text_font_size = '18px'\n",
    "plot_gs_nov.title.text_font_size = '16px'\n",
    "\n",
    "show(plot_gs_nov)\n",
    "figure_legend(f\"Fig. {fig_count}\", title=f\"Novelties and Sizes of Gene Sets Within {library_name} Library\", content=f\"Scatterplot showing the size, publication count, and novelty rating of each gene set within your library.\")\n",
    "fig_count = fig_count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a scatterplot of library size by novelty for all Enrichr libraries\n",
    "xlabel = 'Library Size (Genes)'\n",
    "ylabel = 'Novelty Statistic'\n",
    "\n",
    "source1 = ColumnDataSource(\n",
    "        data=dict(\n",
    "            x = all_enrichr_novs_mix.Genes,\n",
    "            y = all_enrichr_novs_mix.Stat,\n",
    "            alpha = [0.9] * all_enrichr_novs_mix.shape[0],\n",
    "            size = [7] * all_enrichr_novs_mix.shape[0],\n",
    "            novelty = all_enrichr_novs_mix.Novelty,\n",
    "            lib = all_enrichr_novs_mix.Library,\n",
    "        )\n",
    "    )\n",
    "\n",
    "source2 = ColumnDataSource(\n",
    "        data=dict(\n",
    "            x = my_lib_nov.Genes,\n",
    "            y = my_lib_nov.Stat,\n",
    "            alpha = [0.7] * my_lib_nov.shape[0],\n",
    "            size = [10] * my_lib_nov.shape[0],\n",
    "            novelty = my_lib_nov.Novelty,\n",
    "            lib = my_lib_nov.Library,\n",
    "        )\n",
    "    )\n",
    "\n",
    "source3 = ColumnDataSource(\n",
    "    data=dict(\n",
    "            x = all_enrichr_novs_mouse.Genes,\n",
    "            y = all_enrichr_novs_mouse.Stat,\n",
    "            alpha = [0.9] * all_enrichr_novs_mouse.shape[0],\n",
    "            size = [7] * all_enrichr_novs_mouse.shape[0],\n",
    "            novelty = all_enrichr_novs_mouse.Novelty,\n",
    "            lib = all_enrichr_novs_mouse.Library,\n",
    "        )\n",
    "    )\n",
    "\n",
    "\n",
    "#print(embedding.shape[0])\n",
    "\n",
    "hover_nov = HoverTool(names=[\"df\"], tooltips=\"\"\"\n",
    "    <div style=\"margin: 10\">\n",
    "        <div style=\"margin: 0 auto; width:300px;\">\n",
    "            <span style=\"font-size: 12px; font-weight: bold;\">Gene Set Library:</span>\n",
    "            <span style=\"font-size: 12px\">@lib</span>\n",
    "         <div style=\"margin: 0 auto; width:300px;\">\n",
    "            <span style=\"font-size: 12px; font-weight: bold;\">Novelty:</span>\n",
    "            <span style=\"font-size: 12px\">@novelty</span>\n",
    "        <div style=\"margin: 0 auto; width:300px;\">\n",
    "            <span style=\"font-size: 12px; font-weight: bold;\">Total Genes:</span>\n",
    "            <span style=\"font-size: 12px\">@x</span>\n",
    "        </div>\n",
    "    </div>\n",
    "    \"\"\")\n",
    "tools_nov = [hover_nov, 'pan', 'wheel_zoom', 'reset','save']\n",
    "title_nov = f\"Novelty and Size of {library_name} Library Among All Enrichr Libraries\"\n",
    "plot_nov = figure(plot_width=700, plot_height=700, tools=tools_nov, title=title_nov, x_axis_label=xlabel, y_axis_label=ylabel, x_axis_type='log')\n",
    "plot_nov.circle('x', 'y', size='size', \n",
    "                alpha='alpha', line_alpha=0, line_width=0.01, source=source1, name=\"df\", \n",
    "                fill_color=factor_cmap('novelty', palette=Spectral6, factors=novelties),\n",
    "               legend_field='novelty')\n",
    "plot_nov.square('x', 'y', size='size', \n",
    "                alpha='alpha', line_alpha=0, line_width=0.01, source=source3, name=\"df\", \n",
    "                fill_color=factor_cmap('novelty', palette=Spectral6, factors=novelties))\n",
    "if other:\n",
    "    plot_nov.triangle('x', 'y', size='size', \n",
    "                alpha='alpha', line_alpha=0, line_width=0.01, source=source2, name=\"df\", \n",
    "                fill_color='red')\n",
    "if specname=='Mouse':\n",
    "    plot_nov.square('x', 'y', size='size', \n",
    "                    alpha='alpha', line_alpha=0, line_width=0.01, source=source2, name=\"df\", \n",
    "                    fill_color='red')\n",
    "if specname=='Human':\n",
    "    plot_nov.circle('x', 'y', size=10, \n",
    "                alpha='alpha', line_alpha=0, line_width=0.01, source=source2, name=\"df\", \n",
    "                fill_color='red')\n",
    "plot_nov.xaxis.axis_label_text_font_style = 'normal'\n",
    "plot_nov.yaxis.axis_label_text_font_style = 'normal'\n",
    "plot_nov.title.align = 'center'\n",
    "plot_nov.legend.location = (170,520)\n",
    "plot_nov.xaxis.axis_label_text_font_size = '18px'\n",
    "plot_nov.yaxis.axis_label_text_font_size = '18px'\n",
    "plot_nov.title.text_font_size = '16px'\n",
    "\n",
    "from bokeh.models import Legend, LegendItem\n",
    "markers = ['circle','square','triangle']\n",
    "r = plot_nov.scatter(x=0, y=0, color=\"grey\", size=6, marker=markers)\n",
    "r.visible = False\n",
    "\n",
    "shape_legend = Legend(items=[\n",
    "    LegendItem(label=\"human harmonized\", renderers=[r], index=0),\n",
    "    LegendItem(label=\"mouse harmonized\", renderers=[r], index=1),\n",
    "    LegendItem(label=\"other\", renderers=[r], index=2),],\n",
    "    location=(10,520)           \n",
    ")\n",
    "plot_nov.add_layout(shape_legend)\n",
    "\n",
    "show(plot_nov)\n",
    "figure_legend(f\"Fig. {fig_count}\", title=f\"Novelty and Size of {library_name} Library Among All Enrichr Libraries\", content=f\"Scatterplot showing the size and novelty of your library compared with 174 Enrichr libraries. The Library Novelty Statistic ranges from 0 (well-studied) to 3 (highly understudied). The {library_name} library is shown in red.\")\n",
    "fig_count = fig_count + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bokeh barplot\n",
    "def make_bok_barplot(dat, col1name, col2name, title, lab1, lab2, tooltips_vert, tooltips_hor):\n",
    "    barsource_v = ColumnDataSource(\n",
    "            dict(\n",
    "            x = dat[col1name],\n",
    "            y = dat[col2name],\n",
    "            novelty = dat['Novelty'][::-1],\n",
    "        )\n",
    "    )\n",
    "    barsource_h = ColumnDataSource(\n",
    "            dict(\n",
    "            x = dat[col2name][::-1],\n",
    "            y = dat[col1name][::-1],\n",
    "            novelty = dat['Novelty'][::-1],\n",
    "        )\n",
    "    )\n",
    "\n",
    "    bar_title = title\n",
    "\n",
    "    if orient_bar == 'Vertical':\n",
    "        bokbar = figure(x_range=dat[col1name], plot_height=350, title=bar_title, toolbar_location='below', tooltips=tooltips_vert, x_axis_label=lab1, y_axis_label=lab2)\n",
    "        bokbar.vbar(x=dat[col1name], top=dat[col2name], width=.5, color=bar_color, hover_alpha=.7)\n",
    "        bokbar.xaxis.major_label_orientation = math.pi/5\n",
    "        bokbar.xgrid.grid_line_color = None\n",
    "        bokbar.y_range.start = 0\n",
    "\n",
    "        if counts_bar:\n",
    "            labels = LabelSet(x='x', y='y', text='y', level='annotation',\n",
    "                x_offset=-7, y_offset=0, source=barsource_v, render_mode='canvas', text_font_size = '11px')\n",
    "\n",
    "            bokbar.add_layout(labels)\n",
    "\n",
    "    if orient_bar == 'Horizontal':\n",
    "        bokbar = figure(y_range = dat[col1name][::-1], plot_height=400, title=bar_title, toolbar_location='below', tooltips=tooltips_hor, x_axis_label=lab2, y_axis_label=lab1)\n",
    "        bokbar.hbar(y='y',right='x', height=.5, color=bar_color, hover_alpha=.7, source=barsource_h)\n",
    "        bokbar.xgrid.grid_line_color = None\n",
    "\n",
    "        if counts_bar:\n",
    "            labels = LabelSet(x='x', y='y', text='x', level='annotation',\n",
    "                x_offset=2, y_offset=-6, source=barsource_h, render_mode='canvas', text_font_size = '11px')\n",
    "\n",
    "            bokbar.add_layout(labels)\n",
    "\n",
    "    bokbar.xaxis.axis_label_text_font_style = 'normal'\n",
    "    bokbar.yaxis.axis_label_text_font_style = 'normal'\n",
    "    bokbar.title.align = 'center'\n",
    "    \n",
    "    return bokbar  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bokbar_counts_title = '' + str(num_bar_genes) + ' Most Frequent Genes in ' + library_name\n",
    "bokbar_pubs_title = '' + str(num_bar_genes) + ' Most Studied Genes in ' + library_name\n",
    "\n",
    "tooltips_vert = [\n",
    "    (\"count\", \"@top\")\n",
    "    ]\n",
    "\n",
    "tooltips_hor = [\n",
    "    (\"count\", \"@x\")\n",
    "    ]\n",
    "\n",
    "bokbar_counts = make_bok_barplot(top_genes, 'Gene', 'Count', bokbar_counts_title, 'Genes', 'Counts', tooltips_vert, tooltips_hor)\n",
    "show(bokbar_counts)\n",
    "figure_legend(f\"Fig. {fig_count}\", title=f\"Most Frequent Genes in {library_name}\")\n",
    "fig_count = fig_count + 1\n",
    "\n",
    "if len(pubhist_dat) > 0:\n",
    "    tooltips_vert = [\n",
    "        (\"publications\", \"@top\"),\n",
    "        (\"novelty\", \"@novelty\")\n",
    "    ]\n",
    "\n",
    "    tooltips_hor = [\n",
    "        (\"publications\", \"@x\"),\n",
    "        (\"novelty\", \"@novelty\")\n",
    "    ]\n",
    "    \n",
    "    bokbar_pubs = make_bok_barplot(top_pub_genes, 'Gene', 'Publications', bokbar_pubs_title, 'Genes', 'Publications', tooltips_vert, tooltips_hor)\n",
    "    show(bokbar_pubs)\n",
    "    figure_legend(f\"Fig. {fig_count}\", title=f\"Most Studied Genes in {library_name}\")\n",
    "    fig_count = fig_count + 1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "my-first-appyter-2",
   "language": "python",
   "name": "my-first-appyter-2"
  },
  "language_info": {
   "name": "python",
   "version": ""
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}