{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Cell magic `%%appyter` not found.\n"
     ]
    }
   ],
   "source": [
    "%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# tcgaDataLoader\n",
    "\n",
    "An appyter that interfaces with the The Cancer Genome Atlas (TCGA) API to simplify the process of obtaining this dataset's RNA-seq and clinical data tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Script to acquire RNA-Seq and clinical data from TCGA\n",
    "import requests\n",
    "import json\n",
    "import os\n",
    "import re\n",
    "import gzip\n",
    "import shutil\n",
    "import tarfile\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from maayanlab_bioinformatics.harmonization import ncbi_genes\n",
    "import math\n",
    "import io\n",
    "from gzip import GzipFile\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Cell magic `%%appyter` not found.\n"
     ]
    }
   ],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name=\"RNASeq\",\n",
    "    title=\"RNA-seq specifications\",\n",
    "    img = \"tcga-logo.png\"\n",
    ") %}\n",
    "\n",
    "\n",
    "{% do SectionField(\n",
    "    name=\"CLINICAL\",\n",
    "    title=\"Clincial data specifications\",\n",
    "    img = \"tcga-logo.png\"\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Cell magic `%%appyter` not found.\n"
     ]
    }
   ],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "rna_types = {{MultiChoiceField(\n",
    "    name = \"RNA_types\",\n",
    "    label = \"Types of RNAs to include in the output\",\n",
    "    choices = ['protein-coding','pseudo', 'other', 'unknown','ncRNA','tRNA','rRNA','scRNA','snoRNA','snRNA','biological-region'],\n",
    "    section=\"RNASeq\",\n",
    "    default=[],\n",
    ") }}\n",
    "\n",
    "map_ids = {{BoolField(\n",
    "    name = \"map_ids\",\n",
    "    label = \"Map ensembl IDs to Entrez gene symbols?\",\n",
    "    default = True,\n",
    "    section = \"RNASeq\"\n",
    ") }}\n",
    "    \n",
    "min_cases = {{IntField(\n",
    "    name = \"min_documents\",\n",
    "    label = \"Minimum cases per cancer\",\n",
    "    description = \"The minimum number of cases under a single cancer type for that data to be loaded into the dataset.\",\n",
    "    section=\"RNASeq\",\n",
    "    default=0,\n",
    "    min=0,\n",
    "    max=100000\n",
    ") }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Endpoints\n",
    "base_url = 'https://api.gdc.cancer.gov/'\n",
    "files_endpt = base_url + 'files/'\n",
    "genes_endpt = base_url + 'genes/'\n",
    "cases_endpt = base_url + 'cases/'\n",
    "data_endpt = base_url + \"data/\"\n",
    "\n",
    "json_header = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "workflow_type = \"HTSeq - Counts\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build params \n",
    "\n",
    "filters = {\n",
    "        \"op\": \"and\",\n",
    "        \"content\":[\n",
    "             {\n",
    "                \"op\": \"=\",\n",
    "                \"content\":\n",
    "                 {\n",
    "                     \"field\": \"files.experimental_strategy\",\n",
    "                     \"value\": \"RNA-Seq\",\n",
    "                 }\n",
    "             },\n",
    "             {\n",
    "                \"op\": \"=\",\n",
    "                \"content\":\n",
    "                 {\n",
    "                     \"field\": \"access\",\n",
    "                     \"value\": \"open\",\n",
    "\n",
    "                 }\n",
    "             },\n",
    "             {\n",
    "                \"op\": \"=\",\n",
    "                \"content\":\n",
    "                 {\n",
    "                     \"field\": \"files.analysis.workflow_type\",\n",
    "                     \"value\": workflow_type,\n",
    "                 }\n",
    "             },\n",
    "    ]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "198 total cancer types\n",
      "\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'min_cases' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-141-95ac9df886bd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'{len(buckets)} total cancer types\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0mcancer_types\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"key\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0my\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"doc_count\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0mmin_cases\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuckets\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'{len(cancer_types)} cancer types with at least {min_cases} associated files'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-141-95ac9df886bd>\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'{len(buckets)} total cancer types\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0mcancer_types\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"key\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0my\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"doc_count\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0mmin_cases\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuckets\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'{len(cancer_types)} cancer types with at least {min_cases} associated files'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'min_cases' is not defined"
     ]
    }
   ],
   "source": [
    "# Get list of all cancer types returned from the search parameters\n",
    "\n",
    "params = {\n",
    "    \"filters\": filters,\n",
    "    \"size\":\"0\",\n",
    "    \"facets\":\"cases.diagnoses.primary_diagnosis\",\n",
    "}\n",
    "\n",
    "response = requests.post(files_endpt, data = json.dumps(params), headers=json_header).json() # optionally also provide params argument\n",
    "buckets = response[\"data\"][\"aggregations\"][\"cases.diagnoses.primary_diagnosis\"][\"buckets\"]\n",
    "\n",
    "print(f'{len(buckets)} total cancer types\\n')\n",
    "\n",
    "cancer_types = [ y[\"key\"] for y in list(filter(lambda x: x[\"doc_count\"] >= min_cases, buckets)) ]\n",
    "\n",
    "print(f'{len(cancer_types)} cancer types with at least {min_cases} associated files')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_cancer_params(cancer_type):\n",
    "    # filter results for this cancer type\n",
    "    cancer_filters = {\n",
    "        \"op\": \"and\",\n",
    "        \"content\": [\n",
    "            *filters[\"content\"],\n",
    "            {\n",
    "                \"op\": \"=\",\n",
    "                \"content\":\n",
    "                 {\n",
    "                     \"field\": \"cases.diagnoses.primary_diagnosis\",\n",
    "                     \"value\": cancer_type,\n",
    "                 }\n",
    "             }\n",
    "        ]\n",
    "    }\n",
    "    \n",
    "    fields = \"file_id,file_name,cases.case_id\"\n",
    "    \n",
    "    # build parameters object\n",
    "    return {\n",
    "        \"fields\": fields,\n",
    "        \"filters\": json.dumps(cancer_filters),\n",
    "        \"size\": 100000 # do not limit size\n",
    "    }\n",
    "\n",
    "  \n",
    "def get_uuids(cancer_type):\n",
    "    # get list of all files with RNA-seq results\n",
    "    response = requests.get(files_endpt, params = make_cancer_params(cancer_type))\n",
    "    data = json.loads(response.content.decode(\"utf-8\"))\n",
    "    \n",
    "    # get list of results\n",
    "    results = data[\"data\"][\"hits\"]\n",
    "        \n",
    "    # get list of file and case uuids\n",
    "    file_uuid_list = [ entry[\"file_id\"] for entry in results]\n",
    "    case_uuid_list = [ entry[\"cases\"][0][\"case_id\"] for entry in results]\n",
    "    \n",
    "    # create a dictionary to map file uuids to case uuids\n",
    "    files_to_cases = {}\n",
    "    for entry in results:\n",
    "        files_to_cases[entry[\"file_id\"]] = entry[\"cases\"][0][\"case_id\"]        \n",
    "    \n",
    "    return file_uuid_list, case_uuid_list, files_to_cases\n",
    "\n",
    "def get_files(uuid_list,files_to_cases):\n",
    "    params = {\"ids\": uuid_list}\n",
    "    \n",
    "    df = pd.DataFrame({\"ensembl_id\": []}).set_index(\"ensembl_id\")\n",
    "    \n",
    "    # A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "    response = requests.post(data_endpt, data=json.dumps(params), headers=json_header)\n",
    "\n",
    "    # filename is found in the Content-Disposition header of response\n",
    "    response_head_cd = response.headers[\"Content-Disposition\"]\n",
    "    file_name = re.findall(\"filename=(.+)\", response_head_cd)[0]\n",
    "    \n",
    "    # write the file content (bytes) and load the tar.gz file\n",
    "    file_object = io.BytesIO(response.content)\n",
    "    tar = tarfile.open(fileobj=file_object)\n",
    "    \n",
    "    # access files within the zipped file\n",
    "    for file in tar.getmembers():\n",
    "        if (file.name == \"MANIFEST.txt\"): continue\n",
    "        f=tar.extractfile(file)\n",
    "        bytestream=io.BytesIO(f.read())\n",
    "        got_text = GzipFile(None, 'rb', fileobj=bytestream).read().decode('utf-8')\n",
    "        \n",
    "        # set column name to uuid of corresponding case for this file\n",
    "        file_name = file.name.split(\"/\")[1].split(\".\")[0]\n",
    "        \n",
    "        new_df = pd.read_csv(io.StringIO(got_text), sep=\"\\t\",header=None, names=[\"ensembl_id\", file_name])\n",
    "            \n",
    "        # collapse all versioned names of genes to just gene name so we can merge\n",
    "        new_df.ensembl_id.replace(to_replace = r'\\..*$', value = \"\", regex=True, inplace=True)\n",
    "        new_df = new_df.set_index(\"ensembl_id\")\n",
    "        \n",
    "        df = pd.DataFrame.merge(df, new_df, how=\"outer\", left_index = True, right_index = True)\n",
    "    \n",
    "    tar.close()\n",
    "    \n",
    "    # drop rows not corresponding to genes (i.e. metadata)\n",
    "    non_genes = list(filter(lambda val: not \"ENSG\" in val, list(df.index.values)))\n",
    "    df = df.drop(non_genes)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def get_ncbi_df():\n",
    "    # Map Ensebml ids to Entrez gene symbols\n",
    "    ncbi = pd.DataFrame(ncbi_genes.ncbi_genes_fetch())\n",
    "    all_ids = ncbi.dbXrefs.values\n",
    "\n",
    "    def get_ensembl_id(ids):\n",
    "        ids = \"\".join(ids)\n",
    "        ensembl = re.findall(\"Ensembl:(.*)\", ids)\n",
    "        if (len(ensembl) == 1):\n",
    "            return ensembl[0]\n",
    "        else:\n",
    "            return None\n",
    "        \n",
    "    ensembl_ids = [ get_ensembl_id(ids) for ids in all_ids]\n",
    "\n",
    "    ncbi = ncbi[[\"dbXrefs\", \"Symbol\", \"type_of_gene\"]]\n",
    "    ncbi[\"ensembl\"] = ensembl_ids\n",
    "    ncbi = ncbi.drop(columns=[\"dbXrefs\"])\n",
    "    ncbi = ncbi.set_index(\"ensembl\")\n",
    "    \n",
    "    display(ncbi.head())\n",
    "    print(ncbi[\"type_of_gene\"].unique())\n",
    "    return ncbi\n",
    "\n",
    "\n",
    "\n",
    "def map_ncbi_data(df, ncbi, rna_types):\n",
    "    \n",
    "    ensembl_to_gene_type = ncbi.to_dict()[\"type_of_gene\"]\n",
    "    ensembl_to_symbol = ncbi.to_dict()[\"Symbol\"]\n",
    "\n",
    "    data_ensembl_ids = df.index.to_list()\n",
    "\n",
    "    def id_to_type(key):\n",
    "        if (key in ensembl_to_gene_type):\n",
    "            return ensembl_to_gene_type[key]\n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    # if the key is present, return it; otherwise, set the index for the corresponding row as its ensembl id\n",
    "    def id_to_symbol(key):\n",
    "        if (key in ensembl_to_symbol):\n",
    "            return ensembl_to_symbol[key]\n",
    "        else:\n",
    "            return key # if entrez symbol not found, keep as ensembl id\n",
    "\n",
    "    data_types = [ id_to_type(key) for key in data_ensembl_ids ]\n",
    "    data_symbols = [ id_to_symbol(key) for key in data_ensembl_ids ]\n",
    "\n",
    "    df[\"type_of_gene\"] = data_types\n",
    "    df[\"symbol\"] = data_symbols\n",
    "\n",
    "    df = df[df['type_of_gene'].isin(rna_types)]\n",
    "    df = df.reindex(columns=(['symbol','type_of_gene'] + list([a for a in df.columns if not a in ['symbol','type_of_gene']] )))\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000121410</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000175899</th>\n",
       "      <td>A2M</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000256069</th>\n",
       "      <td>A2MP1</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000171428</th>\n",
       "      <td>NAT1</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000156006</th>\n",
       "      <td>NAT2</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Symbol    type_of_gene\n",
       "ensembl                               \n",
       "ENSG00000121410   A1BG  protein-coding\n",
       "ENSG00000175899    A2M  protein-coding\n",
       "ENSG00000256069  A2MP1          pseudo\n",
       "ENSG00000171428   NAT1  protein-coding\n",
       "ENSG00000156006   NAT2  protein-coding"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['protein-coding' 'pseudo' 'other' 'unknown' 'ncRNA' 'tRNA' 'rRNA' 'scRNA'\n",
      " 'snoRNA' 'snRNA' 'biological-region']\n",
      "endometrioid adenocarcinoma, nos: \n",
      "540 files\n",
      "540 cases\n",
      "\n",
      "Got table for endometrioid adenocarcinoma, nos with 60483 genes\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dae5e5d9-b037-4d4a-90e4-b95fbfeed478</th>\n",
       "      <th>b95c9bdd-1cc0-4b8f-959b-4c1de981135c</th>\n",
       "      <th>e250c5cd-1786-4b25-b143-c8da28a3b77c</th>\n",
       "      <th>a638e045-eb92-4f48-afe8-e57f0e4307eb</th>\n",
       "      <th>f2e7119a-29c3-4f45-9af5-28c3610b0385</th>\n",
       "      <th>c13fc444-6e7b-4ae1-9c90-e539d78417ca</th>\n",
       "      <th>ecf78283-0eb3-4710-8e5a-30f56da510bd</th>\n",
       "      <th>51500805-d88a-4f67-9beb-04a7228033c8</th>\n",
       "      <th>98ac488f-bf05-4175-b9f2-b33cb46665e3</th>\n",
       "      <th>f40d273f-5343-4e57-995b-6adcfdc7c7d3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>4511</td>\n",
       "      <td>1884</td>\n",
       "      <td>1374</td>\n",
       "      <td>1407</td>\n",
       "      <td>2747</td>\n",
       "      <td>3043</td>\n",
       "      <td>3054</td>\n",
       "      <td>558</td>\n",
       "      <td>4576</td>\n",
       "      <td>3712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>661</td>\n",
       "      <td>470</td>\n",
       "      <td>146</td>\n",
       "      <td>349</td>\n",
       "      <td>584</td>\n",
       "      <td>591</td>\n",
       "      <td>1213</td>\n",
       "      <td>116</td>\n",
       "      <td>1172</td>\n",
       "      <td>456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>480</td>\n",
       "      <td>1251</td>\n",
       "      <td>272</td>\n",
       "      <td>227</td>\n",
       "      <td>379</td>\n",
       "      <td>343</td>\n",
       "      <td>523</td>\n",
       "      <td>728</td>\n",
       "      <td>257</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>166</td>\n",
       "      <td>218</td>\n",
       "      <td>180</td>\n",
       "      <td>85</td>\n",
       "      <td>126</td>\n",
       "      <td>156</td>\n",
       "      <td>133</td>\n",
       "      <td>81</td>\n",
       "      <td>381</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 dae5e5d9-b037-4d4a-90e4-b95fbfeed478  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  4511   \n",
       "ENSG00000000005                                     1   \n",
       "ENSG00000000419                                   661   \n",
       "ENSG00000000457                                   480   \n",
       "ENSG00000000460                                   166   \n",
       "\n",
       "                 b95c9bdd-1cc0-4b8f-959b-4c1de981135c  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  1884   \n",
       "ENSG00000000005                                     0   \n",
       "ENSG00000000419                                   470   \n",
       "ENSG00000000457                                  1251   \n",
       "ENSG00000000460                                   218   \n",
       "\n",
       "                 e250c5cd-1786-4b25-b143-c8da28a3b77c  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  1374   \n",
       "ENSG00000000005                                     1   \n",
       "ENSG00000000419                                   146   \n",
       "ENSG00000000457                                   272   \n",
       "ENSG00000000460                                   180   \n",
       "\n",
       "                 a638e045-eb92-4f48-afe8-e57f0e4307eb  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  1407   \n",
       "ENSG00000000005                                     2   \n",
       "ENSG00000000419                                   349   \n",
       "ENSG00000000457                                   227   \n",
       "ENSG00000000460                                    85   \n",
       "\n",
       "                 f2e7119a-29c3-4f45-9af5-28c3610b0385  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  2747   \n",
       "ENSG00000000005                                     2   \n",
       "ENSG00000000419                                   584   \n",
       "ENSG00000000457                                   379   \n",
       "ENSG00000000460                                   126   \n",
       "\n",
       "                 c13fc444-6e7b-4ae1-9c90-e539d78417ca  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  3043   \n",
       "ENSG00000000005                                     3   \n",
       "ENSG00000000419                                   591   \n",
       "ENSG00000000457                                   343   \n",
       "ENSG00000000460                                   156   \n",
       "\n",
       "                 ecf78283-0eb3-4710-8e5a-30f56da510bd  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  3054   \n",
       "ENSG00000000005                                    28   \n",
       "ENSG00000000419                                  1213   \n",
       "ENSG00000000457                                   523   \n",
       "ENSG00000000460                                   133   \n",
       "\n",
       "                 51500805-d88a-4f67-9beb-04a7228033c8  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                   558   \n",
       "ENSG00000000005                                     0   \n",
       "ENSG00000000419                                   116   \n",
       "ENSG00000000457                                   728   \n",
       "ENSG00000000460                                    81   \n",
       "\n",
       "                 98ac488f-bf05-4175-b9f2-b33cb46665e3  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  4576   \n",
       "ENSG00000000005                                    33   \n",
       "ENSG00000000419                                  1172   \n",
       "ENSG00000000457                                   257   \n",
       "ENSG00000000460                                   381   \n",
       "\n",
       "                 f40d273f-5343-4e57-995b-6adcfdc7c7d3  \n",
       "ensembl_id                                             \n",
       "ENSG00000000003                                  3712  \n",
       "ENSG00000000005                                     2  \n",
       "ENSG00000000419                                   456  \n",
       "ENSG00000000457                                   161  \n",
       "ENSG00000000460                                    62  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>dae5e5d9-b037-4d4a-90e4-b95fbfeed478</th>\n",
       "      <th>b95c9bdd-1cc0-4b8f-959b-4c1de981135c</th>\n",
       "      <th>e250c5cd-1786-4b25-b143-c8da28a3b77c</th>\n",
       "      <th>a638e045-eb92-4f48-afe8-e57f0e4307eb</th>\n",
       "      <th>f2e7119a-29c3-4f45-9af5-28c3610b0385</th>\n",
       "      <th>c13fc444-6e7b-4ae1-9c90-e539d78417ca</th>\n",
       "      <th>ecf78283-0eb3-4710-8e5a-30f56da510bd</th>\n",
       "      <th>51500805-d88a-4f67-9beb-04a7228033c8</th>\n",
       "      <th>98ac488f-bf05-4175-b9f2-b33cb46665e3</th>\n",
       "      <th>f40d273f-5343-4e57-995b-6adcfdc7c7d3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>4511</td>\n",
       "      <td>1884</td>\n",
       "      <td>1374</td>\n",
       "      <td>1407</td>\n",
       "      <td>2747</td>\n",
       "      <td>3043</td>\n",
       "      <td>3054</td>\n",
       "      <td>558</td>\n",
       "      <td>4576</td>\n",
       "      <td>3712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>661</td>\n",
       "      <td>470</td>\n",
       "      <td>146</td>\n",
       "      <td>349</td>\n",
       "      <td>584</td>\n",
       "      <td>591</td>\n",
       "      <td>1213</td>\n",
       "      <td>116</td>\n",
       "      <td>1172</td>\n",
       "      <td>456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>480</td>\n",
       "      <td>1251</td>\n",
       "      <td>272</td>\n",
       "      <td>227</td>\n",
       "      <td>379</td>\n",
       "      <td>343</td>\n",
       "      <td>523</td>\n",
       "      <td>728</td>\n",
       "      <td>257</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>166</td>\n",
       "      <td>218</td>\n",
       "      <td>180</td>\n",
       "      <td>85</td>\n",
       "      <td>126</td>\n",
       "      <td>156</td>\n",
       "      <td>133</td>\n",
       "      <td>81</td>\n",
       "      <td>381</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   symbol    type_of_gene  \\\n",
       "ensembl_id                                  \n",
       "ENSG00000000003    TSPAN6  protein-coding   \n",
       "ENSG00000000005      TNMD  protein-coding   \n",
       "ENSG00000000419      DPM1  protein-coding   \n",
       "ENSG00000000457     SCYL3  protein-coding   \n",
       "ENSG00000000460  C1orf112  protein-coding   \n",
       "\n",
       "                 dae5e5d9-b037-4d4a-90e4-b95fbfeed478  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  4511   \n",
       "ENSG00000000005                                     1   \n",
       "ENSG00000000419                                   661   \n",
       "ENSG00000000457                                   480   \n",
       "ENSG00000000460                                   166   \n",
       "\n",
       "                 b95c9bdd-1cc0-4b8f-959b-4c1de981135c  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  1884   \n",
       "ENSG00000000005                                     0   \n",
       "ENSG00000000419                                   470   \n",
       "ENSG00000000457                                  1251   \n",
       "ENSG00000000460                                   218   \n",
       "\n",
       "                 e250c5cd-1786-4b25-b143-c8da28a3b77c  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  1374   \n",
       "ENSG00000000005                                     1   \n",
       "ENSG00000000419                                   146   \n",
       "ENSG00000000457                                   272   \n",
       "ENSG00000000460                                   180   \n",
       "\n",
       "                 a638e045-eb92-4f48-afe8-e57f0e4307eb  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  1407   \n",
       "ENSG00000000005                                     2   \n",
       "ENSG00000000419                                   349   \n",
       "ENSG00000000457                                   227   \n",
       "ENSG00000000460                                    85   \n",
       "\n",
       "                 f2e7119a-29c3-4f45-9af5-28c3610b0385  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  2747   \n",
       "ENSG00000000005                                     2   \n",
       "ENSG00000000419                                   584   \n",
       "ENSG00000000457                                   379   \n",
       "ENSG00000000460                                   126   \n",
       "\n",
       "                 c13fc444-6e7b-4ae1-9c90-e539d78417ca  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  3043   \n",
       "ENSG00000000005                                     3   \n",
       "ENSG00000000419                                   591   \n",
       "ENSG00000000457                                   343   \n",
       "ENSG00000000460                                   156   \n",
       "\n",
       "                 ecf78283-0eb3-4710-8e5a-30f56da510bd  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  3054   \n",
       "ENSG00000000005                                    28   \n",
       "ENSG00000000419                                  1213   \n",
       "ENSG00000000457                                   523   \n",
       "ENSG00000000460                                   133   \n",
       "\n",
       "                 51500805-d88a-4f67-9beb-04a7228033c8  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                   558   \n",
       "ENSG00000000005                                     0   \n",
       "ENSG00000000419                                   116   \n",
       "ENSG00000000457                                   728   \n",
       "ENSG00000000460                                    81   \n",
       "\n",
       "                 98ac488f-bf05-4175-b9f2-b33cb46665e3  \\\n",
       "ensembl_id                                              \n",
       "ENSG00000000003                                  4576   \n",
       "ENSG00000000005                                    33   \n",
       "ENSG00000000419                                  1172   \n",
       "ENSG00000000457                                   257   \n",
       "ENSG00000000460                                   381   \n",
       "\n",
       "                 f40d273f-5343-4e57-995b-6adcfdc7c7d3  \n",
       "ensembl_id                                             \n",
       "ENSG00000000003                                  3712  \n",
       "ENSG00000000005                                     2  \n",
       "ENSG00000000419                                   456  \n",
       "ENSG00000000457                                   161  \n",
       "ENSG00000000460                                    62  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# note: this will take a while!\n",
    "ncbi = get_ncbi_df()\n",
    "cancer_dfs = {}\n",
    "\n",
    "for cancer in cancer_types:\n",
    "    file_uuid_list, case_uuid_list, files_to_cases = get_uuids(cancer)\n",
    "\n",
    "    num_files = len(file_uuid_list)\n",
    "    num_cases = len(case_uuid_list)\n",
    "    \n",
    "    print(f\"{cancer}: \\n{num_files} files\\n{num_cases} cases\\n\")\n",
    "    \n",
    "    df_rna = get_files(file_uuid_list,files_to_cases)\n",
    "        \n",
    "    df_rna = map_ncbi_data(df_rna, ncbi, rna_types)\n",
    "    \n",
    "    if (map_ids):\n",
    "        df_rna = df_rna.drop(\"ensembl_id\").set_index(\"symbol\")\n",
    "    \n",
    "    print(f\"Got table for {cancer} with {df_rna.shape[0]} genes\")\n",
    "\n",
    "    display(df_rna.head())\n",
    "    \n",
    "    cancer_dfs[cancer] = df_rna\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TCGA",
   "language": "python",
   "name": "tcga"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
