{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ca2541e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc869493",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# Gene Set Library Synopsis\n",
    "This Appyter counts the total genes, total gene sets, and individual gene frequencies in a submitted gene set library.\n",
    "Results are visualized as tables and bar graphs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b029f8a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%% Imports\n",
    "import appyter\n",
    "import pandas as pd\n",
    "import os\n",
    "from os import path\n",
    "import random\n",
    "import time\n",
    "import math\n",
    "import numpy as np\n",
    "import warnings\n",
    "import base64\n",
    "import h5py\n",
    "from pandas.api.types import CategoricalDtype\n",
    "\n",
    "import umap.umap_ as umap\n",
    "from maayanlab_bioinformatics.enrichment import enrich_crisp\n",
    "import pyarrow.feather as feather\n",
    "\n",
    "import plotly\n",
    "from plotly import tools\n",
    "import plotly.express as px\n",
    "import plotly.graph_objs as go\n",
    "plotly.offline.init_notebook_mode()\n",
    "\n",
    "import matplotlib.pyplot as plt; plt.rcdefaults()\n",
    "from matplotlib import rcParams\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib_venn import venn2, venn3\n",
    "%matplotlib inline\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import IPython\n",
    "from IPython.display import HTML, display, FileLink, Markdown, IFrame\n",
    "\n",
    "import chart_studio\n",
    "import chart_studio.plotly as py\n",
    "\n",
    "import uuid\n",
    "import urllib\n",
    "from textwrap import dedent\n",
    "import json\n",
    "import requests\n",
    "from operator import itemgetter\n",
    "import leidenalg as la\n",
    "import igraph as ig\n",
    "\n",
    "from six.moves import urllib\n",
    "from itertools import combinations\n",
    "from itertools import chain\n",
    "import scipy.spatial.distance as dist\n",
    "import scipy.stats as ss\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import quantile_transform, StandardScaler\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.manifold import TSNE\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "from utils import *\n",
    "\n",
    "from plotly.offline import init_notebook_mode\n",
    "init_notebook_mode(connected = True)\n",
    "\n",
    "from bokeh.io import output_notebook, export_svg\n",
    "from bokeh.io.export import get_screenshot_as_png\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool, CustomJS, ColumnDataSource, Span, ranges, LabelSet\n",
    "from bokeh.layouts import layout, row, column, gridplot\n",
    "from bokeh.palettes import all_palettes\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e3ececc",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "{% do SectionField(\n",
    "    name='introSection', \n",
    "    title='Gene Set Library Synopsis', \n",
    "    subtitle='This Appyter performs basic analysis on a submitted gene set library.', \n",
    "    img='gene-library-3.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='geneEntrySection', \n",
    "    title='1. Enter a Gene Library',\n",
    "    subtitle='Upload a GMT file containing a gene set library or select an existing Enrichr library. You may also choose to use the default gene set library provided.',\n",
    "    img='data-upload-icon.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='barGraphSection', \n",
    "    title='2. Bar Graph',\n",
    "    subtitle='Generate a bar graph of the most common genes in the library.',\n",
    "    img='bar-icon.png'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60ff77e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Inputting libraries and settings\n",
    "{% set library_kind = TabField(\n",
    "    name = 'library_kind',\n",
    "    label = 'Library',\n",
    "    default = 'Upload a library',\n",
    "    description = '',\n",
    "    choices = {\n",
    "        'Upload a library': [\n",
    "            FileField(\n",
    "                name = 'library_filename',\n",
    "                label = 'Gene Library File (.gmt or .txt)',\n",
    "                default = 'Cancer_Cell_Line_Encyclopedia.gmt', \n",
    "                examples = {'Cancer_Cell_Line_Encyclopedia.gmt': url_for('static', filename = 'Cancer_Cell_Line_Encyclopedia.gmt'),\n",
    "                            'Cancer_Cell_Line_Encyclopedia.txt': url_for('static', filename = 'Cancer_Cell_Line_Encyclopedia.txt')}, \n",
    "                description = 'GMT is a tab-delimited file format that describes sets. Visit https://bit.ly/35crtXQ for more information and http://www.molmine.com/magma/fileformats.html to create your own.',\n",
    "                section = 'geneEntrySection')\n",
    "        ],\n",
    "        \n",
    "        'Select a library from Enrichr': [\n",
    "            ChoiceField(\n",
    "                name = 'enrichr_library',\n",
    "                description = 'Select one Enrichr library whose genes will be counted',\n",
    "                label = 'Enrichr Library',\n",
    "                default = 'Cancer_Cell_Line_Encyclopedia',\n",
    "                section = 'geneEntrySection',\n",
    "                choices = [\n",
    "                    'ARCHS4_Cell-lines',\n",
    "                    'ARCHS4_IDG_Coexp',\n",
    "                    'ARCHS4_Kinases_Coexp',\n",
    "                    'ARCHS4_TFs_Coexp',\n",
    "                    'ARCHS4_Tissues',\n",
    "                    'Achilles_fitness_decrease',\n",
    "                    'Achilles_fitness_increase',\n",
    "                    'Aging_Perturbations_from_GEO_down',\n",
    "                    'Aging_Perturbations_from_GEO_up',\n",
    "                    'Allen_Brain_Atlas_10x_scRNA_2021',\n",
    "                    'Allen_Brain_Atlas_down',\n",
    "                    'Allen_Brain_Atlas_up',\n",
    "                    'BioCarta_2013',\n",
    "                    'BioCarta_2015',\n",
    "                    'BioCarta_2016',\n",
    "                    'BioPlanet_2019',\n",
    "                    'BioPlex_2017',\n",
    "                    'CCLE_Proteomics_2020',\n",
    "                    'CORUM',\n",
    "                    'COVID-19_Related_Gene_Sets',\n",
    "                    'Cancer_Cell_Line_Encyclopedia',\n",
    "                    'ChEA_2013',\n",
    "                    'ChEA_2015',\n",
    "                    'ChEA_2016',\n",
    "                    'Chromosome_Location',           \n",
    "                    'Chromosome_Location_hg19',\n",
    "                    'ClinVar_2019',\n",
    "                    'dbGaP',\n",
    "                    'DSigDB',\n",
    "                    'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "                    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "                    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "                    'DisGeNET',\n",
    "                    'Disease_Perturbations_from_GEO_down',\n",
    "                    'Disease_Perturbations_from_GEO_up',\n",
    "                    'Disease_Signatures_from_GEO_down_2014',\n",
    "                    'Disease_Signatures_from_GEO_up_2014',\n",
    "                    'DrugMatrix',\n",
    "                    'Drug_Perturbations_from_GEO_2014',\n",
    "                    'Drug_Perturbations_from_GEO_down',\n",
    "                    'Drug_Perturbations_from_GEO_up',\n",
    "                    'ENCODE_Histone_Modifications_2013',\n",
    "                    'ENCODE_Histone_Modifications_2015',\n",
    "                    'ENCODE_TF_ChIP-seq_2014',\n",
    "                    'ENCODE_TF_ChIP-seq_2015',\n",
    "                    'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "                    'ESCAPE',\n",
    "                    'Elsevier_Pathway_Collection',\n",
    "                    'Enrichr_Libraries_Most_Popular_Genes',\n",
    "                    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "                    'Enrichr_Users_Contributed_Lists_2020',\n",
    "                    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "                    'GO_Biological_Process_2013',\n",
    "                    'GO_Biological_Process_2015',\n",
    "                    'GO_Biological_Process_2017',\n",
    "                    'GO_Biological_Process_2017b',\n",
    "                    'GO_Biological_Process_2018',\n",
    "                    'GO_Cellular_Component_2013',\n",
    "                    'GO_Cellular_Component_2015',\n",
    "                    'GO_Cellular_Component_2017',\n",
    "                    'GO_Cellular_Component_2017b',\n",
    "                    'GO_Cellular_Component_2018',\n",
    "                    'GO_Molecular_Function_2013',\n",
    "                    'GO_Molecular_Function_2015',\n",
    "                    'GO_Molecular_Function_2017',\n",
    "                    'GO_Molecular_Function_2017b',\n",
    "                    'GO_Molecular_Function_2018',\n",
    "                    'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "                    'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "                    'GWAS_Catalog_2019',\n",
    "                    'GeneSigDB',\n",
    "                    'Gene_Perturbations_from_GEO_down',\n",
    "                    'Gene_Perturbations_from_GEO_up',\n",
    "                    'Genes_Associated_with_NIH_Grants',\n",
    "                    'Genome_Browser_PWMs',\n",
    "                    'HMDB_Metabolites',\n",
    "                    'HMS_LINCS_KinomeScan',\n",
    "                    'HomoloGene',\n",
    "                    'HumanCyc_2015',\n",
    "                    'HumanCyc_2016',\n",
    "                    'Human_Gene_Atlas',\n",
    "                    'Human_Phenotype_Ontology',\n",
    "                    'huMAP',\n",
    "                    'InterPro_Domains_2019',\n",
    "                    'Jensen_COMPARTMENTS',\n",
    "                    'Jensen_DISEASES',\n",
    "                    'Jensen_TISSUES',\n",
    "                    'KEA_2013',\n",
    "                    'KEA_2015',\n",
    "                    'KEGG_2013',\n",
    "                    'KEGG_2015',\n",
    "                    'KEGG_2016',\n",
    "                    'KEGG_2019_Human',\n",
    "                    'KEGG_2019_Mouse',\n",
    "                    'Kinase_Perturbations_from_GEO_down',\n",
    "                    'Kinase_Perturbations_from_GEO_up',\n",
    "                    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "                    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "                    'LINCS_L1000_Chem_Pert_down',\n",
    "                    'LINCS_L1000_Chem_Pert_up',\n",
    "                    'LINCS_L1000_Ligand_Perturbations_down',\n",
    "                    'LINCS_L1000_Ligand_Perturbations_up',\n",
    "                    'Ligand_Perturbations_from_GEO_down',\n",
    "                    'Ligand_Perturbations_from_GEO_up',\n",
    "                    'lncHUB_lncRNA_Co-Expression',\n",
    "                    'MCF7_Perturbations_from_GEO_down',\n",
    "                    'MCF7_Perturbations_from_GEO_up',\n",
    "                    'MGI_Mammalian_Phenotype_2013',\n",
    "                    'MGI_Mammalian_Phenotype_2017',\n",
    "                    'MGI_Mammalian_Phenotype_Level_3',\n",
    "                    'MGI_Mammalian_Phenotype_Level_4',\n",
    "                    'MGI_Mammalian_Phenotype_Level_4_2019',\n",
    "                    'MSigDB_Computational',\n",
    "                    'MSigDB_Hallmark_2020',\n",
    "                    'MSigDB_Oncogenic_Signatures',\n",
    "                    'Microbe_Perturbations_from_GEO_down',\n",
    "                    'Microbe_Perturbations_from_GEO_up',\n",
    "                    'miRTarBase_2017',\n",
    "                    'Mouse_Gene_Atlas',\n",
    "                    'NCI-60_Cancer_Cell_Lines',\n",
    "                    'NCI-Nature_2015',\n",
    "                    'NCI-Nature_2016',\n",
    "                    'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "                    'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "                    'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "                    'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "                    'NURSA_Human_Endogenous_Complexome',\n",
    "                    'OMIM_Disease',\n",
    "                    'OMIM_Expanded',\n",
    "                    'Old_CMAP_down',\n",
    "                    'Old_CMAP_up',\n",
    "                    'PPI_Hub_Proteins',\n",
    "                    'Panther_2015',\n",
    "                    'Panther_2016',\n",
    "                    'Pfam_Domains_2019',\n",
    "                    'Pfam_InterPro_Domains',\n",
    "                    'PheWeb_2019',\n",
    "                    'Phosphatase_Substrates_from_DEPOD',\n",
    "                    'ProteomicsDB_2020',\n",
    "                    'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "                    'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "                    'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "                    'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "                    'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "                    'Reactome_2013',\n",
    "                    'Reactome_2015',\n",
    "                    'Reactome_2016',\n",
    "                    'SILAC_Phosphoproteomics',\n",
    "                    'SubCell_BarCode',\n",
    "                    'SysMyo_Muscle_Gene_Sets'\n",
    "                    'TF-LOF_Expression_from_GEO',\n",
    "                    'TF_Perturbations_Followed_by_Expression',\n",
    "                    'TG_GATES_2020',\n",
    "                    'TRANSFAC_and_JASPAR_PWMs',\n",
    "                    'TRRUST_Transcription_Factors_2019',\n",
    "                    'Table_Mining_of_CRISPR_Studies',\n",
    "                    'TargetScan_microRNA',\n",
    "                    'TargetScan_microRNA_2017',\n",
    "                    'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
    "                    'Tissue_Protein_Expression_from_ProteomicsDB.csv',\n",
    "                    'Transcription_Factor_PPIs',\n",
    "                    'UK_Biobank_GWAS_v1',\n",
    "                    'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "                    'VirusMINT',\n",
    "                    'Virus_Perturbations_from_GEO_down',\n",
    "                    'Virus_Perturbations_from_GEO_up',\n",
    "                    'WikiPathways_2013',\n",
    "                    'WikiPathways_2015',\n",
    "                    'WikiPathways_2016',\n",
    "                    'WikiPathways_2019_Human',\n",
    "                    'WikiPathways_2019_Mouse'\n",
    "                ]\n",
    "            )\n",
    "        ],\n",
    "    },\n",
    "    section = 'geneEntrySection',\n",
    ") %}\n",
    "\n",
    "# Bar Graph Field\n",
    "bargraph = {{ BoolField(\n",
    "    name = 'bargraph',\n",
    "    label = 'Bar Graph?',\n",
    "    default = 'true',\n",
    "    description = 'Select \\'Yes\\' if you would like to generate a bar graph. Otherwise, select \\'No\\'',\n",
    "    section = 'barGraphSection'\n",
    ")}}\n",
    "\n",
    "# Choose number of genes in bar graph\n",
    "num_bar_genes = {{ IntField(\n",
    "    name='num_bar_genes', \n",
    "    label='Displayed Genes', \n",
    "    min=2, \n",
    "    max=20, \n",
    "    default=20, \n",
    "    description='Select the number of genes to display in the bar graph.', \n",
    "    section='barGraphSection'\n",
    ")}}\n",
    "\n",
    "# Choose the orientation of the graph: horizontal or vertical bars\n",
    "orient = \"{{ ChoiceField(name = 'orient', label = 'Orientation', choices = ['Horizontal', 'Vertical'], default = 'Horizontal', description = 'Choose whether your bar graph will be displayed horizontally or vertically', section = 'barGraphSection') }}\"\n",
    "\n",
    "# Choose color of bars\n",
    "color = \"{{ ChoiceField(name = 'color', label = 'Color', choices = ['Black', 'Blue', 'Red', 'Green', 'Grey', 'Orange', 'Purple', 'Yellow', 'Pink'], default = 'Black', section = 'barGraphSection') }}\"\n",
    "\n",
    "# Choose whether gene counts are displayed on bar graph\n",
    "counts = {{ BoolField(name = 'counts', label = 'Show Counts?', default = 'true', description = 'Choose \\'Yes\\' to label the bars with the gene count (number of appearances in library).', section = 'barGraphSection') }}\n",
    "\n",
    "bar_file_name = {{ StringField(\n",
    "    name = 'bar_file_name',\n",
    "    label = 'File Name',\n",
    "    default = 'bargraph',\n",
    "    description = 'Enter a name/description to save your bar graph',\n",
    "    section = 'barGraphSection'\n",
    ")}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ba78214",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Color for Bar plot\n",
    "color_conversion = {\n",
    "    'Black': 'black',\n",
    "    'Blue': 'lightskyblue',\n",
    "    'Red': 'tomato',\n",
    "    'Green': 'mediumspringgreen',\n",
    "    'Grey': 'lightgrey',\n",
    "    'Orange': 'orange',\n",
    "    'Purple': 'plum',\n",
    "    'Yellow': 'yellow',\n",
    "    'Pink': 'lightpink'\n",
    "}\n",
    "\n",
    "bar_color = color_conversion[color]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1458fadd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%- if library_kind.raw_value == 'Upload a library' %}\n",
    "library_kind = \"Upload a library\"\n",
    "library_filename = {{ library_kind.value[0] }}\n",
    "library_name = library_filename.replace(\"_\", \" \").replace(\".txt\", \"\").replace(\".gmt\", \"\")\n",
    "\n",
    "{%- else %}\n",
    "library_kind = \"Select a library from Enrichr\"\n",
    "library_filename = \"{{ library_kind.value[0] }}\"\n",
    "library_name = \"{{ library_kind.value[0] }}\"\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efc3d9c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download library from the Enrichr site using its file name\n",
    "def download_library(library_filename):\n",
    "    with open(f\"{library_filename}\", \"w\") as fw:\n",
    "        with urllib.request.urlopen(f'https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName={library_filename}') as f:\n",
    "            for line in f.readlines():\n",
    "                fw.write(line.decode('utf-8'))\n",
    "                fw.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4395225",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load library \n",
    "def load(library_filename):\n",
    "    if library_kind == \"Select a library from Enrichr\":\n",
    "        download_library(library_filename)\n",
    "    library_data, library_genes = load_library(library_filename)\n",
    "    # to upper case\n",
    "    return library_data, library_genes\n",
    "\n",
    "# Returns a dictionary (library_data) where the values are all the elements\n",
    "def load_library(library_filename):\n",
    "    library_data = dict()\n",
    "    with open(library_filename, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "        library_genes = [''] * len(lines)\n",
    "        i = 0\n",
    "        for line in lines:\n",
    "            splited = line.strip().split(\"\\t\")\n",
    "            elements = pd.Series(splited[2:]).dropna()\n",
    "            if len(elements) > 0:\n",
    "                # to upper case\n",
    "                allxs = [x.upper() for x in elements]\n",
    "                library_data[splited[0]] = allxs\n",
    "                library_genes[i] = (' ').join(allxs)\n",
    "            i = i + 1    \n",
    "    return library_data, library_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8af6c1bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Count the number of each gene \n",
    "library_data, library_genes = load(library_filename)\n",
    "if library_kind == \"Select a library from Enrichr\":\n",
    "    library_name = library_name.replace(\"_\", \" \")\n",
    "vals = list(library_data.values())\n",
    "all_genes = list(chain(*vals))\n",
    "all_genes_unique = np.unique(np.array(all_genes))\n",
    "all_sets = list(library_data.keys())\n",
    "\n",
    "count_frame = pd.Series(all_genes).value_counts().sort_index().reset_index().reset_index(drop=True)\n",
    "count_frame.columns = ['GENE', 'COUNT']\n",
    "count_frame.dropna()\n",
    "count_frame.sort_values(by=['COUNT'], inplace=True, ascending=False)\n",
    "count_frame = count_frame.reset_index(drop=True)\n",
    "top_genes = count_frame.iloc[0:num_bar_genes]\n",
    "\n",
    "unique_genes = len(count_frame['GENE'])\n",
    "unique_sets = len(library_data.keys())\n",
    "\n",
    "# Drop skipped row \n",
    "mask = count_frame['GENE'].str.len() > 0\n",
    "count_frame = count_frame.loc[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e2601b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output a table of genes and counts.\n",
    "def create_download_link(df, title, filename):  \n",
    "    csv = df.to_csv(index = False)\n",
    "    b64 = base64.b64encode(csv.encode())\n",
    "    payload = b64.decode()\n",
    "    html = '<a download=\"{filename}\" href=\"data:text/csv;base64,{payload}\" target=\"_blank\">{title}</a>'\n",
    "    html = html.format(payload=payload, title=title, filename=filename)\n",
    "    return HTML(html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be0957ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# Gene Counts\n",
    "The table below displays the counts (number of appearances throughout the entire library) of each gene. This preview has a limited number of genes, but the full chart is also available for download."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eff40025",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Print out a table of the count_frame dataframe \n",
    "counts_filename = library_name.replace(\" \", \"_\") + \"_gene_counts.csv\"\n",
    "\n",
    "# First print total number of genes and gene sets (terms)\n",
    "print(\"Total genes: {}\".format(unique_genes))\n",
    "print(\"Total gene sets: {}\".format(unique_sets))\n",
    "\n",
    "display(HTML(count_frame[0:num_bar_genes].to_html(index=False)))\n",
    "display(Markdown(f\"*Table 1. Gene count results for {library_name} library*\"))\n",
    "display(create_download_link(count_frame, \"Download this table as a CSV\", counts_filename))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47d74376",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# Unmapped Gene Names\n",
    "This Appyter checks whether your gene set library contains unmapped gene names in _-DEC, _-MAR, and _-SEP formats. These conversions frequently occur when gene names are loaded into Excel. For example, either MARC1 or MARCH1 will automatically become '1-MAR'. Read this article for more information: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1044-7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "067282cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for unmapped genes and display them \n",
    "month_genes = all_genes_unique.copy()\n",
    "month_genes.sort()\n",
    "first = -1\n",
    "last = -1\n",
    "for i in range(len(month_genes)):\n",
    "    if len(month_genes[i]) > 4:\n",
    "        substr = month_genes[i][-4:]\n",
    "        if(substr == '-DEC' or substr == '-MAR' or substr == '-SEP'):\n",
    "            if first == -1:\n",
    "                first = i\n",
    "            else:\n",
    "                last = i+1\n",
    "        else:\n",
    "            if first != -1:\n",
    "                break\n",
    "\n",
    "def month_sorter(month):\n",
    "    return month[-3]\n",
    "\n",
    "def date_sorter(month):\n",
    "    dash = month.index('-')\n",
    "    return int(month[:dash])\n",
    "\n",
    "month_genes = month_genes[first:last]\n",
    "month_genes = sorted(month_genes,key=lambda x: (month_sorter(x), date_sorter(x)))\n",
    "month_genes = pd.DataFrame(data=month_genes, columns=['GENE NAME'])\n",
    "\n",
    "# Display if incorrect genes \n",
    "if len(month_genes) > 0:\n",
    "    month_genes_filename = 'unmapped_gene_names_' + library_name\n",
    "    print('' + str(len(month_genes)) + ' unmapped gene names found.')\n",
    "    display(HTML(month_genes.to_html(index=False)))\n",
    "    display(Markdown(f\"*Table 2. Unmapped gene names in {library_name} library*\"))\n",
    "    display(create_download_link(month_genes, \"Download this table as a CSV\", month_genes_filename))\n",
    "else:\n",
    "    print(\"No unmapped gene names found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "551b7917",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# Bar Graph\n",
    "The bar graph displays the most common genes in your gene set library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3e755c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bokeh bar graph\n",
    "if bargraph:\n",
    "    barsource_v = ColumnDataSource(\n",
    "            dict(\n",
    "            x = top_genes['GENE'],\n",
    "            y = top_genes['COUNT']\n",
    "        )\n",
    "    )\n",
    "    barsource_h = ColumnDataSource(\n",
    "            dict(\n",
    "            x = top_genes['COUNT'][::-1],\n",
    "            y = top_genes['GENE'][::-1]\n",
    "        )\n",
    "    )\n",
    "\n",
    "    bar_title = '' + str(num_bar_genes) + ' Most Common Genes in ' + library_name\n",
    "\n",
    "    if orient == 'Vertical':\n",
    "        bokbar = figure(x_range=top_genes['GENE'], plot_height=350, title=bar_title, toolbar_location=None, tools=\"hover\", tooltips='@top', x_axis_label='Genes', y_axis_label='Counts')\n",
    "        bokbar.vbar(x=top_genes['GENE'], top=top_genes['COUNT'], width=.5, color=bar_color, hover_alpha=.7)\n",
    "        bokbar.xaxis.major_label_orientation = math.pi/5\n",
    "        bokbar.xgrid.grid_line_color = None\n",
    "        bokbar.y_range.start = 0\n",
    "\n",
    "        if counts:\n",
    "            labels = LabelSet(x='x', y='y', text='y', level='annotation',\n",
    "                x_offset=-7, y_offset=0, source=barsource_v, render_mode='canvas', text_font_size = '11px')\n",
    "\n",
    "            bokbar.add_layout(labels)\n",
    "\n",
    "    if orient == 'Horizontal':\n",
    "        bokbar = figure(y_range = top_genes['GENE'][::-1], plot_height=400, title=bar_title, toolbar_location=None, tools=\"hover\", tooltips='@right', x_axis_label='Counts', y_axis_label='Genes')\n",
    "        bokbar.hbar(y=top_genes['GENE'][::-1],right=top_genes['COUNT'][::-1], height=.5, color=bar_color, hover_alpha=.7)\n",
    "        bokbar.xgrid.grid_line_color = None\n",
    "        \n",
    "        if counts:\n",
    "            labels = LabelSet(x='x', y='y', text='x', level='annotation',\n",
    "                x_offset=2, y_offset=-6, source=barsource_h, render_mode='canvas', text_font_size = '11px')\n",
    "\n",
    "            bokbar.add_layout(labels)\n",
    "\n",
    "    bokbar.xaxis.axis_label_text_font_style = 'normal'\n",
    "    bokbar.yaxis.axis_label_text_font_style = 'normal'\n",
    "    bokbar.title.align = 'center'\n",
    "\n",
    "    show(bokbar)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "my-first-appyter-2",
   "language": "python",
   "name": "my-first-appyter-2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
