{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "{% do SectionField(\n",
    "    name='PRIMARY',\n",
    "    title='Enrichr Consensus Terms',\n",
    "    subtitle='This appyter returns consensus Enrichr terms using a set of gene sets',\n",
    "    img='enrichr.png'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set title = StringField(\n",
    "    name='title',\n",
    "    label='Notebook name',\n",
    "    default='Enrichr Consensus Terms',\n",
    "    section=\"PRIMARY\",\n",
    ") %}\n",
    "\n",
    "title = {{ title }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "import requests\n",
    "import pandas as pd\n",
    "import json\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import display, IFrame, Markdown\n",
    "import math\n",
    "import scipy.stats as st\n",
    "import fastcluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(Markdown(\"# %s\"%(title)), display_id=\"title\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clustergrammer_url = 'https://amp.pharm.mssm.edu/clustergrammer/matrix_upload/'\n",
    "ENRICHR_URL = 'https://maayanlab.cloud/Enrichr'\n",
    "# libraries = [\"ChEA_2016\", \"GO_Biological_Process_2018\" ,\"GWAS_Catalog_2019\" , \"KEGG_2019_Human\"]\n",
    "table = 1\n",
    "figure = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clustergrammer(df, name, clustergrammer_url, display_id, fignum=1, label=\"Clustergrammer\"):\n",
    "    clustergram_df = df.rename(columns={i:\"Signature: %s\"%i for i in df.columns}, index={i:\"Enriched Term: %s\"%i for i in df.index})\n",
    "    clustergram_df.to_csv(name, sep=\"\\t\")\n",
    "    response = ''\n",
    "    for i in range(5):\n",
    "        try:\n",
    "            res = requests.post(clustergrammer_url, files={'file': open(name, 'rb')})\n",
    "            if not res.ok:\n",
    "                response = res.text\n",
    "                time.sleep(1)\n",
    "            else:\n",
    "                clustergrammer_url = res.text.replace(\"http:\",\"https:\")   \n",
    "                break\n",
    "        except Exception as e:\n",
    "            response = e\n",
    "            time.sleep(2)\n",
    "    else:\n",
    "        if type(response) == Exception:\n",
    "            raise response\n",
    "        else:\n",
    "            raise Exception(response)\n",
    "    display(IFrame(clustergrammer_url, width=\"1000\", height=\"1000\"), display_id=\"clustergram_%s\"%display_id)\n",
    "    display(Markdown(\"**Figure %d** %s [Go to url](%s)\"%(fignum, label, clustergrammer_url)), display_id=\"clustergram_label_%s\"%display_id )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap = sns.cubehelix_palette(50, hue=0.05, rot=0, light=1, dark=0)\n",
    "def heatmap(df, filename, display_id, width=15, height=15):\n",
    "#     fig = plt.figure(figsize=(width,height))\n",
    "    cg = sns.clustermap(df, cmap=cmap, figsize=(width, height), cbar_pos=(0.02, 0.65, 0.05, 0.18),)\n",
    "    cg.ax_row_dendrogram.set_visible(False)\n",
    "    cg.ax_col_dendrogram.set_visible(False)\n",
    "    display(cg, display_id=\"heatmap_%s\"%display_id)\n",
    "    plt.show()\n",
    "    cg.savefig(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dataframe(enrichment, lib, table, display_id):\n",
    "    term_df = pd.DataFrame(index=enrichment.keys())\n",
    "    for k,v in enrichment.items():\n",
    "        sigs = v[\"libraries\"][lib]\n",
    "        for sig in sigs:\n",
    "            term = sig[1]\n",
    "            if term not in term_df.columns:\n",
    "                term_df[term] = 0.0\n",
    "            p = sig[2]\n",
    "            term_df.at[k, term] = -math.log(p)\n",
    "    term_df = term_df.transpose()\n",
    "    term_df.to_csv(\"%s_enrichment.tsv\"%lib, sep=\"\\t\")\n",
    "    display(term_df.head(10), display_id=\"dataframe_%s\"%display_id)\n",
    "    display(Markdown(\"**Table %d** The table below shows the result of the enrichment analysis of %d gene sets \\\n",
    "        with the %s library in Enrichr. Each score is computed by getting the negative logarithm of the p-value \\\n",
    "        ($-\\ln{pval}$). [Download complete table](%s_enrichment.tsv)\"%(table, num_sigs, lib.replace(\"_\",\" \"), lib)),\n",
    "        display_id=\"dataframe_caption_%s\"%display_id\n",
    "    )\n",
    "    table+=1\n",
    "    return term_df, table\n",
    "\n",
    "def get_consensus(df, lib, top_results, table, display_id):\n",
    "    \n",
    "    consensus = df.sum(1).sort_values(ascending=False)[0:top_results].to_frame(name=\"scores\")\n",
    "\n",
    "    # Save to tsv\n",
    "    consensus.to_csv(\"%s_consensus.tsv\"%lib, sep=\"\\t\")\n",
    "    display(consensus.head(10), display_id=\"consensus_%s\"%display_id)\n",
    "    display(Markdown(\"**Table %d** %s Consensus terms. \\\n",
    "            Consensus scores are computed by taking the sum of scores in Table %d. \\\n",
    "            [Download top %d terms](%s_consensus.tsv)\"%(table, lib.replace(\"_\",\" \"), (table-1), top_results, lib)),\n",
    "            display_id=\"consensus_caption_%s\"%display_id\n",
    "    )\n",
    "    table+=1\n",
    "    return consensus, table\n",
    "\n",
    "def stackedBarPlot(df, filename, display_id, width = 15, height = 15):\n",
    "    df['mean'] = df.mean(axis=1)\n",
    "    df = df.sort_values(by = 'mean')[0:top_results]\\\n",
    "        .drop(['mean'], axis = 1)\n",
    "    if df.shape[0]==0:\n",
    "        return False\n",
    "    \n",
    "    plot = df.plot.barh(stacked = True, figsize = (width,height), fontsize = 20)\n",
    "    plt.legend(bbox_to_anchor=(1.7, 0), loc='lower right', prop={'size': 16})\n",
    "    plt.xlabel('-log(p)',labelpad = 20, fontsize = 'xx-large')\n",
    "    display(plot, display_id=\"stacked_%s\"%display_id)\n",
    "    plt.savefig(filename, format = 'svg', bbox_inches='tight')\n",
    "    plt.show()\n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enrichr Functions\n",
    "def addList(genes, description):\n",
    "    payload = {\n",
    "        'list': (None, '\\n'.join(genes)),\n",
    "        'description': (None, description)\n",
    "    }\n",
    "    res = requests.post(ENRICHR_URL + \"/addList\", files=payload)\n",
    "    time.sleep(1)\n",
    "    if not res.ok:\n",
    "        raise Exception('Error analyzing gene list')\n",
    "    data = res.json()\n",
    "    return data[\"userListId\"]\n",
    "\n",
    "def enrich(userListId, library, alpha):\n",
    "    res = requests.get(\n",
    "        ENRICHR_URL +\"/enrich\", params={\"userListId\": userListId, \"backgroundType\": library}\n",
    "     )\n",
    "    time.sleep(1)\n",
    "    if not res.ok:\n",
    "        raise Exception('Error fetching enrichment results')\n",
    "    data = res.json()\n",
    "    return [i for i in data[library] if i[2] < alpha]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "\n",
    "{% set input_gene_set = FileField(\n",
    "    name='input_gene_set',\n",
    "    label='Gene Set',\n",
    "    default='input.gmt',\n",
    "    section=\"PRIMARY\",\n",
    "    examples={\n",
    "        'input.gmt': 'https://appyters.maayanlab.cloud/storage/EnrichrConsensus/sample_input/10input.gmt'\n",
    "    }\n",
    ") %}\n",
    "\n",
    "input_gene_set = {{ input_gene_set }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "transcription_libraries = {{ MultiChoiceField(name='transcription_libraries',\n",
    "                                description='Select the Enrichr libraries you would like in your figure.',\n",
    "                                label='Transcription',\n",
    "                                default=[],\n",
    "                                section = 'PRIMARY',\n",
    "    choices=[\n",
    "    'ARCHS4_TFs_Coexp',\n",
    "    'ChEA_2016',\n",
    "    'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "    'ENCODE_Histone_Modifications_2015',\n",
    "    'ENCODE_TF_ChIP-seq_2015',\n",
    "    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "    'Genome_Browser_PWMs',\n",
    "    'lncHUB_lncRNA_Co-Expression',\n",
    "    'miRTarBase_2017',\n",
    "    'TargetScan_microRNA_2017',\n",
    "    'TF-LOF_Expression_from_GEO',\n",
    "    'TF_Perturbations_Followed_by_Expression',\n",
    "    'Transcription_Factor_PPIs',\n",
    "    'TRANSFAC_and_JASPAR_PWMs',\n",
    "    'TRRUST_Transcription_Factors_2019']) }}\n",
    "\n",
    "\n",
    "pathways_libraries = {{ MultiChoiceField(name='pathways_libraries',\n",
    "                                        description='Select the Enrichr libraries you would like in your figure.',\n",
    "                                        label='Pathways',\n",
    "                                        default=[],\n",
    "                                        section = 'PRIMARY',\n",
    "    choices=[\n",
    "    'ARCHS4_Kinases_Coexp',\n",
    "    'BioCarta_2016',\n",
    "    'BioPlanet_2019',\n",
    "    'BioPlex_2017',\n",
    "    'CORUM',\n",
    "    'Elsevier_Pathway_Collection',\n",
    "    'HMS_LINCS_KinomeScan',\n",
    "    'HumanCyc_2016',\n",
    "    'huMAP',\n",
    "    'KEA_2015',\n",
    "    'KEGG_2019_Human',\n",
    "    'KEGG_2019_Mouse',\n",
    "    'Kinase_Perturbations_from_GEO_down',\n",
    "    'Kinase_Perturbations_from_GEO_up',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "    'NCI-Nature_2016',\n",
    "    'NURSA_Human_Endogenous_Complexome',\n",
    "    'Panther_2016',\n",
    "    'Phosphatase_Substrates_from_DEPOD',\n",
    "    'PPI_Hub_Proteins',\n",
    "    'Reactome_2016',\n",
    "    'SILAC_Phosphoproteomics',\n",
    "    'SubCell_BarCode',\n",
    "    'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "    'WikiPathways_2019_Human',\n",
    "    'WikiPathways_2019_Mouse']) }}    \n",
    "    \n",
    "  \n",
    "ontologies_libraries = {{ MultiChoiceField(name='ontologies_libraries',\n",
    "                                        description='Select the Enrichr libraries you would like in your figure.',\n",
    "                                        label='Ontologies',\n",
    "                                        default=[],\n",
    "                                        section = 'PRIMARY',\n",
    "    choices=[\n",
    "    'GO_Biological_Process_2018',\n",
    "    'GO_Cellular_Component_2018',\n",
    "    'GO_Molecular_Function_2018',\n",
    "    'Human_Phenotype_Ontology',\n",
    "    'Jensen_COMPARTMENTS',\n",
    "    'Jensen_DISEASES',\n",
    "    'Jensen_TISSUES',\n",
    "    'MGI_Mammalian_Phenotype_Level_4_2019']) }} \n",
    "\n",
    "    \n",
    "diseases_drugs_libraries = {{ MultiChoiceField(name='diseases_drugs_libraries',\n",
    "                                        description='Select the Enrichr libraries you would like in your figure.',\n",
    "                                        label='Diseases/Drugs',\n",
    "                                        default=[],\n",
    "                                        section = 'PRIMARY',\n",
    "    choices=[    \n",
    "        'Achilles_fitness_decrease',\n",
    "        'Achilles_fitness_increase',\n",
    "        'ARCHS4_IDG_Coexp',\n",
    "        'ClinVar_2019',\n",
    "        'dbGaP',\n",
    "        'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "        'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "        'DisGeNET',\n",
    "        'DrugMatrix',\n",
    "        'DSigDB',\n",
    "        'GeneSigDB',\n",
    "        'GWAS_Catalog_2019',\n",
    "        'LINCS_L1000_Chem_Pert_down',\n",
    "        'LINCS_L1000_Chem_Pert_up',\n",
    "        'LINCS_L1000_Ligand_Perturbations_down',\n",
    "        'LINCS_L1000_Ligand_Perturbations_up',\n",
    "        'MSigDB_Computational',\n",
    "        'MSigDB_Oncogenic_Signatures',\n",
    "        'Old_CMAP_down',\n",
    "        'Old_CMAP_up',\n",
    "        'OMIM_Disease',\n",
    "        'OMIM_Expanded',\n",
    "        'PheWeb_2019',\n",
    "        'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "        'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "        'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "        'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "        'UK_Biobank_GWAS_v1',\n",
    "        'Virus_Perturbations_from_GEO_down',\n",
    "        'Virus_Perturbations_from_GEO_up',\n",
    "        'VirusMINT']) \n",
    "                            }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraries = transcription_libraries + pathways_libraries + ontologies_libraries + diseases_drugs_libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment = {}\n",
    "with open(input_gene_set) as o:\n",
    "    for line in o:\n",
    "        unpacked = line.strip().split(\"\\t\")\n",
    "        if len(unpacked) == 1:\n",
    "            raise ValueError(\"Line '%s' is either empty or not formatted properly. Please consult README for more information\"%line)\n",
    "        sigid = unpacked[0]\n",
    "        geneset = [i for i in unpacked[1:] if len(i) > 0]\n",
    "        enrichment[sigid] = {\n",
    "            \"genes\": [i.split(\",\")[0] for i in geneset]\n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_sigs = len(enrichment)\n",
    "input_sigs = pd.DataFrame.from_dict(enrichment, orient=\"index\")\n",
    "display(input_sigs.head(10))\n",
    "display(Markdown(\"**Table %d** Input Signatures\"%(table)), display_id=\"input_sigs\")\n",
    "table+=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User defined parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "alpha = {{FloatField(name='alpha', label='p-value cutoff', default=0.05, section='PRIMARY')}}\n",
    "top_results = {{IntField(name='min_count', label='Top results', description=\"Number of top results to keep\", default=25, section='PRIMARY')}}\n",
    "width = {{FloatField(name='width', label='image width', default=15, section='PRIMARY')}}\n",
    "height = {{FloatField(name='height', label='image height', default=15, section='PRIMARY')}}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "failed_userlist = []\n",
    "failed_enrich = {}\n",
    "for description, values in enrichment.items():\n",
    "    print(\"Querying %s\"%(description), end=\"\\r\", flush=True)\n",
    "    genes = values[\"genes\"]\n",
    "    for tries in range(5):\n",
    "        try:\n",
    "            userListId = addList(genes, description)\n",
    "            enrichment[description][\"userListId\"] = userListId\n",
    "            break\n",
    "        except Exception as e:\n",
    "            print(e)\n",
    "            time.sleep(0.5)\n",
    "    else:\n",
    "        failed_userlist.append(description)\n",
    "        continue\n",
    "    time.sleep(0.1)\n",
    "    enrichment[description][\"libraries\"] = {}\n",
    "\n",
    "    for library in libraries:\n",
    "        for tries in range(5):\n",
    "            try:\n",
    "                userlistId = enrichment[description][\"userListId\"]\n",
    "                results = enrich(userListId, library, alpha)\n",
    "                enrichment[description][\"libraries\"][library] = results\n",
    "                break\n",
    "            except Exception as e:\n",
    "                print(e)\n",
    "                time.sleep(0.5)\n",
    "        else:\n",
    "            if description not in failed_enrich:\n",
    "                failed_enrich[description] = []\n",
    "            failed_enrich[description].append(library)\n",
    "            continue\n",
    "        time.sleep(0.1)\n",
    "if len(failed_userlist):\n",
    "    print(\"Failed to add %d list\"%len(failed_userlist))\n",
    "if len(failed_enrich):\n",
    "    print(\"Failed enrichment for %d gene sets\"%len(failed_enrich))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for lib in libraries:\n",
    "    display(Markdown(\"## %s\"%lib.replace(\"_\",\" \")), display_id=\"title_%s\"%lib)\n",
    "    term_df,table = get_dataframe(enrichment, lib, table, display_id=lib)\n",
    "    consensus, table = get_consensus(term_df, lib, top_results, table, display_id=lib)\n",
    "    # Visualize\n",
    "    consensus_df = term_df.loc[consensus.index]\n",
    "    if (consensus_df.shape[1] > 0):\n",
    "        clustergram_filename = \"%s_consensus_clust.tsv\"%lib\n",
    "        clustergram_caption = \"Clustergrammer for the top %d consensus terms for %s \"%(top_results, lib.replace(\"_\",\" \"))\n",
    "        \n",
    "        clustergrammer(consensus_df,\n",
    "                       clustergram_filename,\n",
    "                       clustergrammer_url,\n",
    "                       lib,\n",
    "                       figure, \n",
    "                       clustergram_caption,\n",
    "                      )\n",
    "        figure+=1\n",
    "        results_count = len(consensus.index) if len(consensus.index) < top_results else top_results\n",
    "        heatmap(consensus_df, \"%s_consensus.svg\"%lib, lib, width, height)\n",
    "        display(Markdown(\"**Figure %d** Heatmap for the top %d consensus terms for %s. [Download figure](%s_consensus.svg)\"%(figure, results_count, lib.replace(\"_\",\" \"), lib)), \n",
    "                display_id=\"heatmap_caption_%s\"%lib)\n",
    "        figure+=1\n",
    "#         if num_sigs <=15:\n",
    "        status = stackedBarPlot(consensus_df, \"%s_consensus_barplot.svg\"%lib, display_id=lib)\n",
    "        if status:\n",
    "            display(Markdown(\"**Figure %d** Stacked bar plot for the top %d consensus terms for **%s**. [Download figure](%s_consensus_barplot.svg)\"%(figure, top_results, lib.replace(\"_\",\" \"), lib)), \n",
    "                    display_id=\"stacked_bar_caption_%s\"%lib)\n",
    "            figure +=1 \n",
    "            \n",
    "    else:\n",
    "        print(\"No terms found\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "[1] Chen EY, Tan CM, Kou Y, Duan Q, Wang Z, Meirelles GV, Clark NR, Ma'ayan A. Enrichr: interactive and collaborative HTML5 gene list enrichment analysis tool. BMC Bioinformatics. 2013;128(14).\n",
    "\n",
    "[2] Kuleshov MV, Jones MR, Rouillard AD, Fernandez NF, Duan Q, Wang Z, Koplev S, Jenkins SL, Jagodnik KM, Lachmann A, McDermott MG, Monteiro CD, Gundersen GW, Ma'ayan A. Enrichr: a comprehensive gene set enrichment analysis web server 2016 update. Nucleic Acids Research. 2016; gkw377.\n",
    "\n",
    "[3] Fernandez, N. F. et al. Clustergrammer, a web-based heatmap visualization and analysis tool for high-dimensional biological data. Sci. Data 4:170151 doi: 10.1038/sdata.2017.151 (2017)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter-default",
   "language": "python",
   "name": "jupyter-default"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
