{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichr Compressed Bar Chart Figure \n",
    "\n",
    "This appyter creates a publishable figure summarizing enrichment analysis results from Enrichr (https://amp.pharm.mssm.edu/Enrichr/). \n",
    "\n",
    "The resulting figure will contain the top 5 enriched terms and their corresponding p-values for each library chosen. Colored bars correspond to terms with significant p-values (<0.05). An asterisk (*) next to a p-value indicates the term also has a significant adjusted p-value (<0.05)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import json\n",
    "import requests\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import time\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from IPython.display import display,FileLink, Markdown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(name='section0', title = 'Publish Your Enrichr Results', subtitle = 'Create a publishable figure summarizing your Enrichr analysis results. The resulting figure will contain the top 5 enriched terms and their corresponding p-values for each library chosen. Colored bars correspond to terms with significant p-values (<0.05). An asterisk (*) next to a p-value indicates the term also has a significant adjusted p-value (<0.05).', img = 'screenshot.png')%}\n",
    "{% do SectionField(name='section1', title = '1. Submit Your Gene List', subtitle = 'Upload a text file containing your gene list -OR- copy and paste your gene list into the text box below (One gene per row).', img = 'screenshot.png')%}\n",
    "{% do SectionField(name='section2', title = '2. Choose Enrichr Libraries', subtitle = 'Select the Enrichr libraries you would like in your figure. Multiple libraries can be chosen from different categories.', img = 'screenshot.png')%}\n",
    "{% do SectionField(name='section3', title = '3. Output Options', subtitle = 'Choose a format and name to save your figure.', img = 'screenshot.png')%}\n",
    "{% do SectionField(name='section4', title = '4. Other Options', subtitle = 'Choose a color for your barchart. Additionally, enrichment terms may be too long for the figure and may need to be manually changed. You may upload a XLSX file with two columns containing original terms (left column) and your corresponding preferred shortened terms (right column).', img = 'screenshot.png')%}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "gene_list_filename = {{ FileField(name='gene_list_filename', label='Gene List File', default='', description='Upload your gene list as a text file (One gene per row).',section = 'section1') }}\n",
    "\n",
    "gene_list_input = {{ TextField(name='gene_list_input', label='Gene List', default='Nsun3\\nPolrmt\\nNlrx1\\nSfxn5', description='Paste your gene list (One gene per row).', section = 'section1') }}\n",
    "\n",
    "transcription_libraries = {{ MultiChoiceField(name='transcription_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Transcription', default=['ARCHS4_TFs_Coexp'], section = 'section2',choices=[\n",
    "    'ARCHS4_TFs_Coexp',\n",
    "    'ChEA_2016',\n",
    "    'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "    'ENCODE_Histone_Modifications_2015',\n",
    "    'ENCODE_TF_ChIP-seq_2015',\n",
    "    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "    'Genome_Browser_PWMs',\n",
    "    'lncHUB_lncRNA_Co-Expression',\n",
    "    'miRTarBase_2017',\n",
    "    'TargetScan_microRNA_2017',\n",
    "    'TF-LOF_Expression_from_GEO',\n",
    "    'TF_Perturbations_Followed_by_Expression',\n",
    "    'Transcription_Factor_PPIs',\n",
    "    'TRANSFAC_and_JASPAR_PWMs',\n",
    "    'TRRUST_Transcription_Factors_2019']) }}\n",
    "\n",
    "\n",
    "pathways_libraries = {{ MultiChoiceField(name='pathways_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Pathways', default=['ARCHS4_Kinases_Coexp'], section = 'section2',choices=[\n",
    "    'ARCHS4_Kinases_Coexp',\n",
    "    'BioCarta_2016',\n",
    "    'BioPlanet_2019',\n",
    "    'BioPlex_2017',\n",
    "    'CORUM',\n",
    "    'Elsevier_Pathway_Collection',\n",
    "    'HMS_LINCS_KinomeScan',\n",
    "    'HumanCyc_2016',\n",
    "    'huMAP',\n",
    "    'KEA_2015',\n",
    "    'KEGG_2019_Human',\n",
    "    'KEGG_2019_Mouse',\n",
    "    'Kinase_Perturbations_from_GEO_down',\n",
    "    'Kinase_Perturbations_from_GEO_up',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "    'NCI-Nature_2016',\n",
    "    'NURSA_Human_Endogenous_Complexome',\n",
    "    'Panther_2016',\n",
    "    'Phosphatase_Substrates_from_DEPOD',\n",
    "    'PPI_Hub_Proteins',\n",
    "    'Reactome_2016',\n",
    "    'SILAC_Phosphoproteomics',\n",
    "    'SubCell_BarCode',\n",
    "    'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "    'WikiPathways_2019_Human',\n",
    "    'WikiPathways_2019_Mouse']) }}    \n",
    "    \n",
    "  \n",
    "ontologies_libraries = {{ MultiChoiceField(name='ontologies_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Ontologies', default=['GO_Biological_Process_2018'], section = 'section2',choices=[\n",
    "    'GO_Biological_Process_2018',\n",
    "    'GO_Cellular_Component_2018',\n",
    "    'GO_Molecular_Function_2018',\n",
    "    'Human_Phenotype_Ontology',\n",
    "    'Jensen_COMPARTMENTS',\n",
    "    'Jensen_DISEASES',\n",
    "    'Jensen_TISSUES',\n",
    "    'MGI_Mammalian_Phenotype_Level_4_2019']) }} \n",
    "\n",
    "    \n",
    "diseases_drugs_libraries = {{ MultiChoiceField(name='diseases_drugs_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Diseases/Drugs', default=['Achilles_fitness_decrease'], section = 'section2',choices=[    \n",
    "    'Achilles_fitness_decrease',\n",
    "    'Achilles_fitness_increase',\n",
    "    'ARCHS4_IDG_Coexp',\n",
    "    'ClinVar_2019',\n",
    "    'dbGaP',\n",
    "    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "    'DisGeNET',\n",
    "    'DrugMatrix',\n",
    "    'DSigDB',\n",
    "    'GeneSigDB',\n",
    "    'GWAS_Catalog_2019',\n",
    "    'LINCS_L1000_Chem_Pert_down',\n",
    "    'LINCS_L1000_Chem_Pert_up',\n",
    "    'LINCS_L1000_Ligand_Perturbations_down',\n",
    "    'LINCS_L1000_Ligand_Perturbations_up',\n",
    "    'MSigDB_Computational',\n",
    "    'MSigDB_Oncogenic_Signatures',\n",
    "    'Old_CMAP_down',\n",
    "    'Old_CMAP_up',\n",
    "    'OMIM_Disease',\n",
    "    'OMIM_Expanded',\n",
    "    'PheWeb_2019',\n",
    "    'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "    'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "    'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "    'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "    'UK_Biobank_GWAS_v1',\n",
    "    'Virus_Perturbations_from_GEO_down',\n",
    "    'Virus_Perturbations_from_GEO_up',\n",
    "    'VirusMINT']) }}\n",
    "    \n",
    "    \n",
    "cell_types_libraries = {{ MultiChoiceField(name='cell_types_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Cell Types', default=['Allen_Brain_Atlas_down'], section = 'section2',choices=[        \n",
    "    'Allen_Brain_Atlas_down',\n",
    "    'Allen_Brain_Atlas_up',\n",
    "    'ARCHS4_Cell-lines',\n",
    "    'ARCHS4_Tissues',\n",
    "    'Cancer_Cell_Line_Encyclopedia',\n",
    "    'CCLE_Proteomics_2020',\n",
    "    'ESCAPE',\n",
    "    'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "    'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "    'Human_Gene_Atlas',\n",
    "    'Mouse_Gene_Atlas',\n",
    "    'NCI-60_Cancer_Cell_Lines',\n",
    "    'ProteomicsDB_2020',\n",
    "    'Tissue_Protein_Expression_from_Human_Proteome_Map']) }}    \n",
    "    \n",
    "    \n",
    "    \n",
    "miscellaneous_libraries = {{ MultiChoiceField(name='miscellaneous_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Miscellaneous', default=['Chromosome_Location_hg19'], section = 'section2',choices=[            \n",
    "    'Chromosome_Location_hg19',\n",
    "    'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "    'Enrichr_Libraries_Most_Popular_Genes',\n",
    "    'Genes_Associated_with_NIH_Grants',\n",
    "    'HMDB_Metabolites',\n",
    "    'HomoloGene',\n",
    "    'InterPro_Domains_2019',\n",
    "    'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "    'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "    'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "    'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "    'Pfam_Domains_2019',\n",
    "    'Pfam_InterPro_Domains',\n",
    "    'Table_Mining_of_CRISPR_Studies']) }}    \n",
    "    \n",
    "    \n",
    "legacy_libraries = {{ MultiChoiceField(name='legacy_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Legacy', default=['BioCarta_2013'], section = 'section2',choices=[                \n",
    "    'BioCarta_2013',\n",
    "    'BioCarta_2015',\n",
    "    'ChEA_2013',\n",
    "    'ChEA_2015',\n",
    "    'Chromosome_Location',\n",
    "    'Disease_Signatures_from_GEO_down_2014',\n",
    "    'Disease_Signatures_from_GEO_up_2014',\n",
    "    'Drug_Perturbations_from_GEO_2014',\n",
    "    'ENCODE_Histone_Modifications_2013',\n",
    "    'ENCODE_TF_ChIP-seq_2014',\n",
    "    'GO_Biological_Process_2013',\n",
    "    'GO_Biological_Process_2015',\n",
    "    'GO_Biological_Process_2017',\n",
    "    'GO_Biological_Process_2017b',\n",
    "    'GO_Cellular_Component_2013',\n",
    "    'GO_Cellular_Component_2015',\n",
    "    'GO_Cellular_Component_2017',\n",
    "    'GO_Cellular_Component_2017b',\n",
    "    'GO_Molecular_Function_2013',\n",
    "    'GO_Molecular_Function_2015',\n",
    "    'GO_Molecular_Function_2017',\n",
    "    'GO_Molecular_Function_2017b',\n",
    "    'HumanCyc_2015',\n",
    "    'KEA_2013',\n",
    "    'KEGG_2013',\n",
    "    'KEGG_2015',\n",
    "    'KEGG_2016',\n",
    "    'MGI_Mammalian_Phenotype_2013',\n",
    "    'MGI_Mammalian_Phenotype_2017',\n",
    "    'MGI_Mammalian_Phenotype_Level_3',\n",
    "    'MGI_Mammalian_Phenotype_Level_4',\n",
    "    'NCI-Nature_2015',\n",
    "    'Panther_2015',\n",
    "    'Reactome_2013',\n",
    "    'Reactome_2015',\n",
    "    'TargetScan_microRNA',\n",
    "    'Tissue_Protein_Expression_from_ProteomicsDB',\n",
    "    'WikiPathways_2013',\n",
    "    'WikiPathways_2015',\n",
    "    'WikiPathways_2016']) }} \n",
    "\n",
    "crowd_libraries = {{ MultiChoiceField(name='crowd_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Crowd', default=['Aging_Perturbations_from_GEO_down'], section = 'section2',choices=[                \n",
    "    'Aging_Perturbations_from_GEO_down',\n",
    "    'Aging_Perturbations_from_GEO_up',\n",
    "    'Disease_Perturbations_from_GEO_down',\n",
    "    'Disease_Perturbations_from_GEO_up',\n",
    "    'Drug_Perturbations_from_GEO_down',\n",
    "    'Drug_Perturbations_from_GEO_up',\n",
    "    'Gene_Perturbations_from_GEO_down',\n",
    "    'Gene_Perturbations_from_GEO_up',\n",
    "    'Ligand_Perturbations_from_GEO_down',\n",
    "    'Ligand_Perturbations_from_GEO_up',\n",
    "    'MCF7_Perturbations_from_GEO_down',\n",
    "    'MCF7_Perturbations_from_GEO_up',\n",
    "    'Microbe_Perturbations_from_GEO_down',\n",
    "    'Microbe_Perturbations_from_GEO_up',\n",
    "    'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "    'SysMyo_Muscle_Gene_Sets']) }}\n",
    "\n",
    "\n",
    "figure_file_format = {{ MultiChoiceField(name='figure_file_format', label='File Format', choices = ['png', 'pdf', 'svg' ], default=['png'], description='Select the format(s) to save your figure.', section = 'section3') }}\n",
    "\n",
    "output_file_name = {{ StringField(name='output_file_name', label='File Name', default='Enrichr_results', description='Enter a name/description to save your figure.', section = 'section3') }}\n",
    "\n",
    "\n",
    "color = \"{{ ChoiceField(name='color', description='Select a color for your barchart.', label='Color', default='Red', section = 'section4',choices=[                \n",
    "    'Blue',\n",
    "    'Green',\n",
    "    'Grey',\n",
    "    'Orange',\n",
    "    'Purple',\n",
    "    'Red',\n",
    "    'Yellow']) }}\"\n",
    "\n",
    "annot_dict_filename = {{ FileField(name='annot_dict_filename', label='Term Conversions (Optional)', default='', description='Enrichment terms may be too long for the figure and may need to be manually changed. You may upload a XLSX file with two columns containing original terms (left column) and your corresponding preferred shortened terms (right column).', section = 'section4') }}\n",
    "\n",
    "final_output_file_names = [str(output_file_name+'.'+file_type) for file_type in figure_file_format]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import gene list as file or from text box file\n",
    "# Will choose file upload over textbox if a file is given \n",
    "if gene_list_filename != '':\n",
    "    open_gene_list_file = open(gene_list_filename,'r')\n",
    "    lines = open_gene_list_file.readlines()\n",
    "    genes = [x.strip() for x in lines]\n",
    "    open_gene_list_file.close()\n",
    "else:\n",
    "    genes = gene_list_input.split('\\n')\n",
    "    genes = [x.strip() for x in genes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import term conversion file (Optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Conversion dictionary for long terms\n",
    "annot_dict = {}\n",
    "\n",
    "# Import term conversion file if one is specified \n",
    "if annot_dict_filename != '':\n",
    "    conversion_file = pd.read_excel(annot_dict_filename, 'Sheet1', header=None)\n",
    "    annot_dict = dict(zip([x.strip() for x in conversion_file[0]], [x.strip() for x in conversion_file[1]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map bar color choice "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Color for barchart \n",
    "color_conversion={\n",
    "    'Red':'tomato',\n",
    "    'Blue':'lightskyblue',\n",
    "    'Green':'mediumspringgreen',\n",
    "    'Grey':'lightgrey',\n",
    "    'Orange':'orange',\n",
    "    'Purple':'plum',\n",
    "    'Yellow':'yellow',\n",
    "    'Pink':'lightpink'\n",
    "}\n",
    "\n",
    "color = color_conversion[color]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get Enrichr Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine libraries from all categories \n",
    "enrichr_libraries = np.sort(transcription_libraries+pathways_libraries+ontologies_libraries+diseases_drugs_libraries+cell_types_libraries+miscellaneous_libraries+legacy_libraries+crowd_libraries)\n",
    "\n",
    "# Function to get Enrichr Results \n",
    "# Takes a gene list and Enrichr libraries as input \n",
    "def Enrichr_API(enrichr_gene_list, all_libraries):\n",
    "\n",
    "\n",
    "    all_terms = []\n",
    "    all_pvalues =[] \n",
    "    all_adjusted_pvalues = []\n",
    "    library_success = []\n",
    "    short_id = ''\n",
    "\n",
    "    for library_name in all_libraries : \n",
    "        ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/addList'\n",
    "        genes_str = '\\n'.join(enrichr_gene_list)\n",
    "        description = 'Example gene list'\n",
    "        payload = {\n",
    "            'list': (None, genes_str),\n",
    "            'description': (None, description)\n",
    "        }\n",
    "\n",
    "        response = requests.post(ENRICHR_URL, files=payload)\n",
    "        if not response.ok:\n",
    "            raise Exception('Error analyzing gene list')\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "        time.sleep(0.5)\n",
    "        ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/enrich'\n",
    "        query_string = '?userListId=%s&backgroundType=%s'\n",
    "        user_list_id = data['userListId']\n",
    "        short_id = data[\"shortId\"]\n",
    "        gene_set_library = library_name\n",
    "        response = requests.get(\n",
    "            ENRICHR_URL + query_string % (user_list_id, gene_set_library)\n",
    "         )\n",
    "        if not response.ok:\n",
    "            raise Exception('Error fetching enrichment results')\n",
    "        try:\n",
    "            data = json.loads(response.text)\n",
    "            results_df  = pd.DataFrame(data[library_name][0:5])\n",
    "            all_terms.append(list(results_df[1]))\n",
    "            all_pvalues.append(list(results_df[2]))\n",
    "            all_adjusted_pvalues.append(list(results_df[6]))\n",
    "            library_success.append(library_name)\n",
    "        except:\n",
    "            print('Error for ' + library_name + ' library')\n",
    "\n",
    "    return([all_terms,all_pvalues,all_adjusted_pvalues,str(short_id),library_success])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get enrichr results\n",
    "results = Enrichr_API(genes,enrichr_libraries)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Enrichr Results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function plots results \n",
    "# Takes all terms, all p-values, all adjusted p-values, plot title, Enrichr libraries , and specified figure format\n",
    "def enrichr_figure(all_terms,all_pvalues, all_adjusted_pvalues, plot_names, all_libraries, fig_format, bar_color): \n",
    "    \n",
    "    # rows and columns depend on number of Enrichr libraries submitted \n",
    "    rows = []\n",
    "    cols = []\n",
    "    \n",
    "    # Bar colors\n",
    "    if bar_color!= 'lightgrey':\n",
    "        bar_color_not_sig = 'lightgrey'\n",
    "        edgecolor=None\n",
    "        linewidth=0\n",
    "    else:\n",
    "        bar_color_not_sig = 'white'\n",
    "        edgecolor='black'\n",
    "        linewidth=1\n",
    "    \n",
    "    # If only 1 Enrichr library selected, make simple plot \n",
    "    if len(all_libraries)==1:\n",
    "        #fig,axes = plt.subplots(1, 1,figsize=[8.5,6])\n",
    "        plt.figure(figsize=(12,6))\n",
    "        rows = [0]\n",
    "        cols = [0]\n",
    "        i = 0 \n",
    "        bar_colors = [bar_color if (x < 0.05) else bar_color_not_sig for x in all_pvalues[i]]\n",
    "        fig = sns.barplot(x=np.log10(all_pvalues[i])*-1, y=all_terms[i], palette=bar_colors, edgecolor=edgecolor, linewidth=linewidth)\n",
    "        fig.axes.get_yaxis().set_visible(False)\n",
    "        fig.set_title(all_libraries[i].replace('_',' '),fontsize=26)\n",
    "        fig.set_xlabel('-Log10(p-value)',fontsize=25)\n",
    "        fig.xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "        fig.tick_params(axis='x', which='major', labelsize=20)\n",
    "        if max(np.log10(all_pvalues[i])*-1)<1:\n",
    "            fig.xaxis.set_ticks(np.arange(0, max(np.log10(all_pvalues[i])*-1), 0.1))\n",
    "        for ii,annot in enumerate(all_terms[i]):\n",
    "            if annot in annot_dict.keys():\n",
    "                annot = annot_dict[annot]\n",
    "            if all_adjusted_pvalues[i][ii] < 0.05:\n",
    "                annot = '  *'.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))]) \n",
    "            else:\n",
    "                annot = '  '.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))])\n",
    "\n",
    "            title_start= max(fig.axes.get_xlim())/200\n",
    "            fig.text(title_start,ii,annot,ha='left',wrap = True, fontsize = 26)\n",
    "            fig.patch.set_edgecolor('black')  \n",
    "            fig.patch.set_linewidth('2')\n",
    "        \n",
    "    \n",
    "    # If there are an even number of Enrichr libraries below 6\n",
    "    # Plots 1x2 or 2x2\n",
    "    else:\n",
    "        if len(all_libraries) % 2 == 0 and len(all_libraries) < 5:\n",
    "                for i in range(0,int(len(all_libraries)/2)):    \n",
    "                    rows = rows + [i]*2\n",
    "                    cols = list(range(0,2))*int(len(all_libraries)/2)    \n",
    "                fig, axes = plt.subplots(len(np.unique(rows)), len(np.unique(cols)),figsize=[7,int(2* len(np.unique(rows)))]) \n",
    "    \n",
    "        \n",
    "        # All other # of libraries 6 and above will have 3 columns and a flexible number of rows to accomodate all plots\n",
    "        else:\n",
    "            for i in range(0,int(np.ceil(len(all_libraries)/3))):\n",
    "                rows = rows + [i]*3\n",
    "                cols = list(range(0,3))*int(np.ceil(len(all_libraries)/3))\n",
    "            fig, axes = plt.subplots(len(np.unique(rows)), len(np.unique(cols)),figsize=[8,int(2* len(np.unique(rows)))])\n",
    "           \n",
    "        # If final figure only has one row...\n",
    "        if len(np.unique(rows))==1:\n",
    "            for i,library_name in enumerate(all_libraries):\n",
    "                bar_colors = [bar_color if (x < 0.05) else bar_color_not_sig for x in all_pvalues[i]]\n",
    "                sns.barplot(x=np.log10(all_pvalues[i])*-1, y=all_terms[i],ax=axes[i], palette=bar_colors, edgecolor=edgecolor, linewidth=linewidth)\n",
    "                axes[i].axes.get_yaxis().set_visible(False)\n",
    "                axes[i].set_title(library_name.replace('_',' '),fontsize=36)\n",
    "                axes[i].set_xlabel('-Log10(p-value)',fontsize=35)\n",
    "                axes[i].xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "                axes[i].tick_params(axis='x', which='major', labelsize=30)\n",
    "                if max(np.log10(all_pvalues[i])*-1)<1:\n",
    "                    axes[i].xaxis.set_ticks(np.arange(0, max(np.log10(all_pvalues[i])*-1), 0.1))\n",
    "                for ii,annot in enumerate(all_terms[i]):\n",
    "                    if annot in annot_dict.keys():\n",
    "                        annot = annot_dict[annot]\n",
    "                    if all_adjusted_pvalues[i][ii] < 0.05:\n",
    "                        annot = '  *'.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))]) \n",
    "                    else:\n",
    "                        annot = '  '.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))])\n",
    "\n",
    "                    title_start= max(axes[i].axes.get_xlim())/200\n",
    "                    axes[i].text(title_start,ii,annot,ha='left',wrap = True, fontsize = 36)\n",
    "                    axes[i].patch.set_edgecolor('black')  \n",
    "                    axes[i].patch.set_linewidth('2')\n",
    "\n",
    "            plt.subplots_adjust(top=4.5, right = 4.7,wspace = 0.03,hspace = 0.2)\n",
    "\n",
    "\n",
    "        # If the final figure has more than one row...\n",
    "        else:\n",
    "\n",
    "\n",
    "            for i,library_name in enumerate(all_libraries):\n",
    "                bar_colors = [bar_color if (x < 0.05) else bar_color_not_sig for x in all_pvalues[i]]\n",
    "                sns.barplot(x=np.log10(all_pvalues[i])*-1, y=all_terms[i],ax=axes[rows[i],cols[i]], palette=bar_colors, edgecolor=edgecolor, linewidth=linewidth)\n",
    "                axes[rows[i],cols[i]].axes.get_yaxis().set_visible(False)\n",
    "                axes[rows[i],cols[i]].set_title(library_name.replace('_',' '),fontsize=36)\n",
    "                axes[rows[i],cols[i]].set_xlabel('-Log10(p-value)',fontsize=35)\n",
    "                axes[rows[i],cols[i]].xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "                axes[rows[i],cols[i]].tick_params(axis='x', which='major', labelsize=30)\n",
    "                if max(np.log10(all_pvalues[i])*-1)<1:\n",
    "                    axes[rows[i],cols[i]].xaxis.set_ticks(np.arange(0, max(np.log10(all_pvalues[i])*-1), 0.1))\n",
    "                for ii,annot in enumerate(all_terms[i]):\n",
    "                    if annot in annot_dict.keys():\n",
    "                        annot = annot_dict[annot]\n",
    "                    if all_adjusted_pvalues[i][ii] < 0.05:\n",
    "                        annot = '  *'.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))]) \n",
    "                    else:\n",
    "                        annot = '  '.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))])\n",
    "\n",
    "                    title_start= max(axes[rows[i],cols[i]].axes.get_xlim())/200\n",
    "                    axes[rows[i],cols[i]].text(title_start,ii,annot,ha='left',wrap = True, fontsize = 36)\n",
    "                    axes[rows[i],cols[i]].patch.set_edgecolor('black')  \n",
    "                    axes[rows[i],cols[i]].patch.set_linewidth('2')\n",
    "\n",
    "            plt.subplots_adjust(top=4.8, right = 4.7,wspace = 0.03,hspace = 0.2)\n",
    "\n",
    "\n",
    "        # If >6 libraries are chosen and is not a multiple of 3, delete empty plots\n",
    "        if len(np.unique(rows))*len(np.unique(cols)) != len(all_libraries):\n",
    "            diff = (len(np.unique(rows))*len(np.unique(cols))) - len(all_libraries)\n",
    "            for i in range (1,int(diff+1)):\n",
    "                fig.delaxes(axes[rows[-i]][cols[-i]])\n",
    "    \n",
    "    # Save results \n",
    "    for plot_name in plot_names:\n",
    "        plt.savefig(plot_name,bbox_inches = 'tight')\n",
    "    \n",
    "    # Show plot \n",
    "    plt.show()\n",
    "    \n",
    "    \n",
    "               "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichr_figure(results[0],results[1],results[2],final_output_file_names, results[4],figure_file_format, color)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download bar graphs \n",
    "for i, file in enumerate(final_output_file_names):\n",
    "    display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get complete enrichment analysis results from Enrichr \n",
    "display(Markdown('Access your complete Enrichment results here: ' + str('https://amp.pharm.mssm.edu/Enrichr/enrich?dataset='+ results[3])))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
