{"cells":[{"cell_type":"markdown","metadata":{},"source":[" # Harmonizome ETL: The Cancer Gene Atlas (TCGA)"]},{"cell_type":"markdown","metadata":{},"source":[" Created by: Charles Dai <br>\n"," Credit to: Moshe Silverstein\n","\n"," Data Source: https://gdc.cancer.gov/"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[]},"outputs":[],"source":["# appyter init\n","from appyter import magic\n","magic.init(lambda _=globals: _())"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["import sys\n","import os\n","from datetime import date\n","import gzip\n","import io\n","import requests\n","import json\n","\n","import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","%matplotlib inline\n","from tqdm import tqdm\n","\n","import harmonizome.utility_functions as uf\n","import harmonizome.lookup as lookup"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%load_ext autoreload\n","%autoreload 2"]},{"cell_type":"markdown","metadata":{},"source":[" ### Notebook Information"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[]},"outputs":[],"source":["print('This notebook was run on:', date.today(), '\\nPython version:', sys.version)"]},{"cell_type":"markdown","metadata":{},"source":[" # Initialization"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%appyter hide_code\n","\n","{% do SectionField(\n","    name='settings',\n","    title='Settings',\n","    img='setting_icon.png'\n",") %}"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[]},"outputs":[],"source":["%%appyter code_eval\n","\n","{% do DescriptionField(\n","    name='description',\n","    text='The files used for analysis are downloaded from the NIH Genomic Data Commons Data Portal using the GDC API from <a href=\"https://gdc.cancer.gov/\" target=\"_blank\">gdc.cancer.gov</a>.',\n","    section='settings'\n",") %}\n","\n","{% set project_name = ChoiceField(\n","    name='project_name',\n","    label='TCGA Project Name',\n","    choices=[\n","        'Acute Myeloid Leukemia',\n","        'Adrenocortical Carcinoma',\n","        'Bladder Urothelial Carcinoma',\n","        'Brain Lower Grade Glioma',\n","        'Breast Invasive Carcinoma',\n","        'Cervical Squamous Cell Carcinoma and Endocervical Adenocarcinoma',\n","        'Cholangiocarcinoma',\n","        'Colon Adenocarcinoma',\n","        'Esophageal Carcinoma',\n","        'Glioblastoma Multiforme',\n","        'Head and Neck Squamous Cell Carcinoma',\n","        'Kidney Chromophobe',\n","        'Kidney Renal Clear Cell Carcinoma',\n","        'Kidney Renal Papillary Cell Carcinoma',\n","        'Liver Hepatocellular Carcinoma',\n","        'Lung Adenocarcinoma',\n","        'Lung Squamous Cell Carcinoma',\n","        'Lymphoid Neoplasm Diffuse Large B-cell Lymphoma',\n","        'Mesothelioma',\n","        'Ovarian Serous Cystadenocarcinoma',\n","        'Pancreatic Adenocarcinoma',\n","        'Pheochromocytoma and Paraganglioma',\n","        'Prostate Adenocarcinoma',\n","        'Rectum Adenocarcinoma',\n","        'Sarcoma',\n","        'Skin Cutaneous Melanoma',\n","        'Stomach Adenocarcinoma',\n","        'Testicular Germ Cell Tumors',\n","        'Thymoma',\n","        'Thyroid Carcinoma',\n","        'Uterine Carcinosarcoma',\n","        'Uterine Corpus Endometrial Carcinoma',\n","        'Uveal Melanoma'\n","    ],\n","    default='Acute Myeloid Leukemia',\n","    section='settings'\n",") %}"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["base_url = 'https://api.gdc.cancer.gov/'\n","files_endpt = base_url + 'files/'\n","cases_endpt = base_url + 'cases/'\n","data_endpt = base_url + 'data/'"]},{"cell_type":"markdown","metadata":{},"source":[" ### Load Mapping Dictionaries"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[]},"outputs":[],"source":["symbol_lookup, geneid_lookup = lookup.get_lookups(mapfrom='ensembl')"]},{"cell_type":"markdown","metadata":{},"source":[" ### Output Path"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["output_name = 'tcga'\n","\n","path = 'Output/TCGA'\n","if not os.path.exists(path):\n","    os.makedirs(path)"]},{"cell_type":"markdown","metadata":{},"source":[" # Load Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Fetch Data From NCI"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[]},"outputs":[],"source":["%%appyter code_exec\n","\n","fields = [\n","    'cases.case_id'\n","]\n","\n","filters = {\n","    'op': 'and',\n","    'content': [{\n","        'op': 'in',\n","        'content': {\n","            'field': 'experimental_strategy',\n","            'value': ['RNA-Seq'],\n","        }\n","    }, \n","    {\n","        'op': 'in',\n","        'content': {\n","            'field': 'access',\n","            'value': ['open'],\n","        }\n","    },\n","    {\n","        'op': 'in',\n","        'content': {\n","            'field': 'file_name',\n","            'value': ['*htseq.counts.gz'],\n","        }\n","    },\n","    {\n","        'op': 'in',\n","        'content': {\n","            'field': 'cases.project.name',\n","            'value': ['{{project_name}}'],\n","        }\n","    }\n","    ],\n","}"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["params = {\n","    'fields': ','.join(fields),\n","    'filters': json.dumps(filters),\n","    'size': 100000,\n","    'facets': 'cases.case_id'\n","}\n","response = requests.get('https://api.gdc.cancer.gov/files', params=params)\n","data = response.json()['data']['hits']"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[]},"outputs":[],"source":["files = pd.DataFrame([(f['id'], f['cases'][0]['case_id']) for f in data], columns=['file_id', 'case_id']).set_index('file_id')\n","print(files.shape)\n","files.head()"]},{"cell_type":"markdown","metadata":{},"source":["## Construct Expression Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[]},"outputs":[],"source":["matrix = pd.DataFrame()\n","\n","for file_id in tqdm(files.index, unit='samples'):\n","    response = requests.get(data_endpt + file_id, headers = {\"Content-Type\": \"application/json\"})\n","    string_data = io.StringIO(str(gzip.decompress(response.content), 'utf-8'))\n","    matrix = pd.concat([matrix, pd.read_csv(string_data, sep='\\t', header=None, names=['ENSMBL ID', files.loc[file_id, 'case_id']], index_col=0)], axis=1)\n","matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.index = matrix.index.map(lambda x: x.split('.')[0])\n","matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":["## Construct Sample Meta"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["cases_fields = requests.get(cases_endpt + '_mapping').json()['fields']\n","keyfields = [field for field in cases_fields if \n","    any(word in field for word in ['demographic', 'diagnoses']) and 'treatment' not in field]"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[]},"outputs":[],"source":["sample_meta = pd.DataFrame()\n","for case_id in tqdm(files['case_id'].drop_duplicates(), unit='cases'):\n","    response = requests.get(cases_endpt + case_id, params={'fields': ','.join(keyfields)})\n","    data = response.json()['data']\n","    sample = pd.DataFrame([{'case_id': case_id, **data['demographic'], **data['diagnoses'][0]}])\n","    sample_meta = pd.concat([sample_meta, sample])\n","sample_meta = sample_meta.set_index('case_id')\n","sample_meta.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["sample_meta.shape"]},{"cell_type":"markdown","metadata":{},"source":[" # Pre-process Data"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.index.name = 'Gene Symbol'\n","matrix.columns.name = 'Case'\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Save Unfiltered Matrix to file"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_data(matrix, path, output_name + '_matrix_unfiltered',\n","            compression='gzip', dtype=np.float32)"]},{"cell_type":"markdown","metadata":{},"source":[" # Filter Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Map Gene Symbols to Up-to-date Approved Gene Symbols"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[]},"outputs":[],"source":["matrix = uf.map_symbols(matrix, symbol_lookup)\n","matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Merge Duplicate Genes By Rows and Duplicate Columns"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.merge(matrix, 'row')\n","matrix = uf.merge(matrix, 'column')\n","matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Remove Genes that are More Than 95% Missing or Zero Inference"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = matrix.replace(0, np.nan).dropna(\n","    thresh=0.05 * matrix.shape[1], axis=0).replace(np.nan, 0)\n","matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Normalize Matrix (Quantile Normalize the Matrix by Column)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.quantile_normalize(matrix)\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Normalize Matrix (Z-Score the Rows)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.zscore(matrix)\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Histogram of First Sample"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.iloc[:, 0].hist(bins=100)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Histogram of First Gene"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.iloc[0, :].hist(bins=100)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Save Filtered Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_data(matrix, path, output_name + '_matrix_filtered', \n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" # Analyze Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene List"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["gene_list = uf.gene_list(matrix, geneid_lookup)\n","gene_list.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["gene_list.shape"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_data(gene_list, path, output_name + '_gene_list',\n","            ext='tsv', compression='gzip', index=False)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Attribute List"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["attribute_list = uf.attribute_list(matrix, sample_meta)\n","attribute_list.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["attribute_list.shape"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_data(attribute_list, path, output_name + '_attribute_list',\n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create matrix of Standardized values (values between -1, and 1)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["standard_matrix = uf.standardized_matrix(matrix)\n","standard_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_data(standard_matrix, path, output_name + '_standard_matrix',\n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Plot of A Single Celltype, Normalized Value vs. Standardized Value"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["plt.plot(matrix[matrix.columns[0]],\n","         standard_matrix[standard_matrix.columns[0]], 'bo')\n","plt.xlabel('Normalized Values')\n","plt.ylabel('Standardized Values')\n","plt.title(standard_matrix.columns[0])\n","plt.grid(True)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Ternary Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["ternary_matrix = uf.ternary_matrix(standard_matrix)\n","ternary_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_data(ternary_matrix, path, output_name + '_ternary_matrix',\n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene and Attribute Set Libraries"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_setlib(ternary_matrix, 'gene', 'up', path, output_name + '_gene_up_set')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_setlib(ternary_matrix, 'gene', 'down', path, output_name + '_gene_down_set')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_setlib(ternary_matrix, 'attribute', 'up', path, \n","                           output_name + '_attribute_up_set')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_setlib(ternary_matrix, 'attribute', 'down', path, \n","                             output_name + '_attribute_down_set')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Attribute Similarity Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["attribute_similarity_matrix = uf.similarity_matrix(standard_matrix.T, 'cosine')\n","attribute_similarity_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_data(attribute_similarity_matrix, path,\n","            output_name + '_attribute_similarity_matrix', \n","            compression='npz', symmetric=True, dtype=np.float32)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene Similarity Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["gene_similarity_matrix = uf.similarity_matrix(standard_matrix, 'cosine')\n","gene_similarity_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.save_data(gene_similarity_matrix, path, \n","            output_name + '_gene_similarity_matrix',\n","            compression='npz', symmetric=True, dtype=np.float32)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene-Attribute Edge List"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["edge_list = uf.edge_list(standard_matrix)\n","uf.save_data(edge_list, path, output_name + '_edge_list', \n","        ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" # Create Downloadable Save File"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.archive(path)"]},{"cell_type":"markdown","metadata":{},"source":[" ### Link to download output files: [click here](./output_archive.zip)"]}],"metadata":{"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":3},"orig_nbformat":2,"kernelspec":{"name":"python_defaultSpec_1594072879307","display_name":"Python 3.8.0 64-bit ('harmonizome': venv)"}},"nbformat":4,"nbformat_minor":2}