{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bulk RNA-seq Analysis pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This pipeline enables you to analyze and visualize your bulk RNA sequencing datasets with an array of downstream analysis and visualization tools. The pipeline includes: PCA/UMAP/t-SNE analysis, Clustergrammer interactive heatmap, library size analysis, differential gene expression analysis, enrichment analysis, and L1000 small molecule search."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic libraries\n",
    "import pandas as pd\n",
    "import os\n",
    "import random\n",
    "import time\n",
    "import numpy as np\n",
    "import warnings\n",
    "import base64  \n",
    "import json\n",
    "from pandas.api.types import CategoricalDtype\n",
    "\n",
    "# Visualization\n",
    "import plotly\n",
    "from plotly import tools\n",
    "import plotly.express as px\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "import matplotlib.pyplot as plt; plt.rcdefaults()\n",
    "from matplotlib import rcParams\n",
    "\n",
    "import IPython\n",
    "from IPython.display import HTML, display, Markdown, IFrame\n",
    "\n",
    "# Data analysis\n",
    "from itertools import combinations\n",
    "import scipy.spatial.distance as dist\n",
    "import scipy.stats as ss\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import quantile_transform\n",
    "\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "\n",
    "# External Code\n",
    "from utils import *\n",
    "\n",
    "from plotly.offline import init_notebook_mode\n",
    "init_notebook_mode(connected = True)\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% do SectionField(\n",
    "    name='Data_Section',\n",
    "    title='Load your Data',\n",
    "    subtitle='Load your metadata and expression data in comma/tab separated formats. Genes should be in rows and samples shoud be in columns. Example files are downloadable here: <a href=\"https://appyters.maayanlab.cloud/storage/Bulk_RNA_seq/GSE70466.zip\"> link </a>',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Normalization_Section',\n",
    "    title='Select Normalization Methods',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Visualization_Section',\n",
    "    title='Select Visualization Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='DEG_Section',\n",
    "    title='Select Differentially Exprssed Gene Analysis Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set meta_data_filename = FileField(\n",
    "    name='meta_data_filename', \n",
    "    label='Meta data file (.csv or .txt)', \n",
    "    default='GSE70466_example_metadata.txt',\n",
    "\n",
    "    examples={'GSE70466_example_metadata.txt': \"https://appyters.maayanlab.cloud/storage/Bulk_RNA_seq/GSE70466-metadata.txt\"}, \n",
    "    description='Upload metadata as two-column comma seperated or tab seperated format. One column contains sample ids and the other column contains sample labels', \n",
    "    section='Data_Section')\n",
    "\n",
    "%}\n",
    "{% set rnaseq_data_filename = FileField(\n",
    "    name='rnaseq_data_filename', \n",
    "    label='RNA-seq data file (.csv or .txt)', \n",
    "    default='GSE70466_example_expression.txt',\n",
    "    examples={'GSE70466_example_expression.txt': \"https://appyters.maayanlab.cloud/storage/Bulk_RNA_seq/GSE70466-expression.txt\"}, \n",
    "    description='Upload RNA-seq expression data as comma seperated or tab seperated format. The index of the dataset are genes, the columns are samples.', \n",
    "    section='Data_Section')\n",
    "\n",
    "%}\n",
    "{% set meta_class_column_name = StringField(\n",
    "    name='meta_class_column_name', \n",
    "    label='Class column name in metadata', \n",
    "    default='cell line', \n",
    "    description='class column name of metadata', \n",
    "    section='Data_Section')\n",
    "%}\n",
    "\n",
    "{% set control_name = StringField(\n",
    "    name='control_name', \n",
    "    label='Control label', \n",
    "    default='PrEC', \n",
    "    description='name of control class',\n",
    "    section='Data_Section')\n",
    "%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set filter_genes = BoolField(\n",
    "    name='filter_genes', \n",
    "    label='Filter genes?', \n",
    "    default='true',\n",
    "    description='Check if User wants to filter genes with lowest variances', \n",
    "    section='Normalization_Section',\n",
    ") \n",
    "%}\n",
    "\n",
    "{% set low_expression_threshold = FloatField(\n",
    "    name='low_expression_threshold', \n",
    "    label='Low expression threshold', \n",
    "    default=0.3, \n",
    "    min=0,\n",
    "    max=15.0,\n",
    "    step=0.01,\n",
    "    description='Threshold to filter out low expression genes. The value should vary based on the user dataset.', \n",
    "    section='Normalization_Section'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set logCPM_normalization = BoolField(\n",
    "    name='logCPM_normalization', \n",
    "    label='logCPM normalization?', \n",
    "    default='true', \n",
    "    description='Check if User wants the dataset to be logCPM-transformed', \n",
    "    section='Normalization_Section')\n",
    "%}\n",
    "\n",
    "{% set log_normalization = BoolField(\n",
    "    name='log_normalization', \n",
    "    label='log normalization?', \n",
    "    default='false', \n",
    "    description='Check if User wants the dataset to be log-transformed', \n",
    "    section='Normalization_Section')\n",
    "%}\n",
    "\n",
    "{% set z_normalization = BoolField(\n",
    "    name='z_normalization', \n",
    "    label='Z normalization?', \n",
    "    default='true', \n",
    "    description='Check if User wants the dataset to be normalized with Z-normalized method', \n",
    "    section='Normalization_Section')\n",
    "%}\n",
    "\n",
    "{% set q_normalization = BoolField(\n",
    "    name='q_normalization', \n",
    "    label='Quantile normalization?', \n",
    "    default='false', \n",
    "    description='Check if User wants the dataset to be normalized with Quantile normalization method', \n",
    "    section='Normalization_Section')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set interactive_plot = BoolField(\n",
    "    name='interactive_plot', \n",
    "    label='Interactive plots?', \n",
    "    default='false', \n",
    "    description='Check if User wants interactive plots', \n",
    "    section='Visualization_Section')\n",
    "%}\n",
    "\n",
    "{% set visualization_method = ChoiceField(\n",
    "    name='visualization_method', \n",
    "    label='Visualization Methods', \n",
    "    choices = {'PCA': 'PCA', 'UMAP': 'UMAP', 't-SNE': 't-SNE'},\n",
    "    default='PCA', \n",
    "    description='Select a visualization method', \n",
    "    section='Visualization_Section')\n",
    "%}\n",
    "\n",
    "{% set nr_genes = IntField(\n",
    "    name='nr_genes', \n",
    "    label='Genes for Dimension Reduction', \n",
    "    min=0, \n",
    "    max=30000, \n",
    "    default=2500, \n",
    "    description='The maximum number of genes for dimension reduction', \n",
    "    section='Visualization_Section')\n",
    "%}\n",
    "\n",
    "{% set gene_list_for_clustergrammer = TextField(\n",
    "    name='gene_list_for_clustergrammer', \n",
    "    label='Gene List for Clustergrammer (Optional)', \n",
    "    default='', \n",
    "    description='Paste your gene list (One gene per row) for Clustergrammer heatmap plots.', \n",
    "    section = 'Visualization_Section')\n",
    "%}\n",
    "\n",
    "{% set clustering_topk = IntField(\n",
    "    name='clustering_topk', \n",
    "    label='Genes for clustergrammer', \n",
    "    min=0, \n",
    "    max=1000, \n",
    "    default=800, \n",
    "    description='The number of genes with largest variance for Clustergrammer', \n",
    "    section='Visualization_Section')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set diff_gex_method = ChoiceField(\n",
    "    name='diff_gex_method',\n",
    "    label='Differential expression analysis method',\n",
    "    choices={'limma': 'limma','characteristic direction': 'characteristic_direction', 'edgeR': 'edgeR', 'DESeq2': 'DESeq2'},\n",
    "    default='limma', \n",
    "    description='Set a method to get differentially expressed genes', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "{% set diff_gex_plot_method = ChoiceField(\n",
    "    name='diff_gex_plot_method',\n",
    "    label='Differential expression analysis plotting method',\n",
    "    choices={'Volcano plot': 'volcano','MA plot': 'MA_plot'},\n",
    "    default='Volcano plot', \n",
    "    description='Set a plot method to see differentially expressed genes. Available for limma/edgeR/DESeq2.', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "\n",
    "{% set pvalue_threshold = FloatField(\n",
    "    name='pvalue_threshold', \n",
    "    label='P-value threshold', \n",
    "    min=0, \n",
    "    max=1, \n",
    "    default=0.05, \n",
    "    description='Threshold to highlight significantly differentially expressed genes.', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "{% set logfc_threshold = FloatField(\n",
    "    name='logfc_threshold', \n",
    "    label='logFC threshold',\n",
    "    min=0,\n",
    "    max=1000,\n",
    "    default=1.5, \n",
    "    description='Threshold to highlight diffentially expressed genes.', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "{% set gene_topk = IntField(\n",
    "    name='gene_topk', \n",
    "    label='Maximum genes for Enrichr', \n",
    "    min=0, \n",
    "    max=1000, \n",
    "    default=500, \n",
    "    description='The maximum number of genes discovered by the Characteristic Direction method', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "\n",
    "{% set enrichr_libraries = MultiChoiceField(\n",
    "    name='enrichr_libraries',\n",
    "    label='Enrichr Libraries (upto 2)',\n",
    "    descriptions='Enrichr libraries to be visualized. Select one or two libraries',\n",
    "    choices=['Gene Ontology',\n",
    "            'Pathway',\n",
    "            'Kinase',\n",
    "            'Transcription Factor',\n",
    "            'miRNA'],\n",
    "    default=['Gene Ontology', 'Pathway'],\n",
    "    section='DEG_Section'\n",
    "    )\n",
    "%}\n",
    "\n",
    "\n",
    "{% set nr_genesets = IntField(\n",
    "    name='nr_genesets', \n",
    "    label='Top ranked gene sets', \n",
    "    min=0, \n",
    "    max=20, \n",
    "    default=15, \n",
    "    description='The number of result gene sets', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "{% set small_molecule_method = ChoiceField(\n",
    "    name='small_molecule_method',\n",
    "    label='Small molecule analysis method',\n",
    "    choices={'L1000CDS2': 'L1000CDS2','L1000FWD': 'L1000FWD'},\n",
    "    default='L1000FWD', \n",
    "    description='Set a small molecule analysis method', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "{% set l1000_topk = IntField(\n",
    "    name='l1000_topk', \n",
    "    label='Genes for L1000CDS2 or L1000FWD', \n",
    "    min=0, \n",
    "    max=1000, \n",
    "    default=500, \n",
    "    description='The number of genes to L1000CDS2 or L1000FWD', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "{% set nr_drugs = IntField(\n",
    "    name='nr_drugs', \n",
    "    label='Top ranked drugs from L1000CDS2 or L1000FWD', \n",
    "    min=0, \n",
    "    max=20, \n",
    "    default=7, \n",
    "    description='The number of result drugs', \n",
    "    section='DEG_Section')\n",
    "%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "rnaseq_data_filename = \"{{rnaseq_data_filename.value}}\"\n",
    "meta_data_filename = \"{{meta_data_filename.value}}\"\n",
    "meta_class_column_name = \"{{meta_class_column_name.value}}\"\n",
    "control_name = \"{{control_name.value}}\"\n",
    "\n",
    "interactive_plot = {{interactive_plot.value}}\n",
    "filter_genes = {{filter_genes.value}}\n",
    "low_expression_threshold = {{low_expression_threshold.value}}\n",
    "\n",
    "\n",
    "nr_genes = {{nr_genes.value}}\n",
    "gene_list_for_clustergrammer = \"{{gene_list_for_clustergrammer.value}}\"\n",
    "clustering_topk = {{clustering_topk.value}}\n",
    "\n",
    "diff_gex_method = \"{{diff_gex_method.value}}\"\n",
    "diff_gex_plot_method = \"{{diff_gex_plot_method.value}}\"\n",
    "pvalue_threshold = {{pvalue_threshold.value}}\n",
    "logfc_threshold = {{logfc_threshold.value}}\n",
    "gene_topk = {{gene_topk.value}}\n",
    "enrichr_libraries = {{enrichr_libraries.value}}\n",
    "nr_genesets = {{nr_genesets.value}}\n",
    "\n",
    "small_molecule_method = \"{{small_molecule_method.value}}\"\n",
    "l1000_topk = {{l1000_topk.value}}\n",
    "nr_drugs = {{nr_drugs.value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings('ignore')\n",
    "random.seed(0)\n",
    "pandas2ri.activate()\n",
    "notebook_metadata = dict()\n",
    "notebook_metadata[\"tables\"] = dict()\n",
    "notebook_metadata[\"figures\"] = dict()\n",
    "notebook_metadata[\"input_parameters\"] = dict()\n",
    "if interactive_plot == True:\n",
    "    plot_type='interactive'\n",
    "else:\n",
    "    plot_type='static'\n",
    "results = {}\n",
    "table_counter = 1\n",
    "figure_counter = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "\n",
    "notebook_metadata[\"input_parameters\"][\"rnaseq_data_filename\"] = rnaseq_data_filename\n",
    "notebook_metadata[\"input_parameters\"][\"meta_data_filename\"] = meta_data_filename\n",
    "notebook_metadata[\"input_parameters\"][\"meta_class_column_name\"] = meta_class_column_name\n",
    "notebook_metadata[\"input_parameters\"][\"control_name\"] = control_name\n",
    "\n",
    "notebook_metadata[\"input_parameters\"][\"filter_genes\"] = filter_genes\n",
    "notebook_metadata[\"input_parameters\"][\"low_expression_threshold\"] = low_expression_threshold\n",
    "notebook_metadata[\"input_parameters\"][\"logCPM_normalization\"] = {{logCPM_normalization.value}}\n",
    "notebook_metadata[\"input_parameters\"][\"log_normalization\"] = {{log_normalization.value}}\n",
    "notebook_metadata[\"input_parameters\"][\"z_normalization\"] = {{z_normalization.value}}\n",
    "notebook_metadata[\"input_parameters\"][\"q_normalization\"] = {{q_normalization.value}}\n",
    "\n",
    "notebook_metadata[\"input_parameters\"][\"visualization_method\"] = \"{{visualization_method.value}}\"\n",
    "notebook_metadata[\"input_parameters\"][\"nr_genes\"] = nr_genes\n",
    "notebook_metadata[\"input_parameters\"][\"gene_list_for_clustergrammer\"] = gene_list_for_clustergrammer\n",
    "notebook_metadata[\"input_parameters\"][\"clustering_topk\"] = clustering_topk\n",
    "\n",
    "notebook_metadata[\"input_parameters\"][\"diff_gex_method\"] = diff_gex_method\n",
    "notebook_metadata[\"input_parameters\"][\"diff_gex_plot_method\"] = diff_gex_plot_method\n",
    "notebook_metadata[\"input_parameters\"][\"pvalue_threshold\"] = pvalue_threshold\n",
    "notebook_metadata[\"input_parameters\"][\"logfc_threshold\"] = logfc_threshold\n",
    "notebook_metadata[\"input_parameters\"][\"gene_topk\"] = gene_topk\n",
    "notebook_metadata[\"input_parameters\"][\"enrichr_libraries\"] = enrichr_libraries\n",
    "notebook_metadata[\"input_parameters\"][\"nr_genesets\"] = nr_genesets\n",
    "notebook_metadata[\"input_parameters\"][\"small_molecule_method\"] = small_molecule_method\n",
    "notebook_metadata[\"input_parameters\"][\"l1000_topk\"] = l1000_topk\n",
    "notebook_metadata[\"input_parameters\"][\"nr_drugs\"] = nr_drugs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "try:\n",
    "    check_files(rnaseq_data_filename)\n",
    "except:\n",
    "    print(\"Error! Please load an RNA-seq expression file in txt, tsv or csv format\")\n",
    "    pass\n",
    "    \n",
    "try:    \n",
    "    check_files(meta_data_filename)\n",
    "except:\n",
    "    print(\"Error! Please load a metadata file in txt, tsv or csv format\")\n",
    "    pass\n",
    "\n",
    "if rnaseq_data_filename.endswith(\".csv\"):\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0).sort_index()\n",
    "else:\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "if meta_data_filename.endswith(\".csv\"):\n",
    "    meta_df = pd.read_csv(meta_data_filename, index_col=0, dtype=str)\n",
    "else:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0, dtype=str)\n",
    "    \n",
    "meta_df.index = meta_df.index.map(str)\n",
    "  \n",
    "# Match samples between the metadata and the datasets\n",
    "try:\n",
    "    check_df(meta_df, meta_class_column_name)\n",
    "except:\n",
    "    print(f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "\n",
    "meta_df = meta_df[meta_df.index.isin(expr_df.columns)]\n",
    "\n",
    "# sort metadata by class labels; control first\n",
    "classes = list(meta_df[meta_class_column_name].unique())\n",
    "classes.remove(control_name)\n",
    "classes.insert(0, control_name)\n",
    "meta_df['tmp_class'] = pd.Categorical(meta_df[meta_class_column_name], classes)\n",
    "meta_df = meta_df.sort_values('tmp_class')\n",
    "meta_df = meta_df.drop('tmp_class', axis=1)\n",
    "\n",
    "\n",
    "expr_df = expr_df.loc[:,meta_df.index]\n",
    "expr_df = expr_df.groupby(expr_df.index).sum()\n",
    "try:\n",
    "    assert(meta_df.shape[0]==expr_df.shape[1])\n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format. \\\n",
    "    Please check if the index of the expression data are genes and the columns are sample IDs. \\\n",
    "    Sample IDs in the expression data and the metadata should be matched\")\n",
    "dataset = dict()\n",
    "current_dataset = 'rawdata'\n",
    "dataset[current_dataset] = expr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if filter_genes.value == True %}\n",
    "Filter out low expressed genes \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if filter_genes.value == True %}\n",
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "\n",
    "## Filter out lowly expressed genes\n",
    "mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "expr_df = expr_df.loc[mask_low_vals, :]\n",
    "current_dataset += '+filter_genes'\n",
    "dataset[current_dataset] = expr_df\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset['dataset_metadata'] = meta_df\n",
    "\n",
    "table_counter, notebook_metadata = display_object(table_counter, \"Raw RNA-seq expression data. The table displays the first 5 rows of the quantified RNA-seq expression dataset. Rows represent genes, columns represent samples, and values show the number of mapped reads.\", notebook_metadata, \"raw_exp.csv\", dataset[current_dataset].head(), istable=True)\n",
    "table_counter, notebook_metadata = display_object(table_counter, \"Metadata. The table displays the metadata associated with the samples in the RNA-seq dataset. Rows represent RNA-seq samples, columns represent metadata categories.\", notebook_metadata, \"metadata.csv\", dataset['dataset_metadata'].head(), istable=True)\n",
    "table_counter, notebook_metadata = display_object(table_counter, \"Sample size for each class. The table displays the number of samples in each class.\", notebook_metadata, \"num_of_samples_in_class.csv\", dataset['dataset_metadata'].reset_index().groupby(meta_class_column_name).count(), istable=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if logCPM_normalization.value == True or log_normalization.value == True or z_normalization.value == True or q_normalization.value == True %}\n",
    "Normalization methods (\n",
    "{% if logCPM_normalization.value %}count per million (CPM), {% endif %} {% if log_normalization.value %} log transformation, {% endif %} {% if z_normalization.value %} Z normalization, {% endif %}  {% if q_normalization.value %}quantile normalization {% endif %}) will be applied to convert raw read counts into informative measures of gene expression and remove factors that affect the analysis.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if logCPM_normalization.value == True or log_normalization.value == True or z_normalization.value == True or q_normalization.value == True %}\n",
    "\n",
    "dataset, normalization = normalize(dataset, current_dataset, {{logCPM_normalization}}, {{log_normalization}}, {{z_normalization}}, {{q_normalization}})\n",
    "table_counter, notebook_metadata = display_object(table_counter,\"Normalized data. The table displays the expression values after normalization.\",  notebook_metadata, \"normalized_exp.csv\", dataset[normalization].head(), istable=True)\n",
    "display(create_download_link(dataset[normalization], filename=\"normalized_exp.csv\"))\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualize Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if visualization_method.value == \"PCA\" %}\n",
    "Principal Component Analysis (PCA) (Clark et al. 2011) is a statistical technique used to identify global patterns in high-dimensional datasets. It is commonly used to explore the similarity of biological samples in RNA-seq datasets. To achieve this, gene expression values are transformed into Principal Components (PCs), a set of linearly uncorrelated features which represent the most relevant sources of variance in the data, and subsequently visualized using a scatter plot.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if visualization_method.value == \"PCA\" %}\n",
    "\n",
    "method = \"PCA\"\n",
    "{% elif visualization_method.value == \"UMAP\"%}\n",
    "\n",
    "method = \"UMAP\"\n",
    "{% elif visualization_method.value == \"t-SNE\"%}\n",
    "\n",
    "method = \"t-SNE\"\n",
    "{% endif %}\n",
    "\n",
    "# Run analysis\n",
    "results[method] = run_dimension_reduction(dataset=dataset, method=method,\\\n",
    "                         nr_genes=nr_genes, normalization=normalization, plot_type=plot_type)\n",
    "# Display results\n",
    "plot_name = \"{}_plot_of_samples.png\".format(method)\n",
    "figure_counter, notebook_metadata = plot_samples(results[method], meta_class_column_name=meta_class_column_name, counter=figure_counter, plot_name=plot_name, notebook_metadata=notebook_metadata, plot_type=plot_type)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clustergrammer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clustergrammer (Fernandez et al. 2017) is a web-based tool for visualizing and analyzing high-dimensional data as interactive and hierarchically clustered heatmaps. It is commonly used to explore the similarity between samples in an RNA-seq dataset. In addition to identifying clusters of samples, it also allows to identify the genes which contribute to the clustering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run analysis\n",
    "results['clustergrammer'] = run_clustergrammer(dataset=dataset, meta_class_column_name=meta_class_column_name, nr_genes=clustering_topk, normalization=normalization, z_score=True, gene_list=gene_list_for_clustergrammer)\n",
    "\n",
    "# Display results\n",
    "plot_clustergrammar(results['clustergrammer'])\n",
    "caption = \"Clustered heatmap plot. The figure contains an interactive heatmap displaying gene expression for each sample in the RNA-seq dataset. Every row of the heatmap represents a gene, every column represents a sample, and every cell displays normalized gene expression values. The heatmap additionally features color bars beside each column which represent prior knowledge of each sample, such as the tissue of origin or experimental treatment.\"\n",
    "\n",
    "figure_counter, notebook_metadata = display_object(figure_counter, caption, notebook_metadata, saved_filename=results['clustergrammer'], istable=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Library size analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to quantify gene expression in an RNA-seq dataset, reads generated from the sequencing step are mapped to a reference genome and subsequently aggregated into numeric gene counts. Due to experimental variations and random technical noise, samples in an RNA-seq datasets often have variable amounts of the total RNA. Library size analysis calculates and displays the total number of reads mapped for each sample in the RNA-seq dataset, facilitating the identification of outlying samples and the assessment of the overall quality of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df['sum'] = expr_df.sum().tolist()\n",
    "\n",
    "fig = px.histogram(meta_df[\"sum\"])\n",
    "fig.update_yaxes(title=\"samples/cells\")\n",
    "fig.update_xaxes(title=\"reads\")\n",
    "if plot_type == \"static\":\n",
    "    fig.show(renderer=\"png\")\n",
    "else:\n",
    "    fig.show()\n",
    "plot_name = \"library_size_plot.png\"\n",
    "fig.write_image(plot_name)\n",
    "figure_counter, notebook_metadata = display_object(figure_counter, \"Histogram of the total number of reads mapped for each sample. The figure contains an interactive bar chart which displays the number of samples according to the total number of reads mapped to each RNA-seq sample in the dataset. Additional information for each sample is available by hovering over the bars.\", notebook_metadata, saved_filename=plot_name, istable=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential Gene Expression "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gene expression signatures are alterations in the patterns of gene expression that occur as a result of cellular perturbations such as drug treatments, gene knock-downs or diseases. They can be quantified using differential gene expression (DGE) methods (Ritchie et al. 2015, Clark et al. 2014), which compare gene expression between two groups of samples to identify genes whose expression is significantly altered in the perturbation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "signatures = get_signatures(classes, dataset, normalization, diff_gex_method, meta_class_column_name, filter_genes)\n",
    "\n",
    "for label, signature in signatures.items():\n",
    "    case_label = label.split(\" vs. \")[1]\n",
    "    table_counter, notebook_metadata = display_object(table_counter, \"Differentially expressed genes between {} using {}. The figure displays a browsable table containing the gene expression signature generated from a differential gene expression analysis. Every row of the table represents a gene; the columns display the estimated measures of differential expression.\".format(label, diff_gex_method), notebook_metadata, \"DEG_results_{}.csv\".format(label), signature, istable=True)\n",
    "    display(create_download_link(signature, filename=\"DEG_results_{}.csv\".format(label)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if diff_gex_method.value == \"limma\" or diff_gex_method.value == \"edgeR\" or diff_gex_method.value == \"DESeq2\"%}\n",
    "    \n",
    "{% if diff_gex_plot_method.value == \"volcano\" %}\n",
    "results['volcano_plot'] = {}\n",
    "# Loop through signatures\n",
    "for label, signature in signatures.items():\n",
    "    results['volcano_plot'][label] = run_volcano(signature, label, dataset, pvalue_threshold, logfc_threshold, plot_type)\n",
    "    plot_name = plot_volcano(results['volcano_plot'][label])\n",
    "    figure_counter, notebook_metadata = display_object(figure_counter, \"Volcano plot for {}. The figure contains an interactive scatter plot which displays the log2-fold changes and statistical significance of each gene calculated by performing a differential gene expression analysis. Genes with logFC > {} and p-value < {} in red and genes with logFC < -{} and p-value < {} in blue. Additional information for each gene is available by hovering over it.\".format(label, logfc_threshold, pvalue_threshold, logfc_threshold, pvalue_threshold), notebook_metadata, plot_name, istable=False)\n",
    "\n",
    "{% elif diff_gex_plot_method.value == \"MA_plot\" %}\n",
    "# Initialize results\n",
    "results['ma_plot'] = {}\n",
    "\n",
    "# Loop through signatures\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    results['ma_plot'][label] = run_maplot(signature=signature, signature_label=label, pvalue_threshold=pvalue_threshold, logfc_threshold=logfc_threshold, plot_type=plot_type)\n",
    "    # Display results\n",
    "    plot_name = plot_maplot(results['ma_plot'][label])\n",
    "    figure_counter, notebook_metadata = display_object(figure_counter, \"MA plot for {}. The figure contains an interactive scatter plot which displays the average expression and statistical significance of each gene calculated by performing differential gene expression analysis. Genes with logFC > {} and p-value < {} in red and genes with logFC < -{} and p-value < {} in blue. Additional information for each gene is available by hovering over it.\".format(label, logfc_threshold, pvalue_threshold, logfc_threshold, pvalue_threshold), notebook_metadata, plot_name, istable=False)\n",
    "\n",
    "{% endif %}\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichment Analysis using Enrichr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Enrichment analysis is a statistical procedure used to identify biological terms which are over-represented in a given gene set. These include signaling pathways, molecular functions, diseases, and a wide variety of other biological terms obtained by integrating prior knowledge of gene function from multiple resources. Enrichr (Kuleshov et al. 2016) is a web-based application which allows to perform enrichment analysis using a large collection of gene-set libraries and various interactive approaches to display enrichment results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loop through signatures\n",
    "results = {}\n",
    "results['enrichr']= {}\n",
    "if diff_gex_method == \"characteristic_direction\":\n",
    "    fc_colname = \"CD-coefficient\"\n",
    "    sort_genes_by = \"CD-coefficient\"\n",
    "    ascending = False\n",
    "elif diff_gex_method == \"limma\":\n",
    "    fc_colname = \"logFC\"\n",
    "    sort_genes_by = \"t\"\n",
    "    ascending = False\n",
    "elif diff_gex_method == \"edgeR\":\n",
    "    fc_colname = \"logFC\"\n",
    "    sort_genes_by = \"PValue\"\n",
    "    ascending = True\n",
    "elif diff_gex_method == \"DESeq2\":\n",
    "    fc_colname = \"log2FoldChange\"\n",
    "    sort_genes_by = \"padj\"\n",
    "    ascending = True\n",
    "enrichr_link_dict = dict()\n",
    "for label, signature in signatures.items():    \n",
    "    case_label = label.split(\" vs. \")[1]\n",
    "    # Run analysis\n",
    "    results['enrichr'][label] = run_enrichr(signature=signature, signature_label=label, fc_colname=fc_colname,geneset_size=gene_topk, sort_genes_by = sort_genes_by,ascending=ascending)\n",
    "    tmp_enrichr_link_dict = dict()\n",
    "    title_up = f\"Enrichment Analysis Result: {label} (up-regulated in {case_label})\"  \n",
    "    title_down = f\"Enrichment Analysis Result: {label} (down-regulated in {case_label})\"  \n",
    "    enrichr_link_dict[title_up] = dict()\n",
    "    enrichr_link_dict[title_up][\"link\"] = \"<a href=https://maayanlab.cloud/Enrichr/enrich?dataset={} target=\\\"_blank\\\">link to Enrichr</a>\".format(results['enrichr'][label][\"upregulated\"][\"shortId\"])\n",
    "    enrichr_link_dict[title_down] = dict()\n",
    "    enrichr_link_dict[title_down][\"link\"] = \"<a href=https://maayanlab.cloud/Enrichr/enrich?dataset={} target=\\\"_blank\\\">link to Enrichr</a>\".format(results['enrichr'][label][\"downregulated\"][\"shortId\"])\n",
    "\n",
    "enrichr_link_df = pd.DataFrame.from_dict(enrichr_link_dict).T\n",
    "table_counter, notebook_metadata = display_object(table_counter, \"The table displays links to Enrichr containing the results of enrichment analyses generated by analyzing the up-regulated and down-regulated genes from a differential expression analysis. By clicking on these links, users can interactively explore and download the enrichment results from the Enrichr website.\", notebook_metadata=notebook_metadata, saved_filename=\"enrichr_links.csv\", df=enrichr_link_df, ishtml=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if \"Gene Ontology\" in enrichr_libraries.value %}\n",
    "# GO Enrichment Analysis\n",
    "Gene Ontology (GO) (Ashburner et al. 2000) is a major bioinformatics initiative aimed at unifying the representation of gene attributes across all species. It contains a large collection of experimentally validated and predicted associations between genes and biological terms. This information can be leveraged by Enrichr to identify the biological processes, molecular functions and cellular components which are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if \"Gene Ontology\" in enrichr_libraries.value %}\n",
    "\n",
    "results['go_enrichment'] = {}\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    results['go_enrichment'][label] = get_enrichr_results_by_library(results['enrichr'][label], label, library_type='go', version='2018')\n",
    "    \n",
    "for label, signature in signatures.items():\n",
    "    # Create dataframe\n",
    "    enrichment_results = results['go_enrichment'][label]\n",
    "    enrichment_dataframe = pd.concat([enrichment_results['upregulated'], enrichment_results['downregulated']])\n",
    "\n",
    "    # Plot barcharts\n",
    "    libraries = enrichment_dataframe['gene_set_library'].unique()   \n",
    "    for gene_set_library in libraries:\n",
    "        plot_name = \"{}_barchart_{}.png\".format(gene_set_library, label)\n",
    "        plot_library_barchart(enrichment_results, gene_set_library, enrichment_results['signature_label'], enrichment_results['sort_results_by'], nr_genesets=nr_genesets, plot_type=plot_type, plot_name=plot_name) # 10 300\n",
    "        figure_counter, notebook_metadata = display_object(figure_counter, \"Enrichment Analysis Results for {} in Gene Onotology ({}). The figure contains interactive bar charts displaying the results of the Gene Ontology enrichment analysis generated using Enrichr. The x axis indicates the -log10(P-value) for each term. Significant terms are highlighted in bold. Additional information about enrichment results is available by hovering over each bar.\".format(label, gene_set_library), notebook_metadata, saved_filename=plot_name, istable=False)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if \"Pathway\" in enrichr_libraries.value %}\n",
    "# Pathway Enrichment Analysis\n",
    "Biological pathways are sequences of interactions between biochemical compounds which play a key role in determining cellular behavior. Databases such as KEGG (Kanehisa et al. 2000), Reactome (Croft et al. 2014) and WikiPathways (Kelder et al. 2012) contain a large number of associations between such pathways and genes. This information can be leveraged by Enrichr to identify the biological pathways which are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if \"Pathway\" in enrichr_libraries.value %}\n",
    "# Initialize results\n",
    "results['pathway_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, enrichr_results in results['enrichr'].items():\n",
    "    # Run analysis\n",
    "    results['pathway_enrichment'][label] = get_enrichr_results_by_library(enrichr_results=enrichr_results, signature_label=label, plot_type=plot_type, library_type='pathway', sort_results_by='pvalue')\n",
    "\n",
    "for label, signature in signatures.items():\n",
    "    # Create dataframe\n",
    "    enrichment_results = results['pathway_enrichment'][label]\n",
    "    enrichment_dataframe = pd.concat([enrichment_results['upregulated'], enrichment_results['downregulated']])\n",
    "\n",
    "    # Plot barcharts\n",
    "    libraries = enrichment_dataframe['gene_set_library'].unique()   \n",
    "    for gene_set_library in libraries:\n",
    "        # Display results\n",
    "        plot_name = \"{}_barchart_{}.png\".format(gene_set_library, label)\n",
    "        plot_library_barchart(enrichment_results, gene_set_library, enrichment_results['signature_label'], enrichment_results['sort_results_by'], nr_genesets=nr_genesets, plot_type=plot_type)\n",
    "        figure_counter, notebook_metadata = display_object(figure_counter, \"Enrichment Analysis Results for {} in {}. The figure contains interactive bar charts displaying the results of the pathway enrichment analysis generated using Enrichr. The x axis indicates the -log10(P-value) for each term. Significant terms are highlighted in bold. Additional information about enrichment results is available by hovering over each bar.\".format(label, gene_set_library), notebook_metadata, saved_filename=plot_name, istable=False)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if \"Transcription Factor\" in enrichr_libraries.value %}\n",
    "# Transcription Factor Enrichment Analysis\n",
    "Transcription Factors (TFs) are proteins involved in the transcriptional regulation of gene expression. Databases such as ChEA (Lachmann et al. 2010) and ENCODE (Consortium, 2014) contain a large number of associations between TFs and their transcriptional targets. This information can be leveraged by Enrichr to identify the transcription factors whose targets are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if \"Transcription Factor\" in enrichr_libraries.value %}\n",
    "# Initialize results\n",
    "results['tf_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, enrichr_results in results['enrichr'].items():\n",
    "#     # Run analysis\n",
    "    results['tf_enrichment'][label] = get_enrichr_result_tables_by_library(enrichr_results=enrichr_results, signature_label=label)\n",
    "    table_counter, notebook_metadata = display_table(results['tf_enrichment'][label], \"Transcription Factor\", notebook_metadata, table_counter)\n",
    "    \n",
    "{% endif %}   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if \"Kinase\" in enrichr_libraries.value %}\n",
    "# Kinase Enrichment Analysis\n",
    "Protein kinases are enzymes that modify other proteins by chemically adding phosphate groups. Databases such as KEA (Lachmann et al. 2009) contain a large number of associations between kinases and their substrates. This information can be leveraged by Enrichr to identify the protein kinases whose substrates are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if \"Kinase\" in enrichr_libraries.value %}\n",
    "# Initialize results\n",
    "results['kinase_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, enrichr_results in results['enrichr'].items():\n",
    "    # Run analysis\n",
    "    results['kinase_enrichment'][label] = get_enrichr_result_tables_by_library(enrichr_results=enrichr_results, signature_label=label, library_type=\"ke\")\n",
    "\n",
    "    # Display results\n",
    "    table_counter, notebook_metadata = display_table(results['kinase_enrichment'][label], \"Kinase\", notebook_metadata, table_counter)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if \"miRNA\" in enrichr_libraries.value %}\n",
    "# miRNA Enrichment Analysis\n",
    "microRNAs (miRNAs) are small non-coding RNA molecules which play a key role in the post-transcriptional regulation of gene expression. Databases such as TargetScan (Agarwal et al. 2015) and MiRTarBase (Chou et al. 2016) contain a large number of associations between miRNAs and their targets. This information can be leveraged by Enrichr to identify the miRNAs whose targets are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if \"miRNA\" in enrichr_libraries.value %}\n",
    "\n",
    "results['mirna_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, enrichr_results in results['enrichr'].items():\n",
    "    # Run analysis\n",
    "    results['mirna_enrichment'][label] = get_enrichr_result_tables_by_library(enrichr_results=enrichr_results, signature_label=label, library_type=\"mirna\")\n",
    "\n",
    "    # Display results\n",
    "    table_counter, notebook_metadata = display_table(results['mirna_enrichment'][label], \"miRNA\", notebook_metadata, table_counter)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if small_molecule_method.value == \"L1000CDS2\" %}\n",
    "# L1000CDS2 Query\n",
    "L1000CDS2 (Duan et al. 2016) is a web-based tool for querying gene expression signatures against signatures created from human cell lines treated with over 20,000 small molecules and drugs for the LINCS project. It is commonly used to identify small molecules which mimic or reverse the effects of a gene expression signature generated from a differential gene expression analysis.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if small_molecule_method.value == \"L1000CDS2\" %}\n",
    "# Initialize results\n",
    "results['l1000cds2'] = {}\n",
    "\n",
    "# Loop through signatures\n",
    "for label, signature in signatures.items(): \n",
    "    # Run analysis\n",
    "    results['l1000cds2'][label] = run_l1000cds2(signature=signature, nr_genes=l1000_topk, signature_label=label, plot_type=plot_type)\n",
    "\n",
    "    # Display results\n",
    "    plot_name = \"L1000CDS2_{}.png\".format(label)\n",
    "    figure_counter, notebook_metadata = plot_l1000cds2(results['l1000cds2'][label], counter=figure_counter, nr_drugs=nr_drugs, notebook_metadata=notebook_metadata, plot_name=plot_name)\n",
    "{% endif %}    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if small_molecule_method.value == \"L1000FWD\" %}\n",
    "# L1000FWD Query\n",
    "L1000FWD (Wang et al. 2018) is a web-based tool for querying gene expression signatures against signatures created from human cell lines treated with over 20,000 small molecules and drugs for the LINCS project.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if small_molecule_method.value == \"L1000FWD\" %}\n",
    "# Initialize results\n",
    "results['l1000fwd'] = {}\n",
    "\n",
    "# Loop through signatures\n",
    "for label, signature in signatures.items():\n",
    "    display(Markdown(\"*L1000FWD for {}*\".format(label)))\n",
    "    \n",
    "    # Run analysis\n",
    "    results['l1000fwd'][label] = run_l1000fwd(signature=signature, signature_label=label, nr_genes=l1000_topk)\n",
    "\n",
    "    # Display results\n",
    "    figure_counter, table_counter, notebook_metadata = plot_l1000fwd(results['l1000fwd'][label], figure_counter=figure_counter, table_counter=table_counter, notebook_metadata=notebook_metadata)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save metadata of the notebook as json\n",
    "with open(\"notebook_metadata.json\", \"w\") as fw:\n",
    "    json.dump(notebook_metadata, fw)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Agarwal, Vikram, et al. \"Predicting effective microRNA target sites in mammalian mRNAs.\" elife 4 (2015): e05005.\n",
    "<br>\n",
    "Ashburner, M., Ball, C.A., Blake, J.A., Botstein, D., Butler, H., Cherry, J.M., Davis, A.P., Dolinski, K., Dwight, S.S. and Eppig, J.T. (2000) Gene Ontology: tool for the unification of biology. Nature genetics, 25, 25.\n",
    "<br>\n",
    "Chou, Chih-Hung, et al. \"miRTarBase 2016: updates to the experimentally validated miRNA-target interactions database.\" Nucleic acids research 44.D1 (2016): D239-D247.\n",
    "<br>\n",
    "Clark, N.R. and Ma’ayan, A. (2011) Introduction to statistical methods to analyze large data sets: principal components analysis. Sci. Signal., 4, tr3-tr3.\n",
    "<br>\n",
    "Clark, Neil R., et al. \"The characteristic direction: a geometrical approach to identify differentially expressed genes.\" BMC bioinformatics 15.1 (2014): 79.\n",
    "<br>\n",
    "Consortium, E.P. (2004) The ENCODE (ENCyclopedia of DNA elements) project. Science, 306, 636-640.\n",
    "<br>\n",
    "Croft, David, et al. \"The Reactome pathway knowledgebase.\" Nucleic acids research 42.D1 (2014): D472-D477.\n",
    "<br>\n",
    "Duan, Q., et al. \"L1000cds2: Lincs l1000 characteristic direction signatures search engine. NPJ Syst Biol Appl. 2016; 2: 16015.\" (2016).\n",
    "<br>\n",
    "Fernandez, Nicolas F., et al. \"Clustergrammer, a web-based heatmap visualization and analysis tool for high-dimensional biological data.\" Scientific data 4 (2017): 170151.\n",
    "<br>\n",
    "Kanehisa, M. and Goto, S. (2000) KEGG: kyoto encyclopedia of genes and genomes. Nucleic acids research, 28, 27-30.\n",
    "<br>\n",
    "Kelder, Thomas, et al. \"WikiPathways: building research communities on biological pathways.\" Nucleic acids research 40.D1 (2012): D1301-D1307.\n",
    "<br>\n",
    "Kuleshov, M.V., Jones, M.R., Rouillard, A.D., Fernandez, N.F., Duan, Q., Wang, Z., Koplev, S., Jenkins, S.L., Jagodnik, K.M. and Lachmann, A. (2016) Enrichr: a comprehensive gene set enrichment analysis web server 2016 update. Nucleic acids research, 44, W90-W97.\n",
    "<br>\n",
    "Lachmann, A., Xu, H., Krishnan, J., Berger, S.I., Mazloom, A.R. and Ma'ayan, A. (2010) ChEA: transcription factor regulation inferred from integrating genome-wide ChIP-X experiments. Bioinformatics, 26, 2438-2444.\n",
    "<br>\n",
    "Lachmann, Alexander, and Avi Ma'ayan. \"KEA: kinase enrichment analysis.\" Bioinformatics 25.5 (2009): 684-686.\n",
    "<br>\n",
    "Ritchie, Matthew E., et al. \"limma powers differential expression analyses for RNA-sequencing and microarray studies.\" Nucleic acids research 43.7 (2015): e47-e47.\n",
    "<br>\n",
    "Wang, Zichen, et al. \"L1000FWD: fireworks visualization of drug-induced transcriptomic signatures.\" Bioinformatics 34.12 (2018): 2150-2152."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py39",
   "language": "python",
   "name": "py39"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
