{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bulk RNA-seq Analysis pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\">\n",
    "<li><span><a  href=\"#Bulk-RNA-seq-Analysis-pipeline\" data-toc-modified-id=\"Bulk-RNA-seq-Analysis-pipeline-1\"><span class=\"toc-item-num\"></span>Bulk RNA-seq Analysis pipeline</a></span></li>\n",
    "<li><span><a  href=\"#1.-Load-datasets\" data-toc-modified-id=\"1.-Load-datasets-2\"><span class=\"toc-item-num\"></span>1. Load datasets</a></span></li>\n",
    "<li><span><a  href=\"#2.-PCA\" data-toc-modified-id=\"2.-PCA-3\"><span class=\"toc-item-num\"></span>2. PCA</a></span></li>\n",
    "<li><span><a  href=\"#3.-Clustergrammer\" data-toc-modified-id=\"3.-Clustergrammer-4\"><span class=\"toc-item-num\"></span>3. Clustergrammer</a></span></li>\n",
    "<li><span><a  href=\"#4.-Library-size-analysis\" data-toc-modified-id=\"4.-Library-size-analysis-5\"><span class=\"toc-item-num\"></span>4. Library size analysis</a></span></li>\n",
    "<li><span><a  href=\"#5.-Differential-Gene-Expression\" data-toc-modified-id=\"5.-Differential-Gene-Expression-6\"><span class=\"toc-item-num\"></span>5. Differential Gene Expression</a></span></li>\n",
    "<li><span><a  href=\"#6.-Volcano-Plot\" data-toc-modified-id=\"6.-Volcano-Plot-7\"><span class=\"toc-item-num\"></span>6. Volcano Plot</a></span></li>\n",
    "<li><span><a  href=\"#7.-MA-Plot\" data-toc-modified-id=\"7.-MA-Plot-8\"><span class=\"toc-item-num\"></span>7. MA Plot</a></span></li>\n",
    "<li><span><a  href=\"#8.-Enrichr\" data-toc-modified-id=\"8.-Enrichr-9\"><span class=\"toc-item-num\"></span>8. Enrichr</a></span></li>\n",
    "<li><span><a  href=\"#9.-GO-Enrichment-Analysis\" data-toc-modified-id=\"9.-GO-Enrichment-Analysis-10\"><span class=\"toc-item-num\"></span>9. GO Enrichment Analysis</a></span></li>\n",
    "<li><span><a  href=\"#10.-Pathway-Enrichment-Analysis\" data-toc-modified-id=\"10.-Pathway-Enrichment-Analysis-11\"><span class=\"toc-item-num\"></span>10. Pathway Enrichment Analysis</a></span></li>\n",
    "<li><span><a  href=\"#11.-Transcription-Factor-Enrichment-Analysis\" data-toc-modified-id=\"11.-Transcription-Factor-Enrichment-Analysis-12\"><span class=\"toc-item-num\"></span>11. Transcription Factor Enrichment Analysis</a></span></li>\n",
    "<li><span><a  href=\"#12.-Kinase-Enrichment-Analysis\" data-toc-modified-id=\"12.-Kinase-Enrichment-Analysis-13\"><span class=\"toc-item-num\"></span>12. Kinase Enrichment Analysis</a></span></li>\n",
    "<li><span><a  href=\"#13.-miRNA-Enrichment-Analysis\" data-toc-modified-id=\"13.-miRNA-Enrichment-Analysis-14\"><span class=\"toc-item-num\"></span>13. miRNA Enrichment Analysis</a></span></li>\n",
    "<li><span><a  href=\"#14.-L1000CDS2-Query\" data-toc-modified-id=\"14.-L1000CDS2-Query-15\"><span class=\"toc-item-num\"></span>14. L1000CDS2 Query</a></span></li>\n",
    "<li><span><a  href=\"#15.-L1000FWD-Query\" data-toc-modified-id=\"15.-L1000FWD-Query-16\"><span class=\"toc-item-num\"></span>15. L1000FWD Query</a></span></li>\n",
    "<li><span><a  href=\"#References\" data-toc-modified-id=\"References-17\"><span class=\"toc-item-num\"></span>References</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%javascript\n",
    "require.config({\n",
    "  paths: {\n",
    "    d3: 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.9.2/d3',\n",
    "    jquery: 'https://code.jquery.com/jquery-3.4.1.min',\n",
    "    plotly: 'https://cdn.plot.ly/plotly-latest.min'\n",
    "  },\n",
    "\n",
    "  shim: {\n",
    "    plotly: {\n",
    "      deps: ['d3', 'jquery'],\n",
    "      exports: 'plotly'\n",
    "    }\n",
    "  }\n",
    "});"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "{% do SectionField(\n",
    "    name='PRIMARY',\n",
    "    title='Bulk RNA-seq analysis pipeline',\n",
    "    subtitle='This pipeline enables you to analyze and visualize your bulk RNA sequencing datasets with an array of  downstream analysis and visualization tools. The pipeline includes: PCA analysis, Clustergrammer interactive heatmap, library size analysis, differential gene expression analysis, enrichment analysis, and L1000 small molecule search.',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "meta_data_filename = \"{{ FileField(name='meta_data_filename', label='Meta data file (.csv or .tsv)', default='metadata_GSE147507_series1.csv', description='Upload metadata as two-column csv format. One column contains sample ids and the other column contains sample labels', section='PRIMARY').value }}\"\n",
    "rnaseq_data_filename = \"{{ FileField(name='rnaseq_data_filename', label='RNA-seq data file (.csv or .tsv)', default='GSE147507_RawReadCounts_Human.csv', description='Upload rna-seq dataset as csv format. The index of the dataset are genes, the columns are samples.', section='PRIMARY').value }}\"\n",
    "\n",
    "meta_class_column_name = \"{{ StringField(name='meta_class_column_name', label='Class column name in metadata', default='CLASS', description='class column name of metadata', section='PRIMARY').value }}\"\n",
    "meta_id_column_name = \"{{ StringField(name='meta_id_column_name', label='Sample ID column name in metadata', default='SAMPLE_ID', description='sample ID column name of metadata', section='PRIMARY').value }}\"\n",
    "\n",
    "filter_genes = {{ BoolField(name='filter_genes', label='Filter genes?', default='true', description='Check if User wants the dataset to be log-transformed', section='PRIMARY').value }}\n",
    "low_expression_threshold = {{ FloatField(name='low_expression_threshold', label='Low expression threshold', default='0.3', description='Threshold to filter out low expression genes. The value should vary based on the user dataset.', section='PRIMARY').value }}\n",
    "\n",
    "logCPM_normalization = {{ BoolField(name='logCPM_normalization', label='logCPM normalization?', default='true', description='Check if User wants the dataset to be logCPM-transformed', section='PRIMARY').value }}\n",
    "log_normalization = {{ BoolField(name='log_normalization', label='Log normalization?', default='false', description='Check if User wants the dataset to be log-transformed', section='PRIMARY').value }}\n",
    "z_normalization = {{ BoolField(name='z_normalization', label='Z normalization?', default='true', description='Check if User wants the dataset to be normalized with Z-normalized method', section='PRIMARY').value }}\n",
    "q_normalization = {{ BoolField(name='q_normalization', label='Quantile normalization?', default='false', description='Check if User wants the dataset to be normalized with Quantile normalization method', section='PRIMARY').value }}\n",
    "\n",
    "nr_genes = {{ IntField(name='nr_genes', label='Genes for PCA', min=0, max=30000, default=2500, description='The maximum number of genes discovered by the Characteristic Direction method', section='PRIMARY').value }}\n",
    "clustering_topk = {{ IntField(name='clustering_topk', label='Genes for clustergrammer', min=0, max=1000, default=800, description='The number of genes with largest variance for Clustergrammer', section='PRIMARY').value }}\n",
    "gene_topk = {{ IntField(name='gene_topk', label='Maximum genes for Enrichr', min=0, max=1000, default=500, description='The maximum number of genes discovered by the Characteristic Direction method', section='PRIMARY').value }}\n",
    "l1000_topk = {{ IntField(name='l1000_topk', label='Genes for L1000CDS2 and L1000FWD', min=0, max=1000, default=500, description='The number of genes to be queried to L1000CDS2', section='PRIMARY').value }}\n",
    "nr_drugs = {{ IntField(name='nr_drugs', label='Top ranked drugs', min=0, max=20, default=7, description='the number of result drugs', section='PRIMARY').value }}\n",
    "nr_genesets = {{ IntField(name='nr_genesets', label='Top ranked gene sets', min=0, max=20, default=15, description='the number of result gene sets', section='PRIMARY').value }}\n",
    "\n",
    "\n",
    "pvalue_threshold = {{ FloatField(name='pvalue_threshold', label='P-value threshold', default='0.05', description='Threshold to highlight significantly differentially expressed genes.', section='PRIMARY').value }}\n",
    "logfc_threshold = {{ FloatField(name='logfc_threshold', label='logFC threshold', default='1.5', description='Threshold to highlight diffentially expressed genes.', section='PRIMARY').value }}\n",
    "\n",
    "diff_gex_method = \"{{ ChoiceField(name='diff_gex_method',label='Differential expression analysis method',choices={'limma': 'limma','characteristic_direction': 'characteristic_direction'},default='limma', description='Set a method to get differentially expressed genes', section='PRIMARY').value }}\"\n",
    "interactive_plot = {{ BoolField(name='interactive_plot', label='Interactive plots?', default='true', description='Check if User wants interactive plots', section='PRIMARY').value }}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import urllib3\n",
    "import requests, json\n",
    "import sys\n",
    "import geode\n",
    "\n",
    "import plotly\n",
    "from plotly import tools\n",
    "import plotly.express as px\n",
    "import plotly.graph_objs as go\n",
    "plotly.offline.init_notebook_mode() # To embed plots in the output cell of the notebook\n",
    "\n",
    "import matplotlib.pyplot as plt; plt.rcdefaults()\n",
    "from matplotlib import rcParams\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib_venn import venn2, venn3\n",
    "%matplotlib inline\n",
    "\n",
    "import IPython\n",
    "from IPython.display import HTML, display, Markdown, IFrame\n",
    "\n",
    "from itertools import combinations\n",
    "\n",
    "import scipy.spatial.distance as dist\n",
    "import scipy.stats as ss\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import quantile_transform\n",
    "\n",
    "import random\n",
    "from time import sleep\n",
    "import time\n",
    "import numpy as np\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "random.seed(0)\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "pandas2ri.activate()\n",
    "if interactive_plot == True:\n",
    "    plot_type='interactive'\n",
    "else:\n",
    "    plot_type='static'\n",
    "    \n",
    "import chart_studio\n",
    "import chart_studio.plotly as py\n",
    "chart_studio.tools.set_credentials_file(username='mjjeon', api_key='v0rpMa6lhST28Sq7XqtM')\n",
    "\n",
    "results = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Load datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if rnaseq_data_filename.endswith(\".csv\"):\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0).sort_index()\n",
    "else:\n",
    "    expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "if meta_data_filename.endswith(\".csv\"):\n",
    "    meta_df = pd.read_csv(meta_data_filename)\n",
    "else:\n",
    "    meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\")\n",
    "    \n",
    "# Match samples between the metadata and the datasets\n",
    "meta_df = meta_df[meta_df[meta_id_column_name].isin(expr_df.columns)]\n",
    "expr_df = expr_df.loc[:,meta_df[meta_id_column_name]]\n",
    "\n",
    "assert(meta_df.shape[0]==expr_df.shape[1])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter out low expressed genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter out non-expressed genes\n",
    "expr_df = expr_df.loc[expr_df.sum(axis=1) > 0, :]\n",
    "print (expr_df.shape)\n",
    "\n",
    "if filter_genes == True: \n",
    "    ## Filter out lowly expressed genes\n",
    "    mask_low_vals = (expr_df > low_expression_threshold).sum(axis=1) > 2\n",
    "    expr_df = expr_df.loc[mask_low_vals, :]\n",
    "\n",
    "    print (expr_df.shape)\n",
    "dataset = dict()\n",
    "dataset['rawdata'] = expr_df\n",
    "dataset['dataset_metadata'] = meta_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Various normalization methods such as count per million (CPM), log transformation, Z normalization, and quantile normalization will be applied to convert raw read counts into informative measures of gene expression and remove factors that affect the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CPM(data):\n",
    "\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\")\n",
    "        data = (data/data.sum())*10**6\n",
    "        data = data.fillna(0)\n",
    "        \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def logCPM(data):\n",
    "\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\")\n",
    "        data = (data/data.sum())*10**6\n",
    "        data = data.fillna(0)\n",
    "        data = np.log10(data+1)\n",
    "\n",
    "    # Return\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def log(data):\n",
    "\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\")\n",
    "        data = data.fillna(0)\n",
    "        data = np.log10(data+1)\n",
    "\n",
    "    # Return\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def qnormalization(data):\n",
    "    newdata = pd.DataFrame(quantile_transform(\n",
    "        data, axis=1, output_distribution='normal'))\n",
    "    newdata.columns = data.columns\n",
    "    newdata.index = data.index\n",
    "    return newdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalization = 'rawdata'\n",
    "if logCPM_normalization == True:  \n",
    "    data = dataset[normalization]\n",
    "    normalization += '+logCPM'\n",
    "    dataset[normalization] = logCPM(data)\n",
    "    \n",
    "if log_normalization == True:    \n",
    "    data = dataset[normalization]\n",
    "    normalization += '+log'\n",
    "    dataset[normalization] = log(data)\n",
    "    \n",
    "if z_normalization == True:\n",
    "    data = dataset[normalization]\n",
    "    normalization += '+z_norm'    \n",
    "    dataset[normalization] = data.T.apply(ss.zscore, axis=0).T.dropna()\n",
    "if q_normalization == True:\n",
    "    data = dataset[normalization]\n",
    "    normalization += '+q_norm'\n",
    "    dataset[normalization] = qnormalization(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Display datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "display(expr_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(meta_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(meta_df.groupby(meta_class_column_name).count())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Principal Component Analysis (PCA)[1] is a statistical technique used to identify global patterns in high-dimensional datasets. It is commonly used to explore the similarity of biological samples in RNA-seq datasets. To achieve this, gene expression values are transformed into Principal Components (PCs), a set of linearly uncorrelated features which represent the most relevant sources of variance in the data, and subsequently visualized using a scatter plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_pca(dataset, normalization='logCPM', nr_genes=2500, color_by='auto', color_type='categorical', filter_samples=True, plot_type='interactive'):\n",
    "    # Get data\n",
    "    before_norm = normalization.replace(\"+z_norm\", \"\").replace(\"+q_norm\", \"\")\n",
    "    top_genes = dataset[before_norm].var(axis=1).sort_values(ascending=False)\n",
    "    \n",
    "    expression_dataframe = dataset[normalization].copy()\n",
    "    \n",
    "    # Filter columns\n",
    "    if filter_samples and dataset.get('signature_metadata'):\n",
    "        selected_samples = [sample for samples in list(dataset['signature_metadata'].values())[0].values() for sample in samples]\n",
    "        expression_dataframe = expression_dataframe[selected_samples]\n",
    "\n",
    "    # Filter rows\n",
    "    expression_dataframe = expression_dataframe.loc[top_genes.index[:nr_genes]]\n",
    "    \n",
    "    # Run PCA\n",
    "    pca=PCA(n_components=3)\n",
    "    pca.fit(expression_dataframe)\n",
    "\n",
    "    # Get Variance\n",
    "    var_explained = ['PC'+str((i+1))+'('+str(round(e*100, 1))+'% var. explained)' for i, e in enumerate(pca.explained_variance_ratio_)]\n",
    "\n",
    "    # Estimate colors\n",
    "    if color_by == 'auto':\n",
    "\n",
    "        # Add signature groups\n",
    "        if dataset.get('signature_metadata'):\n",
    "            A_label, B_label = list(dataset.get('signature_metadata').keys())[0].split(' vs ')\n",
    "            col = []\n",
    "            group_dict = list(dataset.get('signature_metadata').values())[0]\n",
    "            for gsm in dataset['sample_metadata'].index:\n",
    "                if gsm in group_dict['A']:\n",
    "                    col.append(A_label)\n",
    "                elif gsm in group_dict['B']:\n",
    "                    col.append(B_label)\n",
    "                else:\n",
    "                    col.append('Other')\n",
    "            dataset['dataset_metadata']['Sample Group'] = col\n",
    "            color_by = 'Sample Group'\n",
    "        else:\n",
    "\n",
    "            # Add group column, if available\n",
    "            if 'Group' in dataset['dataset_metadata'].columns:\n",
    "                color_by = 'Group'\n",
    "            else:\n",
    "                color_by = None\n",
    "\n",
    "\n",
    "    # Return\n",
    "    pca_results = {'pca': pca, 'var_explained': var_explained, \n",
    "                   'dataset_metadata': dataset['dataset_metadata'][dataset['dataset_metadata'][meta_id_column_name] == expression_dataframe.columns], \n",
    "                   'color_by': color_by, 'color_type': color_type, 'nr_genes': nr_genes, \n",
    "                   'normalization': normalization, 'signature_metadata': dataset.get('signature_metadata'), \n",
    "                   'plot_type': plot_type}\n",
    "    return pca_results\n",
    "\n",
    "\n",
    "\n",
    "def plot_pca(pca_results):\n",
    "    pca_transformed = pca_results['pca']\n",
    "    variance_explained = pca_results['var_explained']\n",
    "    meta_df = pca_results['dataset_metadata']\n",
    "    \n",
    "    meta_df['x'] = pca_transformed.components_[0]\n",
    "    meta_df['y'] = pca_transformed.components_[1]\n",
    "    meta_df['z'] = pca_transformed.components_[2]\n",
    "\n",
    "    display(IPython.core.display.HTML('''\n",
    "            <script src=\"/static/components/requirejs/require.js\"></script>\n",
    "            <script>\n",
    "              requirejs.config({\n",
    "                paths: {\n",
    "                  base: '/static/base',\n",
    "                  plotly: 'https://cdn.plot.ly/plotly-latest.min.js?noext',\n",
    "\n",
    "                },\n",
    "              });\n",
    "            </script>\n",
    "            '''))\n",
    "\n",
    "    classes = meta_df[meta_class_column_name].unique().tolist()\n",
    "    SYMBOLS = ['circle', 'square']\n",
    "    \n",
    "    if len(classes) > 10:\n",
    "        def r(): return random.randint(0, 255)\n",
    "        COLORS = ['#%02X%02X%02X' % (r(), r(), r())\n",
    "                          for i in range(len(classes))]\n",
    "    else:\n",
    "        COLORS = [\n",
    "            '#1f77b4',\n",
    "            '#ff7f0e',\n",
    "            '#2ca02c',\n",
    "            '#d62728',\n",
    "            '#9467bd',\n",
    "            '#8c564b',\n",
    "            '#e377c2',\n",
    "            '#7f7f7f',\n",
    "            '#bcbd22',\n",
    "            '#17becf',\n",
    "            ]\n",
    "\n",
    "    data = [] # To collect all Scatter3d instances\n",
    "    for (cls), meta_df_sub in meta_df.groupby([meta_class_column_name]):\n",
    "        # Iteratate through samples grouped by class\n",
    "        display_name = '%s' % (cls)\n",
    "        # Initiate a Scatter3d instance for each group of samples specifying their coordinates\n",
    "        # and displaying attributes including color, shape, size and etc.\n",
    "        trace = go.Scatter3d(\n",
    "            x = meta_df_sub['x'],\n",
    "            y = meta_df_sub['y'],\n",
    "            z = meta_df_sub['z'],\n",
    "\n",
    "            text=meta_df_sub.index,\n",
    "            mode='markers',\n",
    "            marker=dict(\n",
    "                size=10,\n",
    "                color=COLORS[classes.index(cls)], # Color by infection status\n",
    "                opacity=.8,\n",
    "            ),\n",
    "            name=display_name,\n",
    "        )\n",
    "\n",
    "        data.append(trace)\n",
    "\n",
    "    # Configs for layout and axes\n",
    "    layout=dict(height=1000, width=1000, \n",
    "                title='3D PCA plot for samples',\n",
    "                scene=dict(\n",
    "                    xaxis=dict(title=variance_explained[0]),\n",
    "                    yaxis=dict(title=variance_explained[1]),\n",
    "                    zaxis=dict(title=variance_explained[2])\n",
    "                    )\n",
    "    )\n",
    "    fig=dict(data=data, layout=layout)\n",
    "    if plot_type == \"interactive\":\n",
    "        plotly.offline.iplot(fig)\n",
    "    else:\n",
    "        py.image.ishow(fig)\n",
    "        \n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plot_counter = 0\n",
    "# Run analysis\n",
    "results['pca'] = run_pca(dataset=dataset, nr_genes=nr_genes, normalization=normalization, plot_type=plot_type)\n",
    "\n",
    "# Display results\n",
    "plot_pca(results['pca'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Clustergrammer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clustergrammer[2] is a web-based tool for visualizing and analyzing high-dimensional data as interactive and hierarchically clustered heatmaps. It is commonly used to explore the similarity between samples in an RNA-seq dataset. In addition to identifying clusters of samples, it also allows to identify the genes which contribute to the clustering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_link(url):\n",
    "    raw_html = '<a href=\"%s\" target=\"_blank\">%s</a>' % (url, url)\n",
    "    return display(HTML(raw_html))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_clustergrammer(dataset, normalization='logCPM', z_score=True, nr_genes=1500, metadata_cols=None, filter_samples=True):\n",
    "    # Subset the expression DataFrame using top 800 genes with largest variance\n",
    "    data = dataset[normalization].copy()\n",
    "    variances = np.var(data, axis=1)\n",
    "    srt_idx = variances.argsort()[::-1]\n",
    "    expr_df_sub = data.iloc[srt_idx].iloc[:nr_genes]\n",
    "    \n",
    "    # prettify sample names\n",
    "    sample_names = ['::'.join([y, x]) for x,y in\n",
    "                       zip(meta_df[meta_class_column_name], expr_df_sub.columns)]\n",
    "    expr_df_sub.columns = sample_names\n",
    "    expr_df_sub.index = [\"Gene: \"+str(x) for x in expr_df_sub.index]\n",
    "    sample_name = [\"Sample: \"+x for x in sample_names]\n",
    "    expr_df_sub.columns = sample_name\n",
    "\n",
    "\n",
    "    treatment_type = [\"Class: \"+ x.split(\"::\")[1] for x in sample_names]\n",
    "    new_series = pd.DataFrame(treatment_type).T\n",
    "    new_series.columns = expr_df_sub.columns\n",
    "    expr_df_sub = pd.concat([new_series, expr_df_sub], axis=0)\n",
    "\n",
    "    index_list = list(expr_df_sub.index)\n",
    "    index_list = [\"\" if \"Gene\" not in str(x) else x for x in index_list]\n",
    "    expr_df_sub.index = index_list\n",
    "    expr_df_sub_file = \"expr_df_sub_file.txt\"\n",
    "    expr_df_sub.to_csv(\"expr_df_sub_file.txt\", sep='\\t')\n",
    "    # POST the expression matrix to Clustergrammer and get the URL\n",
    "    clustergrammer_url = 'https://amp.pharm.mssm.edu/clustergrammer/matrix_upload/'\n",
    "    r = requests.post(clustergrammer_url, files={'file': open(expr_df_sub_file, 'rb')}).text\n",
    "    return r\n",
    "    \n",
    "#############################################\n",
    "########## 2. Plot\n",
    "#############################################\n",
    "\n",
    "def plot_clustergrammar(clustergrammer_url):\n",
    "    clustergrammer_url = clustergrammer_url.replace(\"http:\", \"https:\")\n",
    "    display_link(clustergrammer_url)\n",
    "    # Embed\n",
    "    display(IPython.display.IFrame(clustergrammer_url, width=\"1000\", height=\"1000\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run analysis\n",
    "results['clustergrammer'] = run_clustergrammer(dataset=dataset, nr_genes=clustering_topk, normalization=normalization, z_score=True)\n",
    "\n",
    "# Display results\n",
    "plot_clustergrammar(results['clustergrammer'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Library size analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to quantify gene expression in an RNA-seq dataset, reads generated from the sequencing step are mapped to a reference genome and subsequently aggregated into numeric gene counts. Due to experimental variations and random technical noise, samples in an RNA-seq datasets often have variable amounts of the total RNA. Library size analysis calculates and displays the total number of reads mapped for each sample in the RNA-seq dataset, facilitating the identification of outlying samples and the assessment of the overall quality of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta_df['sum'] = [x/10**5 for x in expr_df.sum().tolist()]\n",
    "meta_df['sum'] = expr_df.sum().tolist()\n",
    "fig = px.bar(meta_df.sort_values(by='sum'), x=\"SAMPLE_ID\", y='sum')\n",
    "\n",
    "if plot_type=='interactive':\n",
    "        plotly.offline.iplot(fig)\n",
    "else:\n",
    "    py.image.ishow(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Differential Gene Expression "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gene expression signatures are alterations in the patterns of gene expression that occur as a result of cellular perturbations such as drug treatments, gene knock-downs or diseases. They can be quantified using differential gene expression (DGE) methods[3,4], which compare gene expression between two groups of samples to identify genes whose expression is significantly altered in the perturbation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "robjects.r('''limma <- function(rawcount_dataframe, design_dataframe, filter_genes=FALSE, adjust=\"BH\") {\n",
    "    \n",
    "    # Load packages\n",
    "    suppressMessages(require(limma))\n",
    "    suppressMessages(require(edgeR))\n",
    "\n",
    "    # Convert design matrix\n",
    "    design <- as.matrix(design_dataframe)\n",
    "\n",
    "    # Create DGEList object\n",
    "    dge <- DGEList(counts=rawcount_dataframe)\n",
    "\n",
    "    # Filter genes\n",
    "    if (filter_genes) {\n",
    "        keep <- filterByExpr(dge, design)\n",
    "        dge <- dge[keep,]\n",
    "    }\n",
    "\n",
    "    # Calculate normalization factors\n",
    "    dge <- calcNormFactors(dge)\n",
    "\n",
    "    # Run VOOM\n",
    "    v <- voom(dge, plot=FALSE)\n",
    "\n",
    "    # Fit linear model\n",
    "    fit <- lmFit(v, design)\n",
    "\n",
    "    # Make contrast matrix\n",
    "    cont.matrix <- makeContrasts(de=B-A, levels=design)\n",
    "\n",
    "    # Fit\n",
    "    fit2 <- contrasts.fit(fit, cont.matrix)\n",
    "\n",
    "    # Run DE\n",
    "    fit2 <- eBayes(fit2)\n",
    "\n",
    "    # Get results\n",
    "    limma_dataframe <- topTable(fit2, adjust=adjust, number=nrow(rawcount_dataframe))\n",
    "    limma_dataframe$gene_symbol <- rownames(limma_dataframe)\n",
    "\n",
    "    # Return\n",
    "    return(limma_dataframe)\n",
    "}\n",
    "''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_signatures(classes, dataset):\n",
    "    expr_df = dataset['rawdata']\n",
    "    meta_df = dataset[\"dataset_metadata\"]\n",
    "    signatures = dict()\n",
    "    for A, B in combinations(classes, 2):\n",
    "        group_A = meta_df.loc[meta_df[meta_class_column_name]==A, \"SAMPLE_ID\"].tolist()\n",
    "        group_B = meta_df.loc[meta_df[meta_class_column_name]==B, \"SAMPLE_ID\"].tolist()\n",
    "\n",
    "        design_dataframe = pd.DataFrame([{'index': x, 'A': int(x in group_A), 'B': int(x in group_B)} for x in expr_df.columns]).set_index('index')\n",
    "\n",
    "        processed_data = {\"expression\": expr_df, 'design': design_dataframe}\n",
    "\n",
    "        limma = robjects.r['limma']\n",
    "        signature = pandas2ri.conversion.rpy2py(limma(pandas2ri.conversion.py2rpy(processed_data['expression']), pandas2ri.conversion.py2rpy(processed_data['design']), filter_genes))\n",
    "\n",
    "        signature_label = \"vs.\".join([A, B])\n",
    "        signatures[signature_label] = signature.sort_values(\"t\", ascending=False)\n",
    "    return signatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "classes = meta_df[meta_class_column_name].unique().tolist()\n",
    "signatures = get_signatures(classes, dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for label, signature in signatures.items():\n",
    "    print(\"Label:\", label)\n",
    "    display(signature)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Volcano Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Volcano plots are a type of scatter plot commonly used to display the results of a differential gene expression analysis. They can be used to quickly identify genes whose expression is significantly altered in a perturbation, and to assess the global similarity of gene expression in two groups of biological samples. Each point in the scatter plot represents a gene; the axes display the significance versus fold-change estimated by the differential expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_2D_scatter(x, y, text='', title='', xlab='', ylab='', hoverinfo='text', color='black', colorscale='Blues', size=8, showscale=False, symmetric_x=False, symmetric_y=False, pad=0.5, hline=False, vline=False, return_trace=False, labels=False, plot_type='interactive', de_type='ma'):\n",
    "    range_x = [-max(abs(x))-pad, max(abs(x))+pad]if symmetric_x else None\n",
    "    range_y = [-max(abs(y))-pad, max(abs(y))+pad]if symmetric_y else None\n",
    "    trace = go.Scattergl(x=x, y=y, mode='markers', text=text, hoverinfo=hoverinfo, marker={'color': color, 'colorscale': colorscale, 'showscale': showscale, 'size': size})\n",
    "    if return_trace:\n",
    "        return trace\n",
    "    else:\n",
    "        if de_type == 'ma':\n",
    "            annotations = [\n",
    "                {'x': 1, 'y': 0.1, 'text':'<span style=\"color: blue; font-size: 10pt; font-weight: 600;\">Down-regulated in '+labels[-1]+'</span>', 'showarrow': False, 'xref': 'paper', 'yref': 'paper', 'xanchor': 'right', 'yanchor': 'top'},\n",
    "                {'x': 1, 'y': 0.9, 'text':'<span style=\"color: red; font-size: 10pt; font-weight: 600;\">Up-regulated in '+labels[-1]+'</span>', 'showarrow': False, 'xref': 'paper', 'yref': 'paper', 'xanchor': 'right', 'yanchor': 'bottom'}\n",
    "            ] if labels else []\n",
    "        elif de_type == 'volcano':\n",
    "            annotations = [\n",
    "                {'x': 0.25, 'y': 1.07, 'text':'<span style=\"color: blue; font-size: 10pt; font-weight: 600;\">Down-regulated in '+labels[-1]+'</span>', 'showarrow': False, 'xref': 'paper', 'yref': 'paper', 'xanchor': 'center'},\n",
    "                {'x': 0.75, 'y': 1.07, 'text':'<span style=\"color: red; font-size: 10pt; font-weight: 600;\">Up-regulated in '+labels[-1]+'</span>', 'showarrow': False, 'xref': 'paper', 'yref': 'paper', 'xanchor': 'center'}\n",
    "            ] if labels else []\n",
    "        layout = go.Layout(title=title, xaxis={'title': xlab, 'range': range_x}, yaxis={'title': ylab, 'range': range_y}, hovermode='closest', annotations=annotations)\n",
    "        fig = go.Figure(data=[trace], layout=layout)\n",
    "    \n",
    "    if plot_type=='interactive':\n",
    "        plotly.offline.iplot(fig)\n",
    "    else:\n",
    "        py.image.ishow(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_volcano(signature, signature_label, dataset, normalization='logCPM'):\n",
    "    expr_df = dataset['rawdata']\n",
    "    meta_df = dataset[\"dataset_metadata\"]\n",
    "    \n",
    "    color = []\n",
    "    text = []\n",
    "    for index, rowData in signature.iterrows():\n",
    "\n",
    "        # Text\n",
    "        text.append('<b>'+index+'</b><br>Avg Expression = '+str(round(rowData['AveExpr'], ndigits=2))+'<br>logFC = '+str(round(rowData['logFC'], ndigits=2))+'<br>p = '+'{:.2e}'.format(rowData['P.Value'])+'<br>FDR = '+'{:.2e}'.format(rowData['adj.P.Val']))\n",
    "\n",
    "        # Color\n",
    "        if rowData['P.Value'] < pvalue_threshold:\n",
    "            if rowData['logFC'] < -logfc_threshold:\n",
    "                color.append('blue')\n",
    "            elif rowData['logFC'] > logfc_threshold:\n",
    "                color.append('red')\n",
    "            else:\n",
    "                color.append('black')\n",
    "\n",
    "        else:\n",
    "            color.append('black')\n",
    "\n",
    "    volcano_plot_results = {'x': signature['logFC'], 'y': -np.log10(signature['P.Value']), 'text':text, 'color': color, 'signature_label': signature_label, 'plot_type': plot_type}\n",
    "    return volcano_plot_results\n",
    "\n",
    "def plot_volcano(volcano_plot_results):\n",
    "    spacer = ' '*50\n",
    "    plot_2D_scatter(\n",
    "        x=volcano_plot_results['x'],\n",
    "        y=volcano_plot_results['y'],\n",
    "        text=volcano_plot_results['text'],\n",
    "        color=volcano_plot_results['color'],\n",
    "        symmetric_x=True,\n",
    "        xlab='log2FC',\n",
    "        ylab='-log10P',\n",
    "        title='<b>{volcano_plot_results[signature_label]} Signature | Volcano Plot</b>'.format(**locals()),\n",
    "        labels=volcano_plot_results['signature_label'].split('vs.'),\n",
    "        plot_type=volcano_plot_results['plot_type'],\n",
    "        de_type='volcano'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize results\n",
    "results['volcano_plot'] = {}\n",
    "\n",
    "# Loop through signatures\n",
    "for label, signature in signatures.items():\n",
    "    results['volcano_plot'][label] = run_volcano(signature, label, dataset)\n",
    "    plot_volcano(results['volcano_plot'][label])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. MA Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MA plots are a type of scatter plot commonly used to display the results of a differential gene expression analysis. They can be used to quickly identify genes whose expression is significantly altered in a perturbation, and to assess the global similarity of gene expression in two groups of biological samples. Each point in the scatter plot represents a gene; the axes display the average gene expression versus fold-change estimated by the differential expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def run_maplot(signature, signature_label='', pvalue_threshold=0.05, logfc_threshold=1.5, plot_type='interactive'):\n",
    "\n",
    "    # Loop through signature\n",
    "    color = []\n",
    "    text = []\n",
    "    for index, rowData in signature.iterrows():\n",
    "\n",
    "        # Text\n",
    "        text.append('<b>'+index+'</b><br>Avg Expression = '+str(round(rowData['AveExpr'], ndigits=2))+'<br>logFC = '+str(round(rowData['logFC'], ndigits=2))+'<br>p = '+'{:.2e}'.format(rowData['P.Value'])+'<br>FDR = '+'{:.2e}'.format(rowData['adj.P.Val']))\n",
    "\n",
    "        # Color\n",
    "        if rowData['adj.P.Val'] < pvalue_threshold:\n",
    "            if rowData['logFC'] < -logfc_threshold:\n",
    "                color.append('blue')\n",
    "            elif rowData['logFC'] > logfc_threshold:\n",
    "                color.append('red')\n",
    "            else:\n",
    "                color.append('black')\n",
    "\n",
    "        else:\n",
    "            color.append('black')\n",
    "    \n",
    "    # Return \n",
    "    volcano_plot_results = {'x': signature['AveExpr'], 'y': signature['logFC'], 'text':text, 'color': color, 'signature_label': signature_label, 'plot_type': plot_type}\n",
    "    return volcano_plot_results\n",
    "\n",
    "def plot_maplot(volcano_plot_results):    \n",
    "    plot_2D_scatter(\n",
    "        x=volcano_plot_results['x'],\n",
    "        y=volcano_plot_results['y'],\n",
    "        text=volcano_plot_results['text'],\n",
    "        color=volcano_plot_results['color'],\n",
    "        symmetric_y=True,\n",
    "        xlab='Average Expression',\n",
    "        ylab='logFC',\n",
    "        title='<b>{volcano_plot_results[signature_label]} Signature | MA Plot</b>'.format(**locals()),\n",
    "        labels=volcano_plot_results['signature_label'].split('vs.'),\n",
    "        plot_type=volcano_plot_results['plot_type']\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize results\n",
    "results['ma_plot'] = {}\n",
    "\n",
    "# Loop through signatures\n",
    "for label, signature in signatures.items():\n",
    "\n",
    "    # Run analysis\n",
    "    results['ma_plot'][label] = run_maplot(signature=signature, signature_label=label, pvalue_threshold=pvalue_threshold, logfc_threshold=logfc_threshold, plot_type=plot_type)\n",
    "\n",
    "    # Display results\n",
    "    plot_maplot(results['ma_plot'][label])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Enrichr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Enrichment analysis is a statistical procedure used to identify biological terms which are over-represented in a given gene set. These include signaling pathways, molecular functions, diseases, and a wide variety of other biological terms obtained by integrating prior knowledge of gene function from multiple resources. Enrichr[5] is a web-based application which allows to perform enrichment analysis using a large collection of gene-set libraries and various interactive approaches to display enrichment results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_enrichr_geneset(geneset, label=''):\n",
    "    ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/addList'\n",
    "    genes_str = '\\n'.join(geneset)\n",
    "    payload = {\n",
    "        'list': (None, genes_str),\n",
    "        'description': (None, label)\n",
    "    }\n",
    "    response = requests.post(ENRICHR_URL, files=payload)\n",
    "    if not response.ok:\n",
    "        raise Exception('Error analyzing gene list')\n",
    "    time.sleep(0.5)\n",
    "    data = json.loads(response.text)\n",
    "    return data\n",
    "\n",
    "def run_enrichr(signature, signature_label, geneset_size=500, sort_genes_by='t'):\n",
    "\n",
    "    # Sort signature\n",
    "    signature = signature.sort_values(sort_genes_by, ascending=False)\n",
    "\n",
    "    # Get genesets\n",
    "    genesets = {\n",
    "        'upregulated': signature.index[:geneset_size],\n",
    "        'downregulated': signature.index[-geneset_size:]\n",
    "    }\n",
    "\n",
    "    # Submit to Enrichr\n",
    "    enrichr_ids = {geneset_label: submit_enrichr_geneset(geneset=geneset, label=signature_label+', '+geneset_label+', from BioJupies') for geneset_label, geneset in genesets.items()}\n",
    "    enrichr_ids['signature_label'] = signature_label\n",
    "    return enrichr_ids\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_cd(dataset, normalization):\n",
    "    meta_df = dataset['dataset_metadata'].copy()\n",
    "    expr_df = dataset[normalization]\n",
    "    d_platform_cd = {} # to top up/down genes\n",
    "    cd_results = dict()\n",
    "\n",
    "    sample_classes = {}\n",
    "    for layout1 in meta_df[meta_class_column_name].unique():\n",
    "        for layout2 in meta_df[meta_class_column_name].unique():\n",
    "            if layout1 != layout2:            \n",
    "                sample_class = np.zeros(expr_df.shape[1], dtype=np.int32)\n",
    "                sample_class[meta_df[meta_class_column_name].values == layout1] = 1\n",
    "                sample_class[meta_df[meta_class_column_name].values == layout2] = 2\n",
    "                diff_gene_set_name = \"vs.\".join([layout1, layout2])\n",
    "                sample_classes[diff_gene_set_name] = sample_class\n",
    "\n",
    "\n",
    "    for platform, sample_class in sample_classes.items():\n",
    "\n",
    "        cd_res = geode.chdir(expr_df.values, sample_class, expr_df.index, gamma=.5, sort=False, calculate_sig=False)\n",
    "        cd_coefs = np.array([x[0] for x in cd_res])\n",
    "        cd_coefs_series = pd.DataFrame(cd_coefs)\n",
    "        cd_coefs_series.index = expr_df.index\n",
    "        cd_coefs_series.columns = [\"coef\"]\n",
    "        cd_results[platform] = cd_coefs_series\n",
    "        \n",
    "    return cd_results\n",
    "\n",
    "if diff_gex_method == \"characteristic_direction\":\n",
    "    signatures = run_cd(dataset, normalization)\n",
    "    sort_genes_by = 'coef'\n",
    "else:\n",
    "    sort_genes_by = 't'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Loop through signatures\n",
    "results = {}\n",
    "results['enrichr']= {}\n",
    "\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    results['enrichr'][label] = run_enrichr(signature=signature, signature_label=label, geneset_size=gene_topk, sort_genes_by=sort_genes_by)\n",
    "    print(label+\"-up\")\n",
    "    display_link(\"https://amp.pharm.mssm.edu/Enrichr/enrich?dataset={}\".format(results['enrichr'][label][\"upregulated\"][\"shortId\"]))\n",
    "    print(label+\"-down\")\n",
    "    display_link(\"https://amp.pharm.mssm.edu/Enrichr/enrich?dataset={}\".format(results['enrichr'][label][\"downregulated\"][\"shortId\"]))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. GO Enrichment Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gene Ontology (GO)[6] is a major bioinformatics initiative aimed at unifying the representation of gene attributes across all species. It contains a large collection of experimentally validated and predicted associations between genes and biological terms. This information can be leveraged by Enrichr to identify the biological processes, molecular functions and cellular components which are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_enrichr_results(user_list_id, gene_set_libraries, overlappingGenes=True, geneset=None):\n",
    "    ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/enrich'\n",
    "    query_string = '?userListId=%s&backgroundType=%s'\n",
    "    results = []\n",
    "    for gene_set_library, label in gene_set_libraries.items():\n",
    "        response = requests.get(\n",
    "                    ENRICHR_URL +\n",
    "                       query_string % (user_list_id, gene_set_library)\n",
    "                )\n",
    "        if not response.ok:\n",
    "            raise Exception('Error fetching enrichment results')\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "        resultDataframe = pd.DataFrame(data[gene_set_library], columns=[\n",
    "                                       'rank', 'term_name', 'pvalue', 'zscore', 'combined_score', 'overlapping_genes', 'FDR', 'old_pvalue', 'old_FDR'])\n",
    "        selectedColumns = ['term_name', 'zscore', 'combined_score', 'pvalue', 'FDR'] if not overlappingGenes else [\n",
    "            'term_name', 'zscore', 'combined_score', 'FDR', 'pvalue', 'overlapping_genes']\n",
    "        resultDataframe = resultDataframe.loc[:, selectedColumns]\n",
    "        resultDataframe['gene_set_library'] = label\n",
    "        resultDataframe['log10P'] = -np.log10(resultDataframe['pvalue'])\n",
    "        results.append(resultDataframe)\n",
    "    concatenatedDataframe = pd.concat(results)\n",
    "    if geneset:\n",
    "        concatenatedDataframe['geneset'] = geneset\n",
    "    return concatenatedDataframe\n",
    "\n",
    "\n",
    "def run_go(enrichr_results, signature_label, plot_type='interactive', go_version='2018', sort_results_by='pvalue'):\n",
    "\n",
    "    # Libraries\n",
    "    go_version = str(go_version)\n",
    "    libraries = {\n",
    "        'GO_Biological_Process_'+go_version: 'Gene Ontology Biological Process ('+go_version+' version)',\n",
    "        'GO_Molecular_Function_'+go_version: 'Gene Ontology Molecular Function ('+go_version+' version)',\n",
    "        'GO_Cellular_Component_'+go_version: 'Gene Ontology Cellular Component ('+go_version+' version)'\n",
    "    }\n",
    "\n",
    "    # Get Enrichment Results\n",
    "    enrichment_results = {geneset: get_enrichr_results(enrichr_results[geneset]['userListId'], gene_set_libraries=libraries, geneset=geneset) for geneset in ['upregulated', 'downregulated']}\n",
    "    enrichment_results['signature_label'] = signature_label\n",
    "    enrichment_results['plot_type'] = plot_type\n",
    "    enrichment_results['sort_results_by'] = sort_results_by\n",
    "\n",
    "    # Return\n",
    "    return enrichment_results\n",
    "\n",
    "\n",
    "\n",
    "def plot_library_barchart(enrichr_results, gene_set_library, signature_label, sort_results_by='pvalue', nr_genesets=15, height=400, plot_type='interactive'):\n",
    "    sort_results_by = 'log10P' if sort_results_by == 'pvalue' else 'combined_score'\n",
    "    fig = tools.make_subplots(rows=1, cols=2, print_grid=False)\n",
    "    for i, geneset in enumerate(['upregulated', 'downregulated']):\n",
    "        # Get dataframe\n",
    "        enrichment_dataframe = enrichr_results[geneset]\n",
    "        plot_dataframe = enrichment_dataframe[enrichment_dataframe['gene_set_library'] == gene_set_library].sort_values(sort_results_by, ascending=False).iloc[:nr_genesets].iloc[::-1]\n",
    "\n",
    "        # Format\n",
    "        n = 7\n",
    "        plot_dataframe['nr_genes'] = [len(genes) for genes in plot_dataframe['overlapping_genes']]\n",
    "        plot_dataframe['overlapping_genes'] = ['<br>'.join([', '.join(genes[i:i+n]) for i in range(0, len(genes), n)]) for genes in plot_dataframe['overlapping_genes']]\n",
    "\n",
    "        # Get Bar\n",
    "        bar = go.Bar(\n",
    "            x=plot_dataframe[sort_results_by],\n",
    "            y=plot_dataframe['term_name'],\n",
    "            orientation='h',\n",
    "            name=geneset.title(),\n",
    "            showlegend=False,\n",
    "            hovertext=['<b>{term_name}</b><br><b>P-value</b>: <i>{pvalue:.2}</i><br><b>FDR</b>: <i>{FDR:.2}</i><br><b>Z-score</b>: <i>{zscore:.3}</i><br><b>Combined score</b>: <i>{combined_score:.3}</i><br><b>{nr_genes} Genes</b>: <i>{overlapping_genes}</i><br>'.format(**rowData) for index, rowData in plot_dataframe.iterrows()],\n",
    "            hoverinfo='text',\n",
    "            marker={'color': '#FA8072' if geneset == 'upregulated' else '#87CEFA'}\n",
    "        )\n",
    "        fig.append_trace(bar, 1, i+1)\n",
    "\n",
    "        # Get text\n",
    "        text = go.Scatter(\n",
    "            x=[max(bar['x'])/50 for x in range(len(bar['y']))],\n",
    "            y=bar['y'],\n",
    "            mode='text',\n",
    "            hoverinfo='none',\n",
    "            showlegend=False,\n",
    "            text=['*<b>{}</b>'.format(rowData['term_name']) if rowData['FDR'] < 0.1 else '{}'.format(\n",
    "                rowData['term_name']) for index, rowData in plot_dataframe.iterrows()],\n",
    "            textposition=\"middle right\",\n",
    "            textfont={'color': 'black'}\n",
    "        )\n",
    "        fig.append_trace(text, 1, i+1)\n",
    "\n",
    "    # Get annotations\n",
    "    labels = signature_label.split('vs.')\n",
    "    annotations = [\n",
    "        {'x': 0.25, 'y': 1.06, 'text': '<span style=\"color: #FA8072; font-size: 10pt; font-weight: 600;\">Up-regulated in ' +\n",
    "            labels[-1]+'</span>', 'showarrow': False, 'xref': 'paper', 'yref': 'paper', 'xanchor': 'center'},\n",
    "        {'x': 0.75, 'y': 1.06, 'text': '<span style=\"color: #87CEFA; font-size: 10pt; font-weight: 600;\">Down-regulated in ' +\n",
    "            labels[-1]+'</span>', 'showarrow': False, 'xref': 'paper', 'yref': 'paper', 'xanchor': 'center'}\n",
    "    ] if signature_label else []\n",
    "\n",
    "    # Get title\n",
    "    title = signature_label + ' | ' + gene_set_library\n",
    "\n",
    "    fig['layout'].update(height=height, title='<b>{}</b>'.format(title),\n",
    "                         hovermode='closest', annotations=annotations)\n",
    "    fig['layout']['xaxis1'].update(domain=[0, 0.49], title='-log10P' if sort_results_by == 'log10P' else 'Enrichment score')\n",
    "    fig['layout']['xaxis2'].update(domain=[0.51, 1], title='-log10P' if sort_results_by == 'log10P' else 'Enrichment score')\n",
    "    fig['layout']['yaxis1'].update(showticklabels=False)\n",
    "    fig['layout']['yaxis2'].update(showticklabels=False)\n",
    "    fig['layout']['margin'].update(l=0, t=65, r=0, b=35)\n",
    "    if plot_type=='interactive':\n",
    "        plotly.offline.iplot(fig)\n",
    "    else:\n",
    "        py.image.ishow(fig)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results['go_enrichment'] = {}\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    results['go_enrichment'][label] = run_go(results['enrichr'][label], label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for label, signature in signatures.items():\n",
    "    # Create dataframe\n",
    "    enrichment_results = results['go_enrichment'][label]\n",
    "    enrichment_dataframe = pd.concat([enrichment_results['upregulated'], enrichment_results['downregulated']])\n",
    "\n",
    "    # Plot barcharts\n",
    "    libraries = enrichment_dataframe['gene_set_library'].unique()   \n",
    "    for gene_set_library in libraries:\n",
    "        plot_library_barchart(enrichment_results, gene_set_library, enrichment_results['signature_label'], enrichment_results['sort_results_by'], nr_genesets=nr_genesets, plot_type=plot_type) # 10 300\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10. Pathway Enrichment Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Biological pathways are sequences of interactions between biochemical compounds which play a key role in determining cellular behavior. Databases such as KEGG[7], Reactome[8] and WikiPathways[9] contain a large number of associations between such pathways and genes. This information can be leveraged by Enrichr to identify the biological pathways which are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_pathway(enrichr_results, signature_label, plot_type='interactive', sort_results_by='pvalue'):\n",
    "\n",
    "    # Libraries\n",
    "    libraries = {\n",
    "        'KEGG_2016': 'KEGG Pathways',\n",
    "        'WikiPathways_2016': 'WikiPathways',\n",
    "        'Reactome_2016': 'Reactome Pathways'\n",
    "    }\n",
    "\n",
    "    # Get Enrichment Results\n",
    "    enrichment_results = {geneset: get_enrichr_results(enrichr_results[geneset]['userListId'], gene_set_libraries=libraries, geneset=geneset) for geneset in ['upregulated', 'downregulated']}\n",
    "    enrichment_results['signature_label'] = signature_label\n",
    "    enrichment_results['plot_type'] = plot_type\n",
    "    enrichment_results['sort_results_by'] = sort_results_by\n",
    "\n",
    "    # Return\n",
    "    return enrichment_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Initialize results\n",
    "results['pathway_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, enrichr_results in results['enrichr'].items():\n",
    "    # Run analysis\n",
    "    results['pathway_enrichment'][label] = run_pathway(enrichr_results=enrichr_results, signature_label=label, plot_type=plot_type, sort_results_by='pvalue')\n",
    "\n",
    "for label, signature in signatures.items():\n",
    "    # Create dataframe\n",
    "    enrichment_results = results['pathway_enrichment'][label]\n",
    "    enrichment_dataframe = pd.concat([enrichment_results['upregulated'], enrichment_results['downregulated']])\n",
    "\n",
    "    # Plot barcharts\n",
    "    libraries = enrichment_dataframe['gene_set_library'].unique()   \n",
    "    for gene_set_library in libraries:\n",
    "        # Display results\n",
    "        plot_library_barchart(enrichment_results, gene_set_library, enrichment_results['signature_label'], enrichment_results['sort_results_by'], nr_genesets=nr_genesets, plot_type=plot_type)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Transcription Factor Enrichment Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transcription Factors (TFs) are proteins involved in the transcriptional regulation of gene expression. Databases such as ChEA[10] and ENCODE[11] contain a large number of associations between TFs and their transcriptional targets. This information can be leveraged by Enrichr to identify the transcription factors whose targets are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_tf(enrichr_results, signature_label):\n",
    "    \n",
    "    # Libraries\n",
    "    libraries = {\n",
    "        'ChEA_2016': '### A. ChEA (experimentally validated targets)',\n",
    "        'ENCODE_TF_ChIP-seq_2015': '### B. ENCODE (experimentally validated targets)',\n",
    "        'ARCHS4_TFs_Coexp': '### C. ARCHS4 (coexpressed genes)'\n",
    "    }\n",
    "\n",
    "    # Initialize results\n",
    "    results = []\n",
    "\n",
    "    # Loop through genesets\n",
    "    for geneset in ['upregulated', 'downregulated']:\n",
    "\n",
    "        # Append ChEA results\n",
    "        enrichment_dataframe = get_enrichr_results(enrichr_results[geneset]['userListId'], gene_set_libraries=libraries, geneset=geneset)\n",
    "        results.append(enrichment_dataframe)\n",
    "\n",
    "    # Concatenate results\n",
    "    enrichment_dataframe = pd.concat(results)\n",
    "\n",
    "    return {'enrichment_dataframe': enrichment_dataframe, 'signature_label': signature_label}\n",
    "\n",
    "def results_table(enrichment_dataframe, source_label, target_label):\n",
    "\n",
    "    # Get libraries\n",
    "    for gene_set_library in enrichment_dataframe['gene_set_library'].unique():\n",
    "\n",
    "        # Get subset\n",
    "        enrichment_dataframe_subset = enrichment_dataframe[enrichment_dataframe['gene_set_library'] == gene_set_library].copy()\n",
    "\n",
    "        # Get unique values from source column\n",
    "        enrichment_dataframe_subset[source_label] = [x.split('_')[0] for x in enrichment_dataframe_subset['term_name']]\n",
    "        enrichment_dataframe_subset = enrichment_dataframe_subset.sort_values(['FDR', 'pvalue']).rename(columns={'pvalue': 'P-value'}).drop_duplicates(source_label)\n",
    "\n",
    "        # Add links and bold for significant results\n",
    "        enrichment_dataframe_subset[source_label] = ['<a href=\"http://www.mirbase.org/cgi-bin/query.pl?terms={x}\" target=\"_blank\">{x}</a>'.format(**locals()) if '-miR-' in x else '<a href=\"http://amp.pharm.mssm.edu/Harmonizome/gene/{x}\" target=\"_blank\">{x}</a>'.format(**locals())for x in enrichment_dataframe_subset[source_label]]\n",
    "        enrichment_dataframe_subset[source_label] = [rowData[source_label].replace('target=\"_blank\">', 'target=\"_blank\"><b>').replace('</a>', '*</b></a>') if rowData['FDR'] < 0.05 else rowData[source_label] for index, rowData in enrichment_dataframe_subset.iterrows()]\n",
    "\n",
    "        # Add rank\n",
    "        enrichment_dataframe_subset['Rank'] = ['<b>'+str(x+1)+'</b>' for x in range(len(enrichment_dataframe_subset.index))]\n",
    "\n",
    "        # Add overlapping genes with tooltip\n",
    "        enrichment_dataframe_subset['nr_overlapping_genes'] = [len(x) for x in enrichment_dataframe_subset['overlapping_genes']]\n",
    "        enrichment_dataframe_subset['overlapping_genes'] = [', '.join(x) for x in enrichment_dataframe_subset['overlapping_genes']]\n",
    "        enrichment_dataframe_subset[target_label.title()] = ['{nr_overlapping_genes} {geneset} '.format(**rowData)+target_label+'s' for index, rowData in enrichment_dataframe_subset.iterrows()]\n",
    "        # enrichment_dataframe[target_label.title()] = ['<span class=\"gene-tooltip\">{nr_overlapping_genes} {geneset} '.format(**rowData)+target_label+'s<div class=\"gene-tooltip-text\">{overlapping_genes}</div></span>'.format(**rowData) for index, rowData in enrichment_dataframe.iterrows()]\n",
    "\n",
    "        # Convert to HTML\n",
    "        pd.set_option('max.colwidth', -1)\n",
    "        html_table = enrichment_dataframe_subset.head(50)[['Rank', source_label, 'P-value', 'FDR', target_label.title()]].to_html(escape=False, index=False, classes='w-100')\n",
    "        html_results = '<div style=\"max-height: 200px; overflow-y: scroll;\">{}</div>'.format(html_table)\n",
    "\n",
    "        # Add CSS\n",
    "        display(HTML('<style>.w-100{width: 100%;} .text-left th{text-align: left !important;}</style>'))\n",
    "        display(HTML('<style>.slick-cell{overflow: visible;}.gene-tooltip{text-decoration: underline; text-decoration-style: dotted;}.gene-tooltip .gene-tooltip-text{visibility: hidden; position: absolute; left: 60%; width: 250px; z-index: 1000; text-align: center; background-color: black; color: white; padding: 5px 10px; border-radius: 5px;} .gene-tooltip:hover .gene-tooltip-text{visibility: visible;} .gene-tooltip .gene-tooltip-text::after {content: \" \";position: absolute;bottom: 100%;left: 50%;margin-left: -5px;border-width: 5px;border-style: solid;border-color: transparent transparent black transparent;}</style>'))\n",
    "\n",
    "        # Display gene set\n",
    "        print(gene_set_library)\n",
    "        # Display table\n",
    "        display(HTML(html_results))\n",
    "\n",
    "        # Download enrichment results from enrichment_dataframe\n",
    "        # download_button(enrichment_results.to_csv(sep='\\t'), 'Download Enrichment Results', 'enrichment.txt')\n",
    "        \n",
    "def display_table(tf_analysis_results):\n",
    "    \n",
    "    # Plot Table\n",
    "    results_table(tf_analysis_results['enrichment_dataframe'].copy(), source_label='Transcription Factor', target_label='target')\n",
    "\n",
    "    # Download button\n",
    "    results_dataframe = tf_analysis_results['enrichment_dataframe'].copy()\n",
    "    results_dataframe['gene_set_library'] = [x.split('. ')[-1] for x in results_dataframe['gene_set_library']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize results\n",
    "results['tf_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, enrichr_results in results['enrichr'].items():\n",
    "    print(\"Label:\", label)\n",
    "    # Run analysis\n",
    "    results['tf_enrichment'][label] = run_tf(enrichr_results=enrichr_results, signature_label=label)\n",
    "    display_table(results['tf_enrichment'][label])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 12. Kinase Enrichment Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Protein kinases are enzymes that modify other proteins by chemically adding phosphate groups. Databases such as KEA[12] contain a large number of associations between kinases and their substrates. This information can be leveraged by Enrichr to identify the protein kinases whose substrates are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_ke(enrichr_results, signature_label):\n",
    "    # Libraries\n",
    "    libraries = {\n",
    "        'KEA_2015': '### A. KEA (experimentally validated targets)',\n",
    "        'ARCHS4_Kinases_Coexp': '### B. ARCHS4 (coexpressed genes)'\n",
    "    }\n",
    "\n",
    "    # Initialize results\n",
    "    results = []\n",
    "\n",
    "    # Loop through genesets\n",
    "    for geneset in ['upregulated', 'downregulated']:\n",
    "\n",
    "        # Append ChEA results\n",
    "        enrichment_dataframe = get_enrichr_results(enrichr_results[geneset]['userListId'], gene_set_libraries=libraries, geneset=geneset)\n",
    "        results.append(enrichment_dataframe)\n",
    "\n",
    "    # Concatenate results\n",
    "    enrichment_dataframe = pd.concat(results)\n",
    "\n",
    "    return {'enrichment_dataframe': enrichment_dataframe, 'signature_label': signature_label}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize results\n",
    "results['kinase_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, enrichr_results in results['enrichr'].items():\n",
    "    print(\"Label:\", label)\n",
    "    # Run analysis\n",
    "    results['kinase_enrichment'][label] = run_ke(enrichr_results=enrichr_results, signature_label=label)\n",
    "\n",
    "    # Display results\n",
    "    display_table(results['kinase_enrichment'][label])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. miRNA Enrichment Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "microRNAs (miRNAs) are small non-coding RNA molecules which play a key role in the post-transcriptional regulation of gene expression. Databases such as TargetScan[13] and MiRTarBase[14] contain a large number of associations between miRNAs and their targets. This information can be leveraged by Enrichr to identify the miRNAs whose targets are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_mirna(enrichr_results, signature_label):\n",
    "    # Libraries\n",
    "    libraries = {\n",
    "        'TargetScan_microRNA_2017': '### A. TargetScan (experimentally validated targets)',\n",
    "        'miRTarBase_2017': '### B. miRTarBase (experimentally validated targets)'\n",
    "    }\n",
    "\n",
    "    # Initialize results\n",
    "    results = []\n",
    "\n",
    "    # Loop through genesets\n",
    "    for geneset in ['upregulated', 'downregulated']:\n",
    "\n",
    "        # Append ChEA results\n",
    "        enrichment_dataframe = get_enrichr_results(enrichr_results[geneset]['userListId'], gene_set_libraries=libraries, geneset=geneset)\n",
    "        results.append(enrichment_dataframe)\n",
    "\n",
    "    # Concatenate results\n",
    "    enrichment_dataframe = pd.concat(results)\n",
    "\n",
    "    return {'enrichment_dataframe': enrichment_dataframe, 'signature_label': signature_label}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results['mirna_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, enrichr_results in results['enrichr'].items():\n",
    "    print(\"Label:\", label)\n",
    "    \n",
    "    # Run analysis\n",
    "    results['mirna_enrichment'][label] = run_mirna(enrichr_results=enrichr_results, signature_label=label)\n",
    "\n",
    "    # Display results\n",
    "    display_table(results['mirna_enrichment'][label])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 14. L1000CDS2 Query"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "L1000CDS2[15] is a web-based tool for querying gene expression signatures against signatures created from human cell lines treated with over 20,000 small molecules and drugs for the LINCS project. It is commonly used to identify small molecules which mimic or reverse the effects of a gene expression signature generated from a differential gene expression analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_l1000cds2(signature, nr_genes=500, signature_label='', plot_type='interactive'):\n",
    "    # Define results\n",
    "    l1000cds2_results = {'signature_label': signature_label}\n",
    "\n",
    "    # Define upperGenes Function\n",
    "    upperGenes = lambda genes: [gene.upper() for gene in genes]\n",
    "\n",
    "    # Get Data\n",
    "    data = {\"upGenes\":upperGenes(signature.index[:nr_genes]),\"dnGenes\":upperGenes(signature.index[-nr_genes:])}\n",
    "\n",
    "    # Loop through aggravate:\n",
    "    for aggravate in [True, False]:\n",
    "\n",
    "        # Send to API\n",
    "        config = {\"aggravate\":aggravate,\"searchMethod\":\"geneSet\",\"share\":True,\"combination\":False,\"db-version\":\"latest\"}\n",
    "        r = requests.post('http://amp.pharm.mssm.edu/L1000CDS2/query',data=json.dumps({\"data\":data,\"config\":config}),headers={'content-type':'application/json'})\n",
    "        label = 'mimic' if aggravate else 'reverse'\n",
    "\n",
    "        # Add results\n",
    "        resGeneSet = r.json()\n",
    "        if resGeneSet.get('topMeta'):\n",
    "            l1000cds2_dataframe = pd.DataFrame(resGeneSet['topMeta'])[['cell_id', 'pert_desc', 'pert_dose', 'pert_dose_unit', 'pert_id', 'pert_time', 'pert_time_unit', 'pubchem_id', 'score', 'sig_id']].replace('-666', np.nan)\n",
    "            l1000cds2_results[label] = {'url': 'http://amp.pharm.mssm.edu/L1000CDS2/#/result/{}'.format(resGeneSet['shareId']), 'table': l1000cds2_dataframe}\n",
    "        else:\n",
    "            l1000cds2_results[label] = None\n",
    "    l1000cds2_results['plot_type'] = plot_type\n",
    "\n",
    "    # Return\n",
    "    return l1000cds2_results\n",
    "    \n",
    "def plot_l1000cds2(l1000cds2_results, nr_drugs=7, height=300):\n",
    "\n",
    "    # Check if there are results\n",
    "    if not l1000cds2_results['mimic'] or not l1000cds2_results['reverse']:\n",
    "        print('### No results were found.\\n This is likely due to the fact that the gene identifiers were not recognized by L1000CDS<sup>2</sup>. Please note that L1000CDS<sup>2</sup> currently only supports HGNC gene symbols (https://www.genenames.org/). If your dataset uses other gene identifier systems, such as Ensembl IDs or Entrez IDs, consider converting them to HGNC. Automated gene identifier conversion is currently under development.')\n",
    "        \n",
    "    else:\n",
    "        # Bar charts\n",
    "        fig = tools.make_subplots(rows=1, cols=2, print_grid=False);\n",
    "        for i, direction in enumerate(['mimic', 'reverse']):\n",
    "            drug_counts = l1000cds2_results[direction]['table'].groupby('pert_desc').size().sort_values(ascending=False).iloc[:nr_drugs].iloc[::-1]\n",
    "\n",
    "            # Get Bar\n",
    "            bar = go.Bar(\n",
    "                x=drug_counts.values,\n",
    "                y=drug_counts.index,\n",
    "                orientation='h',\n",
    "                name=direction.title(),\n",
    "                hovertext=drug_counts.index,\n",
    "                hoverinfo='text',\n",
    "                marker={'color': '#FF7F50' if direction=='mimic' else '#9370DB'}\n",
    "            )\n",
    "            fig.append_trace(bar, 1, i+1)\n",
    "            \n",
    "            # Get text\n",
    "            text = go.Scatter(\n",
    "                x=[max(bar['x'])/50 for x in range(len(bar['y']))],\n",
    "                y=bar['y'],\n",
    "                mode='text',\n",
    "                hoverinfo='none',\n",
    "                showlegend=False,\n",
    "                text=drug_counts.index,\n",
    "                textposition=\"middle right\",\n",
    "                textfont={'color': 'black'}\n",
    "            )\n",
    "            fig.append_trace(text, 1, i+1)\n",
    "\n",
    "        fig['layout'].update(height=height, title='<b>L1000CDS<sup>2</sup> {}| Small Molecule Query</b><br><i>Top small molecules</i>'.format(l1000cds2_results['signature_label']), hovermode='closest')\n",
    "        fig['layout']['xaxis1'].update(domain=[0,0.5])\n",
    "        fig['layout']['xaxis1'].update(title='<br>Count')\n",
    "        fig['layout']['xaxis2'].update(title='<br>Count')\n",
    "        fig['layout']['xaxis2'].update(domain=[0.5,1])\n",
    "        fig['layout']['yaxis1'].update(showticklabels=False)\n",
    "        fig['layout']['yaxis2'].update(showticklabels=False)\n",
    "        fig['layout']['margin'].update(l=10, t=95, r=0, b=45, pad=5)\n",
    "\n",
    "        if l1000cds2_results['plot_type'] == 'interactive':\n",
    "            plotly.offline.iplot(fig)        \n",
    "        else:\n",
    "            py.image.ishow(fig)\n",
    "\n",
    "        # Download\n",
    "        result_list = []\n",
    "        for direction, result_dict in l1000cds2_results.items():\n",
    "            if direction in ['mimic', 'reverse']:\n",
    "                df = pd.DataFrame(result_dict['table'])\n",
    "                df['direction'] = direction\n",
    "                result_list.append(df)\n",
    "#         result_txt = pd.concat(result_list).to_csv(index=False, sep='\\t')\n",
    "        # Links\n",
    "        \n",
    "        print(' *Mimic Signature Query Results*:')\n",
    "        display_link(l1000cds2_results['mimic']['url'])\n",
    "        print(' *Reverse Signature Query Results*:')\n",
    "        display_link(l1000cds2_results['reverse']['url'])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize results\n",
    "results['l1000cds2'] = {}\n",
    "\n",
    "# Loop through signatures\n",
    "for label, signature in signatures.items():    \n",
    "    # Run analysis\n",
    "    results['l1000cds2'][label] = run_l1000cds2(signature=signature, nr_genes=l1000_topk, signature_label=label, plot_type=plot_type)\n",
    "\n",
    "    # Display results\n",
    "    plot_l1000cds2(results['l1000cds2'][label], nr_drugs=nr_drugs)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 15. L1000FWD Query"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "L1000FWD[16] is a web-based tool for querying gene expression signatures against signatures created from human cell lines treated with over 20,000 small molecules and drugs for the LINCS project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_l1000fwd(signature, nr_genes=500, signature_label=''):\n",
    "    # Define results\n",
    "    l1000fwd_results = {'signature_label': signature_label}\n",
    "\n",
    "    # Define upperGenes Function\n",
    "    upperGenes = lambda genes: [gene.upper() for gene in genes]\n",
    "\n",
    "    # Get Data\n",
    "    payload = {\"up_genes\":upperGenes(signature.index[:nr_genes]),\"down_genes\":upperGenes(signature.index[-nr_genes:])}\n",
    "\n",
    "    # Get URL\n",
    "    L1000FWD_URL = 'https://amp.pharm.mssm.edu/L1000FWD/'\n",
    "\n",
    "    # Get result\n",
    "    response = requests.post(L1000FWD_URL + 'sig_search', json=payload)\n",
    "    if 'KeyError' in response.text:\n",
    "        l1000fwd_results['result_url'] = None\n",
    "    else:\n",
    "        # Get ID and URL\n",
    "        result_id = response.json()['result_id']\n",
    "        l1000fwd_results['result_url'] = 'https://amp.pharm.mssm.edu/l1000fwd/vanilla/result/'+result_id\n",
    "        l1000fwd_results['result_id'] = result_id\n",
    "\n",
    "        # Get Top\n",
    "        l1000fwd_results['signatures'] = requests.get(L1000FWD_URL + 'result/topn/' + result_id).json()\n",
    "\n",
    "\n",
    "    # Return\n",
    "    return l1000fwd_results\n",
    "    \n",
    "def plot_l1000fwd(l1000fwd_results, nr_drugs=7, height=300):\n",
    "    \n",
    "    # Check if results\n",
    "    if l1000fwd_results['result_url']:\n",
    "\n",
    "        # Display IFrame\n",
    "        display(IFrame(l1000fwd_results['result_url'], width=\"1000\", height=\"1000\"))\n",
    "\n",
    "        # Display tables\n",
    "        for direction, signature_list in l1000fwd_results['signatures'].items():\n",
    "\n",
    "            # Fix dataframe\n",
    "            rename_dict = {'sig_id': 'Signature ID', 'pvals': 'P-value', 'qvals': 'FDR', 'zscores': 'Z-score', 'combined_scores': 'Combined Score'}\n",
    "            signature_dataframe = pd.DataFrame(signature_list)[list(rename_dict.keys())].rename(columns=rename_dict).sort_values('P-value').rename_axis('Rank')\n",
    "            signature_dataframe.index = [x + 1 for x in range(len(signature_dataframe.index))]\n",
    "            signature_txt = signature_dataframe.to_csv(sep='\\t')\n",
    "\n",
    "            # Display table\n",
    "            pd.set_option('max.colwidth', -1)\n",
    "            signature_dataframe['Signature ID'] = ['<a href=\"http://amp.pharm.mssm.edu/dmoa/sig/{x}\" target=\"_blank\">{x}</a>'.format(**locals()) for x in signature_dataframe['Signature ID']]\n",
    "            table_html = signature_dataframe.to_html(escape=False, classes='w-100')\n",
    "            print('** {} Signatures: **'.format(direction.title()))\n",
    "            display(HTML('<style>.w-100{{width: 100% !important;}}</style><div style=\"max-height: 250px; overflow-y: auto; margin-bottom: 25px;\">{}</div>'.format(table_html)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Initialize results\n",
    "results['l1000fwd'] = {}\n",
    "\n",
    "# Loop through signatures\n",
    "for label, signature in signatures.items():\n",
    "    print(\"Label:\", label)\n",
    "    # Run analysis\n",
    "    results['l1000fwd'][label] = run_l1000fwd(signature=signature, signature_label=label, nr_genes=l1000_topk)\n",
    "\n",
    "    # Display results\n",
    "    plot_l1000fwd(results['l1000fwd'][label])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[1] Clark, N.R. and Ma’ayan, A. (2011) Introduction to statistical methods to analyze large data sets: principal components analysis. Sci. Signal., 4, tr3-tr3.\n",
    "<br>\n",
    "[2] Fernandez, Nicolas F., et al. \"Clustergrammer, a web-based heatmap visualization and analysis tool for high-dimensional biological data.\" Scientific data 4 (2017): 170151.\n",
    "<br>\n",
    "[3] Ritchie, Matthew E., et al. \"limma powers differential expression analyses for RNA-sequencing and microarray studies.\" Nucleic acids research 43.7 (2015): e47-e47.\n",
    "<br>\n",
    "[4] Clark, Neil R., et al. \"The characteristic direction: a geometrical approach to identify differentially expressed genes.\" BMC bioinformatics 15.1 (2014): 79.\n",
    "<br>\n",
    "[5] Kuleshov, M.V., Jones, M.R., Rouillard, A.D., Fernandez, N.F., Duan, Q., Wang, Z., Koplev, S., Jenkins, S.L., Jagodnik, K.M. and Lachmann, A. (2016) Enrichr: a comprehensive gene set enrichment analysis web server 2016 update. Nucleic acids research, 44, W90-W97.\n",
    "<br>\n",
    "[6] Ashburner, M., Ball, C.A., Blake, J.A., Botstein, D., Butler, H., Cherry, J.M., Davis, A.P., Dolinski, K., Dwight, S.S. and Eppig, J.T. (2000) Gene Ontology: tool for the unification of biology. Nature genetics, 25, 25.\n",
    "<br>\n",
    "[7] Kanehisa, M. and Goto, S. (2000) KEGG: kyoto encyclopedia of genes and genomes. Nucleic acids research, 28, 27-30.\n",
    "<br>\n",
    "[8] Croft, David, et al. \"The Reactome pathway knowledgebase.\" Nucleic acids research 42.D1 (2014): D472-D477.\n",
    "<br>\n",
    "[9] Kelder, Thomas, et al. \"WikiPathways: building research communities on biological pathways.\" Nucleic acids research 40.D1 (2012): D1301-D1307.\n",
    "<br>\n",
    "[10] Lachmann, A., Xu, H., Krishnan, J., Berger, S.I., Mazloom, A.R. and Ma'ayan, A. (2010) ChEA: transcription factor regulation inferred from integrating genome-wide ChIP-X experiments. Bioinformatics, 26, 2438-2444.\n",
    "<br>\n",
    "[11] Consortium, E.P. (2004) The ENCODE (ENCyclopedia of DNA elements) project. Science, 306, 636-640.\n",
    "<br>\n",
    "[12] Lachmann, Alexander, and Avi Ma'ayan. \"KEA: kinase enrichment analysis.\" Bioinformatics 25.5 (2009): 684-686.\n",
    "<br>\n",
    "[13] Agarwal, Vikram, et al. \"Predicting effective microRNA target sites in mammalian mRNAs.\" elife 4 (2015): e05005.\n",
    "<br>\n",
    "[14] Chou, Chih-Hung, et al. \"miRTarBase 2016: updates to the experimentally validated miRNA-target interactions database.\" Nucleic acids research 44.D1 (2016): D239-D247.\n",
    "<br>\n",
    "[15] Hafner, Marc, et al. \"L1000CDS2: LINCS L1000 characteristic direction signatures search engine.\"\n",
    "<br>\n",
    "[16] Wang, Zichen, et al. \"L1000FWD: fireworks visualization of drug-induced transcriptomic signatures.\" Bioinformatics 34.12 (2018): 2150-2152."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py37",
   "language": "python",
   "name": "py37"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
