{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b303d068",
   "metadata": {},
   "source": [
    "# GSEA Appyter\n",
    "### This appyter performs gene set enrichment analysis. A table will display the enrichment statistics for the top gene sets chosen by the input criteria (with the full results available for download as a CSV file), and an interactive Enrichment Plot and Hit Indices Plot pair will also be generated for these sets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19fe9943",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2123832",
   "metadata": {},
   "outputs": [],
   "source": [
    "#imports\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import math\n",
    "from decimal import Decimal\n",
    "\n",
    "#loading Enrichr libraries\n",
    "import requests\n",
    "import urllib.request\n",
    "\n",
    "#performing GSEA\n",
    "import gseapy as gp\n",
    "from gseapy.gsea import GSEA\n",
    "\n",
    "#creating visualizations\n",
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "from IPython.display import display, HTML, Markdown, FileLink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3213bac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide\n",
    "{% do SectionField(\n",
    "    name='gsea', \n",
    "    title='Gene Set Enrichment Analysis', \n",
    "    subtitle='Perform GSEA by inputting the following files. Follow the link to learn about the proper data formats: https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats.',\n",
    "    img='plot_icon.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='library', \n",
    "    title='Library', \n",
    "    subtitle='Choose an Enrichr gene set library or upload your own (.gmt).',\n",
    "    img='library_icon.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='data', \n",
    "    title='Data Submission', \n",
    "    subtitle='Option 1: Upload a preranked gene list (.rnk). Option 2: Upload an expression dataset with gene counts (.gct) and a phenotype labels file (.cls), and then choose a ranking method. Please make sure that the gene symbols used in the expression file match the types used in the library file for the appyter to work properly.',\n",
    "    img='data_icon.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='result', \n",
    "    title='Result Options', \n",
    "    subtitle='Choose how to get the top gene sets. Note: the appyter may not be able to support showing a large amount of gene sets.',\n",
    "    img='results_icon.png'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide\n",
    "{% set library_tab = TabField(\n",
    "    name='library_tab',\n",
    "    label='Submit Your Library',\n",
    "    default='Select an Enrichr Library',\n",
    "    description='',\n",
    "    choices={'Upload':[FileField(\n",
    "            name='geneset_filename',\n",
    "            label='Library',\n",
    "            description='GMT file containing sample gene set.',\n",
    "            default='',\n",
    "            examples={\n",
    "        'Descartes_Cell_Types_and_Tissue_2021.gmt': 'https://appyters.maayanlab.cloud/storage/GSEA_Appyter/Descartes_Cell_Types_and_Tissue_2021.gmt'}\n",
    "            )],\n",
    "        'Select an Enrichr Library': [ChoiceField(\n",
    "            name='enrichr_choice',\n",
    "            label='Enrichr Library',\n",
    "            default='Descartes_Cell_Types_and_Tissue_2021',\n",
    "            description='',\n",
    "            choices=['ARCHS4_Cell-lines',\n",
    "            'ARCHS4_IDG_Coexp',\n",
    "            'ARCHS4_Kinases_Coexp',\n",
    "            'ARCHS4_TFs_Coexp',\n",
    "            'ARCHS4_Tissues',\n",
    "            'Achilles_fitness_decrease',\n",
    "            'Achilles_fitness_increase',\n",
    "            'Aging_Perturbations_from_GEO_down',\n",
    "            'Aging_Perturbations_from_GEO_up',\n",
    "            'Allen_Brain_Atlas_10x_scRNA_2021',\n",
    "            'Allen_Brain_Atlas_down',\n",
    "            'Allen_Brain_Atlas_up',\n",
    "            'BioCarta_2013',\n",
    "            'BioCarta_2015',\n",
    "            'BioCarta_2016',\n",
    "            'BioPlanet_2019',\n",
    "            'BioPlex_2017',\n",
    "            'CCLE_Proteomics_2020',\n",
    "            'CORUM',\n",
    "            'COVID-19_Related_Gene_Sets',\n",
    "            'Cancer_Cell_Line_Encyclopedia',\n",
    "            'ChEA_2013',\n",
    "            'ChEA_2015',\n",
    "            'ChEA_2016',\n",
    "            'Chromosome_Location',\n",
    "            'Chromosome_Location_hg19',\n",
    "            'ClinVar_2019',\n",
    "            'DSigDB',\n",
    "            'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "            'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "            'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "            'Descartes_Cell_Types_and_Tissue_2021',\n",
    "            'DisGeNET',\n",
    "            'Disease_Perturbations_from_GEO_down',\n",
    "            'Disease_Perturbations_from_GEO_up',\n",
    "            'Disease_Signatures_from_GEO_down_2014',\n",
    "            'Disease_Signatures_from_GEO_up_2014',\n",
    "            'DrugMatrix',\n",
    "            'Drug_Perturbations_from_GEO_2014',\n",
    "            'Drug_Perturbations_from_GEO_down',\n",
    "            'Drug_Perturbations_from_GEO_up',\n",
    "            'ENCODE_Histone_Modifications_2013',\n",
    "            'ENCODE_Histone_Modifications_2015',\n",
    "            'ENCODE_TF_ChIP-seq_2014',\n",
    "            'ENCODE_TF_ChIP-seq_2015',\n",
    "            'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "            'ESCAPE',\n",
    "            'Elsevier_Pathway_Collection',\n",
    "            'Enrichr_Libraries_Most_Popular_Genes',\n",
    "            'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "            'Enrichr_Users_Contributed_Lists_2020',\n",
    "            'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "            'GO_Biological_Process_2013',\n",
    "            'GO_Biological_Process_2015',\n",
    "            'GO_Biological_Process_2017',\n",
    "            'GO_Biological_Process_2017b',\n",
    "            'GO_Biological_Process_2018',\n",
    "            'GO_Cellular_Component_2013',\n",
    "            'GO_Cellular_Component_2015',\n",
    "            'GO_Cellular_Component_2017',\n",
    "            'GO_Cellular_Component_2017b',\n",
    "            'GO_Cellular_Component_2018',\n",
    "            'GO_Molecular_Function_2013',\n",
    "            'GO_Molecular_Function_2015',\n",
    "            'GO_Molecular_Function_2017',\n",
    "            'GO_Molecular_Function_2017b',\n",
    "            'GO_Molecular_Function_2018',\n",
    "            'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "            'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "            'GWAS_Catalog_2019',\n",
    "            'GeneSigDB',\n",
    "            'Gene_Perturbations_from_GEO_down',\n",
    "            'Gene_Perturbations_from_GEO_up',\n",
    "            'Genes_Associated_with_NIH_Grants',\n",
    "            'Genome_Browser_PWMs',\n",
    "            'HMDB_Metabolites',\n",
    "            'HMS_LINCS_KinomeScan',\n",
    "            'HomoloGene',\n",
    "            'HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression',\n",
    "            'HumanCyc_2015',\n",
    "            'HumanCyc_2016',\n",
    "            'Human_Gene_Atlas',\n",
    "            'Human_Phenotype_Ontology',\n",
    "            'InterPro_Domains_2019',\n",
    "            'Jensen_COMPARTMENTS',\n",
    "            'Jensen_DISEASES',\n",
    "            'Jensen_TISSUES',\n",
    "            'KEA_2013',\n",
    "            'KEA_2015',\n",
    "            'KEGG_2013',\n",
    "            'KEGG_2015',\n",
    "            'KEGG_2016',\n",
    "            'KEGG_2019_Human',\n",
    "            'KEGG_2019_Mouse',\n",
    "            'KEGG_2021_Human',\n",
    "            'Kinase_Perturbations_from_GEO_down',\n",
    "            'Kinase_Perturbations_from_GEO_up',\n",
    "            'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "            'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "            'LINCS_L1000_Chem_Pert_down',\n",
    "            'LINCS_L1000_Chem_Pert_up',\n",
    "            'LINCS_L1000_Ligand_Perturbations_down',\n",
    "            'LINCS_L1000_Ligand_Perturbations_up',\n",
    "            'Ligand_Perturbations_from_GEO_down',\n",
    "            'Ligand_Perturbations_from_GEO_up',\n",
    "            'MCF7_Perturbations_from_GEO_down',\n",
    "            'MCF7_Perturbations_from_GEO_up',\n",
    "            'MGI_Mammalian_Phenotype_2013',\n",
    "            'MGI_Mammalian_Phenotype_2017',\n",
    "            'MGI_Mammalian_Phenotype_Level_3',\n",
    "            'MGI_Mammalian_Phenotype_Level_4',\n",
    "            'MGI_Mammalian_Phenotype_Level_4_2019',\n",
    "            'MSigDB_Computational',\n",
    "            'MSigDB_Hallmark_2020',\n",
    "            'MSigDB_Oncogenic_Signatures',\n",
    "            'Microbe_Perturbations_from_GEO_down',\n",
    "            'Microbe_Perturbations_from_GEO_up',\n",
    "            'Mouse_Gene_Atlas',\n",
    "            'NCI-60_Cancer_Cell_Lines',\n",
    "            'NCI-Nature_2015',\n",
    "            'NCI-Nature_2016',\n",
    "            'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "            'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "            'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "            'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "            'NURSA_Human_Endogenous_Complexome',\n",
    "            'OMIM_Disease',\n",
    "            'OMIM_Expanded',\n",
    "            'Old_CMAP_down',\n",
    "            'Old_CMAP_up',\n",
    "            'PPI_Hub_Proteins',\n",
    "            'Panther_2015',\n",
    "            'Panther_2016',\n",
    "            'Pfam_Domains_2019',\n",
    "            'Pfam_InterPro_Domains',\n",
    "            'PheWeb_2019',\n",
    "            'Phosphatase_Substrates_from_DEPOD',\n",
    "            'ProteomicsDB_2020',\n",
    "            'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "            'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "            'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "            'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "            'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "            'Reactome_2013',\n",
    "            'Reactome_2015',\n",
    "            'Reactome_2016',\n",
    "            'SILAC_Phosphoproteomics',\n",
    "            'SubCell_BarCode',\n",
    "            'SysMyo_Muscle_Gene_Sets',\n",
    "            'TF-LOF_Expression_from_GEO',\n",
    "            'TF_Perturbations_Followed_by_Expression',\n",
    "            'TG_GATES_2020',\n",
    "            'TRANSFAC_and_JASPAR_PWMs',\n",
    "            'TRRUST_Transcription_Factors_2019',\n",
    "            'Table_Mining_of_CRISPR_Studies',\n",
    "            'TargetScan_microRNA',\n",
    "            'TargetScan_microRNA_2017',\n",
    "            'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
    "            'Tissue_Protein_Expression_from_ProteomicsDB',\n",
    "            'Transcription_Factor_PPIs',\n",
    "            'UK_Biobank_GWAS_v1',\n",
    "            'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "            'VirusMINT',\n",
    "            'Virus_Perturbations_from_GEO_down',\n",
    "            'Virus_Perturbations_from_GEO_up',\n",
    "            'WikiPathway_2021_Human',\n",
    "            'WikiPathways_2013',\n",
    "            'WikiPathways_2015',\n",
    "            'WikiPathways_2016',\n",
    "            'WikiPathways_2019_Human',\n",
    "            'WikiPathways_2019_Mouse',\n",
    "            'dbGaP',\n",
    "            'huMAP',\n",
    "            'lncHUB_lncRNA_Co-Expression',\n",
    "            'miRTarBase_2017'])]\n",
    "        },\n",
    "    section='library')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed49a694",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide\n",
    "{% set data_tab = TabField(\n",
    "    name='data_tab',\n",
    "    label='Submit Your Data',\n",
    "    default='Option 1',\n",
    "    description='',\n",
    "    choices={\n",
    "        'Option 1': [FileField(\n",
    "                name='ranked_filename',\n",
    "                label='Preranked Gene List',\n",
    "                default='',\n",
    "                examples={'logFC_ranked_GSE70466.rnk': 'https://appyters.maayanlab.cloud/storage/GSEA_Appyter/logFC_ranked_GSE70466.rnk'},\n",
    "                description='RNK file containing preranked genelist.',\n",
    "                section = 'data'\n",
    "            )],\n",
    "        'Option 2': [FileField(\n",
    "                name='expression_filename',\n",
    "                label='Expression Dataset',\n",
    "                description='GCT file containing sample expression dataset.',\n",
    "                default='',\n",
    "                examples={\n",
    "        'GSE70466.gct': 'https://appyters.maayanlab.cloud/storage/GSEA_Appyter/GSE70466.gct'},\n",
    "                section='data'),\n",
    "            FileField(\n",
    "                name='phenotype_filename',\n",
    "                label='Phenotype Labels',\n",
    "                description='CLS file containing sample expression dataset.',\n",
    "                default='',\n",
    "                examples={\n",
    "        'GSE70466.cls': 'https://appyters.maayanlab.cloud/storage/GSEA_Appyter/GSE70466.cls'},\n",
    "                section='data'),\n",
    "            ChoiceField(\n",
    "                name='ranking_method',\n",
    "                label='Method for Ranking Genes',\n",
    "                choices={'Log2 Ratio of Class Means': 'log2_ratio_of_classes', 'Difference of Class Means': 'diff_of_classes', 'Ratio of Class Means (Fold Change)': 'ratio_of_classes', 'T-test': 't_test', 'Signal-to-Noise': 'signal_to_noise'},\n",
    "                default='Log2 Ratio of Class Means',\n",
    "                description='The method used to calculate a correlation or ranking.',\n",
    "                section='data')]\n",
    "            },\n",
    "        section='data')\n",
    "    %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide\n",
    "{% set result_number = IntField(\n",
    "    name='result_number',\n",
    "    label='Number of Top Gene Sets',\n",
    "    min=1,\n",
    "    max=50,\n",
    "    default=5,\n",
    "    description='The number of gene sets that will be displayed in the enrichment results table and plots.',\n",
    "    section='result')\n",
    "%}\n",
    "{% set result_criteria = ChoiceField(\n",
    "    name='result_criteria',\n",
    "    label='Criteria for Top Gene Sets',\n",
    "    choices={'Enrichment Score': 'es', 'Normalized Enrichment Score': 'nes', 'P-Value': 'pval', 'False Discovery Rate': 'fdr'},\n",
    "    default='P-Value',\n",
    "    description='The top gene sets will be calculated with the chosen criteria.',\n",
    "    section='result')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dad89d10",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%- if library_tab.raw_value == 'Upload'%}\n",
    "library_tab = 'Upload'\n",
    "library_filename = {{ library_tab.value[0] }}\n",
    "library_name = library_filename.replace('_', ' ').replace('.gmt', '')\n",
    "\n",
    "{%- else %}\n",
    "library_tab = 'Select an Enrichr Library'\n",
    "library_filename = '{{ library_tab.value[0] }}'\n",
    "library_name = '{{ library_tab.value[0] }}'\n",
    "{%- endif %}\n",
    "\n",
    "{%- if data_tab.raw_value == 'Option 1'%}\n",
    "data_tab = 'Option 1'\n",
    "ranked_filename = {{ data_tab.value[0] }}\n",
    "\n",
    "{%- else %}\n",
    "data_tab = 'Option 2'\n",
    "expression_filename = {{ data_tab.value[0] }}\n",
    "phenotype_filename = {{ data_tab.value[1] }}\n",
    "ranking_method = '{{ data_tab.value[2] }}'\n",
    "{%- endif %}\n",
    "\n",
    "result_number = {{ result_number.value }}\n",
    "result_criteria = '{{ result_criteria.value }}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f18f125b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "<h2>Loading Data</h2>\n",
    "\n",
    "This may take a while... Please allow up to around five minutes for the data to load, for inputs similar in size to the default examples and input choices. Once the data is done loading, a message will be printed underneath this section."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "412b5ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#checks if inputs are valid\n",
    "def checkInputs():\n",
    "    if data_tab == 'Option 1':\n",
    "        if library_filename == '' or ranked_filename == '':\n",
    "            raise Exception('Please submit necessary materials for Option 1 to continue.')\n",
    "        if ranked_filename.endswith('.rnk') == False:\n",
    "            raise Exception('Please upload a RNK file (ends in .rnk).')\n",
    "    else:\n",
    "        if (library_filename == '' or expression_filename == '' or\n",
    "        phenotype_filename == ''):\n",
    "            raise Exception('Please submit necessary materials for Option 2 to continue.')\n",
    "        if expression_filename.endswith('.gct') == False:\n",
    "            raise Exception('Please upload a GCT file (ends in .gct).')\n",
    "        if phenotype_filename.endswith('.cls') == False:\n",
    "            raise Exception('Please upload a CLS file (ends in .cls).')\n",
    "    if (library_tab == 'Upload' and library_filename.endswith('.gmt') \n",
    "    == False):\n",
    "        raise Exception('Please upload a GMT file (ends in .gmt).')\n",
    "\n",
    "#performs GSEA\n",
    "def gsea():\n",
    "    if data_tab == 'Option 1':\n",
    "        rnk = pd.read_csv(ranked_filename, header=None, sep=\"\\t\")\n",
    "        results = gp.prerank(rnk=rnk, gene_sets=library_filename, max_size=500)\n",
    "    else:\n",
    "        results = gp.gsea(data=expression_filename, gene_sets=library_filename,\n",
    "        cls=phenotype_filename, max_size=500)\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beb3a14b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#loads a preranked list and converts it to a dataframe for Option 1\n",
    "#creates a ranked list with chosen ranking method and converts to dataframe for Option 2\n",
    "def loadRanked():\n",
    "    if data_tab == 'Option 1':\n",
    "        values = pd.read_csv(ranked_filename, header=None, sep=\"\\t\")\n",
    "        if len(values.index) < 5000:\n",
    "            raise Exception('Current ranked gene list has less than 5000 genes. We expect all human coding genes to be in the list, which is around 20000.')\n",
    "        values.sort_values(by=values.columns[1], ascending=False, inplace=True)\n",
    "        values.reset_index(drop=True, inplace=True)\n",
    "        values.columns = ['Gene', 'Rank']\n",
    "    elif data_tab == 'Option 2':\n",
    "        pos, neg, classes = gp.parser.gsea_cls_parser(phenotype_filename)\n",
    "        obj = GSEA(data=expression_filename, gene_sets=library_filename,\n",
    "        classes=classes)\n",
    "        exp, classesDict = obj.load_data(classes)\n",
    "        values = gp.algorithm.ranking_metric(df=exp, pos=pos, neg=neg, \n",
    "        method=ranking_method, classes=classesDict, ascending=False)\n",
    "        values = pd.DataFrame({'Gene':values.index, 'Rank':values.values})\n",
    "    return values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bea4fe76",
   "metadata": {},
   "outputs": [],
   "source": [
    "#loads libraries\n",
    "def downloadLibrary(name):\n",
    "    with open(f\"{name}\", \"w\") as fw:\n",
    "        with urllib.request.urlopen(f'https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName={name}') as f:\n",
    "            for line in f.readlines():\n",
    "                fw.write(line.decode('utf-8'))\n",
    "                fw.flush()\n",
    "def loadLibrary(library_filename):\n",
    "    if library_tab == 'Select an Enrichr Library':\n",
    "        downloadLibrary(library_filename)\n",
    "    library_data = dict()\n",
    "    with open(library_filename, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "        for line in lines:\n",
    "            splited = line.strip().split(\"\\t\")\n",
    "            elements = splited[2:]\n",
    "            if len(elements) > 0:\n",
    "                # to upper case\n",
    "                library_data[splited[0]] = [x.upper() for x in elements]\n",
    "    if len(library_data) < result_number:\n",
    "        raise Exception('There cannot be less gene sets in the library than the number of gene sets displayed for the results.')\n",
    "    return library_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f57239d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter_code_exec\n",
    "#loading all needed data\n",
    "try:\n",
    "    checkInputs()\n",
    "except Exception as e:\n",
    "    print(e)\n",
    "result = gsea().results\n",
    "ranked = loadRanked()\n",
    "library_data = loadLibrary(library_filename)\n",
    "print('Data loaded successfully!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8d3f7bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "<h2>Enrichment Results Table</h2>\n",
    "\n",
    "The following table displays the enrichment analysis result statistics of the top gene sets with the chosen criteria. The columns can be re-arranged by dragging. \n",
    "\n",
    "The table is available for download as a PNG file by clicking the camera icon on the upper right. The CSV file containing the full version of the results sorted by the input criteria is available for download by clicking the link below the table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#makes values appropriate for table viewing\n",
    "def validateNumber(num):\n",
    "    num = round(num, 4)\n",
    "    if abs(num) < 0.01:\n",
    "        num = '{:.2E}'.format(Decimal(num)) #scientific notation\n",
    "    return num\n",
    "\n",
    "#create a downloadable csv file of all the results\n",
    "def create_download_link(df, title = \"Download CSV file of all results\", filename = \"GSEA_Enrichment_Results_data.csv\"):  \n",
    "    csv = df.to_csv(filename, index = False)\n",
    "    html = f'<a href=\"{filename}\" target=_blank>{title}</a>'\n",
    "    return HTML(html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61409c9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#organizes gsea results into sorted dataframe (for download)\n",
    "result_df = pd.DataFrame.from_dict(result, orient='index')\n",
    "if result_criteria == 'pval' or result_criteria == 'fdr':\n",
    "    ascent = True\n",
    "elif result_criteria == 'es' or result_criteria == 'nes':\n",
    "    ascent = False\n",
    "result_df.index.name = 'gene set'\n",
    "sorted_result_df = result_df.sort_values(by=result_criteria, ascending=ascent)\n",
    "sorted_result_df.reset_index(inplace=True)\n",
    "\n",
    "#gets top gene sets' data (calculated statistics only) needed for the table display\n",
    "top_result_df = sorted_result_df.copy(deep=True)\n",
    "top_result_df = top_result_df.drop(['geneset_size', 'matched_size', 'genes', 'ledge_genes', 'RES', 'hit_indices'], 1)\n",
    "top_result_df = top_result_df.truncate(before=1, after=result_number)\n",
    "top_result_df.set_index(keys='gene set', inplace=True)\n",
    "top_result_df = top_result_df.applymap(validateNumber)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#displays this section\n",
    "fig1 = go.Figure(data=[go.Table(columnwidth=[200, 50, 50, 50, 50],\n",
    "    header=dict(values=['Gene Set', 'ES', 'NES', 'P-value', 'FDR'], height=40), \n",
    "    cells=dict(values=[top_result_df.index, top_result_df.es, top_result_df.nes, \n",
    "    top_result_df.pval, top_result_df.fdr], height=30))])\n",
    "fig1.update_layout(width=750, font_size=14)\n",
    "fig1.show()\n",
    "display(create_download_link(sorted_result_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c99e784",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "<h2>Generating GSEA Plots</h2>\n",
    "\n",
    "For each gene set library, the top gene sets with the best chosen criteria will be plotted below. Choose which gene set to view with the drop down menu. Note that the hit indices plot will only be present for individual gene sets.\n",
    "\n",
    "Various interactive options are available at the toolbar on the upper right. Lines can be toggled on and off by clicking its name in the legend. Hovering over the running sum line on the Enrichment Plot will display the gene, its ranking on the list of genes, and the running sum score at that point. Also be aware that zooming in and out of the Enrichment Plot will simultaneously zoom for the Hit Indices plot.\n",
    "\n",
    "This plot pairing is available for download as a PNG file by clicking the camera icon on the toolbar."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#breaks strings that are too long into new lines\n",
    "def lineBreak(string):\n",
    "    words = string.split(\" \")\n",
    "    name = \"\"\n",
    "    check = \"\"\n",
    "    for i in range(len(words)):\n",
    "        check += words[i] + \" \"\n",
    "        if len(check) > 20 and i < len(words)-1:\n",
    "            name += check + \"<br>\"\n",
    "            check = \"\"\n",
    "    name += check\n",
    "    return name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "904e4361",
   "metadata": {},
   "outputs": [],
   "source": [
    "#converts necessary dataframes to list\n",
    "rankedGenes = ranked['Gene'].tolist()\n",
    "rankedCorrelation = ranked['Rank'].tolist()\n",
    "top = top_result_df.index.tolist()\n",
    "\n",
    "#initialize figure basics\n",
    "fig2 = make_subplots(rows=2, cols=1, row_heights = [0.8, 0.2],\n",
    "        shared_xaxes=True, vertical_spacing = 0.10,\n",
    "        subplot_titles=(\"Enrichment Plot\", \"Hit Indices\"))\n",
    "arbitrary = [dict(type=\"line\", xref=\"x2\", yref=\"y2\", x0=0, y0=0,\n",
    "            x1=len(rankedGenes), y1=0, line=dict(color=\"White\", width=0.5)), \n",
    "            dict(type=\"line\", xref=\"x2\", yref=\"y2\", x0=0, y0=-1,\n",
    "            x1=0, y1=1, line=dict(color=\"White\", width=0.5))]\n",
    "            #pair of arbitrary lines that maintain scale of plot\n",
    "fig2.update_layout(shapes=arbitrary, height=600, width=750, font_size=16)\n",
    "buttons = [dict(label='All Top Gene Sets', method=\"update\",\n",
    "            args=[{\"visible\": [True for i in range(len(top))]}, {\"shapes\": arbitrary}])]\n",
    "\n",
    "for geneSet in top:\n",
    "#plots values for each top gene set \n",
    "    xvals = [i for i in range(len(rankedGenes))]\n",
    "    #gets hit indices and running sum vector\n",
    "    r = gp.algorithm.enrichment_score(rankedGenes, rankedCorrelation, library_data[geneSet])\n",
    "    hits, yvals = r[2], r[3]\n",
    "    index = top.index(geneSet)\n",
    "    #creates enrichment plot\n",
    "    setName = lineBreak(geneSet)\n",
    "    fig2.add_trace(go.Scatter(x=xvals, y=yvals, mode='lines', showlegend=True,\n",
    "        name=setName, text = ['Gene: {}'.format(gene) for gene in rankedGenes],\n",
    "        hovertemplate=('%{text}' + \n",
    "        '<br>Ranking: %{x}' + \n",
    "        '<br>Running Sum: %{y}')),\n",
    "        row=1, col=1)\n",
    "    #creates hit indices\n",
    "    lines = []\n",
    "    for x in hits:\n",
    "        line = dict(type='line', x0=x, y0=-1, x1=x, y1=1, xref = 'x2', yref = 'y2')\n",
    "        lines.append(line)\n",
    "    #creates drop down menu option\n",
    "    visible = [True if i==index else False for i in range(len(top))]\n",
    "    button = dict(label=f'{geneSet}', method=\"update\",\n",
    "        args=[{\"visible\": visible}, {\"shapes\": lines}])\n",
    "    buttons.append(button)\n",
    "\n",
    "#make drop down menu\n",
    "fig2.update_layout(updatemenus=[\n",
    "    dict(buttons=buttons, pad = {\"r\": 10, \"t\": 10}, showactive = True,\n",
    "    x = 0, xanchor = \"left\", y = 1.2, yanchor = \"top\")])\n",
    "\n",
    "fig2.update_xaxes(title_text=\"Gene Rankings\", row=2, col=1)\n",
    "fig2.update_yaxes(title_text=\"Enrichment Score (ES)\", row=1, col=1)\n",
    "fig2.update_yaxes(visible=False, showticklabels=False, row=2, col=1)\n",
    "\n",
    "fig2.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "GSEA",
   "language": "python",
   "name": "gsea"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}