{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a8a5cf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31fb5373",
   "metadata": {},
   "source": [
    "# X2K Appyter\n",
    "\n",
    "The X2K Appyter (Expression2Kinases) predicts upstream regulatory networks associated with user-inputted sets of genes. Discrete query gene sets are compared first to ChEA3 libraries of transcription factor target gene sets assembled from orthogonal 'omics' datasets. Afterwards, ChEA3 results are put through a protein-protein interaction database to determine the transcription factor intermediate protein interactors. Finally, protein interactors are compared to the KEA3 background database—which contains measured and predicted kinase-substrate interactions, kinase-protein interactions, and interactions supported by co-expression and co-occurrence data—to determine which kinases may be most closely associated with the transcription factor intermediate protein interactors. Overall, Expression2Kinases can provide a method of identifying upstream regulators likely responsible for observed patterns in genome-wide gene expression.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc52c431",
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import json\n",
    "import requests\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from time import sleep\n",
    "from tabulate import tabulate\n",
    "from observable_jupyter import embed\n",
    "from IPython.display import HTML, display, Image, FileLink, Markdown\n",
    "import plotly.graph_objects as go\n",
    "import kaleido\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb86bc70",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code \n",
    "\n",
    "{% do SectionField(\n",
    "    name = 'input',\n",
    "    title = 'Gene Set Upload',\n",
    "    subtitle = 'Upload a gene set you wish to analyze',\n",
    "    img = 'images/chea3.png'\n",
    ")%}\n",
    "\n",
    "{% do SectionField(\n",
    "    name = 'analysis',\n",
    "    title = 'Analysis Filtering Options',\n",
    "    subtitle = 'Set thresholds for including top enriched transcription factors and kinases',\n",
    "    img = 'images/chea3.png'\n",
    ")%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79ff40ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code \n",
    "\n",
    "{% set gene_input = TabField(\n",
    "    name = 'gene_input',\n",
    "    label = 'Gene Set Input',\n",
    "    default = 'Paste',\n",
    "    description = 'Input your gene set',\n",
    "    choices = {\n",
    "        'Paste': [\n",
    "            TextField(\n",
    "                name = 'paste_gene_input',\n",
    "                label = 'Input Gene Set',\n",
    "                default = 'KIAA0907\\nKDM5A\\nCDC25A\\nEGR1\\nGADD45B\\nRELB\\nTERF2IP\\nSMNDC1' +\n",
    "                '\\nTICAM1\\nNFKB2\\nRGS2\\nNCOA3\\nTEX10\\nARID4B\\nCHIC2\\nFBXO11\\nMTF2\\nCDK2' +\n",
    "                '\\nDNTTIP2\\nGADD45A\\nGOLT1B\\nPOLR2K\\nNFKBIE\\nGABPB1\\nECD\\nPHKG2\\nRAD9A\\nNET1' +\n",
    "                '\\nKIAA0753\\nEZH2\\nNRAS\\nATP6V0B\\nCDK7\\nCCNH\\nSENP6\\nTIPARP\\nFOS\\nARPP19' +\n",
    "                '\\nTFAP2A\\nKDM5B\\nNPC1\\nTP53BP2\\nNUSAP1\\nSCCPDH\\nKIF20A\\nFZD7\\nUSP22\\nPIP4K2B' +\n",
    "                '\\nCRYZ\\nGNB5\\nEIF4EBP1\\nPHGDH\\nRRAGA\\nSLC25A46\\nRPA1\\nHADH\\nDAG1\\nRPIA\\nP4HA2' +\n",
    "                '\\nMACF1\\nTMEM97\\nMPZL1\\nPSMG1\\nPLK1\\nSLC37A4\\nGLRX\\nCBR3\\nPRSS23\\nNUDCD3' +\n",
    "                '\\nCDC20\\nKIAA0528\\nNIPSNAP1\\nTRAM2\\nSTUB1\\nDERA\\nMTHFD2\\nBLVRA\\nIARS2\\nLIPA' +\n",
    "                '\\nPGM1\\nCNDP2\\nBNIP3\\nCTSL1\\nCDC25B\\nHSPA8\\nEPRS\\nPAX8\\nSACM1L\\nHOXA5\\nTLE1' +\n",
    "                '\\nPYGL\\nTUBB6\\nLOXL1',\n",
    "                description = 'Input your list of genes (one gene per row)',\n",
    "                section = 'input'\n",
    "            )\n",
    "        ],\n",
    "        'Upload': [\n",
    "            FileField(\n",
    "                name = 'upload_gene_input',\n",
    "                label = 'Upload File',\n",
    "                default = '',\n",
    "                description = 'Upload your lsit of genes as a text file (one gene per row)',\n",
    "                section = 'input'\n",
    "            )\n",
    "        ]\n",
    "    },\n",
    "    section = 'input'\n",
    ")%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4654149f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% set tf_filtering = TabField(\n",
    "    name = 'tf_filtering',\n",
    "    label = 'Transcription Factoring Filtering Method',\n",
    "    default = 'Rank Threshold',\n",
    "    description = 'Choose a method to filter the transcription factors by.',\n",
    "    choices = {\n",
    "        'Rank Threshold': [\n",
    "            IntField(\n",
    "                name = 'num_tfs',\n",
    "                label = 'Top ranked transcription factors to display',\n",
    "                description = 'Input the amount of top transcription factors you wish to be analyzed',\n",
    "                default = 10,\n",
    "                min = 1,\n",
    "                max = 30,\n",
    "                section = 'input'\n",
    "            )\n",
    "        ],\n",
    "        'Score Threshold': [\n",
    "            IntField(\n",
    "                name = 'tf_score_threshold',\n",
    "                label = 'Transcription Factor Score Threshold',\n",
    "                description = 'Only transcription factors with a score equal to or below this ' +\n",
    "                'threshold (lower score is better) will be considered significant and ' +\n",
    "                'will be included.',\n",
    "                default = 75,\n",
    "                min = 1,\n",
    "                max = 150,\n",
    "                section = 'input'\n",
    "            )\n",
    "        ]\n",
    "    },\n",
    "    section = 'analysis'\n",
    ")%}\n",
    "\n",
    "{% set kin_filtering = TabField(\n",
    "    name = 'kin_filtering',\n",
    "    label = 'Kinase Filtering Method',\n",
    "    default = 'Rank Threshold',\n",
    "    description = 'Choose a method to filter the kinases by.',\n",
    "    choices = {\n",
    "        'Rank Threshold': [\n",
    "            IntField(\n",
    "                name = 'num_kins',\n",
    "                label = 'Top ranked kinases to display',\n",
    "                description = 'Input the amount of top kinases you wish to be analyzed',\n",
    "                default = 10,\n",
    "                min = 1,\n",
    "                max = 30,\n",
    "                section = 'analysis'\n",
    "            )\n",
    "        ],\n",
    "        'Score Threshold': [\n",
    "            IntField(\n",
    "                name = 'kin_score_threshold',\n",
    "                label = 'Kinase Score Threshold',\n",
    "                description = 'Only kinases with a score equal to or below this ' +\n",
    "                'threshold (lower score is better) will be considered significant and ' +\n",
    "                'will be included.',\n",
    "                default = 75,\n",
    "                min = 1,\n",
    "                max = 150,\n",
    "                section = 'analysis'\n",
    "            )\n",
    "        ]\n",
    "    },\n",
    "    section = 'analysis'\n",
    ")%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5008d59",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Might want to make this an input field? Where you can choose among a variety of databases?\n",
    "#{% set ppi_dataset = ChoiceField(\n",
    "    #name = 'ppi_dataset', \n",
    "    #label = 'Protein-Protein Interaction Database', \n",
    "    #choices = {'PCA': 'PCA', 'UMAP': 'UMAP', 't-SNE': 't-SNE'},\n",
    "    #default = 'BioGrid', \n",
    "    #description = 'Select a protein-protein interaction database to be used in analysis.', \n",
    "    #section = 'input')\n",
    "#%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e1356ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to call the ChEA3 API\n",
    "def get_chea3_results(gene_set, query_name):\n",
    "    ADDLIST_URL = 'https://amp.pharm.mssm.edu/chea3/api/enrich/'\n",
    "    payload = {\n",
    "        'gene_set': gene_set,\n",
    "        'query_name': query_name\n",
    "    }\n",
    "    response = requests.post(ADDLIST_URL, data=json.dumps(payload))\n",
    "    if not response.ok: \n",
    "        # r.ok (where r is the object) returns whether the call to the url was successful\n",
    "        raise Exception('Error analyzing gene list')\n",
    "    sleep(1)\n",
    "    return json.loads(response.text) # .text returns the content of response in unicode\n",
    "\n",
    "def get_kea3_results(gene_set, query_name):\n",
    "    ADDLIST_URL = 'https://maayanlab.cloud/kea3/api/enrich/'\n",
    "    payload = {\n",
    "        'gene_set': gene_set,\n",
    "        'query_name': query_name\n",
    "    }\n",
    "\n",
    "    response = requests.post(ADDLIST_URL, data=json.dumps(payload))\n",
    "    if not response.ok:\n",
    "        raise Exception('Error analyzing gene list')\n",
    "    sleep(1)\n",
    "\n",
    "    return json.loads(response.text)\n",
    "\n",
    "def indexfinder(lib_score_list, value):\n",
    "    index = 1\n",
    "    for num in lib_score_list:\n",
    "        if num == value:\n",
    "            return index\n",
    "        elif num != 0:\n",
    "            index += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f7c6e7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# this cell extracts the gene list, whether it was pasted in, or uploaded as a file\n",
    "\n",
    "{%- if gene_input.raw_value == 'Paste' %} # for if the gene list was pasted in\n",
    "geneset = {{ gene_input.value[0] }}\n",
    "{%- else %} # for if the gene list was uploaded as a file (only two options)\n",
    "geneset_filename = {{ gene_input.value[0] }}\n",
    "{%- endif %}\n",
    "\n",
    "{%- if tf_filtering.raw_value == 'Rank Threshold' %} \n",
    "num_tfs = {{ tf_filtering.value[0] }}\n",
    "{%- else %} # for if the gene list was uploaded as a file (only two options)\n",
    "tf_score_threshold = {{ tf_filtering.value[0] }}\n",
    "{%- endif %}\n",
    "\n",
    "{%- if kin_filtering.raw_value == 'Rank Threshold' %} \n",
    "num_kins = {{ kin_filtering.value[0] }}\n",
    "{%- else %} # for if the gene list was uploaded as a file (only two options)\n",
    "kin_score_threshold = {{ kin_filtering.value[0] }}\n",
    "{%- endif %}\n",
    "\n",
    "# also extracts the integer inputted for number of transcription factors to be returned\n",
    "\n",
    "threshold = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4e2e78a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec \n",
    "\n",
    "# This cell parses the gene list input into an organized python list \n",
    "# (when we extracted the gene list, it was still in a text file format, not a python list)\n",
    "\n",
    "{%- if gene_input.raw_value == 'Paste' %} \n",
    "genes = geneset.split('\\n')\n",
    "genes = [x.strip() for x in genes]\n",
    "{%- else %}\n",
    "open_gene_input = open(geneset_filename, 'r')\n",
    "lines = open_gene_input.readlines()\n",
    "genes = [x.strip() for x in lines]\n",
    "open_gene_input.close()\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e436c04",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### getting results from the ChEA3 API ######\n",
    "chea3results = get_chea3_results(genes, 'query')\n",
    "\n",
    "# so to clarify, \"genes\" are the user inputted genes (from the input page)\n",
    "# \"results\" is the list of transcription factors that is returned by the ChEA3 API in \n",
    "# response to the inputted genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4012a1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Read in the PPI Database being used ######\n",
    "\n",
    "ppi_url = 'https://appyters.maayanlab.cloud/storage/X2K_Appyter/BioGrid-6-23-21.csv'\n",
    "ppi = pd.read_csv(ppi_url)\n",
    "# this will need to be updated in the future so that the appyter works for people who\n",
    "# don't already have this BioGrid file on their local machine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1abbc0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Database Processing ######\n",
    "\n",
    "# This is only needed if the incoming database has no direction (TF could be in either column\n",
    "# 1 or column 2, same for their interactors). Normally, I assume column 1 is ALWAYS the TF\n",
    "# and column 2 is ALWAYS the interactor of the TF, but I'm not sure if I can assume that. \n",
    "# The code below sorts a \"directionless\" dataframe so that TFs are ALWAYs in column 1 and \n",
    "# their interactors are always in column 2\n",
    "\n",
    "#print(ppi.shape)\n",
    "#ppi_addition = pd.DataFrame({'Gene_A':ppi.Gene_B, \n",
    "                             #'Gene_B':ppi.Gene_A})\n",
    "\n",
    "#updated_ppi = pd.concat([ppi,ppi_addition], ignore_index=True)\n",
    "\n",
    "#print(updated_ppi.shape)\n",
    "#print(updated_ppi.drop_duplicates().shape)\n",
    "\n",
    "#updated_ppi.shape[0] - updated_ppi.drop_duplicates().shape[0]\n",
    "# shows that there were 3213 unique interactions with \"direction\"\n",
    "\n",
    "#updated_ppi['check'] = updated_ppi.apply(lambda row: ''.join(sorted([row['Gene_A'], \n",
    "                                                                     #row['Gene_B']])), \n",
    "                                 #axis = 1)\n",
    "\n",
    "#pupdated_ppi = updated_ppi.drop_duplicates('check').loc[:,['Gene_A', 'Gene_B']]\n",
    "\n",
    "#print(updated_ppi.shape)\n",
    "#print(pupdated_ppi.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef111c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "mr_chea3results = chea3results['Integrated--meanRank']\n",
    "\n",
    "chea3libs_sorted = ['ARCHS4 Coexpression','ENCODE ChIP-seq','Enrichr Queries',\n",
    "                    'GTEx Coexpression','Literature ChIP-seq','ReMap ChIP-seq']\n",
    "\n",
    "for i in range(len(mr_chea3results)):\n",
    "    for lib in chea3libs_sorted:\n",
    "        mr_chea3results[i].update({lib:0})\n",
    "        \n",
    "for i in range(len(mr_chea3results)):\n",
    "    thing = mr_chea3results[i]['Library'].split(';')\n",
    "    for a in range(len(thing)):\n",
    "        library, value = thing[a].split(',')\n",
    "        mr_chea3results[i].update({library:int(value)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05b67e95",
   "metadata": {},
   "outputs": [],
   "source": [
    "sortedARCHS4 = sorted(mr_chea3results, key = lambda k: k['ARCHS4 Coexpression'])\n",
    "sortedGTEx = sorted(mr_chea3results, key = lambda k: k['GTEx Coexpression']) \n",
    "sortedEnrichr = sorted(mr_chea3results, key = lambda k: k['Enrichr Queries']) \n",
    "sortedENCODE = sorted(mr_chea3results, key = lambda k: k['ENCODE ChIP-seq']) \n",
    "sortedReMap = sorted(mr_chea3results, key = lambda k: k['ReMap ChIP-seq']) \n",
    "sortedLit = sorted(mr_chea3results, key = lambda k: k['Literature ChIP-seq']) \n",
    "\n",
    "rankedARCHS4 = [entry['ARCHS4 Coexpression'] for entry in sortedARCHS4]\n",
    "rankedENCODE = [entry['ENCODE ChIP-seq'] for entry in sortedENCODE]\n",
    "rankedEnrichr = [entry['Enrichr Queries'] for entry in sortedEnrichr] \n",
    "rankedGTEx = [entry['GTEx Coexpression'] for entry in sortedGTEx]\n",
    "rankedLit = [entry['Literature ChIP-seq'] for entry in sortedLit]\n",
    "rankedReMap = [entry['ReMap ChIP-seq'] for entry in sortedReMap] \n",
    "\n",
    "\n",
    "chea3ranking_dict = {'ARCHS4 Coexpression':rankedARCHS4,\n",
    "                     'ENCODE ChIP-seq':rankedENCODE,\n",
    "                     'Enrichr Queries':rankedEnrichr,\n",
    "                     'GTEx Coexpression':rankedGTEx,\n",
    "                     'Literature ChIP-seq':rankedLit,\n",
    "                     'ReMap ChIP-seq':rankedReMap}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbd10efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "for tfentry in mr_chea3results:\n",
    "    tfentry.update( [('SumRank', 0), ('AvgRank', 0) ])\n",
    "    library_scores = tfentry['Library'].split(';')\n",
    "    lib_counter = 0\n",
    "    for a in library_scores:\n",
    "        l, v = a.split(',')\n",
    "        v = int(v)\n",
    "        scorerank = indexfinder(chea3ranking_dict[l], int(v))\n",
    "        tfentry['SumRank'] += int(scorerank)\n",
    "        lib_counter += 1\n",
    "    tfentry['AvgRank'] = (tfentry['SumRank'] / lib_counter)\n",
    "    \n",
    "sorted_chea3results = sorted(mr_chea3results, key = lambda k: k['AvgRank'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb2e0660",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    tf_score_threshold\n",
    "except NameError: \n",
    "    toptfs = []\n",
    "    tf_index = 0\n",
    "    while (len(toptfs) < num_tfs):\n",
    "        if len(sorted_chea3results[tf_index]['Library'].split(';')) >= threshold:\n",
    "            toptfs.append(sorted_chea3results[tf_index])\n",
    "        tf_index += 1\n",
    "else:\n",
    "    toptfs = [sorted_chea3results[i] for i in range(len(sorted_chea3results)) \\\n",
    "              if len(sorted_chea3results[i]['Library'].split(';')) >= threshold \\\n",
    "              if float(sorted_chea3results[i]['Score']) <= tf_score_threshold]\n",
    "#toptfs = [chea3_lib_results[i] for i in range(len(chea3_lib_results)) \\\n",
    "          #if float(chea3_lib_results[i]['Score']) <= score_threshold]\n",
    "# if a TFs score meets the threshold value, adds this TF (and its data) to the toptfs list\n",
    "\n",
    "toptfnames = [i['TF'] for i in toptfs]\n",
    "# makes a list of just the names of the top TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d91c74c",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Setting up the PPI Dictionary ######\n",
    "\n",
    "# ppi_dict is a dictionary where the key is a tf, and the value is a list of all the possible\n",
    "# interactors of this tf (based on the ppi database)\n",
    "\n",
    "ppi_dict = {}\n",
    "\n",
    "for i in range(ppi.shape[0]):\n",
    "    ppi_dict.update({ppi.Gene_A[i]:[]})\n",
    "    \n",
    "for i in range(ppi.shape[0]):\n",
    "    ppi_dict[ppi.Gene_A[i]].append(ppi.Gene_B[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68ca67e5",
   "metadata": {},
   "source": [
    "### Transcription Factors and Their Interactors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a601ae2",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Makes a dictionary for the top tfs only ######\n",
    "\n",
    "# ppi_dict is a dictionary of the entire ppi databaes\n",
    "# top_dict is that same dictionary, but filtered to only include the tfs in toptfnames\n",
    "\n",
    "top_dict = {}\n",
    "\n",
    "for name in toptfnames:\n",
    "    if ppi_dict.get(name):\n",
    "        top_dict.update({name:ppi_dict.get(name)})\n",
    "    else:\n",
    "        top_dict.update({name: 'No known interactors'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8decf89d",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Makes a list, top_tf_interactors ######\n",
    "\n",
    "# tf_interactors = list containing all the interactors per tf in toptfnames\n",
    "# top_tf_interactors = list containing all the interactors that are shared by at least 2 top tfs\n",
    "\n",
    "tf_interactors = [j for i in top_dict.values() if i != 'No known interactors' for j in i]\n",
    "\n",
    "top_tf_interactors = []\n",
    "\n",
    "for tf in tf_interactors:\n",
    "    count = 0\n",
    "    for i in top_dict.items():\n",
    "        for interactor in i[1]:\n",
    "            if interactor == tf:\n",
    "                count += 1\n",
    "    if count >= 2:\n",
    "        top_tf_interactors.append(tf)\n",
    "\n",
    "top_tf_interactors = [i for n, i in enumerate(top_tf_interactors) \n",
    "                      if i not in top_tf_interactors[:n]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "687ee7d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Now putting the top_tf_interactors through KEA3 #######\n",
    "\n",
    "kea3results = get_kea3_results(top_tf_interactors, 'query')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74a798ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Getting Ready To Investigate the Top Ranked Kinases ######\n",
    "\n",
    "mr_kea3results = kea3results['Integrated--meanRank']\n",
    "\n",
    "\n",
    "\n",
    "kea3libs_sorted = ['BioGRID', 'ChengKSIN', 'ChengPPI', 'HIPPIE', 'mentha', \n",
    "                   'MINT', 'PhosDAll', 'prePPI', 'PTMsigDB', 'STRING', 'STRING.bind']\n",
    "\n",
    "for i in range(len(mr_kea3results)):\n",
    "    for lib in kea3libs_sorted:\n",
    "        mr_kea3results[i].update({lib:0})\n",
    "        \n",
    "for i in range(len(mr_kea3results)):\n",
    "    thing = mr_kea3results[i]['Library'].split(';')\n",
    "    for a in range(len(thing)):\n",
    "        library, value = thing[a].split(',')\n",
    "        mr_kea3results[i].update({library:int(value)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51676b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Sorting Kinases According to Rank in Each Library #######\n",
    "\n",
    "sortedPTMsigDB = sorted(mr_kea3results, key = lambda k: k['PTMsigDB'])\n",
    "sortedprePPI = sorted(mr_kea3results, key = lambda k: k['prePPI']) \n",
    "sortedmentha = sorted(mr_kea3results, key = lambda k: k['mentha']) \n",
    "sortedMINT = sorted(mr_kea3results, key = lambda k: k['MINT']) \n",
    "sortedSTRING = sorted(mr_kea3results, key = lambda k: k['STRING']) \n",
    "sortedHIPPIE = sorted(mr_kea3results, key = lambda k: k['HIPPIE']) \n",
    "sortedChengKSIN = sorted(mr_kea3results, key = lambda k: k['ChengKSIN']) \n",
    "sortedBioGRID = sorted(mr_kea3results, key = lambda k: k['BioGRID']) \n",
    "sortedPhosDAll = sorted(mr_kea3results, key = lambda k: k['PhosDAll']) \n",
    "sortedSTRINGbind = sorted(mr_kea3results, key = lambda k: k['STRING.bind']) \n",
    "sortedChengPPI = sorted(mr_kea3results, key = lambda k: k['ChengPPI']) \n",
    "\n",
    "rankedPTMsigDB = [entry['PTMsigDB'] for entry in sortedPTMsigDB]\n",
    "rankedprePPI = [entry['prePPI'] for entry in sortedprePPI]\n",
    "rankedmentha = [entry['mentha'] for entry in sortedmentha]\n",
    "rankedMINT = [entry['MINT'] for entry in sortedMINT]\n",
    "rankedSTRING = [entry['STRING'] for entry in sortedSTRING]\n",
    "rankedHIPPIE = [entry['HIPPIE'] for entry in sortedHIPPIE]\n",
    "rankedChengKSIN = [entry['ChengKSIN'] for entry in sortedChengKSIN]\n",
    "rankedBioGRID = [entry['BioGRID'] for entry in sortedBioGRID]\n",
    "rankedPhosDAll = [entry['PhosDAll'] for entry in sortedPhosDAll] \n",
    "rankedSTRINGbind = [entry['STRING.bind'] for entry in sortedSTRINGbind]\n",
    "rankedChengPPI = [entry['ChengPPI'] for entry in sortedChengPPI]\n",
    "\n",
    "# Composes a dictionary of all the kinases where each entry is the kinases ranked\n",
    "# according to the respective library (the library name is the key in the dictionar)\n",
    "kea3ranking_dict = {'PTMsigDB':rankedPTMsigDB,\n",
    "                    'prePPI':rankedprePPI,\n",
    "                    'mentha':rankedmentha,\n",
    "                    'MINT':rankedMINT,\n",
    "                    'STRING':rankedSTRING,\n",
    "                    'HIPPIE':rankedHIPPIE,\n",
    "                    'ChengKSIN':rankedChengKSIN, \n",
    "                    'BioGRID':rankedBioGRID, \n",
    "                    'PhosDAll':rankedPhosDAll, \n",
    "                    'STRING.bind':rankedSTRINGbind, \n",
    "                    'ChengPPI':rankedChengPPI}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcc2c32c",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Sorting the kinases by AvgRank #######\n",
    "\n",
    "for tfentry in mr_kea3results:\n",
    "    tfentry.update( [('SumRank', 0), ('AvgRank', 0) ])\n",
    "    library_scores = tfentry['Library'].split(';')\n",
    "    lib_counter = 0\n",
    "    for a in library_scores:\n",
    "        l, v = a.split(',')\n",
    "        v = int(v)\n",
    "        scorerank = indexfinder(kea3ranking_dict[l], int(v))\n",
    "        tfentry['SumRank'] += int(scorerank)\n",
    "        lib_counter += 1\n",
    "    tfentry['AvgRank'] = (tfentry['SumRank'] / lib_counter)\n",
    "    \n",
    "sorted_kea3results = sorted(mr_kea3results, key = lambda k: k['AvgRank'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab16e707",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Getting the List of Sorted Kinases ######\n",
    "\n",
    "try:\n",
    "    kin_score_threshold\n",
    "except NameError:\n",
    "    sorted_topkinresults = []\n",
    "    kin_index = 0\n",
    "    while (len(sorted_topkinresults) < num_kins):\n",
    "        if len(sorted_kea3results[kin_index]['Library'].split(';')) >= threshold:\n",
    "            sorted_topkinresults.append(sorted_kea3results[kin_index])\n",
    "        kin_index += 1\n",
    "else:\n",
    "    sorted_topkinresults = [mr_kea3results[i] for i in range(len(sorted_kea3results)) \\\n",
    "                            if len(sorted_kea3results[i]['Library'].split(';')) >= threshold \\\n",
    "                            if float(mr_kea3results[i]['Score']) <= kin_score_threshold]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a26aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "topkinnames = [i['TF'] for i in sorted_topkinresults]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9c86ab4",
   "metadata": {},
   "source": [
    "# Network Summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1621d1e5",
   "metadata": {},
   "source": [
    "## Transcription Factors and Their Interactors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be4fe3fa",
   "metadata": {},
   "source": [
    "### Highest Ranked Transcription Factors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b59f5a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(HTML(', '.join(toptfnames)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f9ffdff",
   "metadata": {},
   "outputs": [],
   "source": [
    "c_lib_palette = {'ARCHS4 Coexpression':'rgb(196, 8, 8)',\n",
    "                 'ENCODE ChIP-seq':'rgb(244, 109, 67)',\n",
    "                 'Enrichr Queries':'rgb(242, 172, 68)', \n",
    "                 'GTEx Coexpression':'rgb(236, 252, 68)',\n",
    "                 'Literature ChIP-seq':'rgb(165, 242, 162)',\n",
    "                 'ReMap ChIP-seq':'rgb(92, 217, 78)'}\n",
    "\n",
    "tf_amt = len(toptfs)\n",
    "\n",
    "c_lib_means = {'ARCHS4 Coexpression': [0] * tf_amt, 'ENCODE ChIP-seq': [0] * tf_amt, \n",
    "               'Enrichr Queries': [0] * tf_amt, 'GTEx Coexpression': [0] * tf_amt,\n",
    "               'Literature ChIP-seq': [0] * tf_amt, 'ReMap ChIP-seq': [0] * tf_amt}\n",
    "\n",
    "toptfs = toptfs[::-1]\n",
    "\n",
    "# set up a list with all the TFs, sorted by rank (lowest to highest, in line with top_results)\n",
    "sorted_tfs = []\n",
    "for i in range(0, len(toptfs)):\n",
    "    sorted_tfs.append(toptfs[i].get('TF'))    \n",
    "    \n",
    "for i, tfentry in enumerate(toptfs):\n",
    "    libscores = tfentry['Library'].split(';')\n",
    "    for a in libscores:\n",
    "        lib, value = a.split(',')\n",
    "        rank = indexfinder(chea3ranking_dict[lib], int(value))\n",
    "        avg = tfentry['AvgRank']\n",
    "        tot = tfentry['SumRank']\n",
    "        bar_length = (rank*avg)/tot\n",
    "        c_lib_means[lib][i] = float(bar_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfab6fd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting the actual bar chart\n",
    "chea3_fig = go.Figure(data = [go.Bar(name = c_lib, \n",
    "                                     x = c_lib_means[c_lib], \n",
    "                                     y = sorted_tfs,\n",
    "                                     marker = go.bar.Marker(color = c_lib_palette[c_lib]), \n",
    "                                     orientation = 'h') \n",
    "                              for c_lib in chea3libs_sorted])\n",
    "\n",
    "chea3_fig.update_layout(barmode = 'stack')\n",
    "chea3_fig.update_layout(\n",
    "    title = {\n",
    "        'text': 'Stacked Bar Chart of Average Ranks in Different Libraries',\n",
    "        'y': 0.87,\n",
    "        'x': 0.5,\n",
    "        'xanchor': 'center',\n",
    "        'yanchor': 'top',\n",
    "    },\n",
    "    xaxis_title = 'Average of Ranks Across All Libraries',\n",
    "    yaxis_title = 'Transcription Factors',\n",
    "    font = dict(\n",
    "        size = 16,\n",
    "        color = 'black'\n",
    "    )\n",
    ")\n",
    "\n",
    "chea3_fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16deae11",
   "metadata": {},
   "outputs": [],
   "source": [
    "chea3_mr_table = [0] * len(toptfs)\n",
    "\n",
    "chea3_mr_counter = 0\n",
    "# set up a counter for indexing (to fill mrtable)\n",
    "\n",
    "tbl_toptfs = toptfs[::-1]\n",
    "\n",
    "for i in range(len(tbl_toptfs)):\n",
    "    tbl_toptfs[i]['Rank'] = i + 1\n",
    "    \n",
    "for i in tbl_toptfs:\n",
    "    chea3_mr_table[chea3_mr_counter] = [i['Rank'], \n",
    "                                        i['TF'], \n",
    "                                        i['AvgRank'], \n",
    "                                        i['Library'].replace(',', ': ').replace(';', ', '),\n",
    "                                        f\"{', '.join(i['Overlapping_Genes'].split(',')[0:10])}, ...\"]\n",
    "    # filling mrtable, using replace() to reformat a bit\n",
    "    # [0:10] for Overlapping_Genes so that only 10 Overlapping Genes are shown (with '...' after)\n",
    "    chea3_mr_counter += 1\n",
    "\n",
    "# now actually making/printing the table\n",
    "display(HTML(tabulate(chea3_mr_table, ['Rank', \n",
    "                                       'Transcription Factor', \n",
    "                                       'Avg Rank', \n",
    "                                       'Library Ranks',\n",
    "                                       'Overlapping Genes'], \n",
    "                      tablefmt='html')))\n",
    "\n",
    "\n",
    "chea3_tsv_name = 'transcriptionfactors.tsv'\n",
    "with open(chea3_tsv_name, 'w') as tsv_file:\n",
    "    tsv_file.write(tabulate(chea3_mr_table, ['Rank', \n",
    "                                             'Transcription Factor',\n",
    "                                             'Avg Rank', \n",
    "                                             'Library Ranks', \n",
    "                                             'Overlapping Genes'], \n",
    "                            tablefmt='tsv'))\n",
    "    display(HTML(f'<a href=\"{chea3_tsv_name}\">Download table in .tsv</a>'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b4cb6cf",
   "metadata": {},
   "source": [
    "### Transcription Factor Interactors\n",
    "\n",
    "\"Transcription factor interactors\" are any molecules that interact with the given transcription factor according to the protein-protein interaction database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abaa8c96",
   "metadata": {},
   "outputs": [],
   "source": [
    "for name in toptfnames:\n",
    "    if ppi_dict.get(name):\n",
    "        display(HTML(f\"{name} interacts with {len(ppi_dict.get(name))} gene(s)/protein(s):\"))\n",
    "        display(HTML(', '.join(ppi_dict.get(name))))\n",
    "    else:\n",
    "        display(HTML(f\"{name} has no known interactors.\"))\n",
    "    display(HTML(f\"<hr>\"))\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b11ad33",
   "metadata": {},
   "source": [
    "## Relevant Transcription Factor Interactors\n",
    "\n",
    "\"Relevant transcription factor interactors\" are transcription factor interactors that are shared among at least 2 of the top trasncription factors themselves."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0b22f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(HTML(f\"There are {len(top_tf_interactors)} relevant transcription factor interactors:\"))\n",
    "display(HTML(', '.join(top_tf_interactors)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "683fe5cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "rtf_fortable = []\n",
    "    \n",
    "for i in range(len(top_tf_interactors)):\n",
    "    rtf_fortable.append([i+1, top_tf_interactors[i]])\n",
    "    \n",
    "display(HTML(tabulate(rtf_fortable, ['','TF Interactors'], tablefmt='html')))\n",
    "\n",
    "rtf_tsv_name = 'tf_interactors.tsv'\n",
    "with open(rtf_tsv_name, 'w') as tsv_file:\n",
    "    tsv_file.write(tabulate(rtf_fortable, ['','TF Interactors'], tablefmt='tsv'))\n",
    "    display(HTML(f'<a href=\"{rtf_tsv_name}\">Download table in .tsv</a>'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c728e3bb",
   "metadata": {},
   "source": [
    "## Kinases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab45e96b",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(HTML(', '.join(topkinnames)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09ae3de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Preparing to Plot the Kinases Bar Chart ######\n",
    "\n",
    "k_lib_palette = {'BioGRID': 'rgb(196, 8, 8)', 'ChengKSIN': 'rgb(244, 109, 67)', \n",
    "                 'ChengPPI': 'rgb(242, 172, 68)','HIPPIE': 'rgb(236, 252, 68)', \n",
    "                 'mentha': 'rgb(165, 242, 162)', 'MINT': 'rgb(92, 217, 78)',\n",
    "                 'PhosDAll': 'rgb(0, 138, 64)', 'prePPI': 'rgb(96, 191, 235)', \n",
    "                 'PTMsigDB': 'rgb(14, 130, 201)', 'STRING': 'rgb(58, 50, 168)', \n",
    "                 'STRING.bind': 'rgb(158, 50, 168)'}\n",
    "# this sets all the color values for all the libraries that will be displayed in the bar chart\n",
    "\n",
    "# NOTE: removed Integrated mean/topRank since those are compiled from the above 6 libraries \n",
    "# afterwards and so none of the TFs will have Integrated mean/topRank as one of their libraries\n",
    "\n",
    "kin_amt = len(sorted_topkinresults)\n",
    "\n",
    "k_lib_means = {'STRING.bind': [0] * kin_amt, 'ChengPPI': [0] * kin_amt, \n",
    "               'PhosDAll': [0] * kin_amt, 'BioGRID': [0] * kin_amt, \n",
    "               'HIPPIE': [0] * kin_amt, 'ChengKSIN': [0] * kin_amt,\n",
    "               'STRING': [0] * kin_amt, 'MINT': [0] * kin_amt, \n",
    "               'mentha': [0] * kin_amt, 'prePPI': [0] * kin_amt,\n",
    "               'PTMsigDB': [0] * kin_amt}\n",
    "# creates a dictionary where each library is a key, and the values are empty lists with as\n",
    "# many indices/spaces as the user has requested transcription factors (ex: if the user\n",
    "# requests 15 TFs to be returned, the lists will have 15 spaces)\n",
    "\n",
    "# need to reverse the order for plotting (go.Figure plots \"backwards\" sort of)    \n",
    "sorted_topkinresults = sorted_topkinresults[::-1]\n",
    "\n",
    "sorted_kins = []\n",
    "for i in range(0, len(sorted_topkinresults)):\n",
    "    sorted_kins.append(sorted_topkinresults[i].get('TF'))\n",
    "    # this pulls only the TF name from top_results and adds it to sorted_tfs\n",
    "\n",
    "for i, tfentry in enumerate(sorted_topkinresults):\n",
    "    libscores = tfentry['Library'].split(';')\n",
    "    for a in libscores:\n",
    "        lib, value = a.split(',')\n",
    "        rank = indexfinder(kea3ranking_dict[lib], int(value))\n",
    "        avg = tfentry['AvgRank']\n",
    "        tot = tfentry['SumRank']\n",
    "        bar_length = (rank*avg)/tot\n",
    "        k_lib_means[lib][i] = float(bar_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f81b23e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Plotting the mean rank data ######\n",
    "\n",
    "kea3_fig = go.Figure(data = [go.Bar(name = k_lib, \n",
    "                                    x = k_lib_means[k_lib], \n",
    "                                    y = sorted_kins,\n",
    "                                    marker = go.bar.Marker(color = k_lib_palette[k_lib]), \n",
    "                                    orientation = 'h') \n",
    "                             for k_lib in kea3libs_sorted])\n",
    "\n",
    "kea3_fig.update_layout(barmode = 'stack')\n",
    "kea3_fig.update_layout(\n",
    "    title = {\n",
    "        'text': 'Stacked Bar Chart of Average Ranks in Different Libraries',\n",
    "        'y': 0.87,\n",
    "        'x': 0.5,\n",
    "        'xanchor': 'center',\n",
    "        'yanchor': 'top',\n",
    "    },\n",
    "    xaxis_title = 'Average of Ranks Across All Libraries',\n",
    "    yaxis_title = 'Kinases',\n",
    "    font = dict(\n",
    "        size = 16,\n",
    "        color = 'black'\n",
    "    )\n",
    ")\n",
    "\n",
    "kea3_fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "889ed70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "kea3_mr_table = [0] * len(sorted_topkinresults)\n",
    "\n",
    "kea3_mr_counter = 0\n",
    "# set up a counter for indexing (to fill mrtable)\n",
    "\n",
    "# need to \"re-reverse\" the ordering for the table (so now we're back to normal ordering)\n",
    "tbl_sorted_topkinresults = sorted_topkinresults[::-1]\n",
    "for i in range(0, len(tbl_sorted_topkinresults)):\n",
    "    tbl_sorted_topkinresults[i]['Rank'] = i + 1\n",
    "\n",
    "for i in tbl_sorted_topkinresults:\n",
    "    kea3_mr_table[kea3_mr_counter] = [i['Rank'], \n",
    "                                      i['TF'], \n",
    "                                      i['AvgRank'], \n",
    "                                      i['Library'].replace(',', ': ').replace(';', ', '),\n",
    "                                      f\"{', '.join(i['Overlapping_Genes'].split(',')[0:10])}, ...\"]\n",
    "    # filling mr_table, using replace() to reformat a bit\n",
    "    # [0:10] for Overlapping_Genes so that only 10 Overlapping Genes are shown (with '...' after)\n",
    "    kea3_mr_counter += 1\n",
    "                        \n",
    "display(HTML(tabulate(kea3_mr_table, ['Rank', \n",
    "                                      'Kinase', \n",
    "                                      'Avg Rank', \n",
    "                                      'Library Ranks',\n",
    "                                      'Overlapping Genes'], \n",
    "                      tablefmt='html')))\n",
    "\n",
    "\n",
    "kea3_tsv_name = 'kinases.tsv'\n",
    "with open(kea3_tsv_name, 'w') as tsv_file:\n",
    "    tsv_file.write(tabulate(kea3_mr_table, ['Rank', \n",
    "                                            'Kinase',\n",
    "                                            'Avg Rank', \n",
    "                                            'Library Ranks', \n",
    "                                            'Overlapping Genes'], \n",
    "                            tablefmt='tsv'))\n",
    "    display(HTML(f'<a href=\"{kea3_tsv_name}\">Download table in .tsv</a>'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "593c4d71",
   "metadata": {},
   "outputs": [],
   "source": [
    "x2k_visualizer_input = {'nodes':[], 'interactions':[]}\n",
    "for tfentry in toptfs:\n",
    "    x2k_visualizer_input['nodes'].append({'name': tfentry['TF'],\n",
    "                                          'type': 'tf',\n",
    "                                          'pvalue': -1})\n",
    "    \n",
    "for kinentry in sorted_topkinresults:\n",
    "    x2k_visualizer_input['nodes'].append({'name': kinentry['TF'],\n",
    "                                          'type': 'kinase',\n",
    "                                          'pvalue': -1})\n",
    "for tfientry in top_tf_interactors:\n",
    "    x2k_visualizer_input['nodes'].append({'name': tfientry,\n",
    "                                          'type': 'other',\n",
    "                                          'pvalue': -1})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c65d424e",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Creating 3 Dictionaries For Indexing, One for Each Protein Category ######\n",
    "\n",
    "index_counter = 0\n",
    "\n",
    "x2k_visualizer_tfindices = {}\n",
    "for i in range(0, len(toptfnames)):\n",
    "    x2k_visualizer_tfindices.update({x2k_visualizer_input['nodes'][i].get('name'):i})\n",
    "    index_counter = i + 1\n",
    "\n",
    "x2k_visualizer_kinindices = {}\n",
    "for i in range(index_counter, index_counter + len(sorted_topkinresults)):\n",
    "    x2k_visualizer_kinindices.update({x2k_visualizer_input['nodes'][i].get('name'):i})\n",
    "    index_counter = i + 1\n",
    "    \n",
    "x2k_visualizer_intindices = {}\n",
    "for i in range(index_counter, index_counter + len(top_tf_interactors)):\n",
    "    x2k_visualizer_intindices.update({x2k_visualizer_input['nodes'][i].get('name'):i})\n",
    "\n",
    "print(x2k_visualizer_tfindices)\n",
    "print(x2k_visualizer_kinindices)\n",
    "print(x2k_visualizer_intindices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "428f65f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Connecting TFs and Interactors ######\n",
    "\n",
    "for tf in toptfnames:\n",
    "    for i in top_dict.get(tf):\n",
    "        if top_dict.get(tf) != 'No known interactors':\n",
    "            tfint_interaction = {'source': x2k_visualizer_tfindices.get(tf),\n",
    "                                 'target': x2k_visualizer_intindices.get(i)}\n",
    "        if i in top_tf_interactors and tfint_interaction not in x2k_visualizer_input['interactions']:\n",
    "            x2k_visualizer_input['interactions'].append(tfint_interaction)\n",
    "            \n",
    "x2k_visualizer_input['interactions']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc4e6435",
   "metadata": {},
   "outputs": [],
   "source": [
    "overlapping_genelist1 = [i['Overlapping_Genes'] for i in sorted_topkinresults]\n",
    "\n",
    "overlapping_genelist2 = []\n",
    "for genelist in overlapping_genelist1:\n",
    "    gene = genelist.split(',')\n",
    "    overlapping_genelist2.append(gene)\n",
    "    \n",
    "# overlapping_genelist2 is a list of lists, where the list indices correspond to \n",
    "# a single kinases index in sorted_topkinresults (index 3 of sorted_topkinresults will\n",
    "# have the overlapping genes in index 3 of overlapping_genelist2)\n",
    "\n",
    "int_kin_interactions = {}\n",
    "for kin_genelist in overlapping_genelist2:\n",
    "    for gene in kin_genelist:\n",
    "        int_kin_interactions.update({gene:[]})\n",
    "\n",
    "new_sorted_topkinresults = [i.copy() for i in sorted_topkinresults]\n",
    "for i in range(len(new_sorted_topkinresults)):\n",
    "    new_sorted_topkinresults[i].update({'Overlapping_Genes':overlapping_genelist2[i]})\n",
    "\n",
    "for i in int_kin_interactions.keys():\n",
    "    for j in new_sorted_topkinresults:\n",
    "        if i in j['Overlapping_Genes']:\n",
    "            int_kin_interactions[i].append(j['TF'])\n",
    "\n",
    "\n",
    "int_kin_interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c81202b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in top_tf_interactors:\n",
    "    for b in int_kin_interactions.get(a):\n",
    "        intkin_interaction = {'source': x2k_visualizer_intindices.get(a),\n",
    "                              'target': x2k_visualizer_kinindices.get(b)}\n",
    "        if intkin_interaction not in x2k_visualizer_input['interactions']:\n",
    "            x2k_visualizer_input['interactions'].append(intkin_interaction)\n",
    "            \n",
    "            \n",
    "x2k_visualizer_input['interactions']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25563990",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_x2k_visualizer_input = {}\n",
    "full_x2k_visualizer_input['X2K'] = x2k_visualizer_input\n",
    "\n",
    "full_x2k_visualizer_input.update([ ('ChEA', []), ('KEA', []), ('G2N', {}), ('input', []) ])\n",
    "full_x2k_visualizer_input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8de190a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "embed('@maxim-k/x2k-network', cells=['x2k'], inputs={'json': full_x2k_visualizer_input})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
