{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30e2f31d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f1685d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name='primary',\n",
    "    title='Target Identifier',\n",
    "    img='target_identifier_logo.jpg'\n",
    ") %}\n",
    "\n",
    "{% do DescriptionField(\n",
    "    name='data_file_description',\n",
    "    text='''\n",
    "    If you are uploading scRNA-seq data, please only include cells from a single cluster.  Files should be a tsv/csv of the form:<br />\n",
    "    <table class=\"table\">\n",
    "    <tr>\n",
    "      <td>&nbsp;</td>\n",
    "      <th>Replicate 1</th>\n",
    "      <th>Replicate 2</th>\n",
    "      <th>...</th>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>Gene|Transcript 1</th>\n",
    "      <td>0</td>\n",
    "      <td>200</td>\n",
    "      <td>...</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>Gene|Transcript 2</th>\n",
    "      <td>5</td>\n",
    "      <td>180</td>\n",
    "      <td>...</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>...</th>\n",
    "      <td>...</td>\n",
    "      <td>...</td>\n",
    "      <td>...</td>\n",
    "    </tr>\n",
    "    </table>''',\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set file = FileField(\n",
    "    name='input_expression',\n",
    "    label='Input RNA-seq expression vectors, such as from a tumor',\n",
    "    description='Gene/Transcripts on the rows, replicates on the columns',\n",
    "    default='GSE49155-lung-squamous-cell-carcinoma.tsv',\n",
    "    required=True,\n",
    "    examples={\n",
    "        'GSE49155-lung-squamous-cell-carcinoma.tsv': 'https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/GSE49155-patient.tsv',\n",
    "    },\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set input_transcript_level = BoolField(\n",
    "    name='input_transcript_level',\n",
    "    label='Whether the input RNA-seq expression vectors are at the level of transcripts or genes',\n",
    "    default=False,\n",
    "    yes_label='Transcript Level',\n",
    "    no_label='Gene Level',\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set input_single_cell = BoolField(\n",
    "    name='input_single_cell',\n",
    "    label='Whether the input is scRNA-seq or bulk RNA-seq',\n",
    "    default=False,\n",
    "    yes_label='scRNA-seq',\n",
    "    no_label='Bulk RNA-seq',\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set background = TabField(\n",
    "    name='background',\n",
    "    label='Normal tissue background',\n",
    "    description='Input expression will be contrasted against this background',\n",
    "    default='Precomputed',\n",
    "    choices={\n",
    "        'Precomputed': [\n",
    "            ChoiceField(\n",
    "                name='background_dataset',\n",
    "                label='Normal bulk RNA or scRNA-seq background',\n",
    "                description='Choose Gene or Transcript Background variant if your input vectors are at the Gene or Transcript level.',\n",
    "                choices={\n",
    "                    'GTEx (bulk RNA-seq) - Gene': '\"s3://storage/Tumor_Gene_Target_Screener/gtex-gene-stats.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True)',\n",
    "                    'GTEx (bulk RNA-seq) - Transcript': '\"s3://storage/Tumor_Gene_Target_Screener/gtex-transcript-stats.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True)',\n",
    "                    'ARCHS4 Anatomy (bulk RNA-seq) - Gene': '\"s3://storage/Tumor_Gene_Target_Screener/archs4-gene-anatomy-stats.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True)',\n",
    "                    'ARCHS4 Anatomy (bulk RNA-seq) - Transcript': '\"s3://storage/Tumor_Gene_Target_Screener/archs4-transcript-anatomy-stats.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True)',\n",
    "                    'ARCHS4 Extra (bulk RNA-seq) - Gene': '\"s3://storage/Tumor_Gene_Target_Screener/archs4-gene-extra-stats.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True)',\n",
    "                    'ARCHS4 Extra (bulk RNA-seq) - Transcript': '\"s3://storage/Tumor_Gene_Target_Screener/archs4-transcript-extra-stats.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True)',\n",
    "                    'ARCHS4 New (bulk RNA-seq) - Gene': '\"s3://storage/Tumor_Gene_Target_Screener/archs4-gene-stats.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True)',\n",
    "                    'ARCHS4 New (bulk RNA-seq) - Transcript': '\"s3://storage/Tumor_Gene_Target_Screener/archs4-transcript-stats.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True)',\n",
    "                    'Tabula Sapiens (scRNA-seq) - Gene': '\"s3://storage/Tumor_Gene_Target_Screener/ts_10x_cell-ontology-class_donors_tissue-labels_v1.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True)',\n",
    "                    'Human Cell Atlas (scRNA-seq) - Gene': '\"s3://storage/Tumor_Gene_Target_Screener/hca_10x_donors_tissue-labels_v1.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True)',\n",
    "                },\n",
    "                default='GTEx (bulk RNA-seq) - Gene',\n",
    "            ),\n",
    "        ],\n",
    "        'Custom': [\n",
    "            FileField(\n",
    "                name='background_upload',\n",
    "                label='Normal tissue background',\n",
    "                description='Given a matrix (d), genes or transcripts by samples, this matrix can be constructed with `d.T.groupby(sample_tissue_mappings).description().T`',\n",
    "                examples={\n",
    "                    'GTEx': 'https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/gtex-gene-stats.tsv',\n",
    "                },\n",
    "                default=None,\n",
    "                section='primary',\n",
    "            ),\n",
    "            BoolField(\n",
    "                name='background_transcript_level',\n",
    "                label='Whether this file is at the level of transcripts or genes',\n",
    "                default=False,\n",
    "                yes_label='Transcript Level',\n",
    "                no_label='Gene Level',\n",
    "            ),\n",
    "            BoolField(\n",
    "                name='background_single_cell',\n",
    "                label='Whether this file is scRNA-seq or bulk RNA-seq',\n",
    "                default=False,\n",
    "                yes_label='scRNA-seq',\n",
    "                no_label='Bulk RNA-seq',\n",
    "            ),\n",
    "            DescriptionField(\n",
    "                name='background_file_description',\n",
    "                text='''\n",
    "                Files should be a tsv/csv of the form:<br />\n",
    "                <table class=\"table\">\n",
    "                <tr>\n",
    "                  <td>&nbsp;</td>\n",
    "                  <td>&nbsp;</td>\n",
    "                  <th>Tissue|Tissue-Cell Type 1</th>\n",
    "                  <th>...</th>\n",
    "                </tr>\n",
    "                <tr>\n",
    "                  <th>Gene|Transcript 1</th>\n",
    "                  <th>25%</th>\n",
    "                  <td>0</td>\n",
    "                  <td>...</td>\n",
    "                </tr>\n",
    "                <tr>\n",
    "                  <th>...</th>\n",
    "                  <th>...</th>\n",
    "                  <td>...</td>\n",
    "                  <td>...</td>\n",
    "                </tr>\n",
    "                <tr>\n",
    "                  <th>Gene|Transcript n</th>\n",
    "                  <th>mean</th>\n",
    "                  <td>180</td>\n",
    "                  <td>...</td>\n",
    "                </tr>\n",
    "                </table>''',\n",
    "            ),\n",
    "        ],\n",
    "    },\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set background_dataset = background.value[0] %}\n",
    "{% set background_transcript_level = (\n",
    "    background.value[0].raw_value.endswith('- Transcript')\n",
    ") if background.raw_value == 'Precomputed' else (\n",
    "    background.value[1].raw_value or False\n",
    ") %}\n",
    "\n",
    "{% set background_single_cell = (\n",
    "    'scRNA-seq' in background.value[0].raw_value\n",
    ") if background.raw_value == 'Precomputed' else (\n",
    "    background.value[2].raw_value or False\n",
    ") %}\n",
    "\n",
    "{% set membrane_screener = BoolField(\n",
    "    name='membrane_screener',\n",
    "    label='Prioritize membrane genes',\n",
    "    description='Use membranome to identify membrane',\n",
    "    default=True,\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set normalize_to_background = BoolField(\n",
    "    name='normalize_to_background',\n",
    "    label='Normalize to background distribution',\n",
    "    default=True,\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set proteomics_vis = BoolField(\n",
    "    name='proteomics_vis',\n",
    "    label='Show protein expression profiles of gene candidates',\n",
    "    description='View protein expression levels in normal tissues, from Human Proteome Map and Human Protein Atlas proteomics data',\n",
    "    default=True,\n",
    "    section='primary',\n",
    ") %}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "007d75b1",
   "metadata": {},
   "source": [
    "# Overexpressed Candidate Identification\n",
    "\n",
    "This appyter uses RNA-seq expression data from the input, such as a tumor, and identifies over-expressed proteins verses a baseline dataset of either normal tissues (such those in GTEx or ARCHS4) or normal single-cells (such as those in Tabula Sapiens or Human Cell Atlas). It then prioritizes candidates by significance and targetability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b33fd1e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "import qnorm\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import warnings\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.graph_objects as go\n",
    "from IPython.display import HTML, display, Markdown\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib_venn import venn2\n",
    "from maayanlab_bioinformatics.normalization import zscore_normalize, log2_normalize\n",
    "from maayanlab_bioinformatics.dge import limma_voom_differential_expression\n",
    "from maayanlab_bioinformatics.harmonization.ncbi_genes import ncbi_genes_lookup\n",
    "lookup = ncbi_genes_lookup()\n",
    "\n",
    "import sys\n",
    "import contextlib\n",
    "@contextlib.contextmanager\n",
    "def suppress_output(stdout=True, stderr=True, dest='/dev/null'):\n",
    "    ''' Usage:\n",
    "    with suppress_output():\n",
    "        print('hi')\n",
    "    '''\n",
    "    dev_null = open(dest, 'a')\n",
    "    if stdout:\n",
    "        _stdout = sys.stdout\n",
    "        sys.stdout = dev_null\n",
    "    if stderr:\n",
    "        _stderr = sys.stderr\n",
    "        sys.stderr = dev_null\n",
    "    try:\n",
    "        yield\n",
    "    finally:\n",
    "        if stdout:\n",
    "            sys.stdout = _stdout\n",
    "        if stderr:\n",
    "            sys.stderr = _stderr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "254f8067",
   "metadata": {},
   "source": [
    "## Load Input RNA-seq Expression Data\n",
    "\n",
    "Load RNA-seq expression data for the input."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b13e82ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_table(filename):\n",
    "    if filename.endswith('.tsv') or filename.endswith('.tsv.gz'):\n",
    "        return pd.read_csv(filename, sep='\\t', index_col=0)\n",
    "    elif filename.endswith('.csv') or filename.endswith('.csv.gz'):\n",
    "        return pd.read_csv(filename, sep=',', index_col=0)\n",
    "    elif filename.endswith('.gct') or filename.endswith('.gct.gz'):\n",
    "        return pd.read_csv(filename, sep='\\t', index_col=0, skiprows=2)\n",
    "    else:\n",
    "        return pd.read_table(filename, sep=None, engine='python', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5000377",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "df_expr = read_table({{ file }})\n",
    "{% if input_single_cell.raw_value %}\n",
    "df_expr = df_expr.sum(axis=1).to_frame()\n",
    "{% endif %}\n",
    "{% if input_transcript_level.raw_value %}\n",
    "df_expr_transcripts = df_expr.index.map(lambda idx: idx.partition('.')[0])\n",
    "df_expr = df_expr.groupby(df_expr_transcripts, observed=True).sum()\n",
    "{% else %}\n",
    "df_expr_genes = df_expr.index.astype(str).map(lambda idx: lookup(idx.partition('.')[0]))\n",
    "df_expr = df_expr.groupby(df_expr_genes, observed=True).median()\n",
    "{% endif %}\n",
    "df_expr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9bf2192",
   "metadata": {},
   "source": [
    "## Load Background Dataset\n",
    "\n",
    "The background dataset contains expression for normal tissues across many healthy individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cfe29da",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "df_bg_stats = pd.read_csv({{ background_dataset }}, sep='\\t', index_col=[0,1])\n",
    "{% if background_transcript_level %}\n",
    "df_bg_transcripts = df_bg_stats.unstack().index.map(lambda idx: idx.partition('.')[0])\n",
    "df_bg_stats = df_bg_stats.unstack().groupby(df_bg_transcripts, observed=True).sum().stack()\n",
    "{% else %}\n",
    "df_bg_genes = df_bg_stats.unstack().index.map(lambda idx: lookup(idx.partition('.')[0]))\n",
    "df_bg_stats = df_bg_stats.unstack().groupby(df_bg_genes, observed=True).median().stack()\n",
    "{% endif %}\n",
    "df_bg_expr = df_bg_stats.loc[(slice(None), ['25%', '50%', '75%']), :].unstack()\n",
    "df_bg_expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b68ca30",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if background_transcript_level or input_transcript_level.raw_value %}\n",
    "## Load Transcript <=> Gene Mappings\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "761244b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "{% if background_transcript_level or input_transcript_level.raw_value %}\n",
    "df_transcript_gene_map = pd.read_csv(\"s3://storage/Tumor_Gene_Target_Screener/transcript-gene-map.tsv.gz\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True), sep='\\t', header=0, index_col=0, compression='gzip')\n",
    "df_transcript_gene_map\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d270adcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if membrane_screener.raw_value %}\n",
    "## Load Membrane Proteins for Screening\n",
    "\n",
    "Membrane proteins are ideal targets, we can get these from [Membranome](https://membranome.org/) among other places.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2852e5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "{% if membrane_screener.raw_value %}\n",
    "proteins = pd.read_csv('https://lomize-group-membranome.herokuapp.com/proteins?fileFormat=csv')\n",
    "proteins = proteins[proteins['species_name_cache'] == 'Homo sapiens']\n",
    "membrane_proteins = proteins['genename'].map(lookup).dropna()\n",
    "membrane_proteins\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74c487bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if background_transcript_level and input_transcript_level.raw_value %}\n",
    "## Background (Transcript) - Input (Transcript) Gene Interoperability\n",
    "\n",
    "Several features of this appyter operate at the gene level, thus while we will find significant transcripts, we\n",
    "will use the associated gene for the other features.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a1f97c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if background_transcript_level and not input_transcript_level.raw_value %}\n",
    "## Background (Transcript) - Input (Gene) Interoperability\n",
    "\n",
    "As the input RNA-seq expression data is at the gene level, we will map the\n",
    "background to genes for differential expression but later highlight significant transcripts in the background.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cec25b5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "{% if background_transcript_level and not input_transcript_level.raw_value %}\n",
    "df_bg_expr = df_bg_expr.groupby(df_transcript_gene_map['gene_symbol']).sum().groupby(lookup).median()\n",
    "df_bg_expr\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c05fcbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if not background_transcript_level and input_transcript_level.raw_value %}\n",
    "## Background (Gene) - Input (Transcript) Interoperability\n",
    "\n",
    "Although the input RNA-seq expression data is at the transcript level, the background is only at the gene level.\n",
    "This is not ideal, but we can still highlight individual transcripts in the input that are over-expressed\n",
    "when compared to the average background expression across all transcripts.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e56bb2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "{% if not background_transcript_level and input_transcript_level.raw_value %}\n",
    "# \"melt\" from wide format to long format with the columns: (index, type, stat, value)\n",
    "df_bg_expr_melted = df_bg_expr.melt(ignore_index=False)\n",
    "# merge with transcript_gene_map will transform the index from gene_symbol to ensembl_transcript_id\n",
    "#  duplicating entries for each statistic accordingly\n",
    "df_bg_expr_melted_mapped = pd.merge(\n",
    "    left=df_bg_expr_melted, left_index=True,\n",
    "    right=df_transcript_gene_map, right_on='gene_symbol',\n",
    ")\n",
    "# \n",
    "df_bg_expr = df_bg_expr_melted_mapped.pivot(columns=['variable_0', 'variable_1'], values='value')\n",
    "df_bg_expr\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0763ba3a",
   "metadata": {},
   "source": [
    "## Distribution matching between input sample(s) & the background\n",
    "\n",
    "We show the median gene expression distribution in the input and in the background before and after normalization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "416950d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ((ax11, ax12), (ax21, ax22)) = plt.subplots(2, 2)\n",
    "log2_normalize(df_expr).median(axis=1).hist(bins=100, ax=ax11)\n",
    "ax11.set_title('Median Expression')\n",
    "ax11.set_ylabel('Input')\n",
    "log2_normalize(df_expr).median(axis=0).hist(bins=100, ax=ax12)\n",
    "ax12.set_title('Median Sample Expression')\n",
    "log2_normalize(df_bg_expr).median(axis=1).hist(bins=100, ax=ax21)\n",
    "ax21.set_ylabel('Background')\n",
    "log2_normalize(df_bg_expr).median(axis=0).hist(bins=100, ax=ax22)\n",
    "ax21.set_xlabel('$log_2(count)$')\n",
    "ax22.set_xlabel('$log_2(count)$')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "606d0718",
   "metadata": {},
   "source": [
    "The following venn diagram shows the gene/transcript overlap between the input and the background, ideally it should be very high, otherwise additional identifier mapping may be required. If the two are completely disjoint, you've most likely incorrectly labeled your data as gene or transcript data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "739dd065",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_index = list(set(df_expr.index) & set(df_bg_expr.index))\n",
    "venn2([set(df_expr.index), set(df_bg_expr.index)],\n",
    "      ['Input  ', '  Background'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d2e8a1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if normalize_to_background.raw_value %}\n",
    "Here we use quantile normalization [1] to align the expression with the background's median distribution. The implications of this has not been fully explored though presumably given that we have no experimental controls to properly normalize, it should be better than nothing.\n",
    "\n",
    "[1] B.M. Bolstad, R.A Irizarry, M. Åstrand, T.P. Speed, A comparison of normalization methods for high density oligonucleotide array data based on variance and bias, Bioinformatics, Volume 19, Issue 2, 22 January 2003, Pages 185–193, https://doi.org/10.1093/bioinformatics/19.2.185\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f3911b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if normalize_to_background.raw_value %}\n",
    "target_distribution = df_bg_expr.loc[common_index, :].median(axis=1)\n",
    "df_expr_norm = qnorm.quantile_normalize(df_expr.loc[common_index, :], target=target_distribution)\n",
    "df_bg_expr_norm = qnorm.quantile_normalize(df_bg_expr.loc[common_index, :], target=target_distribution)\n",
    "\n",
    "fig, ((ax11, ax12), (ax21, ax22)) = plt.subplots(2, 2)\n",
    "log2_normalize(df_expr_norm).median(axis=1).hist(bins=100, ax=ax11)\n",
    "ax11.set_title('Median Expression')\n",
    "ax11.set_ylabel('Input')\n",
    "log2_normalize(df_expr_norm).median(axis=0).hist(bins=100, ax=ax12)\n",
    "ax12.set_title('Median Sample Expression')\n",
    "log2_normalize(df_bg_expr_norm).median(axis=1).hist(bins=100, ax=ax21)\n",
    "ax21.set_ylabel('Background')\n",
    "log2_normalize(df_bg_expr_norm).median(axis=0).hist(bins=100, ax=ax22)\n",
    "ax21.set_xlabel('$log_2(count)$')\n",
    "ax22.set_xlabel('$log_2(count)$')\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "{% else %}\n",
    "print('Warning: Proceeding without normalization')\n",
    "df_expr_norm = df_expr.loc[common_index, :]\n",
    "df_bg_expr_norm = df_bg_expr.loc[common_index, :]\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2a12359",
   "metadata": {},
   "source": [
    "## Perform Differential Expression between Input & Background\n",
    "\n",
    "We use voom-limma to identify significantly differentially expressed genes between the background healthy expression quantiles and the input expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1bc6df1",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "with suppress_output():\n",
    "    dge = limma_voom_differential_expression(\n",
    "        df_bg_expr_norm, df_expr_norm,\n",
    "        voom_design=True,\n",
    "    )\n",
    "    {% if input_transcript_level.raw_value %}\n",
    "    dge['ensembl_transcript_id'] = dge.index\n",
    "    dge['gene_symbol'] = df_transcript_gene_map.loc[dge.index, 'gene_symbol'].apply(lambda g: lookup(g) or g)\n",
    "    dge['label'] = dge.apply(lambda r: f\"{r['ensembl_transcript_id']} - {r['gene_symbol']}\", axis=1)\n",
    "    {% else %}\n",
    "    dge['gene_symbol'] = dge.index\n",
    "    dge['label'] = dge.index\n",
    "    {% endif %}\n",
    "dge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12778291",
   "metadata": {},
   "source": [
    "## Narrow Down Candidate Set\n",
    "\n",
    "We identify significantly differentially expressed genes with logFC- t-statistic products which significant deviate from the mean, or equivalently, those points which are furthest from the volcano plot origin."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eec43be",
   "metadata": {},
   "outputs": [],
   "source": [
    "dge['-log(adj.P.Val)'] = -np.log(dge['adj.P.Val'])\n",
    "prod = (np.abs(dge['t']) * dge['logFC'])\n",
    "dge['is_deg'] = dge['adj.P.Val'] < 0.05\n",
    "dge['is_significant'] = prod > prod.mean() + 3 * prod.std()\n",
    "dge['score'] = dge['is_deg'].astype(int) + dge['is_significant'].astype(int)\n",
    "#\n",
    "fig = go.Figure()\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[~dge.is_deg, 'logFC'],\n",
    "    y=dge.loc[~dge.is_deg, '-log(adj.P.Val)'],\n",
    "    name='Other',\n",
    "    showlegend=False,\n",
    "    marker=dict(\n",
    "        color='black',\n",
    "    )\n",
    "))\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[dge.is_deg & ~dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[dge.is_deg & ~dge.is_significant, '-log(adj.P.Val)'],\n",
    "    text=dge.loc[dge.is_deg & ~dge.is_significant, 'label'],\n",
    "    name='Differentially Expressed',\n",
    "    marker=dict(\n",
    "        color='#ffdddd',\n",
    "    )\n",
    "))\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[dge.is_significant, '-log(adj.P.Val)'],\n",
    "    text=dge.loc[dge.is_significant, 'label'],\n",
    "    name='Significantly Far from Origin',\n",
    "    marker=dict(\n",
    "        color='red',\n",
    "    )\n",
    "))\n",
    "fig.update_layout(\n",
    "    title='Background vs Input Differential Expression',\n",
    "    xaxis_title='Log Fold Change',\n",
    "    yaxis_title='-Log[Adjusted P-Value]',\n",
    "    autosize=True,\n",
    ")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5726ef67",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if membrane_screener.raw_value %}\n",
    "dge['is_membrane'] = np.in1d(dge['gene_symbol'], membrane_proteins)\n",
    "dge['score'] = dge['score'] + dge['is_membrane'].astype(int)\n",
    "#\n",
    "fig = go.Figure()\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[~dge.is_membrane&~dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[~dge.is_membrane&~dge.is_significant, '-log(adj.P.Val)'],\n",
    "    name='Other',\n",
    "    showlegend=False,\n",
    "    marker=dict(\n",
    "        color='black',\n",
    "    )\n",
    "))\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[~dge.is_membrane&dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[~dge.is_membrane&dge.is_significant, '-log(adj.P.Val)'],\n",
    "    text=dge.loc[~dge.is_membrane&dge.is_significant, 'label'],\n",
    "    name='Significant Other',\n",
    "    showlegend=False,\n",
    "    marker=dict(\n",
    "        color='red',\n",
    "    )\n",
    "))\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[dge.is_membrane&~dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[dge.is_membrane&~dge.is_significant, '-log(adj.P.Val)'],\n",
    "    name='Membrane Protein',\n",
    "    marker=dict(\n",
    "        color='grey',\n",
    "    )\n",
    "))\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[dge.is_membrane&dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[dge.is_membrane&dge.is_significant, '-log(adj.P.Val)'],\n",
    "    text=dge.loc[dge.is_membrane&dge.is_significant, 'label'],\n",
    "    name='Significant Membrane Protein',\n",
    "    marker=dict(\n",
    "        color='orange',\n",
    "    )\n",
    "))\n",
    "fig.update_layout(\n",
    "    title='Background vs Input',\n",
    "    xaxis_title='Log Fold Change',\n",
    "    yaxis_title='-Log[Adjusted P-Value]',\n",
    "    autosize=True,\n",
    ")\n",
    "fig.show()\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff262708",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "dge_final = dge[dge.score >= 1].sort_values(['score', 't'], ascending=False).iloc[:16]\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "dge_final['Link'] = dge_final['gene_symbol'].map(lambda g: f\"<a href=\\\"https://cfde-gene-pages.cloud/gene/{g}\\\">{g}</a>\")\n",
    "{% if membrane_screener.raw_value %}\n",
    "display(HTML(dge_final[[\n",
    "    'AveExpr',\n",
    "    'logFC',\n",
    "    'P.Value',\n",
    "    'adj.P.Val',\n",
    "    'is_deg',\n",
    "    'is_significant',\n",
    "    'is_membrane',\n",
    "    'score',\n",
    "    'Link',\n",
    "]].to_html(notebook=True, escape=False)))\n",
    "{% else %}\n",
    "display(HTML(dge_final[[\n",
    "    'AveExpr',\n",
    "    'logFC',\n",
    "    'P.Value',\n",
    "    'adj.P.Val',\n",
    "    'is_deg',\n",
    "    'is_significant',\n",
    "    'score',\n",
    "    'Link',\n",
    "]].to_html(notebook=True, escape=False)))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0610e875",
   "metadata": {},
   "source": [
    "## Review Expression Levels of Selected Candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2734d08",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if 'Tabula Sapiens' in background.value[0].raw_value %}\n",
    "The background data were obtained from the [Tabula Sapiens](https://tabula-sapiens-portal.ds.czbiohub.org) atlas (The Tabula Sapiens Consortium, Science 376, eabl4896 (2022)).\n",
    "This dataset contains gene expression levels of 469 cell types from 24 normal tissues across 14 donors.  For a given gene, some cell types have zero expression levels and are listed separately.  You can explore the cell types with non-zero expression levels in the interactive plot, but please note that some cell types were only seen in a single donor and thus do not display statistics.\n",
    "{% elif 'Human Cell Atlas' in background.value[0].raw_value %}\n",
    "The background data were obtained from 15 datasets in the [Human Cell Atlas](https://data.humancellatlas.org).\n",
    "This dataset contains gene expression levels of 27 cell types from 14 normal tissues across a number of donors.  For a given gene, some cell types have zero expression levels and are listed separately.  You can explore the cell types with non-zero expression levels in the interactive plot, but please note that some cell types were only seen in a single donor and thus do not display statistics.\n",
    "{% endif %}\n",
    "\n",
    "{% if proteomics_vis.raw_value %}\n",
    "Proteomics data were obtained from the [Human Protein Atlas](https://www.proteinatlas.org/about/download) (HPA) with IHC-based expression profiling, the [Human Proteome Map](https://www.humanproteomemap.org/download.php) (HPM) with MS-based expression quantification, and a [GTEx proteome project](https://doi.org/10.1016/j.cell.2020.08.036) using TMT MS. \n",
    "These datasets contain protein expression levels detected in normal tissues and cell types. Not all differentially-expressed gene candidates may be present in the data from each project (see table for which proteomics data are present/absent). Plots show expression levels (HPA), average spectral counts (HPM), or a log-transformed relative abundance (GTEx) by tissue/cell-type for each gene candidate (excluding expression levels from the HPA where the [reliability score](https://www.proteinatlas.org/about/assays+annotation) was uncertain). \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bf2a941",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if proteomics_vis.raw_value %}\n",
    "hpm = pd.read_csv(\"s3://storage/Tumor_Gene_Target_Screener/hpm.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True), sep='\\t', header=0, index_col=0)\n",
    "hpa = pd.read_csv(\"s3://storage/Tumor_Gene_Target_Screener/hpa.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True), sep='\\t', header=0, index_col=1)\n",
    "gtexp = pd.read_csv(\"s3://storage/Tumor_Gene_Target_Screener/gtex_proteomics.tsv\", storage_options=dict(client_kwargs=dict(endpoint_url=\"https://appyters.maayanlab.cloud\"), anon=True), sep='\\t', header=0)\n",
    "gtexp['Name'] = gtexp['gene.id'].map(lambda idx: lookup(idx))\n",
    "\n",
    "# Show available genes in each dataset\n",
    "available = pd.DataFrame({'Gene': dge_final['gene_symbol'], \n",
    "                          'in HPM': dge_final['gene_symbol'].isin(hpm.index), \n",
    "                          'in HPA': dge_final['gene_symbol'].isin(hpa.index),\n",
    "                          'in GTEx Proteomics': dge_final['gene_symbol'].isin(gtexp.Name)}).drop_duplicates(subset=['Gene'])\n",
    "display(HTML(available.to_html(notebook=True, escape=False)))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51f57cda",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# create a dictionary of tissues with their respective cell types\n",
    "def create_tissue_cell_type_dict(stats):\n",
    "    final_dict = dict()\n",
    "    for cell_type in stats.columns:\n",
    "        tissue_cell_type_list = cell_type.split(\"-\", maxsplit=1)\n",
    "        final_dict.setdefault(tissue_cell_type_list[0], []).append(tissue_cell_type_list[1])\n",
    "    return final_dict\n",
    "\n",
    "# draw a category of boxplots\n",
    "def display_boxplots(stats, tissue=None):\n",
    "    IQR = stats.loc['75%']-stats.loc['25%']\n",
    "    fig.add_trace(go.Box(\n",
    "        lowerfence=np.maximum(\n",
    "            stats.loc['min'],\n",
    "            stats.loc['25%'] - (1.5*IQR),\n",
    "        ),\n",
    "        q1=stats.loc['25%'],\n",
    "        median=stats.loc['50%'],\n",
    "        q3=stats.loc['75%'],\n",
    "        upperfence=np.minimum(\n",
    "            stats.loc['max'],\n",
    "            stats.loc['75%'] + (1.5*IQR),\n",
    "        ),\n",
    "        mean=stats.loc['mean'],\n",
    "        sd=stats.loc['std'],\n",
    "        y=stats.columns,\n",
    "        {% if input_transcript_level.raw_value and background_transcript_level %}\n",
    "        name=f'{% if background_single_cell %}{tissue}{% else%}Background (Transcript Expression){% endif %}',\n",
    "        {% else %}\n",
    "        name=f'{% if background_single_cell %}{tissue}{% else%}Background (Gene Expression){% endif %}',\n",
    "        {% endif %}\n",
    "        orientation='h'\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d41b14f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if proteomics_vis.raw_value %}\n",
    "import plotly.express as px\n",
    "hpa.Tissue = hpa[\"Tissue\"] + \", \" + hpa[\"Cell.type\"]\n",
    "hpa = hpa[hpa['Reliability'] != \"Uncertain\"] \n",
    "gtexp['tissue_specificity'] = gtexp.tissue_specificity.fillna('NA')\n",
    "{% endif %}\n",
    "\n",
    "for index, row in dge_final.iterrows():\n",
    "    gene_symbol = row['gene_symbol']\n",
    "    label = row['label']\n",
    "    display(Markdown(f\"### {label}\"))\n",
    "    {% if background_transcript_level and not input_transcript_level.raw_value %}\n",
    "    # get stats for all transcripts corresponding to this gene symbol\n",
    "    stats = df_bg_stats.loc[(df_transcript_gene_map[df_transcript_gene_map['gene_symbol'] == gene_symbol].index, slice(None))].unstack()\n",
    "    # identify per-tissue mask based top 5 medians, the dropped level here is the stats level since we're going\n",
    "    #  to apply this mask to all stats based on the median (50%)\n",
    "    mask = (stats.loc[:, (slice(None), '50%')].droplevel(1, axis=1).rank(ascending=False, method='first') <= 5)\n",
    "    # apply mask to stats. the resulting mask has at most the top 5 transcripts for each tissue\n",
    "    #  if this results in a transcript missing from all tissues, we'll drop it\n",
    "    stats = stats.stack()[mask].unstack(0).dropna(how='all', axis=1)\n",
    "    # flatten column for next step, creating columns of the form: {tissue} - {transcript_id}\n",
    "    stats.columns = stats.columns.to_flat_index().map(lambda col: ' - '.join(col))\n",
    "    {% elif not background_transcript_level and input_transcript_level.raw_value %}\n",
    "    stats = df_bg_stats.loc[(gene_symbol, slice(None))]\n",
    "    {% else %}\n",
    "    stats = df_bg_stats.loc[(index, slice(None))]\n",
    "    {% endif %}\n",
    "\n",
    "    fig = go.Figure()\n",
    "    \n",
    "    {% if background_single_cell %}\n",
    "    stats_all = stats\n",
    "    # store all of the cell types with non-zero expression\n",
    "    stats = stats_all.loc[:, (stats_all != 0).any(axis=0)]\n",
    "    # store all of the cell types with zero expression\n",
    "    stats_zero_expr = stats_all.loc[:, (stats_all == 0).all(axis=0)]\n",
    "    # display the number of cell types with zero expression\n",
    "    zero_expr_size = len(stats_all.columns)-len(stats.columns)\n",
    "    display(Markdown(f\"#### {zero_expr_size} cell type{'' if zero_expr_size == 1 else 's'} with zero expression:\"))\n",
    "\n",
    "    # create a dictionary of tissues with their respective cell types (for zero expression cell types)\n",
    "    stats_zero_tissues_dict = create_tissue_cell_type_dict(stats_zero_expr)\n",
    "\n",
    "    # display cell types with zero expression\n",
    "    for tissue in stats_zero_tissues_dict.keys():\n",
    "        display(Markdown(f\"##### **{tissue}:** {'; '.join(stats_zero_tissues_dict[tissue])}\"))\n",
    "    \n",
    "    # display the number of cell types with non-zero expression\n",
    "    nonzero_expr_size = len(stats.columns)\n",
    "    display(Markdown(f\"#### {nonzero_expr_size} cell type{'' if nonzero_expr_size == 1 else 's'} with non-zero expression:\"))\n",
    "\n",
    "    # create a dictionary of tissues with their respective cell types (for non-zero expression cell types).  We won't need the values for the keys as implemented, but they're nice to have.\n",
    "    stats_tissues_dict = create_tissue_cell_type_dict(stats)\n",
    "\n",
    "    # display boxplots for each tissue\n",
    "    for tissue in stats_tissues_dict.keys():\n",
    "        # subset the stats into just cell types that correspond to the given tissue\n",
    "        display_boxplots(stats[stats.columns[stats.columns.map(lambda x: x.startswith(tissue))]], tissue)\n",
    "\n",
    "    {% else %}\n",
    "    display_boxplots(stats)\n",
    "    {% endif %}\n",
    "\n",
    "    fig.add_trace(go.Box(\n",
    "        x=df_expr_norm.loc[index],\n",
    "        {% if input_transcript_level.raw_value %}\n",
    "        name=f\"Patient Input\",\n",
    "        {% else %}\n",
    "        name=f\"Patient Input\",\n",
    "        {% endif %}\n",
    "        orientation='h'\n",
    "    ))\n",
    "    \n",
    "    fig.update_layout(title=label+f\" ({% if background_single_cell %}sc{% else %}Bulk {% endif %}RNA-seq)\", height= 300 if len(stats.columns) < 2 else len(stats.columns)*50)\n",
    "    fig.show()\n",
    "    \n",
    "    {% if proteomics_vis.raw_value %}\n",
    "    if gene_symbol in gtexp.Name.values:\n",
    "        d = gtexp[gtexp['Name'] == gene_symbol]\n",
    "        fig = px.strip(d, y=\"tissue\", x=\"value\",  \n",
    "                       orientation='h',\n",
    "                       stripmode=\"overlay\",\n",
    "                       hover_data=[\"tissue_specificity\"],\n",
    "                       height=30*d['tissue'].nunique())\n",
    "        fig.add_trace(go.Box(x=d['value'],\n",
    "                             y=d['tissue'],\n",
    "                             orientation='h',\n",
    "                             marker=dict(color='#636EFA'),\n",
    "                             name=\"n > 1\"))\n",
    "        fig.update_layout(title=label+\" (GTEx Proteomics)\",\n",
    "                          autosize=True,\n",
    "                          showlegend=False)\n",
    "        fig.update_xaxes(title=\"log2(relative abundance)\")\n",
    "        fig.update_yaxes(title=None)\n",
    "        fig.show()\n",
    "       \n",
    "    if gene_symbol in hpm.index:\n",
    "        fig = px.scatter(hpm.loc[[gene_symbol]], \n",
    "                         y=\"Tissue\", x=\"value\", \n",
    "                         height=20*hpm.loc[[gene_symbol]].shape[0])\n",
    "        fig.update_layout(title=label+\" (HPM)\", \n",
    "                          autosize=True)\n",
    "        fig.update_xaxes(title=\"Average Spectral Counts\")\n",
    "        fig.update_yaxes(title=None)\n",
    "        fig.show()\n",
    "    \n",
    "    if gene_symbol in hpa.index:\n",
    "        fig = px.scatter(hpa.loc[[gene_symbol]], \n",
    "                         y=\"Tissue\", x=\"Level\", \n",
    "                         category_orders={\"Level\": [\"Not detected\", \"Low\", \"Medium\", \"High\"]}, \n",
    "                         hover_data=[\"Reliability\"],  \n",
    "                         hover_name=\"Tissue\",\n",
    "                         height=20*hpa.loc[[gene_symbol]].shape[0])\n",
    "        fig.update_layout(title=label+\" (HPA)\", \n",
    "                          showlegend=False, \n",
    "                          autosize=True, \n",
    "                          xaxis={'tickmode':'array', \n",
    "                                 'tickvals':[0, 1, 2, 3], \n",
    "                                 'ticktext':[\"Not detected\", \"Low\", \"Medium\", \"High\"]})\n",
    "        fig.update_xaxes(title=\"Tissue Expression Level\")\n",
    "        fig.update_yaxes(title=None)\n",
    "        fig.show()\n",
    "    {% endif %}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "bc6a549fc5b5a2ab0c472566a8b55e89aadd87484d87304664828dfe61ef602a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
