{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import multiprocessing as mp\n",
    "from MulticoreTSNE import MulticoreTSNE as TSNE\n",
    "import seaborn as sns\n",
    "from IPython.display import HTML, display, Markdown\n",
    "\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import ColumnDataSource, HoverTool, NumeralTickFormatter\n",
    "from bokeh.transform import factor_cmap\n",
    "from bokeh.models import Select\n",
    "from bokeh.models.callbacks import CustomJS\n",
    "from bokeh.layouts import column\n",
    "\n",
    "import scanpy as sc\n",
    "\n",
    "from react_scatter_board import ReactScatterBoard\n",
    "from scipy.stats import zscore\n",
    "import qnorm\n",
    "import itertools\n",
    "\n",
    "output_notebook() #allow bokeh plot show on the notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% do SectionField(\n",
    "    name='Data_Section',\n",
    "    title='Load Your Data',\n",
    "    subtitle='Load your metadata and data matrix in comma/tab separated formats. Genes should be in rows and samples shoud be in columns.',\n",
    "    img='load.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Visualization_Section',\n",
    "    title='Select Visualization Parameters',\n",
    "    subtitle='',\n",
    "    img='select.png'\n",
    "    \n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set rnaseq_data_filename = FileField(\n",
    "    name='rnaseq_data_filename', \n",
    "    label='RNA-seq data file (.csv, .tsv, .tpm or .txt)', \n",
    "    default='Papillary+adenocarcinoma,+NOS_data.csv',\n",
    "    examples={'Papillary+adenocarcinoma,+NOS_data.csv': \"https://appyters.maayanlab.cloud/storage/dimensionality_reduction_visualization/Papillary+adenocarcinoma,+NOS_data.csv\"}, \n",
    "    description='Upload RNA-seq data matrix as comma seperated or tab seperated format. The index of the dataset are cells, the columns are samples.', \n",
    "    section='Data_Section')\n",
    "\n",
    "%}\n",
    "\n",
    "{% set meta_data_filename = FileField(\n",
    "    name='meta_data_filename', \n",
    "    label='Metadata file (.csv, .tsv, .tpm or .txt)', \n",
    "    default='Papillary+adenocarcinoma,+NOS_clinical_metadata.csv',\n",
    "    examples={'Papillary+adenocarcinoma,+NOS_clinical_metadata.csv': \"https://appyters.maayanlab.cloud/storage/dimensionality_reduction_visualization/Papillary+adenocarcinoma,+NOS_clinical_metadata.csv\"}, \n",
    "    description='Upload metadata as two-column comma seperated or tab seperated format. One column contains sample ids and the other column contains sample labels', \n",
    "    section='Data_Section')\n",
    "\n",
    "%}\n",
    "\n",
    "{% set meta_class_column_name = StringField(\n",
    "    name='meta_class_column_name', \n",
    "    label='Class column name in metadata', \n",
    "    default='case_id', \n",
    "    description='One of the column names in metadata that aligns with main data', \n",
    "    section='Data_Section')\n",
    "%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set need_transpose = BoolField(\n",
    "    name='need_transpose', \n",
    "    label='Samples are the columns?',\n",
    "    default='true',\n",
    "    description='Select true if samples are columns and rows are genes', \n",
    "    section='Visualization_Section') \n",
    "%}\n",
    "\n",
    "{% set need_normalize = BoolField(\n",
    "    name='need_normalize', \n",
    "    label='Normalization?',\n",
    "    default='true',\n",
    "    description='Select true if you would like to normalize the data matrix', \n",
    "    section='Visualization_Section') \n",
    "%}\n",
    "\n",
    "{% set need_leiden = BoolField(\n",
    "    name='need_leiden', \n",
    "    label='Leiden graph for clustering?',\n",
    "    default='true',\n",
    "    description='Leiden graph automatic identifies clusters and color them', \n",
    "    section='Visualization_Section') \n",
    "%}\n",
    "\n",
    "{% set plot_3D = ChoiceField(\n",
    "    name='3D_plot', \n",
    "    label='3D plots?', \n",
    "    choices = {'2D': '2D', '3D': '3D', 'All':'All'},\n",
    "    default='All', \n",
    "    description='Choose to have 2D, 3D, or all plots', \n",
    "    section='Visualization_Section')\n",
    "%}\n",
    "\n",
    "{% set visualization_method = ChoiceField(\n",
    "    name='visualization_method', \n",
    "    label='Visualization Methods', \n",
    "    choices = {'All': 'All','PCA': 'PCA', 't-SNE': 't-SNE', 'UMAP': 'UMAP'},\n",
    "    default='All', \n",
    "    description='Select a visualization method', \n",
    "    section='Visualization_Section')\n",
    "%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "rnaseq_data_filename = {{rnaseq_data_filename}}\n",
    "meta_data_filename = {{meta_data_filename}}\n",
    "meta_class_column_name = {{meta_class_column_name}}\n",
    "need_leiden = {{need_leiden}}\n",
    "need_transpose = {{need_transpose}}\n",
    "\n",
    "clustering_topk = 50"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PCA, t-SNE, and UMAP Appyter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This Appyter [1] loads data in a matrix format, it then performs dimensionality reduction algorithms on the data; and then visualizes the dimensionality reduced data with with static and interactive 2D and 3D PCA [2], t-SNE [3], and UMAP [4] plots.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if visualization_method.value == \"All\" or visualization_method.value == \"PCA\" %}\n",
    "## PCA\n",
    "Principal Component Analysis (PCA) [2] is a statistical technique ussed to identify global patterns in high-dimensional datasets. It is commonly used to explore the similarity of biological samples in omics datasets. To perform PCA, variable values are transformed into Principal Components (PCs), a set of linearly orthogonal features which represent the most relevant sources of variance in the data. The top PCs subsequently visualized using a scatter plot for each variable to estimate similarity and differences between the variables in higher dimensions.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if visualization_method.value == \"All\" or visualization_method.value == \"t-SNE\" %}\n",
    "## t-SNE\n",
    "t-distributed stochastic neighbor embedding (t-SNE) [3] is a statistical method used to identify global patterns in high-dimensional datasets. It is a non-linear Dimensionality reduction technique, and it embeds the points from a higher dimension to a lower dimension trying to preserve the neighborhood of that point. It then provides each datapoint a location in a two or three-dimensional space visualized using a scatter plot.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if visualization_method.value == \"All\" or visualization_method.value == \"UMAP\" %}\n",
    "## UMAP\n",
    "Uniform Manifold Approximation and Projection (UMAP) [4] is a statistical method used to identify global patterns in high-dimensional datasets. It creates a graph that accurately reflects the topology of the true high dimensional graph. It then calculates the weight for edges of this graph and builds the low dimensional graph.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if need_leiden.value %}\n",
    "## Leiden Algorithm\n",
    "Leiden algorithm [5] identifies well-connected clusters in networks. In this procedure, the appyter takes into account both modalities of the data by integrating connectivity graphs generated from each modality (ref https://scanpy-tutorials.readthedocs.io/en/multiomics/cite-seq/pbmc5k.html#Clustering). It visualizes the samples colored by clusters.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis Overview\n",
    "The input data matrix that has been normalized first undergoes dimensionality reduction using PCA [2], t-SNE [3], and UMAP [4] with the Python library. Columns are then clustered based on their most-associated highly-variable variables and metadata features. Clusters are visualized using the Bokeh package [6] for the 2D plots, and the React-Scatter-Board package [7] developed by the Ma'ayan lab for interactive visualization of the 3D plots. In addition, user may select to upload data without metadata and use Leiden's algorithm [5] to visualize clusters."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_file(filename, **kwargs):\n",
    "    if filename.endswith(\".csv\"):\n",
    "        temp_df = pd.read_csv(filename, **kwargs).sort_index()\n",
    "    elif filename.endswith(\".txt\") or filename.endswith(\".tsv\") or filename.endswith(\".tpm\"):\n",
    "        temp_df = pd.read_csv(filename, sep=\"\\t\", **kwargs).sort_index()\n",
    "    else:\n",
    "        raise Exception('Error! Incorrect format. Please load file in txt, tsv, tpm, or csv format')\n",
    "    \n",
    "    return temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if rnaseq_data_filename == '':\n",
    "    raise Exception('Please upload data matrix.')\n",
    "    \n",
    "expr_df = load_file(rnaseq_data_filename, index_col=0)\n",
    "display(expr_df.head())\n",
    "\n",
    "if need_transpose:\n",
    "    # transpose input matrix so that column names are genes\n",
    "    expr_df = expr_df.transpose()\n",
    "    \n",
    "expr_df = expr_df.dropna(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if meta_data_filename.value%}\n",
    "meta_df = load_file(meta_data_filename)\n",
    "assert meta_class_column_name in meta_df.columns, f\"Error! Column '{meta_class_column_name}' is not in metadata\"\n",
    "\n",
    "# remove duplicates in index column\n",
    "meta_df = meta_df.drop_duplicates(subset=[meta_class_column_name])\n",
    "\n",
    "# align metedata and main data\n",
    "temp = meta_df.merge(expr_df,left_on=meta_class_column_name, right_index=True)\n",
    "\n",
    "df2 = temp[expr_df.columns].copy()\n",
    "expr_df = df2.set_index(temp[meta_class_column_name])\n",
    "\n",
    "df1 = temp[meta_df.columns].copy()\n",
    "meta_df = df1.set_index(meta_class_column_name)\n",
    "\n",
    "for i in range(meta_df.shape[0]):\n",
    "    if expr_df.index[i] != meta_df.index[i]:\n",
    "        raise Exception('Error! Metadata is not aligned with main data. Please check if you enter the correct metadata column name')\n",
    "        \n",
    "# category NAN value as 'missing' label\n",
    "for feature in meta_df.columns:\n",
    "    if meta_df[feature].isna().any():\n",
    "        meta_df[feature].fillna('Missing', inplace =True)\n",
    "        \n",
    "display(meta_df.head())\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Processing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if need_normalize.value %}\n",
    "# Normalize Data\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if need_normalize.value%}\n",
    "# Normalize Data\n",
    "def log2_normalize(x, offset=1.):\n",
    "    return np.log2(x + offset)\n",
    "\n",
    "df_data_norm = log2_normalize(expr_df, offset=1)\n",
    "df_data_norm = qnorm.quantile_normalize(df_data_norm, axis=0)\n",
    "\n",
    "#convert to zscores\n",
    "expr_df = pd.DataFrame(zscore(df_data_norm, axis=0), index=df_data_norm.index, columns=df_data_norm.columns)\n",
    "\n",
    "\n",
    "# check if data have been normalized correctly\n",
    "missing_percent = np.mean(expr_df.isnull().mean() * 100)\n",
    "print(\"Missing value is {} %.\".format(missing_percent))\n",
    "if missing_percent > 10:\n",
    "    raise Exception(\"Error in normalization! Too many NA values. Try to unselect normalization field.\")\n",
    "    \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if need_leiden.value %}\n",
    "# Compute label and pca based on Leiden Algorithm \n",
    "leiden_df = sc.AnnData(expr_df,dtype=np.float32)\n",
    "\n",
    "try:\n",
    "    sc.pp.pca(leiden_df)\n",
    "except:\n",
    "    raise Exception(\"Error! Incorrect format. Please check if the data have been read/seperated correctly.\")\n",
    "    \n",
    "sc.pp.neighbors(leiden_df)  \n",
    "sc.tl.leiden(leiden_df, key_added=\"leiden\")\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Process metadata\n",
    "df_y = pd.DataFrame()\n",
    "\n",
    "if meta_data_filename:\n",
    "    # exclude features with too many categories or only one feature or non string type\n",
    "    feature_selection = [feature for feature in meta_df\n",
    "                      if meta_df[feature].dtype == object and len(meta_df[feature].unique()) < 50 and len(meta_df[feature].unique()) > 1] \n",
    "    df_y = meta_df[feature_selection].copy()\n",
    "    df_y = df_y.astype(str)\n",
    "\n",
    "if need_leiden:\n",
    "    df_y['leiden'] = list(leiden_df.obs['leiden'].values)\n",
    "else:\n",
    "    df_y['y'] = expr_df.index\n",
    "\n",
    "    \n",
    "for feature in df_y:\n",
    "    # assign 'Unclassified' to low frequency categories\n",
    "    if len(df_y[feature].unique()) > 10:\n",
    "        top9 = df_y[feature].value_counts()[:9].index.tolist()\n",
    "        remain = set(df_y[feature].unique()) - set(top9)\n",
    "        df_y[feature].loc[df_y[feature].isin(remain)] = 'Unclassified'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Data transformation for 2D visualization\n",
    "{% if visualization_method.value == \"All\" or visualization_method.value == \"PCA\" %}\n",
    "# Normalize transformed data to have a better visualization on 3D plot\n",
    "leiden_df.obsm['X_pca'] = zscore(leiden_df.obsm['X_pca'],axis=0)\n",
    "\n",
    "pca_data = pd.DataFrame({'x':leiden_df.obsm['X_pca'][:,0],\n",
    "                       'y':leiden_df.obsm['X_pca'][:,1],\n",
    "                       'z':leiden_df.obsm['X_pca'][:,2]})\n",
    "pca_df = df_y.reset_index().join(pca_data).set_index(df_y.index)\n",
    "{% endif %}\n",
    "\n",
    "{% if visualization_method.value == \"All\" or visualization_method.value == \"t-SNE\" %}\n",
    "n_jobs=mp.cpu_count()\n",
    "tsne = TSNE(n_jobs = n_jobs, n_components=3)\n",
    "leiden_df.obsm['X_tsne'] = tsne.fit_transform(leiden_df.obsm['X_pca'])\n",
    "leiden_df.obsm['X_tsne'] = zscore(leiden_df.obsm['X_tsne'],axis=0)\n",
    "\n",
    "tsne_data = pd.DataFrame({'x':leiden_df.obsm['X_tsne'][:,0],\n",
    "                       'y':leiden_df.obsm['X_tsne'][:,1],\n",
    "                       'z':leiden_df.obsm['X_tsne'][:,2]})\n",
    "tsne_df = df_y.reset_index().join(tsne_data).set_index(df_y.index)\n",
    "{% endif %}\n",
    "\n",
    "{% if visualization_method.value == \"All\" or visualization_method.value == \"UMAP\" %}\n",
    "sc.tl.umap(leiden_df, n_components=3)\n",
    "leiden_df.obsm['X_umap'] = zscore(leiden_df.obsm['X_umap'],axis=0)\n",
    "\n",
    "umap_data = pd.DataFrame({'x':leiden_df.obsm['X_umap'][:,0],\n",
    "                       'y':leiden_df.obsm['X_umap'][:,1],\n",
    "                       'z':leiden_df.obsm['X_umap'][:,2]})\n",
    "umap_df = df_y.reset_index(drop=True).join(umap_data).set_index(df_y.index)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Data processing for 3D visualization\n",
    "def create_3D_df(df_2D):\n",
    "    return [dict(r, opacity=0.7) for i, r in df_2D.iterrows()]\n",
    "\n",
    "{% if (plot_3D.value == \"3D\" or plot_3D.value == \"All\") and \n",
    " (visualization_method.value == \"All\" or visualization_method.value == \"PCA\") %}\n",
    "pca_df_3D = create_3D_df(pca_df)\n",
    "{% endif %}\n",
    "\n",
    "{% if (plot_3D.value == \"3D\" or plot_3D.value == \"All\") and \n",
    " (visualization_method.value == \"All\" or visualization_method.value == \"t-SNE\") %}\n",
    "tsne_df_3D = create_3D_df(tsne_df)\n",
    "{% endif %}\n",
    "\n",
    "{% if (plot_3D.value == \"3D\" or plot_3D.value == \"All\") and \n",
    " (visualization_method.value == \"All\" or visualization_method.value == \"UMAP\") %}\n",
    "umap_df_3D = create_3D_df(umap_df)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualize Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter = iter(itertools.count(start=1))\n",
    "def display_caption(caption):\n",
    "    display(Markdown(\"*Figure {}. {}*\".format(next(counter), caption)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if plot_3D.value == \"2D\" or plot_3D.value == \"All\" %}\n",
    "\n",
    "def generate_colors(input_df, feature):\n",
    "    pal = sns.color_palette()\n",
    "    color = factor_cmap(feature, palette=pal.as_hex(), factors=np.array(input_df[feature].unique()))\n",
    "\n",
    "    return color \n",
    "    \n",
    "def interactive_circle_plot(input_df, x_lab, y_lab, feature):\n",
    "    input_df['legend'] = input_df[feature]\n",
    "    source = ColumnDataSource(input_df)\n",
    "\n",
    "    TOOLTIPS = [\n",
    "        (\"index\", \"$index\"),\n",
    "        (\"(x,y)\", \"($x, $y)\"),\n",
    "        (\"feature\", '@'+feature),\n",
    "    ]\n",
    "    \n",
    "    n = input_df[feature].nunique()\n",
    "    numberOfPoints = input_df.shape[0]\n",
    "    point_size = 10 if numberOfPoints < 100 else 25*np.log(numberOfPoints)/np.sqrt(numberOfPoints)\n",
    "    \n",
    "    p = figure(height=400, tooltips=TOOLTIPS,x_axis_label=x_lab, y_axis_label=y_lab,sizing_mode=\"scale_width\")\n",
    "       \n",
    "    if meta_data_filename or need_leiden:\n",
    "        color = generate_colors(input_df, feature)\n",
    "        p1 = p.circle('x', 'y', size=point_size, source=source, legend_field='legend',fill_color= color, line_color=color)\n",
    "        p.add_layout(p.legend[0], 'right')\n",
    "    else:\n",
    "        p1 = p.circle('x', 'y', size=3, source=source)\n",
    "        \n",
    "    p.output_backend = \"svg\"\n",
    "    p.xgrid.visible = False\n",
    "    p.ygrid.visible = False\n",
    "    p.xaxis.minor_tick_line_color = None \n",
    "    p.yaxis.minor_tick_line_color = None \n",
    "    p.xaxis.axis_label_text_font_size = '12pt'\n",
    "    p.yaxis.axis_label_text_font_size = '12pt'\n",
    "    p.xaxis[0].formatter = NumeralTickFormatter(format=\"0.0\")\n",
    "    p.yaxis[0].formatter = NumeralTickFormatter(format=\"0.0\")\n",
    "        \n",
    "    return p, p1, source\n",
    "\n",
    "def generate_plot(p, p_circle, source_data, input_df):\n",
    "    column_options = list(input_df.columns[:-4])\n",
    "\n",
    "    my_colors = {}\n",
    "    for feature in column_options:\n",
    "        my_colors[feature] = generate_colors(input_df, feature)\n",
    "        \n",
    "    arg = dict(colors=my_colors, plot=p_circle, source = source_data, \n",
    "           data_category = input_df.reset_index().to_dict(orient='list'), feature = input_df.columns[0])\n",
    "    \n",
    "    select = Select(title=\"Plot to show:\", value=column_options[0], options=column_options)\n",
    "    select.js_on_change(\"value\", CustomJS(args=arg, code=\"\"\"\n",
    "\n",
    "    plot.glyph.line_color = colors[this.value]\n",
    "    plot.glyph.fill_color = colors[this.value]\n",
    "    source.data['legend'] = data_category[this.value]\n",
    "    source.data[feature] = data_category[this.value]\n",
    "    \n",
    "    \"\"\"))\n",
    "\n",
    "    layout = column(select, p)\n",
    "    layout.sizing_mode = \"stretch_both\"\n",
    "    return layout\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if (plot_3D.value == \"2D\" or plot_3D.value == \"All\") and \n",
    "(visualization_method.value == \"All\" or visualization_method.value == \"PCA\") %}\n",
    "# 2D PCA plot\n",
    "p, p_circle, source_data = interactive_circle_plot(pca_df, \"PC-1\", \"PC-2\", pca_df.columns[0])\n",
    "\n",
    "if meta_data_filename or need_leiden:\n",
    "    pca_plot = generate_plot(p, p_circle, source_data, pca_df)\n",
    "    show(pca_plot)\n",
    "else:\n",
    "    show(p)\n",
    "caption = \"PCA 2D scatter plot. Each point presents a column from the input data matrix.\"\n",
    "display_caption(caption)\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if (plot_3D.value == \"2D\" or plot_3D.value == \"All\") and \n",
    " (visualization_method.value == \"All\" or visualization_method.value == \"t-SNE\") %}\n",
    "# t-SNE scatter plot\n",
    "\n",
    "p, p_circle, source_data = interactive_circle_plot(tsne_df, \"t-SNE-1\", \"t-SNE-2\", tsne_df.columns[0])\n",
    "\n",
    "if meta_data_filename or need_leiden:\n",
    "    tsne_plot = generate_plot(p, p_circle, source_data, tsne_df)\n",
    "    show(tsne_plot)\n",
    "else:\n",
    "    show(p)\n",
    "\n",
    "caption = \"t-SNE 2D scatter plot. Each point presents a column from the input data matrix.\"\n",
    "display_caption(caption)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if (plot_3D.value == \"2D\" or plot_3D.value == \"All\") and \n",
    " (visualization_method.value == \"All\" or visualization_method.value == \"UMAP\") %}\n",
    "# Umap scatter plot\n",
    "\n",
    "p, p_circle, source_data = interactive_circle_plot(umap_df, \"UMAP-1\", \"UMAP-2\", umap_df.columns[0])\n",
    "if meta_data_filename or need_leiden:\n",
    "    umap_plot = generate_plot(p, p_circle, source_data, umap_df)\n",
    "    show(umap_plot)\n",
    "else:\n",
    "    show(p)\n",
    "\n",
    "caption = \"UMAP 2D scatter plot. Each point presents a column from the input data matrix.\"\n",
    "display_caption(caption)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if (plot_3D.value == \"3D\" or plot_3D.value == \"All\") and \n",
    " (visualization_method.value == \"All\" or visualization_method.value == \"PCA\") %}\n",
    "# PCA 3D scatter plot\n",
    "\n",
    "display(ReactScatterBoard(\n",
    "  is3d=True,\n",
    "  data=pca_df_3D,\n",
    "  shapeKey=pca_df.columns[0],\n",
    "  colorKey=pca_df.columns[0],\n",
    "  scale=4\n",
    "))\n",
    "\n",
    "caption = \"PCA 3D scatter plot. Each point presents a column from the input data matrix. Scroll to zoom, drag to move around.\"\n",
    "\n",
    "display_caption(caption)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if (plot_3D.value == \"3D\" or plot_3D.value == \"All\") and (visualization_method.value == \"All\" or visualization_method.value == \"t-SNE\") %}\n",
    "# t-SNE 3D scatter plot\n",
    "\n",
    "display(ReactScatterBoard(\n",
    "  is3d=True,\n",
    "  data=tsne_df_3D,\n",
    "  shapeKey=tsne_df.columns[0],\n",
    "  colorKey=tsne_df.columns[0],\n",
    "  scale=4\n",
    "))\n",
    "\n",
    "caption = \"t-SNE 3D scatter plot. Each point presents a column from the input data matrix. Scroll to zoom, drag to move around.\"\n",
    "\n",
    "display_caption(caption)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if (plot_3D.value == \"3D\" or plot_3D.value == \"All\") and (visualization_method.value == \"All\" or visualization_method.value == \"UMAP\") %}\n",
    "# Umap 3D scatter plot\n",
    "\n",
    "display(ReactScatterBoard(\n",
    "  is3d=True,\n",
    "  data=umap_df_3D,\n",
    "  shapeKey=umap_df.columns[0],\n",
    "  colorKey=umap_df.columns[0],\n",
    "  scale=4\n",
    "))\n",
    "\n",
    "caption = \"UMAP 3D scatter plot. Each point presents a column from the input data matrix. Scroll to zoom, drag to move around.\"\n",
    "\n",
    "display_caption(caption)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reference\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Clarke DJB, Jeon M, Stein DJ, Moiseyev N, Kropiwnicki E, Dai C, Xie Z, Wojciechowicz ML, Litz S, Hom J, Evangelista JE, Goldman L, Zhang S, Yoon C, Ahamed T, Bhuiyan S, Cheng M, Karam J, Jagodnik KM, Shu I, Lachmann A, Ayling S, Jenkins SL, Ma'ayan A. Appyters: Turning Jupyter Notebooks into data-driven web apps. Patterns (N Y). 2021 Mar 4;2(3):100213.\n",
    "<br>\n",
    "2. Clark NR, Ma'ayan A. Introduction to statistical methods to analyze large data sets: principal components analysis. Sci Signal. 2011 Sep 6;4(190):tr3.\n",
    "<br>\n",
    "3. Maaten, Laurens van der and Geoffrey E. Hinton. “Visualizing Data using t-SNE.” Journal of Machine Learning Research 9 (2008): 2579-2605.\n",
    "<br>\n",
    "4. McInnes, L., Healy, J., Saul, N. & Großberger, L. UMAP: uniform manifold approximation and projection. J. Open Source Softw. 3, 861 (2018).\n",
    "<br>\n",
    "5. Traag, Vincent A., Ludo Waltman, and Nees Jan van Eck. \"From Louvain to Leiden: guaranteeing well-connected communities.\" Scientific reports 9.1 (2019): 1-12.\n",
    "<br>\n",
    "6. Bokeh Development Team (2018). Bokeh: Python library for interactive visualization. URL http://www.bokeh.pydata.org.\n",
    "<br>\n",
    "7. MaayanLab. react-scatter-board: Python library for interactive 3 Dimension plot. URL https://github.com/MaayanLab/react-scatter-board"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3.8-appyter",
   "language": "python",
   "name": "python3.8-appyter"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
