{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STEAP\n",
    "## Post-Processing Notebook \n",
    "\n",
    "1.    [Identifying significant cell types](#1)\n",
    "2.    [Overlap  with enriched brain related cell types (using upsetplot)](#2)\n",
    "3.    [Gene-Set Enrichment Analysis (GSEA)](#3)\n",
    "4.    [Cell type correlation between the pheontype and brain related traits](#4)\n",
    "5.    [Expression Specificity (ES) gene correlation between cell types](#5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we need to import the necessary packages and scripts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import constants # file which consists of constant for the analysis\n",
    "from scripts import convert_output_to_dataframe\n",
    "from scripts import gene_set_enrichment_analysis\n",
    "from scripts import calculate_beta_correlation\n",
    "from scripts import calculate_es_correlation\n",
    "from scripts import upsetplot\n",
    "from scripts import circosplot\n",
    "\n",
    "from bokeh.io import output_notebook\n",
    "from utils import display_bokeh_df, make_caption, display_link, create_download_link, create_zip_file\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% do SectionField(\n",
    "    name='Data_Section',\n",
    "    title='Load your Data'\n",
    ") %}\n",
    "\n",
    "{% set gwas_name = StringField(\n",
    "    name='gwas_name', \n",
    "    label='Name of the phenotype used as GWAS_SUMSTATS in the config.yml file (e.g. SCZ_PGC3_2020)',\n",
    "    default='SCZ_PGC3_2020',\n",
    "    description='Give a name to the phenotype analysed in the prioritization files.',\n",
    "    section='Data_Section'\n",
    ") %}\n",
    "\n",
    "{% set pheno_name = StringField(\n",
    "    name='pheno_name', \n",
    "    label='Name of the phenotype (e.g. Schizophrenia)', \n",
    "    default='Schizophrenia',\n",
    "    description='Give a name to the phenotype analysed in the prioritization files.',\n",
    "    section='Data_Section'\n",
    ") %}\n",
    "\n",
    "{% set hmagma_file = FileField(\n",
    "    name='hmagma_file', \n",
    "    label='The H-MAGMA priotitization.csv file', \n",
    "    examples={'h-magma.csv': 'https://github.com/erwinerdem/appyter_steap/raw/master/h-magma.csv'},\n",
    "    section='Data_Section'\n",
    ") %}\n",
    "\n",
    "{% set ldsc_file = FileField(\n",
    "    name='ldsc_file', \n",
    "    label='The S-LDSC priotitization.csv file', \n",
    "    examples={'ldsc.csv': 'https://github.com/erwinerdem/appyter_steap/raw/master/ldsc.csv'},\n",
    "    section='Data_Section'\n",
    ") %}\n",
    "\n",
    "{% set magma_file = FileField(\n",
    "    name='magma_file', \n",
    "    label='The MAGMA priotitization.csv file', \n",
    "    examples={'magma.csv': 'https://github.com/erwinerdem/appyter_steap/raw/master/magma.csv'},\n",
    "    section='Data_Section'\n",
    ") %}\n",
    "\n",
    "{% set df_all = FileField(\n",
    "    name='df_all', \n",
    "    label='The .h5 data containing the enrichment analysis of brain related phenotypes.', \n",
    "    description='Load the example enrichment_data.h5 file.',\n",
    "    examples={'enrichment_data.h5': 'https://github.com/erwinerdem/STEAP/raw/master/data/enrichment_data.h5'},\n",
    "    section='Data_Section'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% do SectionField(\n",
    "    name='Analysis_Section',\n",
    "    title='Choose your Analysis'\n",
    ") %}\n",
    "\n",
    "\n",
    "{% set other_gwas = MultiChoiceField(\n",
    "    name='other_gwas',\n",
    "    label='Select which other brain related phenotypes you wish to compare against (default is all)',\n",
    "    choices=[\n",
    "        'Structural MRI',\n",
    "        'DTI Tracts',\n",
    "        'rs-fMRI Network',\n",
    "        'Schizophrenia',\n",
    "        \"Alzheimer's disease\",\n",
    "        \"Bipolar\",\n",
    "        \"Depression\",\n",
    "        \"Autism spectrum disorder\"\n",
    "    ],\n",
    "    default=[\n",
    "        'Structural MRI',\n",
    "        'DTI Tracts',\n",
    "        'rs-fMRI Network',\n",
    "        'Schizophrenia',\n",
    "        \"Alzheimer's disease\",\n",
    "        \"Bipolar\",\n",
    "        \"Depression\",\n",
    "        \"Autism spectrum disorder\"\n",
    "    ],\n",
    "    section='Analysis_Section'\n",
    ") %}\n",
    "\n",
    "\n",
    "{% set gsea = BoolField(\n",
    "    name='gsea', \n",
    "    label='Perform gene-set enrichment analysis?', \n",
    "    default='false', \n",
    "    section='Analysis_Section'\n",
    ") %}\n",
    "\n",
    "{% set genesetlist = MultiChoiceField(\n",
    "    name='genesetlist',\n",
    "    label='Select which gene-list(s) to use for gene-set enrichment analysis',\n",
    "    choices=[\n",
    "        'ARCHS4_Cell-lines',\n",
    "        'ARCHS4_IDG_Coexp',\n",
    "        'ARCHS4_Kinases_Coexp',\n",
    "        'ARCHS4_TFs_Coexp',\n",
    "        'ARCHS4_Tissues',\n",
    "        'Achilles_fitness_decrease',\n",
    "        'Achilles_fitness_increase',\n",
    "        'Aging_Perturbations_from_GEO_down',\n",
    "        'Aging_Perturbations_from_GEO_up',\n",
    "        'Allen_Brain_Atlas_down',\n",
    "        'Allen_Brain_Atlas_up',\n",
    "        'BioCarta_2013',\n",
    "        'BioCarta_2015',\n",
    "        'BioCarta_2016',\n",
    "        'BioPlanet_2019',\n",
    "        'BioPlex_2017',\n",
    "        'CCLE_Proteomics_2020',\n",
    "        'CORUM',\n",
    "        'COVID-19_Related_Gene_Sets',\n",
    "        'Cancer_Cell_Line_Encyclopedia',\n",
    "        'ChEA_2013',\n",
    "        'ChEA_2015',\n",
    "        'ChEA_2016',\n",
    "        'Chromosome_Location',\n",
    "        'Chromosome_Location_hg19',\n",
    "        'ClinVar_2019',\n",
    "        'DSigDB',\n",
    "        'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "        'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "        'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "        'DisGeNET',\n",
    "        'Disease_Perturbations_from_GEO_down',\n",
    "        'Disease_Perturbations_from_GEO_up',\n",
    "        'Disease_Signatures_from_GEO_down_2014',\n",
    "        'Disease_Signatures_from_GEO_up_2014',\n",
    "        'DrugMatrix',\n",
    "        'Drug_Perturbations_from_GEO_2014',\n",
    "        'Drug_Perturbations_from_GEO_down',\n",
    "        'Drug_Perturbations_from_GEO_up',\n",
    "        'ENCODE_Histone_Modifications_2013',\n",
    "        'ENCODE_Histone_Modifications_2015',\n",
    "        'ENCODE_TF_ChIP-seq_2014',\n",
    "        'ENCODE_TF_ChIP-seq_2015',\n",
    "        'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "        'ESCAPE',\n",
    "        'Elsevier_Pathway_Collection',\n",
    "        'Enrichr_Libraries_Most_Popular_Genes',\n",
    "        'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "        'Enrichr_Users_Contributed_Lists_2020',\n",
    "        'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "        'GO_Biological_Process_2013',\n",
    "        'GO_Biological_Process_2015',\n",
    "        'GO_Biological_Process_2017',\n",
    "        'GO_Biological_Process_2017b',\n",
    "        'GO_Biological_Process_2018',\n",
    "        'GO_Cellular_Component_2013',\n",
    "        'GO_Cellular_Component_2015',\n",
    "        'GO_Cellular_Component_2017',\n",
    "        'GO_Cellular_Component_2017b',\n",
    "        'GO_Cellular_Component_2018',\n",
    "        'GO_Molecular_Function_2013',\n",
    "        'GO_Molecular_Function_2015',\n",
    "        'GO_Molecular_Function_2017',\n",
    "        'GO_Molecular_Function_2017b',\n",
    "        'GO_Molecular_Function_2018',\n",
    "        'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "        'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "        'GWAS_Catalog_2019',\n",
    "        'GeneSigDB',\n",
    "        'Gene_Perturbations_from_GEO_down',\n",
    "        'Gene_Perturbations_from_GEO_up',\n",
    "        'Genes_Associated_with_NIH_Grants',\n",
    "        'Genome_Browser_PWMs',\n",
    "        'HMDB_Metabolites',\n",
    "        'HMS_LINCS_KinomeScan',\n",
    "        'HomoloGene',\n",
    "        'HumanCyc_2015',\n",
    "        'HumanCyc_2016',\n",
    "        'Human_Gene_Atlas',\n",
    "        'Human_Phenotype_Ontology',\n",
    "        'InterPro_Domains_2019',\n",
    "        'Jensen_COMPARTMENTS',\n",
    "        'Jensen_DISEASES',\n",
    "        'Jensen_TISSUES',\n",
    "        'KEA_2013',\n",
    "        'KEA_2015',\n",
    "        'KEGG_2013',\n",
    "        'KEGG_2015',\n",
    "        'KEGG_2016',\n",
    "        'KEGG_2019_Human',\n",
    "        'KEGG_2019_Mouse',\n",
    "        'Kinase_Perturbations_from_GEO_down',\n",
    "        'Kinase_Perturbations_from_GEO_up',\n",
    "        'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "        'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "        'LINCS_L1000_Chem_Pert_down',\n",
    "        'LINCS_L1000_Chem_Pert_up',\n",
    "        'LINCS_L1000_Ligand_Perturbations_down',\n",
    "        'LINCS_L1000_Ligand_Perturbations_up',\n",
    "        'Ligand_Perturbations_from_GEO_down',\n",
    "        'Ligand_Perturbations_from_GEO_up',\n",
    "        'MCF7_Perturbations_from_GEO_down',\n",
    "        'MCF7_Perturbations_from_GEO_up',\n",
    "        'MGI_Mammalian_Phenotype_2013',\n",
    "        'MGI_Mammalian_Phenotype_2017',\n",
    "        'MGI_Mammalian_Phenotype_Level_3',\n",
    "        'MGI_Mammalian_Phenotype_Level_4',\n",
    "        'MGI_Mammalian_Phenotype_Level_4_2019',\n",
    "        'MSigDB_Computational',\n",
    "        'MSigDB_Hallmark_2020',\n",
    "        'MSigDB_Oncogenic_Signatures',\n",
    "        'Microbe_Perturbations_from_GEO_down',\n",
    "        'Microbe_Perturbations_from_GEO_up',\n",
    "        'Mouse_Gene_Atlas',\n",
    "        'NCI-60_Cancer_Cell_Lines',\n",
    "        'NCI-Nature_2015',\n",
    "        'NCI-Nature_2016',\n",
    "        'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "        'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "        'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "        'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "        'NURSA_Human_Endogenous_Complexome',\n",
    "        'OMIM_Disease',\n",
    "        'OMIM_Expanded',\n",
    "        'Old_CMAP_down',\n",
    "        'Old_CMAP_up',\n",
    "        'PPI_Hub_Proteins',\n",
    "        'Panther_2015',\n",
    "        'Panther_2016',\n",
    "        'Pfam_Domains_2019',\n",
    "        'Pfam_InterPro_Domains',\n",
    "        'PheWeb_2019',\n",
    "        'Phosphatase_Substrates_from_DEPOD',\n",
    "        'ProteomicsDB_2020',\n",
    "        'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "        'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "        'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "        'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "        'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "        'Reactome_2013',\n",
    "        'Reactome_2015',\n",
    "        'Reactome_2016',\n",
    "        'SILAC_Phosphoproteomics',\n",
    "        'SubCell_BarCode',\n",
    "        'SysMyo_Muscle_Gene_Sets',\n",
    "        'TF-LOF_Expression_from_GEO',\n",
    "        'TF_Perturbations_Followed_by_Expression',\n",
    "        'TG_GATES_2020',\n",
    "        'TRANSFAC_and_JASPAR_PWMs',\n",
    "        'TRRUST_Transcription_Factors_2019',\n",
    "        'Table_Mining_of_CRISPR_Studies',\n",
    "        'TargetScan_microRNA',\n",
    "        'TargetScan_microRNA_2017',\n",
    "        'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
    "        'Tissue_Protein_Expression_from_ProteomicsDB',\n",
    "        'Transcription_Factor_PPIs',\n",
    "        'UK_Biobank_GWAS_v1',\n",
    "        'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "        'VirusMINT',\n",
    "        'Virus_Perturbations_from_GEO_down',\n",
    "        'Virus_Perturbations_from_GEO_up',\n",
    "        'WikiPathways_2013',\n",
    "        'WikiPathways_2015',\n",
    "        'WikiPathways_2016',\n",
    "        'WikiPathways_2019_Human',\n",
    "        'WikiPathways_2019_Mouse',\n",
    "        'dbGaP',\n",
    "        'huMAP',\n",
    "        'lncHUB_lncRNA_Co-Expression',\n",
    "        'miRTarBase_2017',\n",
    "    ],\n",
    "    default = [\n",
    "        'ARCHS4_Tissues',\n",
    "        'Aging_Perturbations_from_GEO_down',\n",
    "        'Aging_Perturbations_from_GEO_up',\n",
    "        'Allen_Brain_Atlas_down',\n",
    "        'Allen_Brain_Atlas_up',\n",
    "        'ClinVar_2019',\n",
    "        'Disease_Signatures_from_GEO_down_2014',\n",
    "        'Disease_Signatures_from_GEO_up_2014',\n",
    "        'ENCODE_Histone_Modifications_2013',\n",
    "        'ENCODE_Histone_Modifications_2015',\n",
    "        'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "        'GO_Biological_Process_2018',\n",
    "        'GO_Cellular_Component_2018',\n",
    "        'GO_Molecular_Function_2018',\n",
    "        'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "        'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "        'GWAS_Catalog_2019',\n",
    "        'HMDB_Metabolites',\n",
    "        'Human_Gene_Atlas',\n",
    "        'Jensen_COMPARTMENTS',\n",
    "        'Jensen_DISEASES',\n",
    "        'Jensen_TISSUES',\n",
    "        'KEGG_2019_Human',\n",
    "        'Reactome_2016',\n",
    "        'Tissue_Protein_Expression_from_ProteomicsDB',\n",
    "        'WikiPathways_2019_Human',\n",
    "        'miRTarBase_2017'\n",
    "    ],\n",
    "    section='Analysis_Section'\n",
    ") %}\n",
    "\n",
    "{% set beta_corr = BoolField(\n",
    "    name='beta_corr', \n",
    "    label='Perform cell type correlation?', \n",
    "    default='false', \n",
    "    section='Analysis_Section'\n",
    ") %}\n",
    "\n",
    "{% set circos_plot = BoolField(\n",
    "    name='circos_plot', \n",
    "    label='Visualize cell type correlation with a circosplot?', \n",
    "    default='false', \n",
    "    section='Analysis_Section'\n",
    ") %}\n",
    "\n",
    "{% set esmu_corr = BoolField(\n",
    "    name='esmu_corr', \n",
    "    label='Perform ES gene correlation?', \n",
    "    default='false', \n",
    "    section='Analysis_Section'\n",
    ") %}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "gwas_name = \"{{gwas_name.value}}\"\n",
    "name = \"{{pheno_name.value}}\"\n",
    "\n",
    "hmagma_file = \"{{hmagma_file.value}}\"\n",
    "ldsc_file = \"{{ldsc_file.value}}\"\n",
    "magma_file = \"{{magma_file.value}}\"\n",
    "datah5 = \"{{df_all.value}}\"\n",
    "\n",
    "other_gwas = {{other_gwas.value}}\n",
    "\n",
    "table_counter = 1\n",
    "figure_counter = 1\n",
    "\n",
    "genesetlist = {{genesetlist.value}}\n",
    "gwas_group_dict = {k: v for k,v in constants.GWAS_GROUP_DICT.items() if k in other_gwas}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "gsea = \"{{gsea.value}}\"\n",
    "beta_corr = \"{{beta_corr.value}}\"\n",
    "esmu_corr = \"{{esmu_corr.value}}\"\n",
    "circos_plot = \"{{circos_plot.value}}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The input files will be converted to a pandas dataframe to easily analyse the ouput."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_dict = {\n",
    "    gwas_name: {\n",
    "        'H-MAGMA': hmagma_file,\n",
    "        'LDSC': ldsc_file,\n",
    "        'MAGMA': magma_file\n",
    "    }\n",
    "}\n",
    "df_list_1 = []\n",
    "for d in file_dict.values():\n",
    "    df_list_2 = []\n",
    "    for method, file_path in d.items():\n",
    "        df = pd.read_csv(file_path)\n",
    "        df['method'] = method\n",
    "        df.sort_values(\n",
    "            by=['gwas', 'specificity_id', 'annotation'],\n",
    "            inplace=True)\n",
    "        df_list_2.append(df)\n",
    "    df_list_1.extend(df_list_2)\n",
    "\n",
    "df = pd.concat(df_list_1, ignore_index=True)\n",
    "# count the number of methods (not used atm)\n",
    "df = df.merge(\n",
    "    df.groupby(\n",
    "        ['gwas', 'specificity_id', 'annotation']\n",
    "    ).size().to_frame('n_methods'),\n",
    "    on=['gwas', 'specificity_id', 'annotation'], how='left')\n",
    "# count the number of annotations/celltypes\n",
    "df.sort_values(by=['gwas', 'method'], inplace=True)\n",
    "df.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.    Find Significant Cell-Types <a id='1'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The pvalue of the output must first be corrected.\n",
    "By default it uses bonferroni correction, but other methods can also be applied\n",
    "(see [`statsmodels.stats.multitest.multipletests`](https://www.statsmodels.org/stable/generated/statsmodels.stats.multitest.multipletests.html)\n",
    "documentation)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = convert_output_to_dataframe.pvalue_correction(df, method=constants.PVAL_CORRECTION) # corrects pval\n",
    "\n",
    "# bokeh display df\n",
    "table_counter = display_bokeh_df(df, table_counter, f\"Results of the enrichment analysis on the {name} phenotype.\")\n",
    "create_download_link(df, f\"{name}_enrichment_analysis.xlsx\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To identify the enriched celltypes, we can simply filter out all non-significant hits. In the STEAP pipeline we used three different methods (S-LDSC, MAGMA and H-MAGMA) to determine whether a cell type is enriched or not.\n",
    "Here, we define a cell type to be actually enriched if it is significant in at least 2 methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "significant_counts = df[\n",
    "        (df[f'pvalue_{constants.PVAL_CORRECTION}']<=0.05) # get only significant hits\n",
    "].value_counts(['gwas','specificity_id','annotation']) # count in how many methods it is significant\n",
    "\n",
    "# only get enriched cell-types in >=2 methods\n",
    "significant_counts = significant_counts[\n",
    "    (significant_counts>=len(constants.METHODS)-1)\n",
    "]\n",
    "\n",
    "# bokeh display df\n",
    "bokeh_df = significant_counts.sort_index(level=1).reset_index().rename(columns={0:'significant_method_count'})\n",
    "table_counter = display_bokeh_df(bokeh_df, table_counter, f\"Enriched cell types in the {name} phenotype. Only cell types significant in two or more methods are shown.\")\n",
    "create_download_link(bokeh_df, f\"significant_celltype_counts_{name}.xlsx\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.    Overlap of Enriched Cell-Types Between Phenotypes <a id='2'></a>\n",
    "\n",
    "Let's compare which enriched cell types in the input phenotype are also enriched in other gwas (groups) using the upsetplot. \n",
    "To do this we need data of the other phenotypes as well. \n",
    "Instead of running the pipeline with 500+ different GWAS, we can just load a pre-computed dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.read_hdf(datah5, 'df_all')\n",
    "\n",
    "# bokeh display df\n",
    "bokeh_df = df_all.sample(150) # only sample as df is too large\n",
    "table_counter = display_bokeh_df(bokeh_df, table_counter, f\"Enrichment analysis on a large number of phenotypes. These include structural MRI measurements, white matter DTI tracts and rs-fMRI network phenotypes.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This data contains 538 unique GWAS phenotypes.\n",
    "We can group most of these GWASes in four main groups.\n",
    "1.    Structural MRI Measurements\n",
    "2.    White Matter DTI Tracts\n",
    "3.    Resting State fMRI Network\n",
    "4.    Brain related diseases\n",
    "\n",
    "\n",
    "The `GWAS_GROUP_DICT` variable contains information about these groups:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(gwas_group_dict)\n",
    "# GWAS group name : [(regex) keywords used to match to gwas]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This dictionary contains the GWAS group names as keys and keywords as values.\n",
    "These keywords can be used to search for the GWAS in the dataframe (using regex).\n",
    "For instance `volume` is a keywords used to include `volume: thalamus` phenotype in the `Structural MRI` group.\n",
    "\n",
    "To analyze the phenotype of interest let's include it to the dictionary as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_group_dict[name] = [gwas_name]\n",
    "display(gwas_group_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can visualize the overlap of cell types between phenotypes using the upsetplot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = f'upsetplot_{name}.png'\n",
    "upsetplot.plot_upset(\n",
    "    pd.concat([df_all, df]), # concat the preloaded data with the input data\n",
    "    gwas_group_dict,\n",
    "    sign_threshold=len(constants.METHODS)-1, # only use celltypes significant in >=2 methods\n",
    "    element_size=56,\n",
    "    save=True,\n",
    "    filename=filename\n",
    ")\n",
    "figure_counter =  make_caption(figure_counter, f\"Upsetplot of overlapping cell types between {name} and the brain related phenotypes.\", 'Figure')\n",
    "display_link(filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can investigate the actual cell types which overlap in these phenotypes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shared_celltypes_df = upsetplot.get_shared_celltypes(pd.concat([df_all,df]),\n",
    "                                                     gwas_group_dict,\n",
    "                                                     sign_threshold=len(constants.METHODS)-1,\n",
    "                                                     save_to_excel=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And to see which cell types are overlapping with only the input phenotype:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bokeh display df\n",
    "bokeh_df = shared_celltypes_df[\n",
    "    shared_celltypes_df['group'].str.contains(name)\n",
    "].reset_index(drop=True)\n",
    "table_counter = display_bokeh_df(bokeh_df, table_counter, f\"Cell types overlapping with the three brain measurement groups and {name} phenotype.\")\n",
    "create_download_link(bokeh_df, f\"shared_celltypes_{name}.xlsx\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if gsea.value == True %}\n",
    "## 3.    Gene-Set Enrichment Analysis <a id='3'></a>\n",
    "\n",
    "\n",
    "We can get additional information about the enriched cell types using GSEA.\n",
    "GSEA uses a gene-list as input. \n",
    "Here, we select the top 1% of genes with the highest ES value calculated using CELLEX in enriched cell types.\n",
    "This percentage can be edited in the constant.py file, but it is recommended to not do this. \n",
    "[A range of 15 to 500 genes is recommended by the original GSEA developers](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html). \n",
    "During testing, the top 1% revealed to always have a number of genes in this range.\n",
    "\n",
    "GSEA compares the gene-list to a gene-set. Gene-sets used in the analysis can be found and edited in the constant.py file. \n",
    "The gene-set libraries are from [Enrichr](https://maayanlab.cloud/Enrichr/#stats).\n",
    "The gene-sets used here are:\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if gsea.value == True %}\n",
    "for gene_set in genesetlist:\n",
    "    print(gene_set)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if gsea.value == True %}\n",
    "gsea_dict = gene_set_enrichment_analysis.gsea(\n",
    "    df,\n",
    "    gwas_group_dict={name:[gwas_name]},\n",
    "    gene_set_list=genesetlist,\n",
    ")\n",
    "# since only one gwas is within the gwas group all cell types enriched have rank 1\n",
    "# thus all enriched cell types will be analysed\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if gsea.value == True %}\n",
    "\n",
    "To get the gsea results for all cell types analysed download the .zip file below\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if gsea.value == True %}\n",
    "inputfiles = [f\"gsea_{celltype.replace(', ','-')}.xlsx\" for celltype in gsea_dict.keys()]\n",
    "outfile = f'gsea_{name}.zip'\n",
    "create_zip_file(outfile, inputfiles)\n",
    "display_link(outfile)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if gsea.value == True %}\n",
    "\n",
    "For phenotypes with large number of enriched cell types it can be quite daunting to go through all these gsea files.\n",
    "To quickly see which terms occur the most often in all of the enriched cell types just use the `summarize_gsea` function. This does not take into account the values of the 'Combined Score' gsea calculates.\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if gsea.value == True %}\n",
    "filename = f'gsea_summary_{name}.xlsx'\n",
    "gsea_summary_df = gene_set_enrichment_analysis.summarize_gsea(\n",
    "    gsea_dict,\n",
    "    save_to_excel=True,\n",
    "    filename=filename\n",
    ")\n",
    "\n",
    "table_counter = display_bokeh_df(\n",
    "    gsea_summary_df,\n",
    "    table_counter,\n",
    "    f\"Summary of gsea analysis on enriched cell types for {name}.\\\n",
    "    It shows the terms in the gene-sets which are significant in the {name} enriched cell types.\\\n",
    "    The last columns displays the number of cell types which the term is significant in.\"\n",
    "    )\n",
    "display_link(filename)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if beta_corr.value == True %}\n",
    "\n",
    "## 4.    Cell Type Correlation <a id='4'></a>\n",
    "\n",
    "We can calculate the similarity between phenotypes using the enrichment values acquired from the pipeline. \n",
    "Here, we first concatenate our data to the pre-loaded data, which contains 500+ unqiue phenotypes. \n",
    "\n",
    "These phenotypes were analysed with the same scRNA-seq datasets, which is important for reliable calculation of the cell type correlation.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if beta_corr.value == True %}\n",
    "df_concat = pd.concat([df_all, df])\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if beta_corr.value == True %}\n",
    "\n",
    "Then we calculate the correlation between all of these phenotypes.\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if beta_corr.value == True %}\n",
    "\n",
    "corr_df = calculate_beta_correlation.calculate_celltype_corr(df_concat)\n",
    "table_counter = display_bokeh_df(corr_df.sample(150), table_counter,\n",
    "                                 \"Cell type correlation between all the phenotypes analysed.\\\n",
    "                                 Here only a sample of the full data is shown.\\\n",
    "                                 The full data can be downloaded below.\")\n",
    "create_download_link(corr_df, \"corr_df.xlsx\") \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if beta_corr.value == True %}\n",
    "\n",
    "To only select for significant correlations with our phenotype of interest, we can do:\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if beta_corr.value == True %}\n",
    "\n",
    "bokeh_df = corr_df[\n",
    "    # get only correlations with phenotype\n",
    "    ((corr_df['gwasx'].str.contains(gwas_name))|(corr_df['gwasy'].str.contains(gwas_name)))\n",
    "    &\n",
    "    # get only significant correlations\n",
    "    (corr_df['pval'] <= (0.05/corr_df.shape[0])) # bonferroni correction\n",
    "].pivot(index='method',columns=['gwasx','gwasy'])['corr'].T.reset_index() # where NaN implies not significant correlation\n",
    "table_counter = display_bokeh_df(bokeh_df, table_counter,\n",
    "                                 f\"Significant cell type correlation between {name} and the other analysed phenotypes.\\\n",
    "                                 NaN values are non-significant correlations\")\n",
    "create_download_link(bokeh_df, f\"corr_df_{name}.xlsx\") \n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if beta_corr.value == True %}\n",
    "\n",
    "To visualize the correlation between the phenotypes and out phenotype of interest we can use the circosplot.\n",
    "Again we use the `gwas_group_dict` as input. The other inputs are the correlation dataframe calculated previously and the name of the gwas of interest.\n",
    "\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if beta_corr.value == True and circos_plot.value == True %}\n",
    "\n",
    "filename = f'circosplot_{name}.png'\n",
    "\n",
    "# showing all correlations in the chord plot would get messy so we can limit it to\n",
    "# for example 75 quantile correlations in both positive and negative direction\n",
    "corrs = corr_df[(corr_df['pval'])<=(0.05/corr_df.shape[0])]['corr']\n",
    "corr_limit = (corrs[corrs<0].quantile(.75),corrs[corrs>0].quantile(.75))\n",
    "# plot\n",
    "circosplot.plot(corr_df, gwas_group_dict, gwas_name, \n",
    "                corr_limit=corr_limit,\n",
    "                save=True,\n",
    "                filename=filename\n",
    "               )\n",
    "figure_counter =  make_caption(figure_counter,\n",
    "                               f\"Circosplot {name} and the brain related phenotypes.\\\n",
    "                               The circle itself represents the gwas groups. \\\n",
    "                               The red lines inside are the mean correlations between the gwases inside these groups.\\\n",
    "                               The brigher the red, the higher the correlation.\\\n",
    "                               The three outer barplot circles represent the correlation between {name} and the traits in the gwas groups per methods.\\\n",
    "                               From inside to outside the methods are H-MAGMA. LDSC, MAGMA.\\\n",
    "                               Red bars denote positive correlation and blue negative.\\\n",
    "                               The big bars spanning the three outer circles denote whether the correlation is significant or not in all methods.\",\n",
    "                               'Figure')\n",
    "display_link(filename)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if esmu_corr.value == True %}\n",
    "\n",
    "## 5.     ES Gene Correlation <a id='5'></a>\n",
    "\n",
    "\n",
    "We calculated the correlations between phenotypes, but we can also do this between cell types.\n",
    "This can be useful to double-check an enriched cell type and see what other cell types it is correlated with.\n",
    "We give as input the datasets (specificity_id). \n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if esmu_corr.value == True %}\n",
    "\n",
    "datasets = df['specificity_id'].unique() # get all unique datasets\n",
    "print(datasets)\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if esmu_corr.value == True %}\n",
    "\n",
    "We can calculate the ES gene correlation using:\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if esmu_corr.value == True %}\n",
    "\n",
    "es_corr_df = calculate_es_correlation.calculate_es_corr(datasets) # calculates correlation\n",
    "\n",
    "table_counter = display_bokeh_df(es_corr_df.sample(150), table_counter,\n",
    "                                 \"ES gene correlation between all the cell types in the scRNA-seq datasets.\\\n",
    "                                 Here only a sample of the full data is shown.\\\n",
    "                                 The full data can be downloaded below.\")\n",
    "create_download_link(es_corr_df, \"es_corr_df.xlsx\") \n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if esmu_corr.value == True %}\n",
    "\n",
    "To only select for the correlations with our enriched cell types we can first select these:\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if esmu_corr.value == True %}\n",
    "\n",
    "significant_celltypes = significant_counts.reset_index()[\n",
    "    ['specificity_id','annotation']].drop_duplicates().values.tolist()\n",
    "significant_celltypes = [f'{d}, {c}' for d,c in significant_celltypes]\n",
    "display(significant_celltypes)\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%appyter markdown\n",
    "{% if esmu_corr.value == True %}\n",
    "\n",
    "Then we only show these correlations:\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if esmu_corr.value == True %}\n",
    "\n",
    "bokeh_df = es_corr_df[\n",
    "    (\n",
    "        # get only celltypes significant in phenotype\n",
    "        (es_corr_df['celltypex'].isin(significant_celltypes))\n",
    "        |\n",
    "        (es_corr_df['celltypey'].isin(significant_celltypes))\n",
    "    )\n",
    "    &\n",
    "    # get only significant correlations\n",
    "    (es_corr_df['pval_bonferroni']<=0.05)\n",
    "].sort_values('corr', ascending=False)\n",
    "\n",
    "table_counter = display_bokeh_df(bokeh_df.sample(150), table_counter,\n",
    "                                 f\"ES gene correlation between all the cell types\\\n",
    "                                 in the scRNA-seq datasets and enriched celltypes in {name}.\\\n",
    "                                 Here only a sample of the full data is shown.\\\n",
    "                                 The full data can be downloaded below.\")\n",
    "create_download_link(bokeh_df, f\"es_corr_df_{name}.xlsx\") \n",
    "\n",
    "{% endif %}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
