{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% do SectionField(name='section1', title = '1. Please input a valid gene symbol or Ensembl ID', subtitle = '', img = 'lncRNA_appyter_logo.png')%}\n",
    "{% do SectionField(name='section2', title = '2. Options', subtitle = '', img = 'lncRNA_appyter_logo.png')%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set query = AutocompleteField(name='gene',label='Gene Symbol/Ensembl ID',description='',default='HOTAIR',required=True, choices = load_static(\"lncRNAs.json\"), section='section1') %}\n",
    "{% set options_fast_compute = BoolField(name='fast_compute', label='Precompute', default='true', description='Precompute will retrieve precomputed results from the example files below for a faster run time. Select \\'No\\' to run whole analysis.', section='section2')%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "query = {{ query }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# Report about the Long Non-coding RNA (lncRNA) {{query.raw_value}}\n",
    "\n",
    "Based on lncRNA-gene co-expression, this report provides predictions on the biological functions of {{query.raw_value}}, displays the median expression of {{query.raw_value}} across tissues and cell-lines, and predicts small molecules that may specifically up- or down-regulate the expression of {{query.raw_value}}."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries \n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import h5py as h5\n",
    "from IPython.display import display,FileLink, HTML, Markdown\n",
    "import os\n",
    "from utils import *\n",
    "import s3fs\n",
    "from bokeh.io import output_notebook\n",
    "import json\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save gene name in json\n",
    "with open('gene.json', 'w') as f:\n",
    "    json.dump([query], f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### Import lncRNA-gene co-expression matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This lncRNA-gene co-expression matrix was generated by computing the Pearson correlation coefficients for 10,000 randomly selected bulk RNA-seq samples from Recount3 [1]. NOTE: If an Ensembl ID was entered, it will be converted to its corresponding gene symbol if available. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Import lncRNA-gene co-expression matrix\n",
    "s3 = s3fs.S3FileSystem(anon=True, client_kwargs=dict(endpoint_url='https://s3.appyters.maayanlab.cloud'))\n",
    "f = h5.File(s3.open('storage/lncRNA_Appyter/v0.0.6/Recount3_lncRNA_pcorr_cols.h5', 'rb'), 'r') \n",
    "corr =f[\"data/correlation\"]\n",
    "col_genes = [x.decode('UTF-8') for x in f[\"meta/columns/genes\"]]\n",
    "col_genes_ensembl = [x.decode('UTF-8') for x in f[\"meta/columns/ensembl\"]]\n",
    "row_genes =  [x.decode('UTF-8') for x in f[\"meta/rows/genes\"]]\n",
    "row_genes_ensembl = [x.decode('UTF-8') for x in f[\"meta/rows/ensembl\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert input Ensembl ID to gene symbol\n",
    "ensembl_2_genes = dict(zip(row_genes_ensembl,row_genes))\n",
    "genes_2_ensembl = dict(zip(row_genes,row_genes_ensembl))\n",
    "if query in col_genes_ensembl:\n",
    "    query_new = ensembl_2_genes[query]\n",
    "    if query != query_new:\n",
    "        print('The Ensembl ID ' + query + ' has been converted to the gene symbol ' + query_new )\n",
    "        query = query_new\n",
    "        print('Predicting functions for ' + query + '(' + genes_2_ensembl[query] + ')')\n",
    "    else:\n",
    "        print('Predicting functions for ' + query + '(' + genes_2_ensembl[query] + ')')\n",
    "else:\n",
    "    if query in row_genes:\n",
    "        print('Predicting functions for ' + query + '(' + genes_2_ensembl[query] + ')')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### Top genes correlated with {{query.raw_value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "Using the loaded lncRNA-gene correlation matrix, we report the genes that are most correlated with {{query.raw_value}}."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find most correlated genes and lncRNAs with the input lncRNA\n",
    "if not os.path.exists(\"gene_correlations/\"):\n",
    "        os.makedirs(\"gene_correlations/\", exist_ok=True)\n",
    "\n",
    "# Get index of lncRNA of interest\n",
    "idx_query = np.where(np.asarray(col_genes) == query)[0][0]\n",
    "\n",
    "# Ranks genes based on pearson correlation with the lncRNA of interest\n",
    "lncRNA_coexp = pd.DataFrame(corr[:,idx_query])\n",
    "lncRNA_coexp.index = row_genes\n",
    "lncRNA_coexp.columns = [\"Pearson's Correlation Coefficient\"]\n",
    "lncRNA_coexp = lncRNA_coexp.sort_values(by=\"Pearson's Correlation Coefficient\", ascending=False)\n",
    "\n",
    "# save gene correlations to csv file\n",
    "lncRNA_coexp.to_csv('gene_correlations/'+ query + '_correlated_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(lncRNA_coexp[0:20])\n",
    "display(Markdown(f\"*Table 1. The Top 20 genes correlated with {query} ranked by Pearsonâ€™s correlation coefficients.*\"))\n",
    "display(FileLink('gene_correlations/' + query + '_correlated_genes.csv', result_html_prefix=str('Download Table 1: ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### Interactive network visualization of the top 100 genes correlated with {{query.raw_value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "Interactive network visualization of the top 100 genes correlated with {{query.raw_value}}. Each node represents a gene and is colored by chromosome location, except for the bright red node which represents the lncRNA {{query.raw_value}}. The thickness of the edges corresponds to Pearson correlation coefficients. Clicking on a gene node will highlight its corresponding edges in orange. Hovering over a node will display the gene name and chromosome location.\n",
    "\n",
    "Network Methods: All pairwise correlations between the top 100 genes correlated with {{query.raw_value}} are extracted. The 3 edges with the highest correlation per gene node are used to initialize the network. Edges with weights < 0.3 are dropped. To further prune the network, the edge with the lowest weight for each hub node is dropped. At the start, a hub node is defined as a node with > 10 edges. The pruning process is repeated until the network has an average of < 3 edges per node. The top 5 edges for {{query.raw_value}} are shown regardless of their weights."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Visualize the co-expression network for the top 100 correlated genes\n",
    "if not os.path.exists(\"coexpression_network/\"):\n",
    "        os.makedirs(\"coexpression_network/\", exist_ok=True)\n",
    "g = network_vis(query,lncRNA_coexp,genes_2_ensembl,row_genes)\n",
    "g[1].to_csv(\"coexpression_network/\"+query+'_network_node_metadata.csv')\n",
    "g[2].to_csv(\"coexpression_network/\"+query+'_network_edge_metadata.csv')\n",
    "g[0].show('coexpression_network/' + query + '_network.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(Markdown(f\"*Figure 1. Interactive network visualization of the top 100 genes correlated with {query}.*\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(FileLink('coexpression_network/' + query + '_network.html', result_html_prefix=str('Download Figure 1: ')))\n",
    "display(FileLink(\"coexpression_network/\"+query+'_network_node_metadata.csv', result_html_prefix=str('Download Node metadata: ')))\n",
    "display(FileLink(\"coexpression_network/\"+query+'_network_edge_metadata.csv', result_html_prefix=str('Download Edge metadata: ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### Enrichment analysis applied to the top 200 genes most correlated with {{query.raw_value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "The top 200 genes most correlated with {{query.raw_value}} are submitted to Enrichr [2-4] for enrichment analysis. NOTE: Only genes with official Entrez gene symbols are submitted to Enrichr. Ensembl IDs that do not map to an official gene symbol were dropped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Submit top 200 gene symbols to Enrichr\n",
    "# For users running the notebook locally: To adjust the number of genes submitted to Enrichr, change the value of n below\n",
    "n = 200\n",
    "top_n_genes = [x for x in list(lncRNA_coexp.index) if not x.startswith('ENSG')]\n",
    "top_n_genes = top_n_genes[0:n]\n",
    "enrichr_link = Enrichr_API(top_n_genes,str('Top 200 correlated genes with the lncRNA: ' + query))\n",
    "\n",
    "# Save enrichr link to text file\n",
    "if not os.path.exists(\"enrichment_analysis/\"):\n",
    "        os.makedirs(\"enrichment_analysis/\", exist_ok=True)\n",
    "open_file = open('enrichment_analysis/'+query+'_top_'+str(n)+'_correlated_genes_Enrichr_link.txt','w')\n",
    "open_file.write(enrichr_link)\n",
    "open_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(HTML(\"Access the enrichment analysis results for the top 200 most correlated with {query} here: <a href='{href}'>{link}</a>\".format(href=enrichr_link, link = enrichr_link, query=query)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### Top lncRNAs correlated with {{query.raw_value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "Below we list the top 20 lncRNAs, out of all 15,862 lncRNAs within our database, that correlate most with {{query.raw_value}} based on their Pearson correlation coefficients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download most correlated lncRNAs\n",
    "lncRNA_lncRNA_coexp = lncRNA_coexp.loc[col_genes]\n",
    "lncRNA_lncRNA_coexp = lncRNA_lncRNA_coexp.sort_values(by=\"Pearson's Correlation Coefficient\", ascending=False)\n",
    "\n",
    "# save gene correlations to csv file\n",
    "lncRNA_lncRNA_coexp.to_csv('gene_correlations/' + query + '_correlated_lncRNAs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(lncRNA_lncRNA_coexp[0:20])\n",
    "display(Markdown(f\"*Table 2. Top 20 lncRNAs that correlate most with {query} ranked by Pearson correlation coefficients.*\"))\n",
    "display(FileLink('gene_correlations/' + query + '_correlated_lncRNAs.csv', result_html_prefix=str('Download Table 2: ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### Predicted biological functions of {{query.raw_value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "For each Enrichr [2-4] library, we compute the mean Pearson correlation coefficients for each gene set by averaging the Pearson correlation coefficients between each gene in the gene set and {{query.raw_value}}. Terms with high mean Pearson correlation coefficients are prioritized. These terms are predicted to be associated with {{query.raw_value}}."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if options_fast_compute.raw_value == False %}\n",
    "\n",
    "# Folder for biological function predictions\n",
    "if not os.path.exists(\"predicted_functions/\"):\n",
    "    os.makedirs(\"predicted_functions/\", exist_ok=True)\n",
    "\n",
    "# Make function predictions\n",
    "prediction_libraries = ['MGI_Mammalian_Phenotype_Level_4_2021','GO_Biological_Process_2021','KEGG_2021_Human','DisGeNET']\n",
    "prediction_libraries = np.array_split(prediction_libraries, int(np.ceil(len(prediction_libraries)/2))) \n",
    "for i_group,group in enumerate(prediction_libraries):\n",
    "    predictions = []\n",
    "    library_names = []\n",
    "    for pred_library in group:\n",
    "        predictions.append(predict_functions(pred_library,lncRNA_coexp,query))\n",
    "        library_names.append(pred_library.replace('_',' '))\n",
    "    if i_group == 0:\n",
    "        plot_results(library_names=library_names, results_dfs=predictions,file_name='predicted_functions/'+query+'_biological_function_predictions_figure2')\n",
    "        display(Markdown(f\"*Figure 2. Predicted MGI Mammalian Phenotypes and GO Biological Processes for the lncRNA {query}. Terms are ranked by averaging the mean Pearson correlation coefficients between each gene in a gene set and {query}.*\" ))\n",
    "    if i_group == 1:\n",
    "        plot_results(library_names=library_names, results_dfs=predictions,file_name='predicted_functions/'+query+'_biological_function_predictions_figure3')\n",
    "        display(Markdown(f\"*Figure 3. Predicted KEGG pathways and DisGeNET disease terms for the lncRNA {query}. Terms are ranked by averaging the Pearson correlation coefficients between each gene in a gene set and {query}.*\" ))\n",
    "\n",
    "    \n",
    "    # Save Predictions \n",
    "    for ii,prediction in enumerate(predictions):\n",
    "        prediction.to_csv(\"predicted_functions/\" + library_names[ii]+'_' + query + '.csv')\n",
    "        display(FileLink(\"predicted_functions/\" + library_names[ii]+'_' + query + '.csv', result_html_prefix=str('Download predictions: ')))\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if options_fast_compute.raw_value == True %}\n",
    "\n",
    "# Folder for biological function predictions\n",
    "if not os.path.exists(\"predicted_functions/\"):\n",
    "    os.makedirs(\"predicted_functions/\", exist_ok=True)\n",
    "\n",
    "# Import pre-computed lncRNA functions\n",
    "prediction_libraries = ['MGI_Mammalian_Phenotype_Level_4_2021','GO_Biological_Process_2021','KEGG_2021_Human','DisGeNET','ChEA_2016','ENCODE_TF_ChIP-seq_2015']\n",
    "versions = ['v0.0.6','v0.0.6','v0.0.6','v0.0.6','v0.0.8','v0.0.8']\n",
    "predlib_2_version = dict(zip(prediction_libraries,versions))\n",
    "prediction_libraries = np.array_split(prediction_libraries, int(np.ceil(len(prediction_libraries)/2))) \n",
    "for i_group,group in enumerate(prediction_libraries):\n",
    "    predictions = []\n",
    "    library_names = []\n",
    "    for pred_library in group:\n",
    "        precomputed_df_indexes = list(pd.read_csv(s3.open('storage/lncRNA_Appyter/'+predlib_2_version[pred_library]+'/'+ pred_library + '_lncRNA_avg_coexpression_gene_index_order.txt','rb'),header=None)[0])\n",
    "        query_idx = np.where(np.asarray(precomputed_df_indexes) == query)[0][0] \n",
    "        precomputed_avg_coexp = pd.read_csv(s3.open('storage/lncRNA_Appyter/'+predlib_2_version[pred_library]+'/'+ pred_library + '_lncRNA_avg_coexpression.csv','rb'),index_col=0,header=None, skiprows=query_idx+1,nrows=1)\n",
    "        cols = list(pd.read_csv(s3.open('storage/lncRNA_Appyter/'+predlib_2_version[pred_library]+'/'+ pred_library + '_lncRNA_avg_coexpression.csv','rb'),header=0, index_col = 0, nrows=1).columns)\n",
    "        precomputed_avg_coexp.columns = cols\n",
    "        precomputed_avg_coexp = precomputed_avg_coexp.T\n",
    "        precomputed_avg_coexp = precomputed_avg_coexp.sort_values(by=query,ascending = False).reset_index().rename({'index': 'Term', query: 'Mean Pearson Correlation'}, axis='columns') \n",
    "        cols=0\n",
    "        predictions.append(precomputed_avg_coexp)\n",
    "        library_names.append(pred_library.replace('_',' '))\n",
    "    if i_group == 0:\n",
    "        plot_results(library_names=library_names, results_dfs=predictions,file_name='predicted_functions/'+query+'_biological_function_predictions_figure2')\n",
    "        display(Markdown(f\"*Figure 2. Predicted MGI Mammalian Phenotypes and GO Biological Processes for the lncRNA {query}. Terms are ranked by averaging the mean Pearson correlation coefficients between each gene in a gene set and {query}.*\" ))\n",
    "    if i_group == 1:\n",
    "        plot_results(library_names=library_names, results_dfs=predictions,file_name='predicted_functions/'+query+'_biological_function_predictions_figure3')\n",
    "        display(Markdown(f\"*Figure 3. Predicted KEGG pathways and DisGeNET disease terms for the lncRNA {query}. Terms are ranked by averaging the mean Pearson correlation coefficients between each gene in a gene set and {query}.*\" ))\n",
    "    if i_group == 2:\n",
    "        plot_results(library_names=library_names, results_dfs=predictions,file_name='predicted_functions/'+query+'_biological_function_predictions_figure4')\n",
    "        display(Markdown(f\"*Figure 4. Predicted ChEA and ENCODE terms for the lncRNA {query}. Terms are ranked by averaging the mean Pearson correlation coefficients between each gene in a gene set and {query}.*\" ))\n",
    "\n",
    "    # Save Predictions \n",
    "    for ii,prediction in enumerate(predictions):\n",
    "        prediction.to_csv(\"predicted_functions/\" + library_names[ii]+'_' + query + '.csv')\n",
    "        display(FileLink(\"predicted_functions/\" + library_names[ii]+'_' + query + '.csv', result_html_prefix=str('Download predictions: ')))\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### Expression of {{query.raw_value}} across tissues and cell lines\n",
    "This part of the report provides the Z-score (Normalized Median Expression) for the lncRNA {{query.raw_value}} in various tissues and cell lines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "Samples from Recount3 [1] were automatically labelled by tissue type or cell line of origin. Tissue and cell line samples were log2 transformed and quantile normalized separately. Tissues and cell lines with less than 20 samples were removed, and z-scores were computed along the lncRNA axis to compare expression levels across all tissues and cell lines. The median expression of {{query.raw_value}} was then calculated for each tissue type and cell line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Import z-score data\n",
    "tissue_expr_zscore  = pd.read_csv(s3.open('storage/lncRNA_Appyter/v0.0.6/lncRNA_zscore_median_expr_by_tissue_filtered.csv', 'rb'),header=0, index_col=0) \n",
    "tissue_expr_zscore.index = [x.split(',')[0] for x in tissue_expr_zscore.index]\n",
    "\n",
    "\n",
    "cell_line_expr_zscore = pd.read_csv(s3.open('storage/lncRNA_Appyter/v0.0.6/lncRNA_zscore_median_expr_by_cell_line_filtered.csv', 'rb'),header=0, index_col=0) \n",
    "cell_line_expr_zscore.index = [x.split(',')[0] for x in cell_line_expr_zscore.index]\n",
    "cell_line_expr_zscore.columns = [x.upper().replace(' CELL','') for x in cell_line_expr_zscore.columns]\n",
    "\n",
    "# Create folder for tissue and cell line specific expression\n",
    "if not os.path.exists(\"tissue_and_cell_line_expression/\"):\n",
    "    os.makedirs(\"tissue_and_cell_line_expression\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rank tissues by z-score(median expression)\n",
    "tissue_specific_lncRNA = pd.DataFrame(tissue_expr_zscore.loc[query])\n",
    "tissue_specific_lncRNA = tissue_specific_lncRNA.sort_values(by=query,ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_bar(tissue_specific_lncRNA ,query,'Tissues','Z-score (Median Expression)','tissue_and_cell_line_expression/'+query+'_zscore_tissue_expression')\n",
    "display(Markdown(f\"*Figure 5. Z-score (median expression) for the lncRNA {query} in various tissue types.*\"))\n",
    "tissue_specific_lncRNA.to_csv(\"tissue_and_cell_line_expression/\" + query + '_tissue_zscore' + '.csv')\n",
    "display(FileLink(\"tissue_and_cell_line_expression/\" + query + '_tissue_zscore' + '.csv', result_html_prefix=str('Download table with z-score (median expression) values for ' + query + ' in various tissue types: ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rank cell lines by z-score(median expression)\n",
    "cell_line_specific_lncRNA = pd.DataFrame(cell_line_expr_zscore.loc[query])\n",
    "cell_line_specific_lncRNA = cell_line_specific_lncRNA.sort_values(by=query,ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_bar(cell_line_specific_lncRNA[0:30],query,'Cell Lines','Z-score (Median Expression)','tissue_and_cell_line_expression/'+query+'_zscore_cell_line_expression')\n",
    "display(Markdown(f\"*Figure 6. Z-score (median expression) for the lncRNA {query} in the top 30 cell lines.*\"))\n",
    "cell_line_specific_lncRNA.to_csv(\"tissue_and_cell_line_expression/\" + query + '_cell_line_zscore' + '.csv')\n",
    "display(FileLink(\"tissue_and_cell_line_expression/\" + query + '_cell_line_zscore' + '.csv', result_html_prefix=str('Download table with z-score (median expression) values for ' + query + ' in various cell lines: ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### Visualizing all lncRNAs based on their gene expression similarity across tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "We applied UMAP [5] to visualize lncRNA expression across 3,000 randomly selected samples (with tissue type labels) from Recount3[1]. Samples were first log2 transformed and quantile normalized along the gene axis, then UMAP was applied to the lncRNA expression data with samples as features. Each data point represents a single lncRNA (n=15,862). Use the drop-down menu to color lncRNAs by expression z-score in a specific tissue. The black arrow is pointing to the location of {{query.raw_value}}."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "umap_tissue_results = pd.read_csv(s3.open('storage/lncRNA_Appyter/v0.0.6/umap_tissues.csv', 'rb'),header=0, index_col=0) \n",
    "tissue_expr_median_expr = pd.read_csv(s3.open('storage/lncRNA_Appyter/v0.0.6/lncRNA_zscore_median_expr_by_tissue_filtered.csv', 'rb'),header=0, index_col=0) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "values_dict_tz = dict()\n",
    "unique_tissues = np.unique(tissue_expr_zscore.columns)\n",
    "for t in unique_tissues:\n",
    "    values_dict_tz[t] = list(tissue_expr_zscore[t])\n",
    "\n",
    "\n",
    "# Create folder for tissue and cell line specific expression\n",
    "if not os.path.exists(\"umap/\"):\n",
    "    os.makedirs(\"umap/cell_lines/figures/static\", exist_ok=True)\n",
    "    os.makedirs(\"umap/tissues/figures/static\", exist_ok=True)\n",
    "    \n",
    "plot_dynamic_scatter(umap_df=umap_tissue_results, values_dict=values_dict_tz,option_list=list(unique_tissues) ,sample_names=list(umap_tissue_results.index),caption_text='UMAP was applied to 3,000 randomly selected samples (with tissue type labels) from Recount3. Each data point represents a lncRNA (n=15,862) and are colored by z-score (median expression) in ', figure_counter=7,category_list_dict=None, category=False,dropdown=True,color_by_title='Z-score',highlight_query=query,first_selection=list(tissue_specific_lncRNA.index)[0],static_images_save = list(tissue_specific_lncRNA.index[0:2]), file_path='umap/tissues/figures/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### Visualizing all lncRNAs based on their gene expression similarity across cell lines "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "We applied UMAP [5] to visualize lncRNA expression across 3,000 randomly selected samples (with cell line labels) from Recount3[1]. Samples were first log2 transformed and quantile normalized along the gene axis, then UMAP was applied to the lncRNA expression data with samples as features. Each data point represents a single lncRNA (n=15,862). Use the drop-down menu to color lncRNAs by expression z-score in a specific cell line. The black arrow is pointing to the location of {{query.raw_value}}."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "umap_cell_line_results = pd.read_csv(s3.open('storage/lncRNA_Appyter/v0.0.6/umap_cell_lines.csv', 'rb'),header=0, index_col=0) \n",
    "cell_line_expr_median_expr = pd.read_csv(s3.open('storage/lncRNA_Appyter/v0.0.6/lncRNA_zscore_median_expr_by_cell_line_filtered.csv', 'rb'),header=0, index_col=0) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "values_dict_cz = dict()\n",
    "unique_cell_lines = np.unique(cell_line_expr_zscore.columns)\n",
    "for t in unique_cell_lines:\n",
    "    values_dict_cz[t] = list(cell_line_expr_zscore[t])\n",
    "    \n",
    "plot_dynamic_scatter(umap_df=umap_cell_line_results, values_dict=values_dict_cz,option_list=list(unique_cell_lines) ,sample_names=list(umap_cell_line_results.index),caption_text='UMAP was applied to 3,000 randomly selected samples (with cell line labels) from Recount3. Each data point represents a lncRNA (n=15,862) and are colored by z-score (median expression) in ', figure_counter=8,category_list_dict=None, category=False,dropdown=True,color_by_title='Z-score',highlight_query=query,first_selection=list(cell_line_specific_lncRNA.index)[0], static_images_save = list(cell_line_specific_lncRNA.index[0:2]), file_path='umap/cell_lines/figures/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### L1000 small molecules predicted to modulate {{query.raw_value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "~1.4 million Level 5 L1000 chemical perturbation gene expression signatures were downloaded from SigCom LINCS (https://maayanlab.cloud/sigcom-lincs) [6]. For each unique signature and lncRNA pair, a mean Pearson correlation coefficient was computed by taking the average Pearson coefficient between the lncRNA and all genes in the signature. All 15,862 lncRNAs were then ranked by mean Pearson correlation coefficient, and the top 1,000 lncRNAs with the highest coefficients were retained for each signature. The top 500 lncRNA-L1000 signature associations are reported here for {{query.raw_value}}, separated by direction. If {{query.raw_value}} is highly correlated with the up-regulated genes for a specific small molecule, then this small molecule is predicted to up-regulate {{query.raw_value}}."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### L1000 small molecules predicted to up-Regulate {{query.raw_value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "The prioritized small molecules below are predicted to specifically up-regulate {{query.raw_value}}."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Load predicted small molecules to modulate the lncRNA\n",
    "l1000_prediction_file = pd.read_csv(s3.open('storage/lncRNA_Appyter/v0.0.6/l1000_sm_lncRNAs_final.tsv', 'rb'),sep='\\t',header=None, index_col=0)\n",
    "if query in list(l1000_prediction_file.index):\n",
    "    l1000_prediction_file = l1000_prediction_file.loc[query]\n",
    "    l1000_prediction_file = list(l1000_prediction_file.dropna(axis=0))\n",
    "    drugs_up = []\n",
    "    drugs_up_corr = []\n",
    "    drugs_down = []\n",
    "    drugs_down_corr=[]\n",
    "    for sig in l1000_prediction_file:\n",
    "        sig_id = sig.split(',')[0]\n",
    "        sig_id_corr = float(sig.split(',')[1])\n",
    "        if sig_id.split(' ')[1] == 'down':\n",
    "            drugs_down.append(sig_id)\n",
    "            drugs_down_corr.append(sig_id_corr)\n",
    "        else:\n",
    "            drugs_up.append(sig_id)\n",
    "            drugs_up_corr.append(sig_id_corr)\n",
    "\n",
    "    up_results = pd.DataFrame({'L1000 Signature ID':drugs_up,'Drug': [x.split('_')[4] for x in drugs_up],'Up/Down':[x.split(' ')[1] for x in drugs_up],'Dose':[x.split(' ')[0].split('_')[-1] for x in drugs_up],'Cell line':[x.split('_')[1] for x in drugs_up],'Time point':[x.split('_')[2] for x in drugs_up],'Mean Pearson Correlation':drugs_up_corr})\n",
    "    down_results = pd.DataFrame({'L1000 Signature ID':drugs_down,'Drug': [x.split('_')[4] for x in drugs_down],'Up/Down':[x.split(' ')[1] for x in drugs_down],'Dose':[x.split(' ')[0].split('_')[-1] for x in drugs_down],'Cell line':[x.split('_')[1] for x in drugs_down],'Time point':[x.split('_')[2] for x in drugs_down],'Mean Pearson Correlation':drugs_down_corr})\n",
    "else:\n",
    "    up_results = pd.DataFrame()\n",
    "    down_results = pd.DataFrame()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(up_results) > 0:\n",
    "    display(up_results[0:20])\n",
    "    display(Markdown(f\"*Table 3. L1000 small molecules predicted to up-regulate the lncRNA {query}.*\"))\n",
    "\n",
    "    # Create folder for tissue and cell line specific expression\n",
    "    if not os.path.exists(\"l1000_sm_predictions/\"):\n",
    "        os.makedirs(\"l1000_sm_predictions/\", exist_ok=True)\n",
    "    up_results.to_csv(\"l1000_sm_predictions/\" + query + '_l1000_sm_predictions_up' + '.csv')\n",
    "    display(FileLink(\"l1000_sm_predictions/\" + query + '_l1000_sm_predictions_up' + '.csv', result_html_prefix=str('Download table of L1000 small molecules predicted to up-regulate ' + query + ': ')))\n",
    "else:\n",
    "    display(Markdown(f\"**There are no small molecules predicted to specifically up-regulate the expression of {query}.**\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "### L1000 small molecules predicted to down-regulate {{query.raw_value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "The prioritized small molecules below are predicted to specifically down-regulate {{query.raw_value}}."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(down_results) > 0:\n",
    "    display(down_results[0:20])\n",
    "    display(Markdown(f\"*Table 4. L1000 small molecules predicted to down-regulate the lncRNA {query}.*\"))\n",
    "    # Create folder for tissue and cell line specific expression\n",
    "    if not os.path.exists(\"l1000_sm_predictions/\"):\n",
    "        os.makedirs(\"l1000_sm_predictions/\", exist_ok=True)\n",
    "    down_results.to_csv(\"l1000_sm_predictions/\" + query + '_l1000_sm_predictions_down' + '.csv')\n",
    "    display(FileLink(\"l1000_sm_predictions/\" + query + '_l1000_sm_predictions_down' + '.csv', result_html_prefix=str('Download table of L1000 small molecules predicted to down-regulate ' + query + ': ')))\n",
    "else:\n",
    "    display(Markdown(f\"**There are no small molecules predicted to specifically down-regulate the expression of {query}.**\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# close h5 file\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### References\n",
    "[1] Wilks C, Zheng SC, Chen FY, Charles R, Solomon B, Ling JP, Imada EL, Zhang D, Joseph L, Leek JT: recount3: summaries and queries for large-scale RNA-seq expression and splicing. bioRxiv 2021:2021.2005.2021.445138.\n",
    "\n",
    "[2] Xie Z, Bailey A, Kuleshov MV, Clarke DJB, Evangelista JE, Jenkins SL, Lachmann A, Wojciechowicz ML, Kropiwnicki E, Jagodnik KM: Gene Set Knowledge Discovery with Enrichr. Current Protocols 2021, 1(3):e90.\n",
    "\n",
    "[3] Chen EY, Tan CM, Kou Y, Duan Q, Wang Z, Meirelles GV, Clark NR, Maâ€™ayan A: Enrichr: interactive and collaborative HTML5 gene list enrichment analysis tool. BMC Bioinformatics 2013, 14(1):128.\n",
    "\n",
    "[4] Kuleshov MV, Jones MR, Rouillard AD, Fernandez NF, Duan Q, Wang Z, Koplev S, Jenkins SL, Jagodnik KM, Lachmann A: Enrichr: a comprehensive gene set enrichment analysis web server 2016 update. Nucleic Acids Research 2016, 44(W1):W90-W97.\n",
    "\n",
    "[5] McInnes L, Healy J, Melville J: Umap: Uniform manifold approximation and projection for dimension reduction. arXiv preprint arXiv:180203426 2018.\n",
    "\n",
    "[6] Evangelista et al. SigCom LINCS: Data and Metadata Search Engine for Gene Expression Signatures. 2021. In preparation.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
