{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "<center>\n",
    "    <h1 id = \"top-of-app\"> \n",
    "        <div style=\"font-size:3rem;font-weight:500\"> <img src=\"{{ url_for('static', filename='cluster-icon.svg') }}\" style=\"height:45px;padding:0 5px;display:inline\"/> Patient Cohorts RNA-Seq Viewer</div>\n",
    "    </h1>\n",
    "    <br>\n",
    "    <div style=\"font-size:2rem;font-weight:500\">An appyter for the visualization and analysis of intra-cancer patient clusters based on RNA-Seq profiles and clinical data</div>\n",
    "</center>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[The Cancer Genome Atlas (TCGA)](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga) dataset contains multiomics profiling and clinical data from over 10,000 tumors collected from patients spanning several cancer types. Specifically, TCGA has bulk RNA-sequencing (RNA-Seq) profiling of tumors, which can provide insights into mechanisms and classify tumors by subtype.\n",
    "\n",
    "By default, this appyter provides analysis and visualization of TCGA datasets. Users can optionally upload their own datasets.\n",
    "\n",
    "The appyter provides analysis for RNA-Seq TCGA data for cancers with over 150 cases. The report automatically identifies clusters of patient and determines which clinical features and genes are most associated with each cluster.\n",
    "\n",
    "For the TCGA data, each column in the RNA-Seq dataset corresponds to a row in the clinical dataset; both are referenced by the same identifier (here the case_id as provided by TCGA).\n",
    "\n",
    "The RNA-Seq data loaded from TCGA is in the form of raw counts mapped to genes with the [htseq-count](https://htseq.readthedocs.io/en/release_0.9.0/count.html) analysis package; the same format should be followed for user-uploaded files. The analysis filters out lowly expressed genes, identifies the most variable genes, normalize the counts, and reduces the dimensionality of the dataset further with PCA and UMAP.\n",
    "\n",
    "To determine the ideal number of clusters, the analysis tests a range of possible K clusters, and selects the optimal number based on a modified silhouette score that prioritizes more clusters to avoid missing out small clusters.\n",
    "\n",
    "The appyter also identifies the top genes for each cluster, using these for enrichment analysis and suggestion for drugs and small molecules based on the drugs that mimic or reverse the signatures obtained for each cluster. Such drug suggestions are based on the L1000 dataset, using the L1000FWD API. It should be noted that these are speculative predictions and should not be applied to patients before carefully tested in cell based assays and animal models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import requests\n",
    "import time\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "from umap import UMAP\n",
    "from scipy.stats import zscore\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn import metrics\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_samples, silhouette_score\n",
    "import matplotlib.cm as cm\n",
    "from maayanlab_bioinformatics.dge import characteristic_direction\n",
    "from maayanlab_bioinformatics.normalization import log2_normalize, filter_by_var, zscore_normalize\n",
    "import qnorm\n",
    "from maayanlab_bioinformatics.utils import merge\n",
    "import plotly.express as px\n",
    "import plotly\n",
    "import math\n",
    "from collections import OrderedDict\n",
    "import json\n",
    "from jupyter_d3 import scatter_plot\n",
    "from IPython.display import display, IFrame, Markdown, HTML\n",
    "from textwrap import wrap\n",
    "from lifelines import KaplanMeierFitter\n",
    "from lifelines.statistics import multivariate_logrank_test, pairwise_logrank_test\n",
    "import plotly.graph_objects as go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "\n",
    "{% do SectionField(\n",
    "    name=\"DATASET\",\n",
    "    title=\"Dataset selection\",\n",
    "    subtitle='If using TCGA data, leave both file upload fields blank and simply select the desired cancer type.',\n",
    "    img = \"tcga-logo.png\"\n",
    ") %}\n",
    "\n",
    "\n",
    "{% do SectionField(\n",
    "    name='CONFIG',\n",
    "    title='Analysis parameters',\n",
    "    subtitle='Select various parameters for dimensionality reduction and dataset size. Defaults for n_neighbors and min_cluster_dist are based on the defaults used by the Seurat R Package for single cell genomics analysis.',\n",
    "    img = \"parameters-icon.svg\"\n",
    "\n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name=\"ENRICHR_LIBS\",\n",
    "    title=\"Libraries to include in the Enrichr search\",\n",
    "    img = \"enrichr-logo.png\"\n",
    ") %}\n",
    "\n",
    "\n",
    "{% do SectionField(\n",
    "    name=\"SURVIVAL\",\n",
    "    title=\"Survival analysis parameters\",\n",
    "    subtitle=\"If using TCGA data or not interested in survival analysis, leave any (or all) fields blank.\",\n",
    "    img = \"survival-icon.svg\"\n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name=\"L1000FWD\",\n",
    "    title=\"L1000FWD search parameters\",\n",
    "    img = \"l1000fwd.png\"\n",
    ") %}\n",
    "\n",
    "{% set data_filename = FileField(\n",
    "    name='data_filename', \n",
    "    label='RNA-seq data file (.csv)', \n",
    "    description='Upload RNA-seq dataset in csv format. The index of the dataset are genes, the columns are samples.', \n",
    "    default='',\n",
    "    examples={'Papillary adenocarcinoma, NOS_data.csv': 'https://tcga-enrichr-viewer.s3.us-east-2.amazonaws.com/Papillary+adenocarcinoma%2C+NOS_data.csv'},\n",
    "    section='DATASET'\n",
    ") %}\n",
    "\n",
    "{% set clinical_data_filename = FileField(\n",
    "    name='clinical_data_filename', \n",
    "    label='Clinical data file (.csv)', \n",
    "    description='Upload clinical data in csv format. The first column should contain patient IDs corresponding to those in the RNA-seq file. Each subsequent column is a clinical data field.', \n",
    "    default='',\n",
    "    examples={'Papillary adenocarcinoma, NOS_clinical_data.csv': 'https://tcga-enrichr-viewer.s3.us-east-2.amazonaws.com/Papillary+adenocarcinoma%2C+NOS_clinical_data.csv'},\n",
    "    section='DATASET'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "data_filename = {{ data_filename }}\n",
    "clinical_data_filename = {{ clinical_data_filename }}\n",
    "\n",
    "cancer = '''{{ ChoiceField(\n",
    "        name = \"cancer\",\n",
    "        label = \"Cancer type\",\n",
    "        description=\"The value provided as the primary diagnosis on cases in TCGA.\",\n",
    "        choices=[\"Infiltrating duct carcinoma, NOS\",\"Squamous cell carcinoma, NOS\",\"Lobular carcinoma, NOS\",\"Acute myeloid leukemia, NOS\",\"Hepatocellular carcinoma, NOS\",\"Serous cystadenocarcinoma, NOS\",\"Endometrioid adenocarcinoma, NOS\",\"Adenocarcinoma, NOS\",\"Clear cell adenocarcinoma, NOS\",\"Glioblastoma\",\"Mucinous adenocarcinoma\",\"Transitional cell carcinoma\",\"Malignant melanoma, NOS\",\"Papillary adenocarcinoma, NOS\"],\n",
    "        section=\"DATASET\",\n",
    "        default=\"Papillary adenocarcinoma, NOS\" \n",
    ") }}'''\n",
    "    \n",
    "n_pca_components = {{ IntField(\n",
    "    name='n_pca_components',\n",
    "    label='Number of PCA components to use for UMAP',\n",
    "    default=10,\n",
    "    min=0,\n",
    "    max=100,\n",
    "    section='CONFIG',\n",
    ") }}\n",
    "\n",
    "n_neighbors = {{ IntField(\n",
    "    name='n_neighbors',\n",
    "    label='Number of neighbors for each projected datapoint computed by UMAP',\n",
    "    description='Smaller values preserve local manifold structure in the dataset as opposed to overall global structure',\n",
    "    default=40,\n",
    "    min=2,\n",
    "    max=200,\n",
    "    section='CONFIG',\n",
    ") }}\n",
    "\n",
    "min_cluster_dist = {{ FloatField(\n",
    "    name='min_cluster_dist',\n",
    "    label='Minimum distance between UMAP-projected points',\n",
    "    description='How tightly packed points produced by dimensionality reduction with UMAP are permitted to be.',\n",
    "    default=0.3,\n",
    "    min=0.1,\n",
    "    max=1,\n",
    "    section='CONFIG',\n",
    ") }}\n",
    "\n",
    "top_n_genes = {{ IntField(\n",
    "    name='top_n_genes',\n",
    "    label='Number of top most variable genes to analyze',\n",
    "    description=\"The number of top most variable genes to use for analysis\",\n",
    "    default=2500,\n",
    "    min=100,\n",
    "    max=19000,\n",
    "    section='CONFIG',\n",
    ") }}\n",
    "\n",
    "max_clusters_calculation = {{ RadioField(\n",
    "    name=\"max_clusters_calculation\",\n",
    "    label=\"Method for calculating the maximum possible number of clusters\",\n",
    "    description = \"n is the sample size (number of cases)\",\n",
    "    choices = [\n",
    "         {\n",
    "            \"label\": '''\n",
    "                <math xmlns=\"http://www.w3.org/1998/Math/MathML\">\n",
    "                  <mstyle displaystyle=\"true\">\n",
    "                    <mfrac>\n",
    "                      <msqrt>\n",
    "                        <mrow>\n",
    "                          <mi>n</mi>\n",
    "                        </mrow>\n",
    "                      </msqrt>\n",
    "                      <mn>2</mn>\n",
    "                    </mfrac>\n",
    "                  </mstyle>\n",
    "                </math>\n",
    "            ''', \n",
    "            \"value\":\"root/2\"\n",
    "        },\n",
    "        {\n",
    "            \"label\": '''\n",
    "            <math xmlns=\"http://www.w3.org/1998/Math/MathML\">\n",
    "                <mstyle displaystyle=\"true\">\n",
    "                <msqrt>\n",
    "                  <mrow>\n",
    "                    <mi>n</mi>\n",
    "                  </mrow>\n",
    "                </msqrt>\n",
    "              </mstyle>\n",
    "            </math>\n",
    "            ''', \n",
    "            \"value\":\"root\"\n",
    "        }\n",
    "    ],\n",
    "    default = \"root/2\",\n",
    "    section = 'CONFIG'\n",
    ")}}\n",
    "\n",
    "use_weighted_silhouette_score = {{BoolField(\n",
    "    name=\"use_weighted_silhouette_score\",\n",
    "    label=\"Use a weighted silhouette score to determine ideal number of clusters\",\n",
    "    description = \"A weighted score is a weighted combination of the original score and the value of k, therefore encouraging more clusters.\",\n",
    "    default = True,\n",
    "    section = 'CONFIG'\n",
    ")}}\n",
    "\n",
    "use_second_deriv_method = {{BoolField(\n",
    "    name=\"use_second_deriv_method\",\n",
    "    label=\"Use the most concave down local maxima of the silhouette score to determine k\",\n",
    "    description = \"This method uses the unweighted silhouette score regardless of the choice above. If no maxima are found, we resort to k as determined by the chosen silhouette score method. \",\n",
    "    default = True,\n",
    "    section = 'CONFIG'\n",
    ")}}\n",
    "\n",
    "top_n_genes_enrichment = {{ IntField(\n",
    "    name='top_n_genes_enrichment',\n",
    "    label='Number of top genes to use for enrichment analysis',\n",
    "    description='The number of \\'top\\' genes to use for enrichment analysis',\n",
    "    default=250,\n",
    "    min=100,\n",
    "    max=1000,\n",
    "    section='CONFIG',\n",
    ") }}\n",
    "\n",
    "heatmap_top_n = {{ IntField(\n",
    "    name='cluster_top_n',\n",
    "    label='Number of up and down genes per cluster for heatmap visualizatiton',\n",
    "    default=100,\n",
    "    min=100,\n",
    "    max=1000,\n",
    "    section='CONFIG',\n",
    ") }}\n",
    "\n",
    "top_n_results = {{ IntField(\n",
    "    name='top_n_results',\n",
    "    label='Number of top enrichment results',\n",
    "    default=5,\n",
    "    min=1,\n",
    "    max=100,\n",
    "    section='CONFIG',\n",
    ") }}\n",
    "\n",
    "use_default_libraries = {{BoolField(\n",
    "    name=\"use_default_libraries\",\n",
    "    label=\"Use default Enrichr libraries?\",\n",
    "    description = \"The default libraries include:\\nKEGG_2019_Human,\\nKEGG_2019_Mouse,\\nGO_Biological_Process_2018,\\nMGI_Mammalian_Phenotype_Level_4_2019,\\nGWAS_Catalog_2019,\\nENCODE_TF_ChIP-seq_2015.\",\n",
    "    default = False,\n",
    "    section = 'ENRICHR_LIBS'\n",
    ")}}\n",
    "\n",
    "\n",
    "transcription_libraries = {{MultiChoiceField(name='transcription_libraries', \n",
    "                                            label='Transcription',\n",
    "                                            default=[], \n",
    "                                            section = 'ENRICHR_LIBS',\n",
    "                                            choices=[\n",
    "                                                'ARCHS4_TFs_Coexp',\n",
    "                                                'ChEA_2016',\n",
    "                                                'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "                                                'ENCODE_Histone_Modifications_2015',\n",
    "                                                'ENCODE_TF_ChIP-seq_2015',\n",
    "                                                'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "                                                'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "                                                'Genome_Browser_PWMs',\n",
    "                                                'lncHUB_lncRNA_Co-Expression',\n",
    "                                                'miRTarBase_2017',\n",
    "                                                'TargetScan_microRNA_2017',\n",
    "                                                'TF-LOF_Expression_from_GEO',\n",
    "                                                'TF_Perturbations_Followed_by_Expression',\n",
    "                                                'Transcription_Factor_PPIs',\n",
    "                                                'TRANSFAC_and_JASPAR_PWMs',\n",
    "                                                'TRRUST_Transcription_Factors_2019',\n",
    "                                            ])  }}\n",
    "\n",
    "pathways_libraries = {{ MultiChoiceField(name='pathways_libraries', \n",
    "                                         label='Pathways', \n",
    "                                         default=[], \n",
    "                                         section='ENRICHR_LIBS',\n",
    "                                         choices=[\n",
    "                                                    'ARCHS4_Kinases_Coexp',\n",
    "                                                    'BioCarta_2016',\n",
    "                                                    'BioPlanet_2019',\n",
    "                                                    'BioPlex_2017',\n",
    "                                                    'CORUM',\n",
    "                                                    'Elsevier_Pathway_Collection',\n",
    "                                                    'HMS_LINCS_KinomeScan',\n",
    "                                                    'HumanCyc_2016',\n",
    "                                                    'huMAP',\n",
    "                                                    'KEA_2015',\n",
    "                                                    'KEGG_2019_Human',\n",
    "                                                    'KEGG_2019_Mouse',\n",
    "                                                    'Kinase_Perturbations_from_GEO_down',\n",
    "                                                    'Kinase_Perturbations_from_GEO_up',\n",
    "                                                    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "                                                    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "                                                    'NCI-Nature_2016',\n",
    "                                                    'NURSA_Human_Endogenous_Complexome',\n",
    "                                                    'Panther_2016',\n",
    "                                                    'Phosphatase_Substrates_from_DEPOD',\n",
    "                                                    'PPI_Hub_Proteins',\n",
    "                                                    'Reactome_2016',\n",
    "                                                    'SILAC_Phosphoproteomics',\n",
    "                                                    'SubCell_BarCode',\n",
    "                                                    'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "                                                    'WikiPathways_2019_Human',\n",
    "                                                    'WikiPathways_2019_Mouse']) }}\n",
    "\n",
    "ontologies_libraries = {{ MultiChoiceField(name='ontologies_libraries', \n",
    "                                           label='Ontologies', \n",
    "                                           default=[], \n",
    "                                           section = 'ENRICHR_LIBS',\n",
    "                                           choices=[\n",
    "                                                'GO_Biological_Process_2018',\n",
    "                                                'GO_Cellular_Component_2018',\n",
    "                                                'GO_Molecular_Function_2018',\n",
    "                                                'Human_Phenotype_Ontology',\n",
    "                                                'Jensen_COMPARTMENTS',\n",
    "                                                'Jensen_DISEASES',\n",
    "                                                'Jensen_TISSUES',\n",
    "                                                'MGI_Mammalian_Phenotype_Level_4_2019'\n",
    "                                           ])  }}\n",
    "    \n",
    "diseases_drugs_libraries = {{ MultiChoiceField(name='diseases_drugs_libraries', \n",
    "                                               label='Diseases/Drugs', \n",
    "                                               default=[], \n",
    "                                               section = 'ENRICHR_LIBS',\n",
    "                                               choices=[    \n",
    "                                                    'Achilles_fitness_decrease',\n",
    "                                                    'Achilles_fitness_increase',\n",
    "                                                    'ARCHS4_IDG_Coexp',\n",
    "                                                    'ClinVar_2019',\n",
    "                                                    'dbGaP',\n",
    "                                                    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "                                                    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "                                                    'DisGeNET',\n",
    "                                                    'DrugMatrix',\n",
    "                                                    'DSigDB',\n",
    "                                                    'GeneSigDB',\n",
    "                                                    'GWAS_Catalog_2019',\n",
    "                                                    'LINCS_L1000_Chem_Pert_down',\n",
    "                                                    'LINCS_L1000_Chem_Pert_up',\n",
    "                                                    'LINCS_L1000_Ligand_Perturbations_down',\n",
    "                                                    'LINCS_L1000_Ligand_Perturbations_up',\n",
    "                                                    'MSigDB_Computational',\n",
    "                                                    'MSigDB_Oncogenic_Signatures',\n",
    "                                                    'Old_CMAP_down',\n",
    "                                                    'Old_CMAP_up',\n",
    "                                                    'OMIM_Disease',\n",
    "                                                    'OMIM_Expanded',\n",
    "                                                    'PheWeb_2019',\n",
    "                                                    'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "                                                    'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "                                                    'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "                                                    'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "                                                    'UK_Biobank_GWAS_v1',\n",
    "                                                    'Virus_Perturbations_from_GEO_down',\n",
    "                                                    'Virus_Perturbations_from_GEO_up',\n",
    "                                                    'VirusMINT']) }}\n",
    "    \n",
    "cell_types_libraries = {{ MultiChoiceField(name='cell_types_libraries', \n",
    "                                           label='Cell Types', \n",
    "                                           default=[], \n",
    "                                           section = 'ENRICHR_LIBS',\n",
    "                                           choices=[        \n",
    "                                                'Allen_Brain_Atlas_down',\n",
    "                                                'Allen_Brain_Atlas_up',\n",
    "                                                'ARCHS4_Cell-lines',\n",
    "                                                'ARCHS4_Tissues',\n",
    "                                                'Cancer_Cell_Line_Encyclopedia',\n",
    "                                                'CCLE_Proteomics_2020',\n",
    "                                                'ESCAPE',\n",
    "                                                'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "                                                'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "                                                'Human_Gene_Atlas',\n",
    "                                                'Mouse_Gene_Atlas',\n",
    "                                                'NCI-60_Cancer_Cell_Lines',\n",
    "                                                'ProteomicsDB_2020',\n",
    "                                                'Tissue_Protein_Expression_from_Human_Proteome_Map']) }}\n",
    "    \n",
    "    \n",
    "miscellaneous_libraries = {{ MultiChoiceField(name='miscellaneous_libraries', \n",
    "                                              label='Miscellaneous', \n",
    "                                              default=[], \n",
    "                                              section = 'ENRICHR_LIBS',\n",
    "                                              choices=[            \n",
    "                                                    'Chromosome_Location_hg19',\n",
    "                                                    'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "                                                    'Enrichr_Libraries_Most_Popular_Genes',\n",
    "                                                    'Genes_Associated_with_NIH_Grants',\n",
    "                                                    'HMDB_Metabolites',\n",
    "                                                    'HomoloGene',\n",
    "                                                    'InterPro_Domains_2019',\n",
    "                                                    'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "                                                    'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "                                                    'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "                                                    'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "                                                    'Pfam_Domains_2019',\n",
    "                                                    'Pfam_InterPro_Domains',\n",
    "                                                    'Table_Mining_of_CRISPR_Studies'])  }}\n",
    "    \n",
    "    \n",
    "legacy_libraries = {{ MultiChoiceField(name='legacy_libraries', \n",
    "                                       label='Legacy', \n",
    "                                       default=[], \n",
    "                                       section = 'ENRICHR_LIBS',\n",
    "                                       choices=[                \n",
    "                                            'BioCarta_2013',\n",
    "                                            'BioCarta_2015',\n",
    "                                            'ChEA_2013',\n",
    "                                            'ChEA_2015',\n",
    "                                            'Chromosome_Location',\n",
    "                                            'Disease_Signatures_from_GEO_down_2014',\n",
    "                                            'Disease_Signatures_from_GEO_up_2014',\n",
    "                                            'Drug_Perturbations_from_GEO_2014',\n",
    "                                            'ENCODE_Histone_Modifications_2013',\n",
    "                                            'ENCODE_TF_ChIP-seq_2014',\n",
    "                                            'GO_Biological_Process_2013',\n",
    "                                            'GO_Biological_Process_2015',\n",
    "                                            'GO_Biological_Process_2017',\n",
    "                                            'GO_Biological_Process_2017b',\n",
    "                                            'GO_Cellular_Component_2013',\n",
    "                                            'GO_Cellular_Component_2015',\n",
    "                                            'GO_Cellular_Component_2017',\n",
    "                                            'GO_Cellular_Component_2017b',\n",
    "                                            'GO_Molecular_Function_2013',\n",
    "                                            'GO_Molecular_Function_2015',\n",
    "                                            'GO_Molecular_Function_2017',\n",
    "                                            'GO_Molecular_Function_2017b',\n",
    "                                            'HumanCyc_2015',\n",
    "                                            'KEA_2013',\n",
    "                                            'KEGG_2013',\n",
    "                                            'KEGG_2015',\n",
    "                                            'KEGG_2016',\n",
    "                                            'MGI_Mammalian_Phenotype_2013',\n",
    "                                            'MGI_Mammalian_Phenotype_2017',\n",
    "                                            'MGI_Mammalian_Phenotype_Level_3',\n",
    "                                            'MGI_Mammalian_Phenotype_Level_4',\n",
    "                                            'NCI-Nature_2015',\n",
    "                                            'Panther_2015',\n",
    "                                            'Reactome_2013',\n",
    "                                            'Reactome_2015',\n",
    "                                            'TargetScan_microRNA',\n",
    "                                            'Tissue_Protein_Expression_from_ProteomicsDB',\n",
    "                                            'WikiPathways_2013',\n",
    "                                            'WikiPathways_2015',\n",
    "                                            'WikiPathways_2016']) }}\n",
    "\n",
    "crowd_libraries = {{ MultiChoiceField(name='crowd_libraries', \n",
    "                                      label='Crowd', \n",
    "                                      default=[],\n",
    "                                      section = 'ENRICHR_LIBS',\n",
    "                                      choices=[                \n",
    "                                            'Aging_Perturbations_from_GEO_down',\n",
    "                                            'Aging_Perturbations_from_GEO_up',\n",
    "                                            'Disease_Perturbations_from_GEO_down',\n",
    "                                            'Disease_Perturbations_from_GEO_up',\n",
    "                                            'Drug_Perturbations_from_GEO_down',\n",
    "                                            'Drug_Perturbations_from_GEO_up',\n",
    "                                            'Gene_Perturbations_from_GEO_down',\n",
    "                                            'Gene_Perturbations_from_GEO_up',\n",
    "                                            'Ligand_Perturbations_from_GEO_down',\n",
    "                                            'Ligand_Perturbations_from_GEO_up',\n",
    "                                            'MCF7_Perturbations_from_GEO_down',\n",
    "                                            'MCF7_Perturbations_from_GEO_up',\n",
    "                                            'Microbe_Perturbations_from_GEO_down',\n",
    "                                            'Microbe_Perturbations_from_GEO_up',\n",
    "                                            'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "                                            'SysMyo_Muscle_Gene_Sets']) }}\n",
    "\n",
    "\n",
    "\n",
    "top_n_drugs = {{IntField(\n",
    "    name='top_n_drugs',\n",
    "    label='Number of top drug treatments to suggest for each cluster',\n",
    "    default=5,\n",
    "    min=0,\n",
    "    max=50,\n",
    "    section='L1000FWD',\n",
    ") }}\n",
    "\n",
    "default_libraries = OrderedDict([\n",
    "    ('Diseases/Drugs', ['GWAS_Catalog_2019']), \n",
    "    ('Ontologies', ['GO_Biological_Process_2018','MGI_Mammalian_Phenotype_Level_4_2019']),\n",
    "    ('Pathways', ['KEGG_2019_Human','KEGG_2019_Mouse']),\n",
    "    ('Transcription', ['ENCODE_TF_ChIP-seq_2015']),\n",
    "])\n",
    "\n",
    "time_to_death_col = {{ StringField(\n",
    "    name=\"time_to_death_col\",\n",
    "    label=\"`Time to death` column name\",\n",
    "    section = 'SURVIVAL',\n",
    "    default=\"\",\n",
    ")}}\n",
    "\n",
    "time_to_last_followup_col = {{ StringField(\n",
    "    name=\"time_to_last_followup_col\",\n",
    "    label=\"`Time to last follow-up` column name\",\n",
    "    section = 'SURVIVAL',\n",
    "    default=\"\",\n",
    ")}}\n",
    "\n",
    "vital_status_col = {{ StringField(\n",
    "    name=\"vital_status_col\",\n",
    "    label=\"`Vital status` column name\",\n",
    "    section = 'SURVIVAL',\n",
    "    default=\"\",\n",
    ")}}\n",
    "\n",
    "alive_val = {{ StringField(\n",
    "    name=\"alive_val\",\n",
    "    label=\"Value in vital status column corresponding  to 'alive'\",\n",
    "    section = 'SURVIVAL',\n",
    "    default=\"\",\n",
    ")}}\n",
    "\n",
    "dead_val = {{ StringField(\n",
    "    name=\"dead_val\",\n",
    "    label=\"Value in vital status column corresponding  to 'dead'\",\n",
    "    section = 'SURVIVAL',\n",
    "    default=\"\",\n",
    ")}}\n",
    "\n",
    "survival_time_unit = {{ StringField(\n",
    "    name=\"survival_time_unit\",\n",
    "    label=\"Unit of time used for survival analysis columns\",\n",
    "    description=\"Used for labeling the survival plot x-axis.\",\n",
    "    section = 'SURVIVAL',\n",
    "    default=\"days\",\n",
    ")}}\n",
    "\n",
    "\n",
    "if use_default_libraries:\n",
    "    enrichr_libraries = default_libraries\n",
    "else: \n",
    "    enrichr_libraries = OrderedDict([\n",
    "        ('Diseases/Drugs', diseases_drugs_libraries), \n",
    "        ('Ontologies', ontologies_libraries),\n",
    "        ('Cell Type', cell_types_libraries),\n",
    "        ('Pathways', pathways_libraries),\n",
    "        ('Transcription', transcription_libraries),\n",
    "        ('Legacy', legacy_libraries),\n",
    "        ('Crowd', crowd_libraries)\n",
    "    ])\n",
    "\n",
    "    all_empty = True\n",
    "    for key,libs in enrichr_libraries.items():\n",
    "        if len(libs) > 0:\n",
    "            all_empty = False\n",
    "            break\n",
    "    if all_empty:\n",
    "        enrichr_libraries = default_libraries\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Import dataset <a class=\"anchor\" id=\"import\"></a>\n",
    "\n",
    "If using user-uploaded data, we access those files and set the necessary index names.\n",
    "\n",
    "Otherwise, we first download the data containing RNA-seq profiles for the selected cancer type from TCGA and the corresponding clinical data for those cases.\n",
    "\n",
    "The RNA-seq data that is provided has already been processed using the HTSeq Python package, which calculates the number of mapped reads to each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Notebook display util functions\n",
    "\n",
    "def download_button(content, label, filename):\n",
    "    # Add download button\n",
    "    outname = filename.split('.')[0]\n",
    "    display(HTML('<textarea id=\"textbox_{outname}\" style=\"display: none;\">{content}</textarea> <button style=\"margin:10px 0;\" id=\"create_{outname}\">{label}</button> <a download=\"{filename}\" id=\"downloadlink_{outname}\" style=\"display: none\">Download</a>'.format(**locals())))\n",
    "    display(HTML('<script type=\"text/javascript\">!function(){{var e=null,t=document.getElementById(\"create_{outname}\"),n=document.getElementById(\"textbox_{outname}\");t.addEventListener(\"click\",function(){{var t,l,c=document.getElementById(\"downloadlink_{outname}\");c.href=(t=n.value,l=new Blob([t],{{type:\"text/plain\"}}),null!==e&&window.URL.revokeObjectURL(e),e=window.URL.createObjectURL(l)),c.click()}},!1)}}();</script>'.format(**locals())))\n",
    "\n",
    "def make_clickable(link):\n",
    "    return f'<a target=\"_blank\" href=\"{link}\">{link}</a>'\n",
    "\n",
    "def figure_header(label,title):\n",
    "    display(HTML(f\"<div style='font-size:2rem; padding:1rem 0;'><b>{label}</b>: {title}</div>\"))\n",
    "    \n",
    "def figure_legend(label,title,content=\"\"):\n",
    "    display(HTML(f\"<div><b>{label}</b>: <i>{title}</i>. {content} </div>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_dataframe(file):\n",
    "    ''' Load a file by downloading it or reading it if already downloaded.\n",
    "    '''\n",
    "    ext = os.path.splitext(file)[1]\n",
    "    if ext in {'.tsv', '.txt'}:\n",
    "        df = pd.read_csv(file, sep='\\t', index_col=0)\n",
    "    elif ext == '.csv':\n",
    "        df = pd.read_csv(file, index_col=0)\n",
    "    else:\n",
    "        raise Exception('Unrecognized file format', ext)\n",
    "\n",
    "    # Fix any type coersion on identifiers\n",
    "    df.index = df.index.astype(str)\n",
    "    df.columns = df.columns.astype(str)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_data = False\n",
    "if (data_filename != \"\" and clinical_data_filename != \"\"):\n",
    "    user_data = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_index = \"symbol\"\n",
    "clinical_index = \"case_id\"\n",
    "\n",
    "if user_data:\n",
    "    print(f\"Loading user-uploaded data...\")\n",
    "    df_data = load_dataframe(data_filename).sort_index()\n",
    "    df_clinical = load_dataframe(clinical_data_filename).sort_index()\n",
    "\n",
    "    df_data.index.name = \"symbol\"\n",
    "    df_clinical.index.name = \"case_id\"\n",
    "else:\n",
    "    print(f\"Loading data for patients diagnosed with {cancer}...\")\n",
    "    # Import the RNA-seq and clinical DataFrames with only entries corresponding to\n",
    "    # the selected cancer type\n",
    "    aws_url = f'https://appyters.maayanlab.cloud/storage/Patient_Cohorts_RNASeq_Viewer/{cancer.replace(\" \", \"%20\")}'\n",
    "\n",
    "    df_data = pd.read_csv(f'{aws_url}_data.csv')\n",
    "    df_data = df_data.set_index(\"symbol\") \n",
    "\n",
    "    df_clinical = pd.read_csv(f'{aws_url}_clinical_data.csv')\n",
    "    df_clinical = df_clinical.set_index(\"case_id\")\n",
    "    df_clinical = df_clinical.replace(\"not reported\", np.nan)    \n",
    "\n",
    "print(\"Data loaded\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_header(\"Table 1\", \"RNA-Seq data\")\n",
    "display(df_data.head())\n",
    "figure_legend(\"Table 1\", \"RNA-Seq data\", \"The RNA-Seq data contains a row per gene and a column per case. The column indices are case_ids (from TCGA or the user-uploaded dataset) while the row indices are Entrez gene symbols.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a list of the clinical features with multiple unique values\n",
    "def has_unique_values(feature):\n",
    "    unique = df_clinical[feature].dropna().unique()\n",
    "    if len(unique) >= 1: return True\n",
    "    return False\n",
    "\n",
    "features = df_clinical.columns.values\n",
    "if not user_data:\n",
    "    features = list(filter(\n",
    "        lambda x: not x in [\"primary_diagnosis\",\"submitter_id\", \"updated_datetime\",\"created_datetime\",\"diagnosis_id\", \"demographic_id\"] \n",
    "        and has_unique_values(x), features))\n",
    "    \n",
    "    df_clinical = df_clinical[features]\n",
    "\n",
    "    if \"icd_10_code\" in df_clinical.columns:\n",
    "        \n",
    "        # retrieve the file mapping codes to descriptions \n",
    "        aws_url = 'https://tcga-enrichr-viewer.s3.amazonaws.com/icd10cm_order_2020.txt'\n",
    "        \n",
    "        codes = pd.read_csv(aws_url,sep=\"\\t\",header=None)\n",
    "        indeces = [ val.split(\" \")[1] for val in codes[0] ]\n",
    "\n",
    "        # The description of each row starts at character 77 in this file\n",
    "        def get_description(val):\n",
    "            return val[77:]\n",
    "\n",
    "        descriptions = [get_description(val) for val in codes[0]]\n",
    "        df_icd_codes = pd.DataFrame(descriptions, index=indeces)\n",
    "        \n",
    "        # Add descriptions to the clinical df\n",
    "        codes = df_clinical[\"icd_10_code\"] \n",
    "        \n",
    "        def get_desc_by_code(code):\n",
    "            code_ind = code.replace(\".\", \"\")\n",
    "            if code_ind in df_icd_codes.index.values:\n",
    "                return df_icd_codes.loc[code_ind,0]\n",
    "            return code # map the code to itself if missing in the table\n",
    "            \n",
    "        \n",
    "        df_clinical[\"icd_10_code_desc\"] = [ get_desc_by_code(x) for x in codes]\n",
    "        features = [\"icd_10_code_desc\", *features]\n",
    "\n",
    "    # drop rows for case_ids already represented (i.e. don't have more than one clinical entry per patient)\n",
    "    df_clinical = df_clinical.loc[~df_clinical.index.duplicated(keep='first')]\n",
    "else:\n",
    "    features = list(filter(\n",
    "        lambda x: has_unique_values(x), features))\n",
    "    df_clinical = df_clinical[features]\n",
    "\n",
    "\n",
    "print(f\"{df_data.shape[1]} cases, {len(features)} clinical features\\n\\n\")\n",
    "\n",
    "figure_header(\"Table 2\",\"Clinical metadata\")\n",
    "display(df_clinical)\n",
    "figure_legend(\"Table 2\",\"Clinical metadata\", \"The column indices of the clinical dataset are clinical features and the row indices are case_ids corresponding to the column indices of the RNA-Seq dataset.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Clinical features loaded:\\n\")\n",
    "print_features = [ print(f) for f in features  ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Process data and compute clusters <a class=\"anchor\" id=\"processing\"></a>\n",
    "\n",
    "Next, we'll need to normalize the RNA-seq data, reduce its dimensionality using PCA and UMAP, and compute clusters of RNA-seq profiles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data size stats\n",
    "\n",
    "df_library_size = pd.DataFrame(\n",
    "    {\n",
    "        'n_expressed_genes': df_data[df_data > 0].count(),\n",
    "        'log_n_reads': np.log2(df_data.sum() + 1),\n",
    "        'n_reads': df_data.sum(),\n",
    "    }).sort_values('n_reads', ascending=False)\n",
    "\n",
    "df_library_size.index.name = \"case_id\"\n",
    "\n",
    "figure_header(\"Table 3\",\"Library size\")\n",
    "display(df_library_size.head()) \n",
    "figure_legend(\"Table 3\",\"Library size\", \"By default, the first five entries are shown. A gene read is counted toward n_reads for a single patient if its value is greater than 0.\")\n",
    "\n",
    "figure_header(\"Figure 1\",\"Library size distribution\")\n",
    "sns.distplot(df_library_size[\"n_reads\"]); plt.show()\n",
    "figure_legend(\"Figure 1\",\"Library size distribution\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We normalize two versions of the dataset: one with just the `top_n_genes` most variable genes and one with all genes. This is because we need the former to compute clusters in the first place (after dimensionality reduction), and the latter to compute the characteristic direction of each gene in the dataset for each computed cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalization\n",
    "\n",
    "# maintain the unfiltered dataset to compute characteristic directions later\n",
    "df_data_norm_all_genes = df_data.copy()\n",
    "\n",
    "# compute log normalization of matrix\n",
    "df_data_norm = log2_normalize(df_data, offset=1)\n",
    "df_data_norm_all_genes = log2_normalize(df_data_norm_all_genes, offset=1)\n",
    "\n",
    "#quantile normalize\n",
    "df_data_norm = qnorm.quantile_normalize(df_data_norm, axis=1)\n",
    "df_data_norm_all_genes = qnorm.quantile_normalize(df_data_norm_all_genes, axis=1)\n",
    "\n",
    "# take top_n_genes most variable rows\n",
    "df_data_norm = filter_by_var(df_data,top_n = top_n_genes)\n",
    "\n",
    "# convert to zscores\n",
    "df_data_norm = zscore_normalize(df_data_norm.T).T\n",
    "df_data_norm_all_genes = zscore_normalize(df_data_norm_all_genes.T).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Normalized dataset with all genes has shape \",df_data_norm_all_genes.shape)\n",
    "print(f\"Normalized dataset with {top_n_genes} most variable genes has shape \", df_data_norm.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_header(\"Table 4\",\"Normalized RNA-Seq data\")\n",
    "display(df_data_norm)\n",
    "figure_legend(\"Table 4\",\"Normalized RNA-Seq data\", \"Counts are filtered for the <i>top_n_genes</i> most variable genes. A log transform and normalization is performed on the resultitng dataset, which is then converted to z-scores.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the first gene distribution\n",
    "gene = df_data_norm.index.values[0]\n",
    "figure_header(\"Figure 2\",f\"Sample gene expression distibution for {gene}\")\n",
    "sns.distplot(df_data_norm.iloc[0, :]); plt.show()\n",
    "figure_legend(\"Figure 2\",f\"Sample gene expression distibution for {gene}\", f\"In this dataset, {gene} is the most variably expressed across all samples.\")\n",
    "\n",
    "# plot the last gene distribution\n",
    "gene = df_data_norm.index.values[-1]\n",
    "figure_header(\"Figure 3\",f\"Sample gene expression distibution for {gene}\")\n",
    "sns.distplot(df_data_norm.iloc[-1, :]); plt.show()\n",
    "figure_legend(\"Figure 3\",f\"Sample gene expression distibution for {gene}\", f\"In this dataset, {gene} is the least variably expressed across all samples among the filtered (most variably expressed) genes.\")\n",
    "\n",
    "\n",
    "figure_header(\"Figure 4\",\"Sample individual RNA-Seq profile distribution\")\n",
    "# plot a single RNA seq profile distribution\n",
    "sns.distplot(df_data_norm.iloc[:, 0]); plt.show()\n",
    "figure_legend(\"Figure 4\",\"Sample individual RNA-Seq profile distribution\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's visualize expression of the most variable genes across the dataset with hierachical clustering and a heatmap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalization\n",
    "# take top 800 most variable rows\n",
    "\n",
    "# the \"i\" in the variable here stands for \"initial\", since we'll do a different method \n",
    "# for selecting genes for a heatmap later\n",
    "df_data_norm_heatmap_i = filter_by_var(df_data,top_n=800)\n",
    "\n",
    "# compute log normalization of matrix\n",
    "df_data_norm_heatmap_i = log2_normalize(df_data_norm_heatmap_i)\n",
    "\n",
    "# convert to zscores\n",
    "df_data_norm_heatmap_i = zscore_normalize(df_data_norm_heatmap_i) \n",
    "\n",
    "# Plot heatmap\n",
    "figure_header(\"Figure 5\",\"Heatmap of normalized expression for top 800 most variable genes\")\n",
    "sns.clustermap(df_data_norm_heatmap_i,xticklabels=False); plt.show()\n",
    "figure_legend(\"Figure 5\",\"Heatmap of normalized expression for top 800 most variable genes\", \"This initial heatmap considers only the top most variable genes across the entire dataset, rather than the most differentially expressed genes among potential clusters. As such, there may seem to be a lot of noise preventing us from seeing clear clusters. We will ammend this in a later heatmap.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use PCA to initially reduce the dimensionality of the dataset before clustering while still maintaining most of the variability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PCA\n",
    "data_norm_pca = PCA(\n",
    "  random_state=42,\n",
    ")\n",
    "\n",
    "data_norm_pca.fit(df_data_norm.values.T)\n",
    "\n",
    "df_data_norm_pca = pd.DataFrame(\n",
    "    data_norm_pca.transform(df_data_norm.values.T),\n",
    "    index=df_data_norm.T.index\n",
    ")\n",
    "\n",
    "df_data_norm_pca.columns = [\n",
    "    f'PCA-{c}' # ({r:.3f})'\n",
    "    for c, r in zip(df_data_norm_pca.columns, data_norm_pca.explained_variance_ratio_)\n",
    "]\n",
    "\n",
    "df_data_norm_pca.index.name = \"case_id\"\n",
    "\n",
    "figure_header(\"Table 5\",\"Principle components of RNA-Seq data\")\n",
    "display(df_data_norm_pca.head())\n",
    "figure_legend(\"Table 5\",\"Principle components of RNA-Seq data\", \"The top principle components are the projections of each datapoint onto the axes along which there is the most variation in the dataset.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the two plots below (PCA and UMAP projections), datapoints are color-coded by the selected feature. The size of each point represents the number of reads for that sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pca_data = merge(\n",
    "        df_data_norm_pca[[\"PCA-0\", \"PCA-1\"]],\n",
    "        df_library_size,\n",
    "        df_clinical\n",
    "      )\n",
    "\n",
    "axes = {\n",
    "    \"x\": [min(df_data_norm_pca[\"PCA-0\"])*1.1, max(df_data_norm_pca[\"PCA-0\"])*1.1],\n",
    "    \"y\": [min(df_data_norm_pca[\"PCA-1\"])*1.1, max(df_data_norm_pca[\"PCA-1\"])*1.1]\n",
    "}\n",
    "\n",
    "figure_header(\"Figure 6\",\"Projection of RNA-Seq data onto first two principle directions\")\n",
    "scatter_plot(pca_data,\"PCA-0\", \"PCA-1\",axes,features)\n",
    "figure_legend(\"Figure 6\",\"Projection of RNA-Seq data onto first two principle directions\", \"The size of datapoints corresponds to the number of gene reads. Points can be color-coded by any of the clinical features using the dropdown menu.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We further reduce the dimensionality of the dataset using the UMAP (Uniform Manifold Approximation and Projection) technique on the data projected onto the first `n_pca_components` PCA components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UMAP\n",
    "data_norm_umap = UMAP(\n",
    "  random_state=42,\n",
    "  n_components=2,\n",
    "  n_neighbors=n_neighbors,\n",
    "  metric='cosine',\n",
    "  min_dist=min_cluster_dist,\n",
    ")\n",
    "\n",
    "# use top n_pca_components components of PCA\n",
    "n_pca_components = min(n_pca_components,df_data_norm_pca.shape[1])\n",
    "data_norm_umap.fit(df_data_norm_pca.iloc[:, :n_pca_components].values)\n",
    "\n",
    "df_data_norm_umap = pd.DataFrame(\n",
    "  data_norm_umap.transform(df_data_norm_pca.iloc[:, :n_pca_components].values),\n",
    "  columns=['UMAP-0', 'UMAP-1'],\n",
    "  index=df_data_norm_pca.index,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# project data onto its first 2 UMAP components for visualization\n",
    "umap_data = merge(\n",
    "        df_data_norm_umap[[\"UMAP-0\", \"UMAP-1\"]],\n",
    "        df_library_size,\n",
    "        df_clinical\n",
    "      )\n",
    "\n",
    "axes = {\n",
    "    \"x\": [min(df_data_norm_umap[\"UMAP-0\"])*1.1, max(df_data_norm_umap[\"UMAP-0\"])*1.1],\n",
    "    \"y\": [min(df_data_norm_umap[\"UMAP-1\"])*1.1, max(df_data_norm_umap[\"UMAP-1\"])*1.1]\n",
    "}\n",
    "\n",
    "figure_header(\"Figure 7\",\"First two UMAP components of RNA-Seq data\")\n",
    "scatter_plot(umap_data,\"UMAP-0\", \"UMAP-1\",axes,features)\n",
    "figure_legend(\"Figure 7\",\"First two UMAP components of RNA-Seq data\", \"The size of datapoints corresponds to the number of gene reads. Points can be color-coded by any of the clinical features using the dropdown menu.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We continue using the first 2 UMAP components.\n",
    "\n",
    "To compute clusters, we use the Kmeans method, which requires us to define a total number of clusters. We test a range for the number of total clusters. \n",
    "\n",
    "For each number of clusters, we compute silhouette scores, which are a measure of how similar an entry is to its own cluster compared to other clusters. We want to maximize similarity within a cluster and differrences between clusters, so the ideal number of clusters is that which produces the highest silhouette score.\n",
    "\n",
    "Here, we modify the selection of $k$ slightly to preserve nuances in the data by preferring more clusters over fewer. We also calcuulated a *modified* silhouette score, which takes into account the number of clusters, $k$; each original score is bumped by a factor linear with respect to $k$. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_silhouette_analysis(n_clusters, data, cluster_labels, avg_score, sample_values, centers):\n",
    "    # Create a subplot with 1 row and 2 columns\n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "    fig.set_size_inches(18, 7)\n",
    "\n",
    "    # The 1st subplot is the silhouette plot\n",
    "    # The silhouette coefficient can range from -1, 1 but in this example all\n",
    "    # lie within [-0.1, 1]\n",
    "    ax1.set_xlim([-0.1, 1])\n",
    "    # The (n_clusters+1)*10 is for inserting blank space between silhouette\n",
    "    # plots of individual clusters, to demarcate them clearly.\n",
    "    ax1.set_ylim([0, len(X) + (n_clusters + 1) * 10])\n",
    "\n",
    "    y_lower = 10\n",
    "    for i in range(n_clusters):\n",
    "        # Aggregate the silhouette scores for samples belonging to\n",
    "        # cluster i, and sort them\n",
    "        ith_cluster_silhouette_values = \\\n",
    "            sample_values[cluster_labels == i]\n",
    "\n",
    "        ith_cluster_silhouette_values.sort()\n",
    "\n",
    "        size_cluster_i = ith_cluster_silhouette_values.shape[0]\n",
    "        y_upper = y_lower + size_cluster_i\n",
    "\n",
    "        color = cm.nipy_spectral(float(i) / n_clusters)\n",
    "        ax1.fill_betweenx(np.arange(y_lower, y_upper),\n",
    "                          0, ith_cluster_silhouette_values,\n",
    "                          facecolor=color, edgecolor=color, alpha=0.7)\n",
    "\n",
    "        # Label the silhouette plots with their cluster numbers at the middle\n",
    "        ax1.text(-0.05, y_lower + 0.5 * size_cluster_i, str(i))\n",
    "\n",
    "        # Compute the new y_lower for next plot\n",
    "        y_lower = y_upper + 10  # 10 for the 0 samples\n",
    "\n",
    "    ax1.set_title(\"The silhouette plot for the various clusters.\")\n",
    "    ax1.set_xlabel(\"The silhouette coefficient values\")\n",
    "    ax1.set_ylabel(\"Cluster label\")\n",
    "\n",
    "    # The vertical line for average silhouette score of all the values\n",
    "    ax1.axvline(x=silhouette_avg, color=\"red\", linestyle=\"--\")\n",
    "\n",
    "    ax1.set_yticks([])  # Clear the yaxis labels / ticks\n",
    "    ax1.set_xticks([-0.1, 0, 0.2, 0.4, 0.6, 0.8, 1])\n",
    "\n",
    "    # 2nd Plot showing the actual clusters formed\n",
    "    colors = cm.nipy_spectral(cluster_labels.astype(float) / n_clusters)\n",
    "    ax2.scatter(X[:, 0], X[:, 1], marker='.', s=30, lw=0, alpha=0.7,\n",
    "                c=colors, edgecolor='k')\n",
    "\n",
    "    # Draw white circles at cluster centers\n",
    "    ax2.scatter(centers[:, 0], centers[:, 1], marker='o',\n",
    "                c=\"white\", alpha=1, s=200, edgecolor='k')\n",
    "\n",
    "    for i, c in enumerate(centers):\n",
    "        ax2.scatter(c[0], c[1], marker='$%d$' % i, alpha=1,\n",
    "                    s=50, edgecolor='k')\n",
    "\n",
    "    ax2.set_title(\"The visualization of the clustered data.\")\n",
    "    ax2.set_xlabel(\"Feature space for the 1st feature\")\n",
    "    ax2.set_ylabel(\"Feature space for the 2nd feature\")\n",
    "\n",
    "    plt.suptitle((\"Silhouette analysis for KMeans clustering on sample data \"\n",
    "                  \"with n_clusters = %d\" % n_clusters),\n",
    "                 fontsize=14, fontweight='bold')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "silhouette_scores = []\n",
    "\n",
    "# Set max clusters as a function of the sample size and the user-selected option\n",
    "max_clusters = math.ceil(df_data_norm_umap.shape[0]**0.5)\n",
    "if (max_clusters_calculation == \"root/2\"):\n",
    "    max_clusters = int(math.ceil(max_clusters/2))\n",
    "\n",
    "\n",
    "def calc_weighted_score(silhouette_score, k, max_k):\n",
    "    return silhouette_avg*0.7 + k/max_k*0.3\n",
    "\n",
    "cluster_range = range(2, max_clusters)\n",
    "for n in cluster_range:\n",
    "    X = df_data_norm_umap.values\n",
    "    clusterer = KMeans(n_clusters=n, random_state=42).fit(X)\n",
    "    y_pred = clusterer.predict(X)\n",
    "    \n",
    "    # The silhouette_score gives the average value for all the samples.\n",
    "    # This gives a perspective into the density and separation of the formed\n",
    "    # clusters\n",
    "    silhouette_avg = silhouette_score(X, y_pred, metric='cosine')\n",
    "    \n",
    "    # Compute a weighted score that rewards higher numbers of clusters\n",
    "    weighted_score = calc_weighted_score(silhouette_avg, n, max_clusters)\n",
    "    \n",
    "    # Compute the silhouette scores for each sample\n",
    "    sample_silhouette_values = silhouette_samples(X, y_pred)\n",
    "\n",
    "    silhouette_scores.append({\n",
    "        \"N Clusters\": n,\n",
    "        \"Silhouette Score\": silhouette_avg,\n",
    "        \"Weighted Score\": weighted_score\n",
    "    })\n",
    "    \n",
    "    # Labeling the clusters\n",
    "    centers = clusterer.cluster_centers_\n",
    "    \n",
    "    # Plot the analysis for this number of clusters\n",
    "    #plot_silhouette_analysis(n, X, y_pred, silhouette_avg, sample_silhouette_values, centers)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If using the \"second derivative method\" for ideal $k$ calculation, we also compute estimates of the second derivatives at each posisble point to see if there are local maxima that may present better values of $k$ than the silhouette scores suggest. If such maxima exist, we take the value of $k$ that presents the most negative approximated second derivative (meaning the plot of scores is most concave *down* at this value of $k$). \n",
    "\n",
    "To calculate the second derivatives, we use the following approximation:\n",
    "\n",
    "$$ f''(x) \\approx \\frac{f(x+h) - 2f(x) + f(x-h)}{h^2}$$\n",
    "\n",
    "where, in our case, $h = 1$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "points = {}\n",
    "threshold = 0.3\n",
    "score_type = \"Silhouette Score\"\n",
    "if use_weighted_silhouette_score:\n",
    "    score_type = \"Weighted Score\"\n",
    "    \n",
    "for s in silhouette_scores:\n",
    "    points[s[\"N Clusters\"]] = s[score_type]\n",
    "\n",
    "# No local maxima - resort to previous method of best score calculation\n",
    "k = None\n",
    "\n",
    "if (use_second_deriv_method):\n",
    "    local_maxima = []\n",
    "    # Collect all local maxima\n",
    "    for n in range(cluster_range[1], cluster_range[-2]):\n",
    "        if points[n] > threshold and points[n-1] < points[n] and points[n+1] < points[n]:\n",
    "            local_maxima.append(n)\n",
    "    \n",
    "    if len(local_maxima)>0:\n",
    "        print(\"K values for local maxima and positive silhouette score: \", local_maxima)\n",
    "    else:\n",
    "        print(\"No local maxima with positive silhouette scores found\")\n",
    "\n",
    "    # Otherwise, calculate which of the local max is most concave down (i.e. largest second derivative)\n",
    "    if len(local_maxima) > 0:\n",
    "        running_max = 0\n",
    "        for maxima in local_maxima:\n",
    "            # approximate the second derivative, with step size = 1\n",
    "            second_deriv = points[maxima + 1] -  2*points[maxima] + points[maxima-1]\n",
    "            if second_deriv < running_max:\n",
    "                running_max = second_deriv\n",
    "                k = maxima"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "silhouette_scores = pd.DataFrame(silhouette_scores)\n",
    "figure_header(\"Table 6\", \"Silhouette scores by number of clusters\")\n",
    "display(silhouette_scores.sort_values([\"Silhouette Score\", \"Weighted Score\"], ascending=False).reset_index().head())\n",
    "figure_legend(\"Table 6\", \"Silhouette scores by number of clusters\", \"Values are sorted by the highest Silhouette Score.\")\n",
    "\n",
    "\n",
    "best_score = silhouette_scores.sort_values('Silhouette Score').iloc[-1].to_dict()\n",
    "best_weighted = silhouette_scores.sort_values('Weighted Score').iloc[-1].to_dict()\n",
    "\n",
    "best = {\"Silhouette Score\": best_score, \"Weighted Score\": best_weighted}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not k:\n",
    "    k = int(best[\"Silhouette Score\"][\"N Clusters\"])\n",
    "    if (use_weighted_silhouette_score):\n",
    "        k = int(best[\"Weighted Score\"][\"N Clusters\"])\n",
    "    \n",
    "print(f\"Ideal k: {k} clusters\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot both the uneqeighted and weighted scores as a function of # of clusters\n",
    "colors = {\"Silhouette Score\": \"#7C88FB\", \"Weighted Score\": \"#00CC96\"}\n",
    "\n",
    "for score_type in ['Silhouette Score', 'Weighted Score']:\n",
    "    plt.plot(silhouette_scores['N Clusters'], silhouette_scores[score_type], label=score_type,color=colors[score_type])\n",
    "    plt.scatter([best[score_type]['N Clusters']], [best[score_type][score_type]], label=f\"Best {score_type}: {int(best[score_type]['N Clusters'])} clusters\",color=colors[score_type])\n",
    "plt.axvline(k, label = f\"Ideal k: {k} clusters\", color =\"#EF553B\", alpha=0.8,dashes=(3,3))\n",
    "plt.legend()\n",
    "plt.ylabel('Score')\n",
    "plt.xlabel('Number of Clusters')\n",
    "figure_header(\"Figure 8\", \"Cluster size selection\")\n",
    "plt.show()\n",
    "figure_legend(\"Figure 8\", \"Cluster size selection\", \"The dotted line indicates the value of the 'ideal' <i>k</i> as chosen by the selected scoring method. This value will be used in subsequent clustering.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute the Kmeans dataframe using the ideal number of clusters\n",
    "km = KMeans(n_clusters=k, random_state=42)\n",
    "km_clusters = km.fit_predict(df_data_norm_umap.values)\n",
    "\n",
    "df_data_norm_km = pd.DataFrame({\n",
    "'Cluster': [\n",
    "    str(c)\n",
    "    for c in km_clusters\n",
    "]}, index=df_data_norm_umap.index)\n",
    "\n",
    "\n",
    "print(f'Computed {len(df_data_norm_km[\"Cluster\"].unique())} clusters')\n",
    "df_data_norm_km.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map each cluster to a color for later plots\n",
    "clusters = df_data_norm_km[\"Cluster\"].unique()\n",
    "#plotly_colors = px.colors.qualitative.Plotly\n",
    "plotly_colors = ['#636EFA', '#EF553B', '#00CC96', '#AB63FA', '#FFA15A', '#19D3F3', '#FF6692', '#B6E880', '#FF97FF', '#FECB52']\n",
    "cluster_colors = {}\n",
    "i = 0\n",
    "for c in clusters:\n",
    "    cluster_colors[c] = plotly_colors[i % len(plotly_colors)]\n",
    "    i += 1\n",
    "\n",
    "def cluster_heading(cluster):\n",
    "    display(HTML(f'''\n",
    "    <center>\n",
    "    <div style='background-color:{cluster_colors[cluster] + '98'};\n",
    "        width:100%;height:3rem;display:flex;align-items:center;\n",
    "        justify-content:center;color:white;font-size:2rem'>\n",
    "        <center>Cluster {cluster}</center>\n",
    "    </div>\n",
    "    </center>'''))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we compute differential expression for each cluster. We use the <a href=\"http://www.maayanlab.net/CD/\">Characteristic Direction method</a> for identifying differentially expressed genes among the different clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get differential expression for each cluster, using the dataset containing all genes\n",
    "diff_expr = {}\n",
    "for cluster, samples in df_data_norm_km.groupby('Cluster'):\n",
    "    diff_expr[f\"Cluster {cluster} CD\"] = characteristic_direction(\n",
    "        # expression outside of this cluster\n",
    "        df_data_norm_all_genes.loc[:, df_data_norm_all_genes.columns.difference(samples.index)],\n",
    "        # expression in this cluster\n",
    "        df_data_norm_all_genes.loc[:, samples.index],\n",
    "      )['CD-coefficient']\n",
    "\n",
    "df_diff_expr = pd.DataFrame(diff_expr)\n",
    "df_diff_expr = df_diff_expr.sort_values(by='Cluster 0 CD',ascending=True)\n",
    "df_diff_expr['Symbol'] = df_diff_expr.index.values\n",
    "\n",
    "figure_header(\"Table 7\", \"Differential expression of genes by cluster\")\n",
    "display(df_diff_expr.head())\n",
    "figure_legend(\"Table 7\", \"Differential expression of genes by cluster\", \"By default, the top 5 most differentially expressed genes are shown, along with the corresponding characteristic directions for each cluster.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We fit a logistic regression on each clinical feature to see which most accurately predict the cluster each data point falls into. We also plot ROC curves for features that resulted in the top five highest AUC scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LR\n",
    "aucs = {}\n",
    "num_rocs_plotted = 0\n",
    "rocs = {}\n",
    "\n",
    "for cluster, samples in df_data_norm_km.groupby('Cluster'): \n",
    "    aucs[cluster] = {}\n",
    "    rocs[cluster] = []\n",
    "\n",
    "    for feature in features:\n",
    "        lr = LogisticRegression()\n",
    "        X = df_clinical.copy()\n",
    "        X = X[feature]\n",
    "        X = pd.merge(X, df_data_norm_km, left_index = True, right_index = True)\n",
    "\n",
    "        # drop NAs, and move on if dataset is empty\n",
    "        X.replace(\"not reported\", None)\n",
    "        X = X.dropna()\n",
    "        if (X.shape[0] == 0): continue\n",
    "\n",
    "        cluster_data = X[\"Cluster\"]\n",
    "        X = X.drop(columns= [\"Cluster\"])\n",
    "\n",
    "        # one-hot encode non numerical data\n",
    "        if (not isinstance(X[feature][0], (int, float, complex))):\n",
    "            X = pd.get_dummies(X[feature], prefix=feature)\n",
    "\n",
    "        y_true = (cluster_data == cluster)\n",
    "        \n",
    "        if (len(y_true.unique()) < 2): # if we only have one class in the dataset\n",
    "            print(f\"Not enough data to classify cluster {cluster} based on feature {feature}\")\n",
    "            aucs[cluster][feature] = np.nan\n",
    "            continue \n",
    "                  \n",
    "        lr.fit(X, y_true)\n",
    "\n",
    "        y_score = lr.predict_proba(X)[:, 1]\n",
    "        auc_score = roc_auc_score(y_true, y_score)\n",
    "        aucs[cluster][feature] = auc_score\n",
    "        \n",
    "        # save the ROCs\n",
    "       \n",
    "        rocs[cluster].append({\"auc\":auc_score, \"lr\": lr, \"X\": X, \"y_true\":y_true, \"title\": f'Predictions of cluster {cluster} by feature {feature}'})\n",
    "        \n",
    "df_cluster_aucs = pd.DataFrame(aucs)\n",
    "df_cluster_aucs.index.name=\"Feature\"\n",
    "\n",
    "# sort features by avg AUC across all clusters\n",
    "df_cluster_aucs[\"avg\"] = [ np.mean(df_cluster_aucs.T[f]) for f in df_cluster_aucs.index.values ]\n",
    "df_cluster_aucs = df_cluster_aucs.sort_values(by = \"avg\", ascending=False)\n",
    "df_cluster_aucs = df_cluster_aucs.drop(columns = \"avg\")\n",
    "\n",
    "cols = [('Cluster', col) for col in df_cluster_aucs.columns ]\n",
    "df_cluster_aucs.columns = pd.MultiIndex.from_tuples(cols)\n",
    "\n",
    "figure_header(\"Table 8\", \"Average AUC scores for top-predicting clinical features, by cluster\")\n",
    "display(df_cluster_aucs.head(10))\n",
    "figure_legend(\"Table 8\", \"Average AUC scores for top-predicting clinical features, by cluster\", \"Scores for the top 10 clinical features, as determined by the average AUC score across all clusters, are shown. Higher AUC scores correspond to better classifiers for distinguishing whether or not a datapoint belongs to a certain cluster.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot top 2 (or however many exist) ROCs for each cluster\n",
    "matplotlib.rc('font', size=16)\n",
    "figure_header(\"Figure 9\", \"ROCs for top cluster-predicting clinical features\")\n",
    "\n",
    "for cluster, plots in rocs.items():\n",
    "    plots.sort(reverse=True, key=lambda x: x[\"auc\"])\n",
    "    cluster_heading(cluster)\n",
    "    \n",
    "    if len(plots) < 2:\n",
    "        best_rocs = plots\n",
    "    else:\n",
    "        best_rocs = plots[:2]\n",
    "\n",
    "    num_plots = len(best_rocs)\n",
    "    figure,axes = plt.subplots(int(math.ceil(num_plots / 2.)), 2, figsize=(15,(len(best_rocs)*3.5)))\n",
    "    \n",
    "    axes = axes.flatten()\n",
    "    for i in range(len(axes)):\n",
    "        if i >= len(best_rocs):\n",
    "            axes[i].remove()\n",
    "        else:\n",
    "            plot = best_rocs[i]\n",
    "            fig = metrics.plot_roc_curve(plot[\"lr\"], plot[\"X\"], plot[\"y_true\"], ax=axes[i])\n",
    "\n",
    "            axes[i].set_title('\\n'.join(wrap(plot[\"title\"], 40)))\n",
    "\n",
    "    figure.tight_layout(pad=2)\n",
    "    plt.show()\n",
    "    \n",
    "figure_legend(\"Figure 9\", \"ROCs for top cluster-predicting clinical features\")\n",
    "\n",
    "matplotlib.rcdefaults()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we find the top most up and downregulated genes for each cluster, both to select which data to display in the heatmap below and for Enrichment analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge data\n",
    "df_clustered_umap = pd.merge(left=df_data_norm_km, left_on=\"case_id\", right=df_data_norm_umap, right_on=\"case_id\")\n",
    "df_clustered_pca = pd.merge(left=df_data_norm_km, left_on=\"case_id\", right=df_data_norm_pca, right_on=\"case_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get top Genes for each cluster\n",
    "top_genes = {}\n",
    "all_top_genes = []\n",
    "heatmap_top_n = min(heatmap_top_n, top_n_genes_enrichment)\n",
    "for cluster in df_clustered_umap['Cluster'].unique():\n",
    "    cd_col = f'Cluster {cluster} CD'\n",
    "    if cd_col in df_diff_expr.columns:\n",
    "        # top up genes\n",
    "        up_genes = df_diff_expr.loc[df_diff_expr[cd_col].sort_values(ascending=False).iloc[:top_n_genes_enrichment].index, 'Symbol'].values\n",
    "        # top down genes\n",
    "        dn_genes = df_diff_expr.loc[df_diff_expr[cd_col].sort_values(ascending=True).iloc[:top_n_genes_enrichment].index, 'Symbol'].values\n",
    "    else:\n",
    "        raise Exception('Cant find col for cluster')\n",
    "    all_top_genes.append(up_genes[:heatmap_top_n])\n",
    "    all_top_genes.append(dn_genes[:heatmap_top_n])\n",
    "    # save results\n",
    "    top_genes[cluster] = (up_genes, dn_genes)\n",
    "all_top_genes = [item for sublist in all_top_genes for item in sublist]  # flatten all genes to one list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We select data corresponding to only the `heatmap_top_n` up and downregulated genes for each cluster. We log-transform and normalize as before, plotting a heatmap for the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_data_norm_heatmap_f = df_data.loc[all_top_genes, :]\n",
    "\n",
    "# compute log normalization of matrix\n",
    "df_data_norm_heatmap_f = log2_normalize(df_data_norm_heatmap_f)\n",
    "\n",
    "# convert to zscores\n",
    "df_data_norm_heatmap_f = zscore_normalize(df_data_norm_heatmap_f) \n",
    "\n",
    "# Plot heatmap\n",
    "cases = df_data_norm_heatmap_f.columns\n",
    "heatmap_cluster_colors = [ cluster_colors[x] for x in df_clustered_umap.loc[cases, :][\"Cluster\"] ]\n",
    "\n",
    "figure_header(\"Figure 10\", \"Heatmap of top most differentially expressed genes\")\n",
    "sns.clustermap(df_data_norm_heatmap_f,xticklabels=False,col_colors = heatmap_cluster_colors); plt.show()\n",
    "figure_legend(\"Figure 10\", \"Heatmap of top most differentially expressed genes\", \"Color coding along the top edge indicates cluster designation of the corresponding case.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Perform survival analysis <a class=\"anchor\" id=\"survival\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not user_data:\n",
    "    time_to_death_col = \"days_to_death\"\n",
    "    time_to_last_followup_col = \"days_to_last_follow_up\"\n",
    "    vital_status_col = \"vital_status\"\n",
    "    alive_val = \"Alive\"\n",
    "    dead_val = \"Dead\"\n",
    "    survival_time_unit = \"days\"\n",
    "    \n",
    "    \n",
    "if \"\" in [time_to_death_col, \n",
    "              time_to_last_followup_col, \n",
    "              vital_status_col, \n",
    "              alive_val, \n",
    "              dead_val, \n",
    "              survival_time_unit]:\n",
    "    survival_cols_valid = False\n",
    "else:\n",
    "    survival_cols_valid = True\n",
    "    \n",
    "    try:\n",
    "        df_lifeline = df_clinical[[time_to_death_col,vital_status_col,time_to_last_followup_col]]\n",
    "        df_lifeline = pd.merge(left=df_lifeline,left_on=\"case_id\", right=df_data_norm_km, right_on=\"case_id\")\n",
    "\n",
    "        # map alive and dead values in vital_status_col to 0 and 1, respectively\n",
    "        df_lifeline[vital_status_col] = df_lifeline[vital_status_col].replace({alive_val: 0, dead_val: 1 })\n",
    "\n",
    "        # use the time_to_death_col for \"dead\" cases, \n",
    "        # and use time_to_last_followup_col for \"alive\" (right-censored) cases\n",
    "        time_col = np.zeros(df_lifeline.shape[0])\n",
    "    \n",
    "    except KeyError:\n",
    "        survival_cols_valid = False\n",
    "        \n",
    "    def map_to_time(entry):\n",
    "        vital_status = entry[vital_status_col]\n",
    "        if vital_status == 1: # dead\n",
    "            return entry[time_to_death_col]\n",
    "        elif vital_status == 0: # alive\n",
    "            return entry[time_to_last_followup_col]\n",
    "        else:\n",
    "            return np.nan\n",
    "    \n",
    "    if survival_cols_valid:\n",
    "        df_lifeline[\"time\"] = df_lifeline.apply(map_to_time, axis=1)\n",
    "\n",
    "        df_lifeline = df_lifeline[[\"Cluster\",\"time\",vital_status_col]]\n",
    "        df_lifeline = df_lifeline.sort_values(by = \"time\", ascending = True)\n",
    "        df_lifeline = df_lifeline.dropna() # if NaNs remain after all this, drop them (missing vital status)\n",
    "\n",
    "        figure_header(\"Table 9\", \"Dataframe constructed for survival analysis\")\n",
    "        display(df_lifeline)\n",
    "        figure_legend(\"Table 9\", \"Dataframe constructed for survival analysis\", \"For patients listed as living, the time used for survival analysis is the time to their last followup. For deceased patients, the time used is the time to death. Both of these values are counted from the date of diagnosis. Living patients are included as right-censored datapoints.\")\n",
    "\n",
    "        # Add the KM plot for a cluster to the existing figure\n",
    "        def plot_km(cluster):\n",
    "            kmf = KaplanMeierFitter()\n",
    "            f = df_lifeline[\"Cluster\"] == c\n",
    "            T = df_lifeline[f][\"time\"].apply(lambda x: float(x))\n",
    "            C = df_lifeline[f][vital_status_col].apply(lambda x: float(x))\n",
    "\n",
    "            censored_points = df_lifeline[f]\n",
    "            censored_points = censored_points[censored_points[vital_status_col] == 0][\"time\"].values\n",
    "            censored_points_x = [int(x) for x in censored_points]\n",
    "\n",
    "            kmf.fit(T, event_observed=C, label = f'Cluster {c}')\n",
    "            data = pd.DataFrame(kmf.survival_function_).reset_index()\n",
    "\n",
    "            # Add the plot curve\n",
    "            fig.add_trace(go.Scatter(x=data[\"timeline\"], \n",
    "                                     y=data[f\"Cluster {c}\"], \n",
    "                                     name=f\"Cluster {c}\",\n",
    "                                     line_shape='hv', \n",
    "                                     line_color=cluster_colors[c],\n",
    "                                    ))\n",
    "\n",
    "\n",
    "            # Display 95% confidence intervals\n",
    "            ci = kmf.confidence_interval_\n",
    "            ci_time = ci.index.values\n",
    "            ci_lower = ci[f\"Cluster {c}_lower_0.95\"]\n",
    "            ci_upper = ci[f\"Cluster {c}_upper_0.95\"]\n",
    "            fig.add_trace(go.Scatter(x=np.concatenate((ci_time,ci_time[::-1])), \n",
    "                                     y=np.concatenate((ci_upper,ci_lower[::-1])),  \n",
    "                                     fill='toself', \n",
    "                                     line_shape='hv', \n",
    "                                     fillcolor=cluster_colors[c],\n",
    "                                     line_color=cluster_colors[c],\n",
    "                                     opacity=0.2,\n",
    "                                     showlegend=False,\n",
    "                                     name=f\"Cluster {c}\"))\n",
    "\n",
    "            # Add markers to indicate the right-censored datapoints\n",
    "            data = data.set_index(\"timeline\")\n",
    "            censored_ests = data.loc[censored_points_x,f\"Cluster {c}\"]\n",
    "\n",
    "            fig.add_trace(go.Scatter(x=censored_points, \n",
    "                                     y=censored_ests,\n",
    "                                     mode=\"markers\",\n",
    "                                     marker_symbol=\"line-ns-open\",\n",
    "                                     showlegend=False,\n",
    "                                     marker_size = 8, \n",
    "                                     marker_color=cluster_colors[c],\n",
    "                                     name=\"Censored point\"))\n",
    "\n",
    "\n",
    "        clusters = df_lifeline[\"Cluster\"].unique()\n",
    "        fig = go.Figure()\n",
    "\n",
    "        for c in clusters:\n",
    "            plot_km(c)\n",
    "\n",
    "        fig.update_layout(\n",
    "            xaxis_title=f\"Time ({survival_time_unit})\",\n",
    "            yaxis_title=\"Probability of survival\",\n",
    "        )\n",
    "\n",
    "        figure_header(\"Figure 11\",\"Kaplan-Meier survival plots\")\n",
    "        fig.show()\n",
    "        figure_legend(\"Figure 11\", \"Kaplan-Meier survival plots\", \"Right-censored datapoints (patients with a vital status listed as alive as of their last followup) and indicated by the vertical lines. The 95% confidence interval for each curve is shown.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not survival_cols_valid:\n",
    "    print(\"The proper column names and values necessary for survival analysis were not provided, so K-M curves cannot be generated.\")\n",
    "else:\n",
    "    args = [df_lifeline['time'], df_lifeline['Cluster'], df_lifeline[vital_status_col]]\n",
    "\n",
    "    figure_header(\"Table 10\",\"Multivariate log-rank test\")\n",
    "    result_multi = multivariate_logrank_test(*args)\n",
    "    display(result_multi.summary.round({\"test_statistic\":3}))\n",
    "    figure_legend(\"Table 10\",\"Multivariate log-rank test\", \"Comparison of all clusters' survival functions.\")\n",
    "\n",
    "    figure_header(\"Table 11\",\"Pairwise log-rank test\")\n",
    "    result_pairwise = pairwise_logrank_test(*args)\n",
    "    display(result_pairwise.summary.round({\"test_statistic\":3}))\n",
    "    figure_legend(\"Table 11\",\"Pairwise log-rank test\", \"Pairwise comparisons of any two clusters' survival functions.\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  4. Load Enrichr Data  <a class=\"anchor\" id=\"enrichr\"></a>\n",
    "\n",
    "We query the Enrichr API for enrichment analysis of the top most upregulated and downregulated genes for each cluster calculated above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Util functions\n",
    "def enrichr_link_from_genes(genes, description='', enrichr_link='https://maayanlab.cloud/Enrichr'):\n",
    "    ''' Functional access to Enrichr API\n",
    "    '''\n",
    "    time.sleep(1)\n",
    "    resp = requests.post(enrichr_link + '/addList', files={\n",
    "    'list': (None, '\\n'.join(genes)),\n",
    "    'description': (None, description),\n",
    "    })\n",
    "    if resp.status_code != 200:\n",
    "        raise Exception('Enrichr failed with status {}: {}'.format(\n",
    "          resp.status_code,\n",
    "          resp.text,\n",
    "        ))\n",
    "    # wait a tinybit before returning link (backoff)\n",
    "    time.sleep(3)\n",
    "    result = resp.json()\n",
    "    return dict(result, link=enrichr_link + '/enrich?dataset=' + resp.json()['shortId'])\n",
    "\n",
    "def enrichr_get_top_results(userListId, bg, enrichr_link='https://maayanlab.cloud/Enrichr'):\n",
    "    time.sleep(1)\n",
    "    resp = requests.get(enrichr_link + '/enrich?userListId={}&backgroundType={}'.format(userListId, bg))\n",
    "    if resp.status_code != 200:\n",
    "        raise Exception('Enrichr failed with status {}: {}'.format(\n",
    "          resp.status_code,\n",
    "          resp.text,\n",
    "        ))\n",
    "    time.sleep(3)\n",
    "    return pd.DataFrame(resp.json()[bg], columns=['rank', 'term', 'pvalue', 'zscore', 'combinedscore', 'overlapping_genes', 'adjusted_pvalue', '', ''])\n",
    "\n",
    "def save_enrichr_data(cancer_type,df_diff_expr,df_clustered_umap,df_all_results,df_cluster_aucs):\n",
    "    output=f\"appyter_data/{cancer_type}\"\n",
    "    os.makedirs(output, exist_ok=True)\n",
    "    df_diff_expr.to_csv(\n",
    "        f'{output}/df.tsv',\n",
    "        sep='\\t',\n",
    "        index=None\n",
    "    )\n",
    "    df_clustered_umap.to_csv(\n",
    "        f'{output}/df_umap.tsv',\n",
    "        sep='\\t',\n",
    "        index=None\n",
    "    )\n",
    "    df_all_results.to_csv(\n",
    "        f'{output}/df_enrich.tsv',\n",
    "        sep='\\t',\n",
    "        index=None\n",
    "    )\n",
    "    df_cluster_aucs.to_csv(\n",
    "        f'{output}/cluster_aucs.csv',\n",
    "        sep='\\t',\n",
    "        index=None\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get Enrichr links for each cluster\n",
    "enrichr_links = {}\n",
    "\n",
    "for cluster, (up_genes, dn_genes) in top_genes.items():\n",
    "    up_link, dn_link = None, None\n",
    "    if up_genes.size:\n",
    "        try:\n",
    "            up_link = enrichr_link_from_genes(up_genes, f'cluster {cluster} up')\n",
    "        except:\n",
    "            print(f'Enrichr failed for cluster {cluster} up genes')\n",
    "    else:\n",
    "        print(f'cluster {cluster} up: empty')\n",
    "    if dn_genes.size:\n",
    "        try:\n",
    "            dn_link = enrichr_link_from_genes(dn_genes, f'cluster {cluster} down')\n",
    "        except:\n",
    "            print(f'Enrichr failed for cluster {cluster} down genes')\n",
    "    else:\n",
    "        print(f'cluster {cluster} down: empty')\n",
    "    enrichr_links[cluster] = (up_link, dn_link)\n",
    "\n",
    "# Grab top results for each cluster\n",
    "all_enrichr_results = []\n",
    "for cluster, (up_link, dn_link) in enrichr_links.items():\n",
    "    for link_type, link in [('up', up_link), ('down', dn_link)]:\n",
    "        if link is None:\n",
    "            continue\n",
    "        for category, libraries in enrichr_libraries.items():\n",
    "            for library in libraries:\n",
    "                try:\n",
    "                    results = enrichr_get_top_results(link['userListId'], library).sort_values('pvalue').iloc[:top_n_results]\n",
    "                    results['link'] = link['link']\n",
    "                    results['library'] = library\n",
    "                    results['category'] = category\n",
    "                    results['direction'] = link_type\n",
    "                    results['cluster'] = cluster\n",
    "                    all_enrichr_results.append(results)\n",
    "                except:\n",
    "                    print('{}: {} {} {} cluster {} failed, continuing'.format(link, library, category, link_type, cluster))\n",
    "\n",
    "df_enrichr_results = pd.concat(all_enrichr_results).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display a dataframe with clickable enrichr links\n",
    "figure_header(\"Table 12\", \"Enrichment analysis results from Enrichr\")\n",
    "df_clickable = df_enrichr_results.copy()\n",
    "df_clickable['link'] = df_clickable[\"link\"].apply(make_clickable)\n",
    "table_html = df_clickable.to_html(escape=False)\n",
    "display(HTML(f'<div style=\"max-height: 250px; overflow-y: auto; margin-bottom: 25px;\">{table_html}</div>'))\n",
    "download_button(df_enrichr_results.to_csv(), 'Download Enrichr results', 'Enrichr results.csv')\n",
    "figure_legend(\"Table 12\",\"Enrichment analysis results from Enrichr\", \"Results are grouped by  expression direction (up/down) and gene set library. Within groups, results are sorted by lowest p-value (highest rank) first.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_header(\"Figure 12\", \"First two UMAP components of RNA-Seq data\")\n",
    "fig = px.scatter(\n",
    "    df_clustered_umap,\n",
    "    x = df_clustered_umap['UMAP-0'],\n",
    "    y = df_clustered_umap['UMAP-1'],\n",
    "    color = \"Cluster\",\n",
    ")\n",
    "\n",
    "fig.update_traces(marker=dict(\n",
    "            size=12,\n",
    "            opacity=0.8,\n",
    "            line=dict(\n",
    "                color='white',\n",
    "                width=1\n",
    "            )))\n",
    "fig.show()\n",
    "figure_legend(\"Figure 12\", \"First two UMAP components of RNA-Seq data\", \"Datapoints are color-coded by cluster for reference in the following tables.\")\n",
    "\n",
    "figure_header(\"Figure 13\", \"Enichment results, by cluster\")\n",
    "\n",
    "# Make horizontal barplots to visualize top Enrichr results\n",
    "clusters = df_enrichr_results[\"cluster\"].unique()\n",
    "for cluster in clusters:\n",
    "    cluster_results = df_enrichr_results.loc[df_enrichr_results[\"cluster\"] == cluster, :]\n",
    "    libraries = cluster_results[\"library\"].unique()\n",
    "    num_rows = len(libraries)\n",
    "\n",
    "    count = 1 # keep track of which subplot we're on\n",
    "    fig = plt.figure(figsize=(15,5*num_rows))\n",
    "    \n",
    "    for library in cluster_results[\"library\"].unique():\n",
    "            library_results = cluster_results.loc[cluster_results[\"library\"] == library, :]\n",
    "            for direction in library_results[\"direction\"].unique():\n",
    "                plot_results = library_results.loc[cluster_results[\"direction\"] == direction, :]\n",
    "                plot_results = plot_results.sort_values(\"pvalue\",ascending=False)\n",
    "                labels = plot_results[\"term\"]\n",
    "                labels = [ '\\n'.join(wrap(l, 20)) for l in labels ]\n",
    "                values = plot_results[\"pvalue\"]\n",
    "                values = -np.log(values)\n",
    "                \n",
    "                # normalize values to map from 0-1 -> color, with opacity also based on normalized pvalue\n",
    "                cmap = plt.get_cmap('cool')\n",
    "                norm_values = [ 0.3 + (x - min(values))/(max(values) - min(values))*0.7 for x in values]\n",
    "                colors = [ [*cmap(val)[:3], 0.4  + 0.2*val] for val in norm_values]\n",
    "                \n",
    "                # plot result\n",
    "                ax = fig.add_subplot(num_rows,2,count)\n",
    "                ax.barh(labels,values,color = colors)\n",
    "                ax.set_title(f'{library}\\n{direction} genes')\n",
    "                ax.set_xlabel(' – log(pvalue)')\n",
    "                count += 1\n",
    "                \n",
    "    cluster_heading(cluster)\n",
    "    #title = plt.suptitle(f'Cluster {cluster}',fontsize=24, color=\"white\",bbox=(dict(facecolor=cluster_colors[cluster], alpha=0.8,pad=5)))\n",
    "    fig.tight_layout(pad=3, w_pad=2, h_pad=6)\n",
    "    plt.show()\n",
    "    display(HTML(\"<br><br>\"))\n",
    "    \n",
    "figure_legend(\"Figure 13\", \"Enichment results, by cluster\", \"Bar plots indicate the negative log of the p-value for the specified term. One plot is presented per cluster, per gene-set library, per expression direction (up/down).\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Load L1000 Data  <a class=\"anchor\" id=\"l1000\"></a>\n",
    "\n",
    "We query the <a href=\"https://amp.pharm.mssm.edu/L1000FWD/\" target=\"_blank\">L1000 Fireworks Display</a> (L1000FWD) API to find the most similar and most dissimilar gene expression signatures from the L1000 database for each cluster based on the up and down genes. We also provide links to the L1000FWD interactive projection of those results for each cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Util functions\n",
    "def l1000fwd_results_from_genes(up_genes, down_genes, description='', l100fwd_link='http://maayanlab.cloud/L1000FWD/'):\n",
    "    ''' Functional access to L1000FWD API\n",
    "    '''\n",
    "    import time\n",
    "    time.sleep(1)\n",
    "    response = requests.post(l100fwd_link + 'sig_search', json={\n",
    "    'up_genes': list(up_genes),\n",
    "    'down_genes': list(down_genes),\n",
    "    })\n",
    "    l1000fwd_results =  {}\n",
    "    if response.status_code != 200:\n",
    "        raise Exception('L1000FWD failed with status {}: {}'.format(\n",
    "          response.status_code,\n",
    "          response.text,\n",
    "        ))\n",
    "    if 'KeyError' in response.text:\n",
    "        l1000fwd_results['result_url'] = None\n",
    "    else:\n",
    "        # Get ID and URL\n",
    "        result_id = response.json()['result_id']\n",
    "        l1000fwd_results['result_url'] = 'https://maayanlab.cloud/l1000fwd/vanilla/result/'+result_id\n",
    "        l1000fwd_results['result_id'] = result_id\n",
    "\n",
    "        # Get Top\n",
    "        l1000fwd_results['signatures'] = requests.get(l100fwd_link + 'result/topn/' + result_id).json()\n",
    "\n",
    "    # wait a tinybit before returning link (backoff)\n",
    "    time.sleep(1)\n",
    "    return l1000fwd_results\n",
    "\n",
    "def l1000fwd_sig_link(sig_id):\n",
    "    return 'https://maayanlab.cloud/dmoa/sig/' + sig_id\n",
    "\n",
    "def get_signature_by_id(sig_id):\n",
    "    response = requests.get(\"http://maayanlab.cloud/L1000FWD/sig/\" + sig_id)\n",
    "    if response.status_code != 200:\n",
    "        raise Exception('L1000FWD signature query  failed with status {}: {}'.format(\n",
    "          response.status_code,\n",
    "          response.text,\n",
    "        ))\n",
    "    return response.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_l1000fwd_results(l1000fwd_results, plot_counter,cluster_id,nr_drugs=7, height=300):\n",
    "    # Check if results\n",
    "    if l1000fwd_results['result_url']:\n",
    "\n",
    "        # Display cluster title\n",
    "        display(HTML('<br><br>'))\n",
    "        cluster_heading(cluster)\n",
    "\n",
    "        # Display IFrae\n",
    "        display(HTML(f\"<a href='{l1000fwd_results['result_url']}' target='_blank'> View L1000FWD for cluster {cluster_id}</a>\"))\n",
    "    \n",
    "        # Display tables\n",
    "        for direction, signature_list in l1000fwd_results['signatures'].items():\n",
    "\n",
    "            # Fix dataframe\n",
    "            rename_dict = {'sig_id': 'Signature ID', 'pvals': 'P-value', 'qvals': 'FDR', 'zscores': 'Z-score', 'combined_scores': 'Combined Score'}\n",
    "            signature_dataframe = pd.DataFrame(signature_list)[list(rename_dict.keys())].rename(columns=rename_dict).sort_values('P-value').rename_axis('Rank')\n",
    "            signature_dataframe.index = [x + 1 for x in range(len(signature_dataframe.index))]\n",
    "            signature_csv = signature_dataframe.to_csv(sep=\",\")\n",
    "\n",
    "            # Display table\n",
    "            pd.set_option('max.colwidth', None)\n",
    "            signature_dataframe['Signature ID'] = [f'<a href={l1000fwd_sig_link(x)} target=\"_blank\">{x}</a>' for x in signature_dataframe['Signature ID']]\n",
    "            table_html = signature_dataframe.to_html(escape=False, classes='w-100')\n",
    "            display(HTML(f'<h3>{direction.title()} Signatures: </h3>'))\n",
    "            display(HTML(f'<style>.w-100{{width: 100% !important;}}</style><div style=\"max-height: 250px; overflow-y: auto; margin-bottom: 25px;\">{table_html}</div>'))\n",
    "\n",
    "            # Display download button\n",
    "            download_button(signature_csv, f'Download {direction.title()} Signatures', f'Cluster {cluster_id} L1000FWD {direction.title()} signatures.csv')\n",
    "        # Link\n",
    "        display(HTML('Full results available at: <a href=\"{result_url}\" target=\"_blank\">{result_url}</a>.'.format(**l1000fwd_results)))\n",
    "        \n",
    "    # Display error\n",
    "    else:\n",
    "        display(Markdown('### No results were found.\\n This is likely due to the fact that the gene identifiers were not recognized by L1000FWD. Please note that L1000FWD currently only supports HGNC gene symbols (https://www.genenames.org/). If your dataset uses other gene identifier systems, such as Ensembl IDs or Entrez IDs, consider converting them to HGNC. Automated gene identifier conversion is currently under development.'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_counter = 0\n",
    "all_l1000fwd_results = {}\n",
    "figure_header(\"Figure 14\", \"Most similar and opposite L1000 signatures, by cluster\")\n",
    "for cluster, (up_genes, dn_genes) in top_genes.items():\n",
    "    try:\n",
    "        results = l1000fwd_results_from_genes(up_genes,dn_genes)\n",
    "        all_l1000fwd_results[cluster] = results\n",
    "        display_l1000fwd_results(results,plot_counter,cluster)\n",
    "        plot_counter += 1\n",
    "    except:\n",
    "        print(f'L1000FWD API failed for cluster {cluster}, continuing')\n",
    "\n",
    "        \n",
    "figure_legend(\"Figure 14\", \"Most similar and opposite L1000 signatures, by cluster\", \"Results are sorted by smallest p-value.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the signatures most \"opposite\" to the profile of a given cluster (i.e. up and down genes are reversed), we can obtain a set of drugs that may \"perturb\" that cluster into the healthy direction by upregulating its downregulated genes and downregulating the upregulated genes. These may present effective treatments for patients belonging to that cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_drugs = pd.read_csv(\"https://maayanlab.cloud/l1000fwd/download/Drugs_metadata.csv\")\n",
    "\n",
    "# Load top drug suggestions for each cluster based on the drugs used to produce the top_n_drugs opposite signatures\n",
    "drug_results = {}\n",
    "for cluster, results in all_l1000fwd_results.items():\n",
    "    opposite_sigs = results[\"signatures\"][\"opposite\"][:top_n_drugs]\n",
    "    sig_ids = [sig[\"sig_id\"] for sig in opposite_sigs]\n",
    "    pert_ids = []\n",
    "    for sig_id in sig_ids:\n",
    "        try:\n",
    "            signature = get_signature_by_id(sig_id)\n",
    "            pert_ids.append(signature[\"pert_id\"])\n",
    "        except: \n",
    "            print(f'L1000FWD API failed for cluster {cluster}, sig_id {sig_id}, continuing')\n",
    "    \n",
    "    df_cluster_drugs = df_drugs[df_drugs[\"pert_id\"].isin(pert_ids)].copy()\n",
    "    df_cluster_drugs[\"cluster\"] = cluster\n",
    "    df_cluster_drugs = df_cluster_drugs[[\"cluster\", *list(filter(lambda x: x!=\"cluster\", df_cluster_drugs.columns))]]\n",
    "    drug_results[cluster] = df_cluster_drugs\n",
    "    \n",
    "df_all_drugs = pd.concat(drug_results).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display a dataframe with clickable L1000FWD links\n",
    "figure_header(\"Table 13\", \"Drugs used to produce most opposite signatures for each cluster\")\n",
    "df_clickable = df_all_drugs.copy()\n",
    "df_clickable['pert_url'] = df_clickable[\"pert_url\"].apply(make_clickable)\n",
    "table_html = df_clickable.to_html(escape=False)\n",
    "display(HTML(f'<div style=\"max-height: 250px; overflow-y: auto; margin-bottom: 25px;\">{table_html}</div>'))\n",
    "download_button(df_all_drugs.to_csv(), 'Download L1000FWD drug results', 'L1000FWD drugs.csv')\n",
    "figure_legend(\"Table 13\", \"Drugs used to produce most opposite signatures for each cluster\", \"Each entry is a drug/chemical used for perturbation in the L1000 experiments that resulted in a gene-expression signature most opposite to that of the specified cluster.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "data_filename = ''\n",
    "clinical_data_filename = ''\n",
    "cancer = '''Papillary adenocarcinoma, NOS'''\n",
    "n_pca_components = 10\n",
    "n_neighbors = 40\n",
    "min_cluster_dist = 0.3\n",
    "top_n_genes = 2500\n",
    "max_clusters_calculation = 'root/2'\n",
    "use_weighted_silhouette_score = True\n",
    "use_second_deriv_method = True\n",
    "top_n_genes_enrichment = 250\n",
    "heatmap_top_n = 100\n",
    "top_n_results = 5\n",
    "use_default_libraries = False\n",
    "transcription_libraries = []\n",
    "pathways_libraries = []\n",
    "ontologies_libraries = []\n",
    "diseases_drugs_libraries = []\n",
    "cell_types_libraries = []\n",
    "miscellaneous_libraries = []\n",
    "legacy_libraries = []\n",
    "crowd_libraries = []\n",
    "top_n_drugs = 5\n",
    "default_libraries = OrderedDict([\n",
    "    ('Diseases/Drugs', ['GWAS_Catalog_2019']), \n",
    "    ('Ontologies', ['GO_Biological_Process_2018','MGI_Mammalian_Phenotype_Level_4_2019']),\n",
    "    ('Pathways', ['KEGG_2019_Human','KEGG_2019_Mouse']),\n",
    "    ('Transcription', ['ENCODE_TF_ChIP-seq_2015']),\n",
    "])\n",
    "time_to_death_col = ''\n",
    "time_to_last_followup_col = ''\n",
    "vital_status_col = ''\n",
    "alive_val = ''\n",
    "dead_val = ''\n",
    "survival_time_unit = 'days'\n",
    "if use_default_libraries:\n",
    "    enrichr_libraries = default_libraries\n",
    "else: \n",
    "    enrichr_libraries = OrderedDict([\n",
    "        ('Diseases/Drugs', diseases_drugs_libraries), \n",
    "        ('Ontologies', ontologies_libraries),\n",
    "        ('Cell Type', cell_types_libraries),\n",
    "        ('Pathways', pathways_libraries),\n",
    "        ('Transcription', transcription_libraries),\n",
    "        ('Legacy', legacy_libraries),\n",
    "        ('Crowd', crowd_libraries)\n",
    "    ])\n",
    "    all_empty = True\n",
    "    for key,libs in enrichr_libraries.items():\n",
    "        if len(libs) > 0:\n",
    "            all_empty = False\n",
    "            break\n",
    "    if all_empty:\n",
    "        enrichr_libraries = default_libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "cancers=[\"Infiltrating duct carcinoma, NOS\",\n",
    "         \"Squamous cell carcinoma, NOS\",\n",
    "         \"Lobular carcinoma, NOS\",\n",
    "         \"Acute myeloid leukemia, NOS\",\n",
    "         \"Hepatocellular carcinoma, NOS\",\n",
    "         \"Serous cystadenocarcinoma, NOS\",\n",
    "         \"Endometrioid adenocarcinoma, NOS\",\n",
    "         \"Adenocarcinoma, NOS\",\n",
    "         \"Clear cell adenocarcinoma, NOS\",\n",
    "         \"Glioblastoma\",\"Mucinous adenocarcinoma\",\n",
    "         \"Transitional cell carcinoma\",\n",
    "         \"Malignant melanoma, NOS\",\n",
    "         \"Papillary adenocarcinoma, NOS\"]\n",
    "cancer = cancers[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TCGA",
   "language": "python",
   "name": "tcga"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
