{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import requests\n",
    "from maayanlab_bioinformatics.enrichment import enrich_crisp\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as colors\n",
    "import base64\n",
    "from IPython.display import display, Markdown\n",
    "\n",
    "# bokeh\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool, CustomJS, ColumnDataSource, Span\n",
    "from bokeh.layouts import layout, column, gridplot\n",
    "from bokeh.layouts import row as bkrow\n",
    "from bokeh.palettes import all_palettes\n",
    "\n",
    "# display graphics\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichr Volcano Plot Visualizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide\n",
    "{% do SectionField(\n",
    "    name = 'GENE_SETS',\n",
    "    title = 'Gene Set Upload',\n",
    "    subtitle = 'Upload up and down gene sets.',\n",
    "    img = 'enrichr.png'\n",
    ")%}\n",
    "\n",
    "{% do SectionField(\n",
    "    name = 'ENRICHR_LIB',\n",
    "    title = 'Enrichr Library Selection',\n",
    "    subtitle = 'Select background library for enrichment analysis.',\n",
    "    img = 'enrichr.png'\n",
    ")%}\n",
    "\n",
    "{% set up_genes_kind = TabField(\n",
    "    name = 'up_genes_kind',\n",
    "    label = 'Up Genes',\n",
    "    default = 'Paste',\n",
    "    description = 'Paste or upload your up-regulated gene list.',\n",
    "    choices = {\n",
    "        'Paste': [\n",
    "            TextField(\n",
    "                name = 'up_genes_input',\n",
    "                label = 'Up Gene List',\n",
    "                default = 'KIAA0907\\nKDM5A\\nCDC25A\\nEGR1\\nGADD45B\\nRELB\\nTERF2IP\\nSMNDC1\\nTICAM1\\nNFKB2\\nRGS2\\nNCOA3\\nICAM1\\nTEX10\\nCNOT4\\nARID4B\\nCLPX\\nCHIC2\\nCXCL2\\nFBXO11\\nMTF2\\nCDK2\\nDNTTIP2\\nGADD45A\\nGOLT1B\\nPOLR2K\\nNFKBIE\\nGABPB1\\nECD\\nPHKG2\\nRAD9A\\nNET1\\nKIAA0753\\nEZH2\\nNRAS\\nATP6V0B\\nCDK7\\nCCNH\\nSENP6\\nTIPARP\\nFOS\\nARPP19\\nTFAP2A\\nKDM5B\\nNPC1\\nTP53BP2\\nNUSAP1',\n",
    "                description = 'Paste your list of up-regulated genes (one gene per row).',\n",
    "                section = 'GENE_SETS'\n",
    "            ),\n",
    "        ],\n",
    "        'Upload': [\n",
    "            FileField(\n",
    "                name = 'up_genes_filename',\n",
    "                label = 'Up Gene List File',\n",
    "                default = '',\n",
    "                description = 'Upload your list of up-regulated genes as a text file (one gene per row).',\n",
    "                section = 'GENE_SETS'\n",
    "            ),\n",
    "        ],\n",
    "    },\n",
    "    section = 'GENE_SETS'\n",
    ") %}\n",
    "\n",
    "{% set dn_genes_kind = TabField(\n",
    "    name = 'dn_genes_kind',\n",
    "    label = 'Down Genes',\n",
    "    default = 'Paste',\n",
    "    description = 'Paste or upload your down-regulated gene list.',\n",
    "    choices = {\n",
    "        'Paste': [\n",
    "            TextField(\n",
    "                name = 'dn_genes_input',\n",
    "                label = 'Down Gene List',\n",
    "                default = 'SCCPDH\\nKIF20A\\nFZD7\\nUSP22\\nPIP4K2B\\nCRYZ\\nGNB5\\nEIF4EBP1\\nPHGDH\\nRRAGA\\nSLC25A46\\nRPA1\\nHADH\\nDAG1\\nRPIA\\nP4HA2\\nMACF1\\nTMEM97\\nMPZL1\\nPSMG1\\nPLK1\\nSLC37A4\\nGLRX\\nCBR3\\nPRSS23\\nNUDCD3\\nCDC20\\nKIAA0528\\nNIPSNAP1\\nTRAM2\\nSTUB1\\nDERA\\nMTHFD2\\nBLVRA\\nIARS2\\nLIPA\\nPGM1\\nCNDP2\\nBNIP3\\nCTSL1\\nCDC25B\\nHSPA8\\nEPRS\\nPAX8\\nSACM1L\\nHOXA5\\nTLE1\\nPYGL\\nTUBB6\\nLOXL1',\n",
    "                description = 'Paste your list of down-regulated genes (one gene per row).',\n",
    "                section = 'GENE_SETS'\n",
    "            ),\n",
    "        ],\n",
    "        'Upload': [\n",
    "            FileField(\n",
    "                name = 'dn_genes_filename',\n",
    "                label = 'Down Gene List File',\n",
    "                default = '',\n",
    "                description = 'Upload your list of down-regulated genes as a text file (one gene per row).',\n",
    "                section = 'GENE_SETS'\n",
    "            ),\n",
    "        ],\n",
    "    },\n",
    "    section = 'GENE_SETS'\n",
    ") %}\n",
    "\n",
    "{% set enrichr_library = ChoiceField(\n",
    "    name='enrichr_library', \n",
    "    description='Select one Enrichr library for which to create visualizations.', \n",
    "    label='Enrichr Library', \n",
    "    default='WikiPathways_2019_Human', \n",
    "    section = 'ENRICHR_LIB',\n",
    "    choices=[\n",
    "        'ARCHS4_Cell-lines',\n",
    "        'ARCHS4_IDG_Coexp',\n",
    "        'ARCHS4_Kinases_Coexp',\n",
    "        'ARCHS4_TFs_Coexp',\n",
    "        'ARCHS4_Tissues',\n",
    "        'Achilles_fitness_decrease',\n",
    "        'Achilles_fitness_increase',\n",
    "        'Aging_Perturbations_from_GEO_down',\n",
    "        'Aging_Perturbations_from_GEO_up',\n",
    "        'Allen_Brain_Atlas_10x_scRNA_2021',\n",
    "        'Allen_Brain_Atlas_down',\n",
    "        'Allen_Brain_Atlas_up',\n",
    "        'BioCarta_2013',\n",
    "        'BioCarta_2015',\n",
    "        'BioCarta_2016',\n",
    "        'BioPlanet_2019',\n",
    "        'BioPlex_2017',\n",
    "        'CCLE_Proteomics_2020',\n",
    "        'CORUM',\n",
    "        'COVID-19_Related_Gene_Sets',\n",
    "        'Cancer_Cell_Line_Encyclopedia',\n",
    "        'ChEA_2013',\n",
    "        'ChEA_2015',\n",
    "        'ChEA_2016',\n",
    "        'Chromosome_Location',           \n",
    "        'Chromosome_Location_hg19',\n",
    "        'ClinVar_2019',\n",
    "        'dbGaP',\n",
    "        'DSigDB',\n",
    "        'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "        'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "        'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "        'DisGeNET',\n",
    "        'Disease_Perturbations_from_GEO_down',\n",
    "        'Disease_Perturbations_from_GEO_up',\n",
    "        'Disease_Signatures_from_GEO_down_2014',\n",
    "        'Disease_Signatures_from_GEO_up_2014',\n",
    "        'DrugMatrix',\n",
    "        'Drug_Perturbations_from_GEO_2014',\n",
    "        'Drug_Perturbations_from_GEO_down',\n",
    "        'Drug_Perturbations_from_GEO_up',\n",
    "        'ENCODE_Histone_Modifications_2013',\n",
    "        'ENCODE_Histone_Modifications_2015',\n",
    "        'ENCODE_TF_ChIP-seq_2014',\n",
    "        'ENCODE_TF_ChIP-seq_2015',\n",
    "        'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "        'ESCAPE',\n",
    "        'Elsevier_Pathway_Collection',\n",
    "        'Enrichr_Libraries_Most_Popular_Genes',\n",
    "        'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "        'Enrichr_Users_Contributed_Lists_2020',\n",
    "        'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "        'GO_Biological_Process_2013',\n",
    "        'GO_Biological_Process_2015',\n",
    "        'GO_Biological_Process_2017',\n",
    "        'GO_Biological_Process_2017b',\n",
    "        'GO_Biological_Process_2018',\n",
    "        'GO_Cellular_Component_2013',\n",
    "        'GO_Cellular_Component_2015',\n",
    "        'GO_Cellular_Component_2017',\n",
    "        'GO_Cellular_Component_2017b',\n",
    "        'GO_Cellular_Component_2018',\n",
    "        'GO_Molecular_Function_2013',\n",
    "        'GO_Molecular_Function_2015',\n",
    "        'GO_Molecular_Function_2017',\n",
    "        'GO_Molecular_Function_2017b',\n",
    "        'GO_Molecular_Function_2018',\n",
    "        'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "        'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "        'GWAS_Catalog_2019',\n",
    "        'GeneSigDB',\n",
    "        'Gene_Perturbations_from_GEO_down',\n",
    "        'Gene_Perturbations_from_GEO_up',\n",
    "        'Genes_Associated_with_NIH_Grants',\n",
    "        'Genome_Browser_PWMs',\n",
    "        'HMDB_Metabolites',\n",
    "        'HMS_LINCS_KinomeScan',\n",
    "        'HomoloGene',\n",
    "        'HumanCyc_2015',\n",
    "        'HumanCyc_2016',\n",
    "        'Human_Gene_Atlas',\n",
    "        'Human_Phenotype_Ontology',\n",
    "        'huMAP',\n",
    "        'InterPro_Domains_2019',\n",
    "        'Jensen_COMPARTMENTS',\n",
    "        'Jensen_DISEASES',\n",
    "        'Jensen_TISSUES',\n",
    "        'KEA_2013',\n",
    "        'KEA_2015',\n",
    "        'KEGG_2013',\n",
    "        'KEGG_2015',\n",
    "        'KEGG_2016',\n",
    "        'KEGG_2019_Human',\n",
    "        'KEGG_2019_Mouse',\n",
    "        'Kinase_Perturbations_from_GEO_down',\n",
    "        'Kinase_Perturbations_from_GEO_up',\n",
    "        'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "        'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "        'LINCS_L1000_Chem_Pert_down',\n",
    "        'LINCS_L1000_Chem_Pert_up',\n",
    "        'LINCS_L1000_Ligand_Perturbations_down',\n",
    "        'LINCS_L1000_Ligand_Perturbations_up',\n",
    "        'Ligand_Perturbations_from_GEO_down',\n",
    "        'Ligand_Perturbations_from_GEO_up',\n",
    "        'lncHUB_lncRNA_Co-Expression',\n",
    "        'MCF7_Perturbations_from_GEO_down',\n",
    "        'MCF7_Perturbations_from_GEO_up',\n",
    "        'MGI_Mammalian_Phenotype_2013',\n",
    "        'MGI_Mammalian_Phenotype_2017',\n",
    "        'MGI_Mammalian_Phenotype_Level_3',\n",
    "        'MGI_Mammalian_Phenotype_Level_4',\n",
    "        'MGI_Mammalian_Phenotype_Level_4_2019',\n",
    "        'MSigDB_Computational',\n",
    "        'MSigDB_Hallmark_2020',\n",
    "        'MSigDB_Oncogenic_Signatures',\n",
    "        'Microbe_Perturbations_from_GEO_down',\n",
    "        'Microbe_Perturbations_from_GEO_up',\n",
    "        'miRTarBase_2017',\n",
    "        'Mouse_Gene_Atlas',\n",
    "        'NCI-60_Cancer_Cell_Lines',\n",
    "        'NCI-Nature_2015',\n",
    "        'NCI-Nature_2016',\n",
    "        'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "        'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "        'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "        'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "        'NURSA_Human_Endogenous_Complexome',\n",
    "        'OMIM_Disease',\n",
    "        'OMIM_Expanded',\n",
    "        'Old_CMAP_down',\n",
    "        'Old_CMAP_up',\n",
    "        'PPI_Hub_Proteins',\n",
    "        'Panther_2015',\n",
    "        'Panther_2016',\n",
    "        'Pfam_Domains_2019',\n",
    "        'Pfam_InterPro_Domains',\n",
    "        'PheWeb_2019',\n",
    "        'Phosphatase_Substrates_from_DEPOD',\n",
    "        'ProteomicsDB_2020',\n",
    "        'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "        'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "        'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "        'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "        'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "        'Reactome_2013',\n",
    "        'Reactome_2015',\n",
    "        'Reactome_2016',\n",
    "        'SILAC_Phosphoproteomics',\n",
    "        'SubCell_BarCode',\n",
    "        'SysMyo_Muscle_Gene_Sets'\n",
    "        'TF-LOF_Expression_from_GEO',\n",
    "        'TF_Perturbations_Followed_by_Expression',\n",
    "        'TG_GATES_2020',\n",
    "        'TRANSFAC_and_JASPAR_PWMs',\n",
    "        'TRRUST_Transcription_Factors_2019',\n",
    "        'Table_Mining_of_CRISPR_Studies',\n",
    "        'TargetScan_microRNA',\n",
    "        'TargetScan_microRNA_2017',\n",
    "        'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
    "        'Tissue_Protein_Expression_from_ProteomicsDB.csv',\n",
    "        'Transcription_Factor_PPIs',\n",
    "        'UK_Biobank_GWAS_v1',\n",
    "        'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "        'VirusMINT',\n",
    "        'Virus_Perturbations_from_GEO_down',\n",
    "        'Virus_Perturbations_from_GEO_up',\n",
    "        'WikiPathways_2013',\n",
    "        'WikiPathways_2015',\n",
    "        'WikiPathways_2016',\n",
    "        'WikiPathways_2019_Human',\n",
    "        'WikiPathways_2019_Mouse'\n",
    "    ]\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if up_genes_kind.raw_value == 'Paste' %}\n",
    "up_genes_input = {{ up_genes_kind.value[0] }}\n",
    "{%- else %}\n",
    "up_genes_filename = {{ up_genes_kind.value[0] }}\n",
    "{%- endif %}\n",
    "\n",
    "{%- if dn_genes_kind.raw_value == 'Paste' %}\n",
    "dn_genes_input = {{ dn_genes_kind.value[0] }}\n",
    "{%- else %}\n",
    "dn_genes_filename = {{ dn_genes_kind.value[0] }}\n",
    "{%- endif %}\n",
    "\n",
    "enrichr_library = '{{ enrichr_library }}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if up_genes_kind.raw_value == 'Paste' %}\n",
    "up_genes = up_genes_input.split('\\n')\n",
    "up_genes = [x.strip() for x in up_genes]\n",
    "{%- else %}\n",
    "open_up_genes_file = open(up_genes_filename,'r')\n",
    "lines = open_up_genes_file.readlines()\n",
    "up_genes = [x.strip() for x in lines]\n",
    "open_up_genes_file.close()\n",
    "{%- endif %}\n",
    "\n",
    "{%- if dn_genes_kind.raw_value == 'Paste' %}\n",
    "dn_genes = dn_genes_input.split('\\n')\n",
    "dn_genes = [x.strip() for x in dn_genes]\n",
    "{%- else %}\n",
    "open_dn_genes_file = open(dn_genes_filename,'r')\n",
    "lines = open_dn_genes_file.readlines()\n",
    "dn_genes = [x.strip() for x in lines]\n",
    "open_dn_genes_file.close()\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters and Functions\n",
    "significance_value = 0.05\n",
    "\n",
    "def get_library(lib_name):\n",
    "    ''' Returns a dictionary mapping each term from the input library to\n",
    "    its associated geneset. '''\n",
    "\n",
    "    resp = requests.get('https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=json&libraryName=' + lib_name)\n",
    "    if resp.status_code == 200:\n",
    "        lib_data = resp.json()[lib_name]['terms']\n",
    "        return { term: lib_data[term].keys() for term in lib_data.keys() }\n",
    "    else:\n",
    "        return {}\n",
    "\n",
    "\n",
    "def enrich_genes(direction, genelist, libname):\n",
    "    ''' Returns dataframe of enrichment results for input gene list and\n",
    "    libray. '''\n",
    "\n",
    "    enrich_results = enrich_crisp(genelist, libname, 21000, True)\n",
    "    df = pd.DataFrame(\n",
    "        [ [\n",
    "            term, \n",
    "            res.pvalue, \n",
    "            res.odds_ratio,\n",
    "            direction,\n",
    "            ] for (term, res) in enrich_results ], \n",
    "        columns=['term', 'pvalue', 'odds_ratio', 'direction']\n",
    "    )\n",
    "    df['log_pval'] = np.negative(np.log10(df['pvalue']))\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# access Enrichr background library\n",
    "lib = get_library(enrichr_library)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The two volcano plots below provide visualizations of the enrichment results for the up-regulated and down-regulated gene lists when compared to the selected library. \n",
    "\n",
    "Each point represents a term/gene set from the selected library. The x-position gives the odds ratio, while the y-position gives the -log(p-value). Significant terms are larger and colored either blue (up-regulated genes) or red (down-regulated genes). The darker the color, the more significant the term. \n",
    "\n",
    "Hover over a point to view the gene set it represents and its corresponding odds ratio and p-value. \n",
    "\n",
    "Use the toolbar to the right of each plot to pan, zoom, or download the plot as an SVG.\n",
    "\n",
    "*Note: The two plots may have different x and y scales.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if lib == {}:\n",
    "    # handle failure to retrieve library\n",
    "    display(Markdown('## Failed to access library, please try again later.'))\n",
    "\n",
    "else:\n",
    "    # get enrichment results\n",
    "    up_res_df = enrich_genes('up', up_genes, lib)\n",
    "    dn_res_df = enrich_genes('down', dn_genes, lib)\n",
    "\n",
    "    # set plot color scheme\n",
    "    up_cmap = cm.get_cmap('Blues_r')\n",
    "    up_cnorm = colors.Normalize(vmin = up_res_df['pvalue'].min(), vmax = 0.1)\n",
    "\n",
    "    dn_cmap = cm.get_cmap('Reds_r')\n",
    "    dn_cnorm = colors.Normalize(vmin = dn_res_df['pvalue'].min(), vmax = 0.1)\n",
    "\n",
    "    # set point colors and sizes\n",
    "    up_colors = []\n",
    "    up_sizes = []\n",
    "    for r in up_res_df.itertuples():\n",
    "        if r.pvalue < 0.05:\n",
    "            up_colors.append(colors.to_hex(up_cmap(up_cnorm(r.pvalue))))\n",
    "            up_sizes.append(12)\n",
    "        else:\n",
    "            up_colors.append('#808080')\n",
    "            up_sizes.append(6)\n",
    "    dn_colors = []\n",
    "    dn_sizes = []\n",
    "    for r in dn_res_df.itertuples():\n",
    "        if r.pvalue < 0.05:\n",
    "            dn_colors.append(colors.to_hex(dn_cmap(dn_cnorm(r.pvalue))))\n",
    "            dn_sizes.append(12)\n",
    "        else:\n",
    "            dn_colors.append('#808080')\n",
    "            dn_sizes.append(6)\n",
    "\n",
    "    # create data sources\n",
    "    up_source = ColumnDataSource(\n",
    "        data=dict(\n",
    "            x = up_res_df['odds_ratio'],\n",
    "            y = up_res_df['log_pval'],\n",
    "            gene_set = up_res_df['term'],\n",
    "            pvalue = up_res_df['pvalue'],\n",
    "            oddsratio = up_res_df['odds_ratio'],\n",
    "            colors = up_colors,\n",
    "            sizes = up_sizes\n",
    "        )\n",
    "    )\n",
    "    dn_source = ColumnDataSource(\n",
    "        data=dict(\n",
    "            x = dn_res_df['odds_ratio'],\n",
    "            y = dn_res_df['log_pval'],\n",
    "            gene_set = dn_res_df['term'],\n",
    "            pvalue = dn_res_df['pvalue'],\n",
    "            oddsratio = dn_res_df['odds_ratio'],\n",
    "            colors = dn_colors,\n",
    "            sizes = dn_sizes\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # create hover tooltips\n",
    "    tools = [\n",
    "        (\"Gene Set\", \"@gene_set\"),\n",
    "        (\"P-Value\", \"@pvalue\"),\n",
    "        (\"Odds Ratio\", \"@oddsratio\"),\n",
    "    ]\n",
    "\n",
    "    # create plots\n",
    "    up_plot = figure(\n",
    "        plot_width = 700, \n",
    "        plot_height = 700,\n",
    "        tooltips=tools\n",
    "    )\n",
    "    up_plot.circle(\n",
    "        'x', 'y', size = 'sizes', \n",
    "        alpha = 0.7, line_alpha = 0, \n",
    "        line_width = 0.01, source = up_source, \n",
    "        fill_color = 'colors', name = 'up_res_df'\n",
    "    )    \n",
    "\n",
    "    dn_plot = figure(\n",
    "        plot_width = 700, \n",
    "        plot_height = 700,\n",
    "        tooltips=tools\n",
    "    )\n",
    "    dn_plot.circle(\n",
    "        'x', 'y', size = 'sizes',\n",
    "        alpha = 0.7, line_alpha = 0,\n",
    "        line_width = 0.01, source = dn_source,\n",
    "        fill_color = 'colors', name = 'dn_res_df'\n",
    "    )\n",
    "\n",
    "    # set axis labels and title \n",
    "    up_plot.xaxis.axis_label = \"Odds Ratio\"\n",
    "    up_plot.yaxis.axis_label = \"-log10(p-value)\"\n",
    "    up_plot.title.text = f\"Enriched terms for up-regulated genes from {enrichr_library}\"\n",
    "    up_plot.title.align = \"center\"\n",
    "    up_plot.title.text_font_size = \"16px\"\n",
    "\n",
    "    dn_plot.xaxis.axis_label = \"Odds Ratio\"\n",
    "    dn_plot.yaxis.axis_label = \"-log10(p-value)\"\n",
    "    dn_plot.title.text = f\"Enriched terms for down-regulated genes from {enrichr_library}\"\n",
    "    dn_plot.title.align = \"center\"\n",
    "    dn_plot.title.text_font_size = \"16px\"\n",
    "\n",
    "    up_plot.output_backend = \"svg\"\n",
    "    dn_plot.output_backend = \"svg\"\n",
    "\n",
    "    # show plot\n",
    "    grid = gridplot([[up_plot, dn_plot]], plot_width=250, plot_height=250, sizing_mode=\"scale_both\", merge_tools=False)\n",
    "    show(grid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.5 64-bit",
   "metadata": {
    "interpreter": {
     "hash": "718219df2fd2c80bcb49c8a0c7604dc011bfde52c5cacdfbdfbee3bd8a62893c"
    }
   },
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5-final"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
