{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichr Scatterplot Visualizer\n",
    "\n",
    "This appyter creates a scatterplot visualizing enrichment analysis results from Enrichr (https://amp.pharm.mssm.edu/Enrichr/).\n",
    "\n",
    "The resulting figure will contain a scatterplot plot of each selected Enrichr library grouped by gene set similiarity, with blue points indicating gene sets with a significant similarity to the inputted gene set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import umap\n",
    "from maayanlab_bioinformatics.enrichment import enrich_crisp\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as colors\n",
    "from IPython.display import display, HTML\n",
    "import base64\n",
    "\n",
    "# Bokeh\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool, CustomJS, ColumnDataSource, Legend, LegendItem, Span\n",
    "from bokeh.layouts import layout, row, column, gridplot\n",
    "from bokeh.palettes import all_palettes\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(name='section1', title = '1. Submit Your Gene List', subtitle = 'Upload a text file containing your gene list -OR- copy and paste your gene list into the text box below (One gene per row). You can also try the default gene list provided.', img = 'enrichr-icon.png')%}\n",
    "{% do SectionField(name='section2', title = '2. Choose Enrichr Library', subtitle = 'Select between 1 and 9 Enrichr libraries you would like in your figure.', img = 'enrichr-icon.png')%}\n",
    "{% do SectionField(name='section3', title = '3. Choose a Significance Value', subtitle = 'Choose a significance value for the p-values when performing enrichment analysis.', img = 'enrichr-icon.png')%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "gene_list_filename = {{ FileField(name='gene_list_filename', label='Gene List File', default='', description='Upload your gene list as a text file (One gene per row).',section = 'section1') }}\n",
    "\n",
    "gene_list_input = {{ TextField(name='gene_list_input', label='Gene List', default='NSUN3\\nPOLRMT\\nNLRX1\\nSFXN5\\nZC3H12C\\nSLC25A39\\nARSG\\nDEFB29\\nNDUFB6\\nZFAND1\\nTMEM77\\n5730403B10RIK\\nRP23-195K8.6\\nTLCD1\\nPSMC6\\nSLC30A6\\nLOC100047292\\nLRRC40\\nORC5L\\nMPP7\\nUNC119B\\nPRKACA\\nTCN2\\nPSMC3IP\\nPCMTD2\\nACAA1A\\nLRRC1\\n2810432D09RIK\\nSEPHS2\\nSAC3D1\\nTMLHE\\nLOC623451\\nTSR2\\nPLEKHA7\\nGYS2\\nARHGEF12\\nHIBCH\\nLYRM2\\nZBTB44\\nENTPD5\\nRAB11FIP2\\nLIPT1\\nINTU\\nANXA13\\nKLF12\\nSAT2\\nGAL3ST2\\nVAMP8\\nFKBPL\\nAQP11\\nTRAP1\\nPMPCB\\nTM7SF3\\nRBM39\\nBRI3\\nKDR\\nZFP748\\nNAP1L1\\nDHRS1\\nLRRC56\\nWDR20A\\nSTXBP2\\nKLF1\\nUFC1\\nCCDC16\\n9230114K14RIK\\nRWDD3\\n2610528K11RIK\\nACO1\\nCABLES1\\nLOC100047214\\nYARS2\\nLYPLA1\\nKALRN\\nGYK\\nZFP787\\nZFP655\\nRABEPK\\nZFP650\\n4732466D17RIK\\nEXOSC4\\nWDR42A\\nGPHN\\n2610528J11RIK\\n1110003E01RIK\\nMDH1\\n1200014M14RIK\\nAW209491\\nMUT\\n1700123L14RIK\\n2610036D13RIK\\nCOX15\\nTMEM30A\\nNSMCE4A\\nTM2D2\\nRHBDD3\\nATXN2\\nNFS1\\n3110001I20RIK\\nBC038156\\nLOC100047782\\n2410012H22RIK\\nRILP\\nA230062G08RIK\\nPTTG1IP\\nRAB1\\nAFAP1L1\\nLYRM5\\n2310026E23RIK\\nC330002I19RIK\\nZFYVE20\\nPOLI\\nTOMM70A\\nSLC7A6OS\\nMAT2B\\n4932438A13RIK\\nLRRC8A\\nSMO\\nNUPL2\\nTRPC2\\nARSK\\nD630023B12RIK\\nMTFR1\\n5730414N17RIK\\nSCP2\\nZRSR1\\nNOL7\\nC330018D20RIK\\nIFT122\\nLOC100046168\\nD730039F16RIK\\nSCYL1\\n1700023B02RIK\\n1700034H14RIK\\nFBXO8\\nPAIP1\\nTMEM186\\nATPAF1\\nLOC100046254\\nLOC100047604\\nCOQ10A\\nFN3K\\nSIPA1L1\\nSLC25A16\\nSLC25A40\\nRPS6KA5\\nTRIM37\\nLRRC61\\nABHD3\\nGBE1\\nPARP16\\nHSD3B2\\nESM1\\nDNAJC18\\nDOLPP1\\nLASS2\\nWDR34\\nRFESD\\nCACNB4\\n2310042D19RIK\\nSRR\\nBPNT1\\n6530415H11RIK\\nCLCC1\\nTFB1M\\n4632404H12RIK\\nD4BWG0951E\\nMED14\\nADHFE1\\nTHTPA\\nCAT\\nELL3\\nAKR7A5\\nMTMR14\\nTIMM44\\nSF1\\nIPP\\nIAH1\\nTRIM23\\nWDR89\\nGSTZ1\\nCRADD\\n2510006D16RIK\\nFBXL6\\nLOC100044400\\nZFP106\\nCD55\\n0610013E23RIK\\nAFMID\\nTMEM86A\\nALDH6A1\\nDALRD3\\nSMYD4\\nNME7\\nFARS2\\nTASP1\\nCLDN10\\nA930005H10RIK\\nSLC9A6\\nADK\\nRBKS\\n2210016F16RIK\\nVWCE\\n4732435N03RIK\\nZFP11\\nVLDLR\\n9630013D21RIK\\n4933407N01RIK\\nFAHD1\\nMIPOL1\\n1810019D21RIK\\n1810049H13RIK\\nTFAM\\nPAICS\\n1110032A03RIK\\nLOC100044139\\nDNAJC19\\nBC016495\\nA930041I02RIK\\nRQCD1\\nUSP34\\nZCCHC3\\nH2AFJ\\nPHF7\\n4921508D12RIK\\nKMO\\nPRPF18\\nMCAT\\nTXNDC4\\n4921530L18RIK\\nVPS13B\\nSCRN3\\nTOR1A\\nAI316807\\nACBD4\\nFAH\\nAPOOL\\nCOL4A4\\nLRRC19\\nGNMT\\nNR3C1\\nSIP1\\nASCC1\\nFECH\\nABHD14A\\nARHGAP18\\n2700046G09RIK\\nYME1L1\\nGK5\\nGLO1\\nSBK1\\nCISD1\\n2210011C24RIK\\nNXT2\\nNOTUM\\nANKRD42\\nUBE2E1\\nNDUFV1\\nSLC33A1\\nCEP68\\nRPS6KB1\\nHYI\\nALDH1A3\\nMYNN\\n3110048L19RIK\\nRDH14\\nPROZ\\nGORASP1\\nLOC674449\\nZFP775\\n5430437P03RIK\\nNPY\\nADH5\\nSYBL1\\n4930432O21RIK\\nNAT9\\nLOC100048387\\nMETTL8\\nENY2\\n2410018G20RIK\\nPGM2\\nFGFR4\\nMOBKL2B\\nATAD3A\\n4932432K03RIK\\nDHTKD1\\nUBOX5\\nA530050D06RIK\\nZDHHC5\\nMGAT1\\nNUDT6\\nTPMT\\nWBSCR18\\nLOC100041586\\nCDK5RAP1\\n4833426J09RIK\\nMYO6\\nCPT1A\\nGADD45GIP1\\nTMBIM4\\n2010309E21RIK\\nASB9\\n2610019F03RIK\\n7530414M10RIK\\nATP6V1B2\\n2310068J16RIK\\nDDT\\nKLHDC4\\nHPN\\nLIFR\\nOVOL1\\nNUDT12\\nCDAN1\\nFBXO9\\nFBXL3\\nHOXA7\\nALDH8A1\\n3110057O12RIK\\nABHD11\\nPSMB1\\nENSMUSG00000074286\\nCHPT1\\nOXSM\\n2310009A05RIK\\n1700001L05RIK\\nZFP148\\n39509\\nMRPL9\\nTMEM80\\n9030420J04RIK\\nNAGLU\\nPLSCR2\\nAGBL3\\nPEX1\\nCNO\\nNEO1\\nASF1A\\nTNFSF5IP1\\nPKIG\\nAI931714\\nD130020L05RIK\\nCNTD1\\nCLEC2H\\nZKSCAN1\\n1810044D09RIK\\nMETTL7A\\nSIAE\\nFBXO3\\nFZD5\\nTMEM166\\nTMED4\\nGPR155\\nRNF167\\nSPTLC1\\nRIOK2\\nTGDS\\nPMS1\\nPITPNC1\\nPCSK7\\n4933403G14RIK\\nEI24\\nCREBL2\\nTLN1\\nMRPL35\\n2700038C09RIK\\nUBIE\\nOSGEPL1\\n2410166I05RIK\\nWDR24\\nAP4S1\\nLRRC44\\nB3BP\\nITFG1\\nDMXL1\\nC1D', description='Paste your gene list (One gene per row).', section = 'section1') }}\n",
    "\n",
    "enrichr_libraries = {{ MultiCheckboxField(name='transcription_libraries', description='Select up to 9 Enrichr libraries you would like in your figure.', label='Library Selection', default='ChEA_2016', section = 'section2',choices=[\n",
    "    'ARCHS4_TFs_Coexp',\n",
    "    'ChEA_2016',\n",
    "    'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "    'ENCODE_Histone_Modifications_2015',\n",
    "    'ENCODE_TF_ChIP-seq_2015',\n",
    "    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "    'Genome_Browser_PWMs',\n",
    "    'lncHUB_lncRNA_Co-Expression',\n",
    "    'miRTarBase_2017',\n",
    "    'TargetScan_microRNA_2017',\n",
    "    'TF_Perturbations_Followed_by_Expression',\n",
    "    'TRANSFAC_and_JASPAR_PWMs',\n",
    "    'TRRUST_Transcription_Factors_2019',\n",
    "    'ARCHS4_Kinases_Coexp',\n",
    "    'BioCarta_2016',\n",
    "    'BioPlanet_2019',\n",
    "    'CORUM',\n",
    "    'Elsevier_Pathway_Collection',\n",
    "    'HMS_LINCS_KinomeScan',\n",
    "    'HumanCyc_2016',\n",
    "    'huMAP',\n",
    "    'KEA_2015',\n",
    "    'KEGG_2019_Human',\n",
    "    'KEGG_2019_Mouse',\n",
    "    'Kinase_Perturbations_from_GEO_down',\n",
    "    'Kinase_Perturbations_from_GEO_up',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "    'NCI-Nature_2016',\n",
    "    'NURSA_Human_Endogenous_Complexome',\n",
    "    'Panther_2016',\n",
    "    'Phosphatase_Substrates_from_DEPOD',\n",
    "    'PPI_Hub_Proteins',\n",
    "    'Reactome_2016',\n",
    "    'SILAC_Phosphoproteomics',\n",
    "    'SubCell_BarCode',\n",
    "    'WikiPathways_2019_Human',\n",
    "    'WikiPathways_2019_Mouse',\n",
    "    'GO_Cellular_Component_2018',\n",
    "    'GO_Molecular_Function_2018',\n",
    "    'Human_Phenotype_Ontology',\n",
    "    'Jensen_COMPARTMENTS',\n",
    "    'Jensen_DISEASES',\n",
    "    'Jensen_TISSUES',\n",
    "    'Achilles_fitness_decrease',\n",
    "    'Achilles_fitness_increase',\n",
    "    'ARCHS4_IDG_Coexp',\n",
    "    'ClinVar_2019',\n",
    "    'dbGaP',\n",
    "    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "    'Allen_Brain_Atlas_down',\n",
    "    'Allen_Brain_Atlas_up',\n",
    "    'ARCHS4_Cell-lines',\n",
    "    'ARCHS4_Tissues',\n",
    "    'Cancer_Cell_Line_Encyclopedia',\n",
    "    'CCLE_Proteomics_2020',\n",
    "    'ESCAPE',\n",
    "    'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "    'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "    'Human_Gene_Atlas',\n",
    "    'Mouse_Gene_Atlas',\n",
    "    'NCI-60_Cancer_Cell_Lines',\n",
    "    'ProteomicsDB_2020',\n",
    "    'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
    "    'Chromosome_Location_hg19',\n",
    "    'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "    'Enrichr_Libraries_Most_Popular_Genes',\n",
    "    'HMDB_Metabolites',\n",
    "    'HomoloGene',\n",
    "    'InterPro_Domains_2019',\n",
    "    'Pfam_Domains_2019',\n",
    "    'Pfam_InterPro_Domains',\n",
    "    'Table_Mining_of_CRISPR_Studies',\n",
    "    ]) }}\n",
    "\n",
    "significance_value = {{ FloatField(name='significance_value', label='Significance Value', default='0.05', description='Enter a value at which p-values are significant.', section = 'section3') }}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import gene list as file or from text box file\n",
    "# Will choose file upload over textbox if a file is given \n",
    "if gene_list_filename != '':\n",
    "    open_gene_list_file = open(gene_list_filename,'r')\n",
    "    lines = open_gene_list_file.readlines()\n",
    "    genes = [x.strip() for x in lines]\n",
    "    open_gene_list_file.close()\n",
    "else:\n",
    "    genes = gene_list_input.split('\\n')\n",
    "    genes = [x.strip() for x in genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "def download_library(library_name):\n",
    "    # download pre-processed library data\n",
    "    df = pd.read_csv('https://raw.githubusercontent.com/MaayanLab/Enrichr-Viz-Appyter/master/Enrichr-Processed-Library-Storage/Scatterplot/Libraries/' + library_name + '.csv')\n",
    "\n",
    "    name = df['Name'].tolist()\n",
    "    gene_list = df['Genes'].tolist()\n",
    "    library_data = [list(a) for a in zip(name, gene_list)]\n",
    "    return genes, library_data, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# enrichment analysis\n",
    "def get_library_iter(library_data):\n",
    "    for member in library_data:\n",
    "        term = member[0]\n",
    "        gene_set = member[1].split(' ')\n",
    "        yield term, gene_set\n",
    "\n",
    "def get_enrichment_results(genes, library_data):\n",
    "    return sorted(enrich_crisp(genes, get_library_iter(library_data), 20000, True), key=lambda r: r[1].pvalue)\n",
    "\n",
    "def get_pvalue(row, unzipped_results, all_results):\n",
    "    if row['Name'] in list(unzipped_results[0]):\n",
    "        index = list(unzipped_results[0]).index(row['Name'])\n",
    "        return all_results[index][1].pvalue\n",
    "    else:\n",
    "        return 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# call enrichment results\n",
    "def get_plot(library_name):\n",
    "    genes, library_data, df = download_library(library_name)\n",
    "    all_results = get_enrichment_results(genes, library_data)\n",
    "    unzipped_results = list(zip(*all_results))\n",
    "\n",
    "    if len(all_results) == 0:\n",
    "        print(\"There are no enriched terms with your inputted gene set and the \", library_name, \" library.\")\n",
    "        my_colors = ['#808080'] * len(df.index)\n",
    "\n",
    "        source = ColumnDataSource(\n",
    "            data=dict(\n",
    "                x = df['x'],\n",
    "                y = df['y'],\n",
    "                gene_set = df['Name'],\n",
    "                colors = my_colors,\n",
    "                sizes = [6] * len(df.index)\n",
    "            )\n",
    "        )\n",
    "\n",
    "        hover_emb = HoverTool(names=[\"df\"], tooltips=\"\"\"\n",
    "            <div style=\"margin: 10\">\n",
    "                <div style=\"margin: 0 auto; width:200px;\">\n",
    "                    <span style=\"font-size: 12px; font-weight: bold;\">Gene Set:</span>\n",
    "                    <span style=\"font-size: 12px\">@gene_set</span>\n",
    "                </div>\n",
    "            </div>\n",
    "            \"\"\")\n",
    "    else:\n",
    "        # add p value to the dataframe\n",
    "        df['p value'] = df.apply (lambda row: get_pvalue(row, unzipped_results, all_results), axis=1)\n",
    "\n",
    "        # normalize p values for color scaling\n",
    "        cmap = mpl.cm.get_cmap('Blues_r')\n",
    "        norm = colors.Normalize(vmin = df['p value'].min(), vmax=significance_value*2)\n",
    "\n",
    "        my_colors = []\n",
    "        my_sizes = []\n",
    "        for index, row in df.iterrows():\n",
    "            if row['p value'] < significance_value:\n",
    "                my_colors += [mpl.colors.to_hex(cmap(norm(row['p value'])))]\n",
    "                my_sizes += [12]\n",
    "            else:\n",
    "                my_colors += ['#808080']\n",
    "                my_sizes += [6]\n",
    "\n",
    "        source = ColumnDataSource(\n",
    "                data=dict(\n",
    "                    x = df['x'],\n",
    "                    y = df['y'],\n",
    "                    gene_set = df['Name'],\n",
    "                    p_value = df['p value'],\n",
    "                    colors = my_colors,\n",
    "                    sizes = my_sizes\n",
    "                )\n",
    "            )\n",
    "\n",
    "        hover_emb = HoverTool(names=[\"df\"], tooltips=\"\"\"\n",
    "            <div style=\"margin: 10\">\n",
    "                <div style=\"margin: 0 auto; width:200px;\">\n",
    "                    <span style=\"font-size: 12px; font-weight: bold;\">Gene Set:</span>\n",
    "                    <span style=\"font-size: 12px\">@gene_set</span>\n",
    "                    <span style=\"font-size: 12px; font-weight: bold;\">p-value:</span>\n",
    "                    <span style=\"font-size: 12px\">@p_value</span>\n",
    "                </div>\n",
    "            </div>\n",
    "            \"\"\")\n",
    "\n",
    "        # add a legend with most significant enriched term\n",
    "        most_sig_name = str(df[df['p value'] == df['p value'].min()]['Name'].values[0])\n",
    "        most_sig_value = str(df['p value'].min())\n",
    "\n",
    "        if len(enrichr_libraries) == 1:\n",
    "            legend = Legend(items = [LegendItem(label = most_sig_name + ', p-value: ' + most_sig_value)], location = (70, 2))\n",
    "        else:\n",
    "            legend = Legend(items = [LegendItem(label = library_name.replace('_', ' ') + '     ')], location = (0, 10))\n",
    "\n",
    "\n",
    "    tools_emb = [hover_emb, 'pan', 'wheel_zoom', 'reset', 'save']\n",
    "\n",
    "    # create different sized plots depending on how many are selected\n",
    "    if len(enrichr_libraries) == 1:\n",
    "        plot_emb = figure(plot_width=700, plot_height=800, tools=tools_emb)\n",
    "    elif len(enrichr_libraries) == 2:\n",
    "        plot_emb = figure(plot_width=600, plot_height=600, tools=tools_emb)\n",
    "    elif len(enrichr_libraries) == 3 or len(enrichr_libraries) == 5 or len(enrichr_libraries) == 6:\n",
    "        plot_emb = figure(plot_width=400, plot_height=400, tools=tools_emb)\n",
    "    elif len(enrichr_libraries) == 4:\n",
    "        plot_emb = figure(plot_width=500, plot_height=500, tools=tools_emb)\n",
    "    elif 7 <= len(enrichr_libraries) and len(enrichr_libraries) <= 9:\n",
    "        plot_emb = figure(plot_width=350, plot_height=350, tools=tools_emb)\n",
    "\n",
    "\n",
    "    if len(all_results) != 0:\n",
    "        plot_emb.add_layout(legend, 'below')\n",
    "\n",
    "    # hide axis labels and grid lines\n",
    "    plot_emb.xaxis.major_tick_line_color = None\n",
    "    plot_emb.xaxis.minor_tick_line_color = None\n",
    "    plot_emb.yaxis.major_tick_line_color = None\n",
    "    plot_emb.yaxis.minor_tick_line_color = None\n",
    "    plot_emb.xaxis.major_label_text_font_size = '0pt'\n",
    "    plot_emb.yaxis.major_label_text_font_size = '0pt' \n",
    "\n",
    "    plot_emb.circle('x', 'y', size = 'sizes', alpha = 0.7, line_alpha = 0, \n",
    "                    line_width = 0.01, source = source, fill_color = 'colors', name = \"df\")\n",
    "\n",
    "    plot_emb.output_backend = \"svg\"\n",
    "    \n",
    "    return plot_emb, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list = []\n",
    "plot_list = []\n",
    "for library in enrichr_libraries:\n",
    "    plot, df = get_plot(library)\n",
    "    df_list += [df]\n",
    "    plot_list += [plot]\n",
    "\n",
    "if len(enrichr_libraries) == 1:\n",
    "    show(plot_list[0])\n",
    "elif len(enrichr_libraries) == 2:\n",
    "    show(row(plot_list[0], plot_list[1]))\n",
    "elif len(enrichr_libraries) == 3:\n",
    "    show(row(plot_list[0], plot_list[1], plot_list[2]))\n",
    "elif len(enrichr_libraries) == 4:\n",
    "    show(layout([plot_list[0], plot_list[1]], [plot_list[2], plot_list[3]]))\n",
    "elif len(enrichr_libraries) == 5:\n",
    "    show(layout([plot_list[0], plot_list[1], plot_list[2]], [plot_list[3], plot_list[4]]))\n",
    "elif len(enrichr_libraries) == 6:\n",
    "    show(layout([plot_list[0], plot_list[1], plot_list[2]], [plot_list[3], plot_list[4], plot_list[5]]))\n",
    "elif len(enrichr_libraries) == 7:\n",
    "    show(layout([plot_list[0], plot_list[1], plot_list[2]], [plot_list[3], plot_list[4], plot_list[5]], [plot_list[6]]))\n",
    "elif len(enrichr_libraries) == 8:\n",
    "    show(layout([plot_list[0], plot_list[1], plot_list[2]], [plot_list[3], plot_list[4], plot_list[5]], [plot_list[6], plot_list[7]]))\n",
    "elif len(enrichr_libraries) == 9:\n",
    "    show(layout([plot_list[0], plot_list[1], plot_list[2]], [plot_list[3], plot_list[4], plot_list[5]], [plot_list[6], plot_list[7], plot_list[8]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You may have to zoom in to see the details in densely-populated portions of the plots.\n",
    "\n",
    "The larger blue points represent significantly enriched terms (the darker the blue, the more significant).  \n",
    "The gray points are not significant.\n",
    "\n",
    "Hovering over points will display the associated gene set name and the p-value.  \n",
    "\n",
    "Below, tables of significant gene sets and their p-values are displayed and available for download (one for each library)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output significant p-values\n",
    "\n",
    "def create_download_link( df, title = \"Download CSV file of the table above\", filename = \"data.csv\"):  \n",
    "    csv = df.to_csv(index = False)\n",
    "    b64 = base64.b64encode(csv.encode())\n",
    "    payload = b64.decode()\n",
    "    html = '<a download=\"{filename}\" href=\"data:text/csv;base64,{payload}\" target=\"_blank\">{title}</a>'\n",
    "    html = html.format(payload=payload, title=title, filename=filename)\n",
    "    return HTML(html)\n",
    "\n",
    "for index, df in enumerate(df_list):\n",
    "    if 'p value' in df.columns:\n",
    "        sorted_df = df.sort_values(by = ['p value'])\n",
    "        filtered_df = sorted_df[sorted_df['p value'] <= significance_value]\n",
    "        if len(filtered_df) != 0:\n",
    "            display(HTML(f\"<strong>{enrichr_libraries[index].replace('_', ' ')}</strong>\"))\n",
    "            display(HTML(filtered_df[['Name', 'p value']].to_html(index = False)))\n",
    "            display(create_download_link(filtered_df[['Name', 'p value']]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('base': conda)",
   "name": "python37664bitbasecondad81f4d096e8b4b9885e3d702e0bb3be1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6-final"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
