{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e747f423",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0836d1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "{% do SectionField(\n",
    "    name='primary',\n",
    "    title='CFDE Gene Partnership',\n",
    "    subtitle='Resolve gene-centric information from CF DCCs',\n",
    "    img='geneicon.png',\n",
    ") %}\n",
    "{% set input = TabField(\n",
    "    name='input',\n",
    "    label='Gene or Variant',\n",
    "    default='Gene',\n",
    "    choices={\n",
    "        'Gene': [\n",
    "            AutocompleteField(\n",
    "                name='gene',\n",
    "                label='Gene Identifier',\n",
    "                description='Accepts valid HUGO Gene symbols, Entrez, Ensembl or UniProt Gene IDs among others',\n",
    "                default='ACE2',\n",
    "                file_path=url_for('static', filename='genes.json'),\n",
    "                examples={'ACE2': 'ACE2', 'ENSG00000170248': 'ENSG00000170248', 'P24941': 'P24941', 'SHOC2': 'SHOC2'},\n",
    "            ),\n",
    "        ],\n",
    "        'Variant': [\n",
    "            StringField(\n",
    "                name='variant',\n",
    "                label='Variant Identifier',\n",
    "                description='Please submit a HGVS or rsid variant identifier',\n",
    "                default='rs28897756',\n",
    "                examples={'rs28897756': 'rs28897756', 'chr6:g.152708291G>A': 'chr6:g.152708291G>A'},\n",
    "            ),\n",
    "        ],\n",
    "    },\n",
    "    section='primary',\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "764ab600",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "# CFDE Gene-Centric Appyter: {{ input.value[0].raw_value }}\n",
    "\n",
    "{%- if input.raw_value == \"Gene\" %}\n",
    "Given the gene *{{ input.value[0].raw_value }}*, we request information about it from several different DCCs in hopes of creating a comprehensive knowledge report for it.\n",
    "{%- elif input.raw_value == \"Variant\" %}\n",
    "Given the variant *{{ input.value[0].raw_value }}*, we map it to a gene and request information about it from several different DCCs in hopes of creating a comprehensive knowledge report for it.\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d880f5b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "import json\n",
    "import uuid\n",
    "import time\n",
    "import logging\n",
    "import requests\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import urllib.parse\n",
    "import plotly.express as px\n",
    "import scipy.stats as st\n",
    "import traceback\n",
    "from contextlib import contextmanager\n",
    "from textwrap import dedent\n",
    "from functools import partial, lru_cache\n",
    "from IPython.display import HTML, Markdown, display\n",
    "\n",
    "logger = logging.getLogger('')\n",
    "formatter = logging.Formatter('%(message)s')\n",
    "console = logging.StreamHandler()\n",
    "console.setFormatter(formatter)\n",
    "logger.addHandler(console)\n",
    "\n",
    "from itables_javascript import show\n",
    "import itables.options as opt\n",
    "opt.maxBytes = 0\n",
    "opt.showIndex = False\n",
    "\n",
    "class RenderJSON(object):\n",
    "    ''' https://gist.github.com/t27/48b3ac73a1479914f9fe9383e5d45325 '''\n",
    "    def __init__(self, json_data):\n",
    "        if isinstance(json_data, dict) or isinstance(json_data, list):\n",
    "            self.json_str = json.dumps(json_data)\n",
    "        else:\n",
    "            self.json_str = json_data\n",
    "        self.uuid = str(uuid.uuid4())\n",
    "        \n",
    "    def _ipython_display_(self):\n",
    "        display(HTML('<div id=\"{}\" style=\"height: auto; width:100%;\"></div>'.format(self.uuid)))\n",
    "        display(HTML(\"\"\"<script>\n",
    "        require([\"https://rawgit.com/caldwell/renderjson/master/renderjson.js\"], function() {\n",
    "          renderjson.set_show_to_level(1)\n",
    "          document.getElementById('%s').appendChild(renderjson(%s))\n",
    "        });</script>\n",
    "        \"\"\" % (self.uuid, self.json_str)))\n",
    "\n",
    "def combine_pvalues(pvalues, method='fisher', select='pvalue'):\n",
    "    ''' A helper for accessing this method via pd.agg which expects a scaler result\n",
    "    '''\n",
    "    statistic, pvalue = st.combine_pvalues(pvalues, method=method)\n",
    "    return dict(statistic=statistic, pvalue=pvalue)[select]\n",
    "\n",
    "def combine_zscores(zscores):\n",
    "    ''' A helper method for combining zscores via Stouffer method\n",
    "    '''\n",
    "    return zscores.sum() / np.sqrt(zscores.shape[0])\n",
    "\n",
    "def one(it):\n",
    "    try:\n",
    "        return next(iter(it))\n",
    "    except:\n",
    "        return None\n",
    "\n",
    "@contextmanager\n",
    "def ignore_exceptions(error_message=None):\n",
    "    ''' Show but ignore exceptions that occur within the context\n",
    "    Usage:\n",
    "    with ignore_exceptions():\n",
    "        my_fallible_operation()\n",
    "    '''\n",
    "    try:\n",
    "        yield\n",
    "    except Exception as e:\n",
    "        if error_message:\n",
    "            logger.error(error_message)\n",
    "        else:\n",
    "            traceback.print_exc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddd3d9c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input.raw_value == \"Variant\" %}\n",
    "variant_input = {{ input.value[0] }}\n",
    "{% elif input.raw_value == \"Gene\" %}\n",
    "gene_input = {{ input.value[0] }}\n",
    "{% else %}\n",
    "raise NotImplementedError\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b02a80b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if input.raw_value == \"Gene\" %}\n",
    "## MyGeneInfo: Query\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/mygeneinfo.png') }}\" />\n",
    "\n",
    "<https://mygene.info/>\n",
    "\n",
    "To interoperate with different APIs which support different gene identifier schemes. We'll first use mygene.info to resolve gene identifiers.\n",
    "{% elif input.raw_value == \"Variant\" %}\n",
    "## MyVariantInfo\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/myvariantinfo.png') }}\" />\n",
    "\n",
    "<https://myvariant.info/>\n",
    "\n",
    "To map Variants to the nearest Gene, we utilize the MyVariantInfo service which provides information about a variant including the gene mapping.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c816dbea",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input.raw_value == \"Gene\" %}\n",
    "@lru_cache()\n",
    "def mygene_info_query(geneId):\n",
    "    res = requests.get('https://mygene.info/v3/query', params=dict(q=geneId))\n",
    "    results = res.json()\n",
    "    return results\n",
    "\n",
    "{% elif input.raw_value == \"Variant\" %}\n",
    "@lru_cache()\n",
    "def myvariant_info(variantId):\n",
    "    res = requests.get(f\"https://myvariant.info/v1/variant/{urllib.parse.quote(variantId)}\")\n",
    "    results = res.json()\n",
    "    return results if type(results) == list else [results]\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a290a48a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input.raw_value == \"Gene\" %}\n",
    "gene_results = mygene_info_query(gene_input)\n",
    "display(RenderJSON(gene_results))\n",
    "{% elif input.raw_value == \"Variant\" %}\n",
    "variant_results = myvariant_info(variant_input)\n",
    "display(RenderJSON(variant_results))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b17c3b77",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input.raw_value == \"Gene\" %}\n",
    "if len(gene_results['hits']) == 0:\n",
    "    raise Exception(f\"Gene with identifier {gene_input} could not be found in MyGeneInfo!\")\n",
    "\n",
    "# find exact match\n",
    "geneId = one(hit['_id'] for hit in gene_results['hits'] if hit['symbol'] == gene_input)\n",
    "if not geneId:\n",
    "    # Select the most likely gene (mygeneinfo sorts by likelyhood)\n",
    "    _, geneId = max(\n",
    "        (hit['_score'], hit['_id'])\n",
    "        for hit in gene_results['hits']\n",
    "        if hit.keys() & {'_score', 'symbol'}\n",
    "    )\n",
    "display(Markdown(f\"**GeneID: {geneId}**\"))\n",
    "{% elif input.raw_value == \"Variant\" %}\n",
    "# Identify the gene name\n",
    "genes = list(filter(None, {\n",
    "    str(gene)\n",
    "    for result in variant_results\n",
    "    for gene in {\n",
    "        result.get('dbsnp', {}).get('gene', {}).get('geneid'),\n",
    "        result.get('clinvar', {}).get('gene', {}).get('id'),\n",
    "    }\n",
    "    if gene\n",
    "}))\n",
    "display(genes)\n",
    "if len(genes) == 0:\n",
    "    raise Exception('Gene could not be found in MyVariantInfo results!')\n",
    "elif len(genes) > 1:\n",
    "    display(HTML('<span style=\"font-size: 32pt; color: red; font-weight: bold;\">Gene Ambiguity!</span>'))\n",
    "# grab the first gene\n",
    "geneId = genes[0]\n",
    "display(Markdown(f\"### GeneID: {geneId}\"))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69348be9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "## MyGeneInfo\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/mygeneinfo.png') }}\" />\n",
    "\n",
    "<https://mygene.info/>\n",
    "\n",
    "With the Entrez Gene ID, we can resolve lots of different identifiers and identifiability information from mygene.info."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a823735a",
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache()\n",
    "def mygene_info(geneId):\n",
    "    req = requests.get(\n",
    "        f\"https://mygene.info/v3/gene/{geneId}\"\n",
    "    )\n",
    "    return req.json()\n",
    "\n",
    "gene_info = mygene_info(geneId)\n",
    "display(RenderJSON(gene_info))\n",
    "display(Markdown(f\"### Gene Symbol: {gene_info['symbol']}\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bf5c3ec",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Primary Information\n",
    "\n",
    "We query DCC APIs to gain insights about the primary information they collect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d8c3af7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "### GTEx\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/gtex.png') }}\" />\n",
    "\n",
    "<https://gtexportal.org/home/>\n",
    "\n",
    "We query the GTEx Data through the [GTEx API](https://gtexportal.org/home/api-docs/index.html) to identify tissue sites that significantly express the gene question."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d9d8240",
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache()\n",
    "def gtex_singleTissueEqtl(geneSymbol, datasetId='gtex_v8'):\n",
    "    res = requests.get(\n",
    "        'https://gtexportal.org/rest/v1/association/singleTissueEqtl',\n",
    "        params=dict(\n",
    "            format='json',\n",
    "            geneSymbol=geneSymbol,\n",
    "            datasetId=datasetId,\n",
    "        )\n",
    "    )\n",
    "    results = res.json()['singleTissueEqtl']\n",
    "    if len(results) == 0: raise Exception(f\"No information for gene with identifier {geneSymbol} found in GTEx\")\n",
    "    return pd.DataFrame(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0dbfd5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions(f\"Gene with identifier {gene_info['symbol']} currently not available in GTEx\"):\n",
    "    gtex_results = gtex_singleTissueEqtl(gene_info['symbol'])\n",
    "    columns = list(gtex_results.columns)\n",
    "    columns.insert(0, columns.pop(columns.index('nes')))\n",
    "    columns.insert(0, columns.pop(columns.index('pValue')))\n",
    "    columns.insert(0, columns.pop(columns.index('tissueSiteDetailId')))\n",
    "    gtex_results = gtex_results[columns]\n",
    "    show(gtex_results, order=[[gtex_results.columns.get_loc('pValue'), 'asc']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e34e7d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions('Could not process GTEx output'):\n",
    "    gtex_combined_stouffer_statistic = gtex_results.groupby('tissueSiteDetailId')['pValue'] \\\n",
    "        .agg(partial(combine_pvalues, method='stouffer', select='statistic')) \\\n",
    "        .to_frame('combined_stouffer_statistic') \\\n",
    "        .reset_index() \\\n",
    "        .sort_values('combined_stouffer_statistic', ascending=False)\n",
    "    gtex_combined_stouffer_statistic['group'] = gtex_combined_stouffer_statistic['tissueSiteDetailId'].apply(lambda name: name.split('_', maxsplit=1)[0])\n",
    "\n",
    "    fig = px.bar(\n",
    "        gtex_combined_stouffer_statistic,\n",
    "        y='combined_stouffer_statistic',\n",
    "        x='tissueSiteDetailId',\n",
    "        color='group',\n",
    "        orientation='v',\n",
    "        title=f\"Tissues with significant expression of {gene_info['symbol']} in GTEx\",\n",
    "        height=1000,\n",
    "    )\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d0fc7b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "### LINCS\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/LINCS.jpg') }}\" />\n",
    "\n",
    "<https://lincsproject.org/>\n",
    "\n",
    "#### L1000 RNAseq Gene Centric Signature Reverse Search (RGCSRS)\n",
    "\n",
    "[An appyter](https://appyters.maayanlab.cloud/L1000_RNAseq_Gene_Search/) was built for performing Gene Centric signature reverse searches against the LINCS data. Its functionality is repeated here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce1d2198",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as colors\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show as bokeh_show\n",
    "from bokeh.models import ColumnDataSource\n",
    "\n",
    "# display graphics\n",
    "output_notebook()\n",
    "\n",
    "root_path = 'https://appyters.maayanlab.cloud/storage/L1000_RNAseq_Gene_Search'\n",
    "\n",
    "def map_color(cd, logfc, red_norm, blue_norm):\n",
    "    v = cd*logfc\n",
    "\n",
    "    if v < 0: \n",
    "        return '#D3D3D3'\n",
    "    elif cd < 0:\n",
    "        return colors.to_hex(cm.get_cmap('Reds')(red_norm(cd*logfc)))\n",
    "    else:\n",
    "        return colors.to_hex(cm.get_cmap('Blues')(blue_norm(cd*logfc)))\n",
    "\n",
    "def make_plot(comb_df, gene, pert_type):\n",
    "    # check if there are any results\n",
    "    if comb_df.shape[0] == 0: \n",
    "        display(Markdown(\"### **There are no signatures in the pre-processed dataset for the chosen gene, cell line, and perturbation type inputs.**\"))\n",
    "        return\n",
    "\n",
    "    # set color and size for each point on plot\n",
    "    v = (comb_df['logFC']*comb_df['CD']).to_numpy()\n",
    "    red_norm = colors.Normalize(vmin=min(v)-0.005, vmax=max(v)+0.005)\n",
    "    blue_norm = colors.Normalize(vmin=min(v)-0.005, vmax=max(v)+0.005)\n",
    "\n",
    "    plot_colors = [map_color(row.CD, row.logFC, red_norm, blue_norm) for row in comb_df.itertuples()]\n",
    "\n",
    "    # generate data source\n",
    "    data_source = ColumnDataSource(\n",
    "        data=dict(\n",
    "            x = comb_df['logFC'],\n",
    "            y = comb_df['CD'].apply(abs),\n",
    "            cd = comb_df['CD'],\n",
    "            sig = pd.Series(comb_df.index),\n",
    "            fc = comb_df['FC'], \n",
    "            logfc = comb_df['logFC'],\n",
    "            colors = plot_colors, \n",
    "            sizes = [8] * comb_df.shape[0],\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # create hover tooltip\n",
    "    tools = [\n",
    "        (\"Signature\", \"@sig\"),\n",
    "        (\"CD Coeff\", \"@cd\"),\n",
    "        (\"Fold Change\", \"@fc\"),\n",
    "        (\"Log2 Fold Change\", \"@logfc\")\n",
    "    ]\n",
    "    # generate plot and relevant plot labels\n",
    "    plot = figure(\n",
    "        plot_width=700,\n",
    "        plot_height=500,\n",
    "        tooltips=tools\n",
    "    )\n",
    "    \n",
    "    plot.circle(\n",
    "        'x', 'y', \n",
    "        size='sizes',\n",
    "        alpha=0.7, \n",
    "        line_alpha=1,\n",
    "        line_width=1, \n",
    "        line_color='colors',\n",
    "        source=data_source,\n",
    "        fill_color='colors', \n",
    "        name=f\"{gene}_expression_in_L1000_to_RNAseq_{pert_type.replace(' ','')}_volcano_plot\"\n",
    "    )\n",
    "\n",
    "    plot.yaxis.axis_label = 'Abs(CD-Coefficient)'\n",
    "    plot.xaxis.axis_label = 'log2(Fold Change)'\n",
    "    plot.title.text = f\"Differential Expression of {gene} in RNA-seq-like {pert_type} Signatures\"\n",
    "    plot.title.align = 'center'\n",
    "    plot.title.text_font_size = '14px'\n",
    "    bokeh_show(plot)\n",
    "\n",
    "def make_tables(comb_df, pert, is_upreg):\n",
    "    dir_df = comb_df[comb_df['FC'] > 1] if is_upreg else comb_df[comb_df['FC'] < 1]\n",
    "    if dir_df.shape[0] == 0: \n",
    "        display(Markdown(f\"### There are no {'up-regulated' if is_upreg else 'down-regulated'} signatures for the chosen gene and cell line inputs.\"))\n",
    "        return dir_df\n",
    "    dir_df = dir_df.sort_values(by='FC', ascending=not is_upreg)\n",
    "    dir_df['FC'] = dir_df['FC'].apply(lambda x: f'{x:.4f}')\n",
    "    dir_df['CD'] = dir_df['CD'].apply(lambda x: f'{x:.4f}')\n",
    "    if pert == 'xpr':\n",
    "        dir_df['KO Gene'] = dir_df.index.map(lambda x: x.split('_')[4])\n",
    "    else:\n",
    "        dir_df['Perturbagen'] = dir_df.index.map(lambda x: x.split('_')[4])\n",
    "        dir_df['Dose'] = dir_df.index.map(lambda x: x.split('_')[5] if len(x.split('_')) == 6 else '')\n",
    "    dir_df['Cell Line'] = dir_df.index.map(lambda x: x.split('_')[1])\n",
    "    dir_df['Timepoint'] = dir_df.index.map(lambda x: x.split('_')[2].lower())\n",
    "    dir_df = dir_df.rename(columns={\n",
    "            'FC': 'Fold Change', \n",
    "            'logFC': 'Log2(Fold Change)', \n",
    "            'CD': 'CD Coefficient',\n",
    "            'Rank': 'Rank in Signature'})\n",
    "    dir_df.index.names = ['Signature']\n",
    "    return dir_df\n",
    "\n",
    "def download_link(df, fname):\n",
    "    if df.shape[0] == 0: return ''\n",
    "    csv = df.to_csv(fname, sep='\\t', index=True)\n",
    "    link = f'<div>Download full results: <a href=\"{fname}\" target=_blank>{fname}</a></div>'\n",
    "    return link\n",
    "\n",
    "def enrichr(pert, top_perts, direction, gene):\n",
    "    if pert == 'CRISPR':\n",
    "        desc = f\"Top CRISPR targets from RNA-seq-like signatures that {direction}-regulate {gene}\"\n",
    "        list_url = 'https://maayanlab.cloud/Enrichr/addList'\n",
    "        enrich_url = 'https://maayanlab.cloud/Enrichr/enrich'\n",
    "    else:\n",
    "        desc = f\"Top compounds from RNA-seq-like signatures that {direction}-regulate {gene}\"\n",
    "        list_url = 'https://maayanlab.cloud/DrugEnrichr/addList'\n",
    "        enrich_url = 'https://maayanlab.cloud/DrugEnrichr/enrich'\n",
    "    payload = {\n",
    "        'list': (None, '\\n'.join(top_perts)),\n",
    "        'description': (None, desc)\n",
    "    }\n",
    "    response = requests.post(list_url, files=payload)\n",
    "    if not response.ok:\n",
    "        raise Exception('Error analyzing gene list')\n",
    "    time.sleep(0.5)\n",
    "    return f\"{enrich_url}?dataset={response.json()['shortId']}\"\n",
    "\n",
    "def enrichr_link(pert, df, direction, gene): \n",
    "    # check if there are any results\n",
    "    if df.shape[0] < 5: \n",
    "        return(f\"There are not enough {direction}-regulated signatures to submit to {'Enrichr' if pert == 'CRISPR' else 'DrugEnrichr'}.\")\n",
    "    comb_df = df.copy()\n",
    "    comb_df['pert'] = comb_df.index.map(lambda x: x.split('_')[4])\n",
    "    if direction == 'up':\n",
    "        top_perts = comb_df.sort_values(by='FC', ascending=False) \\\n",
    "            .drop_duplicates(subset=['pert'],keep='first')['pert'][:20]\n",
    "    else:\n",
    "        top_perts = comb_df.sort_values(by='FC', ascending=True) \\\n",
    "            .drop_duplicates(subset=['pert'],keep='first')['pert'][:20]\n",
    "    pert_type = 'CRISPR target genes' if pert == 'CRISPR' else 'chemical compounds'\n",
    "    results_url = enrichr(pert, top_perts, direction, gene)\n",
    "    return f'<a href={results_url} target=\"_blank\">Enrichr analysis of top 20 {pert_type} that {direction}-regulate {gene}</a>'\n",
    "\n",
    "def l1000_rnaseq_directional_gene_search(gene=None, gene_data=None, pert=None, pert_full_name=None, pert_name=None, is_upreg=None):\n",
    "    direction = 'up' if is_upreg else 'down'\n",
    "    display(HTML(f'<div style=\"font-size:1rem;padding=1rem;\"><b>Top {pert_full_name} signatures where {gene} is {direction}-regulated (based on fold change)</b></div>'))\n",
    "    tbl = make_tables(gene_data, pert='cp', is_upreg=is_upreg)\n",
    "    display(HTML(tbl[:10].to_html(escape=False, col_space=70)))\n",
    "    display(HTML(download_link(tbl[:100], f\"{gene}_{direction.capitalize()}Reg_L1000_{pert_name}_signatures.tsv\")))\n",
    "    display(HTML(enrichr_link(pert_name, gene_data, direction, gene)))\n",
    "\n",
    "def l1000_rnaseq_gene_search(gene=None, gene_data=None, pert=None, pert_full_name=None, pert_name=None):\n",
    "    make_plot(gene_data, gene_info['symbol'], pert_name)\n",
    "    l1000_rnaseq_directional_gene_search(\n",
    "        gene=gene, gene_data=gene_data, pert=pert, pert_full_name=pert_full_name, pert_name=pert_name, is_upreg=True,\n",
    "    )\n",
    "    l1000_rnaseq_directional_gene_search(\n",
    "        gene=gene, gene_data=gene_data, pert=pert, pert_full_name=pert_full_name, pert_name=pert_name, is_upreg=False,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7767f98a",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions(f\"No information for gene with identifier {gene_info['symbol']} found in L1000\"):\n",
    "  l1000_rnaseq_gene_search(\n",
    "    gene=gene_info['symbol'],\n",
    "    gene_data=pd.read_feather(f\"{root_path}/gene_files/{gene_info['symbol']}.f\").set_index('index'),\n",
    "    pert='xpr',\n",
    "    pert_name='CRISPR',\n",
    "    pert_full_name='CRISPR KO',\n",
    "  )\n",
    "  l1000_rnaseq_gene_search(\n",
    "    gene=gene_info['symbol'],\n",
    "    gene_data=pd.read_feather(f\"{root_path}/cp_gene_files/{gene_info['symbol']}.f\").set_index('index'),\n",
    "    pert='cp',\n",
    "    pert_name='chemical',\n",
    "    pert_full_name='Chemical Perturbation',\n",
    "  )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a85f18",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "### International Mouse Phenotyping Consortium (IMPC)\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/IMPC.svg') }}\" />\n",
    "\n",
    "https://www.mousephenotype.org/\n",
    "\n",
    "IMPC contains serves mouse phenotype information associated with gene markers. Its [API is described here](https://www.mousephenotype.org/help/programmatic-data-access/) and allows us to identify phenotypes significantly associated with a gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a769606",
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache()\n",
    "def impc_phenotype(marker_gene, rows=20):\n",
    "    res = requests.get(\n",
    "        f\"https://www.ebi.ac.uk/mi/impc/solr/genotype-phenotype/select\",\n",
    "        params=dict(\n",
    "            q=f\"marker_symbol:{marker_gene}\",\n",
    "            rows=rows,\n",
    "        ),\n",
    "    )\n",
    "    response = res.json()['response']\n",
    "    if response['numFound'] == 0: raise Exception('No phenotypes found')\n",
    "    return pd.DataFrame(response['docs']).sort_values('p_value', ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5c38ee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions(f\"No information for gene with identifier {gene_info['symbol']} found in IPMC\"):\n",
    "    impc_results = impc_phenotype(gene_info['symbol'].capitalize())\n",
    "    show(impc_results[[\n",
    "        'marker_accession_id',\n",
    "        'mp_term_id',\n",
    "        'mp_term_name',\n",
    "        'assertion_type',\n",
    "        'p_value',\n",
    "        'phenotyping_center',\n",
    "        'percentage_change',\n",
    "        'statistical_method',\n",
    "    ]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dba1820e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions('IPMC Results could not be processed'):\n",
    "    impc_combined_stouffer_statistic = impc_results.groupby('mp_term_name')['p_value'] \\\n",
    "        .agg([\n",
    "            ('combined_stouffer_statistic', partial(combine_pvalues, method='stouffer', select='pvalue')),\n",
    "        ]) \\\n",
    "        .reset_index() \\\n",
    "        .sort_values('combined_stouffer_statistic', ascending=False)\n",
    "    impc_combined_stouffer_statistic['-logp(combined_stouffer_statistic)'] = -np.log10(impc_combined_stouffer_statistic['combined_stouffer_statistic'])\n",
    "\n",
    "    fig = px.bar(\n",
    "        impc_combined_stouffer_statistic,\n",
    "        x='-logp(combined_stouffer_statistic)',\n",
    "        y='mp_term_name',\n",
    "        text='mp_term_name',\n",
    "        orientation='h',\n",
    "        title=f\"Phenotype known to be associated with {gene_info['symbol']} from IMPC\",\n",
    "    )\n",
    "    fig.update_yaxes(showticklabels=False)\n",
    "    fig.update_traces(texttemplate='%{text}', textposition='auto', insidetextanchor='start')\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fef6b1ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "### GlyGen\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px; padding: 5px; background-color: #2f78b7\" src=\"{{ url_for('static', path='logos/glygen.svg') }}\" />\n",
    "\n",
    "<https://www.glygen.org/>\n",
    "\n",
    "GlyGen collects extensive protein product information related to Glycans and permits accessing that information over [their API](https://api.glygen.org/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de66baf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache()\n",
    "def glygen_geneNameSearch(recommended_gene_name, organism_taxon_id=9606):\n",
    "    res = requests.get(\n",
    "        'https://api.glygen.org/directsearch/gene/',\n",
    "        params=dict(\n",
    "            query=json.dumps(dict(\n",
    "                recommended_gene_name=recommended_gene_name,\n",
    "                organism=dict(\n",
    "                    id=organism_taxon_id\n",
    "                ),\n",
    "            )),\n",
    "        ),\n",
    "        verify=False, # not sure why on my system I get SSL errors\n",
    "    )\n",
    "    return res.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c69d960e",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions(f\"No information for gene with identifier {gene_info['symbol']} found in GlyGen\"):\n",
    "    glygen_geneInfo = glygen_geneNameSearch(gene_info['symbol'])\n",
    "    display(RenderJSON(glygen_geneInfo))\n",
    "    d = pd.DataFrame(glygen_geneInfo['results'][0]['glycosylation'])\n",
    "    d['evidence'] = d['evidence'].apply(\n",
    "        lambda evidence: ' '.join(f\"<a href='{e['url']}'>{e['url']}></a>\" for e in evidence if 'url' in e)\n",
    "    )\n",
    "    show(d, order=[[d.columns.get_loc('residue'), 'asc']])\n",
    "    display(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "418ef5cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "### exRNA\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/exRNA.png') }}\" />\n",
    "\n",
    "<https://ldh.clinicalgenome.org/ldh/ui/>\n",
    "\n",
    "The exRNA Linked Data Hub (LDH) facilitates efficient access to collated information such as links and select data from different data sources, which are made available using RESTful APIs. Currently, LDH focuses on linking information about human genes and variants to support exRNA curation efforts.\n",
    "\n",
    "{% if input.raw_value == \"Gene\" %}\n",
    "We provide the gene symbol to exRNA and obtain the reported linked data. The query will produce a document with all associated regulatory element in the +/- 10kb range or overlapping the gene.\n",
    "{% elif input.raw_value == \"Variant\" %}\n",
    "We provide the variant to exRNA and obtain the reported linked data.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "577cf4af",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input.raw_value == \"Gene\" %}\n",
    "@lru_cache()\n",
    "def ldh_gene_lookup(id):\n",
    "    req = requests.get(\n",
    "        f\"https://genboree.org/cfde-gene/Gene/id/{urllib.parse.quote(id)}\",\n",
    "        headers={\n",
    "            'Accept': 'application/json',\n",
    "        },\n",
    "    )\n",
    "    assert req.status_code == 200, f\"Gene with identifier {id} currently not available in exRNA LDH\"\n",
    "    results = req.json()\n",
    "    return results\n",
    "{% elif input.raw_value == \"Variant\" %}\n",
    "@lru_cache()\n",
    "def ldh_variant_lookup(id):\n",
    "    req = requests.get(\n",
    "        f\"https://ldh.clinicalgenome.org/ldh/id/{urllib.parse.quote(id)}\",\n",
    "        headers={\n",
    "            'Accept': 'application/json',\n",
    "        },\n",
    "    )\n",
    "    assert req.status_code == 200, f\"Variant with identifier {id} currently not available in exRNA LDH\"\n",
    "    results = req.json()\n",
    "    return results\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bec45c63",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input.raw_value == \"Gene\" %}\n",
    "with ignore_exceptions(f\"Gene with identifier {gene_info['symbol']} not available in exRNA LDH\"):\n",
    "    ldh_results = ldh_gene_lookup(gene_info['symbol'])\n",
    "    display(Markdown(f\"[View On LDH](https://genboree.org/cfde-gene-dev/ui/id/{urllib.parse.quote(gene_info['symbol'])})\"))\n",
    "    display(RenderJSON(ldh_results))\n",
    "{% elif input.raw_value == \"Variant\" %}\n",
    "with ignore_exceptions(f\"Variant with identifier {variant_input} not available in exRNA LDH\"):\n",
    "    ldh_results = ldh_variant_lookup(variant_input)\n",
    "    display(Markdown(f\"[View On LDH](https://ldh.clinicalgenome.org/ldh/ui/id/{urllib.parse.quote(variant_input)})\"))\n",
    "    display(RenderJSON(ldh_results))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73405715",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "### HuBMAP\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/HuBMAP.png') }}\" />\n",
    "\n",
    "<https://hubmapconsortium.org/>\n",
    "\n",
    "The goal of the Human BioMolecular Atlas Program (HuBMAP) is to develop an open and global platform to map healthy cells in the human body.\n",
    "\n",
    "The HuBMAP ASCT+B Data was processed and is served by [Enrichr](https://maayanlab.cloud/Enrichr). This data can be used to associate genes with cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd21f2ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions(f\"No information for gene with identifier {gene_info['symbol']} found in HuBMAP ASCT+B\"):\n",
    "  req = requests.get('https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName=HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression')\n",
    "  HuBMAP_ASCT_plus_B_gmt = pd.DataFrame({\n",
    "    term: {\n",
    "      gene: True\n",
    "      for gene in geneset\n",
    "      if gene\n",
    "    }\n",
    "    for line in req.text.splitlines()\n",
    "    for (term, _, *geneset) in (line.strip().split(\"\\t\"),)\n",
    "  })\n",
    "  show(HuBMAP_ASCT_plus_B_gmt.loc[gene_info['symbol']].dropna().reset_index().rename({ 'index': 'Term' }, axis=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "951b15f7",
   "metadata": {},
   "source": [
    "### Metabolomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33f8ab4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "display(Markdown(f\"\"\"\n",
    "<div style=\"display: flex; align-items: flex-start\">\n",
    "    <img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/Metabolomics.png') }}\" />\n",
    "    <div>\n",
    "        <p><a href=\"https://metabolomicsworkbench.org/\">https://metabolomicsworkbench.org/</a></p>\n",
    "        <p>The National Institutes of Health (NIH) Common Fund Metabolomics Program was developed with the goal of increasing national capacity in metabolomics by supporting the development of next generation technologies, providing training and mentoring opportunities, increasing the inventory and availability of high quality reference standards, and promoting data sharing and collaboration.</p>\n",
    "        <p><a href=\"https://sc-cfdewebdev.sdsc.edu/MetGENE/index.php?GeneID={geneId}&species=hsa&GeneIDType=ENTREZID\">MetGENE</a> identifies the pathways and reactions catalyzed by the given gene {gene_info['symbol']}, its related metabolites and the studies in Metabolomics Workbench with data on such metabolites.</a></p>\n",
    "    </div>\n",
    "</div>\n",
    "\"\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ead760b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "with ignore_exceptions(f\"No information for gene with identifier {gene_info['symbol']} found in MetGene\"):\n",
    "    metgene_table = requests.get(f\"https://sc-cfdewebdev.sdsc.edu/MetGENE/mgSummary.php?species=hsa&GeneSym={gene_info['symbol']}&GeneID={geneId}&viewType=json\").json()\n",
    "\n",
    "    display(HTML(f\"\"\"\n",
    "    <div style=\"display: flex; align-items: flex-start\">\n",
    "        <table>\n",
    "            <thead>\n",
    "                <tr>\n",
    "                <th>&nbsp;</th>\n",
    "                <th><a href=\"https://sc-cfdewebdev.sdsc.edu/MetGENE/pathways.php?species=hsa&GeneIDType=ENTREZID&anatomy=NA&disease=NA&phenotype=NA&GeneInfoStr={geneId}\">Pathways</a></th>\n",
    "                <th><a href=\"https://sc-cfdewebdev.sdsc.edu/MetGENE/reactions.php?species=hsa&GeneIDType=ENTREZID&anatomy=NA&disease=NA&phenotype=NA&GeneInfoStr={geneId}\">Reactions</a></th>\n",
    "                <th><a href=\"https://sc-cfdewebdev.sdsc.edu/MetGENE/metabolites.php?species=hsa&GeneIDType=ENTREZID&anatomy=NA&disease=NA&phenotype=NA&GeneInfoStr={geneId}\">Metabolites</a></th>\n",
    "                <th><a href=\"https://sc-cfdewebdev.sdsc.edu/MetGENE/studies.php?species=hsa&GeneIDType=ENTREZID&anatomy=NA&disease=NA&phenotype=NA&GeneInfoStr={geneId}\">Studies</a></th>\n",
    "                </tr>\n",
    "            </thead>\n",
    "            <tbody>{''.join(f\"<tr><td>{r['Genes']}</td><td>{r['Pathways']}</td><td>{r['Reactions']}</td><td>{r['Metabolites']}</td><td>{r['Studies']}</td></tr>\" for r in metgene_table)}</tbody>\n",
    "        </table>\n",
    "    <div style=\"display: flex; flex-wrap: wrap;\">\n",
    "        <img\n",
    "            style=\"width: 100%;\"\n",
    "            src=\"https://sc-cfdewebdev.sdsc.edu/MetGENE/mgSummary.php?species=hsa&GeneSym={gene_info['symbol']}&GeneID={geneId}&viewType=png\"\n",
    "            alt=\"MetGene Summary\"\n",
    "        />\n",
    "    </div>\n",
    "    </div>\n",
    "    \"\"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb1a06c5",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Secondary Information\n",
    "\n",
    "Each DCC has assembled a large repository of knowledge besides the data directly collected by the data generation centers they coordinate. We can access this expanded knowledge as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b120e66a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "### IDG\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 500px\" src=\"{{ url_for('static', path='logos/IDG.png') }}\" />\n",
    "\n",
    "<https://druggablegenome.net/>\n",
    "\n",
    "#### Pharos\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/pharos.svg') }}\" />\n",
    "\n",
    "We query IDG's knowledge base of targets and their Disease associations through the [Pharos API](https://pharos.nih.gov/api)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1417fb80",
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache()\n",
    "def idg_pharos_geneDiseaseAssociations(associatedTarget):\n",
    "    req = requests.post(\n",
    "        'https://pharos-api.ncats.io/graphql',\n",
    "        headers={\n",
    "            'Content-Type': 'application/json',\n",
    "            'Accept': 'application/json',\n",
    "        },\n",
    "        json={\n",
    "            'query': dedent(f'''\n",
    "            query associatedDiseases {{\n",
    "                diseases(filter: {{associatedTarget: {json.dumps(associatedTarget)} }}) {{\n",
    "                    diseases {{\n",
    "                        associations {{\n",
    "                            name\n",
    "                            evidence\n",
    "                            pvalue\n",
    "                            zscore\n",
    "                        }}\n",
    "                    }}\n",
    "                }}\n",
    "            }}\n",
    "            ''')\n",
    "        },\n",
    "    )\n",
    "    results = req.json()['data']\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc0f92de",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions(f\"No information for gene with identifier {gene_info['symbol']} found in Pharos\"):\n",
    "    idg_pharos_api_results = idg_pharos_geneDiseaseAssociations(gene_info['symbol'])\n",
    "\n",
    "    idg_pharos_results = pd.DataFrame(\n",
    "        association\n",
    "        for disease in idg_pharos_api_results['diseases']['diseases']\n",
    "        for association in disease['associations']\n",
    "    )\n",
    "\n",
    "    def compute_score(r):\n",
    "        if pd.notnull(r['pvalue']):\n",
    "            return r['pvalue']\n",
    "        elif pd.notnull(r['zscore']):\n",
    "            return 1-st.distributions.norm.cdf(r['zscore'])\n",
    "        else:\n",
    "            # we give na's 0.5 pvalue to show up in the plot albiet, insignificantly\n",
    "            return 0.5\n",
    "\n",
    "    # convert zscores to pvalues if present instead of pvalue\n",
    "    idg_pharos_results['pvalue'] = idg_pharos_results.apply(compute_score, axis=1)\n",
    "    show(idg_pharos_results, order=[[idg_pharos_results.columns.get_loc('pvalue'), 'asc']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a8c829f",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions('Pharos results could not be processed'):\n",
    "    idg_pharos_combined_stouffer_statistic = idg_pharos_results.groupby('name')['pvalue'] \\\n",
    "        .agg([\n",
    "            ('combined_stouffer_statistic', partial(combine_pvalues, method='stouffer', select='pvalue')),\n",
    "        ]) \\\n",
    "        .reset_index() \\\n",
    "        .sort_values('combined_stouffer_statistic', ascending=False)\n",
    "    idg_pharos_combined_stouffer_statistic['-logp(combined_stouffer_statistic)'] = -np.log10(idg_pharos_combined_stouffer_statistic['combined_stouffer_statistic'])\n",
    "\n",
    "    fig = px.bar(\n",
    "        idg_pharos_combined_stouffer_statistic,\n",
    "        x='-logp(combined_stouffer_statistic)',\n",
    "        y='name',\n",
    "        text='name',\n",
    "        orientation='h',\n",
    "        title=f\"Disease known to be associated with {gene_info['symbol']} from IDG's Pharos\",\n",
    "    )\n",
    "    fig.update_yaxes(showticklabels=False)\n",
    "    fig.update_traces(texttemplate='%{text}', textposition='auto', insidetextanchor='start')\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddd64978",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "#### Harmonizome\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 150px\" src=\"{{ url_for('static', path='logos/harmonizome.png') }}\" />\n",
    "\n",
    "We query the [Harmonizome API](https://maayanlab.cloud/Harmonizome/documentation) for associations with various biological entities in a standardized set of numerous omics datasets, as detailed [here](https://maayanlab.cloud/Harmonizome/about)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43b411d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache()\n",
    "def idg_harmonizome_geneInfo(gene, showAssociations=True, version='1.0'):\n",
    "    res = requests.get(\n",
    "        f\"https://maayanlab.cloud/Harmonizome/api/{urllib.parse.quote(version)}/gene/{urllib.parse.quote(gene_info['symbol'])}\",\n",
    "        params=dict(\n",
    "            showAssociations=json.dumps(showAssociations),\n",
    "        ),\n",
    "    )\n",
    "    return res.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c125b72",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions(f\"No information for gene with identifier {gene_info['symbol']} found in Harmonizome\"):\n",
    "    idg_harmonizome_api_geneInfo = idg_harmonizome_geneInfo(gene_info['symbol'])\n",
    "    display(RenderJSON(idg_harmonizome_api_geneInfo))\n",
    "    idg_harmonizome_geneAssociations = pd.DataFrame([\n",
    "        dict(\n",
    "            **geneAssociation['geneSet'],\n",
    "            **geneAssociation,\n",
    "        )\n",
    "        for geneAssociation in idg_harmonizome_api_geneInfo['associations']\n",
    "    ]).drop(['geneSet', 'thresholdValue', 'href'], axis=1).dropna()\n",
    "    show(\n",
    "        idg_harmonizome_geneAssociations,\n",
    "        order=[[idg_harmonizome_geneAssociations.columns.get_loc('standardizedValue'), 'desc']]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97e40aa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions('Harmonizome results could not be procesed'):\n",
    "    idg_harmonizome_geneAssociations['direction'] = idg_harmonizome_geneAssociations['standardizedValue'].apply(\n",
    "        lambda v: 'up' if v > 0 else 'down'\n",
    "    )\n",
    "    idg_harmonizome_geneAssociations['absoluteZscore'] = np.abs(idg_harmonizome_geneAssociations['standardizedValue'])\n",
    "    idg_harmonizome_geneAssociations = idg_harmonizome_geneAssociations.sort_values('absoluteZscore')\n",
    "    idg_harmonizome_geneAssociations_top_10_up = idg_harmonizome_geneAssociations[idg_harmonizome_geneAssociations['direction'] == 'up'].iloc[-10:]\n",
    "    idg_harmonizome_geneAssociations_top_10_down = idg_harmonizome_geneAssociations[idg_harmonizome_geneAssociations['direction'] == 'down'].iloc[-10:]\n",
    "\n",
    "    fig = px.bar(\n",
    "        pd.concat([\n",
    "            idg_harmonizome_geneAssociations_top_10_up,\n",
    "            idg_harmonizome_geneAssociations_top_10_down,\n",
    "        ], axis=0),\n",
    "        x='absoluteZscore',\n",
    "        y='name',\n",
    "        orientation='h',\n",
    "        barmode='group',\n",
    "        color='direction',\n",
    "        facet_row='direction',\n",
    "        text='name',\n",
    "        title=f\"Significant associations with {gene_info['symbol']} in IDG's Harmonizome\",\n",
    "        height=1000,\n",
    "        width=1000,\n",
    "    )\n",
    "    fig.update_yaxes(matches=None, showticklabels=False)\n",
    "    fig.update_traces(texttemplate='%{text}', textposition='inside', insidetextanchor='start')\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a4ef3c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "### ARCHS4\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 250px\" src=\"{{ url_for('static', path='logos/archs4.png') }}\" />\n",
    "\n",
    "https://maayanlab.cloud/archs4/\n",
    "\n",
    "ARCHS4 has processed numerous GEO studies and also has Tissue expression data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "215e9631",
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache()\n",
    "def archs4_tissue_expression(search, species='human'):\n",
    "    res = requests.get(\n",
    "        f\"https://maayanlab.cloud/archs4/search/loadExpressionTissue.php\",\n",
    "        params=dict(\n",
    "            search=search,\n",
    "            species='human',\n",
    "            type='tissue',\n",
    "        ),\n",
    "    )\n",
    "    df = pd.read_csv(io.StringIO(res.text)).dropna()\n",
    "    df.index = pd.MultiIndex.from_tuples(\n",
    "        df['id'].apply(lambda id: id.split('.')),\n",
    "        names=['', 'system', 'organ', 'tissue'],\n",
    "    )\n",
    "    return df.reset_index().drop(['id', ''], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a0abdd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions(f\"No information for gene with identifier {gene_info['symbol']} found in ARCHS4\"):\n",
    "    archs4_tissue_results = archs4_tissue_expression(gene_info['symbol'])\n",
    "    show(archs4_tissue_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53c06b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "### UnitProt\n",
    "\n",
    "<img class=\"img-fluid\" style=\"width: 200px; padding: 5px; background-color: #5cadcb\" src=\"{{ url_for('static', path='logos/uniprot.png') }}\" />\n",
    "\n",
    "https://www.uniprot.org/\n",
    "\n",
    "UniProt is a comprehensive database on protein function information. Their Proteins REST API, [documented here](https://www.ebi.ac.uk/proteins/api/doc/), can be used for gene-centric queries.\n",
    "\n",
    "https://www.ebi.ac.uk/proteins/api/genecentric?offset=0&size=100&gene=STAT3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fa61572",
   "metadata": {},
   "outputs": [],
   "source": [
    "@lru_cache()\n",
    "def uniprot_genecentric(gene, offset=0, size=100):\n",
    "    req = requests.get(\n",
    "        f\"https://www.ebi.ac.uk/proteins/api/genecentric\",\n",
    "        params=dict(\n",
    "            gene=gene,\n",
    "            offset=offset,\n",
    "            size=size,\n",
    "        )\n",
    "    )\n",
    "    return req.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f657e63",
   "metadata": {},
   "outputs": [],
   "source": [
    "with ignore_exceptions(f\"No information for gene with identifier {gene_info['symbol']} found in UniProt\"):\n",
    "    uniprot_geneinfo = uniprot_genecentric(gene_info['symbol'])\n",
    "    show(pd.DataFrame([\n",
    "        dict(related_record, subject=record['gene']['accession'])\n",
    "        for record in uniprot_geneinfo\n",
    "        for related_record in [record['gene'], *record.get('relatedGene',[])]\n",
    "    ]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
