{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA-seq analysis pipeline with enrichment analysis from Enrichr \n",
    "---\n",
    "\n",
    "By default, we use TCGA data. Users can optionally also upload their own datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import requests\n",
    "import time\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "from umap import UMAP\n",
    "from scipy.stats import zscore\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn import metrics\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_score\n",
    "from maayanlab_bioinformatics.dge import characteristic_direction\n",
    "from maayanlab_bioinformatics.normalization import log2_normalize, filter_by_var, zscore_normalize\n",
    "from maayanlab_bioinformatics.utils import merge\n",
    "import plotly.express as px\n",
    "import math\n",
    "from collections import OrderedDict\n",
    "import json\n",
    "from jupyter_d3 import scatter_plot\n",
    "from IPython.display import display, IFrame, Markdown, HTML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name=\"DATASET\",\n",
    "    title=\"Dataset selection\",\n",
    "    subtitle='If using TCGA data, leave both file upload fields blank and simply select the desired cancer type.',\n",
    "    img = \"tcga-logo.png\"\n",
    ") %}\n",
    "\n",
    "\n",
    "{% do SectionField(\n",
    "    name='CONFIG',\n",
    "    title='Analysis parameters',\n",
    "    img = \"cluster-icon.svg\"\n",
    "\n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name=\"ENRICHR_LIBS\",\n",
    "    title=\"Libraries to include in the Enrichr search\",\n",
    "    img = \"enrichr-logo.png\"\n",
    "\n",
    "\n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name=\"L1000FWD\",\n",
    "    title=\"L1000FWD search parameters\",\n",
    "    img = \"l1000fwd.png\"\n",
    ") %}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "data_filename = \"{{ FileField(name='data_filename', label='RNA-seq data file (.csv)', description='Upload RNA-seq dataset in csv format. The index of the dataset are genes, the columns are samples.', section='DATASET').value }}\"\n",
    "\n",
    "clinical_data_filename = \"{{ FileField(name='clinical_data_filename', label='Clinical data file (.csv)', description='Upload clinical data in csv format. The first column should contain patient IDs corresponding to those in the RNA-seq file. Each subsequent column is a clinical data field.', section='DATASET').value }}\"\n",
    "\n",
    "cancer = '''{{ ChoiceField(\n",
    "        name = \"cancer\",\n",
    "        label = \"Cancer type\",\n",
    "        description=\"The value provided as the primary diagnosis on cases in TCGA.\",\n",
    "        choices=[\"Infiltrating duct carcinoma, NOS\",\"Squamous cell carcinoma, NOS\",\"Lobular carcinoma, NOS\",\"Acute myeloid leukemia, NOS\",\"Hepatocellular carcinoma, NOS\",\"Serous cystadenocarcinoma, NOS\",\"Endometrioid adenocarcinoma, NOS\",\"Adenocarcinoma, NOS\",\"Clear cell adenocarcinoma, NOS\",\"Glioblastoma\",\"Mucinous adenocarcinoma\",\"Transitional cell carcinoma\",\"Malignant melanoma, NOS\",\"Papillary adenocarcinoma, NOS\"],\n",
    "        section=\"DATASET\",\n",
    "        default=\"Serous cystadenocarcinoma, NOS\" ) \n",
    "}}'''\n",
    "    \n",
    "# Dimensionality reduction parameters\n",
    "\n",
    "n_pca_components = {{IntField(\n",
    "    name='n_pca_components',\n",
    "    label='Number of PCA components to use for UMAP',\n",
    "    description='The number of \\'top\\' PCA components to use for further dimensionality reduction using the UMAP algorithm',\n",
    "    default=10,\n",
    "    min=0,\n",
    "    max=100,\n",
    "    section='CONFIG',\n",
    ")}}\n",
    "\n",
    "n_neighbors = {{IntField(\n",
    "    name='n_neighbors',\n",
    "    label='Number of neighbors for each projected datapoint computed by UMAP',\n",
    "    description='Smaller values preserve local manifold structure in the dataset as opposed to overall global structure',\n",
    "    default=2,\n",
    "    min=2,\n",
    "    max=20,\n",
    "    section='CONFIG',\n",
    ")}}\n",
    "\n",
    "min_cluster_dist = {{FloatField(\n",
    "    name='min_cluster_dist',\n",
    "    label='Minimum distance between UMAP-projected points',\n",
    "    description='How tightly packed points produced by dimensionality reduction with UMAP are permitted to be.',\n",
    "    default=0.3,\n",
    "    min=0.1,\n",
    "    max=1,\n",
    "    section='CONFIG',\n",
    ")}}\n",
    "\n",
    "\n",
    "# The number of 'top' genes to use for differential expression\n",
    "top_n_genes = {{IntField(\n",
    "    name='top_n_genes',\n",
    "    label='Number of Genes',\n",
    "    description='The number of \\'top\\' genes to use for differential expression',\n",
    "    default=250,\n",
    "    min=100,\n",
    "    max=1000,\n",
    "    section='CONFIG',\n",
    ")}}\n",
    "\n",
    "# The number of 'top' results to keep from enrichment analysis\n",
    "top_n_results = {{ IntField(\n",
    "    name='top_n_results',\n",
    "    label='Number of Top Enrichment Results',\n",
    "    description='The number of \\'top\\' results to keep from enrichment analysis',\n",
    "    default=5,\n",
    "    min=1,\n",
    "    max=100,\n",
    "    section='CONFIG',\n",
    ")}}\n",
    "\n",
    "transcription_libraries = {{ MultiChoiceField(name='transcription_libraries', \n",
    "                                              description='Select the Enrichr libraries you would like in your figure.', \n",
    "                                              label='Transcription', \n",
    "                                              default=[], \n",
    "                                              section = 'ENRICHR_LIBS',\n",
    "                                              choices=[\n",
    "                                                    'ARCHS4_TFs_Coexp',\n",
    "                                                    'ChEA_2016',\n",
    "                                                    'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "                                                    'ENCODE_Histone_Modifications_2015',\n",
    "                                                    'ENCODE_TF_ChIP-seq_2015',\n",
    "                                                    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "                                                    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "                                                    'Genome_Browser_PWMs',\n",
    "                                                    'lncHUB_lncRNA_Co-Expression',\n",
    "                                                    'miRTarBase_2017',\n",
    "                                                    'TargetScan_microRNA_2017',\n",
    "                                                    'TF-LOF_Expression_from_GEO',\n",
    "                                                    'TF_Perturbations_Followed_by_Expression',\n",
    "                                                    'Transcription_Factor_PPIs',\n",
    "                                                    'TRANSFAC_and_JASPAR_PWMs',\n",
    "                                                    'TRRUST_Transcription_Factors_2019'\n",
    "                                                ]) }}\n",
    "\n",
    "pathways_libraries = {{ MultiChoiceField(name='pathways_libraries', \n",
    "                                         description='Select the Enrichr libraries you would like in your figure.', \n",
    "                                         label='Pathways', \n",
    "                                         default=[], \n",
    "                                         section='ENRICHR_LIBS',\n",
    "                                         choices=[\n",
    "                                                    'ARCHS4_Kinases_Coexp',\n",
    "                                                    'BioCarta_2016',\n",
    "                                                    'BioPlanet_2019',\n",
    "                                                    'BioPlex_2017',\n",
    "                                                    'CORUM',\n",
    "                                                    'Elsevier_Pathway_Collection',\n",
    "                                                    'HMS_LINCS_KinomeScan',\n",
    "                                                    'HumanCyc_2016',\n",
    "                                                    'huMAP',\n",
    "                                                    'KEA_2015',\n",
    "                                                    'KEGG_2019_Human',\n",
    "                                                    'KEGG_2019_Mouse',\n",
    "                                                    'Kinase_Perturbations_from_GEO_down',\n",
    "                                                    'Kinase_Perturbations_from_GEO_up',\n",
    "                                                    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "                                                    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "                                                    'NCI-Nature_2016',\n",
    "                                                    'NURSA_Human_Endogenous_Complexome',\n",
    "                                                    'Panther_2016',\n",
    "                                                    'Phosphatase_Substrates_from_DEPOD',\n",
    "                                                    'PPI_Hub_Proteins',\n",
    "                                                    'Reactome_2016',\n",
    "                                                    'SILAC_Phosphoproteomics',\n",
    "                                                    'SubCell_BarCode',\n",
    "                                                    'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "                                                    'WikiPathways_2019_Human',\n",
    "                                                    'WikiPathways_2019_Mouse']) }} \n",
    "\n",
    "ontologies_libraries = {{ MultiChoiceField(name='ontologies_libraries', \n",
    "                                           description='Select the Enrichr libraries you would like in your figure.', \n",
    "                                           label='Ontologies', \n",
    "                                           default=[], \n",
    "                                           section = 'ENRICHR_LIBS',\n",
    "                                           choices=[\n",
    "                                                'GO_Biological_Process_2018',\n",
    "                                                'GO_Cellular_Component_2018',\n",
    "                                                'GO_Molecular_Function_2018',\n",
    "                                                'Human_Phenotype_Ontology',\n",
    "                                                'Jensen_COMPARTMENTS',\n",
    "                                                'Jensen_DISEASES',\n",
    "                                                'Jensen_TISSUES',\n",
    "                                                'MGI_Mammalian_Phenotype_Level_4_2019'\n",
    "                                           ]) }}\n",
    "    \n",
    "diseases_drugs_libraries = {{ MultiChoiceField(name='diseases_drugs_libraries', \n",
    "                                               description='Select the Enrichr libraries you would like in your figure.', \n",
    "                                               label='Diseases/Drugs', \n",
    "                                               default=[], \n",
    "                                               section = 'ENRICHR_LIBS',\n",
    "                                               choices=[    \n",
    "                                                    'Achilles_fitness_decrease',\n",
    "                                                    'Achilles_fitness_increase',\n",
    "                                                    'ARCHS4_IDG_Coexp',\n",
    "                                                    'ClinVar_2019',\n",
    "                                                    'dbGaP',\n",
    "                                                    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "                                                    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "                                                    'DisGeNET',\n",
    "                                                    'DrugMatrix',\n",
    "                                                    'DSigDB',\n",
    "                                                    'GeneSigDB',\n",
    "                                                    'GWAS_Catalog_2019',\n",
    "                                                    'LINCS_L1000_Chem_Pert_down',\n",
    "                                                    'LINCS_L1000_Chem_Pert_up',\n",
    "                                                    'LINCS_L1000_Ligand_Perturbations_down',\n",
    "                                                    'LINCS_L1000_Ligand_Perturbations_up',\n",
    "                                                    'MSigDB_Computational',\n",
    "                                                    'MSigDB_Oncogenic_Signatures',\n",
    "                                                    'Old_CMAP_down',\n",
    "                                                    'Old_CMAP_up',\n",
    "                                                    'OMIM_Disease',\n",
    "                                                    'OMIM_Expanded',\n",
    "                                                    'PheWeb_2019',\n",
    "                                                    'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "                                                    'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "                                                    'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "                                                    'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "                                                    'UK_Biobank_GWAS_v1',\n",
    "                                                    'Virus_Perturbations_from_GEO_down',\n",
    "                                                    'Virus_Perturbations_from_GEO_up',\n",
    "                                                    'VirusMINT']) }}\n",
    "    \n",
    "cell_types_libraries = {{ MultiChoiceField(name='cell_types_libraries', \n",
    "                                           description='Select the Enrichr libraries you would like in your figure.', \n",
    "                                           label='Cell Types', \n",
    "                                           default=[], \n",
    "                                           section = 'ENRICHR_LIBS',\n",
    "                                           choices=[        \n",
    "                                                'Allen_Brain_Atlas_down',\n",
    "                                                'Allen_Brain_Atlas_up',\n",
    "                                                'ARCHS4_Cell-lines',\n",
    "                                                'ARCHS4_Tissues',\n",
    "                                                'Cancer_Cell_Line_Encyclopedia',\n",
    "                                                'CCLE_Proteomics_2020',\n",
    "                                                'ESCAPE',\n",
    "                                                'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "                                                'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "                                                'Human_Gene_Atlas',\n",
    "                                                'Mouse_Gene_Atlas',\n",
    "                                                'NCI-60_Cancer_Cell_Lines',\n",
    "                                                'ProteomicsDB_2020',\n",
    "                                                'Tissue_Protein_Expression_from_Human_Proteome_Map']) }}    \n",
    "    \n",
    "    \n",
    "    \n",
    "miscellaneous_libraries = {{ MultiChoiceField(name='miscellaneous_libraries', \n",
    "                                              description='Select the Enrichr libraries you would like in your figure.', \n",
    "                                              label='Miscellaneous', \n",
    "                                              default=[], \n",
    "                                              section = 'ENRICHR_LIBS',\n",
    "                                              choices=[            \n",
    "                                                    'Chromosome_Location_hg19',\n",
    "                                                    'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "                                                    'Enrichr_Libraries_Most_Popular_Genes',\n",
    "                                                    'Genes_Associated_with_NIH_Grants',\n",
    "                                                    'HMDB_Metabolites',\n",
    "                                                    'HomoloGene',\n",
    "                                                    'InterPro_Domains_2019',\n",
    "                                                    'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "                                                    'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "                                                    'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "                                                    'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "                                                    'Pfam_Domains_2019',\n",
    "                                                    'Pfam_InterPro_Domains',\n",
    "                                                    'Table_Mining_of_CRISPR_Studies']) }}    \n",
    "    \n",
    "    \n",
    "legacy_libraries = {{ MultiChoiceField(name='legacy_libraries', \n",
    "                                       description='Select the Enrichr libraries you would like in your figure.', \n",
    "                                       label='Legacy', \n",
    "                                       default=[], \n",
    "                                       section = 'ENRICHR_LIBS',\n",
    "                                       choices=[                \n",
    "                                            'BioCarta_2013',\n",
    "                                            'BioCarta_2015',\n",
    "                                            'ChEA_2013',\n",
    "                                            'ChEA_2015',\n",
    "                                            'Chromosome_Location',\n",
    "                                            'Disease_Signatures_from_GEO_down_2014',\n",
    "                                            'Disease_Signatures_from_GEO_up_2014',\n",
    "                                            'Drug_Perturbations_from_GEO_2014',\n",
    "                                            'ENCODE_Histone_Modifications_2013',\n",
    "                                            'ENCODE_TF_ChIP-seq_2014',\n",
    "                                            'GO_Biological_Process_2013',\n",
    "                                            'GO_Biological_Process_2015',\n",
    "                                            'GO_Biological_Process_2017',\n",
    "                                            'GO_Biological_Process_2017b',\n",
    "                                            'GO_Cellular_Component_2013',\n",
    "                                            'GO_Cellular_Component_2015',\n",
    "                                            'GO_Cellular_Component_2017',\n",
    "                                            'GO_Cellular_Component_2017b',\n",
    "                                            'GO_Molecular_Function_2013',\n",
    "                                            'GO_Molecular_Function_2015',\n",
    "                                            'GO_Molecular_Function_2017',\n",
    "                                            'GO_Molecular_Function_2017b',\n",
    "                                            'HumanCyc_2015',\n",
    "                                            'KEA_2013',\n",
    "                                            'KEGG_2013',\n",
    "                                            'KEGG_2015',\n",
    "                                            'KEGG_2016',\n",
    "                                            'MGI_Mammalian_Phenotype_2013',\n",
    "                                            'MGI_Mammalian_Phenotype_2017',\n",
    "                                            'MGI_Mammalian_Phenotype_Level_3',\n",
    "                                            'MGI_Mammalian_Phenotype_Level_4',\n",
    "                                            'NCI-Nature_2015',\n",
    "                                            'Panther_2015',\n",
    "                                            'Reactome_2013',\n",
    "                                            'Reactome_2015',\n",
    "                                            'TargetScan_microRNA',\n",
    "                                            'Tissue_Protein_Expression_from_ProteomicsDB',\n",
    "                                            'WikiPathways_2013',\n",
    "                                            'WikiPathways_2015',\n",
    "                                            'WikiPathways_2016']) }} \n",
    "\n",
    "crowd_libraries = {{ MultiChoiceField(name='crowd_libraries', \n",
    "                                      description='Select the Enrichr libraries you would like in your figure.', \n",
    "                                      label='Crowd', \n",
    "                                      default=[],\n",
    "                                      section = 'ENRICHR_LIBS',\n",
    "                                      choices=[                \n",
    "                                            'Aging_Perturbations_from_GEO_down',\n",
    "                                            'Aging_Perturbations_from_GEO_up',\n",
    "                                            'Disease_Perturbations_from_GEO_down',\n",
    "                                            'Disease_Perturbations_from_GEO_up',\n",
    "                                            'Drug_Perturbations_from_GEO_down',\n",
    "                                            'Drug_Perturbations_from_GEO_up',\n",
    "                                            'Gene_Perturbations_from_GEO_down',\n",
    "                                            'Gene_Perturbations_from_GEO_up',\n",
    "                                            'Ligand_Perturbations_from_GEO_down',\n",
    "                                            'Ligand_Perturbations_from_GEO_up',\n",
    "                                            'MCF7_Perturbations_from_GEO_down',\n",
    "                                            'MCF7_Perturbations_from_GEO_up',\n",
    "                                            'Microbe_Perturbations_from_GEO_down',\n",
    "                                            'Microbe_Perturbations_from_GEO_up',\n",
    "                                            'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "                                            'SysMyo_Muscle_Gene_Sets']) }}\n",
    "\n",
    "\n",
    "\n",
    "top_n_drugs = {{IntField(\n",
    "    name='top_n_drugs',\n",
    "    label='Number of top drug treatments to suggest for each cluster',\n",
    "    default=5,\n",
    "    min=0,\n",
    "    max=50,\n",
    "    section='L1000FWD',\n",
    ")}}\n",
    "\n",
    "\n",
    "enrichr_libraries = OrderedDict([\n",
    "    ('Diseases/Drugs', diseases_drugs_libraries), \n",
    "    ('Ontologies', ontologies_libraries),\n",
    "    ('Cell Type', cell_types_libraries),\n",
    "    ('Pathways', pathways_libraries),\n",
    "    ('Transcription', transcription_libraries),\n",
    "    ('Legacy', legacy_libraries),\n",
    "    ('Crowd', crowd_libraries)\n",
    "])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import dataset\n",
    "\n",
    "If using user-uploaded data, we access those files and set the necessary index names.\n",
    "\n",
    "Otherwise, we first download the data containing RNA-seq profiles for the selected cancer type from TCGA and the corresponding clinical data for those cases.\n",
    "\n",
    "The RNA-seq data that is provided has already been processed using the HTSeq Python package, which calculates the number of mapped reads to each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "# Utility function for saving the dataset\n",
    "def save_data(cancer_type,\n",
    "              df_data_norm_km,\n",
    "              df_data_norm_pca,\n",
    "              df_data_norm_umap,\n",
    "              df_diff_expr,\n",
    "              df_aucs):\n",
    "    \n",
    "    # create the root dir\n",
    "    root_dir = f'./processed_data/{cancer_type}'\n",
    "    os.makedirs(root_dir, exist_ok=True)\n",
    "\n",
    "    # /clustering/graphclust/clusters.csv\n",
    "    #   Barcode,Cluster\n",
    "    os.makedirs(f'{root_dir}/clustering/graphclust', exist_ok=True)\n",
    "    df_data_norm_km.to_csv(f'{root_dir}/clustering/graphclust/clusters.csv')\n",
    "    \n",
    "    # /pca/10_components/projection.csv\n",
    "    #   Barcode,PC-1,PC-2,PC-3,PC-4,PC-5,PC-6,PC-7,PC-8,PC-9,PC-10\n",
    "    os.makedirs(f'{root_dir}/pca/{n_pca_components}_components', exist_ok=True)\n",
    "    df_data_norm_pca.to_csv(f'{root_dir}/pca/{n_pca_components}_components/projection.csv')\n",
    "    \n",
    "    # /umap/2_components/projection.csv\n",
    "    #   Barcode,UMAP-1,UMAP-2\n",
    "    os.makedirs(f'{root_dir}/umap/2_components', exist_ok=True)\n",
    "    df_data_norm_umap.to_csv(f'{root_dir}/umap/2_components/projection.csv')\n",
    "    \n",
    "     # /diffexp/graphclust/differential_expression.csv\n",
    "    #   Feature Name,Cluster 2 Log2 fold change,Cluster 0 Log2 fold change,Cluster 1 Log2 fold change,Cluster 3 Log2 fold change,Cluster 2 Adjusted p value,Cluster 0 Adjusted p value,Cluster 1 Adjusted p value,Cluster 3 Adjusted p value,Cluster 2 Mean Counts,Cluster 0 Mean Counts,Cluster 1 Mean Counts,Cluster 3 Mean Counts,Feature Name\n",
    "    os.makedirs(f'{root_dir}/diffexp/graphclust', exist_ok=True)\n",
    "    df_diff_expr.to_csv(f'{root_dir}/diffexp/graphclust/differential_expression.csv')\n",
    "    \n",
    "    # /cluster_aucs.csv\n",
    "    #   ,2,0,1,3\n",
    "    df_aucs.to_csv(f'{root_dir}/cluster_aucs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Notebook display util functions\n",
    "\n",
    "def download_button(content, label, filename):\n",
    "    # Add download button\n",
    "    outname = filename.split('.')[0]\n",
    "    display(HTML('<textarea id=\"textbox_{outname}\" style=\"display: none;\">{content}</textarea> <button style=\"margin:10px 0;\" id=\"create_{outname}\">{label}</button> <a download=\"{filename}\" id=\"downloadlink_{outname}\" style=\"display: none\">Download</a>'.format(**locals())))\n",
    "    display(HTML('<script type=\"text/javascript\">!function(){{var e=null,t=document.getElementById(\"create_{outname}\"),n=document.getElementById(\"textbox_{outname}\");t.addEventListener(\"click\",function(){{var t,l,c=document.getElementById(\"downloadlink_{outname}\");c.href=(t=n.value,l=new Blob([t],{{type:\"text/plain\"}}),null!==e&&window.URL.revokeObjectURL(e),e=window.URL.createObjectURL(l)),c.click()}},!1)}}();</script>'.format(**locals())))\n",
    "\n",
    "def make_clickable(link):\n",
    "    return f'<a target=\"_blank\" href=\"{link}\">{link}</a>'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_data = False\n",
    "if (data_filename != \"\" and clinical_data_filename != \"\"): user_data = True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "data_index = \"symbol\"\n",
    "clinical_index = \"case_id\"\n",
    "\n",
    "if (user_data):\n",
    "    df_data = pd.read_csv(data_filename, index_col=0).sort_index()\n",
    "    df_clinical = pd.read_csv(clinical_data_filename, index_col=0).sort_index()\n",
    "\n",
    "    df_data.index.name = \"symbol\"\n",
    "    df_clinical.index.name = \"case_id\"\n",
    "    \n",
    "    display(df_data.head())\n",
    "    display(df_clinical.head())\n",
    "else:\n",
    "    # Import the RNA-seq and clinical DataFrames with only entries corresponding \n",
    "    # the selected cancer type\n",
    "    aws_url = f'https://tcga-enrichr-viewer.s3.amazonaws.com/{cancer.replace(\" \", \"%20\")}'\n",
    "\n",
    "    df_data = pd.read_csv(f'{aws_url}_data.csv')\n",
    "    df_data = df_data.set_index(\"symbol\")\n",
    "\n",
    "    df_clinical = pd.read_csv(f'{aws_url}_clinical_data.csv')\n",
    "    df_clinical = df_clinical.set_index(\"case_id\")\n",
    "    df_clinical = df_clinical.replace(\"not reported\", np.nan)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a list of the clinical features with multiple unique values\n",
    "features = df_clinical.columns.values\n",
    "\n",
    "def has_unique_values(feature):\n",
    "    unique = df_clinical[feature].dropna().unique()\n",
    "    if len(unique) >= 1: return True\n",
    "    return False\n",
    "\n",
    "if (not user_data):\n",
    "    features = list(filter(\n",
    "        lambda x: not x in [\"primary_diagnosis\",\"submitter_id\", \"updated_datetime\",\"created_datetime\",\"diagnosis_id\", \"demographic_id\"] \n",
    "        and has_unique_values(x), features))\n",
    "\n",
    "    df_clinical = df_clinical[features]\n",
    "else:\n",
    "    features = list(filter(\n",
    "        lambda x: has_unique_values(x), features))\n",
    "    df_clinical = df_clinical[features]\n",
    "\n",
    "\n",
    "    \n",
    "print(f\"{df_clinical.shape[0]} cases, {len(features)} features\")\n",
    "\n",
    "display(df_clinical.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process data and compute clusters\n",
    "\n",
    "Next, we'll need to normalize the RNA-seq data, reduce its dimensionality using PCA and UMAP, and compute clusters of RNA-seq profiles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Data size stats\n",
    "\n",
    "df_library_size = pd.DataFrame(\n",
    "    {\n",
    "        'n_reads': df_data[df_data > 0].count(),\n",
    "        'log_n_reads': np.log2(df_data[df_data > 0].count() + 1),\n",
    "        'n_expressed_genes': df_data.sum(),\n",
    "    }).sort_values('n_reads', ascending=False)\n",
    "\n",
    "df_library_size.index.name = \"case_id\"\n",
    "display(df_library_size.head()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Normalization\n",
    "\n",
    "# take 2500 top most variable rows\n",
    "df_data_norm = filter_by_var(df_data)\n",
    "\n",
    "# compute log normalization of matrix\n",
    "df_data_norm = log2_normalize(df_data_norm)\n",
    "\n",
    "# convert to zscores\n",
    "df_data_norm = zscore_normalize(df_data_norm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(df_data_norm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot 2 random gene distributions\n",
    "sns.distplot(df_data_norm.iloc[0, :]); plt.show()\n",
    "sns.distplot(df_data_norm.iloc[200, :]); plt.show()\n",
    "\n",
    "# plot a single RNA seq profile distribution\n",
    "sns.distplot(df_data_norm.iloc[:, 0]); plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's visualize hierarchical clusters of up and downregulated genes with a heatmap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ax = sns.clustermap(df_data_norm,row_cluster=False,xticklabels=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use PCA to initially reduce the dimensionality of the dataset before clustering while still maintaining most of the variability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PCA\n",
    "data_norm_pca = PCA(\n",
    "  random_state=42,\n",
    ")\n",
    "\n",
    "data_norm_pca.fit(df_data_norm.values.T)\n",
    "\n",
    "df_data_norm_pca = pd.DataFrame(\n",
    "    data_norm_pca.transform(df_data_norm.values.T),\n",
    "    index=df_data_norm.T.index\n",
    ")\n",
    "\n",
    "df_data_norm_pca.columns = [\n",
    "    f'PCA-{c}' # ({r:.3f})'\n",
    "    for c, r in zip(df_data_norm_pca.columns, data_norm_pca.explained_variance_ratio_)\n",
    "]\n",
    "\n",
    "df_data_norm_pca.index.name = \"case_id\"\n",
    "\n",
    "display(df_data_norm_pca.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the two plots below (PCA and UMAP projections), datapoints are color-coded by the selected feature. The size of each point represents the number of reads for that sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pca_data = merge(\n",
    "        df_data_norm_pca[[\"PCA-0\", \"PCA-1\"]],\n",
    "        df_library_size,\n",
    "        df_clinical\n",
    "      )\n",
    "\n",
    "axes = {\n",
    "    \"x\": [min(df_data_norm_pca[\"PCA-0\"])*1.1, max(df_data_norm_pca[\"PCA-0\"])*1.1],\n",
    "    \"y\": [min(df_data_norm_pca[\"PCA-1\"])*1.1, max(df_data_norm_pca[\"PCA-1\"])*1.1]\n",
    "}\n",
    "    \n",
    "scatter_plot(pca_data,\"PCA-0\", \"PCA-1\",axes,features)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We further reduce the dimensionality of the dataset using the UMAP (Uniform Manifold Approximation and Projection) technique on the data projected onto the first `n_pca_components` PCA components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# UMAP\n",
    "data_norm_umap = UMAP(\n",
    "  random_state=42,\n",
    "  n_components=2,\n",
    "  n_neighbors=n_neighbors,\n",
    "  metric='cosine',\n",
    "  min_dist=min_cluster_dist,\n",
    ")\n",
    "\n",
    "# use top n_pca_components components of PCA\n",
    "n_pca_components = min(n_pca_components,df_data_norm_pca.shape[1])\n",
    "data_norm_umap.fit(df_data_norm_pca.iloc[:, :n_pca_components].values)\n",
    "\n",
    "df_data_norm_umap = pd.DataFrame(\n",
    "  data_norm_umap.transform(df_data_norm_pca.iloc[:, :n_pca_components].values),\n",
    "  columns=['UMAP-0', 'UMAP-1'],\n",
    "  index=df_data_norm_pca.index,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# project data onto its first 2 UMAP components for visualization\n",
    "umap_data = merge(\n",
    "        df_data_norm_umap[[\"UMAP-0\", \"UMAP-1\"]],\n",
    "        df_library_size,\n",
    "        df_clinical\n",
    "      )\n",
    "\n",
    "axes = {\n",
    "    \"x\": [min(df_data_norm_umap[\"UMAP-0\"])*1.1, max(df_data_norm_umap[\"UMAP-0\"])*1.1],\n",
    "    \"y\": [min(df_data_norm_umap[\"UMAP-1\"])*1.1, max(df_data_norm_umap[\"UMAP-1\"])*1.1]\n",
    "}\n",
    "\n",
    "scatter_plot(umap_data,\"UMAP-0\", \"UMAP-1\",axes,features)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We continue using the first 2 UMAP components.\n",
    "\n",
    "To compute clusters, we use the Kmeans method, which requires us to define a total number of clusters. We test a range for the number of total clusters. \n",
    "\n",
    "For each number of clusters, we compute silhouette scores, which are a measure of how similar an entry is to its own cluster compared to other clusters. We want to maximize similarity within a cluster and differrences between clusters, so the ideal number of clusters is that which produces the highest silhouette score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "silhouette_scores = {}\n",
    "\n",
    "max_clusters = 25\n",
    "if (df_data_norm_umap.shape[0] < 25):\n",
    "    max_clusters = df_data_norm_umap.shape[0]\n",
    "    max_clusters = max(3,max_clusters)\n",
    "\n",
    "for n in range(2, max_clusters):\n",
    "    y_pred = KMeans(n_clusters=n, random_state=42).fit_predict(df_data_norm_umap.values)\n",
    "    silhouette_scores[n] = silhouette_score(df_data_norm_umap.values, y_pred, metric='cosine')\n",
    "silhouette_scores = pd.DataFrame([\n",
    "    {'N Clusters': k, 'Silhouette Score': v}\n",
    "    for k, v in silhouette_scores.items()\n",
    "])\n",
    "best = silhouette_scores.sort_values('Silhouette Score').iloc[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the scores as a function of # of clusters\n",
    "plt.plot(silhouette_scores['N Clusters'], silhouette_scores['Silhouette Score'])\n",
    "plt.scatter([best['N Clusters']], [best['Silhouette Score']], label='Best: {} clusters'.format(int(best[\"N Clusters\"])))\n",
    "plt.legend()\n",
    "plt.title('Cluster size selection')\n",
    "plt.ylabel('Silhouette Score')\n",
    "plt.xlabel('Number of Clusters')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Compute the Kmeans dataframe using the ideal number of clusters\n",
    "km = KMeans(n_clusters=int(best['N Clusters']), random_state=42)\n",
    "df_data_norm_km = pd.DataFrame({\n",
    "'Cluster': [\n",
    "    str(c)\n",
    "    for c in km.fit_predict(df_data_norm_umap.values)\n",
    "]}, index=df_data_norm_umap.index)\n",
    "\n",
    "print(f'Computed {len(df_data_norm_km[\"Cluster\"].unique())} clusters')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we compute differential expression for each cluster. We use the <a href=\"http://www.maayanlab.net/CD/\">Characteristic Direction method</a> for identifying differentially expressed genes among the different clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get differential expression for each cluster\n",
    "diff_expr = {}\n",
    "for cluster, samples in df_data_norm_km.groupby('Cluster'):\n",
    "    diff_expr[f\"Cluster {cluster} CD\"] = characteristic_direction(\n",
    "        # expression outside of this cluster\n",
    "        df_data_norm.loc[:, df_data_norm.columns.difference(samples.index)],\n",
    "        # expression in this cluster\n",
    "        df_data_norm.loc[:, samples.index],\n",
    "      )['CD-coefficient']\n",
    "\n",
    "df_diff_expr = pd.DataFrame(diff_expr)\n",
    "df_diff_expr = df_diff_expr.sort_values(by='Cluster 0 CD',ascending=True)\n",
    "df_diff_expr['Symbol'] = df_diff_expr.index.values\n",
    "\n",
    "display(df_diff_expr.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We fit a logistic regression on each clinical feature to see which most accurately predict the cluster each datapoint falls into. We also plot up to 5 ROC curves for features that resulted in an AUC $\\gt 0.7$. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# LR\n",
    "aucs = {}\n",
    "num_rocs_plotted = 0\n",
    "\n",
    "for cluster, samples in df_data_norm_km.groupby('Cluster'): \n",
    "    aucs[cluster] = {}\n",
    "\n",
    "    for feature in features:\n",
    "        lr = LogisticRegression()\n",
    "        X = df_clinical.copy()\n",
    "        X = X[feature]\n",
    "        X = pd.merge(X, df_data_norm_km, left_index = True, right_index = True)\n",
    "\n",
    "        # drop NAs, and move on if dataset is empty\n",
    "        X.replace(\"not reported\", None)\n",
    "        X = X.dropna()\n",
    "        if (X.shape[0] == 0): continue\n",
    "\n",
    "        cluster_data = X[\"Cluster\"]\n",
    "        X = X.drop(columns= [\"Cluster\"])\n",
    "\n",
    "        # one-hot encode non numerical data\n",
    "        if (not isinstance(X[feature][0], (int, float, complex))):\n",
    "            X = pd.get_dummies(X[feature], prefix=feature)\n",
    "\n",
    "        y_true = (cluster_data == cluster)\n",
    "\n",
    "        if (len(set(y_true)) < 2): continue # if we only have one class in the dataset\n",
    "        lr.fit(X, y_true)\n",
    "\n",
    "\n",
    "        y_score = lr.predict_proba(X)[:, 1]\n",
    "        auc_score = roc_auc_score(y_true, y_score)\n",
    "        aucs[cluster][feature] = auc_score\n",
    "        \n",
    "        # plot a few good ROCs, but no more than 5\n",
    "        if (auc_score > 0.7 and num_rocs_plotted < 6):\n",
    "            fig = metrics.plot_roc_curve(lr, X, y_true)\n",
    "            fig.figure_.suptitle(f'Predictions of cluster {cluster} by feature {feature}')\n",
    "            plt.show()\n",
    "            num_rocs_plotted += 1\n",
    "\n",
    "df_cluster_aucs = pd.DataFrame(aucs)\n",
    "df_cluster_aucs.index.name=\"Feature\"\n",
    "display(df_cluster_aucs.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichr data\n",
    "\n",
    "We query the Enrichr API for enrichment analysis of the top most upregulated and downregulated genes for each cluster calculated above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Util functions\n",
    "def enrichr_link_from_genes(genes, description='', enrichr_link='https://amp.pharm.mssm.edu/Enrichr'):\n",
    "    ''' Functional access to Enrichr API\n",
    "    '''\n",
    "    time.sleep(1)\n",
    "    resp = requests.post(enrichr_link + '/addList', files={\n",
    "    'list': (None, '\\n'.join(genes)),\n",
    "    'description': (None, description),\n",
    "    })\n",
    "    if resp.status_code != 200:\n",
    "        raise Exception('Enrichr failed with status {}: {}'.format(\n",
    "          resp.status_code,\n",
    "          resp.text,\n",
    "        ))\n",
    "    # wait a tinybit before returning link (backoff)\n",
    "    time.sleep(1)\n",
    "    result = resp.json()\n",
    "    return dict(result, link=enrichr_link + '/enrich?dataset=' + resp.json()['shortId'])\n",
    "\n",
    "def enrichr_get_top_results(userListId, bg, enrichr_link='https://amp.pharm.mssm.edu/Enrichr'):\n",
    "    time.sleep(1)\n",
    "    resp = requests.get(enrichr_link + '/enrich?userListId={}&backgroundType={}'.format(userListId, bg))\n",
    "    if resp.status_code != 200:\n",
    "        raise Exception('Enrichr failed with status {}: {}'.format(\n",
    "          resp.status_code,\n",
    "          resp.text,\n",
    "        ))\n",
    "    time.sleep(1)\n",
    "    return pd.DataFrame(resp.json()[bg], columns=['rank', 'term', 'pvalue', 'zscore', 'combinedscore', 'overlapping_genes', 'adjusted_pvalue', '', ''])\n",
    "\n",
    "def save_enrichr_data(cancer_type,df_diff_expr,df_clustered_umap,df_all_results,df_cluster_aucs):\n",
    "    output=f\"appyter_data/{cancer_type}\"\n",
    "    os.makedirs(output, exist_ok=True)\n",
    "    df_diff_expr.to_csv(\n",
    "        f'{output}/df.tsv',\n",
    "        sep='\\t',\n",
    "        index=None\n",
    "    )\n",
    "    df_clustered_umap.to_csv(\n",
    "        f'{output}/df_umap.tsv',\n",
    "        sep='\\t',\n",
    "        index=None\n",
    "    )\n",
    "    df_all_results.to_csv(\n",
    "        f'{output}/df_enrich.tsv',\n",
    "        sep='\\t',\n",
    "        index=None\n",
    "    )\n",
    "    df_cluster_aucs.to_csv(\n",
    "        f'{output}/cluster_aucs.csv',\n",
    "        sep='\\t',\n",
    "        index=None\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge data\n",
    "df_clustered_umap = pd.merge(left=df_data_norm_km, left_on=\"case_id\", right=df_data_norm_umap, right_on=\"case_id\")\n",
    "df_clustered_pca = pd.merge(left=df_data_norm_km, left_on=\"case_id\", right=df_data_norm_pca, right_on=\"case_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get top Genes for each cluster\n",
    "top_genes = {}\n",
    "for cluster in df_clustered_umap['Cluster'].unique():\n",
    "    fc_col = f'Cluster {cluster} Log2 fold change'\n",
    "    p_col = f'Cluster {cluster} Adjusted p value'\n",
    "    cd_col = f'Cluster {cluster} CD'\n",
    "    if p_col in df_diff_expr.columns:\n",
    "        # significant and positive fold change sorted by p value\n",
    "        up_genes = df_diff_expr.loc[\n",
    "          df_diff_expr[((df_diff_expr[p_col] <= 0.05) & (df_diff_expr[fc_col] > 0))][p_col].sort_values().index,\n",
    "          'Symbol'\n",
    "        ].iloc[:top_n_genes].dropna().values\n",
    "\n",
    "        # significant and negative fold change sorted by p value\n",
    "        dn_genes = df_diff_expr.loc[\n",
    "          df_diff_expr[((df_diff_expr[p_col] <= 0.05) & (df_diff_expr[fc_col] < 0))][p_col].sort_values().index,\n",
    "          'Symbol'\n",
    "        ].iloc[:top_n_genes].dropna().values\n",
    "    elif cd_col in df_diff_expr.columns:\n",
    "        # top up genes\n",
    "        up_genes = df_diff_expr.loc[df_diff_expr[cd_col].sort_values(ascending=False).iloc[:top_n_genes].index, 'Symbol'].values\n",
    "        # top down genes\n",
    "        dn_genes = df_diff_expr.loc[df_diff_expr[cd_col].sort_values(ascending=True).iloc[:top_n_genes].index, 'Symbol'].values\n",
    "    else:\n",
    "        raise Exception('Cant find col for cluster')\n",
    "    # save results\n",
    "    top_genes[cluster] = (up_genes, dn_genes)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get Enrichr links for each cluster\n",
    "enrichr_links = {}\n",
    "\n",
    "for cluster, (up_genes, dn_genes) in top_genes.items():\n",
    "    up_link, dn_link = None, None\n",
    "    if up_genes.size:\n",
    "        up_link = enrichr_link_from_genes(up_genes, 'cluster %s up' % (cluster))\n",
    "    else:\n",
    "        print('cluster %s up: empty' % (cluster))\n",
    "    if dn_genes.size:\n",
    "        dn_link = enrichr_link_from_genes(dn_genes, 'cluster %s down' % (cluster))\n",
    "    else:\n",
    "        print('cluster %s down: empty' % (cluster))\n",
    "    enrichr_links[cluster] = (up_link, dn_link)\n",
    "\n",
    "# Grab top results for each cluster\n",
    "all_enrichr_results = []\n",
    "for cluster, (up_link, dn_link) in enrichr_links.items():\n",
    "    for link_type, link in [('up', up_link), ('down', dn_link)]:\n",
    "        if link is None:\n",
    "            continue\n",
    "    for category, libraries in enrichr_libraries.items():\n",
    "        for library in libraries:\n",
    "            try:\n",
    "                results = enrichr_get_top_results(link['userListId'], library).sort_values('pvalue').iloc[:top_n_results]\n",
    "                results['link'] = link['link']\n",
    "                results['library'] = library\n",
    "                results['category'] = category\n",
    "                results['direction'] = link_type\n",
    "                results['cluster'] = cluster\n",
    "                all_enrichr_results.append(results)\n",
    "            except:\n",
    "                print('{}: {} {} {} cluster {} failed, continuing'.format(link, library, category, link_type, cluster))\n",
    "\n",
    "df_enrichr_results = pd.concat(all_enrichr_results).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display a dataframe with clickable enrichr links\n",
    "df_clickable = df_enrichr_results.copy()\n",
    "df_clickable['link'] = df_clickable[\"link\"].apply(make_clickable)\n",
    "table_html = df_clickable.to_html(escape=False)\n",
    "display(HTML(f'<div style=\"max-height: 250px; overflow-y: auto; margin-bottom: 25px;\">{table_html}</div>'))\n",
    "download_button(df_enrichr_results.to_csv(), 'Download Enrichr results', 'Enrichr results.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## L1000 Data\n",
    "\n",
    "We query the <a href=\"https://amp.pharm.mssm.edu/L1000FWD/\" target=\"_blank\">L1000 Fireworks Display</a> (L1000FWD) API to find the most similar and most dissimilar RNA-seq signatures from the L1000 database for each cluster based on the up and down genes. We also display the interactive projection of those results for each cluster as provided by L1000FWD."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Util functions\n",
    "def l1000fwd_results_from_genes(up_genes, down_genes, description='', l100fwd_link='http://amp.pharm.mssm.edu/L1000FWD/'):\n",
    "    ''' Functional access to L1000FWD API\n",
    "    '''\n",
    "    import time\n",
    "    time.sleep(1)\n",
    "    response = requests.post(l100fwd_link + 'sig_search', json={\n",
    "    'up_genes': list(up_genes),\n",
    "    'down_genes': list(down_genes),\n",
    "    })\n",
    "    l1000fwd_results =  {}\n",
    "    if response.status_code != 200:\n",
    "        raise Exception('L1000FWD failed with status {}: {}'.format(\n",
    "          response.status_code,\n",
    "          response.text,\n",
    "        ))\n",
    "    if 'KeyError' in response.text:\n",
    "        l1000fwd_results['result_url'] = None\n",
    "    else:\n",
    "        # Get ID and URL\n",
    "        result_id = response.json()['result_id']\n",
    "        l1000fwd_results['result_url'] = 'https://amp.pharm.mssm.edu/l1000fwd/vanilla/result/'+result_id\n",
    "        l1000fwd_results['result_id'] = result_id\n",
    "\n",
    "        # Get Top\n",
    "        l1000fwd_results['signatures'] = requests.get(l100fwd_link + 'result/topn/' + result_id).json()\n",
    "\n",
    "    # wait a tinybit before returning link (backoff)\n",
    "    time.sleep(1)\n",
    "    return l1000fwd_results\n",
    "\n",
    "def l1000fwd_sig_link(sig_id):\n",
    "    return 'https://amp.pharm.mssm.edu/dmoa/sig/' + sig_id\n",
    "\n",
    "def get_signature_by_id(sig_id):\n",
    "    response = requests.get(\"http://amp.pharm.mssm.edu/L1000FWD/sig/\" + sig_id)\n",
    "    if response.status_code != 200:\n",
    "        raise Exception('L1000FWD signature query  failed with status {}: {}'.format(\n",
    "          response.status_code,\n",
    "          response.text,\n",
    "        ))\n",
    "    return response.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_l1000fwd_results(l1000fwd_results, plot_counter,cluster_id,nr_drugs=7, height=300):\n",
    "    # Check if results\n",
    "    if l1000fwd_results['result_url']:\n",
    "\n",
    "        # Display cluster title\n",
    "        display(HTML('<h1>Cluster {}</h1>'.format(cluster_id)))\n",
    "\n",
    "        # Display IFrame\n",
    "        display(IFrame(l1000fwd_results['result_url'], width=\"1000\", height=\"1000\"))\n",
    "    \n",
    "        # Display tables\n",
    "        for direction, signature_list in l1000fwd_results['signatures'].items():\n",
    "\n",
    "            # Fix dataframe\n",
    "            rename_dict = {'sig_id': 'Signature ID', 'pvals': 'P-value', 'qvals': 'FDR', 'zscores': 'Z-score', 'combined_scores': 'Combined Score'}\n",
    "            signature_dataframe = pd.DataFrame(signature_list)[list(rename_dict.keys())].rename(columns=rename_dict).sort_values('P-value').rename_axis('Rank')\n",
    "            signature_dataframe.index = [x + 1 for x in range(len(signature_dataframe.index))]\n",
    "            signature_csv = signature_dataframe.to_csv(sep=\",\")\n",
    "\n",
    "            # Display table\n",
    "            pd.set_option('max.colwidth', None)\n",
    "            signature_dataframe['Signature ID'] = [f'<a href={l1000fwd_sig_link(x)} target=\"_blank\">{x}</a>' for x in signature_dataframe['Signature ID']]\n",
    "            table_html = signature_dataframe.to_html(escape=False, classes='w-100')\n",
    "            display(HTML(f'<h2>{direction.title()} Signatures: </h2>'))\n",
    "            display(HTML(f'<style>.w-100{{width: 100% !important;}}</style><div style=\"max-height: 250px; overflow-y: auto; margin-bottom: 25px;\">{table_html}</div>'))\n",
    "\n",
    "            # Display download button\n",
    "            download_button(signature_csv, f'Download {direction.title()} Signatures', f'Cluster {cluster_id} L1000FWD {direction.title()} signatures.csv')\n",
    "        # Link\n",
    "        display(HTML('Full results available at: <a href=\"{result_url}\" target=\"_blank\">{result_url}</a>.'.format(**l1000fwd_results)))\n",
    "\n",
    "    # Display error\n",
    "    else:\n",
    "        display(Markdown('### No results were found.\\n This is likely due to the fact that the gene identifiers were not recognized by L1000FWD. Please note that L1000FWD currently only supports HGNC gene symbols (https://www.genenames.org/). If your dataset uses other gene identifier systems, such as Ensembl IDs or Entrez IDs, consider converting them to HGNC. Automated gene identifier conversion is currently under development.'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_counter = 0\n",
    "all_l1000fwd_results = {}\n",
    "for cluster, (up_genes, dn_genes) in top_genes.items():\n",
    "    results = l1000fwd_results_from_genes(up_genes,dn_genes)\n",
    "    all_l1000fwd_results[cluster] = results\n",
    "    plot_l1000fwd_results(results,plot_counter,cluster)\n",
    "    plot_counter += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the signatures most \"opposite\" to the profile of a given cluster (i.e. up and down genes are reversed), we can obtain a set of drugs that may \"perturb\" that cluster into the healthy direction by upregulating its downregulated genes and downregulating the upregulated genes. These may present effective treatments for patients belonging to that cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_drugs = pd.read_csv(\"https://amp.pharm.mssm.edu/l1000fwd/download/Drugs_metadata.csv\")\n",
    "\n",
    "# Load top drug suggestions for each cluster based on the drugs used to produce the top_n_drugs opposite signatures\n",
    "drug_results = {}\n",
    "for cluster, results in all_l1000fwd_results.items():\n",
    "    opposite_sigs = results[\"signatures\"][\"opposite\"][:top_n_drugs]\n",
    "    sig_ids = [sig[\"sig_id\"] for sig in opposite_sigs]\n",
    "    pert_ids = []\n",
    "    for sig_id in sig_ids:\n",
    "        signature = get_signature_by_id(sig_id)\n",
    "        pert_ids.append(signature[\"pert_id\"])\n",
    "    \n",
    "    df_cluster_drugs = df_drugs[df_drugs[\"pert_id\"].isin(pert_ids)].copy()\n",
    "    df_cluster_drugs[\"cluster\"] = cluster\n",
    "    df_cluster_drugs = df_cluster_drugs[[\"cluster\", *list(filter(lambda x: x!=\"cluster\", df_cluster_drugs.columns))]]\n",
    "    drug_results[cluster] = df_cluster_drugs\n",
    "    \n",
    "df_all_drugs = pd.concat(drug_results).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display a dataframe with clickable enrichr links\n",
    "df_clickable = df_all_drugs.copy()\n",
    "df_clickable['pert_url'] = df_clickable[\"pert_url\"].apply(make_clickable)\n",
    "table_html = df_clickable.to_html(escape=False)\n",
    "display(HTML(f'<div style=\"max-height: 250px; overflow-y: auto; margin-bottom: 25px;\">{table_html}</div>'))\n",
    "download_button(df_all_drugs.to_csv(), 'Download L1000FWD drug results', 'L1000FWD drugs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sample values for testing\n",
    "\n",
    "data_filename = \"\"\n",
    "clinical_data_filename = \"\"\n",
    "cancer = '''Malignant melanoma, NOS'''\n",
    "# Dimensionality reduction parameters\n",
    "n_pca_components = 10\n",
    "n_neighbors = 2\n",
    "min_cluster_dist = 0.3\n",
    "# The number of 'top' genes to use for differential expression\n",
    "top_n_genes = 250\n",
    "# The number of 'top' results to keep from enrichment analysis\n",
    "top_n_results = 5\n",
    "transcription_libraries = []\n",
    "pathways_libraries = [] \n",
    "ontologies_libraries = []\n",
    "diseases_drugs_libraries = ['ARCHS4_IDG_Coexp']\n",
    "cell_types_libraries = ['Allen_Brain_Atlas_up']    \n",
    "miscellaneous_libraries = ['Enrichr_Libraries_Most_Popular_Genes']    \n",
    "legacy_libraries = ['BioCarta_2015'] \n",
    "crowd_libraries = []\n",
    "top_n_drugs = 5\n",
    "enrichr_libraries = OrderedDict([\n",
    "    ('Diseases/Drugs', diseases_drugs_libraries), \n",
    "    ('Ontologies', ontologies_libraries),\n",
    "    ('Cell Type', cell_types_libraries),\n",
    "    ('Pathways', pathways_libraries),\n",
    "    ('Transcription', transcription_libraries),\n",
    "    ('Legacy', legacy_libraries),\n",
    "    ('Crowd', crowd_libraries)\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TCGA",
   "language": "python",
   "name": "tcga"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
