{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichr Manhattan Plot Figure\n",
    "\n",
    "This appyter creates a figure visualizing enrichment analysis results from Enrichr (https://amp.pharm.mssm.edu/Enrichr/) in a manhattan plot. \n",
    "\n",
    "The resulting figure will contain a manhattan plot of the p-values of all the gene sets in the Enrichr libraries selected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import math\n",
    "import json\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "import numpy as np\n",
    "import requests\n",
    "import time\n",
    "import bokeh.io\n",
    "from operator import itemgetter\n",
    "from IPython.display import display, FileLink, Markdown\n",
    "from bokeh.plotting import ColumnDataSource, figure, output_notebook, show\n",
    "from bokeh.models import Legend, LegendItem, Span\n",
    "from bokeh.resources import INLINE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(name='section0', title = 'Visualize Your Enrichr Results', subtitle = 'Create a manhattan plot summarizing your Enrichr analysis results. The resulting figure will contain a manhattan plot of the p-values of all the gene sets in the Enrichr libraries selected.', img = 'enrichr-icon.png')%}\n",
    "{% do SectionField(name='section1', title = '1. Submit Your Gene List', subtitle = 'Upload a text file containing your gene list -OR- copy and paste your gene list into the text box below (One gene per row). You can also try it with the default gene list provided.', img = 'enrichr-icon.png')%}\n",
    "{% do SectionField(name='section2', title = '2. Choose Enrichr Libraries', subtitle = 'Select the Enrichr libraries you would like in your figure.', img = 'enrichr-icon.png')%}\n",
    "{% do SectionField(name='section3', title = '3. Other Options', subtitle = 'Choose color scheme of the plot, add a horizontal line to designate a significance level, and choose location of the legend and point labels. NOTE: for the static plot, legends will only appear if there is more than one Enrichr library selected.', img = 'enrichr-icon.png')%}\n",
    "{% do SectionField(name='section4', title = '4. Output Options', subtitle = 'Choose a format and name to save your static figure.', img = 'enrichr-icon.png')%}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "gene_list_filename = {{ FileField(name='gene_list_filename', label='Gene List File', default='', description='Upload your gene list as a text file (One gene per row).',section = 'section1') }}\n",
    "\n",
    "gene_list_input = {{ TextField(name='gene_list_input', label='Gene List', default='NSUN3\\nPOLRMT\\nNLRX1\\nSFXN5\\nZC3H12C\\nSLC25A39\\nARSG\\nDEFB29\\nNDUFB6\\nZFAND1\\nTMEM77\\n5730403B10RIK\\nRP23-195K8.6\\nTLCD1\\nPSMC6\\nSLC30A6\\nLOC100047292\\nLRRC40\\nORC5L\\nMPP7\\nUNC119B\\nPRKACA\\nTCN2\\nPSMC3IP\\nPCMTD2\\nACAA1A\\nLRRC1\\n2810432D09RIK\\nSEPHS2\\nSAC3D1\\nTMLHE\\nLOC623451\\nTSR2\\nPLEKHA7\\nGYS2\\nARHGEF12\\nHIBCH\\nLYRM2\\nZBTB44\\nENTPD5\\nRAB11FIP2\\nLIPT1\\nINTU\\nANXA13\\nKLF12\\nSAT2\\nGAL3ST2\\nVAMP8\\nFKBPL\\nAQP11\\nTRAP1\\nPMPCB\\nTM7SF3\\nRBM39\\nBRI3\\nKDR\\nZFP748\\nNAP1L1\\nDHRS1\\nLRRC56\\nWDR20A\\nSTXBP2\\nKLF1\\nUFC1\\nCCDC16\\n9230114K14RIK\\nRWDD3\\n2610528K11RIK\\nACO1\\nCABLES1\\nLOC100047214\\nYARS2\\nLYPLA1\\nKALRN\\nGYK\\nZFP787\\nZFP655\\nRABEPK\\nZFP650\\n4732466D17RIK\\nEXOSC4\\nWDR42A\\nGPHN\\n2610528J11RIK\\n1110003E01RIK\\nMDH1\\n1200014M14RIK\\nAW209491\\nMUT\\n1700123L14RIK\\n2610036D13RIK\\nCOX15\\nTMEM30A\\nNSMCE4A\\nTM2D2\\nRHBDD3\\nATXN2\\nNFS1\\n3110001I20RIK\\nBC038156\\nLOC100047782\\n2410012H22RIK\\nRILP\\nA230062G08RIK\\nPTTG1IP\\nRAB1\\nAFAP1L1\\nLYRM5\\n2310026E23RIK\\nC330002I19RIK\\nZFYVE20\\nPOLI\\nTOMM70A\\nSLC7A6OS\\nMAT2B\\n4932438A13RIK\\nLRRC8A\\nSMO\\nNUPL2\\nTRPC2\\nARSK\\nD630023B12RIK\\nMTFR1\\n5730414N17RIK\\nSCP2\\nZRSR1\\nNOL7\\nC330018D20RIK\\nIFT122\\nLOC100046168\\nD730039F16RIK\\nSCYL1\\n1700023B02RIK\\n1700034H14RIK\\nFBXO8\\nPAIP1\\nTMEM186\\nATPAF1\\nLOC100046254\\nLOC100047604\\nCOQ10A\\nFN3K\\nSIPA1L1\\nSLC25A16\\nSLC25A40\\nRPS6KA5\\nTRIM37\\nLRRC61\\nABHD3\\nGBE1\\nPARP16\\nHSD3B2\\nESM1\\nDNAJC18\\nDOLPP1\\nLASS2\\nWDR34\\nRFESD\\nCACNB4\\n2310042D19RIK\\nSRR\\nBPNT1\\n6530415H11RIK\\nCLCC1\\nTFB1M\\n4632404H12RIK\\nD4BWG0951E\\nMED14\\nADHFE1\\nTHTPA\\nCAT\\nELL3\\nAKR7A5\\nMTMR14\\nTIMM44\\nSF1\\nIPP\\nIAH1\\nTRIM23\\nWDR89\\nGSTZ1\\nCRADD\\n2510006D16RIK\\nFBXL6\\nLOC100044400\\nZFP106\\nCD55\\n0610013E23RIK\\nAFMID\\nTMEM86A\\nALDH6A1\\nDALRD3\\nSMYD4\\nNME7\\nFARS2\\nTASP1\\nCLDN10\\nA930005H10RIK\\nSLC9A6\\nADK\\nRBKS\\n2210016F16RIK\\nVWCE\\n4732435N03RIK\\nZFP11\\nVLDLR\\n9630013D21RIK\\n4933407N01RIK\\nFAHD1\\nMIPOL1\\n1810019D21RIK\\n1810049H13RIK\\nTFAM\\nPAICS\\n1110032A03RIK\\nLOC100044139\\nDNAJC19\\nBC016495\\nA930041I02RIK\\nRQCD1\\nUSP34\\nZCCHC3\\nH2AFJ\\nPHF7\\n4921508D12RIK\\nKMO\\nPRPF18\\nMCAT\\nTXNDC4\\n4921530L18RIK\\nVPS13B\\nSCRN3\\nTOR1A\\nAI316807\\nACBD4\\nFAH\\nAPOOL\\nCOL4A4\\nLRRC19\\nGNMT\\nNR3C1\\nSIP1\\nASCC1\\nFECH\\nABHD14A\\nARHGAP18\\n2700046G09RIK\\nYME1L1\\nGK5\\nGLO1\\nSBK1\\nCISD1\\n2210011C24RIK\\nNXT2\\nNOTUM\\nANKRD42\\nUBE2E1\\nNDUFV1\\nSLC33A1\\nCEP68\\nRPS6KB1\\nHYI\\nALDH1A3\\nMYNN\\n3110048L19RIK\\nRDH14\\nPROZ\\nGORASP1\\nLOC674449\\nZFP775\\n5430437P03RIK\\nNPY\\nADH5\\nSYBL1\\n4930432O21RIK\\nNAT9\\nLOC100048387\\nMETTL8\\nENY2\\n2410018G20RIK\\nPGM2\\nFGFR4\\nMOBKL2B\\nATAD3A\\n4932432K03RIK\\nDHTKD1\\nUBOX5\\nA530050D06RIK\\nZDHHC5\\nMGAT1\\nNUDT6\\nTPMT\\nWBSCR18\\nLOC100041586\\nCDK5RAP1\\n4833426J09RIK\\nMYO6\\nCPT1A\\nGADD45GIP1\\nTMBIM4\\n2010309E21RIK\\nASB9\\n2610019F03RIK\\n7530414M10RIK\\nATP6V1B2\\n2310068J16RIK\\nDDT\\nKLHDC4\\nHPN\\nLIFR\\nOVOL1\\nNUDT12\\nCDAN1\\nFBXO9\\nFBXL3\\nHOXA7\\nALDH8A1\\n3110057O12RIK\\nABHD11\\nPSMB1\\nENSMUSG00000074286\\nCHPT1\\nOXSM\\n2310009A05RIK\\n1700001L05RIK\\nZFP148\\n39509\\nMRPL9\\nTMEM80\\n9030420J04RIK\\nNAGLU\\nPLSCR2\\nAGBL3\\nPEX1\\nCNO\\nNEO1\\nASF1A\\nTNFSF5IP1\\nPKIG\\nAI931714\\nD130020L05RIK\\nCNTD1\\nCLEC2H\\nZKSCAN1\\n1810044D09RIK\\nMETTL7A\\nSIAE\\nFBXO3\\nFZD5\\nTMEM166\\nTMED4\\nGPR155\\nRNF167\\nSPTLC1\\nRIOK2\\nTGDS\\nPMS1\\nPITPNC1\\nPCSK7\\n4933403G14RIK\\nEI24\\nCREBL2\\nTLN1\\nMRPL35\\n2700038C09RIK\\nUBIE\\nOSGEPL1\\n2410166I05RIK\\nWDR24\\nAP4S1\\nLRRC44\\nB3BP\\nITFG1\\nDMXL1\\nC1D\\n', description='Paste your gene list (One gene per row).', section = 'section1') }}\n",
    "\n",
    "transcription_libraries = {{ MultiCheckboxField(name='transcription_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Transcription', default=[], section = 'section2',choices=[\n",
    "    'ARCHS4_TFs_Coexp',\n",
    "    'ChEA_2016',\n",
    "    'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "    'ENCODE_Histone_Modifications_2015',\n",
    "    'ENCODE_TF_ChIP-seq_2015',\n",
    "    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "    'Genome_Browser_PWMs',\n",
    "    'lncHUB_lncRNA_Co-Expression',\n",
    "    'miRTarBase_2017',\n",
    "    'TargetScan_microRNA_2017',\n",
    "    'TF-LOF_Expression_from_GEO',\n",
    "    'TF_Perturbations_Followed_by_Expression',\n",
    "    'Transcription_Factor_PPIs',\n",
    "    'TRANSFAC_and_JASPAR_PWMs',\n",
    "    'TRRUST_Transcription_Factors_2019']) }}\n",
    "\n",
    "\n",
    "pathways_libraries = {{ MultiCheckboxField(name='pathways_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Pathways', default=[], section = 'section2',choices=[\n",
    "    'ARCHS4_Kinases_Coexp',\n",
    "    'BioCarta_2016',\n",
    "    'BioPlanet_2019',\n",
    "    'BioPlex_2017',\n",
    "    'CORUM',\n",
    "    'Elsevier_Pathway_Collection',\n",
    "    'HMS_LINCS_KinomeScan',\n",
    "    'HumanCyc_2016',\n",
    "    'huMAP',\n",
    "    'KEA_2015',\n",
    "    'KEGG_2019_Human',\n",
    "    'KEGG_2019_Mouse',\n",
    "    'Kinase_Perturbations_from_GEO_down',\n",
    "    'Kinase_Perturbations_from_GEO_up',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "    'NCI-Nature_2016',\n",
    "    'NURSA_Human_Endogenous_Complexome',\n",
    "    'Panther_2016',\n",
    "    'Phosphatase_Substrates_from_DEPOD',\n",
    "    'PPI_Hub_Proteins',\n",
    "    'Reactome_2016',\n",
    "    'SILAC_Phosphoproteomics',\n",
    "    'SubCell_BarCode',\n",
    "    'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "    'WikiPathways_2019_Human',\n",
    "    'WikiPathways_2019_Mouse']) }}    \n",
    "    \n",
    "  \n",
    "ontologies_libraries = {{ MultiCheckboxField(name='ontologies_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Ontologies', default=[], section = 'section2',choices=[\n",
    "    'GO_Biological_Process_2018',\n",
    "    'GO_Cellular_Component_2018',\n",
    "    'GO_Molecular_Function_2018',\n",
    "    'Human_Phenotype_Ontology',\n",
    "    'Jensen_COMPARTMENTS',\n",
    "    'Jensen_DISEASES',\n",
    "    'Jensen_TISSUES',\n",
    "    'MGI_Mammalian_Phenotype_Level_4_2019']) }} \n",
    "\n",
    "diseases_drugs_libraries = {{ MultiCheckboxField(name='diseases_drugs_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Diseases/Drugs', default=[], section = 'section2',choices=[    \n",
    "    'Achilles_fitness_decrease',\n",
    "    'Achilles_fitness_increase',\n",
    "    'ARCHS4_IDG_Coexp',\n",
    "    'ClinVar_2019',\n",
    "    'dbGaP',\n",
    "    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "    'DisGeNET',\n",
    "    'DrugMatrix',\n",
    "    'DSigDB',\n",
    "    'GeneSigDB',\n",
    "    'GWAS_Catalog_2019',\n",
    "    'LINCS_L1000_Chem_Pert_down',\n",
    "    'LINCS_L1000_Chem_Pert_up',\n",
    "    'LINCS_L1000_Ligand_Perturbations_down',\n",
    "    'LINCS_L1000_Ligand_Perturbations_up',\n",
    "    'MSigDB_Computational',\n",
    "    'MSigDB_Oncogenic_Signatures',\n",
    "    'Old_CMAP_down',\n",
    "    'Old_CMAP_up',\n",
    "    'OMIM_Disease',\n",
    "    'OMIM_Expanded',\n",
    "    'PheWeb_2019',\n",
    "    'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "    'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "    'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "    'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "    'UK_Biobank_GWAS_v1',\n",
    "    'Virus_Perturbations_from_GEO_down',\n",
    "    'Virus_Perturbations_from_GEO_up',\n",
    "    'VirusMINT']) }}\n",
    "    \n",
    "    \n",
    "cell_types_libraries = {{ MultiCheckboxField(name='cell_types_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Cell Types', default=[], section = 'section2',choices=[        \n",
    "    'Allen_Brain_Atlas_down',\n",
    "    'Allen_Brain_Atlas_up',\n",
    "    'ARCHS4_Cell-lines',\n",
    "    'ARCHS4_Tissues',\n",
    "    'Cancer_Cell_Line_Encyclopedia',\n",
    "    'CCLE_Proteomics_2020',\n",
    "    'ESCAPE',\n",
    "    'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "    'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "    'Human_Gene_Atlas',\n",
    "    'Mouse_Gene_Atlas',\n",
    "    'NCI-60_Cancer_Cell_Lines',\n",
    "    'ProteomicsDB_2020',\n",
    "    'Tissue_Protein_Expression_from_Human_Proteome_Map']) }}    \n",
    "    \n",
    "    \n",
    "    \n",
    "miscellaneous_libraries = {{ MultiCheckboxField(name='miscellaneous_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Miscellaneous', default=[], section = 'section2',choices=[            \n",
    "    'Chromosome_Location_hg19',\n",
    "    'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "    'Enrichr_Libraries_Most_Popular_Genes',\n",
    "    'Genes_Associated_with_NIH_Grants',\n",
    "    'HMDB_Metabolites',\n",
    "    'HomoloGene',\n",
    "    'InterPro_Domains_2019',\n",
    "    'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "    'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "    'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "    'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "    'Pfam_Domains_2019',\n",
    "    'Pfam_InterPro_Domains',\n",
    "    'Table_Mining_of_CRISPR_Studies']) }}    \n",
    "    \n",
    "    \n",
    "legacy_libraries = {{ MultiCheckboxField(name='legacy_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Legacy', default=[], section = 'section2',choices=[                \n",
    "    'BioCarta_2013',\n",
    "    'BioCarta_2015',\n",
    "    'ChEA_2013',\n",
    "    'ChEA_2015',\n",
    "    'Chromosome_Location',\n",
    "    'Disease_Signatures_from_GEO_down_2014',\n",
    "    'Disease_Signatures_from_GEO_up_2014',\n",
    "    'Drug_Perturbations_from_GEO_2014',\n",
    "    'ENCODE_Histone_Modifications_2013',\n",
    "    'ENCODE_TF_ChIP-seq_2014',\n",
    "    'GO_Biological_Process_2013',\n",
    "    'GO_Biological_Process_2015',\n",
    "    'GO_Biological_Process_2017',\n",
    "    'GO_Biological_Process_2017b',\n",
    "    'GO_Cellular_Component_2013',\n",
    "    'GO_Cellular_Component_2015',\n",
    "    'GO_Cellular_Component_2017',\n",
    "    'GO_Cellular_Component_2017b',\n",
    "    'GO_Molecular_Function_2013',\n",
    "    'GO_Molecular_Function_2015',\n",
    "    'GO_Molecular_Function_2017',\n",
    "    'GO_Molecular_Function_2017b',\n",
    "    'HumanCyc_2015',\n",
    "    'KEA_2013',\n",
    "    'KEGG_2013',\n",
    "    'KEGG_2015',\n",
    "    'KEGG_2016',\n",
    "    'MGI_Mammalian_Phenotype_2013',\n",
    "    'MGI_Mammalian_Phenotype_2017',\n",
    "    'MGI_Mammalian_Phenotype_Level_3',\n",
    "    'MGI_Mammalian_Phenotype_Level_4',\n",
    "    'NCI-Nature_2015',\n",
    "    'Panther_2015',\n",
    "    'Reactome_2013',\n",
    "    'Reactome_2015',\n",
    "    'TargetScan_microRNA',\n",
    "    'Tissue_Protein_Expression_from_ProteomicsDB',\n",
    "    'WikiPathways_2013',\n",
    "    'WikiPathways_2015',\n",
    "    'WikiPathways_2016']) }} \n",
    "\n",
    "crowd_libraries = {{ MultiCheckboxField(name='crowd_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Crowd', default=[], section = 'section2',choices=[                \n",
    "    'Aging_Perturbations_from_GEO_down',\n",
    "    'Aging_Perturbations_from_GEO_up',\n",
    "    'Disease_Perturbations_from_GEO_down',\n",
    "    'Disease_Perturbations_from_GEO_up',\n",
    "    'Drug_Perturbations_from_GEO_down',\n",
    "    'Drug_Perturbations_from_GEO_up',\n",
    "    'Gene_Perturbations_from_GEO_down',\n",
    "    'Gene_Perturbations_from_GEO_up',\n",
    "    'Ligand_Perturbations_from_GEO_down',\n",
    "    'Ligand_Perturbations_from_GEO_up',\n",
    "    'MCF7_Perturbations_from_GEO_down',\n",
    "    'MCF7_Perturbations_from_GEO_up',\n",
    "    'Microbe_Perturbations_from_GEO_down',\n",
    "    'Microbe_Perturbations_from_GEO_up',\n",
    "    'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "    'SysMyo_Muscle_Gene_Sets']) }}\n",
    "\n",
    "\n",
    "color_choice = '{{ ChoiceField(name='color_choice', label='Color Scheme', default='blue/purple/orange', description='Choose the color scheme you want for your plots.', section = 'section3', choices=[\n",
    "    'blue/purple/orange',\n",
    "    'orange',\n",
    "    'red/orange',\n",
    "    'blue/purple',\n",
    "    'purple/pink',\n",
    "    'green',\n",
    "    'rainbow',\n",
    "    'grayscale']) }}'\n",
    "\n",
    "significance_line = {{ StringField(name='significance_line', label='Significance Line Value', default='', description='Enter a value at which you want a horizontal line to appear (generally used to indicate that values above the line are significant).', section = 'section3') }}\n",
    "\n",
    "legend_location = '{{ ChoiceField(name='legend_location', label='Legend Location', default='below', description='Choose if you want the plot legend to be to the side or below the figure (static plot).', choices= ['below', 'side'], section = 'section3') }}'\n",
    "\n",
    "label_location = '{{ ChoiceField(name='label_location', label='Point Label Location', default='right', description='Choose where you want the labels of the significant points to be located (static plot).', choices= ['right', 'above', 'below'], section = 'section3') }}'\n",
    "\n",
    "figure_file_format = {{ MultiCheckboxField(name='figure_file_format', label='File Format', choices = ['png', 'pdf', 'svg' ], default=['png'], description='Select the format(s) to save your figure (static plot).', section = 'section4') }}\n",
    "\n",
    "output_file_name = {{ StringField(name='output_file_name', label='File Name', default='Enrichr_results', description='Enter a name/description to save your figure (static plot).', section = 'section4') }}\n",
    "\n",
    "final_output_file_names = [str(output_file_name+'.'+file_type) for file_type in figure_file_format]\n",
    "\n",
    "enrichr_libraries = np.sort(transcription_libraries+pathways_libraries+ontologies_libraries+diseases_drugs_libraries+cell_types_libraries+miscellaneous_libraries+legacy_libraries+crowd_libraries)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import gene list as file or from text box file\n",
    "# Will choose file upload over textbox if a file is given \n",
    "if gene_list_filename != '':\n",
    "    open_gene_list_file = open(gene_list_filename,'r')\n",
    "    lines = open_gene_list_file.readlines()\n",
    "    genes = [x.strip() for x in lines]\n",
    "    open_gene_list_file.close()\n",
    "else:\n",
    "    genes = gene_list_input.split('\\n')\n",
    "    genes = [x.strip() for x in genes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get Enrichr Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get Enrichr Results \n",
    "# Takes a gene list and Enrichr libraries as input\n",
    "def Enrichr_API(enrichr_gene_list, all_libraries):\n",
    "\n",
    "    for library_name in all_libraries : \n",
    "        ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/addList'\n",
    "        genes_str = '\\n'.join(enrichr_gene_list)\n",
    "        description = 'Example gene list'\n",
    "        payload = {\n",
    "            'list': (None, genes_str),\n",
    "            'description': (None, description)\n",
    "        }\n",
    "\n",
    "        response = requests.post(ENRICHR_URL, files=payload)\n",
    "        if not response.ok:\n",
    "            raise Exception('Error analyzing gene list')\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "        time.sleep(0.5)\n",
    "        ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/enrich'\n",
    "        query_string = '?userListId=%s&backgroundType=%s'\n",
    "        user_list_id = data['userListId']\n",
    "        short_id = data[\"shortId\"]\n",
    "        gene_set_library = library_name\n",
    "        response = requests.get(\n",
    "            ENRICHR_URL + query_string % (user_list_id, gene_set_library)\n",
    "         )\n",
    "        if not response.ok:\n",
    "            raise Exception('Error fetching enrichment results')\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "\n",
    "        #results_df  = pd.DataFrame(data[library_name][0:5])\n",
    "        \n",
    "        results_df  = pd.DataFrame(data[library_name])\n",
    "        # adds library name to the data frame so the libraries can be distinguished\n",
    "        results_df['library'] = library_name.replace('_', '')\n",
    "\n",
    "\n",
    "    return([results_df, str(short_id)])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assign Color Scheme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = []\n",
    "\n",
    "if color_choice == 'orange':\n",
    "    colors = ['#FF5A00', '#FFA700', '#FF7400', '#FFDB00']\n",
    "if color_choice == 'red/orange':\n",
    "    colors = ['#FF0000',  '#FFCC00', '#FF6600', '#FF9800']\n",
    "if color_choice == 'blue/purple':\n",
    "    colors = ['#0000FF', '#A3A3FF', '#4949FF', '#7879FF']\n",
    "if color_choice == 'green':\n",
    "    colors = ['#2eb62c', '#abe098', '#57c84d', '#c5e8b7']\n",
    "if color_choice == 'rainbow':\n",
    "    colors = ['red', '#fef200', 'green', 'blue', 'purple']\n",
    "if color_choice == 'blue/purple/orange':\n",
    "    colors = ['#003f5c', '#7a5195', '#ef5675', '#ffa600']\n",
    "if color_choice == 'purple/pink':\n",
    "    colors = ['#9800b0', '#ef83bd', '#bc37b1', '#d95db5']\n",
    "if color_choice == 'grayscale':\n",
    "    colors = ['#000000', '#7a7a7a', '#3c3c3c', '#bdbdbd']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Enrichr Results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function plots results \n",
    "\n",
    "def enrichr_figure():\n",
    "\n",
    "    if len(enrichr_libraries) == 1:\n",
    "\n",
    "        results_df = Enrichr_API(genes, enrichr_libraries)[0]\n",
    "\n",
    "        all_terms = []\n",
    "        all_pvalues = []\n",
    "\n",
    "        all_terms.append(list(results_df[1]))\n",
    "        all_pvalues.append(list(results_df[2]))\n",
    "\n",
    "        # make a simple scatterplot\n",
    "        fig, ax = plt.subplots(figsize=(10,4))\n",
    "\n",
    "        # sort the elements alphabetically\n",
    "        x=np.log10(all_pvalues[0])*-1\n",
    "        sorted_terms = list(zip(all_terms[0], x))\n",
    "        sorted_terms = sorted(sorted_terms, key = itemgetter(0))\n",
    "        unzipped_sorted_list = list(zip(*sorted_terms))\n",
    "\n",
    "        data = pd.DataFrame({\"Gene Set\": unzipped_sorted_list[0], \"-log(p value)\": unzipped_sorted_list[1]})\n",
    "\n",
    "        # add significance line and label significant points\n",
    "        if significance_line != '':\n",
    "            ax.axes.axhline(y = float(significance_line), color = 'black', lw = 1)\n",
    "\n",
    "            # label points above the significance line\n",
    "\n",
    "            if label_location == 'right':\n",
    "                coords = (6, -3)\n",
    "            elif label_location == 'below':\n",
    "                coords = (-3, -14)\n",
    "            elif label_location == 'above':\n",
    "                coords = (-4, 5)\n",
    "\n",
    "            point_label = 1\n",
    "            sig_point_handles = []\n",
    "            for index, row in data.iterrows():\n",
    "                if row[\"-log(p value)\"] > float(significance_line):\n",
    "                    ax.annotate(point_label, xy = (row[\"Gene Set\"], row[\"-log(p value)\"]), xytext = coords, textcoords='offset points')\n",
    "                    actual_pvalue = \"{:.5e}\".format(10**(-1*row[\"-log(p value)\"]))\n",
    "                    sig_point_handles += [mpatches.Patch(color = 'white', label = str(point_label) + \": \" + row[\"Gene Set\"] + \", \" + str(actual_pvalue))]\n",
    "                    point_label += 1\n",
    "\n",
    "            # create a legend for the significant points\n",
    "            if point_label != 1:\n",
    "                leg = plt.legend(handles = sig_point_handles, handlelength=0, handletextpad=0, loc='center left', bbox_to_anchor=(1, 0.5), title=\"Significant Points\")\n",
    "                ax.add_artist(leg)\n",
    "\n",
    "        ax.scatter(unzipped_sorted_list[0], unzipped_sorted_list[1], color = colors[0])\n",
    "        ax.axes.get_xaxis().set_ticks([])\n",
    "        plt.ylabel(\"-log(p value)\")\n",
    "        plt.xlabel(enrichr_libraries[0])\n",
    "\n",
    "    elif len(enrichr_libraries) > 1:\n",
    "        # make a manhattan plot\n",
    "\n",
    "        sorted_data = pd.DataFrame({\"Gene Set\": [], \"-log(p value)\": [], \"Library\": []})\n",
    "        fig, ax = plt.subplots(figsize=(10,4))\n",
    "\n",
    "        for i in range(len(enrichr_libraries)):\n",
    "            # get enrichr results from the library selected\n",
    "            results_df = Enrichr_API(genes, [enrichr_libraries[i]])[0]\n",
    "\n",
    "            all_terms = []\n",
    "            all_pvalues = []\n",
    "            library_names = []\n",
    "\n",
    "            all_terms.append(list(results_df[1]))\n",
    "            all_pvalues.append(list(results_df[2]))\n",
    "            library_names.append(list(results_df['library']))\n",
    "\n",
    "            x=np.log10(all_pvalues[0])*-1\n",
    "            sorted_terms = list(zip(all_terms[0], x, library_names[0]))\n",
    "            sorted_terms = sorted(sorted_terms, key = itemgetter(0))\n",
    "            unzipped_sorted_list = list(zip(*sorted_terms))\n",
    "\n",
    "            data = pd.DataFrame({\"Gene Set\": unzipped_sorted_list[0], \"-log(p value)\": unzipped_sorted_list[1], \"Library\": unzipped_sorted_list[2]})\n",
    "        \n",
    "            sorted_data = pd.concat([sorted_data, data])\n",
    "\n",
    "        # group data by library\n",
    "        groups = sorted_data.groupby(\"Library\")\n",
    "\n",
    "        # plot points\n",
    "        color_index = 0\n",
    "        for name, group in groups:\n",
    "            if color_index >= len(colors):\n",
    "                color_index = 0\n",
    "            plt.plot(group[\"Gene Set\"], group[\"-log(p value)\"], marker=\"o\", linestyle=\"\", label=name, color = colors[color_index])\n",
    "            color_index += 1\n",
    "\n",
    "        # remove labels and tick marks on the x-axis\n",
    "        ax.axes.get_xaxis().set_ticks([])\n",
    "\n",
    "        # now sort dataframe by p-value so the significant points are labeled in order\n",
    "        sorted_pvalue_data = sorted_data.sort_values(by = [\"-log(p value)\"], ascending = False)\n",
    "\n",
    "        # add significance line and label significant points\n",
    "        if significance_line != '':\n",
    "            ax.axes.axhline(y = float(significance_line), color = 'black', lw = 1)\n",
    "\n",
    "            # label points above the significance line\n",
    "            if label_location == 'right':\n",
    "                coords = (6, -3)\n",
    "            elif label_location == 'below':\n",
    "                coords = (-3, -14)\n",
    "            elif label_location == 'above':\n",
    "                coords = (-4, 5)\n",
    "\n",
    "            point_label = 1\n",
    "            sig_point_handles = []\n",
    "            for index, row in sorted_pvalue_data.iterrows():\n",
    "                if row[\"-log(p value)\"] > float(significance_line):\n",
    "                    ax.annotate(point_label, xy = (row[\"Gene Set\"], row[\"-log(p value)\"]), xytext = coords, textcoords='offset points')\n",
    "                    actual_pvalue = \"{:.5e}\".format(10**(-1*row[\"-log(p value)\"]))\n",
    "                    sig_point_handles += [mpatches.Patch(color = 'white', label = str(point_label) + \": \" + row[\"Gene Set\"] + \", \" + str(actual_pvalue))]\n",
    "                    point_label += 1\n",
    "\n",
    "            # create a legend for the significant points\n",
    "            if legend_location == 'side' and point_label != 1:\n",
    "                leg = plt.legend(handles = sig_point_handles, handlelength=0, handletextpad=0, loc='center left', bbox_to_anchor=(1.5, 0.5), title=\"Significant Points\")\n",
    "                ax.add_artist(leg)\n",
    "            elif point_label != 1:\n",
    "                leg = plt.legend(handles = sig_point_handles, handlelength=0, handletextpad=0, loc='center left', bbox_to_anchor=(1, 0.5), title=\"Significant Points\")\n",
    "                ax.add_artist(leg)\n",
    "\n",
    "        # adds a legend for the libraries in the location specified\n",
    "        if legend_location == 'below':\n",
    "            # shrink current axis's height by 10% on the bottom\n",
    "            box = ax.get_position()\n",
    "            ax.set_position([box.x0, box.y0 + box.height * 0.1,\n",
    "                            box.width, box.height * 0.9])\n",
    "\n",
    "            # put a legend below current axis\n",
    "            ax.legend(loc='upper center', bbox_to_anchor=(0.5, -0.05),\n",
    "                    fancybox=True, shadow=True, ncol=5)\n",
    "\n",
    "        elif legend_location == 'side':\n",
    "            # shrink current axis by 20%\n",
    "            box = ax.get_position()\n",
    "            ax.set_position([box.x0, box.y0, box.width * 0.8, box.height])\n",
    "\n",
    "            # put a legend to the right of the current axis\n",
    "            ax.legend(loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "\n",
    "        plt.ylabel(\"-log(p value)\")\n",
    "\n",
    "        # save results\n",
    "    for plot_name in final_output_file_names:\n",
    "        plt.savefig(plot_name, bbox_inches = 'tight')\n",
    "\n",
    "    plt.show()\n",
    "    return sorted_data, groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_data, groups = enrichr_figure()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download plots\n",
    "for i, file in enumerate(final_output_file_names):\n",
    "    display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Having trouble with overlapping point labels?\n",
    "Try moving the labels to a different location, plotting fewer libraries, or plot only one library at a time."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Interactive plot using Bokeh\n",
    "You can hover over the data points in this plot to see their associated gene set and p-value. Also check out the toolbar on the right side of the plot which will allow you to pan, box zoom, reset view, save the plot, and turn hover viewing on/off. This plot may open in the new window."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bokeh.io.output_notebook(INLINE)\n",
    "\n",
    "# split data frame into smaller data frames by library\n",
    "list_of_df = []\n",
    "for library_name in enrichr_libraries:\n",
    "    library_name = library_name.replace('_', '')\n",
    "    df_new = sorted_data[sorted_data['Library'] == library_name]\n",
    "    list_of_df += [df_new]\n",
    "\n",
    "list_of_xaxis_values = []\n",
    "for df in list_of_df:  \n",
    "    list_of_xaxis_values += df[\"Gene Set\"].values.tolist()\n",
    "\n",
    "# define the output figure and the features we want\n",
    "p = figure(x_range = list_of_xaxis_values, plot_height=300, plot_width=1000, tools='pan, box_zoom, hover, reset, save')\n",
    "\n",
    "# loop over all libraries\n",
    "r = []\n",
    "color_index = 0\n",
    "for df in list_of_df:\n",
    "    if color_index >= len(colors):\n",
    "        color_index = 0 \n",
    "\n",
    "    # calculate actual p value from -log(p value)\n",
    "    actual_pvalues = []\n",
    "    for log_value in df[\"-log(p value)\"].values.tolist():\n",
    "        actual_pvalues += [\"{:.5e}\".format(10**(-1*log_value))]\n",
    "\n",
    "    # define ColumnDataSource with our data for this library\n",
    "    source = ColumnDataSource(data=dict(\n",
    "        x = df[\"Gene Set\"].values.tolist(),\n",
    "        y = df[\"-log(p value)\"].values.tolist(),\n",
    "        pvalue = actual_pvalues,\n",
    "    ))\n",
    "    \n",
    "    # plot data from this library\n",
    "    r += [p.circle(x = 'x', y = 'y', size=5, fill_color=colors[color_index], line_color= colors[color_index], line_width=1, source = source)]\n",
    "    color_index += 1\n",
    "\n",
    "# create custom legend for the libraries\n",
    "color_index = 0\n",
    "renderer_index = 0\n",
    "legend_items = []\n",
    "for library_name in enrichr_libraries:\n",
    "    legend_items += [LegendItem(label = library_name, renderers = [r[renderer_index]])]\n",
    "    renderer_index += 1\n",
    "\n",
    "legend = Legend(items = legend_items, location = (0, 160))\n",
    "p.add_layout(legend, 'right')\n",
    "\n",
    "# add significance line\n",
    "if significance_line != '':\n",
    "    hline = Span(location = float(significance_line), dimension='width', line_color='black', line_width=1)\n",
    "    p.renderers.extend([hline])\n",
    "\n",
    "p.background_fill_color = 'white'\n",
    "p.xaxis.major_tick_line_color = None \n",
    "p.xaxis.major_label_text_font_size = '0pt'\n",
    "p.y_range.start = 0\n",
    "p.yaxis.axis_label = '-log(p value)'\n",
    "\n",
    "p.hover.tooltips = [\n",
    "    (\"Gene Set\", \"@x\"),\n",
    "    (\"p value\", \"@pvalue\"),\n",
    "]\n",
    "\n",
    "show(p)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6-final"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}