{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%jupyter_template init\n",
    "from jupyter_template import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichr Figure \n",
    "\n",
    "This template creates a publishable figure summarizing enrichment analysis results from Enrichr. \n",
    "\n",
    "The resulting figure will contain the top 5 enriched terms and their corresponding p-values for each library chosen. Red bars correspond to terms with significant p-values (<0.05). An asterisk (*) next to a p-value indicates the term also has a significant adjusted p-value (<0.05)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import json\n",
    "import requests\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "import datetime\n",
    "from datetime import date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%jupyter_template hide_code\n",
    "\n",
    "{% do SectionField(name='section1', title = 'Publish Your Enrichr Results', subtitle = 'Create a publishable figure summarizing your Enrichr analysis results. The resulting figure will contain the top 5 enriched terms and their corresponding p-values for each library chosen. Red bars correspond to terms with significant p-values (<0.05). An asterisk (*) next to a p-value indicates the term also has a significant adjusted p-value (<0.05).', img = 'enrichr-icon.png')%}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%jupyter_template code_eval\n",
    "\n",
    "gene_list_filename = \"{{ FileField(name='gene_list_filename', label='Gene List File', default='', description='Upload your gene list as a text file. One gene per row ONLY.',section = 'section1').value }}\"\n",
    "\n",
    "enrichr_libraries = {{ MultiChoiceField(name='enrichr_libraries', description='Select the Enrichr libraries you would like in your figure.', label='Enrichr Libraries', default=[], section = 'section1',choices=[\n",
    "    'Genes_Associated_with_NIH_Grants',\n",
    "    'Cancer_Cell_Line_Encyclopedia',\n",
    "    'Achilles_fitness_decrease',\n",
    "    'Achilles_fitness_increase',\n",
    "    'Aging_Perturbations_from_GEO_down',\n",
    "    'Aging_Perturbations_from_GEO_up',\n",
    "    'Allen_Brain_Atlas_down',\n",
    "    'Allen_Brain_Atlas_up',\n",
    "    'ARCHS4_Cell-lines',\n",
    "    'ARCHS4_IDG_Coexp',\n",
    "    'ARCHS4_Kinases_Coexp',\n",
    "    'ARCHS4_TFs_Coexp',\n",
    "    'ARCHS4_Tissues',\n",
    "    'BioCarta_2013',\n",
    "    'BioCarta_2015',\n",
    "    'BioCarta_2016',\n",
    "    'BioPlanet_2019',\n",
    "    'BioPlex_2017',\n",
    "    'CCLE_Proteomics_2020',\n",
    "    'ChEA_2013',\n",
    "    'ChEA_2015',\n",
    "    'ChEA_2016',\n",
    "    'Chromosome_Location',\n",
    "    'Chromosome_Location_hg19',\n",
    "    'ClinVar_2019',\n",
    "    'CORUM',\n",
    "    'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "    'dbGaP',\n",
    "    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "    'Disease_Perturbations_from_GEO_down',\n",
    "    'Disease_Perturbations_from_GEO_up',\n",
    "    'Disease_Signatures_from_GEO_down_2014',\n",
    "    'Disease_Signatures_from_GEO_up_2014',\n",
    "    'DisGeNET',\n",
    "    'Drug_Perturbations_from_GEO_2014',\n",
    "    'Drug_Perturbations_from_GEO_down',\n",
    "    'Drug_Perturbations_from_GEO_up',\n",
    "    'DrugMatrix',\n",
    "    'DSigDB',\n",
    "    'Elsevier_Pathway_Collection',\n",
    "    'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "    'ENCODE_Histone_Modifications_2013',\n",
    "    'ENCODE_Histone_Modifications_2015',\n",
    "    'ENCODE_TF_ChIP-seq_2014',\n",
    "    'ENCODE_TF_ChIP-seq_2015',\n",
    "    'Enrichr_Libraries_Most_Popular_Genes',\n",
    "    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "    'ESCAPE',\n",
    "    'Gene_Perturbations_from_GEO_down',\n",
    "    'Gene_Perturbations_from_GEO_up',\n",
    "    'GeneSigDB',\n",
    "    'Genome_Browser_PWMs',\n",
    "    'GO_Biological_Process_2013',\n",
    "    'GO_Biological_Process_2015',\n",
    "    'GO_Biological_Process_2017',\n",
    "    'GO_Biological_Process_2017b',\n",
    "    'GO_Biological_Process_2018',\n",
    "    'GO_Cellular_Component_2013',\n",
    "    'GO_Cellular_Component_2015',\n",
    "    'GO_Cellular_Component_2017',\n",
    "    'GO_Cellular_Component_2017b',\n",
    "    'GO_Cellular_Component_2018',\n",
    "    'GO_Molecular_Function_2013',\n",
    "    'GO_Molecular_Function_2015',\n",
    "    'GO_Molecular_Function_2017',\n",
    "    'GO_Molecular_Function_2017b',\n",
    "    'GO_Molecular_Function_2018',\n",
    "    'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "    'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "    'GWAS_Catalog_2019',\n",
    "    'HMDB_Metabolites',\n",
    "    'HMS_LINCS_KinomeScan',\n",
    "    'HomoloGene',\n",
    "    'Human_Gene_Atlas',\n",
    "    'Human_Phenotype_Ontology',\n",
    "    'HumanCyc_2015',\n",
    "    'HumanCyc_2016',\n",
    "    'huMAP',\n",
    "    'InterPro_Domains_2019',\n",
    "    'Jensen_COMPARTMENTS',\n",
    "    'Jensen_DISEASES',\n",
    "    'Jensen_TISSUES',\n",
    "    'KEA_2013',\n",
    "    'KEA_2015',\n",
    "    'KEGG_2013',\n",
    "    'KEGG_2015',\n",
    "    'KEGG_2016',\n",
    "    'KEGG_2019_Human',\n",
    "    'KEGG_2019_Mouse',\n",
    "    'Kinase_Perturbations_from_GEO_down',\n",
    "    'Kinase_Perturbations_from_GEO_up',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "    'Ligand_Perturbations_from_GEO_down',\n",
    "    'Ligand_Perturbations_from_GEO_up',\n",
    "    'LINCS_L1000_Chem_Pert_down',\n",
    "    'LINCS_L1000_Chem_Pert_up',\n",
    "    'LINCS_L1000_Ligand_Perturbations_down',\n",
    "    'LINCS_L1000_Ligand_Perturbations_up',\n",
    "    'lncHUB_lncRNA_Co-Expression',\n",
    "    'MCF7_Perturbations_from_GEO_down',\n",
    "    'MCF7_Perturbations_from_GEO_up',\n",
    "    'MGI_Mammalian_Phenotype_2013',\n",
    "    'MGI_Mammalian_Phenotype_2017',\n",
    "    'MGI_Mammalian_Phenotype_Level_3',\n",
    "    'MGI_Mammalian_Phenotype_Level_4',\n",
    "    'MGI_Mammalian_Phenotype_Level_4_2019',\n",
    "    'Microbe_Perturbations_from_GEO_down',\n",
    "    'Microbe_Perturbations_from_GEO_up',\n",
    "    'miRTarBase_2017',\n",
    "    'Mouse_Gene_Atlas',\n",
    "    'MSigDB_Computational',\n",
    "    'MSigDB_Oncogenic_Signatures',\n",
    "    'NCI-60_Cancer_Cell_Lines',\n",
    "    'NCI-Nature_2015',\n",
    "    'NCI-Nature_2016',\n",
    "    'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "    'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "    'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "    'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "    'NURSA_Human_Endogenous_Complexome',\n",
    "    'Old_CMAP_down',\n",
    "    'Old_CMAP_up',\n",
    "    'OMIM_Disease',\n",
    "    'OMIM_Expanded',\n",
    "    'Panther_2015',\n",
    "    'Panther_2016',\n",
    "    'Pfam_Domains_2019',\n",
    "    'Pfam_InterPro_Domains',\n",
    "    'PheWeb_2019',\n",
    "    'Phosphatase_Substrates_from_DEPOD',\n",
    "    'PPI_Hub_Proteins',\n",
    "    'ProteomicsDB_2020',\n",
    "    'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "    'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "    'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "    'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "    'Reactome_2013',\n",
    "    'Reactome_2015',\n",
    "    'Reactome_2016',\n",
    "    'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "    'SILAC_Phosphoproteomics',\n",
    "    'SubCell_BarCode',\n",
    "    'SysMyo_Muscle_Gene_Sets',\n",
    "    'TargetScan_microRNA',\n",
    "    'TargetScan_microRNA_2017',\n",
    "    'TF-LOF_Expression_from_GEO',\n",
    "    'TF_Perturbations_Followed_by_Expression',\n",
    "    'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
    "    'Tissue_Protein_Expression_from_ProteomicsDB',\n",
    "    'Transcription_Factor_PPIs',\n",
    "    'TRANSFAC_and_JASPAR_PWMs',\n",
    "    'TRRUST_Transcription_Factors_2019',\n",
    "    'UK_Biobank_GWAS_v1',\n",
    "    'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "    'Virus_Perturbations_from_GEO_down',\n",
    "    'Virus_Perturbations_from_GEO_up',\n",
    "    'VirusMINT',\n",
    "    'WikiPathways_2013',\n",
    "    'WikiPathways_2015',\n",
    "    'WikiPathways_2016',\n",
    "    'WikiPathways_2019_Human',\n",
    "    'WikiPathways_2019_Mouse']).value}}\n",
    "\n",
    "\n",
    "figure_file_format = \"{{ ChoiceField(name='figure_file_format', label='File Format', choices = ['PNG', 'SVG', 'PDF'], default=['PNG'], description='Choose the format to save your figure.', section = 'section1').value }}\"\n",
    "\n",
    "output_file_name = \"{{ StringField(name='output_file_name', label='File Name (Optional)', default='', description='Enter a name/description to save your figure.', section = 'section1').safe_value }}\"\n",
    "\n",
    "annot_dict_filename = \"{{ FileField(name='annot_dict_filename', label='Term Conversions (Optional)', default=' ', description='Enrichment terms may be too long for the figure and may need to be manually changed. You may upload a TSV file with two columns containing original terms (left column) and your corresponding preferred shortened terms (right column).', section = 'section1').value }}\"\n",
    "\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import gene list file \n",
    "open_gene_list_file = open(gene_list_filename,'r')\n",
    "lines = open_gene_list_file.readlines()\n",
    "genes = [x.strip() for x in lines]\n",
    "open_gene_list_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import term conversion file (Optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Conversion dictionary for long terms\n",
    "annot_dict = {}\n",
    "\n",
    "# Import term conversion file if one is specified \n",
    "if annot_dict_filename != '':\n",
    "    open_conversion_file = open(annot_dict_filename,'r')\n",
    "    lines = open_conversion_file.readlines()\n",
    "    for line in lines:\n",
    "        line = line.strip()\n",
    "        line = line.split('\\t')\n",
    "        annot_dict[line[0]]=line[1]\n",
    "    open_conversion_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get Enrichr Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get Enrichr Results \n",
    "# Takes a gene list and Enrichr libraries as input \n",
    "def Enrichr_API(enrichr_gene_list, all_libraries):\n",
    "\n",
    "\n",
    "    all_terms = []\n",
    "    all_pvalues =[] \n",
    "    all_adjusted_pvalues = []\n",
    "\n",
    "\n",
    "    for library_name in all_libraries : \n",
    "        ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/addList'\n",
    "        genes_str = '\\n'.join(enrichr_gene_list)\n",
    "        description = 'Example gene list'\n",
    "        payload = {\n",
    "            'list': (None, genes_str),\n",
    "            'description': (None, description)\n",
    "        }\n",
    "\n",
    "        response = requests.post(ENRICHR_URL, files=payload)\n",
    "        if not response.ok:\n",
    "            raise Exception('Error analyzing gene list')\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "\n",
    "        ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/enrich'\n",
    "        query_string = '?userListId=%s&backgroundType=%s'\n",
    "        user_list_id = data['userListId']\n",
    "        gene_set_library = library_name\n",
    "        response = requests.get(\n",
    "            ENRICHR_URL + query_string % (user_list_id, gene_set_library)\n",
    "         )\n",
    "        if not response.ok:\n",
    "            raise Exception('Error fetching enrichment results')\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "\n",
    "        results_df  = pd.DataFrame(data[library_name][0:5])\n",
    "        all_terms.append(list(results_df[1]))\n",
    "        all_pvalues.append(list(results_df[2]))\n",
    "        all_adjusted_pvalues.append(list(results_df[6]))\n",
    "    \n",
    "    return([all_terms,all_pvalues,all_adjusted_pvalues])\n",
    "\n",
    "\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get enrichr results\n",
    "results = Enrichr_API(genes,enrichr_libraries)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Enrichr Results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function plots results \n",
    "# Takes all terms, all p-values, all adjusted p-values, plot title, Enrichr libraries , and specified figure format\n",
    "def enrichr_figure(all_terms,all_pvalues, all_adjusted_pvalues, plot_name, all_libraries, fig_format): \n",
    "    \n",
    "    # rows and columns depend on number of Enrichr libraries submitted \n",
    "    rows = []\n",
    "    cols = []\n",
    "    \n",
    "    # If only 1 Enrichr library selected, make simple plot \n",
    "    if len(all_libraries)==1:\n",
    "        #fig,axes = plt.subplots(1, 1,figsize=[8.5,6])\n",
    "        plt.figure(figsize=(12,6))\n",
    "        rows = [0]\n",
    "        cols = [0]\n",
    "        i = 0 \n",
    "        bar_colors = ['tomato' if (x < 0.05) else 'lightgrey' for x in all_pvalues[i]]\n",
    "        fig = sns.barplot(x=np.log10(all_pvalues[i])*-1, y=all_terms[i], palette=bar_colors)\n",
    "        fig.axes.get_yaxis().set_visible(False)\n",
    "        fig.set_title(all_libraries[i].replace('_',' '),fontsize=26)\n",
    "        fig.set_xlabel('-Log10(p-value)',fontsize=25)\n",
    "        fig.xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "        fig.tick_params(axis='x', which='major', labelsize=20)\n",
    "        if max(np.log10(all_pvalues[i])*-1)<1:\n",
    "            fig.xaxis.set_ticks(np.arange(0, max(np.log10(all_pvalues[i])*-1), 0.1))\n",
    "        for ii,annot in enumerate(all_terms[i]):\n",
    "            if annot in annot_dict.keys():\n",
    "                annot = annot_dict[annot]\n",
    "            if all_adjusted_pvalues[i][ii] < 0.05:\n",
    "                annot = '  *'.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))]) \n",
    "            else:\n",
    "                annot = '  '.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))])\n",
    "\n",
    "            title_start= max(fig.axes.get_xlim())/200\n",
    "            fig.text(title_start,ii,annot,ha='left',wrap = True, fontsize = 26)\n",
    "            fig.patch.set_edgecolor('black')  \n",
    "            fig.patch.set_linewidth('2')\n",
    "        \n",
    "    \n",
    "    # If there are an even number of Enrichr libraries below 6\n",
    "    # Plots 1x2 or 2x2\n",
    "    else:\n",
    "        if len(all_libraries) % 2 == 0 and len(all_libraries) < 5:\n",
    "                for i in range(0,int(len(all_libraries)/2)):    \n",
    "                    rows = rows + [i]*2\n",
    "                    cols = list(range(0,2))*int(len(all_libraries)/2)    \n",
    "                fig, axes = plt.subplots(len(np.unique(rows)), len(np.unique(cols)),figsize=[7,int(2* len(np.unique(rows)))]) \n",
    "    \n",
    "        \n",
    "        # All other # of libraries 6 and above will have 3 columns and a flexible number of rows to accomodate all plots\n",
    "        else:\n",
    "            for i in range(0,int(np.ceil(len(all_libraries)/3))):\n",
    "                rows = rows + [i]*3\n",
    "                cols = list(range(0,3))*int(np.ceil(len(all_libraries)/3))\n",
    "            fig, axes = plt.subplots(len(np.unique(rows)), len(np.unique(cols)),figsize=[8,int(2* len(np.unique(rows)))])\n",
    "           \n",
    "        # If final figure only has one row...\n",
    "        if len(np.unique(rows))==1:\n",
    "            for i,library_name in enumerate(all_libraries):\n",
    "                bar_colors = ['tomato' if (x < 0.05) else 'lightgrey' for x in all_pvalues[i]]\n",
    "                sns.barplot(x=np.log10(all_pvalues[i])*-1, y=all_terms[i],ax=axes[i], palette=bar_colors)\n",
    "                axes[i].axes.get_yaxis().set_visible(False)\n",
    "                axes[i].set_title(library_name.replace('_',' '),fontsize=36)\n",
    "                axes[i].set_xlabel('-Log10(p-value)',fontsize=35)\n",
    "                axes[i].xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "                axes[i].tick_params(axis='x', which='major', labelsize=30)\n",
    "                if max(np.log10(all_pvalues[i])*-1)<1:\n",
    "                    axes[i].xaxis.set_ticks(np.arange(0, max(np.log10(all_pvalues[i])*-1), 0.1))\n",
    "                for ii,annot in enumerate(all_terms[i]):\n",
    "                    if annot in annot_dict.keys():\n",
    "                        annot = annot_dict[annot]\n",
    "                    if all_adjusted_pvalues[i][ii] < 0.05:\n",
    "                        annot = '  *'.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))]) \n",
    "                    else:\n",
    "                        annot = '  '.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))])\n",
    "\n",
    "                    title_start= max(axes[i].axes.get_xlim())/200\n",
    "                    axes[i].text(title_start,ii,annot,ha='left',wrap = True, fontsize = 36)\n",
    "                    axes[i].patch.set_edgecolor('black')  \n",
    "                    axes[i].patch.set_linewidth('2')\n",
    "\n",
    "            plt.subplots_adjust(top=4.5, right = 4.7,wspace = 0.03,hspace = 0.2)\n",
    "\n",
    "\n",
    "        # If the final figure has more than one row...\n",
    "        else:\n",
    "\n",
    "\n",
    "            for i,library_name in enumerate(all_libraries):\n",
    "                bar_colors = ['tomato' if (x < 0.05) else 'lightgrey' for x in all_pvalues[i]]\n",
    "                sns.barplot(x=np.log10(all_pvalues[i])*-1, y=all_terms[i],ax=axes[rows[i],cols[i]], palette=bar_colors)\n",
    "                axes[rows[i],cols[i]].axes.get_yaxis().set_visible(False)\n",
    "                axes[rows[i],cols[i]].set_title(library_name.replace('_',' '),fontsize=36)\n",
    "                axes[rows[i],cols[i]].set_xlabel('-Log10(p-value)',fontsize=35)\n",
    "                axes[rows[i],cols[i]].xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "                axes[rows[i],cols[i]].tick_params(axis='x', which='major', labelsize=30)\n",
    "                if max(np.log10(all_pvalues[i])*-1)<1:\n",
    "                    axes[rows[i],cols[i]].xaxis.set_ticks(np.arange(0, max(np.log10(all_pvalues[i])*-1), 0.1))\n",
    "                for ii,annot in enumerate(all_terms[i]):\n",
    "                    if annot in annot_dict.keys():\n",
    "                        annot = annot_dict[annot]\n",
    "                    if all_adjusted_pvalues[i][ii] < 0.05:\n",
    "                        annot = '  *'.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))]) \n",
    "                    else:\n",
    "                        annot = '  '.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii],precision=2)))])\n",
    "\n",
    "                    title_start= max(axes[rows[i],cols[i]].axes.get_xlim())/200\n",
    "                    axes[rows[i],cols[i]].text(title_start,ii,annot,ha='left',wrap = True, fontsize = 36)\n",
    "                    axes[rows[i],cols[i]].patch.set_edgecolor('black')  \n",
    "                    axes[rows[i],cols[i]].patch.set_linewidth('2')\n",
    "\n",
    "            plt.subplots_adjust(top=4.8, right = 4.7,wspace = 0.03,hspace = 0.2)\n",
    "\n",
    "\n",
    "        # If >6 libraries are chosen and is not a multiple of 3, delete empty plots\n",
    "        if len(np.unique(rows))*len(np.unique(cols)) != len(all_libraries):\n",
    "            diff = (len(np.unique(rows))*len(np.unique(cols))) - len(all_libraries)\n",
    "            for i in range (1,int(diff+1)):\n",
    "                fig.delaxes(axes[rows[-i]][cols[-i]])\n",
    "\n",
    "    # Save results \n",
    "    if plot_name != '':\n",
    "        plot_name = str(plot_name + '_')\n",
    "    plt.savefig(str(plot_name.replace(' ','_')+'enrichment_results_' + date.today().strftime(\"%Y%m%d\") + '.'+fig_format.lower() ),bbox_inches = 'tight')\n",
    "    \n",
    "    # Show plot \n",
    "    plt.show()\n",
    "    \n",
    "    \n",
    "               "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichr_figure(results[0],results[1],results[2],output_file_name, enrichr_libraries,figure_file_format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
