{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Augmented Pathway Enrichment Analysis (APEA)\n",
    "The APEA Appyter performs enrichment analysis against 4 major pathway databases given an input set of genes. The APEA Appyter performs enrichment analysis with the Fisher's exact test, and reports the results as bar graphs and sorted tables. By augmenting the pathway databases with gene-gene similarity matrices, APEA improves pathway enrichment analysis while also providing additional genes that are likely members of the enriched pathways and are relevant to the biological process under investigation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from maayanlab_bioinformatics.enrichment.crisp import enrich_crisp, fisher_overlap\n",
    "\n",
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from IPython.display import display, FileLink, Markdown, HTML\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from collections import OrderedDict\n",
    "import urllib\n",
    "\n",
    "import base64\n",
    "import csv\n",
    "import itertools\n",
    "import json\n",
    "import requests\n",
    "import time\n",
    "import urllib\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "{% do SectionField(\n",
    "    name='Set_Section',\n",
    "    title='Submit Your Set',\n",
    "    subtitle='Upload a text file containing your set or copy and paste your set into the text box below (one item per row). You can also try the default set provided.',\n",
    "    img='gene-library-3.png'\n",
    "    \n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='Similarity_Section',\n",
    "    title='Choose Your Similarity',\n",
    "    subtitle='Choose a similarity for your augmented pathway enrichment analysis.',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% set set_kind = TabField(\n",
    "    name='set_kind',\n",
    "    label='Set',\n",
    "    default='Try Example Gene Set',\n",
    "    description='Paste or upload your set',\n",
    "    choices={\n",
    "        'Paste': [ \n",
    "            TextField(\n",
    "                name='set_input1',\n",
    "                label='Set',\n",
    "                default='',\n",
    "                description='Paste your set (one item per row). Names in the set should match the names in the GMT file.',\n",
    "                section = 'Set_Section'\n",
    "            )\n",
    "        ],\n",
    "        \n",
    "        'Upload': [\n",
    "            FileField(\n",
    "                name='set_filename',\n",
    "                label='Set File',\n",
    "                default='',\n",
    "                description='Upload your set as a text file (one item per row). Names in the set should match the names in the GMT file.',\n",
    "                section = 'Set_Section'\n",
    "            ),\n",
    "        ],\n",
    "        \n",
    "        'Try Example Gene Set': [\n",
    "            TextField(\n",
    "                name='set_input2',\n",
    "                label='Set',\n",
    "                default='TAAR9\\nEBF2\\nWDR78\\nRRAGA\\nSPATA18\\nSPINT2\\nMRGPRD\\nCD9\\nRBP1\\nCYB5RL\\nMXRA8\\nPM20D1\\nITIH5\\nEPAS1\\nAHCYL2\\nPANK2\\nPON2\\nLRP5\\nSLC5A3\\nNSL1\\nCLDN2\\nLRP8\\nAQP1\\nCLDN1\\nTMEM72\\nGNG4\\nNHLH2\\nC10ORF107\\nS100A13\\nLY6G6C\\nPOF1B\\nWLS\\nC2ORF82\\nFZD4\\nCOG7\\nFZD6\\nFOXF1\\nFZD7\\nERLIN2\\nTYSND1\\nACADSB\\nOR51I2\\nPARP12\\nPPFIBP2\\nATP4A\\nALDH7A1\\nTCN2\\nSLCO5A1\\nSFXN4\\nPRR15\\nMOXD1\\nCAPSL\\nCOL13A1\\nC1ORF177\\nWFDC2\\nSLC6A2\\nDNALI1\\nTNS1\\nLGALS2\\nT\\nBLOC1S1\\nHMOX1\\nPDK4\\nLRAT\\nMNX1\\nSLC19A1\\nHOXC9\\nSCARF2\\nAS3MT\\nARGLU1\\nACE\\nANXA2\\nCARD9\\nPAX7\\nSORCS1\\nRAB33B\\nPHOX2A\\nKIF9\\nCLDN16\\nPTPRB\\nID3\\nITPKB\\nNCR1\\nGAS6\\nCC2D1B\\nATR\\nMYCBP\\nIGSF6\\nTPH1\\nWFIKKN2\\nIGSF5\\nACY3\\nMAOA\\nCAB39L\\nCTSZ\\nPRDM16\\nCYP7A1\\nLIMD1\\nTMEM27\\nSLC22A18\\nKRT28\\nTIMP3\\nEMB\\nRNF152\\nPLEKHN1\\nCLIC3\\nSTRA6\\nCTSC\\nCGNL1\\nPARP4\\nTMEM176B\\nELOVL7\\nSORBS3\\nGPR4\\nF5\\nGUCA2B\\nSERPINB6\\nHADHB\\nFOXR1\\nNBR1\\nSHKBP1\\nRLIM\\nDHRS13\\nHRSP12\\nCD63\\nCCL11\\nF13A1\\nFAM69C\\nKCNA7\\nHCCS\\nGUCA1A\\nADAMTSL2\\nLMAN1\\nING3\\nEGFLAM\\nSCML4\\nOLFML1\\nSOSTDC1\\nCTNNA1\\nC16ORF78\\nFADS1\\nCCDC157\\nPDGFRB\\nCA12\\nCD164\\nPRLR\\nLRRC69\\nUNC5CL\\nMPEG1\\nSLC31A1\\nTECRL\\nVCAM1\\nATP11A\\nUBXN10\\nZNF558\\nDYDC1\\nCD69\\nS100A8\\nFIGF\\nPHLDB2\\nERVFRD-1\\nCD82\\nASB14\\nGPR65\\nVWCE\\nTEKT1\\nTEKT4\\nMSX1\\nSLC16A9\\nZNF423\\nCA14\\nIGFBP2\\nSLC30A7\\nLRRC46\\nPDIA2\\nPPEF1\\nEPHX1\\nFANCM\\nRBPMS\\nTTC21A\\nMR1\\nDDX52\\nLSM5\\nKRT31\\nMAVS\\nTMEM237\\nSMO\\nC6ORF118\\nPGPEP1L\\nIL7R\\nC21ORF62\\nC11ORF97\\nDOCK6\\nAKNA\\nISYNA1\\nCD151\\nCBFB\\nPYROXD2\\nSLC2A1\\nGSTCD\\nLGALS3BP\\nHIGD1B\\nAK7\\nLTBP1\\nARHGAP5\\nRGS5\\nSALL1\\nCOBLL1\\nFHAD1\\nMAEL\\nBTLA\\nIGFBP7\\nODF1\\nACAA2\\nKL\\nTTC16\\nEMX2\\nTTC12\\nGGH\\nCCDC37\\nCFLAR\\nGPR98\\nLAMB2\\nBICC1\\nBMP6\\nCUL4B\\nDNAJC3\\nSP1\\nDAP\\nDNAJC1\\nPIKFYVE\\nDMRTA1\\nALPL\\nMTRF1L\\nBCAR3\\nKDM5D\\nSHC4\\nTTC25\\nDBH\\nDBI\\nCHD1\\nWNT6\\nSPN\\nTTC23L\\nPLTP\\nCYP26B1\\nCASP6\\nTMEM204\\nTMEM207\\nCCDC180\\nCCDC34\\nCA9\\nOVGP1\\nPLEKHG2\\nCPT1A\\nPLEKHG3\\nMYO10\\nRNASET2\\nTBC1D9\\nNAGA\\nPCOLCE\\nMUT\\nFOXJ1\\nSOD3\\nATOX1\\nKRT73\\nSNTB1\\nRP2\\nRPIA\\nCOL8A1\\nALS2\\nCOL8A2\\nSMPDL3A\\nPCOLCE2\\nSLC25A13\\nTAF3\\nFOLR1\\nITGB2\\nHEMGN\\nPRPS2\\nSLC24A5\\nFLT1\\nALAS2\\nLSP1\\nSYCP2\\nSEMA3B\\nETFB\\nPRELP\\nZBTB40\\nPBXIP1\\nSLC4A5\\nCLN8\\nEFS\\nTTR\\nRBM3\\nHECTD3\\nNAGLU\\nALDH2\\nCTNNAL1\\nPCBD1\\nCYTH2',\n",
    "                description='Paste your set (one item per row). Names in the set should match the names in the GMT file.',\n",
    "                section = 'Set_Section'\n",
    "            )\n",
    "        ],\n",
    "        \n",
    "    },\n",
    "    section = 'Set_Section',\n",
    ") %}\n",
    "\n",
    "{% set input_name = StringField(\n",
    "    name='input_name', \n",
    "    label='Input Set Name', \n",
    "    default='M1 Genes', \n",
    "    description='Input net name', \n",
    "    section='Set_Section')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set similarity_kind = TabField(\n",
    "    name='similarity_kind',\n",
    "    label='Similarity',\n",
    "    default='Select a Similarity',\n",
    "    description='',\n",
    "    choices={\n",
    "        'Select a Similarity': [\n",
    "            ChoiceField(\n",
    "                name='similarity', \n",
    "                description='Select one similarity for enrichment analysis.', \n",
    "                label='Similarity', \n",
    "                default='ARCHS4_mRNA_Coexpression', \n",
    "                section = 'Similarity_Section',\n",
    "                choices=[\n",
    "                    'ARCHS4_mRNA_Coexpression',\n",
    "                    'Enrichr_Gene_Coocurrence',\n",
    "                    'GeneRIF_Coocurrence',\n",
    "                    'Tagger_Cooccurence',\n",
    "                    'Proteomics_Coexpression',\n",
    "                ]\n",
    "            )\n",
    "        ],\n",
    "        \n",
    "        \n",
    "    },\n",
    "    section = 'Similarity_Section',\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Input Set and Parameters\n",
    "\n",
    "{%- if set_kind.raw_value == 'Paste' or set_kind.raw_value == 'Try Example Gene Set'%}\n",
    "set_input = {{ set_kind.value[0] }}\n",
    "{%- else %}\n",
    "set_filename = {{ set_kind.value[0] }}\n",
    "{%- endif %}\n",
    "\n",
    "input_name = \"{{ input_name }}\"\n",
    "similarity_kind = \"Select a Similarity\"\n",
    "similarity_name = \"{{ similarity_kind.value[0] }}\"\n",
    "data_dir=\"https://appyters.maayanlab.cloud/storage/Augmented_Pathway_Enrichment_Analysis/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Table Parameters\n",
    "significance_value = 0.05\n",
    "display_topk = 20\n",
    "\n",
    "# Bar Chart Parameters\n",
    "figure_file_format = ['png', 'svg']\n",
    "color = 'deepskyblue'\n",
    "topk = 10\n",
    "\n",
    "# Tiebow Parameters\n",
    "# display_num = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if set_kind.raw_value == 'Paste' or set_kind.raw_value == 'Try Example Gene Set' %}\n",
    "items = set_input.split('\\n')\n",
    "items = [x.strip() for x in items]\n",
    "{%- else %}\n",
    "open_set_file = open(set_filename,'r')\n",
    "lines = open_set_file.readlines()\n",
    "items = [x.strip() for x in lines]\n",
    "open_set_file.close()\n",
    "{%- endif %}\n",
    "\n",
    "# remove duplicates in items\n",
    "items = list(OrderedDict.fromkeys(items))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#loading and validating data\n",
    "def load(similarity_name, items, separate=False):\n",
    "\n",
    "    similarity_data, pathways = load_similarity(similarity_name, separate)\n",
    "    # to upper case\n",
    "    items = [x.upper() for x in items]\n",
    "    validate_inputs(items, similarity_data)\n",
    "\n",
    "    return similarity_data, pathways, items\n",
    "\n",
    "def load_similarity(similarity_name, separate=False):\n",
    "    similarity_filename = similarity_name+'.csv'\n",
    "    pathway_filename = similarity_name.split('_')[0]+'_Pathways.txt'\n",
    "    \n",
    "    if separate == False:\n",
    "        pathway_filename = 'All_Pathways.txt'\n",
    "    \n",
    "    similarity_data = dict()\n",
    "    pathways = []\n",
    "    lib=[]\n",
    "\n",
    "    file = urllib.request.urlopen(data_dir+pathway_filename)\n",
    "    for line in file:\n",
    "        decoded_line = line.decode(\"utf-8\")[:-1]\n",
    "        pathways.append(decoded_line)\n",
    "        \n",
    "    response = urllib.request.urlopen(data_dir+similarity_filename)\n",
    "    lines = [l.decode('utf-8') for l in response.readlines()]\n",
    "    cr = csv.reader(lines)\n",
    "    for row in cr:\n",
    "        lib.append(row)\n",
    "    similarity_data=lib\n",
    "    \n",
    "    \n",
    "    return similarity_data, pathways\n",
    "\n",
    "def validate_inputs(items, similarity_data):\n",
    "    if len(items) == 0:\n",
    "        raise Exception('No items in the input set. Please check the background information.') \n",
    "    if not bool(set(items) & set(itertools.chain(*similarity_data))):\n",
    "        raise Exception('No matches in the input set and similarity.')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Enrichment analysis\n",
    "def get_similarity_iter(similarity_data, pathways):\n",
    "    for i, gset in enumerate(similarity_data):\n",
    "        yield pathways[i], gset\n",
    "\n",
    "def get_enrichment_results(items, similarity_data, pathways):\n",
    "    return sorted(enrich_crisp(items, get_similarity_iter(similarity_data, pathways), 20000, True), key=lambda r: r[1].pvalue)\n",
    "\n",
    "\n",
    "def get_pvalue(row, unzipped_results, all_results):\n",
    "    if row['Name'] in list(unzipped_results[0]):\n",
    "        index = list(unzipped_results[0]).index(row['Name'])\n",
    "        return all_results[index][1].pvalue\n",
    "    else:\n",
    "        return 1\n",
    "    \n",
    "def get_qvalue(p_vals):\n",
    "    r = multipletests(p_vals, method=\"fdr_bh\")\n",
    "    return r[1]\n",
    "    \n",
    "def get_values(obj_list):\n",
    "    pvals = []\n",
    "    odds_ratio = []\n",
    "    n_overlap = []\n",
    "    overlap = []\n",
    "    for i in obj_list:\n",
    "        pvals.append(i.pvalue)\n",
    "        odds_ratio.append(i.odds_ratio)\n",
    "        n_overlap.append(i.n_overlap)\n",
    "        overlap.append(i.overlap)\n",
    "    return pvals, odds_ratio, n_overlap, overlap\n",
    "    \n",
    "def enrichment_analysis(items, similarity_data, pathways):    \n",
    "    all_results = get_enrichment_results(items, similarity_data, pathways)\n",
    "    unzipped_results = list(zip(*all_results))\n",
    "    pvals, odds_ratio, n_overlap, overlap = get_values(unzipped_results[1])\n",
    "    df = pd.DataFrame({\"Name\":unzipped_results[0], \"p value\": pvals, \\\n",
    "                       \"odds_ratio\": odds_ratio, \"n_overlap\": n_overlap, \"overlap\": overlap})\n",
    "    df[\"-log(p value)\"] = -np.log10(df[\"p value\"])\n",
    "    df[\"q value\"] = get_qvalue(df[\"p value\"].tolist())\n",
    "    return [list(unzipped_results[0])], [pvals], df\n",
    "\n",
    "def create_download_link(df, title = \"Download CSV file of this table\", filename = \"data.csv\"):  \n",
    "    csv = df.to_csv(index = False)\n",
    "    b64 = base64.b64encode(csv.encode())\n",
    "    payload = b64.decode()\n",
    "    html = '<a download=\"{filename}\" href=\"data:text/csv;base64,{payload}\" target=\"_blank\">{title}</a>'\n",
    "    html = html.format(payload=payload, title=title, filename=filename)\n",
    "    return HTML(html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bold_augmented(df):\n",
    "    h = df.to_html(index=False)\n",
    "    for i, row in df.iterrows():\n",
    "        pathway = pathways.index(row['Name'])\n",
    "        ol = ', '.join(row['overlap'])\n",
    "        ol2 = ol\n",
    "        for g in row['overlap']:\n",
    "            if not g in similarity_data[pathway][:-100]:\n",
    "                ol2 = ol2.replace(g,'<b>'+g+'*</b>')\n",
    "        h = h.replace(ol, ol2)\n",
    "    display(HTML(h))\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bar Chart Functions\n",
    "def enrichr_figure(all_terms, all_pvalues, all_qvalues, all_libraries, bar_color, topk=10): \n",
    "    all_terms = [all_terms[0][:topk]]\n",
    "    all_pvalues = [all_pvalues[0][:topk]]\n",
    "    all_qvalues = [all_qvalues[:topk]]\n",
    "    # Bar colors\n",
    "    if bar_color != 'lightgrey':\n",
    "        bar_color_not_sig = 'lightgrey'\n",
    "        edgecolor=None\n",
    "        linewidth=0\n",
    "    else:\n",
    "        bar_color_not_sig = 'white'\n",
    "        edgecolor='black'\n",
    "        linewidth=1    \n",
    "\n",
    "    plt.figure(figsize=(24, 12))\n",
    "    \n",
    "    i = 0\n",
    "    bar_colors = [bar_color if (x < 0.05) else bar_color_not_sig for x in all_pvalues[i]]\n",
    "    fig = sns.barplot(x=np.log10(all_pvalues[i])*-1, y=all_terms[i], alpha =.3, palette=bar_colors, edgecolor=edgecolor, linewidth=linewidth)\n",
    "    fig.axes.get_yaxis().set_visible(False)\n",
    "    fig.set_title(all_libraries[i].replace('_', ' '), fontsize=26)\n",
    "    fig.set_xlabel('−log₁₀(p‐value)', fontsize=25)\n",
    "    fig.tick_params(axis='x', which='major', labelsize=20)\n",
    "    if max(np.log10(all_pvalues[i])*-1)<1:\n",
    "        fig.xaxis.set_ticks(np.arange(0, max(np.log10(all_pvalues[i])*-1), 0.1))\n",
    "    for ii,annot in enumerate(all_terms[i]):\n",
    "        if all_qvalues[i][ii] < 0.05:\n",
    "            annot = '  *'.join([annot, str(str(np.format_float_scientific(all_qvalues[i][ii], precision=2)))]) \n",
    "        else:\n",
    "            annot = '  '.join([annot, str(str(np.format_float_scientific(all_qvalues[i][ii], precision=2)))])\n",
    "\n",
    "        title_start= max(fig.axes.get_xlim())/200\n",
    "        fig.text(title_start, ii, annot, ha='left', wrap = True, fontsize = 26)\n",
    "\n",
    "    fig.spines['right'].set_visible(False)\n",
    "    fig.spines['top'].set_visible(False)\n",
    "    # Save results \n",
    "    output_file_names = ['{0}_bar.{1}'.format(all_libraries[i], file_type) for file_type in figure_file_format]\n",
    "    for f in output_file_names:\n",
    "        plt.savefig(f, bbox_inches = 'tight')\n",
    "    \n",
    "    # Show plot \n",
    "    plt.show()\n",
    "    \n",
    "    \n",
    "    return output_file_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tiebow\n",
    "def tiebow(df, augmented=True):\n",
    "    i=0\n",
    "    p=df['Name'][i]\n",
    "    pathway = pathways.index(p)\n",
    "    if(len(p.split())>4):\n",
    "        p=add_breaks3(p)\n",
    "        \n",
    "    if len(items)>len(similarity_data[pathway]):\n",
    "        b=5000\n",
    "        a=5000*(len(items)/len(similarity_data[pathway]))\n",
    "    else:\n",
    "        a=5000\n",
    "        b=5000*(len(similarity_data[pathway])/len(items))\n",
    "    d=b*len(df['overlap'][i])/len(similarity_data[pathway])\n",
    "\n",
    "    plt.figure(figsize=(12,12))\n",
    "    G = nx.Graph()\n",
    "    G.add_node('Input\\n Set', size=a, pos=(-25,25), color='mediumorchid', s='s')\n",
    "    G.add_node('Pathway\\n Set', size=b, pos=(25,25), color='salmon', s='s')\n",
    "    G.add_node('', size=d, pos=(25,25), color='mediumorchid', s='s')\n",
    "    l=len(df['overlap'][0])\n",
    "    for j,g in enumerate(df['overlap'][i]):\n",
    "        c = 'lightskyblue'\n",
    "        if augmented and not g in similarity_data[pathway][:-100]:\n",
    "            c = 'lightgreen'\n",
    "        G.add_node(g, size= 15000, pos=(0,j*50/(l-1)), color=c, s='_')\n",
    "        G.add_edge(g, 'Input\\n Set')\n",
    "        G.add_edge(g, 'Pathway\\n Set')\n",
    "\n",
    "    # #         bbox=dict(facecolor='none', edgecolor='black', boxstyle='round,pad=0.2'),\n",
    "    #         alpha=0.5, with_labels=True, node_shape='_')\n",
    "\n",
    "    nodeShapes = set((aShape[1]['s'] for aShape in G.nodes(data = True)))\n",
    "    #For each node class...\n",
    "    for aShape in nodeShapes:\n",
    "        if aShape=='_':\n",
    "            l=50\n",
    "        else:\n",
    "            l=1\n",
    "        #...filter and draw the subset of nodes with the same symbol in the positions that are now known through the use of the layout.\n",
    "        nx.draw_networkx_nodes(G,nx.get_node_attributes(G,'pos'),\n",
    "                               node_shape = aShape, alpha = 1,\n",
    "                               linewidths = l,\n",
    "                               node_size = [sNode[1]['size'] for sNode in filter(lambda x: x[1][\"s\"]==aShape,G.nodes(data = True))],\n",
    "                               node_color = [sNode[1]['color'] for sNode in filter(lambda x: x[1][\"s\"]==aShape,G.nodes(data = True))],\n",
    "                               nodelist = [sNode[0] for sNode in filter(lambda x: x[1][\"s\"]==aShape,G.nodes(data = True))])\n",
    "\n",
    "    #Finally, draw the edges between the nodes\n",
    "    nx.draw_networkx_edges(G,nx.get_node_attributes(G,'pos'))    \n",
    "    nx.draw_networkx_labels(G,nx.get_node_attributes(G,'pos'), font_size=20)\n",
    "\n",
    "    \n",
    "\n",
    "    # Legend\n",
    "    if augmented:\n",
    "        plt.text(-30,-4.5, 'P-value for augmented pathway: {:.2e}\\nQ-value for augmented pathway: {:.3e}'.format(df['p value'][i],df['q value'][i]), fontsize=18)\n",
    "        plt.scatter([],[], c='lightskyblue', label='Original Pathway Member')\n",
    "        plt.scatter([],[], c='lightgreen', label='Augmented Pathway Member')\n",
    "        l=plt.legend(loc=4, scatterpoints=1, fontsize=15)\n",
    "        l.legendHandles[0]._sizes= [100]\n",
    "        l.legendHandles[1]._sizes= [100]\n",
    "    else:\n",
    "        plt.text(-30,-4.5, 'P-value for unaugmented pathway: {:.2e}\\nQ-value for unaugmented pathway: {:.3e}'.format(df['p value'][i],df['q value'][i]), fontsize=18)\n",
    "    \n",
    "    plt.text(-25,40, input_name+'\\n\\u2193', fontsize=20,\n",
    "        horizontalalignment=\"center\")\n",
    "    plt.text(25,40, p+'\\n\\u2193', fontsize=20,\n",
    "        horizontalalignment=\"center\")\n",
    "     \n",
    "    plt.axis('off')\n",
    "    axis = plt.gca()\n",
    "    axis.set_xlim([1.3*x for x in axis.get_xlim()])\n",
    "#     axis.set_ylim([1.2*y for y in axis.get_ylim()])\n",
    "    plt.tight_layout() \n",
    "    \n",
    "    # Save results \n",
    "    output_file_names = ['{0}_tiebow.{1}'.format(df['Name'][i], file_type) for file_type in figure_file_format]\n",
    "    for f in output_file_names:\n",
    "        plt.savefig(f, bbox_inches = 'tight')\n",
    "        \n",
    "    plt.show()\n",
    "    \n",
    "    return output_file_names\n",
    "    \n",
    "    \n",
    "#     display(tab)\n",
    "\n",
    "\n",
    "def add_breaks3(string):\n",
    "    words = string.split()\n",
    "    grouped_words = [' '.join(words[i: i + 3]) for i in range(0, len(words), 3)]\n",
    "    return '\\n'.join(grouped_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec \n",
    "display(Markdown(f\"The bar charts display the top {topk} enriched terms for the chosen pathway database similarity based on their p-values, along with their q-values. Colored bars correspond to terms with significant p-values (<0.05). An asterisk next to a q-value indicates the term also has a significant q-value (<0.05). The bar chart is downloadable as an image in the PNG and SVG formats. \"))\n",
    "display(Markdown(f\"The tables display the top {display_topk} enrichment analysis matching pathways for the chosen database. The table contains the pathway name, p-value, odds ratio, the number of overlapping genes, the overlapping genes, and -log(p-value). The bolded genes with an asterisk in the overlap column are the additional genes added to the pathway via the augmentation. The table is sorted by p-values in ascending order. The full results are downloadable in CSV format.\"))\n",
    "display(Markdown(f\"The tiebow graphs of the overlapping genes between the input set and the top pathway result based on p-values is shown. For the augmented results, the genes ar colored by whether they were in the original pathway set or from augmentation. The set squares are sizes proportionally, and the purple square within the pathway set node indicates the size of overlap relative to the pathway set. The tiebow graph is downloadable as an image in the PNG and SVG formats.\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec \n",
    "plasma = mpl.cm.get_cmap('plasma')\n",
    "display(Markdown(f\"# Enrichment Analysis with Unaugmented Pathway Databases\"))\n",
    "display(Markdown(f\"Enrichment analysis results are separated for each unaugmented database.\"))\n",
    "db = ['KEGG', 'PANTHER', 'Reactome', 'WikiPathways']\n",
    "for i,l in enumerate(db):\n",
    "    file = l+'_Unaugmented'\n",
    "    display(Markdown(f\"### {l+' Unaugmented'}\"))\n",
    "    similarity_data, pathways, items = load(file, items, separate=True)\n",
    "    results, pvals, results_df = enrichment_analysis(items, similarity_data, pathways)\n",
    "\n",
    "    output_file_names = enrichr_figure(results, pvals, results_df[\"q value\"].tolist(), [file], plasma(i*0.25), topk)\n",
    "    display(Markdown(f\"*Figure {2*i+1}. Bar chart of the top {topk} enriched terms from the {file.replace('_', ' ')} similarity based on p-values, along with their corresponding q-values. Colored bars correspond to terms with significant p-values (<0.05). An asterisk next to a q-value indicates the term also has a significant q-value (<0.05)*\"))     \n",
    "\n",
    "    # Download Bar Chart\n",
    "    for k, file in enumerate(output_file_names):\n",
    "        display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[k] + ': ')))\n",
    "\n",
    "    if 'p value' in results_df.columns:\n",
    "        sorted_df = results_df.sort_values(by = ['p value'])\n",
    "        filtered_df = sorted_df.iloc[:display_topk]\n",
    "        if len(filtered_df) != 0:\n",
    "            display(HTML(filtered_df.to_html(index = False)))\n",
    "            display(Markdown(f\"*Table {i+1}. Enrichment analysis results from the {l.replace('_', ' ')} similarity.*\"))        \n",
    "            display(create_download_link(sorted_df))\n",
    "            \n",
    "    output_file_names = tiebow(sorted_df, augmented=False)\n",
    "    display(Markdown(f\"*Figure {2*i+2}. Tiebow graph of the overlapping genes between the input set and the top pathway result based on p-values. The purple square within the pathway set indicates the size of overlap relative to the pathway set.*\"))     \n",
    "\n",
    "    # Download Tiebow Graph\n",
    "    for i, file in enumerate(output_file_names):\n",
    "        display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec \n",
    "plasma = mpl.cm.get_cmap('plasma')\n",
    "display(Markdown(f\"# Enrichment Analysis by Pathway Database Augmented with {similarity_name.replace('_', ' ')}\"))\n",
    "display(Markdown(f\"Enrichment analysis results are separated for each augmented pathway database.\"))\n",
    "suf = similarity_name.split('_')[0]\n",
    "db = ['KEGG', 'PANTHER', 'Reactome', 'WikiPathways']\n",
    "for i,l in enumerate(db):\n",
    "    file = l+'_Augmented_with_'+suf\n",
    "    display(Markdown(f\"### {l+' Augmented'}\"))\n",
    "    similarity_data, pathways, items = load(file, items, separate=True)\n",
    "    results, pvals, results_df = enrichment_analysis(items, similarity_data, pathways)\n",
    "\n",
    "    output_file_names = enrichr_figure(results, pvals, results_df[\"q value\"].tolist(), [l+'_Augmented_with_'+similarity_name], plasma(i*0.25), topk)\n",
    "    display(Markdown(f\"*Figure {2*i+9}. Bar chart of the top {topk} enriched terms from the {l+' similarity augmented with '+similarity_name.replace('_', ' ')} based on p-values, along with their corresponding q-values. Colored bars correspond to terms with significant p-values (<0.05). An asterisk next to a q-value indicates the term also has a significant q-value (<0.05)*\"))     \n",
    "\n",
    "    # Download Bar Chart\n",
    "    for k, file in enumerate(output_file_names):\n",
    "        display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[k] + ': ')))\n",
    "\n",
    "    if 'p value' in results_df.columns:\n",
    "        sorted_df = results_df.sort_values(by = ['p value'])\n",
    "        filtered_df = sorted_df.iloc[:display_topk]\n",
    "        if len(filtered_df) != 0:\n",
    "            bold_augmented(filtered_df)\n",
    "            display(Markdown(f\"*Table {i+5}. Enrichment analysis results from the augmented {l} similarity. Bolded terms with an asterisk are genes from augmenting the {l} pathways with {similarity_name.replace('_', ' ')}.*\"))        \n",
    "            display(create_download_link(sorted_df))\n",
    "            \n",
    "    output_file_names = tiebow(sorted_df)\n",
    "    display(Markdown(f\"*Figure {2*i+10}. Tiebow graph of the overlapping genes between the input set and the top pathway result based on p-values. The purple square within the pathway set indicates the size of overlap relative to the pathway set.*\"))     \n",
    "\n",
    "    # Download Tiebow Graph\n",
    "    for i, file in enumerate(output_file_names):\n",
    "        display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "display(Markdown(f\"# Combined Enrichment Analysis with All 4 Unaugmented Databases\"))\n",
    "display(Markdown(f\"Enrichment analysis based on the unaugmented data.\"))\n",
    "\n",
    "lib = 'Unaugmented'\n",
    "similarity_data, pathways, items = load(lib, items)\n",
    "results, pvals, results_df = enrichment_analysis(items, similarity_data, pathways)\n",
    "\n",
    "display(Markdown(f\"## Bar Chart\"))\n",
    "output_file_names = enrichr_figure(results, pvals, results_df[\"q value\"].tolist(), ['Combined Unaugmented Databases'], color, topk)\n",
    "display(Markdown(f\"*Figure 17. Bar chart of the top {topk} enriched terms across all 4 pathway databases based on p-values, along with their corresponding q-values. Colored bars correspond to terms with significant p-values (<0.05). An asterisk next to a q-value indicates the term also has a significant q-value (<0.05)*\"))     \n",
    "# Download Bar Chart\n",
    "for i, file in enumerate(output_file_names):\n",
    "    display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))\n",
    "\n",
    "display(Markdown(f\"## Results Table\"))\n",
    "if 'p value' in results_df.columns:\n",
    "    sorted_df = results_df.sort_values(by = ['p value'])\n",
    "    filtered_df = sorted_df.iloc[:display_topk]\n",
    "    if len(filtered_df) != 0:\n",
    "        display(HTML(filtered_df.to_html(index = False)))\n",
    "        display(Markdown(f\"*Table 9. Enrichment analysis results of {lib.replace('_', ' ')}*\"))        \n",
    "        display(create_download_link(sorted_df))    \n",
    "        \n",
    "output_file_names = tiebow(sorted_df, augmented=False)\n",
    "display(Markdown(f\"*Figure 18. Tiebow graph of the overlapping genes between the input set and the top pathway result based on p-values. The purple square within the pathway set indicates the size of overlap relative to the pathway set.*\"))     \n",
    "\n",
    "# Download Tiebow Graph\n",
    "for i, file in enumerate(output_file_names):\n",
    "    display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "display(Markdown(f\"# Combined Enrichment Analysis with All 4 Pathway Databases Augmented with {similarity_name.replace('_', ' ')}\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "## Bar Chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "similarity_data, pathways, items = load(similarity_name, items)\n",
    "results, pvals, results_df = enrichment_analysis(items, similarity_data, pathways)\n",
    "\n",
    "output_file_names = enrichr_figure(results, pvals, results_df[\"q value\"].tolist(), ['Combined Databases Augmented with '+similarity_name], color, topk)\n",
    "display(Markdown(f\"*Figure 19. Bar chart of the top {topk} enriched terms across all 4 pathway augmented databases based on p-values, along with their corresponding q-values. Colored bars correspond to terms with significant p-values (<0.05). An asterisk next to a q-value indicates the term also has a significant q-value (<0.05)*\"))     \n",
    "\n",
    "# Download Bar Chart\n",
    "for i, file in enumerate(output_file_names):\n",
    "    display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec    \n",
    "display(Markdown(f\"## Results Table\"))\n",
    "if 'p value' in results_df.columns:\n",
    "    sorted_df = results_df.sort_values(by = ['p value'])\n",
    "    filtered_df = sorted_df.iloc[:display_topk]\n",
    "    if len(filtered_df) != 0:\n",
    "        bold_augmented(filtered_df)\n",
    "        display(Markdown(f\"*Table 10. Enrichment analysis results from the augmented combined similarity. Bolded terms with an asterisk are genes from augmenting the pathways with {similarity_name.replace('_', ' ')}.*\")) \n",
    "        display(create_download_link(sorted_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "## Tiebow Graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec   \n",
    "output_file_names = tiebow(sorted_df)\n",
    "display(Markdown(f\"*Figure 20. Tiebow graph of the overlapping genes between the input set and the top pathway result based on p-values. The purple square within the pathway set indicates the size of overlap relative to the pathway set.*\"))     \n",
    "\n",
    "# Download Tiebow Graph\n",
    "for i, file in enumerate(output_file_names):\n",
    "    display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Link to Enrichr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get complete enrichment analysis results from Enrichr \n",
    "\n",
    "ENRICHR_URL = 'https://maayanlab.cloud/Enrichr/addList'\n",
    "genes_str = '\\n'.join(items)\n",
    "description = 'Input gene list'\n",
    "payload = {\n",
    "    'list': (None, genes_str),\n",
    "    'description': (None, description)\n",
    "}\n",
    "\n",
    "response = requests.post(ENRICHR_URL, files=payload)\n",
    "if not response.ok:\n",
    "    raise Exception('Error analyzing gene list')\n",
    "\n",
    "data = json.loads(response.text)\n",
    "short_id = data[\"shortId\"]\n",
    "\n",
    "\n",
    "if short_id:\n",
    "    url = 'https://maayanlab.cloud/Enrichr/enrich?dataset=' + short_id\n",
    "    display(HTML(f\"<div style='font-size:1.25rem; padding:1rem 0;'><a href='https://maayanlab.cloud/Enrichr/enrich?dataset={short_id}'>Access additional enrichment analysis results for the input set from the Enrichr website. </a></div>\"))\n",
    "else:\n",
    "    display(HTML(\"<div style='font-size:1.5rem; padding:1rem 0;'><b>No Enrichr results available for the current query</b></div>\"))\n",
    "    display(HTML(\"<div style='font-size:1rem; padding:1rem 0;'>Please try again with a different input list and/or similarity.\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
