# Augmented Pathway Enrichment Analysis

Augmented Pathway Enrichment Analysis Appyter performs enrichment analysis with augmented pathway libraries. The required input is a gene set. The Appyter performs the Fisher exact test to compute enrichment p-value and q-values, and reports the results as a bar graph along with a sorted table. If a combined library is chosen, the results for separate pathway databases are shown on the bottom as well. Unaugmented and augmented results are shown.

## **File Format**
To run the Appyter, a set of genes is required as input. There is an example gene set for use as well.

* Please input a set of items as follows:  
ITEM 1  
ITEM 2  
ITEM 3

## **Example**
* Example: Gene Set Enrichment Analysis  
The example input set is a gene set of up-regulated in a part of brain. The input set library (Example1_GeneSet_Allen_Brain_Atlas_scRNAseq_10x_2021.gmt) is about cell types in the primary motor cortex from human and mouse brains. The library contains up- or down-regulated genes in each cell type in the cortex. The library was generated by processing Allen Brain Map's Human/Mouse M1 10x single cell RNA-seq datasets (https://portal.brain-map.org/atlases-and-data/rnaseq/human-m1-10x). Through the Appyter results, you can find enriched cell types in the primary motor cortex in brain in the given gene set. Various gene set libraries are available at Enrichr (https://maayanlab.cloud/Enrichr/)  

