{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from appyter import magic\n",
    "magic.init(lambda _ = globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GeneSet Library Set Appyter\n",
    "This appyter is designed to perform basic statistics, analysis, and visualizations on a Gene Matrix Transpose (.GMT) file. This will allow bioinformatics researchers to analyze relationships between many different gene sets from several gene set libraries.\n",
    " To create your own GMT file, please see Enrichr. Enrichr, hosted by the Ma'ayan Laboratory at Mt. Sinai Icahn School of Medicine, is a collection of geneset libraries. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"p1008\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(\"p1008\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.1.0.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(\"p1008\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"p1008\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.1.0.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\nif (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"p1008\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np \n",
    "import pandas as pd\n",
    "import itertools \n",
    "import bokeh\n",
    "import pathlib\n",
    "import scanpy as sc\n",
    "from IPython.display import display, FileLink, HTML\n",
    "import anndata\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from maayanlab_bioinformatics.enrichment import crisp\n",
    "from collections import OrderedDict\n",
    "from bokeh.palettes import Category20\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.models import HoverTool, ColumnDataSource, RangeSlider\n",
    "from bokeh.plotting import figure, show, save, output_file\n",
    "import os\n",
    "\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "\n",
    "{% do SectionField(name='GMTSubmission', title='1. Submit a GMT file', subtitle = 'Sumbit a GMT (Gene Matrix Transpose file) for analysis.', img = 'bulb.png') %}\n",
    "{% do SectionField (name = 'UMAP_visualization', title = '2. Scatterplot Visualization', subtitle= 'Visualize relative Geneset similarities on an interactive scatterplot', img = 'bulb.png') %}\n",
    "{% do SectionField(name = 'Pairwise Similarity Table', title = '3. GeneSet Pairwise Similarity Table', subtitle = 'In this table, the value in row A, column B, is the size of the intersection of A and B. If you would like to get a list of genes from a specific intersection of two library terms, please see the Intersection Search Section.', img = 'bulb.png') %}\n",
    "{% do SectionField(name = 'Jaccard Similarity Table', title = '4. Jaccard Similarity Table', subtitle = '##TODO', img = 'bulb.png') %}\n",
    "{% do SectionField(name = 'Intersection Search', title = '5. Gene Intersection Search', subtitle = '###TODO', img = 'bulb.png')%}\n",
    "{% do SectionField(name = 'GMT Descriptive Statistics', title = '6. Descriptive Statistics', subtitle = '#TODO', img = 'bulb.png') %}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Submitted Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "```python\n",
       "gs = 'static/gene_sets_for_breast_cancer.gmt'\n",
       "int_tbl = True\n",
       "jaccard_tbl = True\n",
       "umap = True\n",
       "umap_num_neighbors = 5\n",
       "umap_maxdf = 0.5\n",
       "umap_mindf = 0.25\n",
       "```"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set gs = TabField(\n",
    "    name='gs_type',\n",
    "    label='Gene Sets',\n",
    "    default='Upload',\n",
    "    choices={'Upload': [\n",
    "            FileField(\n",
    "                name='gs',\n",
    "                label='Gene Set Files',\n",
    "                default='static/gene_sets_for_breast_cancer.gmt',\n",
    "                example={\n",
    "                    'example.gmt': url_for('static', filename = 'Geneshot_PainGenes.gmt')\n",
    "                }\n",
    "            ),\n",
    "        ],},\n",
    "    \n",
    "section = 'GMTSubmission',)%}\n",
    "\n",
    "gs = {{gs.value[0]}}\n",
    "\n",
    "\n",
    "\n",
    "int_tbl = {{BoolField(name = 'SimilarityTbl', label = 'Intersection Size Table', default = 'true', description = 'In this table, the value in row A, column B, is the size of the intersection of A and B. If you would like to get a list of genes from a specific intersection of two library terms, please see the Intersection Search Section. Select \\'Yes\\' if you would like to generate a Intersection Size Table. Otherwise, select \\'No\\'', section = 'Pairwise Similarity Table') }}\n",
    "jaccard_tbl = {{BoolField(name = 'JaccardTbl', label = 'Jaccard Similarity Table', default = 'true', description = '##TODO', section = 'Jaccard Similarity Table') }}\n",
    "\n",
    "\n",
    "umap = {{ BoolField(name = 'umap', label = 'ScatterPlot Visualization', default = 'true', description = 'Select \\'Yes\\' if you would like to generate a Scatter Plot. Otherwise, select \\'No\\'', section = 'UMAP_visualization')}}\n",
    "\n",
    "umap_num_neighbors = {{ IntField(name = 'nneighbors', label = 'Number of Neighbors', default = 5, min = 1, max = 30, description = '##TODO: Play around with parameter settings', section = 'UMAP_visualization')}}\n",
    "umap_maxdf = {{ ChoiceField(name = 'max_df', label = 'Max df setting', choices = {'0.5': '0.5', '0.75': '.75', '0.9': '.9', '1.0': '1'}, default = '0.5',  description = '##TODO: Play around with parameter settings', section = 'UMAP_visualization')}}\n",
    "umap_mindf = {{ ChoiceField(name = 'min_df', label = 'Min df setting', choices = {'0.1' : '0.1', '0.25' : '0.25', '0.5': '0.5' }, default = '0.25', description = '##TODO: Play around with parameter settings', section = 'UMAP_visualization')}}\n",
    "\n",
    "{% if gs == '' %}:\n",
    "    raise Exception('Please upload a GMT File!')\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Process the GMT FILE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "```python\n",
       "def series_to_list(gene_list):\n",
       "    ##helper function to convert a gene pd.series to a gene list\n",
       "    ret_list = []\n",
       "    for genes in gene_list:\n",
       "        if type(genes) is str:\n",
       "            ret_list.append(genes)\n",
       "        else: ##pd series case\n",
       "            genes = genes.tolist()\n",
       "            ret_list.append(' '.join(genes))\n",
       "    return ret_list\n",
       "def load_set(file):\n",
       "    ''' Load a set of files into pairs of labeled sets\n",
       "    '''\n",
       "    lst= []\n",
       "    path = pathlib.Path(file)\n",
       "    with open(path) as f:\n",
       "        lines = f.readlines()\n",
       "        for line in lines:\n",
       "            parsed_line = line.split('\\t')\n",
       "            term, library, genes = parsed_line[0], parsed_line[1], parsed_line[2:]\n",
       "            if genes[-1][:-2] == '\\n':\n",
       "                genes[-1] = genes[-1][:-2] ##trim off newline regex '\\n'\n",
       "            lst.append((term,  library, ' '.join(genes)))\n",
       "    zip_lst = [list(i) for i in zip(*lst)]\n",
       "    term, library, genes = zip_lst[0], zip_lst[1], zip_lst[2]\n",
       "    genes = series_to_list(genes)\n",
       "    df = pd.DataFrame({'Genes': genes, 'Library': library}, index = term)\n",
       "    return df\n",
       "```"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "def series_to_list(gene_list):\n",
    "    ##helper function to convert a gene pd.series to a gene list\n",
    "    ret_list = []\n",
    "    for genes in gene_list:\n",
    "        if type(genes) is str:\n",
    "            ret_list.append(genes)\n",
    "        else: ##pd series case\n",
    "            genes = genes.tolist()        \n",
    "            ret_list.append(' '.join(genes))\n",
    "    return ret_list\n",
    "\n",
    "def load_set(file):\n",
    "    ''' Load a set of files into pairs of labeled sets\n",
    "    '''\n",
    "    lst= []\n",
    "    path = pathlib.Path(file)\n",
    "    with open(path) as f:\n",
    "        lines = f.readlines()\n",
    "        for line in lines:\n",
    "            parsed_line = line.split('\\t')\n",
    "            term, library, genes = parsed_line[0], parsed_line[1], parsed_line[2:]\n",
    "            if genes[-1][:-2] == '\\n':\n",
    "                genes[-1] = genes[-1][:-2] ##trim off newline regex '\\n'\n",
    "            lst.append((term,  library, ' '.join(genes)))\n",
    "    zip_lst = [list(i) for i in zip(*lst)]\n",
    "    term, library, genes = zip_lst[0], zip_lst[1], zip_lst[2]\n",
    "    genes = series_to_list(genes)\n",
    "\n",
    "    df = pd.DataFrame({'Genes': genes, 'Library': library}, index = term)\n",
    "    return df              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df = load_set(gs)\n",
    "if df.shape[0] < umap_num_neighbors:\n",
    "    umap_num_neighbors = int(np.ceil(df.shape[0]/2)) ##arbitrary right now. May want to change based on parameter settings\n",
    "    print('Number of Neighbors parameter in scatterplot is too large for the submitted dataset. Resetting number of neighbors to '+ str(umap_num_neighbors)+'')\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_FET(set1, set2, background = 20000):\n",
    "    ##inputs: set1, set2 - python sets\n",
    "    ##output: p-value of the fisher exact test\n",
    "    res = crisp.fisher_overlap(set1, set2, n_background_entities= background, preserve_overlap=True)\n",
    "    if res == None:\n",
    "        return 0\n",
    "    else:\n",
    "        return res.pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "def series_to_str(el):\n",
    "    if type(el) == str:\n",
    "        return el\n",
    "    else:\n",
    "        return ' '.join(el.tolist())\n",
    "\n",
    "def generate_pairs_df(df, background = 20000):\n",
    "    ##inputs: df - pandas dataframe that is the result of GMT_to_df transformation\n",
    "    ##output: pair_df - pandas dataframe whose rows are indexed by a tuple/ pair of terms in the set of Gene set \n",
    "    # #terms and columns represent calculated set properties between the two sets\n",
    "\n",
    "    intersection = []\n",
    "    in_A_not_B = []\n",
    "    in_B_not_A = []\n",
    "    union = []\n",
    "    jaccard = []\n",
    "    FET_pval = []\n",
    "\n",
    "    to_set = lambda el: set(series_to_str(el).split(' '))\n",
    "    space_counter = lambda str1: str1.count(\" \") +1\n",
    "\n",
    "\n",
    "    terms = list(df.index.values)\n",
    "    int_df = pd.DataFrame(index = terms, columns = terms)\n",
    "    jac_df = pd.DataFrame(index = terms, columns = terms)\n",
    "    pairwise_perms = list(itertools.combinations(terms,2))\n",
    "    for term1,term2 in pairwise_perms:\n",
    "        setA, setB = df.loc[term1]['Genes'], df.loc[term2]['Genes']\n",
    "        set1, set2 = to_set(setA), to_set(setB)\n",
    "        intersect = set1 & set2\n",
    "        union_set = set1 | set2\n",
    "        intersection.append(' '.join(intersect))\n",
    "        in_A_not_B.append(' '.join(list(set1 -set2)))\n",
    "        in_B_not_A.append(' '.join(list(set2 - set1)))\n",
    "        union.append(' '.join(list(union_set)))\n",
    "        pval = calculate_FET(set1, set2)\n",
    "        FET_pval.append(pval)\n",
    "\n",
    "        int_size = len(intersect)\n",
    "        uni_size = len(union_set)\n",
    "        jaccard = int_size/uni_size\n",
    "        \n",
    "        jac_df.loc[term1, term2] = jaccard\n",
    "        jac_df.loc[term2, term1] = jaccard\n",
    "        \n",
    "\n",
    "        int_df.loc[term1, term2] = int_size\n",
    "        int_df.loc[term2,term1] = int_size\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    pair_df = pd.DataFrame({'Intersection' : intersection, 'A-B' : in_A_not_B, 'B-A' : in_B_not_A, 'Union': union, 'FET_pval': FET_pval}, index = pairwise_perms)\n",
    "    pair_df['intersect_size'] = pair_df['Intersection'].map(space_counter)\n",
    "    pair_df['union_size'] = pair_df['Union'].map(space_counter)\n",
    "    pair_df['Jaccard'] = pair_df['intersect_size'] / pair_df['union_size']\n",
    "    \n",
    "    np.fill_diagonal(int_df.values,0)\n",
    "    np.fill_diagonal(jac_df.values, 0)\n",
    "    \n",
    "    \n",
    "\n",
    "    return pair_df, int_df, jac_df\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "pair_df, int_df, jac_df = generate_pairs_df(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Pairwise Intersection Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Integrated breast cancer pathway</th>\n",
       "      <th>Stathmin and breast cancer resistance to antimicrotubule agents</th>\n",
       "      <th>Integrated breast cancer pathway WP1984</th>\n",
       "      <th>Breast cancer pathway WP4262</th>\n",
       "      <th>Breast cancer</th>\n",
       "      <th>NOTCH1 Signaling in Breast Cancer</th>\n",
       "      <th>Genes with Mutations Associated with Breast Cancer</th>\n",
       "      <th>Genes with Mutations Associated with Hereditary Breast and/or Ovarian Cancer Syndrome</th>\n",
       "      <th>Proteins Involved in Breast Cancer Related to ERBB2/VEGFR/Akt Signaling Pathway</th>\n",
       "      <th>Hereditary Breast and Ovarian Cancer Syndrome</th>\n",
       "      <th>...</th>\n",
       "      <th>estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:39</th>\n",
       "      <th>Interleukin-13 human Breast cancer - MCF10CA1a cell line (pLKO-shSCR) GSE57677 ligand:242</th>\n",
       "      <th>estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:41</th>\n",
       "      <th>Interleukin-13 human Breast cancer - MCF10CA1a cell line (pLKO-shSCR) GSE57677 ligand:242</th>\n",
       "      <th>estradiol human MCF-7 breast cancer cells GDS3283 ligand:42</th>\n",
       "      <th>estradiol human MCF-7 breast cancer cells GDS3283 ligand:43</th>\n",
       "      <th>Interleukin-13 human Breast cancer - MCF10CA1a cell line (pLKO-shIL13RA2) GSE57677 ligand:243</th>\n",
       "      <th>estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:39</th>\n",
       "      <th>estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:40</th>\n",
       "      <th>estradiol human MCF-7 breast cancer cells GDS3105 ligand:38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Integrated breast cancer pathway</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>146</td>\n",
       "      <td>29</td>\n",
       "      <td>25</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stathmin and breast cancer resistance to antimicrotubule agents</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Integrated breast cancer pathway WP1984</th>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>25</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>23</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Breast cancer pathway WP4262</th>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>139</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>27</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Breast cancer</th>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>139</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>27</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Integrated breast cancer pathway   \n",
       "Integrated breast cancer pathway                                                  0  \\\n",
       "Stathmin and breast cancer resistance to antimi...                                0   \n",
       "Integrated breast cancer pathway WP1984                                         146   \n",
       "Breast cancer pathway WP4262                                                     29   \n",
       "Breast cancer                                                                    25   \n",
       "\n",
       "                                                   Stathmin and breast cancer resistance to antimicrotubule agents   \n",
       "Integrated breast cancer pathway                                                                    0               \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  0                \n",
       "Integrated breast cancer pathway WP1984                                                             0                \n",
       "Breast cancer pathway WP4262                                                                        0                \n",
       "Breast cancer                                                                                       0                \n",
       "\n",
       "                                                   Integrated breast cancer pathway WP1984   \n",
       "Integrated breast cancer pathway                                                       146  \\\n",
       "Stathmin and breast cancer resistance to antimi...                                       0   \n",
       "Integrated breast cancer pathway WP1984                                                  0   \n",
       "Breast cancer pathway WP4262                                                            30   \n",
       "Breast cancer                                                                           25   \n",
       "\n",
       "                                                   Breast cancer pathway WP4262   \n",
       "Integrated breast cancer pathway                                             29  \\\n",
       "Stathmin and breast cancer resistance to antimi...                            0   \n",
       "Integrated breast cancer pathway WP1984                                      30   \n",
       "Breast cancer pathway WP4262                                                  0   \n",
       "Breast cancer                                                               139   \n",
       "\n",
       "                                                   Breast cancer   \n",
       "Integrated breast cancer pathway                              25  \\\n",
       "Stathmin and breast cancer resistance to antimi...             0   \n",
       "Integrated breast cancer pathway WP1984                       25   \n",
       "Breast cancer pathway WP4262                                 139   \n",
       "Breast cancer                                                  0   \n",
       "\n",
       "                                                   NOTCH1 Signaling in Breast Cancer   \n",
       "Integrated breast cancer pathway                                                  12  \\\n",
       "Stathmin and breast cancer resistance to antimi...                                 1   \n",
       "Integrated breast cancer pathway WP1984                                           12   \n",
       "Breast cancer pathway WP4262                                                      14   \n",
       "Breast cancer                                                                     15   \n",
       "\n",
       "                                                   Genes with Mutations Associated with Breast Cancer   \n",
       "Integrated breast cancer pathway                                                                    4  \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  0   \n",
       "Integrated breast cancer pathway WP1984                                                             4   \n",
       "Breast cancer pathway WP4262                                                                        6   \n",
       "Breast cancer                                                                                       3   \n",
       "\n",
       "                                                   Genes with Mutations Associated with Hereditary Breast and/or Ovarian Cancer Syndrome   \n",
       "Integrated breast cancer pathway                                                                    6                                     \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  0                                      \n",
       "Integrated breast cancer pathway WP1984                                                             6                                      \n",
       "Breast cancer pathway WP4262                                                                        6                                      \n",
       "Breast cancer                                                                                       6                                      \n",
       "\n",
       "                                                   Proteins Involved in Breast Cancer Related to ERBB2/VEGFR/Akt Signaling Pathway   \n",
       "Integrated breast cancer pathway                                                                   23                               \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  1                                \n",
       "Integrated breast cancer pathway WP1984                                                            23                                \n",
       "Breast cancer pathway WP4262                                                                       27                                \n",
       "Breast cancer                                                                                      27                                \n",
       "\n",
       "                                                   Hereditary Breast and Ovarian Cancer Syndrome   \n",
       "Integrated breast cancer pathway                                                              16  \\\n",
       "Stathmin and breast cancer resistance to antimi...                                             0   \n",
       "Integrated breast cancer pathway WP1984                                                       16   \n",
       "Breast cancer pathway WP4262                                                                  14   \n",
       "Breast cancer                                                                                 10   \n",
       "\n",
       "                                                    ...   \n",
       "Integrated breast cancer pathway                    ...  \\\n",
       "Stathmin and breast cancer resistance to antimi...  ...   \n",
       "Integrated breast cancer pathway WP1984             ...   \n",
       "Breast cancer pathway WP4262                        ...   \n",
       "Breast cancer                                       ...   \n",
       "\n",
       "                                                   estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:39   \n",
       "Integrated breast cancer pathway                                                                    5                                          \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  5                                           \n",
       "Integrated breast cancer pathway WP1984                                                             6                                           \n",
       "Breast cancer pathway WP4262                                                                        5                                           \n",
       "Breast cancer                                                                                       5                                           \n",
       "\n",
       "                                                   Interleukin-13 human Breast cancer - MCF10CA1a cell line (pLKO-shSCR) GSE57677 ligand:242   \n",
       "Integrated breast cancer pathway                                                                    0                                         \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  1                                          \n",
       "Integrated breast cancer pathway WP1984                                                             0                                          \n",
       "Breast cancer pathway WP4262                                                                        1                                          \n",
       "Breast cancer                                                                                       1                                          \n",
       "\n",
       "                                                   estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:41   \n",
       "Integrated breast cancer pathway                                                                    6                                          \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  3                                           \n",
       "Integrated breast cancer pathway WP1984                                                             7                                           \n",
       "Breast cancer pathway WP4262                                                                        6                                           \n",
       "Breast cancer                                                                                       6                                           \n",
       "\n",
       "                                                   Interleukin-13 human Breast cancer - MCF10CA1a cell line (pLKO-shSCR) GSE57677 ligand:242   \n",
       "Integrated breast cancer pathway                                                                    0                                         \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  1                                          \n",
       "Integrated breast cancer pathway WP1984                                                             0                                          \n",
       "Breast cancer pathway WP4262                                                                        1                                          \n",
       "Breast cancer                                                                                       1                                          \n",
       "\n",
       "                                                   estradiol human MCF-7 breast cancer cells GDS3283 ligand:42   \n",
       "Integrated breast cancer pathway                                                                    3           \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  1            \n",
       "Integrated breast cancer pathway WP1984                                                             3            \n",
       "Breast cancer pathway WP4262                                                                       11            \n",
       "Breast cancer                                                                                      11            \n",
       "\n",
       "                                                   estradiol human MCF-7 breast cancer cells GDS3283 ligand:43   \n",
       "Integrated breast cancer pathway                                                                   12           \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  1            \n",
       "Integrated breast cancer pathway WP1984                                                            12            \n",
       "Breast cancer pathway WP4262                                                                       13            \n",
       "Breast cancer                                                                                      13            \n",
       "\n",
       "                                                   Interleukin-13 human Breast cancer - MCF10CA1a cell line (pLKO-shIL13RA2) GSE57677 ligand:243   \n",
       "Integrated breast cancer pathway                                                                    2                                             \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  0                                              \n",
       "Integrated breast cancer pathway WP1984                                                             2                                              \n",
       "Breast cancer pathway WP4262                                                                        1                                              \n",
       "Breast cancer                                                                                       2                                              \n",
       "\n",
       "                                                   estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:39   \n",
       "Integrated breast cancer pathway                                                                    5                                          \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  5                                           \n",
       "Integrated breast cancer pathway WP1984                                                             6                                           \n",
       "Breast cancer pathway WP4262                                                                        5                                           \n",
       "Breast cancer                                                                                       5                                           \n",
       "\n",
       "                                                   estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:40   \n",
       "Integrated breast cancer pathway                                                                    5                                          \\\n",
       "Stathmin and breast cancer resistance to antimi...                                                  6                                           \n",
       "Integrated breast cancer pathway WP1984                                                             5                                           \n",
       "Breast cancer pathway WP4262                                                                        3                                           \n",
       "Breast cancer                                                                                       3                                           \n",
       "\n",
       "                                                   estradiol human MCF-7 breast cancer cells GDS3105 ligand:38  \n",
       "Integrated breast cancer pathway                                                                    6           \n",
       "Stathmin and breast cancer resistance to antimi...                                                  2           \n",
       "Integrated breast cancer pathway WP1984                                                             6           \n",
       "Breast cancer pathway WP4262                                                                        9           \n",
       "Breast cancer                                                                                       9           \n",
       "\n",
       "[5 rows x 90 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Download Pairwise Intersection Matrix:<a href='P.I_matrix/intersection_matrix.csv' target='_blank'>P.I_matrix/intersection_matrix.csv</a><br>"
      ],
      "text/plain": [
       "C:\\Users\\groga\\OneDrive\\Desktop\\MtSinai\\appyter-catalog\\appyters\\GeneSetLibrarySynopsis\\P.I_matrix\\intersection_matrix.csv"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "if int_tbl:\n",
    "    os.makedirs(\"P.I_matrix\", exist_ok = True)\n",
    "    int_df.to_csv('P.I_matrix/intersection_matrix.csv')\n",
    "    display(int_df.head())\n",
    "    display(FileLink('P.I_matrix/intersection_matrix.csv', result_html_prefix= str('Download Pairwise Intersection Matrix:')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Jaccard Similarity Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "if jaccard_tbl:\n",
    "    display(jac_df.head())\n",
    "    jac_df.to_csv('P.I_matrix/jaccard_matrix.csv')\n",
    "    display(FileLink('P.I_matrix/jaccard_matrix.csv', result_html_prefix= str('Download Pairwise Intersection Matrix:')))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. ScatterPlot Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NoResults(Exception):\n",
    "    pass\n",
    "\n",
    "\n",
    "class APIFailure(Exception):\n",
    "    pass\n",
    "\n",
    "\n",
    "class NotValidFile(Exception):\n",
    "    pass\n",
    "\n",
    "\n",
    "class UMAP_Visualization:\n",
    "\n",
    "    def __init__(self, query_set=[], gene_libraries=[], sig_value=.05, gmt_files=[], gmt_df = []):\n",
    "        self.query_set = [gene.strip() for gene in query_set]\n",
    "        self.gene_libraries = gene_libraries\n",
    "        self.significant_value = sig_value\n",
    "        self.term_library_map = {}\n",
    "        self.dataset = OrderedDict()\n",
    "        self.dataset.update(self.process_gmt_df(gmt_df))\n",
    "        \n",
    "        \n",
    "    def process_gmt_df(self, gmt_df):\n",
    "        if gmt_df == []:\n",
    "            return OrderedDict() ##return the empty Dictionary when no gmt passed in\n",
    "        else:\n",
    "            gmt_df = gmt_df[0]\n",
    "            self.term_library_map.update(pd.Series(gmt_df['Library'].values,index=gmt_df.index.values).to_dict())\n",
    "            print(self.term_library_map)\n",
    "            return OrderedDict(pd.Series(gmt_df['Genes'].values,index=gmt_df.index.values).to_dict())\n",
    "            \n",
    "   \n",
    "\n",
    "    def process_scatterplot(self, nneighbors=30, mindist=0.1, spread=1.0, maxdf=1.0, mindf=1):\n",
    "        libdict = self.dataset\n",
    "        print(\"\\tTF-IDF vectorizing gene set data...\")\n",
    "        # computes tdfidf score--look this up\n",
    "        vec = TfidfVectorizer(max_df=maxdf, min_df=mindf)\n",
    "        X = vec.fit_transform(libdict.values())\n",
    "        print(X.shape)\n",
    "        adata = anndata.AnnData(X)\n",
    "        adata.obs.index = libdict.keys()\n",
    "\n",
    "        print(\"\\tPerforming Leiden clustering...\")\n",
    "        # the n_neighbors and min_dist parameters can be altered\n",
    "        sc.pp.neighbors(adata, n_neighbors=nneighbors)\n",
    "        sc.tl.leiden(adata, resolution=1.0)\n",
    "        sc.tl.umap(adata, min_dist=mindist, spread=spread, random_state=42)\n",
    "\n",
    "        new_order = adata.obs.sort_values(by='leiden').index.tolist()\n",
    "        adata = adata[new_order, :]\n",
    "        adata.obs['leiden'] = 'Cluster ' + adata.obs['leiden'].astype('object')\n",
    "\n",
    "        df = pd.DataFrame(adata.obsm['X_umap'])\n",
    "        df.columns = ['x', 'y']\n",
    "\n",
    "        df['cluster'] = adata.obs['leiden'].values\n",
    "        df['term'] = adata.obs.index\n",
    "        df['genes'] = [libdict[l] for l in df['term']]\n",
    "        df['library'] = [self.term_library_map[l] for l in df['term']]\n",
    "\n",
    "        return df\n",
    "\n",
    "    def get_scatter_colors(self, df):\n",
    "        clusters = pd.unique(df['cluster']).tolist()\n",
    "        colors = list(Category20[20])[::2] + list(Category20[20])[1::2]\n",
    "        color_mapper = {clusters[i]: colors[i % 20]\n",
    "                        for i in range(len(clusters))}\n",
    "        return color_mapper\n",
    "\n",
    "    # def get_marker_mapper(self, df):\n",
    "    #     markers = [\"circle\", \"square\", \"triangle\",\n",
    "    #                \"hex\", \"inverted_triangle\", \"diamond\"]\n",
    "    #     libs = pd.unique(df['library']).tolist()\n",
    "    #     marker_mapper = {libs[i]: markers[i] for i in range(len(libs))}\n",
    "    #     return marker_mapper\n",
    "\n",
    "    def get_scatterplot(self, scatterdf):\n",
    "        df = scatterdf.copy()\n",
    "        color_mapper = self.get_scatter_colors(df)\n",
    "        # marker_mapper = self.get_marker_mapper(df)\n",
    "        df['color'] = df['cluster'].apply(lambda x: color_mapper[x])\n",
    "        # df['marker'] = df['library'].apply(lambda x: marker_mapper[x])\n",
    "\n",
    "        # range_slider = RangeSlider(\"title = Adjust x-axis\",\n",
    "        #                            start=0,\n",
    "        #                            end=10,\n",
    "        #                            step=1)\n",
    "\n",
    "        tooltips = [\n",
    "            (\"Gene Set\", \"@gene_set\"),\n",
    "            (\"Cluster\", \"@label\"),\n",
    "            (\"Library\", \"@library\")\n",
    "        ]\n",
    "\n",
    "        hover_emb = HoverTool(tooltips=tooltips)\n",
    "        tools_emb = [hover_emb, 'pan', 'wheel_zoom', 'reset', 'save']\n",
    "\n",
    "        plot_emb = figure(\n",
    "            width=900,\n",
    "            height=700,\n",
    "            tools=tools_emb\n",
    "        )\n",
    "\n",
    "        source = ColumnDataSource(\n",
    "            data=dict(\n",
    "                x=df['x'],\n",
    "                y=df['y'],\n",
    "                gene_set=df['term'],\n",
    "                colors=df['color'],\n",
    "                label=df['cluster'],\n",
    "                library=df['library'],\n",
    "                # markers=df['marker']\n",
    "\n",
    "            )\n",
    "        )\n",
    "\n",
    "        # hide axis labels and grid lines\n",
    "        plot_emb.xaxis.major_tick_line_color = None\n",
    "        plot_emb.xaxis.minor_tick_line_color = None\n",
    "        plot_emb.yaxis.major_tick_line_color = None\n",
    "        plot_emb.yaxis.minor_tick_line_color = None\n",
    "        plot_emb.xaxis.major_label_text_font_size = '0pt'\n",
    "        plot_emb.yaxis.major_label_text_font_size = '0pt'\n",
    "\n",
    "        plot_emb.output_backend = \"svg\"\n",
    "\n",
    "        plot_emb.xaxis.axis_label = \"UMAP_1\"\n",
    "        plot_emb.yaxis.axis_label = \"UMAP_2\"\n",
    "\n",
    "        s = plot_emb.scatter(\n",
    "            'x',\n",
    "            'y',\n",
    "            size=4,\n",
    "            source=source,\n",
    "            color='colors',\n",
    "            legend_group='label',\n",
    "            # marker='markers'\n",
    "        )\n",
    "\n",
    "        plot_emb.add_layout(plot_emb.legend[0], 'right')\n",
    "\n",
    "        return plot_emb\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "```python\n",
       "if umap:\n",
       "    umap = UMAP_Visualization(gmt_df = [df])\n",
       "    umap_df = umap.process_scatterplot(maxdf = umap_maxdf, mindf = umap_mindf, nneighbors = umap_num_neighbors)\n",
       "    fig = umap.get_scatterplot(umap_df)\n",
       "    show(fig)\n",
       "```"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Integrated breast cancer pathway': 'BioPlanet_2019', 'Stathmin and breast cancer resistance to antimicrotubule agents': 'BioPlanet_2019', 'Integrated breast cancer pathway WP1984': 'WikiPathway_2021_Human', 'Breast cancer pathway WP4262': 'WikiPathways_2019_Human', 'Breast cancer': 'KEGG_2021_Human', 'NOTCH1 Signaling in Breast Cancer': 'Elsevier_Pathway_Collection', 'Genes with Mutations Associated with Breast Cancer': 'Elsevier_Pathway_Collection', 'Genes with Mutations Associated with Hereditary Breast and/or Ovarian Cancer Syndrome': 'Elsevier_Pathway_Collection', 'Proteins Involved in Breast Cancer Related to ERBB2/VEGFR/Akt Signaling Pathway': 'Elsevier_Pathway_Collection', 'Hereditary Breast and Ovarian Cancer Syndrome': 'Elsevier_Pathway_Collection', 'Proteins Involved in Breast Cancer Related to ESR1 Signaling Pathway': 'Elsevier_Pathway_Collection', 'Proteins Involved in Breast Cancer Related to IGF1R/Akt Signaling Pathway': 'Elsevier_Pathway_Collection', 'Proteins Involved in Breast Cancer Related to NOTCH Signaling Pathway': 'Elsevier_Pathway_Collection', 'Proteins Involved in Breast Cancer Related to WNT Signaling Pathway': 'Elsevier_Pathway_Collection', 'Breast Cancer': 'Elsevier_Pathway_Collection', 'WNT Signaling in Breast Cancer': 'Elsevier_Pathway_Collection', 'IGF1R/AKT Signaling in Breast Cancer': 'Elsevier_Pathway_Collection', 'Proteins with Altered Expression in Breast Cancer': 'Elsevier_Pathway_Collection', 'ERBB/VEGFR/Akt Signaling in Breast Cancer': 'Elsevier_Pathway_Collection', 'ESR1 Signaling in Breast Cancer': 'Elsevier_Pathway_Collection', 'ESR1/ERBB Positive Luminal Breast Cancer': 'Elsevier_Pathway_Collection', 'Stathmin and breast cancer resistance to antimicrotubule agents Homo sapiens h stathminPathway': 'BioCarta_2016', 'Cancer Stem cell:Breast': 'CellMarker_Augmented_2021', 'Integrated Breast Cancer Pathway WP1984': 'WikiPathways_2019_Human', 'Integrated Breast Cancer Pathway Homo sapiens WP1984': 'WikiPathways_2016', 'IRF1-19129219-H3396 breast cancer cells-human': 'ChEA_2013', 'breast cancer cell': 'Tissue_Protein_Expression_from_ProteomicsDB', 'Control Nrde2-Depleted Breast Cancer GSE119827 1': 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'Multi-Treatment Models Breast Cancer GSE136823 1': 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'Multi-Treatment Models Breast Cancer GSE136823 2': 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'Multi-Treatment Models Breast Cancer GSE136823 3': 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'Multi-Treatment Models Breast Cancer GSE136823 4': 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'Multi-Treatment Models Breast Cancer GSE136823 5': 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'Multi-Treatment Models Breast Cancer GSE136823 6': 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'Multi-Treatment Models Breast Cancer GSE136823 7': 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'Differentially Breast Cancer Diabetes GSE150586 1': 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'Breast cancer DOID-1612 mouse GSE24594 sample 616': 'Disease_Perturbations_from_GEO_up', 'Breast Cancer DOID-1612 human GSE34925 sample 478': 'Disease_Perturbations_from_GEO_up', 'Breast Cancer C0006142 human GSE2429 sample 148': 'Disease_Perturbations_from_GEO_up', 'Breast Cancer C0006142 human GSE1379 sample 392': 'Disease_Perturbations_from_GEO_up', 'breast cancer DOID-1612 human GSE9574 sample 448': 'Disease_Perturbations_from_GEO_up', 'breast cancer DOID-1612 human GSE14943 sample 504': 'Disease_Perturbations_from_GEO_up', 'breast cancer DOID-1612 human GSE26910 sample 602': 'Disease_Perturbations_from_GEO_up', 'Breast Cancer C0006142 human GSE1378 sample 52': 'Disease_Perturbations_from_GEO_up', 'sporadic breast cancer DOID-8029 human GSE3744 sample 979': 'Disease_Perturbations_from_GEO_up', 'Breast Cancer C0006142 human GSE2155 sample 39': 'Disease_Perturbations_from_GEO_up', 'breast cancer DOID-1612 human GSE3744 sample 978': 'Disease_Perturbations_from_GEO_up', 'Breast Cancer C0006142 mouse GSE2528 sample 28': 'Disease_Perturbations_from_GEO_up', 'Breast Cancer C0006142 human GSE3744 sample 24': 'Disease_Perturbations_from_GEO_up', 'Breast Cancer C0006142 rat GSE1872 sample 63': 'Disease_Perturbations_from_GEO_up', 'estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:41': 'Ligand_Perturbations_from_GEO_down', 'estradiol human MCF-7 breast cancer cells GDS3283 ligand:43': 'Ligand_Perturbations_from_GEO_down', 'estradiol human MCF-7 breast cancer cells GDS3283 ligand:42': 'Ligand_Perturbations_from_GEO_down', 'Interleukin-13 human Breast cancer - MCF10CA1a cell line (pLKO-shIL13RA2) GSE57677 ligand:243': 'Ligand_Perturbations_from_GEO_down', 'estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:40': 'Ligand_Perturbations_from_GEO_down', 'estradiol human MCF-7 breast cancer cells GDS3105 ligand:38': 'Ligand_Perturbations_from_GEO_down', 'estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:39': 'Ligand_Perturbations_from_GEO_down', 'Interleukin-13 human Breast cancer - MCF10CA1a cell line (pLKO-shSCR) GSE57677 ligand:242': 'Ligand_Perturbations_from_GEO_down'}\n",
      "\tTF-IDF vectorizing gene set data...\n",
      "(58, 15)\n",
      "\tPerforming Leiden clustering...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\groga\\AppData\\Local\\Temp\\ipykernel_16512\\3212203020.py:53: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs['leiden'] = 'Cluster ' + adata.obs['leiden'].astype('object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"fa1245eb-af74-426e-b931-39e957f460da\" data-root-id=\"p1010\" style=\"display: contents;\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function embed_document(root) {\n",
       "  const docs_json = {\"21fc5328-2d3f-4384-ae3e-8e121c9afa5b\":{\"version\":\"3.1.0\",\"title\":\"Bokeh Application\",\"defs\":[],\"roots\":[{\"type\":\"object\",\"name\":\"Figure\",\"id\":\"p1010\",\"attributes\":{\"width\":900,\"height\":700,\"x_range\":{\"type\":\"object\",\"name\":\"DataRange1d\",\"id\":\"p1012\"},\"y_range\":{\"type\":\"object\",\"name\":\"DataRange1d\",\"id\":\"p1011\"},\"x_scale\":{\"type\":\"object\",\"name\":\"LinearScale\",\"id\":\"p1023\"},\"y_scale\":{\"type\":\"object\",\"name\":\"LinearScale\",\"id\":\"p1025\"},\"title\":{\"type\":\"object\",\"name\":\"Title\",\"id\":\"p1014\"},\"renderers\":[{\"type\":\"object\",\"name\":\"GlyphRenderer\",\"id\":\"p1060\",\"attributes\":{\"data_source\":{\"type\":\"object\",\"name\":\"ColumnDataSource\",\"id\":\"p1051\",\"attributes\":{\"selected\":{\"type\":\"object\",\"name\":\"Selection\",\"id\":\"p1052\",\"attributes\":{\"indices\":[],\"line_indices\":[]}},\"selection_policy\":{\"type\":\"object\",\"name\":\"UnionRenderers\",\"id\":\"p1053\"},\"data\":{\"type\":\"map\",\"entries\":[[\"x\",{\"type\":\"ndarray\",\"array\":{\"type\":\"bytes\",\"data\":\"hgzyQZlq70GwavBBEpnvQfsy6kFj5/RBsNT0QekF6UEAx/FBOtzzQYPF8kFCFvFBoFnnQSDh8EGJ+/FB/s73QGKK4kADD+RAbmLvQJUh4UC9xutAC7fQQNhe90BDgehAU9PMQJX85kBz+IM/S2LIP5I2gj6iFTQ/HKbsP8Co2T9ZIsk+bGnmPxtP9j+SAGE/u2jAQOE7okC8NZlA+6qyQKkgk0AYSKRAx6CBwHlPhsByrYXAHIV4wKpohsAIboPArMQSv1Je0L76mEO/gYJhv726QL9JIjFBqjUuQQcdMkHuUTBBdZEzQQ==\"},\"shape\":[58],\"dtype\":\"float32\",\"order\":\"little\"}],[\"y\",{\"type\":\"ndarray\",\"array\":{\"type\":\"bytes\",\"data\":\"VAZkQWecY0F7jWpBr9BVQV/IVUF072NBBd1bQfaGVEH7Y1xBfy5gQYCpaUFLlWRBIElUQRRsXUGL6FVBcptAQILYOUBw2ClADd43QMzoQ0DzSRxAQDobQLKQKkBF/kNAVEZLQC08IkA3D+ZAdkzgQKCk7EAGiuRA97PXQJnU5EDC2ulA7vHnQHdD60BDLOZA9wd8QBQSpUDxzKRAnZCKQN/Hn0D855lAVPsKQayZCkEVlwVBjekCQSezDkETbghBxvT4QCpp7UC10fZAK1HuQDGJ8UAaU/1Abu4AQUKD70Aw8fpAIir0QA==\"},\"shape\":[58],\"dtype\":\"float32\",\"order\":\"little\"}],[\"gene_set\",{\"type\":\"ndarray\",\"array\":[\"Interleukin-13 human Breast cancer - MCF10CA1a cell line (pLKO-shSCR) GSE57677 ligand:242\",\"Breast Cancer C0006142 human GSE1379 sample 392\",\"Breast Cancer DOID-1612 human GSE34925 sample 478\",\"Breast Cancer C0006142 human GSE1378 sample 52\",\"breast cancer cell\",\"breast cancer DOID-1612 human GSE3744 sample 978\",\"breast cancer DOID-1612 human GSE26910 sample 602\",\"estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:41\",\"estradiol human MCF-7 breast cancer cells GDS3283 ligand:43\",\"Breast Cancer C0006142 human GSE3744 sample 24\",\"breast cancer DOID-1612 human GSE14943 sample 504\",\"estradiol human MCF-7 breast cancer cells GDS3105 ligand:38\",\"estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:40\",\"Interleukin-13 human Breast cancer - MCF10CA1a cell line (pLKO-shIL13RA2) GSE57677 ligand:243\",\"estradiol human MCF-7 breast cancer cells GDS3283 ligand:42\",\"Stathmin and breast cancer resistance to antimicrotubule agents\",\"Breast Cancer C0006142 human GSE2429 sample 148\",\"Differentially Breast Cancer Diabetes GSE150586 1\",\"sporadic breast cancer DOID-8029 human GSE3744 sample 979\",\"Multi-Treatment Models Breast Cancer GSE136823 2\",\"Control Nrde2-Depleted Breast Cancer GSE119827 1\",\"Cancer Stem cell:Breast\",\"IRF1-19129219-H3396 breast cancer cells-human\",\"Genes with Mutations Associated with Breast Cancer\",\"Genes with Mutations Associated with Hereditary Breast and/or Ovarian Cancer Syndrome\",\"Stathmin and breast cancer resistance to antimicrotubule agents Homo sapiens h stathminPathway\",\"Proteins Involved in Breast Cancer Related to ERBB2/VEGFR/Akt Signaling Pathway\",\"NOTCH1 Signaling in Breast Cancer\",\"Proteins Involved in Breast Cancer Related to IGF1R/Akt Signaling Pathway\",\"ERBB/VEGFR/Akt Signaling in Breast Cancer\",\"Breast Cancer C0006142 human GSE2155 sample 39\",\"Proteins Involved in Breast Cancer Related to NOTCH Signaling Pathway\",\"IGF1R/AKT Signaling in Breast Cancer\",\"Proteins Involved in Breast Cancer Related to WNT Signaling Pathway\",\"WNT Signaling in Breast Cancer\",\"Breast Cancer\",\"Breast Cancer C0006142 mouse GSE2528 sample 28\",\"Multi-Treatment Models Breast Cancer GSE136823 6\",\"Multi-Treatment Models Breast Cancer GSE136823 3\",\"Breast Cancer C0006142 rat GSE1872 sample 63\",\"Breast cancer DOID-1612 mouse GSE24594 sample 616\",\"Multi-Treatment Models Breast Cancer GSE136823 5\",\"Integrated breast cancer pathway\",\"Integrated Breast Cancer Pathway Homo sapiens WP1984\",\"Integrated Breast Cancer Pathway WP1984\",\"ESR1/ERBB Positive Luminal Breast Cancer\",\"Proteins with Altered Expression in Breast Cancer\",\"Integrated breast cancer pathway WP1984\",\"ESR1 Signaling in Breast Cancer\",\"Proteins Involved in Breast Cancer Related to ESR1 Signaling Pathway\",\"Hereditary Breast and Ovarian Cancer Syndrome\",\"Breast cancer\",\"Breast cancer pathway WP4262\",\"Multi-Treatment Models Breast Cancer GSE136823 7\",\"Multi-Treatment Models Breast Cancer GSE136823 4\",\"estradiol human estrogen receptor (ER)-positive MCF7 breast cancer cells GDS3217 ligand:39\",\"breast cancer DOID-1612 human GSE9574 sample 448\",\"Multi-Treatment Models Breast Cancer GSE136823 1\"],\"shape\":[58],\"dtype\":\"object\",\"order\":\"little\"}],[\"colors\",{\"type\":\"ndarray\",\"array\":[\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#1f77b4\",\"#ff7f0e\",\"#ff7f0e\",\"#ff7f0e\",\"#ff7f0e\",\"#ff7f0e\",\"#ff7f0e\",\"#ff7f0e\",\"#ff7f0e\",\"#ff7f0e\",\"#ff7f0e\",\"#ff7f0e\",\"#2ca02c\",\"#2ca02c\",\"#2ca02c\",\"#2ca02c\",\"#2ca02c\",\"#2ca02c\",\"#2ca02c\",\"#2ca02c\",\"#2ca02c\",\"#2ca02c\",\"#d62728\",\"#d62728\",\"#d62728\",\"#d62728\",\"#d62728\",\"#d62728\",\"#9467bd\",\"#9467bd\",\"#9467bd\",\"#9467bd\",\"#9467bd\",\"#9467bd\",\"#8c564b\",\"#8c564b\",\"#8c564b\",\"#8c564b\",\"#8c564b\",\"#e377c2\",\"#e377c2\",\"#e377c2\",\"#e377c2\",\"#e377c2\"],\"shape\":[58],\"dtype\":\"object\",\"order\":\"little\"}],[\"label\",{\"type\":\"ndarray\",\"array\":[\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 0\",\"Cluster 1\",\"Cluster 1\",\"Cluster 1\",\"Cluster 1\",\"Cluster 1\",\"Cluster 1\",\"Cluster 1\",\"Cluster 1\",\"Cluster 1\",\"Cluster 1\",\"Cluster 1\",\"Cluster 2\",\"Cluster 2\",\"Cluster 2\",\"Cluster 2\",\"Cluster 2\",\"Cluster 2\",\"Cluster 2\",\"Cluster 2\",\"Cluster 2\",\"Cluster 2\",\"Cluster 3\",\"Cluster 3\",\"Cluster 3\",\"Cluster 3\",\"Cluster 3\",\"Cluster 3\",\"Cluster 4\",\"Cluster 4\",\"Cluster 4\",\"Cluster 4\",\"Cluster 4\",\"Cluster 4\",\"Cluster 5\",\"Cluster 5\",\"Cluster 5\",\"Cluster 5\",\"Cluster 5\",\"Cluster 6\",\"Cluster 6\",\"Cluster 6\",\"Cluster 6\",\"Cluster 6\"],\"shape\":[58],\"dtype\":\"object\",\"order\":\"little\"}],[\"library\",{\"type\":\"ndarray\",\"array\":[\"Ligand_Perturbations_from_GEO_down\",\"Disease_Perturbations_from_GEO_up\",\"Disease_Perturbations_from_GEO_up\",\"Disease_Perturbations_from_GEO_up\",\"Tissue_Protein_Expression_from_ProteomicsDB\",\"Disease_Perturbations_from_GEO_up\",\"Disease_Perturbations_from_GEO_up\",\"Ligand_Perturbations_from_GEO_down\",\"Ligand_Perturbations_from_GEO_down\",\"Disease_Perturbations_from_GEO_up\",\"Disease_Perturbations_from_GEO_up\",\"Ligand_Perturbations_from_GEO_down\",\"Ligand_Perturbations_from_GEO_down\",\"Ligand_Perturbations_from_GEO_down\",\"Ligand_Perturbations_from_GEO_down\",\"BioPlanet_2019\",\"Disease_Perturbations_from_GEO_up\",\"RNAseq_Automatic_GEO_Signatures_Human_Up\",\"Disease_Perturbations_from_GEO_up\",\"RNAseq_Automatic_GEO_Signatures_Human_Up\",\"RNAseq_Automatic_GEO_Signatures_Human_Up\",\"CellMarker_Augmented_2021\",\"ChEA_2013\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"BioCarta_2016\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"Disease_Perturbations_from_GEO_up\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"Disease_Perturbations_from_GEO_up\",\"RNAseq_Automatic_GEO_Signatures_Human_Up\",\"RNAseq_Automatic_GEO_Signatures_Human_Up\",\"Disease_Perturbations_from_GEO_up\",\"Disease_Perturbations_from_GEO_up\",\"RNAseq_Automatic_GEO_Signatures_Human_Up\",\"BioPlanet_2019\",\"WikiPathways_2016\",\"WikiPathways_2019_Human\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"WikiPathway_2021_Human\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"Elsevier_Pathway_Collection\",\"KEGG_2021_Human\",\"WikiPathways_2019_Human\",\"RNAseq_Automatic_GEO_Signatures_Human_Up\",\"RNAseq_Automatic_GEO_Signatures_Human_Up\",\"Ligand_Perturbations_from_GEO_down\",\"Disease_Perturbations_from_GEO_up\",\"RNAseq_Automatic_GEO_Signatures_Human_Up\"],\"shape\":[58],\"dtype\":\"object\",\"order\":\"little\"}]]}}},\"view\":{\"type\":\"object\",\"name\":\"CDSView\",\"id\":\"p1061\",\"attributes\":{\"filter\":{\"type\":\"object\",\"name\":\"AllIndices\",\"id\":\"p1062\"}}},\"glyph\":{\"type\":\"object\",\"name\":\"Scatter\",\"id\":\"p1057\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"x\"},\"y\":{\"type\":\"field\",\"field\":\"y\"},\"line_color\":{\"type\":\"field\",\"field\":\"colors\"},\"fill_color\":{\"type\":\"field\",\"field\":\"colors\"},\"hatch_color\":{\"type\":\"field\",\"field\":\"colors\"}}},\"nonselection_glyph\":{\"type\":\"object\",\"name\":\"Scatter\",\"id\":\"p1058\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"x\"},\"y\":{\"type\":\"field\",\"field\":\"y\"},\"line_color\":{\"type\":\"field\",\"field\":\"colors\"},\"line_alpha\":{\"type\":\"value\",\"value\":0.1},\"fill_color\":{\"type\":\"field\",\"field\":\"colors\"},\"fill_alpha\":{\"type\":\"value\",\"value\":0.1},\"hatch_color\":{\"type\":\"field\",\"field\":\"colors\"},\"hatch_alpha\":{\"type\":\"value\",\"value\":0.1}}},\"muted_glyph\":{\"type\":\"object\",\"name\":\"Scatter\",\"id\":\"p1059\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"x\"},\"y\":{\"type\":\"field\",\"field\":\"y\"},\"line_color\":{\"type\":\"field\",\"field\":\"colors\"},\"line_alpha\":{\"type\":\"value\",\"value\":0.2},\"fill_color\":{\"type\":\"field\",\"field\":\"colors\"},\"fill_alpha\":{\"type\":\"value\",\"value\":0.2},\"hatch_color\":{\"type\":\"field\",\"field\":\"colors\"},\"hatch_alpha\":{\"type\":\"value\",\"value\":0.2}}}}}],\"toolbar\":{\"type\":\"object\",\"name\":\"Toolbar\",\"id\":\"p1016\",\"attributes\":{\"tools\":[{\"type\":\"object\",\"name\":\"HoverTool\",\"id\":\"p1009\",\"attributes\":{\"renderers\":\"auto\",\"tooltips\":[[\"Gene Set\",\"@gene_set\"],[\"Cluster\",\"@label\"],[\"Library\",\"@library\"]]}},{\"type\":\"object\",\"name\":\"PanTool\",\"id\":\"p1041\"},{\"type\":\"object\",\"name\":\"WheelZoomTool\",\"id\":\"p1042\"},{\"type\":\"object\",\"name\":\"ResetTool\",\"id\":\"p1043\"},{\"type\":\"object\",\"name\":\"SaveTool\",\"id\":\"p1044\"}]}},\"left\":[{\"type\":\"object\",\"name\":\"LinearAxis\",\"id\":\"p1034\",\"attributes\":{\"ticker\":{\"type\":\"object\",\"name\":\"BasicTicker\",\"id\":\"p1035\",\"attributes\":{\"mantissas\":[1,2,5]}},\"formatter\":{\"type\":\"object\",\"name\":\"BasicTickFormatter\",\"id\":\"p1036\"},\"axis_label\":\"UMAP_2\",\"major_label_policy\":{\"type\":\"object\",\"name\":\"AllLabels\",\"id\":\"p1037\"},\"major_label_text_font_size\":\"0pt\",\"major_tick_line_color\":null,\"minor_tick_line_color\":null}}],\"right\":[{\"type\":\"object\",\"name\":\"Legend\",\"id\":\"p1063\",\"attributes\":{\"items\":[{\"type\":\"object\",\"name\":\"LegendItem\",\"id\":\"p1064\",\"attributes\":{\"label\":{\"type\":\"value\",\"value\":\"Cluster 0\"},\"renderers\":[{\"id\":\"p1060\"}],\"index\":0}},{\"type\":\"object\",\"name\":\"LegendItem\",\"id\":\"p1065\",\"attributes\":{\"label\":{\"type\":\"value\",\"value\":\"Cluster 1\"},\"renderers\":[{\"id\":\"p1060\"}],\"index\":15}},{\"type\":\"object\",\"name\":\"LegendItem\",\"id\":\"p1066\",\"attributes\":{\"label\":{\"type\":\"value\",\"value\":\"Cluster 2\"},\"renderers\":[{\"id\":\"p1060\"}],\"index\":26}},{\"type\":\"object\",\"name\":\"LegendItem\",\"id\":\"p1067\",\"attributes\":{\"label\":{\"type\":\"value\",\"value\":\"Cluster 3\"},\"renderers\":[{\"id\":\"p1060\"}],\"index\":36}},{\"type\":\"object\",\"name\":\"LegendItem\",\"id\":\"p1068\",\"attributes\":{\"label\":{\"type\":\"value\",\"value\":\"Cluster 4\"},\"renderers\":[{\"id\":\"p1060\"}],\"index\":42}},{\"type\":\"object\",\"name\":\"LegendItem\",\"id\":\"p1069\",\"attributes\":{\"label\":{\"type\":\"value\",\"value\":\"Cluster 5\"},\"renderers\":[{\"id\":\"p1060\"}],\"index\":48}},{\"type\":\"object\",\"name\":\"LegendItem\",\"id\":\"p1070\",\"attributes\":{\"label\":{\"type\":\"value\",\"value\":\"Cluster 6\"},\"renderers\":[{\"id\":\"p1060\"}],\"index\":53}}]}}],\"below\":[{\"type\":\"object\",\"name\":\"LinearAxis\",\"id\":\"p1027\",\"attributes\":{\"ticker\":{\"type\":\"object\",\"name\":\"BasicTicker\",\"id\":\"p1028\",\"attributes\":{\"mantissas\":[1,2,5]}},\"formatter\":{\"type\":\"object\",\"name\":\"BasicTickFormatter\",\"id\":\"p1029\"},\"axis_label\":\"UMAP_1\",\"major_label_policy\":{\"type\":\"object\",\"name\":\"AllLabels\",\"id\":\"p1030\"},\"major_label_text_font_size\":\"0pt\",\"major_tick_line_color\":null,\"minor_tick_line_color\":null}}],\"center\":[{\"type\":\"object\",\"name\":\"Grid\",\"id\":\"p1033\",\"attributes\":{\"axis\":{\"id\":\"p1027\"}}},{\"type\":\"object\",\"name\":\"Grid\",\"id\":\"p1040\",\"attributes\":{\"dimension\":1,\"axis\":{\"id\":\"p1034\"}}},{\"id\":\"p1063\"}],\"output_backend\":\"svg\"}}],\"callbacks\":{\"type\":\"map\"}}};\n",
       "  const render_items = [{\"docid\":\"21fc5328-2d3f-4384-ae3e-8e121c9afa5b\",\"roots\":{\"p1010\":\"fa1245eb-af74-426e-b931-39e957f460da\"},\"root_ids\":[\"p1010\"]}];\n",
       "  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "  }\n",
       "  if (root.Bokeh !== undefined) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    let attempts = 0;\n",
       "    const timer = setInterval(function(root) {\n",
       "      if (root.Bokeh !== undefined) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else {\n",
       "        attempts++;\n",
       "        if (attempts > 100) {\n",
       "          clearInterval(timer);\n",
       "          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n",
       "        }\n",
       "      }\n",
       "    }, 10, root)\n",
       "  }\n",
       "})(window);"
      ],
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "p1010"
      }
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%appyter code_eval\n",
    "if umap:\n",
    "    umap = UMAP_Visualization(gmt_df = [df])\n",
    "    umap_df = umap.process_scatterplot(maxdf = umap_maxdf, mindf = umap_mindf, nneighbors = umap_num_neighbors)\n",
    "    fig = umap.get_scatterplot(umap_df)\n",
    "    show(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'umap_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[40], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m umap_fig \u001b[38;5;241m=\u001b[39m umap\u001b[38;5;241m.\u001b[39mget_scatterplot(\u001b[43mumap_df\u001b[49m)\n\u001b[0;32m      2\u001b[0m show(fig)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'umap_df' is not defined"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
