{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Independent Enrichment Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This Appyter performs enrichment analysis given an input set of items, for example, gene symbols, and a library of sets in GMT format, for example, a gene set library. The Appyter performs the Fisher exact test to compute enrichment p-value and q-values, and reports the results as a sorted table, a bar graph, and a Manhattan plot. For ranked items, the appyter is able to return a bridge plot for each set in the library and reports the p- and q-values by the Mann-Whitney U test. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from maayanlab_bioinformatics.enrichment.crisp import enrich_crisp, fisher_overlap\n",
    "from maayanlab_bioinformatics.plotting import bridge_plot\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from IPython.display import display, FileLink, Markdown, HTML\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from scipy.stats import mannwhitneyu\n",
    "from collections import OrderedDict\n",
    "import urllib\n",
    "# Manhattan Plot Imports\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "\n",
    "# Bokeh\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool, CustomJS, ColumnDataSource, Span, Select, PreText, Paragraph\n",
    "from bokeh.layouts import layout, row, column, gridplot\n",
    "from bokeh.palettes import all_palettes\n",
    "\n",
    "import base64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% do SectionField(\n",
    "    name='Set_Section',\n",
    "    title='Submit Your Set',\n",
    "    subtitle='Upload a text file containing your set or copy and paste your set into the text box below (one item per row). You can also try the default set provided. ',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='Library_Section',\n",
    "    title='Submit Your Library',\n",
    "    subtitle='Upload a GMT file containing your library. You can also load the default library.',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='Background_Section',\n",
    "    title='(Optional) Submit Your Background',\n",
    "    subtitle='Upload a text file containing a list of background items to use as a filter for both the input and the libary sets. You can also copy and paste your background items into the text box below (one item per row). The default is no background and no filtering out any items.',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% set set_kind = TabField(\n",
    "    name='set_kind',\n",
    "    label='Set',\n",
    "    default='Try Example 1 (Gene Set)',\n",
    "    description='Paste or upload your set',\n",
    "    choices={\n",
    "        'Paste': [ \n",
    "            TextField(\n",
    "                name='set_input1',\n",
    "                label='Set',\n",
    "                default='',\n",
    "                description='Paste your set (one item per row). Names in the set should match the names in the GMT file.',\n",
    "                section = 'Set_Section'\n",
    "            )\n",
    "        ],\n",
    "        \n",
    "        'Upload': [\n",
    "            FileField(\n",
    "                name='set_filename',\n",
    "                label='Set File',\n",
    "                default='',\n",
    "                description='Upload your set as a text file (one item per row). Names in the set should match the names in the GMT file.',\n",
    "                section = 'Set_Section'\n",
    "            ),\n",
    "        ],\n",
    "        \n",
    "        'Try Example 1 (Gene Set)': [\n",
    "            TextField(\n",
    "                name='set_input2',\n",
    "                label='Set',\n",
    "                default='TAAR9\\nEBF2\\nWDR78\\nRRAGA\\nSPATA18\\nSPINT2\\nMRGPRD\\nCD9\\nRBP1\\nCYB5RL\\nMXRA8\\nPM20D1\\nITIH5\\nEPAS1\\nAHCYL2\\nPANK2\\nPON2\\nLRP5\\nSLC5A3\\nNSL1\\nCLDN2\\nLRP8\\nAQP1\\nCLDN1\\nTMEM72\\nGNG4\\nNHLH2\\nC10ORF107\\nS100A13\\nLY6G6C\\nPOF1B\\nWLS\\nC2ORF82\\nFZD4\\nCOG7\\nFZD6\\nFOXF1\\nFZD7\\nERLIN2\\nTYSND1\\nACADSB\\nOR51I2\\nPARP12\\nPPFIBP2\\nATP4A\\nALDH7A1\\nTCN2\\nSLCO5A1\\nSFXN4\\nPRR15\\nMOXD1\\nCAPSL\\nCOL13A1\\nC1ORF177\\nWFDC2\\nSLC6A2\\nDNALI1\\nTNS1\\nLGALS2\\nT\\nBLOC1S1\\nHMOX1\\nPDK4\\nLRAT\\nMNX1\\nSLC19A1\\nHOXC9\\nSCARF2\\nAS3MT\\nARGLU1\\nACE\\nANXA2\\nCARD9\\nPAX7\\nSORCS1\\nRAB33B\\nPHOX2A\\nKIF9\\nCLDN16\\nPTPRB\\nID3\\nITPKB\\nNCR1\\nGAS6\\nCC2D1B\\nATR\\nMYCBP\\nIGSF6\\nTPH1\\nWFIKKN2\\nIGSF5\\nACY3\\nMAOA\\nCAB39L\\nCTSZ\\nPRDM16\\nCYP7A1\\nLIMD1\\nTMEM27\\nSLC22A18\\nKRT28\\nTIMP3\\nEMB\\nRNF152\\nPLEKHN1\\nCLIC3\\nSTRA6\\nCTSC\\nCGNL1\\nPARP4\\nTMEM176B\\nELOVL7\\nSORBS3\\nGPR4\\nF5\\nGUCA2B\\nSERPINB6\\nHADHB\\nFOXR1\\nNBR1\\nSHKBP1\\nRLIM\\nDHRS13\\nHRSP12\\nCD63\\nCCL11\\nF13A1\\nFAM69C\\nKCNA7\\nHCCS\\nGUCA1A\\nADAMTSL2\\nLMAN1\\nING3\\nEGFLAM\\nSCML4\\nOLFML1\\nSOSTDC1\\nCTNNA1\\nC16ORF78\\nFADS1\\nCCDC157\\nPDGFRB\\nCA12\\nCD164\\nPRLR\\nLRRC69\\nUNC5CL\\nMPEG1\\nSLC31A1\\nTECRL\\nVCAM1\\nATP11A\\nUBXN10\\nZNF558\\nDYDC1\\nCD69\\nS100A8\\nFIGF\\nPHLDB2\\nERVFRD-1\\nCD82\\nASB14\\nGPR65\\nVWCE\\nTEKT1\\nTEKT4\\nMSX1\\nSLC16A9\\nZNF423\\nCA14\\nIGFBP2\\nSLC30A7\\nLRRC46\\nPDIA2\\nPPEF1\\nEPHX1\\nFANCM\\nRBPMS\\nTTC21A\\nMR1\\nDDX52\\nLSM5\\nKRT31\\nMAVS\\nTMEM237\\nSMO\\nC6ORF118\\nPGPEP1L\\nIL7R\\nC21ORF62\\nC11ORF97\\nDOCK6\\nAKNA\\nISYNA1\\nCD151\\nCBFB\\nPYROXD2\\nSLC2A1\\nGSTCD\\nLGALS3BP\\nHIGD1B\\nAK7\\nLTBP1\\nARHGAP5\\nRGS5\\nSALL1\\nCOBLL1\\nFHAD1\\nMAEL\\nBTLA\\nIGFBP7\\nODF1\\nACAA2\\nKL\\nTTC16\\nEMX2\\nTTC12\\nGGH\\nCCDC37\\nCFLAR\\nGPR98\\nLAMB2\\nBICC1\\nBMP6\\nCUL4B\\nDNAJC3\\nSP1\\nDAP\\nDNAJC1\\nPIKFYVE\\nDMRTA1\\nALPL\\nMTRF1L\\nBCAR3\\nKDM5D\\nSHC4\\nTTC25\\nDBH\\nDBI\\nCHD1\\nWNT6\\nSPN\\nTTC23L\\nPLTP\\nCYP26B1\\nCASP6\\nTMEM204\\nTMEM207\\nCCDC180\\nCCDC34\\nCA9\\nOVGP1\\nPLEKHG2\\nCPT1A\\nPLEKHG3\\nMYO10\\nRNASET2\\nTBC1D9\\nNAGA\\nPCOLCE\\nMUT\\nFOXJ1\\nSOD3\\nATOX1\\nKRT73\\nSNTB1\\nRP2\\nRPIA\\nCOL8A1\\nALS2\\nCOL8A2\\nSMPDL3A\\nPCOLCE2\\nSLC25A13\\nTAF3\\nFOLR1\\nITGB2\\nHEMGN\\nPRPS2\\nSLC24A5\\nFLT1\\nALAS2\\nLSP1\\nSYCP2\\nSEMA3B\\nETFB\\nPRELP\\nZBTB40\\nPBXIP1\\nSLC4A5\\nCLN8\\nEFS\\nTTR\\nRBM3\\nHECTD3\\nNAGLU\\nALDH2\\nCTNNAL1\\nPCBD1\\nCYTH2',\n",
    "                description='Paste your set (one item per row). Names in the set should match the names in the GMT file.',\n",
    "                section = 'Set_Section'\n",
    "            )\n",
    "        ],\n",
    "        'Try Example 2 (Drug Set)': [\n",
    "            TextField(\n",
    "                name='set_input3',\n",
    "                label='Set',\n",
    "                default='hexachlorophene\\nlopinavir\\nbazedoxifene\\nabemaciclib\\ncamostat\\nmefloquine\\ncyclosporine\\nanidulafungin\\nchloroquine\\namodiaquine\\nloperamide\\nalmitrine\\nhydroxychloroquine\\nniclosamide\\nivacaftor\\nproscillaridin\\nremdesivir',\n",
    "                description='Paste your set (one item per row). Names in the set should match the names in the GMT file.',\n",
    "                section = 'Set_Section'\n",
    "            )\n",
    "        ],\n",
    "        \n",
    "    },\n",
    "    section = 'Set_Section',\n",
    ") %}\n",
    "\n",
    "{% set is_ranked = BoolField(\n",
    "    name='is_ranked', \n",
    "    label='Ranked?', \n",
    "    default='false',\n",
    "    description='Check if your input set (>10,000 items) is ranked.', \n",
    "    section='Set_Section',\n",
    ") \n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set library_kind = TabField(\n",
    "    name='library_kind',\n",
    "    label='Library',\n",
    "    default='Upload',\n",
    "    description='',\n",
    "    choices={\n",
    "        'Upload': [ \n",
    "            FileField(\n",
    "                name='library_filename', \n",
    "                label='Library file (.gmt or .txt)', \n",
    "                default='GeneSet_Allen_Brain_Atlas_scRNAseq_10x.gmt',\n",
    "                examples={'GeneSet_Allen_Brain_Atlas_scRNAseq_10x.gmt': \"https://appyters.maayanlab.cloud/storage/Independent_Enrichment_Analysis/Example1_GeneSet_Allen_Brain_Atlas_scRNAseq_10x_2021.gmt\", 'DrugSet_L1000FWD_Signature_Down.txt': \"https://maayanlab.cloud/DrugEnrichr/geneSetLibrary?mode=text&libraryName=L1000FWD_Signature_Down\"}, \n",
    "                description='A tab-delimited file format that describes sets. Visit https://bit.ly/35crtXQ for more information.', \n",
    "                section='Library_Section')\n",
    "        ],\n",
    "        \n",
    "        'Select a library from Enrichr': [\n",
    "            ChoiceField(\n",
    "                name='enrichr_library', \n",
    "                description='Select one Enrichr library for which to an enrichment analysis.', \n",
    "                label='Enrichr Library', \n",
    "                default='WikiPathways_2019_Human', \n",
    "                section = 'section2',\n",
    "                choices=[\n",
    "                    'ARCHS4_Cell-lines',\n",
    "                    'ARCHS4_IDG_Coexp',\n",
    "                    'ARCHS4_Kinases_Coexp',\n",
    "                    'ARCHS4_TFs_Coexp',\n",
    "                    'ARCHS4_Tissues',\n",
    "                    'Achilles_fitness_decrease',\n",
    "                    'Achilles_fitness_increase',\n",
    "                    'Aging_Perturbations_from_GEO_down',\n",
    "                    'Aging_Perturbations_from_GEO_up',\n",
    "                    'Allen_Brain_Atlas_10x_scRNA_2021',\n",
    "                    'Allen_Brain_Atlas_down',\n",
    "                    'Allen_Brain_Atlas_up',\n",
    "                    'BioCarta_2013',\n",
    "                    'BioCarta_2015',\n",
    "                    'BioCarta_2016',\n",
    "                    'BioPlanet_2019',\n",
    "                    'BioPlex_2017',\n",
    "                    'CCLE_Proteomics_2020',\n",
    "                    'CORUM',\n",
    "                    'COVID-19_Related_Gene_Sets',\n",
    "                    'Cancer_Cell_Line_Encyclopedia',\n",
    "                    'ChEA_2013',\n",
    "                    'ChEA_2015',\n",
    "                    'ChEA_2016',\n",
    "                    'Chromosome_Location',           \n",
    "                    'Chromosome_Location_hg19',\n",
    "                    'ClinVar_2019',\n",
    "                    'dbGaP',\n",
    "                    'DSigDB',\n",
    "                    'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "                    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "                    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "                    'DisGeNET',\n",
    "                    'Disease_Perturbations_from_GEO_down',\n",
    "                    'Disease_Perturbations_from_GEO_up',\n",
    "                    'Disease_Signatures_from_GEO_down_2014',\n",
    "                    'Disease_Signatures_from_GEO_up_2014',\n",
    "                    'DrugMatrix',\n",
    "                    'Drug_Perturbations_from_GEO_2014',\n",
    "                    'Drug_Perturbations_from_GEO_down',\n",
    "                    'Drug_Perturbations_from_GEO_up',\n",
    "                    'ENCODE_Histone_Modifications_2013',\n",
    "                    'ENCODE_Histone_Modifications_2015',\n",
    "                    'ENCODE_TF_ChIP-seq_2014',\n",
    "                    'ENCODE_TF_ChIP-seq_2015',\n",
    "                    'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "                    'ESCAPE',\n",
    "                    'Elsevier_Pathway_Collection',\n",
    "                    'Enrichr_Libraries_Most_Popular_Genes',\n",
    "                    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "                    'Enrichr_Users_Contributed_Lists_2020',\n",
    "                    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "                    'GO_Biological_Process_2013',\n",
    "                    'GO_Biological_Process_2015',\n",
    "                    'GO_Biological_Process_2017',\n",
    "                    'GO_Biological_Process_2017b',\n",
    "                    'GO_Biological_Process_2018',\n",
    "                    'GO_Cellular_Component_2013',\n",
    "                    'GO_Cellular_Component_2015',\n",
    "                    'GO_Cellular_Component_2017',\n",
    "                    'GO_Cellular_Component_2017b',\n",
    "                    'GO_Cellular_Component_2018',\n",
    "                    'GO_Molecular_Function_2013',\n",
    "                    'GO_Molecular_Function_2015',\n",
    "                    'GO_Molecular_Function_2017',\n",
    "                    'GO_Molecular_Function_2017b',\n",
    "                    'GO_Molecular_Function_2018',\n",
    "                    'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "                    'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "                    'GWAS_Catalog_2019',\n",
    "                    'GeneSigDB',\n",
    "                    'Gene_Perturbations_from_GEO_down',\n",
    "                    'Gene_Perturbations_from_GEO_up',\n",
    "                    'Genes_Associated_with_NIH_Grants',\n",
    "                    'Genome_Browser_PWMs',\n",
    "                    'HMDB_Metabolites',\n",
    "                    'HMS_LINCS_KinomeScan',\n",
    "                    'HomoloGene',\n",
    "                    'HumanCyc_2015',\n",
    "                    'HumanCyc_2016',\n",
    "                    'Human_Gene_Atlas',\n",
    "                    'Human_Phenotype_Ontology',\n",
    "                    'huMAP',\n",
    "                    'InterPro_Domains_2019',\n",
    "                    'Jensen_COMPARTMENTS',\n",
    "                    'Jensen_DISEASES',\n",
    "                    'Jensen_TISSUES',\n",
    "                    'KEA_2013',\n",
    "                    'KEA_2015',\n",
    "                    'KEGG_2013',\n",
    "                    'KEGG_2015',\n",
    "                    'KEGG_2016',\n",
    "                    'KEGG_2019_Human',\n",
    "                    'KEGG_2019_Mouse',\n",
    "                    'Kinase_Perturbations_from_GEO_down',\n",
    "                    'Kinase_Perturbations_from_GEO_up',\n",
    "                    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "                    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "                    'LINCS_L1000_Chem_Pert_down',\n",
    "                    'LINCS_L1000_Chem_Pert_up',\n",
    "                    'LINCS_L1000_Ligand_Perturbations_down',\n",
    "                    'LINCS_L1000_Ligand_Perturbations_up',\n",
    "                    'Ligand_Perturbations_from_GEO_down',\n",
    "                    'Ligand_Perturbations_from_GEO_up',\n",
    "                    'lncHUB_lncRNA_Co-Expression',\n",
    "                    'MCF7_Perturbations_from_GEO_down',\n",
    "                    'MCF7_Perturbations_from_GEO_up',\n",
    "                    'MGI_Mammalian_Phenotype_2013',\n",
    "                    'MGI_Mammalian_Phenotype_2017',\n",
    "                    'MGI_Mammalian_Phenotype_Level_3',\n",
    "                    'MGI_Mammalian_Phenotype_Level_4',\n",
    "                    'MGI_Mammalian_Phenotype_Level_4_2019',\n",
    "                    'MSigDB_Computational',\n",
    "                    'MSigDB_Hallmark_2020',\n",
    "                    'MSigDB_Oncogenic_Signatures',\n",
    "                    'Microbe_Perturbations_from_GEO_down',\n",
    "                    'Microbe_Perturbations_from_GEO_up',\n",
    "                    'miRTarBase_2017',\n",
    "                    'Mouse_Gene_Atlas',\n",
    "                    'NCI-60_Cancer_Cell_Lines',\n",
    "                    'NCI-Nature_2015',\n",
    "                    'NCI-Nature_2016',\n",
    "                    'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "                    'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "                    'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "                    'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "                    'NURSA_Human_Endogenous_Complexome',\n",
    "                    'OMIM_Disease',\n",
    "                    'OMIM_Expanded',\n",
    "                    'Old_CMAP_down',\n",
    "                    'Old_CMAP_up',\n",
    "                    'PPI_Hub_Proteins',\n",
    "                    'Panther_2015',\n",
    "                    'Panther_2016',\n",
    "                    'Pfam_Domains_2019',\n",
    "                    'Pfam_InterPro_Domains',\n",
    "                    'PheWeb_2019',\n",
    "                    'Phosphatase_Substrates_from_DEPOD',\n",
    "                    'ProteomicsDB_2020',\n",
    "                    'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "                    'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "                    'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "                    'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "                    'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "                    'Reactome_2013',\n",
    "                    'Reactome_2015',\n",
    "                    'Reactome_2016',\n",
    "                    'SILAC_Phosphoproteomics',\n",
    "                    'SubCell_BarCode',\n",
    "                    'SysMyo_Muscle_Gene_Sets'\n",
    "                    'TF-LOF_Expression_from_GEO',\n",
    "                    'TF_Perturbations_Followed_by_Expression',\n",
    "                    'TG_GATES_2020',\n",
    "                    'TRANSFAC_and_JASPAR_PWMs',\n",
    "                    'TRRUST_Transcription_Factors_2019',\n",
    "                    'Table_Mining_of_CRISPR_Studies',\n",
    "                    'TargetScan_microRNA',\n",
    "                    'TargetScan_microRNA_2017',\n",
    "                    'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
    "                    'Tissue_Protein_Expression_from_ProteomicsDB.csv',\n",
    "                    'Transcription_Factor_PPIs',\n",
    "                    'UK_Biobank_GWAS_v1',\n",
    "                    'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "                    'VirusMINT',\n",
    "                    'Virus_Perturbations_from_GEO_down',\n",
    "                    'Virus_Perturbations_from_GEO_up',\n",
    "                    'WikiPathways_2013',\n",
    "                    'WikiPathways_2015',\n",
    "                    'WikiPathways_2016',\n",
    "                    'WikiPathways_2019_Human',\n",
    "                    'WikiPathways_2019_Mouse'\n",
    "                ]\n",
    "            )\n",
    "        ],\n",
    "        \n",
    "        \n",
    "    },\n",
    "    section = 'Library_Section',\n",
    ") %}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% set background_kind = TabField(\n",
    "    name='background_kind',\n",
    "    label='Background',\n",
    "    default='Paste',\n",
    "    description='Paste or upload your background set',\n",
    "    choices={\n",
    "        'Paste': [\n",
    "            TextField(\n",
    "                name='background_input',\n",
    "                label='Background Set',\n",
    "                default='',\n",
    "                description='Paste your background set (one item per row). Names in the background set should match the names in the GMT file.',\n",
    "                section = 'Background_Section'\n",
    "            ),\n",
    "        ],\n",
    "        'Upload': [\n",
    "            FileField(\n",
    "                name='background_filename',\n",
    "                label='Background File',\n",
    "                default='',\n",
    "                description='Upload your background set as a text file (one item per row). Names in the background set should match the names in the GMT file.',\n",
    "                section = 'Background_Section'\n",
    "            ),\n",
    "        ],\n",
    "    },\n",
    "    section = 'Background_Section',\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if set_kind.raw_value == 'Paste' or set_kind.raw_value == 'Try Example 1 (Gene Set)' or set_kind.raw_value == 'Try Example 2 (Drug Set)'%}\n",
    "set_input = {{ set_kind.value[0] }}\n",
    "{%- else %}\n",
    "set_filename = {{ set_kind.value[0] }}\n",
    "{%- endif %}\n",
    "\n",
    "{%- if library_kind.raw_value == 'Upload' %}\n",
    "library_kind = \"Upload\"\n",
    "library_filename = {{ library_kind.value[0] }}\n",
    "library_name = library_filename.replace(\"_\", \" \").replace(\".txt\", \"\").replace(\".gmt\", \"\")\n",
    "\n",
    "{%- else %}\n",
    "library_kind = \"Select a library from Enrichr\"\n",
    "library_filename = \"{{ library_kind.value[0] }}\"\n",
    "library_name = \"{{ library_kind.value[0] }}\"\n",
    "{%- endif %}\n",
    "\n",
    "\n",
    "{%- if background_kind.raw_value == 'Paste' %}\n",
    "background_input = {{ background_kind.value[0] }}\n",
    "{%- else %}\n",
    "background_filename = {{ background_kind.value[0] }}\n",
    "{%- endif %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_notebook()\n",
    "# Table Parameters\n",
    "significance_value = 0.05\n",
    "display_topk = 20\n",
    "\n",
    "# Bar Chart Parameters\n",
    "figure_file_format = ['png', 'svg']\n",
    "output_file_name = 'Enrichment_analysis_results_bar'\n",
    "color = 'lightskyblue'\n",
    "final_output_file_names = ['{0}.{1}'.format(output_file_name, file_type) for file_type in figure_file_format]\n",
    "topk = 10\n",
    "\n",
    "# Manhattan Plot Parameters\n",
    "manhattan_colors = ['#003f5c', '#7a5195', '#ef5675', '#ffa600']\n",
    "\n",
    "# Bridge Plot Parameters\n",
    "bridge_plot_topk = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if set_kind.raw_value == 'Paste' or set_kind.raw_value == 'Try Example 1 (Gene Set)' or set_kind.raw_value == 'Try Example 2 (Drug Set)' %}\n",
    "items = set_input.split('\\n')\n",
    "items = [x.strip() for x in items]\n",
    "{%- else %}\n",
    "open_set_file = open(set_filename,'r')\n",
    "lines = open_set_file.readlines()\n",
    "items = [x.strip() for x in lines]\n",
    "open_set_file.close()\n",
    "{%- endif %}\n",
    "\n",
    "# remove duplicates in items\n",
    "items = list(OrderedDict.fromkeys(items)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if background_kind.raw_value == 'Paste' %}\n",
    "background_items = background_input.split('\\n')\n",
    "background_items = [x.strip() for x in background_items]\n",
    "{%- else %}\n",
    "open_background_file = open(background_filename,'r')\n",
    "lines = open_background_file.readlines()\n",
    "background_items = [x.strip() for x in lines]\n",
    "open_background_file.close()\n",
    "{%- endif %}\n",
    "condition1 = len(background_items) == 1 and background_items[0] == \"\"\n",
    "condition2 = len(background_items) > 0\n",
    "if condition1 == False and condition2 == True:\n",
    "    items = [x for x in items if x in background_items]\n",
    "    background_items_bool = True\n",
    "else:\n",
    "    background_items_bool = False\n",
    "\n",
    "is_ranked = {{is_ranked.value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_library(library_name):\n",
    "    with open(f\"{library_name}\", \"w\") as fw:\n",
    "        with urllib.request.urlopen(f'https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName={library_name}') as f:\n",
    "            for line in f.readlines():\n",
    "                fw.write(line.decode('utf-8'))\n",
    "                fw.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load(library_filename, items, background_items):\n",
    "    if library_kind == \"Select a library from Enrichr\":\n",
    "        download_library(library_filename)\n",
    "    library_data, average_items_in_library = load_library(library_filename, background_items)\n",
    "    # to upper case\n",
    "    items = [x.upper() for x in items]\n",
    "    validate_inputs(items, library_data)\n",
    "    \n",
    "    try:\n",
    "        validate_ranked_items(items, average_items_in_library)\n",
    "        \n",
    "    except Exception as error:\n",
    "        display(Markdown(f'#### **Warning: {error}**'))\n",
    "    return library_data, items\n",
    "\n",
    "def load_library(library_filename, background_items):\n",
    "    library_data = dict()\n",
    "    items_in_library = list()\n",
    "    with open(library_filename, \"r\") as f:\n",
    "        lines = f.readlines()\n",
    "        for line in lines:\n",
    "            splited = line.strip().split(\"\\t\")\n",
    "            \n",
    "            if background_items_bool == True:\n",
    "                elements = [x for x in splited[2:] if x in background_items]\n",
    "            else:\n",
    "                elements = splited[2:]\n",
    "            if len(elements) > 0:\n",
    "                # to upper case\n",
    "                library_data[splited[0]] = [x.upper() for x in elements]\n",
    "                items_in_library.append(len(elements))\n",
    "    return library_data, (sum(items_in_library)+0.0)/len(items_in_library)\n",
    "\n",
    "def validate_inputs(items, library_data):\n",
    "    library_items = set()\n",
    "    for key, values in library_data.items():\n",
    "        library_items.update(set(values))\n",
    "    if len(items) == 0:\n",
    "        raise Exception('No items in the input set. Please check the background information.') \n",
    "    if len(library_data.keys()) == 0:\n",
    "        raise Exception('No items in the input library. Please check the background information.') \n",
    "    if len(set(items).intersection(library_items)) == 0:\n",
    "        raise Exception('No matches in the input set and library.')    \n",
    "        \n",
    "def validate_ranked_items(items, average_items_in_library):\n",
    "    if is_ranked == True and len(items) < average_items_in_library*10:\n",
    "        raise Exception(f'We recommend that the ranked input list should be ~10 times longer than the sets in the library.')\n",
    "\n",
    "# Enrichment analysis\n",
    "def get_library_iter(library_data):\n",
    "    for term in library_data.keys():\n",
    "        single_set = library_data[term]\n",
    "        yield term, single_set\n",
    "\n",
    "def get_enrichment_results(items, library_data):\n",
    "    return sorted(enrich_crisp(items, get_library_iter(library_data), 20000, True), key=lambda r: r[1].pvalue)\n",
    "\n",
    "\n",
    "def get_pvalue(row, unzipped_results, all_results):\n",
    "    if row['Name'] in list(unzipped_results[0]):\n",
    "        index = list(unzipped_results[0]).index(row['Name'])\n",
    "        return all_results[index][1].pvalue\n",
    "    else:\n",
    "        return 1\n",
    "    \n",
    "# Call enrichment results and return a plot and dataframe for Scatter Plot\n",
    "def get_values(obj_list):\n",
    "    pvals = []\n",
    "    odds_ratio = []\n",
    "    n_overlap = []\n",
    "    overlap = []\n",
    "    for i in obj_list:\n",
    "        pvals.append(i.pvalue)\n",
    "        odds_ratio.append(i.odds_ratio)\n",
    "        n_overlap.append(i.n_overlap)\n",
    "        overlap.append(i.overlap)\n",
    "    return pvals, odds_ratio, n_overlap, overlap\n",
    "\n",
    "def get_qvalue(p_vals):\n",
    "    r = multipletests(p_vals, method=\"fdr_bh\")\n",
    "    return r[1]\n",
    "\n",
    "\n",
    "def enrichment_analysis(items, library_data):    \n",
    "    all_results = get_enrichment_results(items, library_data)\n",
    "    unzipped_results = list(zip(*all_results))\n",
    "    pvals, odds_ratio, n_overlap, overlap = get_values(unzipped_results[1])\n",
    "    df = pd.DataFrame({\"Name\":unzipped_results[0], \"p value\": pvals, \\\n",
    "                       \"odds_ratio\": odds_ratio, \"n_overlap\": n_overlap, \"overlap\": overlap})\n",
    "    df[\"-log(p value)\"] = -np.log10(df[\"p value\"])\n",
    "    df[\"q value\"] = get_qvalue(df[\"p value\"].tolist())\n",
    "    return [list(unzipped_results[0])], [pvals], df\n",
    "\n",
    "# Output a table of significant p-values\n",
    "def create_download_link(df, title = \"Download CSV file of this table\", filename = \"data.csv\"):  \n",
    "    csv = df.to_csv(index = False)\n",
    "    b64 = base64.b64encode(csv.encode())\n",
    "    payload = b64.decode()\n",
    "    html = '<a download=\"{filename}\" href=\"data:text/csv;base64,{payload}\" target=\"_blank\">{title}</a>'\n",
    "    html = html.format(payload=payload, title=title, filename=filename)\n",
    "    return HTML(html)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bar Chart Functions\n",
    "def enrichr_figure(all_terms, all_pvalues, all_qvalues, plot_names, all_libraries, bar_color, topk=10): \n",
    "    all_terms = [all_terms[0][:topk]]\n",
    "    all_pvalues = [all_pvalues[0][:topk]]\n",
    "    all_qvalues = [all_qvalues[:topk]]\n",
    "    # Bar colors\n",
    "    if bar_color != 'lightgrey':\n",
    "        bar_color_not_sig = 'lightgrey'\n",
    "        edgecolor=None\n",
    "        linewidth=0\n",
    "    else:\n",
    "        bar_color_not_sig = 'white'\n",
    "        edgecolor='black'\n",
    "        linewidth=1    \n",
    "\n",
    "    plt.figure(figsize=(24, 12))\n",
    "    \n",
    "    i = 0\n",
    "    bar_colors = [bar_color if (x < 0.05) else bar_color_not_sig for x in all_pvalues[i]]\n",
    "    fig = sns.barplot(x=np.log10(all_pvalues[i])*-1, y=all_terms[i], palette=bar_colors, edgecolor=edgecolor, linewidth=linewidth)\n",
    "    fig.axes.get_yaxis().set_visible(False)\n",
    "    fig.set_title(all_libraries[i], fontsize=26)\n",
    "    fig.set_xlabel('−log₁₀(p‐value)', fontsize=25)\n",
    "    fig.tick_params(axis='x', which='major', labelsize=20)\n",
    "    if max(np.log10(all_pvalues[i])*-1)<1:\n",
    "        fig.xaxis.set_ticks(np.arange(0, max(np.log10(all_pvalues[i])*-1), 0.1))\n",
    "    for ii,annot in enumerate(all_terms[i]):\n",
    "        if all_qvalues[i][ii] < 0.05:\n",
    "            annot = '  *'.join([annot, str(str(np.format_float_scientific(all_qvalues[i][ii], precision=2)))]) \n",
    "        else:\n",
    "            annot = '  '.join([annot, str(str(np.format_float_scientific(all_qvalues[i][ii], precision=2)))])\n",
    "\n",
    "        title_start= max(fig.axes.get_xlim())/200\n",
    "        fig.text(title_start, ii, annot, ha='left', wrap = True, fontsize = 26)\n",
    "\n",
    "    fig.spines['right'].set_visible(False)\n",
    "    fig.spines['top'].set_visible(False)\n",
    "    # Save results \n",
    "    for plot_name in plot_names:\n",
    "        plt.savefig(plot_name, bbox_inches = 'tight')\n",
    "    \n",
    "    # Show plot \n",
    "    plt.show()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Manhattan Plots\n",
    "def manhattan(df):\n",
    "    df = df.sort_values(\"Name\")\n",
    "    list_of_xaxis_values = df[\"Name\"].values.tolist()\n",
    "\n",
    "    # define the output figure and the features we want\n",
    "    p = figure(x_range = list_of_xaxis_values, plot_height=300, plot_width=750, tools='pan, box_zoom, hover, reset, save')\n",
    "\n",
    "    # loop over all libraries\n",
    "    r = []\n",
    "    color_index = 0\n",
    "    if color_index >= len(manhattan_colors):\n",
    "        color_index = 0 \n",
    "\n",
    "    # calculate actual p value from -log(p value)\n",
    "    actual_pvalues = []\n",
    "    for log_value in df[\"-log(p value)\"].values.tolist():\n",
    "        actual_pvalues += [\"{:.5e}\".format(10**(-1*log_value))]\n",
    "\n",
    "    # define ColumnDataSource with our data for this library\n",
    "    source = ColumnDataSource(data=dict(\n",
    "        x = df[\"Name\"].values.tolist(),\n",
    "        y = df[\"-log(p value)\"].values.tolist(),\n",
    "        pvalue = actual_pvalues,\n",
    "    ))\n",
    "\n",
    "    # plot data from this library\n",
    "    r += [p.circle(x = 'x', y = 'y', size=5, fill_color=manhattan_colors[color_index], line_color = manhattan_colors[color_index], line_width=1, source = source)]\n",
    "    color_index += 1\n",
    "\n",
    "    p.background_fill_color = 'white'\n",
    "    p.xaxis.major_tick_line_color = None \n",
    "    p.xaxis.major_label_text_font_size = '0pt'\n",
    "    p.y_range.start = 0\n",
    "    p.yaxis.axis_label = '-log(p value)'\n",
    "\n",
    "    p.hover.tooltips = [\n",
    "        (\"Name\", \"@x\"),\n",
    "        (\"p value\", \"@pvalue\"),\n",
    "    ]\n",
    "    p.output_backend = \"svg\"\n",
    "    \n",
    "    # returns the plot\n",
    "    return p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create bridge plots and Perform MannWhitney U tests\n",
    "def get_mannwhitneyu_pvalue(set_name):\n",
    "    all_ranks = list(range(len(items)))\n",
    "    common_items = list(set(items).intersection(library_data[set_name]))\n",
    "    selected_ranks = list()\n",
    "    for common_item in common_items:\n",
    "        selected_ranks.append(items.index(common_item))\n",
    "        \n",
    "    if len(common_items) == 0:\n",
    "        return float('NaN')\n",
    "    _, p_value = mannwhitneyu(all_ranks, selected_ranks)\n",
    "    return p_value\n",
    "\n",
    "def test_mannwhitneyu(library_data):\n",
    "    mann_whitney_results = dict()\n",
    "    set_names = library_data.keys()\n",
    "    mann_whitney_results[\"p value\"] = dict()\n",
    "    for set_name in set_names:\n",
    "        mann_whitney_results[\"p value\"][set_name] = get_mannwhitneyu_pvalue(set_name)\n",
    "    df = pd.DataFrame(mann_whitney_results)\n",
    "    df = df.dropna()\n",
    "    df = df.sort_values(\"p value\", ascending=True).reset_index()\n",
    "    df[\"q value\"] = get_qvalue(df[\"p value\"].tolist())\n",
    "    df.columns = [\"Set Name\", \"p value\", \"q value\"]\n",
    "    return df\n",
    "\n",
    "def return_bridge_plot(key):\n",
    "    result = dict()\n",
    "    series = pd.Series(items)\n",
    "    input_series = series.isin(library_data[key])\n",
    "    x, y = bridge_plot(input_series)\n",
    "    result[\"x\"] = x\n",
    "    result[\"y\"] = y\n",
    "    return result\n",
    "\n",
    "def extract_pval(dictionary, key):\n",
    "    if key in dictionary:\n",
    "        return str(dictionary[key])\n",
    "    else:\n",
    "        return \"undefined\"\n",
    "    \n",
    "def plot_bridge(mannwhitney_df, bridge_plot_topk):\n",
    "    \n",
    "    pval = mannwhitney_df.set_index(\"Set Name\").to_dict()['p value']    \n",
    "    pval = {key: round(item, 6) for key, item in pval.items()}\n",
    "    \n",
    "    qval = mannwhitney_df.set_index(\"Set Name\").to_dict()['q value']    \n",
    "    qval = {key: round(item, 6) for key, item in qval.items()}\n",
    "    \n",
    "    # select topk\n",
    "    mannwhitney_df = mannwhitney_df.iloc[:bridge_plot_topk, :]\n",
    "    options = mannwhitney_df[\"Set Name\"].tolist()\n",
    "    \n",
    "    \n",
    "    # run bridge plot for all sets\n",
    "    bridge_plot_results = dict()\n",
    "    for key in options:\n",
    "        bridge_plot_results[key] = return_bridge_plot(key)\n",
    "    \n",
    "    # init\n",
    "    source = ColumnDataSource(data=dict(x=bridge_plot_results[options[0]][\"x\"], y=bridge_plot_results[options[0]][\"y\"]))\n",
    "    plot = figure(plot_width=400, plot_height=400)\n",
    "    plot.line('x', 'y', source=source, line_width=3, line_alpha=0.6)\n",
    "    plot.output_backend = \"svg\"\n",
    "    \n",
    "    select = Select(title=\"Select a set:\", value=options[0], options=options)\n",
    "    \n",
    "    default_caption_text = \"Figure 1. Bridge plot of {} (p-value {}, q-value {}). The x-axis shows the ranks of items, and the y-axis shows scores.\"\n",
    "\n",
    "    pre = Paragraph(text = default_caption_text.format(select.value, extract_pval(pval, select.value), extract_pval(qval, select.value)),\n",
    "    width=500, height=100, style={\"font-family\":'Helvetica', \"font-style\": \"italic\"})\n",
    "    \n",
    "    callback = CustomJS(args=dict(source=source, pre=pre, pval=pval, bridge_plot_results=bridge_plot_results), code=\"\"\"        \n",
    "        var data = source.data\n",
    "        const val = cb_obj.value;\n",
    "        data.x = bridge_plot_results[val]['x']\n",
    "        data.y = bridge_plot_results[val]['y']\n",
    "        \n",
    "        pre.text = \"Figure 1. Bridge plot of \" + val + \" (p-value \" + pval[val] +\"). The x-axis shows the ranks of items, and the y-axis shows scores.\" \n",
    "        \n",
    "        source.change.emit();\n",
    "        \n",
    "    \"\"\")\n",
    "    select.js_on_change('value', callback)\n",
    "    \n",
    "    \n",
    "    col = column(select, plot, pre)\n",
    "    show(col)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if is_ranked.value == False %}\n",
    "# Enrichment Analysis\n",
    "The table below displays the top 20 enrichment analysis results for the given set library. The table contains the sets name, p-value, odds ratio, the number of overlapping items, overlapping items, and -log(p-value). The table is sorted by p-values in ascending order. The full results are downloadable in CSV format.\n",
    "{% endif %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if is_ranked.value == False %}\n",
    "library_data, items = load(library_filename, items, background_items)\n",
    "results, pvals, results_df = enrichment_analysis(items, library_data)\n",
    "if 'p value' in results_df.columns:\n",
    "    sorted_df = results_df.sort_values(by = ['p value'])\n",
    "    filtered_df = sorted_df.iloc[:display_topk]\n",
    "    if len(filtered_df) != 0:\n",
    "        display(HTML(filtered_df.to_html(index = False)))\n",
    "        display(Markdown(f\"*Table 1. Enrichment analysis results of {library_name}*\"))        \n",
    "        display(create_download_link(sorted_df))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if is_ranked.value == False %}\n",
    "# Bar Chart\n",
    "{% endif %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if is_ranked.value == False %}\n",
    "display(Markdown(f\"The bar chart below shows the top {topk} enriched terms in a given library. Colored bars correspond to terms with significant p-values (<0.05). The bar chart is downloadable as an image in the PNG and SVG formats. \"))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if is_ranked.value == False %}\n",
    "enrichr_figure(results, pvals, results_df[\"q value\"].tolist(), final_output_file_names, [library_name], color, topk)\n",
    "display(Markdown(f\"*Figure 1. Bar chart of the top {topk} enriched terms in {library_name}, along with their corresponding p-values. Colored bars correspond to terms with significant p-values (<0.05). An asterisk next to a p-value indicates the term also has a significant q-value (<0.05)*\"))     \n",
    "    \n",
    "# Download Bar Chart\n",
    "for i, file in enumerate(final_output_file_names):\n",
    "    display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))\n",
    "    \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if is_ranked.value == False %}\n",
    "# Manhattan Plot\n",
    "In the Manhattan plot below, each line on the x-axis denotes a single set from the library, while the y-axis measures the −log₁₀(p‐value) for each set. Hovering over a point will display the name of the set and the associated p-value. You can also zoom, pan, and save the plot as an svg using the toolbar on the right.\n",
    "{% endif %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if is_ranked.value == False %}\n",
    "show(manhattan(results_df))\n",
    "display(Markdown(f\"*Figure 2. Manhattan plot that displays sets from {library_name} and their p-values on a -log10 scale.*\"))     \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if is_ranked.value == True %}\n",
    "# Bridge Plot\n",
    "In a bridge plot, the x-axis shows the ranks of items, and the y-axis shows scores that are calculated as follows. By walking down the ranked list of input items, if an item is in a set in the given library, the score will be incremented, otherwise, the score will be decremented. You can explore and see a bridge plot for each set of the library. You can zoom, pan, and download the plot in the SVG format using the functions on the toolbar next to the plot. P-values are calculated by the Mann-Whitney U test.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if is_ranked.value == True %}\n",
    "library_data, items = load(library_filename, items, background_items)\n",
    "# calculate p-values\n",
    "mannwhitney_df = test_mannwhitneyu(library_data)\n",
    "plot_bridge(mannwhitney_df, bridge_plot_topk)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if is_ranked.value == True %}\n",
    "# Table of top-ranked sets\n",
    "The Appyter performs the Mann-Whitney U tests for the given ranked list against given sets in the library. The table shows the top-ranked terms based on p-values and the full results are downloadable.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if is_ranked.value == True %}\n",
    "display(mannwhitney_df.iloc[:display_topk].set_index(\"Set Name\"))\n",
    "display(Markdown(f\"*Table 1. Top {display_topk} sets and their associated p-values by the Mann-Whitney U test.*\"))     \n",
    "\n",
    "display(create_download_link(mannwhitney_df, title=\"Download CSV file of this table\"))\n",
    "\n",
    "{% endif %}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "appyter_new",
   "language": "python",
   "name": "appyter_new"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
