# Independent Enrichment Analysis

Independent Enrichment Analysis Appyter performs enrichment analysis without the need to use external libraries or API. The required inputs are a  set and a set library. The Appyter performs the Fisher exact test to compute enrichment p-value and q-values, and reports the results as a sorted table, a bar graph, and a Manhattan plot. For ranked items, the appyter is able to return a bridge plot for each set in the library and to report its p-value by the Mann-Whitney U test. 

## **File Format**
To run the Appyter, a set of items and a library of sets are requiredas input. Background information input is optional.

* Please input a set of items as follows:  
ITEM 1  
ITEM 2  
ITEM 3  

* Please input a library of sets in GMT file format as follows:  
SET 1\t\tITEM 1\tITEM 2\tITEM 3  
SET 2\t\tITEM 3\tITEM 4\tITEM 5  
SET 3\t\tITEM 2\tITEM 5\tITEM 6  
SET 4\t\tITEM 1\tITEM 2\tITEM 6  

* Please input a set of background as follows:  
ITEM 1  
ITEM 2  
ITEM 3  


## **Examples**
We provide two examples for the Appyter.  

* Example 1: Gene Set Enrichment Analysis  
The example input set is a gene set of up-regulated in a part of brain. The input set library (Example1_GeneSet_Allen_Brain_Atlas_scRNAseq_10x_2021.gmt) is about cell types in the primary motor cortex from human and mouse brains. The library contains up- or down-regulated genes in each cell type in the cortex. The library was generated by processing Allen Brain Map's Human/Mouse M1 10x single cell RNA-seq datasets (https://portal.brain-map.org/atlases-and-data/rnaseq/human-m1-10x). Through the Appyter results, you can find enriched cell types in the primary motor cortex in brain in the given gene set. Various gene set libraries are available at Enrichr (https://maayanlab.cloud/Enrichr/)  

* Example 2: Drug Set Enrichment Analysis  
The example input set is a drug set of potential COVID-19 therapeutics. The input set library (Example2_DrugSet_L1000FWD_Signature_Down.txt) is about genes and their corresponding sets of drugs that can down-regulated the genes after treatment. The input set library was processed using LINCS L1000 datasets. After you run the example through the Appyter, you can find enriched genes that are likly to be down-regulated by the potential COVID-19 therapeutics in common. Various drug set libraries are available at DrugEnrichr (https://maayanlab.cloud/DrugEnrichr/) 

