{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichment Analysis Visualizer\n",
    "\n",
    "This appyter creates a variety of visualizations for enrichment analysis results for one selected Enrichr library, and may be run as either a standalone appyter from the [Appyter Catalog](https://appyters.maayanlab.cloud/#/Enrichment_Analysis_Visualizer) or programmatically from the [Enrichr](https://maayanlab.cloud/Enrichr/) results page. \n",
    "\n",
    "For simplicity, the only inputs for this appyter are a gene list and one library. Other parameters are set to default values in the cell below. You can download the notebook, change these parameters, and rerun it if you wish.\n",
    "\n",
    "The pre-processed libraries used to create the scatter plot and hexagonal canvas visualizations can be found [here](https://github.com/MaayanLab/Enrichr-Viz-Appyter/tree/master/Enrichr-Processed-Library-Storage). \n",
    "\n",
    "**A link to the full analysis results on the Enrichr website can be found at the bottom of this page.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scatter Plot Imports\n",
    "from maayanlab_bioinformatics.enrichment import enrich_crisp\n",
    "import matplotlib as mpl\n",
    "import matplotlib.colors as colors\n",
    "import base64\n",
    "\n",
    "# Bar Chart Imports\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import json\n",
    "import requests\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import time\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from IPython.display import display, FileLink, Markdown, HTML\n",
    "\n",
    "# Hexagonal Canvas Imports\n",
    "import json\n",
    "import math\n",
    "import uuid\n",
    "import urllib\n",
    "from textwrap import dedent\n",
    "from string import Template\n",
    "from operator import itemgetter\n",
    "\n",
    "# Manhattan Plot Imports\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "\n",
    "# Bokeh\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool, CustomJS, ColumnDataSource, Span\n",
    "from bokeh.layouts import layout, row, column, gridplot\n",
    "from bokeh.palettes import all_palettes\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name='section1', \n",
    "    title = '1. Submit Your Gene List', \n",
    "    subtitle = 'Upload a text file containing your gene list or copy and paste your gene list into the text box below (One gene per row). You can also try the default gene list provided.', \n",
    "    img = 'enrichr-icon.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='section2', \n",
    "    title = '2. Choose Enrichr Library', \n",
    "    subtitle = 'Select one Enrichr library.', \n",
    "    img = 'enrichr-icon.png'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% set gene_list_kind = TabField(\n",
    "    name='gene_list_kind',\n",
    "    label='Gene List',\n",
    "    default='Paste',\n",
    "    description='Paste or upload your gene list',\n",
    "    choices={\n",
    "        'Paste': [\n",
    "            TextField(\n",
    "                name='gene_list_input',\n",
    "                label='Gene List',\n",
    "                default='NSUN3\\nPOLRMT\\nNLRX1\\nSFXN5\\nZC3H12C\\nSLC25A39\\nARSG\\nDEFB29\\nNDUFB6\\nZFAND1\\nTMEM77\\n5730403B10RIK\\nRP23-195K8.6\\nTLCD1\\nPSMC6\\nSLC30A6\\nLOC100047292\\nLRRC40\\nORC5L\\nMPP7\\nUNC119B\\nPRKACA\\nTCN2\\nPSMC3IP\\nPCMTD2\\nACAA1A\\nLRRC1\\n2810432D09RIK\\nSEPHS2\\nSAC3D1\\nTMLHE\\nLOC623451\\nTSR2\\nPLEKHA7\\nGYS2\\nARHGEF12\\nHIBCH\\nLYRM2\\nZBTB44\\nENTPD5\\nRAB11FIP2\\nLIPT1\\nINTU\\nANXA13\\nKLF12\\nSAT2\\nGAL3ST2\\nVAMP8\\nFKBPL\\nAQP11\\nTRAP1\\nPMPCB\\nTM7SF3\\nRBM39\\nBRI3\\nKDR\\nZFP748\\nNAP1L1\\nDHRS1\\nLRRC56\\nWDR20A\\nSTXBP2\\nKLF1\\nUFC1\\nCCDC16\\n9230114K14RIK\\nRWDD3\\n2610528K11RIK\\nACO1\\nCABLES1\\nLOC100047214\\nYARS2\\nLYPLA1\\nKALRN\\nGYK\\nZFP787\\nZFP655\\nRABEPK\\nZFP650\\n4732466D17RIK\\nEXOSC4\\nWDR42A\\nGPHN\\n2610528J11RIK\\n1110003E01RIK\\nMDH1\\n1200014M14RIK\\nAW209491\\nMUT\\n1700123L14RIK\\n2610036D13RIK\\nCOX15\\nTMEM30A\\nNSMCE4A\\nTM2D2\\nRHBDD3\\nATXN2\\nNFS1\\n3110001I20RIK\\nBC038156\\nLOC100047782\\n2410012H22RIK\\nRILP\\nA230062G08RIK\\nPTTG1IP\\nRAB1\\nAFAP1L1\\nLYRM5\\n2310026E23RIK\\nC330002I19RIK\\nZFYVE20\\nPOLI\\nTOMM70A\\nSLC7A6OS\\nMAT2B\\n4932438A13RIK\\nLRRC8A\\nSMO\\nNUPL2\\nTRPC2\\nARSK\\nD630023B12RIK\\nMTFR1\\n5730414N17RIK\\nSCP2\\nZRSR1\\nNOL7\\nC330018D20RIK\\nIFT122\\nLOC100046168\\nD730039F16RIK\\nSCYL1\\n1700023B02RIK\\n1700034H14RIK\\nFBXO8\\nPAIP1\\nTMEM186\\nATPAF1\\nLOC100046254\\nLOC100047604\\nCOQ10A\\nFN3K\\nSIPA1L1\\nSLC25A16\\nSLC25A40\\nRPS6KA5\\nTRIM37\\nLRRC61\\nABHD3\\nGBE1\\nPARP16\\nHSD3B2\\nESM1\\nDNAJC18\\nDOLPP1\\nLASS2\\nWDR34\\nRFESD\\nCACNB4\\n2310042D19RIK\\nSRR\\nBPNT1\\n6530415H11RIK\\nCLCC1\\nTFB1M\\n4632404H12RIK\\nD4BWG0951E\\nMED14\\nADHFE1\\nTHTPA\\nCAT\\nELL3\\nAKR7A5\\nMTMR14\\nTIMM44\\nSF1\\nIPP\\nIAH1\\nTRIM23\\nWDR89\\nGSTZ1\\nCRADD\\n2510006D16RIK\\nFBXL6\\nLOC100044400\\nZFP106\\nCD55\\n0610013E23RIK\\nAFMID\\nTMEM86A\\nALDH6A1\\nDALRD3\\nSMYD4\\nNME7\\nFARS2\\nTASP1\\nCLDN10\\nA930005H10RIK\\nSLC9A6\\nADK\\nRBKS\\n2210016F16RIK\\nVWCE\\n4732435N03RIK\\nZFP11\\nVLDLR\\n9630013D21RIK\\n4933407N01RIK\\nFAHD1\\nMIPOL1\\n1810019D21RIK\\n1810049H13RIK\\nTFAM\\nPAICS\\n1110032A03RIK\\nLOC100044139\\nDNAJC19\\nBC016495\\nA930041I02RIK\\nRQCD1\\nUSP34\\nZCCHC3\\nH2AFJ\\nPHF7\\n4921508D12RIK\\nKMO\\nPRPF18\\nMCAT\\nTXNDC4\\n4921530L18RIK\\nVPS13B\\nSCRN3\\nTOR1A\\nAI316807\\nACBD4\\nFAH\\nAPOOL\\nCOL4A4\\nLRRC19\\nGNMT\\nNR3C1\\nSIP1\\nASCC1\\nFECH\\nABHD14A\\nARHGAP18\\n2700046G09RIK\\nYME1L1\\nGK5\\nGLO1\\nSBK1\\nCISD1\\n2210011C24RIK\\nNXT2\\nNOTUM\\nANKRD42\\nUBE2E1\\nNDUFV1\\nSLC33A1\\nCEP68\\nRPS6KB1\\nHYI\\nALDH1A3\\nMYNN\\n3110048L19RIK\\nRDH14\\nPROZ\\nGORASP1\\nLOC674449\\nZFP775\\n5430437P03RIK\\nNPY\\nADH5\\nSYBL1\\n4930432O21RIK\\nNAT9\\nLOC100048387\\nMETTL8\\nENY2\\n2410018G20RIK\\nPGM2\\nFGFR4\\nMOBKL2B\\nATAD3A\\n4932432K03RIK\\nDHTKD1\\nUBOX5\\nA530050D06RIK\\nZDHHC5\\nMGAT1\\nNUDT6\\nTPMT\\nWBSCR18\\nLOC100041586\\nCDK5RAP1\\n4833426J09RIK\\nMYO6\\nCPT1A\\nGADD45GIP1\\nTMBIM4\\n2010309E21RIK\\nASB9\\n2610019F03RIK\\n7530414M10RIK\\nATP6V1B2\\n2310068J16RIK\\nDDT\\nKLHDC4\\nHPN\\nLIFR\\nOVOL1\\nNUDT12\\nCDAN1\\nFBXO9\\nFBXL3\\nHOXA7\\nALDH8A1\\n3110057O12RIK\\nABHD11\\nPSMB1\\nENSMUSG00000074286\\nCHPT1\\nOXSM\\n2310009A05RIK\\n1700001L05RIK\\nZFP148\\n39509\\nMRPL9\\nTMEM80\\n9030420J04RIK\\nNAGLU\\nPLSCR2\\nAGBL3\\nPEX1\\nCNO\\nNEO1\\nASF1A\\nTNFSF5IP1\\nPKIG\\nAI931714\\nD130020L05RIK\\nCNTD1\\nCLEC2H\\nZKSCAN1\\n1810044D09RIK\\nMETTL7A\\nSIAE\\nFBXO3\\nFZD5\\nTMEM166\\nTMED4\\nGPR155\\nRNF167\\nSPTLC1\\nRIOK2\\nTGDS\\nPMS1\\nPITPNC1\\nPCSK7\\n4933403G14RIK\\nEI24\\nCREBL2\\nTLN1\\nMRPL35\\n2700038C09RIK\\nUBIE\\nOSGEPL1\\n2410166I05RIK\\nWDR24\\nAP4S1\\nLRRC44\\nB3BP\\nITFG1\\nDMXL1\\nC1D',\n",
    "                description='Paste your gene list (One gene per row).',\n",
    "                section = 'section1'\n",
    "            ),\n",
    "        ],\n",
    "        'Upload': [\n",
    "            FileField(\n",
    "                name='gene_list_filename',\n",
    "                label='Gene List File',\n",
    "                default='',\n",
    "                description='Upload your gene list as a text file (One gene per row).',\n",
    "                section = 'section1'\n",
    "            ),\n",
    "        ],\n",
    "    },\n",
    "    section = 'section1',\n",
    ") %}\n",
    "\n",
    "{% set enrichr_library = ChoiceField(\n",
    "    name='enrichr_library', \n",
    "    description='Select one Enrichr library for which to create visualizations.', \n",
    "    label='Enrichr Library', \n",
    "    default='WikiPathways_2019_Human', \n",
    "    section = 'section2',\n",
    "    choices=[\n",
    "        'ARCHS4_Cell-lines',\n",
    "        'ARCHS4_IDG_Coexp',\n",
    "        'ARCHS4_Kinases_Coexp',\n",
    "        'ARCHS4_TFs_Coexp',\n",
    "        'ARCHS4_Tissues',\n",
    "        'Achilles_fitness_decrease',\n",
    "        'Achilles_fitness_increase',\n",
    "        'Aging_Perturbations_from_GEO_down',\n",
    "        'Aging_Perturbations_from_GEO_up',\n",
    "        'Allen_Brain_Atlas_10x_scRNA_2021',\n",
    "        'Allen_Brain_Atlas_down',\n",
    "        'Allen_Brain_Atlas_up',\n",
    "        'BioCarta_2013',\n",
    "        'BioCarta_2015',\n",
    "        'BioCarta_2016',\n",
    "        'BioPlanet_2019',\n",
    "        'BioPlex_2017',\n",
    "        'CCLE_Proteomics_2020',\n",
    "        'CORUM',\n",
    "        'COVID-19_Related_Gene_Sets',\n",
    "        'Cancer_Cell_Line_Encyclopedia',\n",
    "        'ChEA_2013',\n",
    "        'ChEA_2015',\n",
    "        'ChEA_2016',\n",
    "        'Chromosome_Location',           \n",
    "        'Chromosome_Location_hg19',\n",
    "        'ClinVar_2019',\n",
    "        'dbGaP',\n",
    "        'DSigDB',\n",
    "        'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "        'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "        'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "        'DisGeNET',\n",
    "        'Disease_Perturbations_from_GEO_down',\n",
    "        'Disease_Perturbations_from_GEO_up',\n",
    "        'Disease_Signatures_from_GEO_down_2014',\n",
    "        'Disease_Signatures_from_GEO_up_2014',\n",
    "        'DrugMatrix',\n",
    "        'Drug_Perturbations_from_GEO_2014',\n",
    "        'Drug_Perturbations_from_GEO_down',\n",
    "        'Drug_Perturbations_from_GEO_up',\n",
    "        'ENCODE_Histone_Modifications_2013',\n",
    "        'ENCODE_Histone_Modifications_2015',\n",
    "        'ENCODE_TF_ChIP-seq_2014',\n",
    "        'ENCODE_TF_ChIP-seq_2015',\n",
    "        'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "        'ESCAPE',\n",
    "        'Elsevier_Pathway_Collection',\n",
    "        'Enrichr_Libraries_Most_Popular_Genes',\n",
    "        'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "        'Enrichr_Users_Contributed_Lists_2020',\n",
    "        'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "        'GO_Biological_Process_2013',\n",
    "        'GO_Biological_Process_2015',\n",
    "        'GO_Biological_Process_2017',\n",
    "        'GO_Biological_Process_2017b',\n",
    "        'GO_Biological_Process_2018',\n",
    "        'GO_Cellular_Component_2013',\n",
    "        'GO_Cellular_Component_2015',\n",
    "        'GO_Cellular_Component_2017',\n",
    "        'GO_Cellular_Component_2017b',\n",
    "        'GO_Cellular_Component_2018',\n",
    "        'GO_Molecular_Function_2013',\n",
    "        'GO_Molecular_Function_2015',\n",
    "        'GO_Molecular_Function_2017',\n",
    "        'GO_Molecular_Function_2017b',\n",
    "        'GO_Molecular_Function_2018',\n",
    "        'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "        'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "        'GWAS_Catalog_2019',\n",
    "        'GeneSigDB',\n",
    "        'Gene_Perturbations_from_GEO_down',\n",
    "        'Gene_Perturbations_from_GEO_up',\n",
    "        'Genes_Associated_with_NIH_Grants',\n",
    "        'Genome_Browser_PWMs',\n",
    "        'HMDB_Metabolites',\n",
    "        'HMS_LINCS_KinomeScan',\n",
    "        'HomoloGene',\n",
    "        'HumanCyc_2015',\n",
    "        'HumanCyc_2016',\n",
    "        'Human_Gene_Atlas',\n",
    "        'Human_Phenotype_Ontology',\n",
    "        'huMAP',\n",
    "        'InterPro_Domains_2019',\n",
    "        'Jensen_COMPARTMENTS',\n",
    "        'Jensen_DISEASES',\n",
    "        'Jensen_TISSUES',\n",
    "        'KEA_2013',\n",
    "        'KEA_2015',\n",
    "        'KEGG_2013',\n",
    "        'KEGG_2015',\n",
    "        'KEGG_2016',\n",
    "        'KEGG_2019_Human',\n",
    "        'KEGG_2019_Mouse',\n",
    "        'Kinase_Perturbations_from_GEO_down',\n",
    "        'Kinase_Perturbations_from_GEO_up',\n",
    "        'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "        'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "        'LINCS_L1000_Chem_Pert_down',\n",
    "        'LINCS_L1000_Chem_Pert_up',\n",
    "        'LINCS_L1000_Ligand_Perturbations_down',\n",
    "        'LINCS_L1000_Ligand_Perturbations_up',\n",
    "        'Ligand_Perturbations_from_GEO_down',\n",
    "        'Ligand_Perturbations_from_GEO_up',\n",
    "        'lncHUB_lncRNA_Co-Expression',\n",
    "        'MCF7_Perturbations_from_GEO_down',\n",
    "        'MCF7_Perturbations_from_GEO_up',\n",
    "        'MGI_Mammalian_Phenotype_2013',\n",
    "        'MGI_Mammalian_Phenotype_2017',\n",
    "        'MGI_Mammalian_Phenotype_Level_3',\n",
    "        'MGI_Mammalian_Phenotype_Level_4',\n",
    "        'MGI_Mammalian_Phenotype_Level_4_2019',\n",
    "        'MSigDB_Computational',\n",
    "        'MSigDB_Hallmark_2020',\n",
    "        'MSigDB_Oncogenic_Signatures',\n",
    "        'Microbe_Perturbations_from_GEO_down',\n",
    "        'Microbe_Perturbations_from_GEO_up',\n",
    "        'miRTarBase_2017',\n",
    "        'Mouse_Gene_Atlas',\n",
    "        'NCI-60_Cancer_Cell_Lines',\n",
    "        'NCI-Nature_2015',\n",
    "        'NCI-Nature_2016',\n",
    "        'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "        'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "        'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "        'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "        'NURSA_Human_Endogenous_Complexome',\n",
    "        'OMIM_Disease',\n",
    "        'OMIM_Expanded',\n",
    "        'Old_CMAP_down',\n",
    "        'Old_CMAP_up',\n",
    "        'PPI_Hub_Proteins',\n",
    "        'Panther_2015',\n",
    "        'Panther_2016',\n",
    "        'Pfam_Domains_2019',\n",
    "        'Pfam_InterPro_Domains',\n",
    "        'PheWeb_2019',\n",
    "        'Phosphatase_Substrates_from_DEPOD',\n",
    "        'ProteomicsDB_2020',\n",
    "        'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "        'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "        'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "        'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "        'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "        'Reactome_2013',\n",
    "        'Reactome_2015',\n",
    "        'Reactome_2016',\n",
    "        'SILAC_Phosphoproteomics',\n",
    "        'SubCell_BarCode',\n",
    "        'SysMyo_Muscle_Gene_Sets'\n",
    "        'TF-LOF_Expression_from_GEO',\n",
    "        'TF_Perturbations_Followed_by_Expression',\n",
    "        'TG_GATES_2020',\n",
    "        'TRANSFAC_and_JASPAR_PWMs',\n",
    "        'TRRUST_Transcription_Factors_2019',\n",
    "        'Table_Mining_of_CRISPR_Studies',\n",
    "        'TargetScan_microRNA',\n",
    "        'TargetScan_microRNA_2017',\n",
    "        'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
    "        'Tissue_Protein_Expression_from_ProteomicsDB.csv',\n",
    "        'Transcription_Factor_PPIs',\n",
    "        'UK_Biobank_GWAS_v1',\n",
    "        'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "        'VirusMINT',\n",
    "        'Virus_Perturbations_from_GEO_down',\n",
    "        'Virus_Perturbations_from_GEO_up',\n",
    "        'WikiPathways_2013',\n",
    "        'WikiPathways_2015',\n",
    "        'WikiPathways_2016',\n",
    "        'WikiPathways_2019_Human',\n",
    "        'WikiPathways_2019_Mouse'\n",
    "    ]\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if gene_list_kind.raw_value == 'Paste' %}\n",
    "gene_list_input = {{ gene_list_kind.value[0] }}\n",
    "{%- else %}\n",
    "gene_list_filename = {{ gene_list_kind.value[0] }}\n",
    "{%- endif %}\n",
    "enrichr_library = '{{ enrichr_library }}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if gene_list_kind.raw_value == 'Paste' %}\n",
    "genes = gene_list_input.split('\\n')\n",
    "genes = [x.strip() for x in genes]\n",
    "{%- else %}\n",
    "open_gene_list_file = open(gene_list_filename,'r')\n",
    "lines = open_gene_list_file.readlines()\n",
    "genes = [x.strip() for x in lines]\n",
    "open_gene_list_file.close()\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Error handling\n",
    "class NoResults(Exception):\n",
    "    pass \n",
    "class APIFailure(Exception):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enrichr API Function for Manhattan Plot and Bar Chart\n",
    "# Takes a gene list and Enrichr libraries as input\n",
    "def Enrichr_API(enrichr_gene_list, all_libraries):\n",
    "\n",
    "    all_terms = []\n",
    "    all_pvalues =[] \n",
    "    all_adjusted_pvalues = []\n",
    "\n",
    "    for library_name in all_libraries : \n",
    "        ENRICHR_URL = 'https://maayanlab.cloud/Enrichr/addList'\n",
    "        genes_str = '\\n'.join(enrichr_gene_list)\n",
    "        description = ''\n",
    "        payload = {\n",
    "            'list': (None, genes_str),\n",
    "            'description': (None, description)\n",
    "        }\n",
    "\n",
    "        response = requests.post(ENRICHR_URL, files=payload)\n",
    "        if not response.ok:\n",
    "            raise APIFailure\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "        time.sleep(0.5)\n",
    "        ENRICHR_URL = 'https://maayanlab.cloud/Enrichr/enrich'\n",
    "        query_string = '?userListId=%s&backgroundType=%s'\n",
    "        user_list_id = data['userListId']\n",
    "        short_id = data[\"shortId\"]\n",
    "        gene_set_library = library_name\n",
    "        response = requests.get(\n",
    "            ENRICHR_URL + query_string % (user_list_id, gene_set_library)\n",
    "         )\n",
    "        if not response.ok:\n",
    "            raise APIFailure\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "\n",
    "        if len(data[library_name]) == 0:\n",
    "            raise NoResults\n",
    "\n",
    "        short_results_df  = pd.DataFrame(data[library_name][0:10])\n",
    "        all_terms.append(list(short_results_df[1]))\n",
    "        all_pvalues.append(list(short_results_df[2]))\n",
    "        all_adjusted_pvalues.append(list(short_results_df[6]))\n",
    "        \n",
    "        results_df  = pd.DataFrame(data[library_name])\n",
    "        # adds library name to the data frame so the libraries can be distinguished\n",
    "        results_df['library'] = library_name.replace('_', '')\n",
    "\n",
    "    return [results_df, short_results_df, all_terms, all_pvalues, all_adjusted_pvalues, str(short_id)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scatter Plot Parameters\n",
    "significance_value = 0.05\n",
    "\n",
    "# Bar Chart Parameters\n",
    "figure_file_format = ['png', 'svg']\n",
    "output_file_name = 'Enrichr_results_bar'\n",
    "color = 'lightskyblue'\n",
    "final_output_file_names = ['{0}.{1}'.format(output_file_name, file_type) for file_type in figure_file_format]\n",
    "\n",
    "# Hexagonal Canvas Parameters\n",
    "canvas_color = 'Blue'\n",
    "num_hex_colored = 10\n",
    "\n",
    "# Manhattan Plot Parameters\n",
    "manhattan_colors = ['#003f5c', '#7a5195', '#ef5675', '#ffa600']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scatter Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The scatterplot is organized so that simliar gene sets are clustered together.\n",
    "The larger blue points represent significantly enriched terms - the darker the blue, the more significant the term and the smaller the p-value. The gray points are not significant. \n",
    "\n",
    "Hovering over points will display the associated gene set name and the p-value. You may have to zoom in using the toolbar next to the plot in order to see details in densely-populated portions. Plots can also be downloaded as an svg using the save function on the toolbar.\n",
    "\n",
    "For creating and comparing up to 9 scatter plots at once, use the standalone [Scatter Plot Visualization Appyter](https://appyters.maayanlab.cloud/#Enrichr_Scatterplot_Appyter)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scatter Plot Functions\n",
    "\n",
    "def download_library(library_name):\n",
    "    # Download pre-processed library data\n",
    "    try:\n",
    "        df = pd.read_csv('https://raw.githubusercontent.com/MaayanLab/Enrichr-Viz-Appyter/master/Enrichr-Processed-Library-Storage/Scatterplot/Libraries/' + library_name + '.csv')\n",
    "    except:\n",
    "        display(HTML(\"<div style='font-size:1.5rem; padding:1.5rem 0;'>Failed to retrieve the selected pre-processed library.</div>\"))\n",
    "        return -1, -1, -1\n",
    "\n",
    "    name = df['Name'].tolist()\n",
    "    gene_list = df['Genes'].tolist()\n",
    "    library_data = [list(a) for a in zip(name, gene_list)]\n",
    "    return genes, library_data, df\n",
    "\n",
    "# Enrichment analysis\n",
    "def get_library_iter(library_data):\n",
    "    for member in library_data:\n",
    "        term = member[0]\n",
    "        try:\n",
    "            gene_set = member[1].split(' ')\n",
    "        except:\n",
    "            continue\n",
    "        yield term, gene_set\n",
    "\n",
    "def get_enrichment_results(genes, library_data):\n",
    "    return sorted(enrich_crisp(genes, get_library_iter(library_data), 20000, True), key=lambda r: r[1].pvalue)\n",
    "\n",
    "def get_pvalue(row, unzipped_results, all_results):\n",
    "    if row['Name'] in list(unzipped_results[0]):\n",
    "        index = list(unzipped_results[0]).index(row['Name'])\n",
    "        return all_results[index][1].pvalue\n",
    "    else:\n",
    "        return 1\n",
    "    \n",
    "# Call enrichment results and return a plot and dataframe for Scatter Plot\n",
    "def get_plot(library_name):\n",
    "    genes, library_data, df = download_library(library_name)\n",
    "\n",
    "    # library not supported\n",
    "    if genes == -1:\n",
    "        return -1 ,-1\n",
    "\n",
    "    all_results = get_enrichment_results(genes, library_data)\n",
    "    unzipped_results = list(zip(*all_results))\n",
    "\n",
    "    if len(all_results) == 0:\n",
    "        display(HTML(f\"<div style='font-size:1.5rem; padding:1rem 0;'>There are no enriched terms for your input gene set in the <b>{library_name}</b> library.</div>\"))\n",
    "        my_colors = ['#808080'] * len(df.index)\n",
    "\n",
    "        source = ColumnDataSource(\n",
    "            data=dict(\n",
    "                x = df['x'],\n",
    "                y = df['y'],\n",
    "                gene_set = df['Name'],\n",
    "                colors = my_colors,\n",
    "                sizes = [6] * len(df.index)\n",
    "            )\n",
    "        )\n",
    "\n",
    "        hover_emb = HoverTool(names=[\"df\"], tooltips=\"\"\"\n",
    "            <div style=\"margin: 10\">\n",
    "                <div style=\"margin: 0 auto; width:200px;\">\n",
    "                    <span style=\"font-size: 12px; font-weight: bold;\">Gene Set:</span>\n",
    "                    <span style=\"font-size: 12px\">@gene_set</span>\n",
    "                </div>\n",
    "            </div>\n",
    "            \"\"\")\n",
    "    else:\n",
    "        # add p value to the dataframe\n",
    "        df['p value'] = df.apply (lambda row: get_pvalue(row, unzipped_results, all_results), axis=1)\n",
    "\n",
    "        # normalize p values for color scaling\n",
    "        cmap = mpl.cm.get_cmap('Blues_r')\n",
    "        norm = colors.Normalize(vmin = df['p value'].min(), vmax=significance_value*2)\n",
    "\n",
    "        my_colors = []\n",
    "        my_sizes = []\n",
    "        for index, row in df.iterrows():\n",
    "            if row['p value'] < significance_value:\n",
    "                my_colors += [mpl.colors.to_hex(cmap(norm(row['p value'])))]\n",
    "                my_sizes += [12]\n",
    "            else:\n",
    "                my_colors += ['#808080']\n",
    "                my_sizes += [6]\n",
    "\n",
    "        source = ColumnDataSource(\n",
    "                data=dict(\n",
    "                    x = df['x'],\n",
    "                    y = df['y'],\n",
    "                    gene_set = df['Name'],\n",
    "                    p_value = df['p value'],\n",
    "                    colors = my_colors,\n",
    "                    sizes = my_sizes\n",
    "                )\n",
    "            )\n",
    "\n",
    "        hover_emb = HoverTool(names=[\"df\"], tooltips=\"\"\"\n",
    "            <div style=\"margin: 10\">\n",
    "                <div style=\"margin: 0 auto; width:200px;\">\n",
    "                    <span style=\"font-size: 12px; font-weight: bold;\">Gene Set:</span>\n",
    "                    <span style=\"font-size: 12px\">@gene_set</span>\n",
    "                    <span style=\"font-size: 12px; font-weight: bold;\">p-value:</span>\n",
    "                    <span style=\"font-size: 12px\">@p_value</span>\n",
    "                </div>\n",
    "            </div>\n",
    "            \"\"\")\n",
    "\n",
    "    tools_emb = [hover_emb, 'pan', 'wheel_zoom', 'reset', 'save']\n",
    "\n",
    "    plot_emb = figure(plot_width=700, plot_height=700, tools=tools_emb)\n",
    "\n",
    "    # hide axis labels and grid lines\n",
    "    plot_emb.xaxis.major_tick_line_color = None\n",
    "    plot_emb.xaxis.minor_tick_line_color = None\n",
    "    plot_emb.yaxis.major_tick_line_color = None\n",
    "    plot_emb.yaxis.minor_tick_line_color = None\n",
    "    plot_emb.xaxis.major_label_text_font_size = '0pt'\n",
    "    plot_emb.yaxis.major_label_text_font_size = '0pt' \n",
    "\n",
    "    plot_emb.circle('x', 'y', size = 'sizes', alpha = 0.7, line_alpha = 0, \n",
    "                    line_width = 0.01, source = source, fill_color = 'colors', name = \"df\")\n",
    "\n",
    "    plot_emb.output_backend = \"svg\"\n",
    "    \n",
    "    return plot_emb, df    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display Scatter Plot\n",
    "plot, df = get_plot(enrichr_library)\n",
    "if plot == -1:\n",
    "    display(HTML(\"<div style='font-size:1rem;'>Unable to create scatter plot visualization.</div>\"))\n",
    "else:\n",
    "    show(plot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bar Chart"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The bar chart shows the top 10 enriched terms in the chosen library, along with their corresponding p-values. Colored bars correspond to terms with significant p-values (<0.05). An asterisk (*) next to a p-value indicates the term also has a significant adjusted p-value (<0.05).\n",
    "\n",
    "The bar chart can be downloaded as an image using the links below the figure. \n",
    "\n",
    "For creating customized bar charts for multiple libraries at once, use the standalone [Bar Chart Appyter](https://appyters.maayanlab.cloud/Enrichr_compressed_bar_chart_figure/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bar Chart Functions\n",
    "# Takes all terms, all p-values, all adjusted p-values, plot title, Enrichr libraries, and specified figure format\n",
    "def enrichr_figure(all_terms, all_pvalues, all_adjusted_pvalues, plot_names, all_libraries, bar_color): \n",
    "    # Bar colors\n",
    "    if bar_color != 'lightgrey':\n",
    "        bar_color_not_sig = 'lightgrey'\n",
    "        edgecolor=None\n",
    "        linewidth=0\n",
    "    else:\n",
    "        bar_color_not_sig = 'white'\n",
    "        edgecolor='black'\n",
    "        linewidth=1    \n",
    "\n",
    "    plt.figure(figsize=(24, 12))\n",
    "    \n",
    "    i = 0\n",
    "    bar_colors = [bar_color if (x < 0.05) else bar_color_not_sig for x in all_pvalues[i]]\n",
    "    fig = sns.barplot(x=np.log10(all_pvalues[i])*-1, y=all_terms[i], palette=bar_colors, edgecolor=edgecolor, linewidth=linewidth)\n",
    "    fig.axes.get_yaxis().set_visible(False)\n",
    "    fig.set_title(all_libraries[i].replace('_', ' '), fontsize=26)\n",
    "    fig.set_xlabel('−log₁₀(p‐value)', fontsize=25)\n",
    "    fig.xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "    fig.tick_params(axis='x', which='major', labelsize=20)\n",
    "    if max(np.log10(all_pvalues[i])*-1)<1:\n",
    "        fig.xaxis.set_ticks(np.arange(0, max(np.log10(all_pvalues[i])*-1), 0.1))\n",
    "    for ii,annot in enumerate(all_terms[i]):\n",
    "        if all_adjusted_pvalues[i][ii] < 0.05:\n",
    "            annot = '  *'.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii], precision=2)))]) \n",
    "        else:\n",
    "            annot = '  '.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii], precision=2)))])\n",
    "\n",
    "        title_start= max(fig.axes.get_xlim())/200\n",
    "        fig.text(title_start, ii, annot, ha='left', wrap = True, fontsize = 26)\n",
    "\n",
    "    fig.spines['right'].set_visible(False)\n",
    "    fig.spines['top'].set_visible(False)\n",
    "\n",
    "    for plot_name in plot_names:\n",
    "        plt.savefig(plot_name, bbox_inches = 'tight')\n",
    "    \n",
    "    # Show plot \n",
    "    plt.show()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display Bar Chart\n",
    "try:\n",
    "    results = Enrichr_API(genes, [enrichr_library])\n",
    "    enrichr_figure(results[2], results[3], results[4], final_output_file_names, [enrichr_library], color)\n",
    "    # Download Bar Chart\n",
    "    for i, file in enumerate(final_output_file_names):\n",
    "        display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))\n",
    "except APIFailure:\n",
    "    results = []\n",
    "    display(HTML(\"<div style='font-size:1.5rem; padding:1rem 0;'><b>Unable to display plot because of a bad response from the Enrichr API</b></div>\"))\n",
    "    display(HTML(\"<div style='font-size:1rem; padding:1rem 0;'>Please try again later.\"))\n",
    "except NoResults:\n",
    "    results = []\n",
    "    display(HTML(\"<div style='font-size:1.5rem; padding:1rem 0;'><b>Unable to display plot because there were no results returned from Enrichr</b></div>\"))\n",
    "    display(HTML(\"<div style='font-size:1rem; padding:1rem 0;'>Please try again with a different Enrichr library, or a different query gene list.\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hexagonal Canvas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each hexagon in the hexagonal canvas plot represents one gene set from the selected library. The hexagons are colored based on the Jaccard similarity index between the input gene list and the gene set represented by the hexagon, with brighter color indicating higher similarity. Hexagons that are grouped together represent similar gene sets. \n",
    "\n",
    "Hovering over a hexagon will display the name of the gene set and the associated similarity index. \n",
    "\n",
    "For creating customized hexagonal canvas plots for up to two libraries at once, use the standalone [Hexagonal Canvas Appyter](https://appyters.maayanlab.cloud/#Enrichr_Canvas_Appyter)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hexagonal Canvas Functions\n",
    "\n",
    "def library_processing():\n",
    "    # Downloads library data for the hexagonal canvas\n",
    "    # Library data is pre-annealed so the canvas will have the most similar gene sets closest together\n",
    "    raw_library_data = []\n",
    "\n",
    "    try:\n",
    "        library_name = enrichr_library\n",
    "        with urllib.request.urlopen('https://raw.githubusercontent.com/MaayanLab/Enrichr-Viz-Appyter/master/Enrichr-Processed-Library-Storage/Annealing/Annealed-Libraries/' + library_name + '.txt') as f:\n",
    "            for line in f.readlines():\n",
    "                raw_library_data.append(line.decode('utf-8').split(\"\\t\\t\"))\n",
    "        name = []\n",
    "        gene_list = []\n",
    "    except:\n",
    "        display(HTML(\"<div style='font-size:1.5rem; padding:1.5rem 0;'>Failed to retrieve the selected annealed library.</div>\"))\n",
    "        return [], -1, -1\n",
    "\n",
    "    for i in range(len(raw_library_data)):\n",
    "        name += [raw_library_data[i][0]]\n",
    "        raw_genes = raw_library_data[i][1].split('\\t')\n",
    "        gene_list += [raw_genes[:-1]]\n",
    "\n",
    "    library_data = [list(a) for a in zip(name, gene_list)]\n",
    "\n",
    "    # raw_library_data: a 2D list where the first element is the name and the second element is a list of genes associated with that name\n",
    "\n",
    "    jaccard_indices = []\n",
    "    indices = []\n",
    "\n",
    "    for gene_set in library_data:\n",
    "        intersection = [value for value in gene_set[1] if value in genes]\n",
    "        index = len(intersection)/(len(gene_set[1]) + len(genes))\n",
    "        jaccard_indices += [[gene_set[0], index]]\n",
    "        indices += [round(index, 5)]\n",
    "\n",
    "    # determine the dimensions of the canvas\n",
    "    x_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "    y_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "\n",
    "    # zip name, gene_list, indices, and blank list for neighbor score then add dummy entries to the zipped list\n",
    "    anneal_list = list(zip(name, gene_list, indices))\n",
    "\n",
    "    return anneal_list, x_dimension, y_dimension\n",
    "\n",
    "def unzip_list(anneal_list):\n",
    "    unzipped_list = zip(*anneal_list)\n",
    "    return list(unzipped_list)\n",
    "\n",
    "# define a list of colors for the hexagonal canvas\n",
    "def get_color(anneal_list, cut_off_value, x_dimension, y_dimension):\n",
    "\n",
    "    # Deal with cut_off_value (only color the most significant 10/20 hexagons)\n",
    "    if cut_off_value == 2.0:\n",
    "        sort_list = sorted(anneal_list, key=itemgetter(2), reverse=True)\n",
    "        cut_off_value = sort_list[int(num_hex_colored)-1][2]\n",
    "\n",
    "    r_value = 0\n",
    "    g_value = 0\n",
    "    b_value = 0\n",
    "\n",
    "    if canvas_color == 'Red':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.8\n",
    "        b_value = 0.8\n",
    "    if canvas_color == 'Yellow':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.3\n",
    "        b_value = 1.0\n",
    "    if canvas_color == 'Purple':\n",
    "        r_value = 0.5\n",
    "        g_value = 1.0\n",
    "        b_value = 0.0\n",
    "    if canvas_color == 'Pink':\n",
    "        r_value = 0.0\n",
    "        g_value = 1.0\n",
    "        b_value = 0.2\n",
    "    if canvas_color == 'Orange':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.45\n",
    "        b_value = 1.0\n",
    "    if canvas_color == 'Green':\n",
    "        r_value = 1.0\n",
    "        g_value = 0.0\n",
    "        b_value = 1.0\n",
    "    if canvas_color == 'Blue':\n",
    "        r_value = 1.0\n",
    "        g_value = 0.9\n",
    "        b_value = 0.0\n",
    "\n",
    "    color_list = []\n",
    "\n",
    "    unzipped_anneal_list = unzip_list(anneal_list)\n",
    "\n",
    "    max_index = max(unzipped_anneal_list[2])\n",
    "\n",
    "    if max_index != 0:\n",
    "        scaled_list = [i/max_index for i in unzipped_anneal_list[2]]\n",
    "    else:\n",
    "        scaled_list = unzipped_anneal_list[2]\n",
    "\n",
    "    for i in range(x_dimension*y_dimension):\n",
    "        if i < len(unzipped_anneal_list[2]) and float(unzipped_anneal_list[2][i]) >= cut_off_value:\n",
    "            color_list += [mpl.colors.to_hex((1-scaled_list[i]*r_value, \n",
    "            1-scaled_list[i]*g_value, 1-scaled_list[i]*b_value))]\n",
    "        elif i < len(unzipped_anneal_list[2]):\n",
    "            color_list += [mpl.colors.to_hex((1-scaled_list[i], \n",
    "            1-scaled_list[i], 1-scaled_list[i]))]\n",
    "        else:\n",
    "            color_list += [\"#FFFFFF\"]\n",
    "    return color_list, max_index, cut_off_value\n",
    "\n",
    "def init_chart():\n",
    "  chart_id = 'mychart-' + str(uuid.uuid4())\n",
    "  display(HTML('<script src=\"/static/components/requirejs/require.js\"></script>'))\n",
    "  display(HTML(Template(dedent('''\n",
    "  <script>\n",
    "  require.config({\n",
    "    paths: {\n",
    "      'd3': 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min',\n",
    "      'd3-hexbin': 'https://d3js.org/d3-hexbin.v0.2.min',\n",
    "    },\n",
    "    shim: {\n",
    "      'd3-hexbin': ['d3']\n",
    "    }\n",
    "  })\n",
    "\n",
    "  // If we configure mychart via url, we can eliminate this define here\n",
    "  define($chart_id, ['d3', 'd3-hexbin'], function(d3, d3_hexbin) {\n",
    "    return function (figure_id, numA, numB, colorList, libraryList, indices) {\n",
    "      var margin = {top: 50, right: 20, bottom: 20, left: 50},\n",
    "        width = 850 - margin.left - margin.right,\n",
    "        height = 350 - margin.top - margin.bottom;\n",
    "\n",
    "      // append the svg object to the body of the page\n",
    "      var svG = d3.select('#' + figure_id)\n",
    "                  .attr(\"width\", width + margin.left + margin.right)\n",
    "                  .attr(\"height\", height + margin.top + margin.bottom)\n",
    "                  .append(\"g\")\n",
    "                  .attr(\"transform\", \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
    "      \n",
    "      //The number of columns and rows of the heatmap\n",
    "      var MapColumns = numA,\n",
    "          MapRows = numB;\n",
    "\n",
    "      //The maximum radius the hexagons can have to still fit the screen\n",
    "      var hexRadius = d3.min([width/((MapColumns + 0.5) * Math.sqrt(3)), height/((MapRows + 1/3) * 1.5)]);\n",
    "\n",
    "      //Calculate the center position of each hexagon\n",
    "      var points = [];\n",
    "      for (var i = 0; i < MapRows; i++) {\n",
    "          for (var j = 0; j < MapColumns; j++) {\n",
    "              var x = hexRadius * j * Math.sqrt(3)\n",
    "              //Offset each uneven row by half of a \"hex-width\" to the right\n",
    "              if(i%2 === 1) x += (hexRadius * Math.sqrt(3))/2\n",
    "              var y = hexRadius * i * 1.5\n",
    "              points.push([x,y])\n",
    "          }\n",
    "      }\n",
    "\n",
    "      //Set the hexagon radius\n",
    "      var hexbin = d3_hexbin.hexbin().radius(hexRadius);\n",
    "\n",
    "      svG.append(\"g\")\n",
    "        .selectAll(\".hexagon\")\n",
    "        .data(hexbin(points))\n",
    "        .enter().append(\"path\")\n",
    "        .attr(\"class\", \"hexagon\")\n",
    "        .attr(\"d\", function (d) {\n",
    "            return \"M\" + d.x + \",\" + d.y + hexbin.hexagon();\n",
    "        })\n",
    "        .attr(\"stroke\", \"black\")\n",
    "        .attr(\"stroke-width\", \"1px\")\n",
    "        .style(\"fill\", function (d,i) { return colorList[i]; })\n",
    "        .on(\"mouseover\", mover)\n",
    "        .on(\"mouseout\", mout)\n",
    "        .append(\"svg:title\")\n",
    "        .text(function(d,i) { return libraryList[i].concat(\" \").concat(indices[i]); });\n",
    "\n",
    "      // Mouseover function\n",
    "      function mover(d) {\n",
    "      d3.select(this)\n",
    "        .transition().duration(10)  \n",
    "        .style(\"fill-opacity\", 0.3)\n",
    "      };\n",
    "\n",
    "      // Mouseout function\n",
    "      function mout(d) { \n",
    "      d3.select(this)\n",
    "        .transition().duration(10)\n",
    "        .style(\"fill-opacity\", 1)\n",
    "      };\n",
    "\n",
    "  }\n",
    "\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({ 'chart_id': repr(chart_id) })))\n",
    "  return chart_id\n",
    "\n",
    "def Canvas(numA, numB, colorList, libraryList, indices):\n",
    "  chart_id = init_chart()\n",
    "  display(HTML(Template(dedent('''\n",
    "  <svg id=$figure_id></svg>\n",
    "  <script>\n",
    "  require([$chart_id], function(mychart) {\n",
    "    mychart($figure_id, $numA, $numB, $colorList, $libraryList, $indices)\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({\n",
    "      'chart_id': repr(chart_id),\n",
    "      'figure_id': repr('fig-' + str(uuid.uuid4())),\n",
    "      'numA': repr(numA),\n",
    "      'numB': repr(numB),\n",
    "      'colorList': repr(colorList),\n",
    "      'libraryList': repr(libraryList),\n",
    "      'indices': repr(indices)\n",
    "  })))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display Hexagonal Canvas\n",
    "anneal_list, x_dimension, y_dimension = library_processing()\n",
    "if x_dimension < 0:\n",
    "    display(HTML(\"<div style='font-size:1rem;'>Unable to create hexagonal canvas visualization.</div>\"))\n",
    "else:\n",
    "    color_list, scaling_factor, cut_off_value = get_color(anneal_list, 2.0, x_dimension, y_dimension)\n",
    "    unzipped_anneal_list = unzip_list(anneal_list)\n",
    "    if len(set(unzipped_anneal_list[2])) > 1:\n",
    "        Canvas(x_dimension, y_dimension, color_list, list(unzipped_anneal_list[0]), list(unzipped_anneal_list[2]))\n",
    "    else:\n",
    "        display(HTML(\"<div style='font-size:1.5rem; padding:1rem 0;'><b>Unable to display plot because there were no results returned from Enrichr</b></div>\"))\n",
    "        display(HTML(\"<div style='font-size:1rem; padding:1rem 0;'>Please try again with a different Enrichr library, or a different query gene list.\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Manhattan Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the Manhattan plot below, each line on the x-axis denotes a single gene set from the selected library, while the y-axis measures the −log₁₀(p‐value) for each gene set. \n",
    "\n",
    "Hovering over a point will display the name of the gene set and the associated p-value. You can also zoom, pan, and save the plot as an svg using the toolbar on the right.\n",
    "\n",
    "For creating customized static and dynamic Manhattan plots to compare multiple libraries at once, use the standalone [Manhattan Plot Appyter](https://appyters.maayanlab.cloud/#Enrichr_Manhattan_Plot)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manhattan Plot Functions\n",
    "\n",
    "# Processes Enrichr data for Manhattan plots\n",
    "def get_data(genes):\n",
    "    # Process Enrichr data\n",
    "    sorted_data = pd.DataFrame({\"Gene Set\": [], \"-log(p value)\": [], \"Library\": []})\n",
    "\n",
    "    # get enrichr results from the library selected\n",
    "    results_df = Enrichr_API(genes, [enrichr_library])[0]\n",
    "\n",
    "    all_terms = []\n",
    "    all_pvalues = []\n",
    "    library_names = []\n",
    "\n",
    "    all_terms.append(list(results_df[1]))\n",
    "    all_pvalues.append(list(results_df[2]))\n",
    "    library_names.append(list(results_df['library']))\n",
    "\n",
    "    x=np.log10(all_pvalues[0])*-1\n",
    "    sorted_terms = list(zip(all_terms[0], x, library_names[0]))\n",
    "    sorted_terms = sorted(sorted_terms, key = itemgetter(0))\n",
    "    unzipped_sorted_list = list(zip(*sorted_terms))\n",
    "\n",
    "    data = pd.DataFrame({\"Gene Set\": unzipped_sorted_list[0], \"-log(p value)\": unzipped_sorted_list[1], \"Library\": unzipped_sorted_list[2]})\n",
    "\n",
    "    sorted_data = pd.concat([sorted_data, data])\n",
    "\n",
    "    # group data by library\n",
    "    groups = sorted_data.groupby(\"Library\")\n",
    "    return sorted_data, groups\n",
    "\n",
    "# Create Manhattan Plots\n",
    "def manhattan(sorted_data):\n",
    "    # split data frame into smaller data frames by library\n",
    "    list_of_df = []\n",
    "    for library_name in [enrichr_library]:\n",
    "        library_name = library_name.replace('_', '')\n",
    "        df_new = sorted_data[sorted_data['Library'] == library_name]\n",
    "        list_of_df += [df_new]\n",
    "\n",
    "    list_of_xaxis_values = []\n",
    "    for df in list_of_df:  \n",
    "        list_of_xaxis_values += df[\"Gene Set\"].values.tolist()\n",
    "\n",
    "    # define the output figure and the features we want\n",
    "    p = figure(x_range = list_of_xaxis_values, plot_height=300, plot_width=750, tools='pan, box_zoom, hover, reset, save')\n",
    "\n",
    "    # loop over all libraries\n",
    "    r = []\n",
    "    color_index = 0\n",
    "    for df in list_of_df:\n",
    "        if color_index >= len(manhattan_colors):\n",
    "            color_index = 0 \n",
    "\n",
    "        # calculate actual p value from -log(p value)\n",
    "        actual_pvalues = []\n",
    "        for log_value in df[\"-log(p value)\"].values.tolist():\n",
    "            actual_pvalues += [\"{:.5e}\".format(10**(-1*log_value))]\n",
    "\n",
    "        # define ColumnDataSource with our data for this library\n",
    "        source = ColumnDataSource(data=dict(\n",
    "            x = df[\"Gene Set\"].values.tolist(),\n",
    "            y = df[\"-log(p value)\"].values.tolist(),\n",
    "            pvalue = actual_pvalues,\n",
    "        ))\n",
    "    \n",
    "        # plot data from this library\n",
    "        r += [p.circle(x = 'x', y = 'y', size=5, fill_color=manhattan_colors[color_index], line_color = manhattan_colors[color_index], line_width=1, source = source)]\n",
    "        color_index += 1\n",
    "\n",
    "    p.background_fill_color = 'white'\n",
    "    p.xaxis.major_tick_line_color = None \n",
    "    p.xaxis.major_label_text_font_size = '0pt'\n",
    "    p.y_range.start = 0\n",
    "    p.yaxis.axis_label = '-log(p value)'\n",
    "\n",
    "    p.hover.tooltips = [\n",
    "        (\"Gene Set\", \"@x\"),\n",
    "        (\"p value\", \"@pvalue\"),\n",
    "    ]\n",
    "    p.output_backend = \"svg\"\n",
    "    \n",
    "    # returns the plot\n",
    "    return p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display Manhattan Plot\n",
    "try:\n",
    "    sorted_data, groups = get_data(genes)\n",
    "    show(manhattan(sorted_data))\n",
    "except APIFailure:\n",
    "    display(HTML(\"<div style='font-size:1.5rem; padding:1rem 0;'><b>Unable to display plot because of a bad response from the Enrichr API</b></div>\"))\n",
    "    display(HTML(\"<div style='font-size:1rem; padding:1rem 0;'>Please try again later.\"))\n",
    "except NoResults:\n",
    "    display(HTML(\"<div style='font-size:1.5rem; padding:1rem 0;'><b>Unable to display plot because there were no results returned from Enrichr</b></div>\"))\n",
    "    display(HTML(\"<div style='font-size:1rem; padding:1rem 0;'>Please try again with a different Enrichr library, or a different query gene list.\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Volcano Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The volcano plot shows the significance of each gene set from the selected library versus its odds ratio. Each point represents a single geneset; the x-axis measures the odds ratio (0, inf) calculated for the gene set, while the y-axis gives the -log(p-value) of the gene set. \n",
    "\n",
    "Larger blue points represent significant terms (p-value < 0.05); smaller gray points represent non-significant terms. The darker the blue color of a point, the more significant it is. \n",
    "\n",
    "Hovering over points will display the corresponding gene set term, the p-value, and the odds ratio. You may have to zoom in using the toolbar next to the plot in order to see details in densely-populated portions. Plots can also be downloaded as an svg using the save function on the toolbar."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_library(lib_name):\n",
    "    '''\n",
    "    Returns a dictionary mapping each term from the input library to \n",
    "    its associated geneset. \n",
    "    '''\n",
    "    raw_lib_data = []\n",
    "\n",
    "    with urllib.request.urlopen('https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName=' + lib_name) as f:\n",
    "        for line in f.readlines():\n",
    "            raw_lib_data.append(line.decode(\"utf-8\").split(\"\\t\\t\"))\n",
    "\n",
    "    name = []\n",
    "    gene_list = []\n",
    "    lib_data = {}\n",
    "\n",
    "    for i in range(len(raw_lib_data)):\n",
    "        name += [raw_lib_data[i][0]]\n",
    "        raw_genes = raw_lib_data[i][1].replace('\\t', ' ')\n",
    "        gene_list += [raw_genes[:-1]]\n",
    "    \n",
    "    lib_data = {a[0]:a[1].split(' ') for a in zip(name, gene_list)}\n",
    "    return lib_data\n",
    "\n",
    "\n",
    "def volcano_plot(library_name, lib):\n",
    "    '''\n",
    "    Make volcano plot of odds ratio vs. significance for input library.\n",
    "    '''\n",
    "    enrich_results = enrich_crisp(genes, lib, 21000, True)\n",
    "\n",
    "    res_df = pd.DataFrame(\n",
    "        [ [\n",
    "            term, \n",
    "            res.pvalue, \n",
    "            res.odds_ratio\n",
    "        ] for (term, res) in enrich_results ], \n",
    "        columns=['term', 'pvalue', 'odds_ratio']\n",
    "    )\n",
    "\n",
    "    res_df['log_pval'] = np.negative(np.log10(res_df['pvalue']))\n",
    "\n",
    "    cmap = mpl.cm.get_cmap('Blues_r')\n",
    "    cnorm = colors.Normalize(vmin = res_df['pvalue'].min(), vmax = 0.1)\n",
    "\n",
    "    my_colors = []\n",
    "    my_sizes = []\n",
    "    for row in res_df.itertuples():\n",
    "        if row.pvalue < 0.05:\n",
    "            my_colors += [mpl.colors.to_hex(cmap(cnorm(row.pvalue)))]\n",
    "            my_sizes += [12]\n",
    "        else:\n",
    "            my_colors += ['#808080']\n",
    "            my_sizes += [6]\n",
    "\n",
    "    source = ColumnDataSource(\n",
    "        data=dict(\n",
    "            x = res_df['odds_ratio'],\n",
    "            y = res_df['log_pval'],\n",
    "            gene_set = res_df['term'],\n",
    "            p_value = res_df['pvalue'],\n",
    "            odds_r = res_df['odds_ratio'],\n",
    "            colors = my_colors,\n",
    "            sizes = my_sizes\n",
    "        )\n",
    "    )\n",
    "\n",
    "    hover_emb = HoverTool(\n",
    "        names=[\"res_df\"], \n",
    "        tooltips=\"\"\"\n",
    "        <div style=\"margin: 10\">\n",
    "            <div style=\"margin: 0 auto; width:200px;\">\n",
    "                <span style=\"font-size: 12px; font-weight: bold;\">Term:</span>\n",
    "                <span style=\"font-size: 12px\">@gene_set<br></span>\n",
    "                <span style=\"font-size: 12px; font-weight: bold;\">P-Value:</span>\n",
    "                <span style=\"font-size: 12px\">@p_value<br></span>\n",
    "                <span style=\"font-size: 12px; font-weight: bold;\">Odds Ratio:</span>\n",
    "                <span style=\"font-size: 12px\">@odds_r<br></span>\n",
    "            </div>\n",
    "        </div>\n",
    "        \"\"\"\n",
    "    )\n",
    "\n",
    "    tools_emb = [hover_emb, 'pan', 'wheel_zoom', 'reset', 'save']\n",
    "\n",
    "    plot_emb = figure(\n",
    "        plot_width = 700, \n",
    "        plot_height = 700,\n",
    "        tools=tools_emb\n",
    "    )\n",
    "\n",
    "    plot_emb.circle(\n",
    "        'x', 'y', size = 'sizes', \n",
    "        alpha = 0.7, line_alpha = 0, \n",
    "        line_width = 0.01, source = source, \n",
    "        fill_color = 'colors', name = \"res_df\"\n",
    "    )\n",
    "\n",
    "    plot_emb.xaxis.axis_label = \"Odds Ratio\"\n",
    "    plot_emb.yaxis.axis_label = \"-log10(p-value)\"\n",
    "\n",
    "    plot_emb.output_backend = \"svg\"\n",
    "    \n",
    "    return plot_emb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lib_data = get_library(enrichr_library)\n",
    "if lib_data == {}:\n",
    "    display(HTML(\"<div style='font-size:1.5rem; padding:1.5rem 0;'>Failed to access Enrichr library API, please try again later.</div>\"))\n",
    "else:\n",
    "    try:\n",
    "        plot = volcano_plot(enrichr_library, lib_data)\n",
    "        show(plot)\n",
    "    except NoResults:\n",
    "        display(HTML(\"<div style='font-size:1.5rem; padding:1rem 0;'><b>Unable to display plot because there were no results returned from Enrichr</b></div>\"))\n",
    "        display(HTML(\"<div style='font-size:1rem; padding:1rem 0;'>Please try again with a different Enrichr library, or a different query gene list.\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of significant p-values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A downloadable table displaying the names, p-values, and q-values of significant terms in the selected library. The `q-value` is an adjusted p-value calculated using the Benjamini-Hochberg method for correction for multiple hypotheses testing. Only the top 10 significant results are displayed in this notebook, but the full table containing all significant results can be downloaded using the *Download CSV file of the full table of significant results* link."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output a table of significant p-values and q-values\n",
    "\n",
    "def create_download_link(df, title = \"Download CSV file of the full table of significant results\", filename = \"data.csv\"):  \n",
    "    csv = df.to_csv(index = False)\n",
    "    b64 = base64.b64encode(csv.encode())\n",
    "    payload = b64.decode()\n",
    "    html = '<a download=\"{filename}\" href=\"data:text/csv;base64,{payload}\" target=\"_blank\">{title}</a>'\n",
    "    html = html.format(payload=payload, title=title, filename=filename)\n",
    "    return HTML(html)\n",
    "\n",
    "if len(results) > 0:\n",
    "    res_df = pd.DataFrame(results[0]).rename(columns={\n",
    "        0: 'rank',\n",
    "        1: 'term',\n",
    "        2: 'p-value',\n",
    "        3: 'zscore',\n",
    "        4: 'combined_score',\n",
    "        5: 'overlap_genes',\n",
    "        6: 'q-value'\n",
    "    })\n",
    "    sorted_res_df = res_df.sort_values(by='p-value', ascending=True)\n",
    "    filtered_res_df = sorted_res_df[sorted_res_df['p-value'] <= significance_value].reset_index()\n",
    "    display(HTML(f\"<strong>Table of top 10 significant p-values and q-values for {enrichr_library.replace('_', ' ')}</strong>\"))\n",
    "    display(HTML(filtered_res_df[['term', 'p-value', 'q-value']][:10].to_html(index = False)))\n",
    "    display(create_download_link(filtered_res_df[['term', 'p-value', 'q-value']]))\n",
    "else:\n",
    "    display(HTML(\"<div style='font-size:1.5rem; padding:1rem 0;'><b>Unable to display table</b></div>\"))\n",
    "    display(HTML(\"<div style='font-size:1rem; padding:1rem 0;'>Please try again later, or start over with a new input list and/or library.\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Link to Enrichr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get complete enrichment analysis results from Enrichr \n",
    "if len(results) > 0:\n",
    "    url = 'https://maayanlab.cloud/Enrichr/enrich?dataset=' + results[5]\n",
    "    display(HTML(f\"<div style='font-size:1.25rem; padding:1rem 0;'><a href='https://maayanlab.cloud/Enrichr/enrich?dataset={results[5]}'>Access the complete enrichment analysis results on the Enrichr website. </a></div>\"))\n",
    "else:\n",
    "    display(HTML(\"<div style='font-size:1.5rem; padding:1rem 0;'><b>No Enrichr results available for the current query</b></div>\"))\n",
    "    display(HTML(\"<div style='font-size:1rem; padding:1rem 0;'>Please try again with a different input list and/or library.\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (base)",
   "language": "python",
   "name": "base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8-final"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
