{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichment Analysis Visualizer\n",
    "\n",
    "This appyter creates a variety of visualizations for enrichment analysis data for one selected Enrichr library. The scatter plot and hexagonal canvas may not be available for every library.\n",
    "\n",
    "Here are the visualizations created: \n",
    " \n",
    "* **Scatter Plot** \n",
    "\n",
    "The scatterplot is organized so that simliar gene sets are clustered together. Enriched gene sets will appear blue instead of gray; the darker the blue the smaller the p-value. The name of the gene set and the associated p-value are displayed when a point is hovered over.\n",
    "\n",
    "Plots can be downloaded as an svg using the save function on the toolbar next to the plot.\n",
    "\n",
    "* **Bar Chart**\n",
    "\n",
    "The bar chart contains the top 10 enriched terms and their corresponding p-values for the chosen library. Colored bars correspond to terms with significant p-values (<0.05). An asterisk (*) next to a p-value indicates the term also has a significant adjusted p-value (<0.05).\n",
    "\n",
    "* **Hexagonal Canvas**\n",
    "\n",
    "The hexagonal canvas is made up of hexagons, each of which represents one gene set from the library selected. The hexagons are colored based on the Jaccard similarity index between the inputted gene list and the gene set the hexagon represents (the brighter, the more similar). The hexagons representing the most similar gene sets are grouped together. The name of the gene set and the associated similarity index are displayed when a hexagon is hovered over.\n",
    "\n",
    "* **Manhattan Plot**\n",
    "\n",
    "A manhattan plot is created displaying all the gene sets from the select library and their p-value. The x-axis of the plot is made up of gene sets from the library. The y-axis of the plot has the -log(p value) for each gene set. The name of the gene set and the associated p-value are displayed when a point is hovered over. You can also zoom, pan, and save the plot as an svg using the toolbar on the right.\n",
    "\n",
    "Additionally, the names and p-values of significant terms in the library are provided in a table output below the plot which can be downloaded. \n",
    "\n",
    "There is a link at the bottom to the full analysis results on the Enrichr website."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scatter Plot Imports\n",
    "from maayanlab_bioinformatics.enrichment import enrich_crisp\n",
    "import matplotlib as mpl\n",
    "import matplotlib.colors as colors\n",
    "import base64\n",
    "\n",
    "# Bar Chart Imports\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import json\n",
    "import requests\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import time\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from IPython.display import display, FileLink, Markdown, HTML\n",
    "\n",
    "# Hexagonal Canvas Imports\n",
    "import json\n",
    "import math\n",
    "import uuid\n",
    "import urllib\n",
    "from textwrap import dedent\n",
    "from string import Template\n",
    "from operator import itemgetter\n",
    "\n",
    "# Manhattan Plot Imports\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "\n",
    "# Bokeh\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool, CustomJS, ColumnDataSource, Span\n",
    "from bokeh.layouts import layout, row, column, gridplot\n",
    "from bokeh.palettes import all_palettes\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(name='section1', title = '1. Submit Your Gene List', subtitle = 'Upload a text file containing your gene list -OR- copy and paste your gene list into the text box below (One gene per row). You can also try the default gene list provided.', img = 'enrichr-icon.png')%}\n",
    "{% do SectionField(name='section2', title = '2. Choose Enrichr Library', subtitle = 'Select one Enrichr library.', img = 'enrichr-icon.png')%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "gene_list_filename = {{ FileField(name='gene_list_filename', label='Gene List File', default='', description='Upload your gene list as a text file (One gene per row).',section = 'section1') }}\n",
    "\n",
    "gene_list_input = {{ TextField(name='gene_list_input', label='Gene List', default='NSUN3\\nPOLRMT\\nNLRX1\\nSFXN5\\nZC3H12C\\nSLC25A39\\nARSG\\nDEFB29\\nNDUFB6\\nZFAND1\\nTMEM77\\n5730403B10RIK\\nRP23-195K8.6\\nTLCD1\\nPSMC6\\nSLC30A6\\nLOC100047292\\nLRRC40\\nORC5L\\nMPP7\\nUNC119B\\nPRKACA\\nTCN2\\nPSMC3IP\\nPCMTD2\\nACAA1A\\nLRRC1\\n2810432D09RIK\\nSEPHS2\\nSAC3D1\\nTMLHE\\nLOC623451\\nTSR2\\nPLEKHA7\\nGYS2\\nARHGEF12\\nHIBCH\\nLYRM2\\nZBTB44\\nENTPD5\\nRAB11FIP2\\nLIPT1\\nINTU\\nANXA13\\nKLF12\\nSAT2\\nGAL3ST2\\nVAMP8\\nFKBPL\\nAQP11\\nTRAP1\\nPMPCB\\nTM7SF3\\nRBM39\\nBRI3\\nKDR\\nZFP748\\nNAP1L1\\nDHRS1\\nLRRC56\\nWDR20A\\nSTXBP2\\nKLF1\\nUFC1\\nCCDC16\\n9230114K14RIK\\nRWDD3\\n2610528K11RIK\\nACO1\\nCABLES1\\nLOC100047214\\nYARS2\\nLYPLA1\\nKALRN\\nGYK\\nZFP787\\nZFP655\\nRABEPK\\nZFP650\\n4732466D17RIK\\nEXOSC4\\nWDR42A\\nGPHN\\n2610528J11RIK\\n1110003E01RIK\\nMDH1\\n1200014M14RIK\\nAW209491\\nMUT\\n1700123L14RIK\\n2610036D13RIK\\nCOX15\\nTMEM30A\\nNSMCE4A\\nTM2D2\\nRHBDD3\\nATXN2\\nNFS1\\n3110001I20RIK\\nBC038156\\nLOC100047782\\n2410012H22RIK\\nRILP\\nA230062G08RIK\\nPTTG1IP\\nRAB1\\nAFAP1L1\\nLYRM5\\n2310026E23RIK\\nC330002I19RIK\\nZFYVE20\\nPOLI\\nTOMM70A\\nSLC7A6OS\\nMAT2B\\n4932438A13RIK\\nLRRC8A\\nSMO\\nNUPL2\\nTRPC2\\nARSK\\nD630023B12RIK\\nMTFR1\\n5730414N17RIK\\nSCP2\\nZRSR1\\nNOL7\\nC330018D20RIK\\nIFT122\\nLOC100046168\\nD730039F16RIK\\nSCYL1\\n1700023B02RIK\\n1700034H14RIK\\nFBXO8\\nPAIP1\\nTMEM186\\nATPAF1\\nLOC100046254\\nLOC100047604\\nCOQ10A\\nFN3K\\nSIPA1L1\\nSLC25A16\\nSLC25A40\\nRPS6KA5\\nTRIM37\\nLRRC61\\nABHD3\\nGBE1\\nPARP16\\nHSD3B2\\nESM1\\nDNAJC18\\nDOLPP1\\nLASS2\\nWDR34\\nRFESD\\nCACNB4\\n2310042D19RIK\\nSRR\\nBPNT1\\n6530415H11RIK\\nCLCC1\\nTFB1M\\n4632404H12RIK\\nD4BWG0951E\\nMED14\\nADHFE1\\nTHTPA\\nCAT\\nELL3\\nAKR7A5\\nMTMR14\\nTIMM44\\nSF1\\nIPP\\nIAH1\\nTRIM23\\nWDR89\\nGSTZ1\\nCRADD\\n2510006D16RIK\\nFBXL6\\nLOC100044400\\nZFP106\\nCD55\\n0610013E23RIK\\nAFMID\\nTMEM86A\\nALDH6A1\\nDALRD3\\nSMYD4\\nNME7\\nFARS2\\nTASP1\\nCLDN10\\nA930005H10RIK\\nSLC9A6\\nADK\\nRBKS\\n2210016F16RIK\\nVWCE\\n4732435N03RIK\\nZFP11\\nVLDLR\\n9630013D21RIK\\n4933407N01RIK\\nFAHD1\\nMIPOL1\\n1810019D21RIK\\n1810049H13RIK\\nTFAM\\nPAICS\\n1110032A03RIK\\nLOC100044139\\nDNAJC19\\nBC016495\\nA930041I02RIK\\nRQCD1\\nUSP34\\nZCCHC3\\nH2AFJ\\nPHF7\\n4921508D12RIK\\nKMO\\nPRPF18\\nMCAT\\nTXNDC4\\n4921530L18RIK\\nVPS13B\\nSCRN3\\nTOR1A\\nAI316807\\nACBD4\\nFAH\\nAPOOL\\nCOL4A4\\nLRRC19\\nGNMT\\nNR3C1\\nSIP1\\nASCC1\\nFECH\\nABHD14A\\nARHGAP18\\n2700046G09RIK\\nYME1L1\\nGK5\\nGLO1\\nSBK1\\nCISD1\\n2210011C24RIK\\nNXT2\\nNOTUM\\nANKRD42\\nUBE2E1\\nNDUFV1\\nSLC33A1\\nCEP68\\nRPS6KB1\\nHYI\\nALDH1A3\\nMYNN\\n3110048L19RIK\\nRDH14\\nPROZ\\nGORASP1\\nLOC674449\\nZFP775\\n5430437P03RIK\\nNPY\\nADH5\\nSYBL1\\n4930432O21RIK\\nNAT9\\nLOC100048387\\nMETTL8\\nENY2\\n2410018G20RIK\\nPGM2\\nFGFR4\\nMOBKL2B\\nATAD3A\\n4932432K03RIK\\nDHTKD1\\nUBOX5\\nA530050D06RIK\\nZDHHC5\\nMGAT1\\nNUDT6\\nTPMT\\nWBSCR18\\nLOC100041586\\nCDK5RAP1\\n4833426J09RIK\\nMYO6\\nCPT1A\\nGADD45GIP1\\nTMBIM4\\n2010309E21RIK\\nASB9\\n2610019F03RIK\\n7530414M10RIK\\nATP6V1B2\\n2310068J16RIK\\nDDT\\nKLHDC4\\nHPN\\nLIFR\\nOVOL1\\nNUDT12\\nCDAN1\\nFBXO9\\nFBXL3\\nHOXA7\\nALDH8A1\\n3110057O12RIK\\nABHD11\\nPSMB1\\nENSMUSG00000074286\\nCHPT1\\nOXSM\\n2310009A05RIK\\n1700001L05RIK\\nZFP148\\n39509\\nMRPL9\\nTMEM80\\n9030420J04RIK\\nNAGLU\\nPLSCR2\\nAGBL3\\nPEX1\\nCNO\\nNEO1\\nASF1A\\nTNFSF5IP1\\nPKIG\\nAI931714\\nD130020L05RIK\\nCNTD1\\nCLEC2H\\nZKSCAN1\\n1810044D09RIK\\nMETTL7A\\nSIAE\\nFBXO3\\nFZD5\\nTMEM166\\nTMED4\\nGPR155\\nRNF167\\nSPTLC1\\nRIOK2\\nTGDS\\nPMS1\\nPITPNC1\\nPCSK7\\n4933403G14RIK\\nEI24\\nCREBL2\\nTLN1\\nMRPL35\\n2700038C09RIK\\nUBIE\\nOSGEPL1\\n2410166I05RIK\\nWDR24\\nAP4S1\\nLRRC44\\nB3BP\\nITFG1\\nDMXL1\\nC1D', description='Paste your gene list (One gene per row).', section = 'section1') }}\n",
    "\n",
    "enrichr_library = '{{ ChoiceField(name='enrichr_library', description='Select one Enrichr library for which to create visualizations.', label='Enrichr Library', default='WikiPathways_2019_Human', section = 'section2',choices=[\n",
    "    'ARCHS4_TFs_Coexp',\n",
    "    'ChEA_2016',\n",
    "    'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "    'ENCODE_Histone_Modifications_2015',\n",
    "    'ENCODE_TF_ChIP-seq_2015',\n",
    "    'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "    'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "    'Genome_Browser_PWMs',\n",
    "    'lncHUB_lncRNA_Co-Expression',\n",
    "    'miRTarBase_2017',\n",
    "    'TargetScan_microRNA_2017',\n",
    "    'TF-LOF_Expression_from_GEO',\n",
    "    'TF_Perturbations_Followed_by_Expression',\n",
    "    'Transcription_Factor_PPIs',\n",
    "    'TRANSFAC_and_JASPAR_PWMs',\n",
    "    'TRRUST_Transcription_Factors_2019',\n",
    "    'ARCHS4_Kinases_Coexp',\n",
    "    'BioCarta_2016',\n",
    "    'BioPlanet_2019',\n",
    "    'BioPlex_2017',\n",
    "    'CORUM',\n",
    "    'Elsevier_Pathway_Collection',\n",
    "    'HMS_LINCS_KinomeScan',\n",
    "    'HumanCyc_2016',\n",
    "    'huMAP',\n",
    "    'KEA_2015',\n",
    "    'KEGG_2019_Human',\n",
    "    'KEGG_2019_Mouse',\n",
    "    'Kinase_Perturbations_from_GEO_down',\n",
    "    'Kinase_Perturbations_from_GEO_up',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "    'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "    'NCI-Nature_2016',\n",
    "    'NURSA_Human_Endogenous_Complexome',\n",
    "    'Panther_2016',\n",
    "    'Phosphatase_Substrates_from_DEPOD',\n",
    "    'PPI_Hub_Proteins',\n",
    "    'Reactome_2016',\n",
    "    'SILAC_Phosphoproteomics',\n",
    "    'SubCell_BarCode',\n",
    "    'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "    'WikiPathways_2019_Human',\n",
    "    'WikiPathways_2019_Mouse',\n",
    "    'GO_Biological_Process_2018',\n",
    "    'GO_Cellular_Component_2018',\n",
    "    'GO_Molecular_Function_2018',\n",
    "    'Human_Phenotype_Ontology',\n",
    "    'Jensen_COMPARTMENTS',\n",
    "    'Jensen_DISEASES',\n",
    "    'Jensen_TISSUES',\n",
    "    'MGI_Mammalian_Phenotype_Level_4_2019',  \n",
    "    'Achilles_fitness_decrease',\n",
    "    'Achilles_fitness_increase',\n",
    "    'ARCHS4_IDG_Coexp',\n",
    "    'ClinVar_2019',\n",
    "    'dbGaP',\n",
    "    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "    'DisGeNET',\n",
    "    'DrugMatrix',\n",
    "    'DSigDB',\n",
    "    'GeneSigDB',\n",
    "    'GWAS_Catalog_2019',\n",
    "    'LINCS_L1000_Chem_Pert_down',\n",
    "    'LINCS_L1000_Chem_Pert_up',\n",
    "    'LINCS_L1000_Ligand_Perturbations_down',\n",
    "    'LINCS_L1000_Ligand_Perturbations_up',\n",
    "    'MSigDB_Computational',\n",
    "    'MSigDB_Oncogenic_Signatures',\n",
    "    'Old_CMAP_down',\n",
    "    'Old_CMAP_up',\n",
    "    'OMIM_Disease',\n",
    "    'OMIM_Expanded',\n",
    "    'PheWeb_2019',\n",
    "    'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "    'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "    'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "    'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "    'UK_Biobank_GWAS_v1',\n",
    "    'Virus_Perturbations_from_GEO_down',\n",
    "    'Virus_Perturbations_from_GEO_up',\n",
    "    'VirusMINT',       \n",
    "    'Allen_Brain_Atlas_down',\n",
    "    'Allen_Brain_Atlas_up',\n",
    "    'ARCHS4_Cell-lines',\n",
    "    'ARCHS4_Tissues',\n",
    "    'Cancer_Cell_Line_Encyclopedia',\n",
    "    'CCLE_Proteomics_2020',\n",
    "    'ESCAPE',\n",
    "    'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
    "    'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
    "    'Human_Gene_Atlas',\n",
    "    'Mouse_Gene_Atlas',\n",
    "    'NCI-60_Cancer_Cell_Lines',\n",
    "    'ProteomicsDB_2020',\n",
    "    'Tissue_Protein_Expression_from_Human_Proteome_Map',           \n",
    "    'Chromosome_Location_hg19',\n",
    "    'Data_Acquisition_Method_Most_Popular_Genes',\n",
    "    'Enrichr_Libraries_Most_Popular_Genes',\n",
    "    'Genes_Associated_with_NIH_Grants',\n",
    "    'HMDB_Metabolites',\n",
    "    'HomoloGene',\n",
    "    'InterPro_Domains_2019',\n",
    "    'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
    "    'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
    "    'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
    "    'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
    "    'Pfam_Domains_2019',\n",
    "    'Pfam_InterPro_Domains',\n",
    "    'Table_Mining_of_CRISPR_Studies',               \n",
    "    'BioCarta_2013',\n",
    "    'BioCarta_2015',\n",
    "    'ChEA_2013',\n",
    "    'ChEA_2015',\n",
    "    'Chromosome_Location',\n",
    "    'Disease_Signatures_from_GEO_down_2014',\n",
    "    'Disease_Signatures_from_GEO_up_2014',\n",
    "    'Drug_Perturbations_from_GEO_2014',\n",
    "    'ENCODE_Histone_Modifications_2013',\n",
    "    'ENCODE_TF_ChIP-seq_2014',\n",
    "    'GO_Biological_Process_2013',\n",
    "    'GO_Biological_Process_2015',\n",
    "    'GO_Biological_Process_2017',\n",
    "    'GO_Biological_Process_2017b',\n",
    "    'GO_Cellular_Component_2013',\n",
    "    'GO_Cellular_Component_2015',\n",
    "    'GO_Cellular_Component_2017',\n",
    "    'GO_Cellular_Component_2017b',\n",
    "    'GO_Molecular_Function_2013',\n",
    "    'GO_Molecular_Function_2015',\n",
    "    'GO_Molecular_Function_2017',\n",
    "    'GO_Molecular_Function_2017b',\n",
    "    'HumanCyc_2015',\n",
    "    'KEA_2013',\n",
    "    'KEGG_2013',\n",
    "    'KEGG_2015',\n",
    "    'KEGG_2016',\n",
    "    'MGI_Mammalian_Phenotype_2013',\n",
    "    'MGI_Mammalian_Phenotype_2017',\n",
    "    'MGI_Mammalian_Phenotype_Level_3',\n",
    "    'MGI_Mammalian_Phenotype_Level_4',\n",
    "    'NCI-Nature_2015',\n",
    "    'Panther_2015',\n",
    "    'Reactome_2013',\n",
    "    'Reactome_2015',\n",
    "    'TargetScan_microRNA',\n",
    "    'Tissue_Protein_Expression_from_ProteomicsDB',\n",
    "    'WikiPathways_2013',\n",
    "    'WikiPathways_2015',\n",
    "    'WikiPathways_2016',               \n",
    "    'Aging_Perturbations_from_GEO_down',\n",
    "    'Aging_Perturbations_from_GEO_up',\n",
    "    'Disease_Perturbations_from_GEO_down',\n",
    "    'Disease_Perturbations_from_GEO_up',\n",
    "    'Drug_Perturbations_from_GEO_down',\n",
    "    'Drug_Perturbations_from_GEO_up',\n",
    "    'Gene_Perturbations_from_GEO_down',\n",
    "    'Gene_Perturbations_from_GEO_up',\n",
    "    'Ligand_Perturbations_from_GEO_down',\n",
    "    'Ligand_Perturbations_from_GEO_up',\n",
    "    'MCF7_Perturbations_from_GEO_down',\n",
    "    'MCF7_Perturbations_from_GEO_up',\n",
    "    'Microbe_Perturbations_from_GEO_down',\n",
    "    'Microbe_Perturbations_from_GEO_up',\n",
    "    'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
    "    'SysMyo_Muscle_Gene_Sets']) }}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set Other Parameters\n",
    "For simplicity, the only inputs for this appyter are a gene list and one library. Other parameters are set to default values in the cell below. You can download the notebook, change these parameters, and rerun it if you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scatter Plot Parameters\n",
    "significance_value = 0.05\n",
    "\n",
    "# Bar Chart Parameters\n",
    "figure_file_format = ['png', 'svg']\n",
    "output_file_name = 'Enrichr_results'\n",
    "color = 'lightskyblue'\n",
    "final_output_file_names = [str(output_file_name+'.'+file_type) for file_type in figure_file_format]\n",
    "\n",
    "# Hexagonal Canvas Parameters\n",
    "canvas_color = 'Blue'\n",
    "num_hex_colored = 10\n",
    "\n",
    "# Manhattan Plot Parameters\n",
    "manhattan_colors = ['#003f5c', '#7a5195', '#ef5675', '#ffa600']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import gene list as file or from text box file\n",
    "# Will choose file upload over textbox if a file is given \n",
    "if gene_list_filename != '':\n",
    "    open_gene_list_file = open(gene_list_filename,'r')\n",
    "    lines = open_gene_list_file.readlines()\n",
    "    genes = [x.strip() for x in lines]\n",
    "    open_gene_list_file.close()\n",
    "else:\n",
    "    genes = gene_list_input.split('\\n')\n",
    "    genes = [x.strip() for x in genes]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scatter Plot Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_library(library_name):\n",
    "    # download pre-processed library data\n",
    "    try:\n",
    "        df = pd.read_csv('https://raw.githubusercontent.com/skylar73/Enrichr-Processed-Library-Storage/master/Scatterplot/Libraries/' + library_name + '.csv')\n",
    "    except:\n",
    "        return -1, -1, -1\n",
    "\n",
    "    name = df['Name'].tolist()\n",
    "    gene_list = df['Genes'].tolist()\n",
    "    library_data = [list(a) for a in zip(name, gene_list)]\n",
    "    return genes, library_data, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# enrichment analysis\n",
    "def get_library_iter(library_data):\n",
    "    for member in library_data:\n",
    "        term = member[0]\n",
    "        gene_set = member[1].split(' ')\n",
    "        yield term, gene_set\n",
    "\n",
    "def get_enrichment_results(genes, library_data):\n",
    "    return sorted(enrich_crisp(genes, get_library_iter(library_data), 20000, True), key=lambda r: r[1].pvalue)\n",
    "\n",
    "def get_pvalue(row, unzipped_results, all_results):\n",
    "    if row['Name'] in list(unzipped_results[0]):\n",
    "        index = list(unzipped_results[0]).index(row['Name'])\n",
    "        return all_results[index][1].pvalue\n",
    "    else:\n",
    "        return 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# call enrichment results and return a plot and dataframe for Scatter Plot\n",
    "def get_plot(library_name):\n",
    "    genes, library_data, df = download_library(library_name)\n",
    "\n",
    "    # library not supported\n",
    "    if genes == -1:\n",
    "        return -1 ,-1\n",
    "\n",
    "    all_results = get_enrichment_results(genes, library_data)\n",
    "    unzipped_results = list(zip(*all_results))\n",
    "\n",
    "    if len(all_results) == 0:\n",
    "        print(\"There are no enriched terms with your inputted gene set and the \", library_name, \" library.\")\n",
    "        my_colors = ['#808080'] * len(df.index)\n",
    "\n",
    "        source = ColumnDataSource(\n",
    "            data=dict(\n",
    "                x = df['x'],\n",
    "                y = df['y'],\n",
    "                gene_set = df['Name'],\n",
    "                colors = my_colors,\n",
    "                sizes = [6] * len(df.index)\n",
    "            )\n",
    "        )\n",
    "\n",
    "        hover_emb = HoverTool(names=[\"df\"], tooltips=\"\"\"\n",
    "            <div style=\"margin: 10\">\n",
    "                <div style=\"margin: 0 auto; width:200px;\">\n",
    "                    <span style=\"font-size: 12px; font-weight: bold;\">Gene Set:</span>\n",
    "                    <span style=\"font-size: 12px\">@gene_set</span>\n",
    "                </div>\n",
    "            </div>\n",
    "            \"\"\")\n",
    "    else:\n",
    "        # add p value to the dataframe\n",
    "        df['p value'] = df.apply (lambda row: get_pvalue(row, unzipped_results, all_results), axis=1)\n",
    "\n",
    "        # normalize p values for color scaling\n",
    "        cmap = mpl.cm.get_cmap('Blues_r')\n",
    "        norm = colors.Normalize(vmin = df['p value'].min(), vmax=significance_value*2)\n",
    "\n",
    "        my_colors = []\n",
    "        my_sizes = []\n",
    "        for index, row in df.iterrows():\n",
    "            if row['p value'] < significance_value:\n",
    "                my_colors += [mpl.colors.to_hex(cmap(norm(row['p value'])))]\n",
    "                my_sizes += [12]\n",
    "            else:\n",
    "                my_colors += ['#808080']\n",
    "                my_sizes += [6]\n",
    "\n",
    "        source = ColumnDataSource(\n",
    "                data=dict(\n",
    "                    x = df['x'],\n",
    "                    y = df['y'],\n",
    "                    gene_set = df['Name'],\n",
    "                    p_value = df['p value'],\n",
    "                    colors = my_colors,\n",
    "                    sizes = my_sizes\n",
    "                )\n",
    "            )\n",
    "\n",
    "        hover_emb = HoverTool(names=[\"df\"], tooltips=\"\"\"\n",
    "            <div style=\"margin: 10\">\n",
    "                <div style=\"margin: 0 auto; width:200px;\">\n",
    "                    <span style=\"font-size: 12px; font-weight: bold;\">Gene Set:</span>\n",
    "                    <span style=\"font-size: 12px\">@gene_set</span>\n",
    "                    <span style=\"font-size: 12px; font-weight: bold;\">p-value:</span>\n",
    "                    <span style=\"font-size: 12px\">@p_value</span>\n",
    "                </div>\n",
    "            </div>\n",
    "            \"\"\")\n",
    "\n",
    "    tools_emb = [hover_emb, 'pan', 'wheel_zoom', 'reset', 'save']\n",
    "\n",
    "    plot_emb = figure(plot_width=700, plot_height=800, tools=tools_emb)\n",
    "\n",
    "    # hide axis labels and grid lines\n",
    "    plot_emb.xaxis.major_tick_line_color = None\n",
    "    plot_emb.xaxis.minor_tick_line_color = None\n",
    "    plot_emb.yaxis.major_tick_line_color = None\n",
    "    plot_emb.yaxis.minor_tick_line_color = None\n",
    "    plot_emb.xaxis.major_label_text_font_size = '0pt'\n",
    "    plot_emb.yaxis.major_label_text_font_size = '0pt' \n",
    "\n",
    "    plot_emb.circle('x', 'y', size = 'sizes', alpha = 0.7, line_alpha = 0, \n",
    "                    line_width = 0.01, source = source, fill_color = 'colors', name = \"df\")\n",
    "\n",
    "    plot_emb.output_backend = \"svg\"\n",
    "    \n",
    "    return plot_emb, df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Enrichr API Function for Manhattan Plot and Bar Chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Takes a gene list and Enrichr libraries as input\n",
    "def Enrichr_API(enrichr_gene_list, all_libraries):\n",
    "\n",
    "    all_terms = []\n",
    "    all_pvalues =[] \n",
    "    all_adjusted_pvalues = []\n",
    "\n",
    "    for library_name in all_libraries : \n",
    "        ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/addList'\n",
    "        genes_str = '\\n'.join(enrichr_gene_list)\n",
    "        description = 'Example gene list'\n",
    "        payload = {\n",
    "            'list': (None, genes_str),\n",
    "            'description': (None, description)\n",
    "        }\n",
    "\n",
    "        response = requests.post(ENRICHR_URL, files=payload)\n",
    "        if not response.ok:\n",
    "            raise Exception('Error analyzing gene list')\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "        time.sleep(0.5)\n",
    "        ENRICHR_URL = 'http://amp.pharm.mssm.edu/Enrichr/enrich'\n",
    "        query_string = '?userListId=%s&backgroundType=%s'\n",
    "        user_list_id = data['userListId']\n",
    "        short_id = data[\"shortId\"]\n",
    "        gene_set_library = library_name\n",
    "        response = requests.get(\n",
    "            ENRICHR_URL + query_string % (user_list_id, gene_set_library)\n",
    "         )\n",
    "        if not response.ok:\n",
    "            raise Exception('Error fetching enrichment results')\n",
    "\n",
    "        data = json.loads(response.text)\n",
    "\n",
    "        short_results_df  = pd.DataFrame(data[library_name][0:10])\n",
    "        all_terms.append(list(short_results_df[1]))\n",
    "        all_pvalues.append(list(short_results_df[2]))\n",
    "        all_adjusted_pvalues.append(list(short_results_df[6]))\n",
    "        \n",
    "        results_df  = pd.DataFrame(data[library_name])\n",
    "        # adds library name to the data frame so the libraries can be distinguished\n",
    "        results_df['library'] = library_name.replace('_', '')\n",
    "\n",
    "    return([results_df, short_results_df, all_terms, all_pvalues, all_adjusted_pvalues, str(short_id)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bar Chart Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function plots results in bar chart\n",
    "# Takes all terms, all p-values, all adjusted p-values, plot title, Enrichr libraries, and specified figure format\n",
    "def enrichr_figure(all_terms, all_pvalues, all_adjusted_pvalues, plot_names, all_libraries, fig_format, bar_color): \n",
    "    \n",
    "    # rows and columns depend on number of Enrichr libraries submitted \n",
    "    rows = []\n",
    "    cols = []\n",
    "    \n",
    "    # Bar colors\n",
    "    if bar_color != 'lightgrey':\n",
    "        bar_color_not_sig = 'lightgrey'\n",
    "        edgecolor=None\n",
    "        linewidth=0\n",
    "    else:\n",
    "        bar_color_not_sig = 'white'\n",
    "        edgecolor='black'\n",
    "        linewidth=1\n",
    "    \n",
    "\n",
    "    plt.figure(figsize=(12, 12))\n",
    "    rows = [0]\n",
    "    cols = [0]\n",
    "    i = 0 \n",
    "    bar_colors = [bar_color if (x < 0.05) else bar_color_not_sig for x in all_pvalues[i]]\n",
    "    fig = sns.barplot(x=np.log10(all_pvalues[i])*-1, y=all_terms[i], palette=bar_colors, edgecolor=edgecolor, linewidth=linewidth)\n",
    "    fig.axes.get_yaxis().set_visible(False)\n",
    "    fig.set_title(all_libraries[i].replace('_', ' '), fontsize=26)\n",
    "    fig.set_xlabel('-Log10(p-value)', fontsize=25)\n",
    "    fig.xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "    fig.tick_params(axis='x', which='major', labelsize=20)\n",
    "    if max(np.log10(all_pvalues[i])*-1)<1:\n",
    "        fig.xaxis.set_ticks(np.arange(0, max(np.log10(all_pvalues[i])*-1), 0.1))\n",
    "    for ii,annot in enumerate(all_terms[i]):\n",
    "        if all_adjusted_pvalues[i][ii] < 0.05:\n",
    "            annot = '  *'.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii], precision=2)))]) \n",
    "        else:\n",
    "            annot = '  '.join([annot, str(str(np.format_float_scientific(all_pvalues[i][ii], precision=2)))])\n",
    "\n",
    "        title_start= max(fig.axes.get_xlim())/200\n",
    "        fig.text(title_start, ii, annot, ha='left', wrap = True, fontsize = 26)\n",
    "        fig.patch.set_edgecolor('black')  \n",
    "        fig.patch.set_linewidth('2')\n",
    "    \n",
    "    # Save results \n",
    "    for plot_name in plot_names:\n",
    "        plt.savefig(plot_name, bbox_inches = 'tight')\n",
    "    \n",
    "    # Show plot \n",
    "    plt.show()  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hexagonal Canvas Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def library_processing():\n",
    "    # downloads library data for the hexagonal canvas\n",
    "    # library data is pre-annealed so the canvas will have the most similar gene sets closest together\n",
    "\n",
    "    raw_library_data = []\n",
    "    library_data = []\n",
    "\n",
    "    try:\n",
    "        library_name = enrichr_library\n",
    "        with urllib.request.urlopen('https://raw.githubusercontent.com/skylar73/Enrichr-Processed-Library-Storage/master/Annealing/Annealed-Libraries/' + enrichr_library + '.txt') as f:\n",
    "            for line in f.readlines():\n",
    "                raw_library_data.append(line.decode('utf-8').split(\"\\t\\t\"))\n",
    "        name = []\n",
    "        gene_list = []\n",
    "    except:\n",
    "        return [], -1, -1\n",
    "\n",
    "    for i in range(len(raw_library_data)):\n",
    "        name += [raw_library_data[i][0]]\n",
    "        raw_genes = raw_library_data[i][1].split('\\t')\n",
    "        gene_list += [raw_genes[:-1]]\n",
    "\n",
    "    library_data = [list(a) for a in zip(name, gene_list)]\n",
    "\n",
    "    # raw_library_data: a 2D list where the first element is the name and the second element is a list of genes associated with that name\n",
    "\n",
    "    jaccard_indices = []\n",
    "    indices = []\n",
    "\n",
    "    for gene_set in library_data:\n",
    "        intersection = [value for value in gene_set[1] if value in genes]\n",
    "        index = len(intersection)/(len(gene_set[1]) + len(genes))\n",
    "        jaccard_indices += [[gene_set[0], index]]\n",
    "        indices += [round(index, 5)]\n",
    "\n",
    "    # determine the dimensions of the canvas\n",
    "    x_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "    y_dimension = math.ceil(math.sqrt(len(indices)))\n",
    "\n",
    "    # zip name, gene_list, indices, and blank list for neighbor score then add dummy entries to the zipped list\n",
    "    num_hex = x_dimension*y_dimension\n",
    "    anneal_list = list(zip(name, gene_list, indices))\n",
    "\n",
    "    return anneal_list, x_dimension, y_dimension\n",
    "\n",
    "def unzip_list(anneal_list):\n",
    "    unzipped_list = zip(*anneal_list)\n",
    "    return list(unzipped_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a list of colors for the hexagonal canvas\n",
    "def get_color(anneal_list, cut_off_value, x_dimension, y_dimension):\n",
    "\n",
    "    # Deal with cut_off_value (only color the most significant 10/20 hexagons)\n",
    "    if cut_off_value == 2.0:\n",
    "        sort_list = sorted(anneal_list, key=itemgetter(2), reverse=True)\n",
    "        cut_off_value = sort_list[int(num_hex_colored)-1][2]\n",
    "\n",
    "    r_value = 0\n",
    "    g_value = 0\n",
    "    b_value = 0\n",
    "\n",
    "    if canvas_color == 'Red':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.8\n",
    "        b_value = 0.8\n",
    "    if canvas_color == 'Yellow':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.3\n",
    "        b_value = 1.0\n",
    "    if canvas_color == 'Purple':\n",
    "        r_value = 0.5\n",
    "        g_value = 1.0\n",
    "        b_value = 0.0\n",
    "    if canvas_color == 'Pink':\n",
    "        r_value = 0.0\n",
    "        g_value = 1.0\n",
    "        b_value = 0.2\n",
    "    if canvas_color == 'Orange':\n",
    "        r_value = 0.0\n",
    "        g_value = 0.45\n",
    "        b_value = 1.0\n",
    "    if canvas_color == 'Green':\n",
    "        r_value = 1.0\n",
    "        g_value = 0.0\n",
    "        b_value = 1.0\n",
    "    if canvas_color == 'Blue':\n",
    "        r_value = 1.0\n",
    "        g_value = 0.9\n",
    "        b_value = 0.0\n",
    "\n",
    "    color_list = []\n",
    "\n",
    "    unzipped_anneal_list = unzip_list(anneal_list)\n",
    "\n",
    "    max_index = max(unzipped_anneal_list[2])\n",
    "\n",
    "    if max_index != 0:\n",
    "        scaled_list = [i/max_index for i in unzipped_anneal_list[2]]\n",
    "    else:\n",
    "        scaled_list = unzipped_anneal_list[2]\n",
    "\n",
    "    for i in range(x_dimension*y_dimension):\n",
    "        if i < len(unzipped_anneal_list[2]) and float(unzipped_anneal_list[2][i]) >= cut_off_value:\n",
    "            color_list += [mpl.colors.to_hex((1-scaled_list[i]*r_value, \n",
    "            1-scaled_list[i]*g_value, 1-scaled_list[i]*b_value))]\n",
    "        elif i < len(unzipped_anneal_list[2]):\n",
    "            color_list += [mpl.colors.to_hex((1-scaled_list[i], \n",
    "            1-scaled_list[i], 1-scaled_list[i]))]\n",
    "        else:\n",
    "            color_list += [\"#FFFFFF\"]\n",
    "    return color_list, max_index, cut_off_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def init_chart():\n",
    "  chart_id = 'mychart-' + str(uuid.uuid4())\n",
    "  display(HTML('<script src=\"/static/components/requirejs/require.js\"></script>'))\n",
    "  display(HTML(Template(dedent('''\n",
    "  <script>\n",
    "  require.config({\n",
    "    paths: {\n",
    "      'd3': 'https://cdnjs.cloudflare.com/ajax/libs/d3/5.16.0/d3.min',\n",
    "      'd3-hexbin': 'https://d3js.org/d3-hexbin.v0.2.min',\n",
    "    },\n",
    "    shim: {\n",
    "      'd3-hexbin': ['d3']\n",
    "    }\n",
    "  })\n",
    "\n",
    "  // If we configure mychart via url, we can eliminate this define here\n",
    "  define($chart_id, ['d3', 'd3-hexbin'], function(d3, d3_hexbin) {\n",
    "    return function (figure_id, numA, numB, colorList, libraryList, indices) {\n",
    "      var margin = {top: 50, right: 20, bottom: 20, left: 50},\n",
    "        width = 850 - margin.left - margin.right,\n",
    "        height = 350 - margin.top - margin.bottom;\n",
    "\n",
    "      // append the svg object to the body of the page\n",
    "      var svG = d3.select('#' + figure_id)\n",
    "                  .attr(\"width\", width + margin.left + margin.right)\n",
    "                  .attr(\"height\", height + margin.top + margin.bottom)\n",
    "                  .append(\"g\")\n",
    "                  .attr(\"transform\", \"translate(\" + margin.left + \",\" + margin.top + \")\");\n",
    "      \n",
    "      //The number of columns and rows of the heatmap\n",
    "      var MapColumns = numA,\n",
    "          MapRows = numB;\n",
    "\n",
    "      //The maximum radius the hexagons can have to still fit the screen\n",
    "      var hexRadius = d3.min([width/((MapColumns + 0.5) * Math.sqrt(3)), height/((MapRows + 1/3) * 1.5)]);\n",
    "\n",
    "      //Calculate the center position of each hexagon\n",
    "      var points = [];\n",
    "      for (var i = 0; i < MapRows; i++) {\n",
    "          for (var j = 0; j < MapColumns; j++) {\n",
    "              var x = hexRadius * j * Math.sqrt(3)\n",
    "              //Offset each uneven row by half of a \"hex-width\" to the right\n",
    "              if(i%2 === 1) x += (hexRadius * Math.sqrt(3))/2\n",
    "              var y = hexRadius * i * 1.5\n",
    "              points.push([x,y])\n",
    "          }\n",
    "      }\n",
    "\n",
    "      //Set the hexagon radius\n",
    "      var hexbin = d3_hexbin.hexbin().radius(hexRadius);\n",
    "\n",
    "      svG.append(\"g\")\n",
    "        .selectAll(\".hexagon\")\n",
    "        .data(hexbin(points))\n",
    "        .enter().append(\"path\")\n",
    "        .attr(\"class\", \"hexagon\")\n",
    "        .attr(\"d\", function (d) {\n",
    "            return \"M\" + d.x + \",\" + d.y + hexbin.hexagon();\n",
    "        })\n",
    "        .attr(\"stroke\", \"black\")\n",
    "        .attr(\"stroke-width\", \"1px\")\n",
    "        .style(\"fill\", function (d,i) { return colorList[i]; })\n",
    "        .on(\"mouseover\", mover)\n",
    "        .on(\"mouseout\", mout)\n",
    "        .append(\"svg:title\")\n",
    "        .text(function(d,i) { return libraryList[i].concat(\" \").concat(indices[i]); });\n",
    "\n",
    "      // Mouseover function\n",
    "      function mover(d) {\n",
    "      d3.select(this)\n",
    "        .transition().duration(10)  \n",
    "        .style(\"fill-opacity\", 0.3)\n",
    "      };\n",
    "\n",
    "      // Mouseout function\n",
    "      function mout(d) { \n",
    "      d3.select(this)\n",
    "        .transition().duration(10)\n",
    "        .style(\"fill-opacity\", 1)\n",
    "      };\n",
    "\n",
    "  }\n",
    "\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({ 'chart_id': repr(chart_id) })))\n",
    "  return chart_id\n",
    "\n",
    "def Canvas(numA, numB, colorList, libraryList, indices):\n",
    "  chart_id = init_chart()\n",
    "  display(HTML(Template(dedent('''\n",
    "  <svg id=$figure_id></svg>\n",
    "  <script>\n",
    "  require([$chart_id], function(mychart) {\n",
    "    mychart($figure_id, $numA, $numB, $colorList, $libraryList, $indices)\n",
    "  })\n",
    "  </script>\n",
    "  ''')).substitute({\n",
    "      'chart_id': repr(chart_id),\n",
    "      'figure_id': repr('fig-' + str(uuid.uuid4())),\n",
    "      'numA': repr(numA),\n",
    "      'numB': repr(numB),\n",
    "      'colorList': repr(colorList),\n",
    "      'libraryList': repr(libraryList),\n",
    "      'indices': repr(indices)\n",
    "  })))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Manhattan Plot Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Processes Enrichr data for Manhattan plots\n",
    "\n",
    "def get_data(genes):\n",
    "    # Process Enrichr data\n",
    "    sorted_data = pd.DataFrame({\"Gene Set\": [], \"-log(p value)\": [], \"Library\": []})\n",
    "\n",
    "    # get enrichr results from the library selected\n",
    "    results_df = Enrichr_API(genes, [enrichr_library])[0]\n",
    "\n",
    "    all_terms = []\n",
    "    all_pvalues = []\n",
    "    library_names = []\n",
    "\n",
    "    all_terms.append(list(results_df[1]))\n",
    "    all_pvalues.append(list(results_df[2]))\n",
    "    library_names.append(list(results_df['library']))\n",
    "\n",
    "    x=np.log10(all_pvalues[0])*-1\n",
    "    sorted_terms = list(zip(all_terms[0], x, library_names[0]))\n",
    "    sorted_terms = sorted(sorted_terms, key = itemgetter(0))\n",
    "    unzipped_sorted_list = list(zip(*sorted_terms))\n",
    "\n",
    "    data = pd.DataFrame({\"Gene Set\": unzipped_sorted_list[0], \"-log(p value)\": unzipped_sorted_list[1], \"Library\": unzipped_sorted_list[2]})\n",
    "\n",
    "    sorted_data = pd.concat([sorted_data, data])\n",
    "\n",
    "    # group data by library\n",
    "    groups = sorted_data.groupby(\"Library\")\n",
    "    return sorted_data, groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Manhattan Plots\n",
    "def manhattan(sorted_data, groups):\n",
    "    # split data frame into smaller data frames by library\n",
    "    list_of_df = []\n",
    "    for library_name in [enrichr_library]:\n",
    "        library_name = library_name.replace('_', '')\n",
    "        df_new = sorted_data[sorted_data['Library'] == library_name]\n",
    "        list_of_df += [df_new]\n",
    "\n",
    "    list_of_xaxis_values = []\n",
    "    for df in list_of_df:  \n",
    "        list_of_xaxis_values += df[\"Gene Set\"].values.tolist()\n",
    "\n",
    "    # define the output figure and the features we want\n",
    "    p = figure(x_range = list_of_xaxis_values, plot_height=300, plot_width=750, tools='pan, box_zoom, hover, reset, save')\n",
    "\n",
    "    # loop over all libraries\n",
    "    r = []\n",
    "    categories = []\n",
    "    color_index = 0\n",
    "    for df in list_of_df:\n",
    "        if color_index >= len(manhattan_colors):\n",
    "            color_index = 0 \n",
    "\n",
    "        # calculate actual p value from -log(p value)\n",
    "        actual_pvalues = []\n",
    "        for log_value in df[\"-log(p value)\"].values.tolist():\n",
    "            actual_pvalues += [\"{:.5e}\".format(10**(-1*log_value))]\n",
    "\n",
    "        # define ColumnDataSource with our data for this library\n",
    "        source = ColumnDataSource(data=dict(\n",
    "            x = df[\"Gene Set\"].values.tolist(),\n",
    "            y = df[\"-log(p value)\"].values.tolist(),\n",
    "            pvalue = actual_pvalues,\n",
    "        ))\n",
    "    \n",
    "        # plot data from this library\n",
    "        r += [p.circle(x = 'x', y = 'y', size=5, fill_color=manhattan_colors[color_index], line_color = manhattan_colors[color_index], line_width=1, source = source)]\n",
    "        color_index += 1\n",
    "\n",
    "    p.background_fill_color = 'white'\n",
    "    p.xaxis.major_tick_line_color = None \n",
    "    p.xaxis.major_label_text_font_size = '0pt'\n",
    "    p.y_range.start = 0\n",
    "    p.yaxis.axis_label = '-log(p value)'\n",
    "\n",
    "    p.hover.tooltips = [\n",
    "        (\"Gene Set\", \"@x\"),\n",
    "        (\"p value\", \"@pvalue\"),\n",
    "    ]\n",
    "    p.output_backend = \"svg\"\n",
    "    \n",
    "    # returns the plot\n",
    "    return p"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Outputs\n",
    "## Scatter Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display Scatter Plot\n",
    "plot, df = get_plot(enrichr_library)\n",
    "if plot == -1:\n",
    "    display(Markdown(\"The Scatter Plot visualization is not available for the selected library at this time.\"))\n",
    "else:\n",
    "    show(plot)\n",
    "    display(Markdown('You may have to zoom in to see the details in densely-populated portions of the plots.'))\n",
    "    display(Markdown('The larger blue points represent significantly enriched terms (the darker the blue, the more significant).'))\n",
    "    display(Markdown('The gray points are not significant.'))\n",
    "    display(Markdown('Hovering over points will display the associated gene set name and the p-value.'))\n",
    "    display(Markdown('Plots can be downloaded as an svg using the save function on the toolbar next to the plot.'))\n",
    "    display(Markdown('[Access the complete Scatter Plot Visualization Appyter here to create and compare up to 9 scatter plots at once](https://appyters.maayanlab.cloud/#Enrichr_Scatterplot_Appyter)'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bar Chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display Bar Chart\n",
    "results = Enrichr_API(genes, [enrichr_library])\n",
    "enrichr_figure(results[2], results[3], results[4], final_output_file_names, [enrichr_library], figure_file_format, color)\n",
    "# Download Bar Chart\n",
    "for i, file in enumerate(final_output_file_names):\n",
    "    display(FileLink(file, result_html_prefix=str('Download ' + figure_file_format[i] + ': ')))\n",
    "display(Markdown('The bar chart contains the top 10 enriched terms and their corresponding p-values for the chosen library. Colored bars correspond to terms with significant p-values (<0.05). An asterisk (*) next to a p-value indicates the term also has a significant adjusted p-value (<0.05).'))\n",
    "display(Markdown('[Access the complete Bar Chart Appyter here to create customized bar charts for multiple libraries at once.](https://appyters.maayanlab.cloud/Enrichr_compressed_bar_chart_figure/)'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hexagonal Canvas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display Hexagonal Canvas\n",
    "anneal_list, x_dimension, y_dimension = library_processing()\n",
    "if x_dimension < 0:\n",
    "    display(Markdown(\"The Hexagonal Canvas visualization is not available for the selected library at this time.\"))\n",
    "else:\n",
    "    color_list, scaling_factor, cut_off_value = get_color(anneal_list, 2.0, x_dimension, y_dimension)\n",
    "    unzipped_anneal_list = unzip_list(anneal_list)\n",
    "    Canvas(x_dimension, y_dimension, color_list, list(unzipped_anneal_list[0]), list(unzipped_anneal_list[2]))\n",
    "    print(\"The color scaling factor for this canvas is: \" + str(scaling_factor))\n",
    "    print(\"Gene sets with a similarity index less than \" + str(cut_off_value) + \" are not colored\")\n",
    "    display(Markdown('Each of the hexagons on the canvas represent one gene set from the library selected. The hexagons are colored based on the Jaccard similarity index between the inputted gene list and the gene set the hexagon represents (the brighter, the more similar). The name of the gene set and the associated similarity index are displayed when a hexagon is hovered over.'))\n",
    "    display(Markdown('[Access the complete Hexagonal Canvas Appyter here to create customized canvases for 1 or 2 libraries at once.](https://appyters.maayanlab.cloud/#Enrichr_Canvas_Appyter)'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manhattan Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display Manhattan Plot\n",
    "sorted_data, groups = get_data(genes)\n",
    "show(manhattan(sorted_data, groups))\n",
    "display(Markdown('The x-axis of the plot is made up of gene sets from the library. The y-axis of the plot has the -log(p value) for each gene set. The name of the gene set and the associated p-value are displayed when a point is hovered over. You can also zoom, pan, and save the plot as an svg using the toolbar on the right.'))\n",
    "display(Markdown('[Access the complete Manhattan Plot Appyter here to create customized static and dynamic Manhattan plots to compare multiple libraries at once.](https://appyters.maayanlab.cloud/#Enrichr_Manhattan_Plot)'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of significant p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output a table of significant p-values\n",
    "def create_download_link(df, title = \"Download CSV file of this table\", filename = \"data.csv\"):  \n",
    "    csv = df.to_csv(index = False)\n",
    "    b64 = base64.b64encode(csv.encode())\n",
    "    payload = b64.decode()\n",
    "    html = '<a download=\"{filename}\" href=\"data:text/csv;base64,{payload}\" target=\"_blank\">{title}</a>'\n",
    "    html = html.format(payload=payload, title=title, filename=filename)\n",
    "    return HTML(html)\n",
    "\n",
    "if plot != -1 and 'p value' in df.columns:\n",
    "    sorted_df = df.sort_values(by = ['p value'])\n",
    "    filtered_df = sorted_df[sorted_df['p value'] <= significance_value]\n",
    "    if len(filtered_df) != 0:\n",
    "        display(HTML(f\"<strong>Table of significant p-values for {enrichr_library.replace('_', ' ')}</strong>\"))\n",
    "        display(HTML(filtered_df[['Name', 'p value']].to_html(index = False)))\n",
    "        display(create_download_link(filtered_df[['Name', 'p value']]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Link to Enrichr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get complete enrichment analysis results from Enrichr \n",
    "url = 'https://amp.pharm.mssm.edu/Enrichr/enrich?dataset=' + results[5]\n",
    "display(HTML(f'<span style=\"font-size:1.5em;\"><a href = \"https://amp.pharm.mssm.edu/Enrichr/enrich?dataset={results[5]}\"> Access your complete Enrichment results on Enrichr here. </a></span>'))"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6-final"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python37664bitbasecondad81f4d096e8b4b9885e3d702e0bb3be1",
   "display_name": "Python 3.7.6 64-bit ('base': conda)"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}