{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f9b5363",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b303d068",
   "metadata": {},
   "source": [
    "# AGSEA Appyter\n",
    "\n",
    "The Augmented Gene Set Enrichment Analysis (AGSEA) Appyter performs and visualizes standard gene set enrichment analysis (GSEA) (Subramanian et al., 2005), as well as augmented GSEA. The augmented feature adds novel genes that are potentially missing from the annotated prior knowledge gene set that are also relevant to the input signature.  \n",
    "\n",
    "AGSEA outputs a table containing the enrichment results statistics for the top gene sets from the chosen gene set library selected by the user. The full results are available for download as a CSV file, and an interactive GSEA plot is also generated. The AGSEA Appyter also highlights genes that are co-expressed with each annotated gene set and are highly ranked within the input signature. These genes are typically understudied genes that should be further explored as relevant to the biological process under investigation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2123832",
   "metadata": {},
   "outputs": [],
   "source": [
    "#imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import math\n",
    "from decimal import Decimal\n",
    "\n",
    "#loading libraries\n",
    "import requests\n",
    "import urllib.request\n",
    "from flask import url_for\n",
    "\n",
    "#performing GSEA\n",
    "import gseapy as gp\n",
    "from gseapy.gsea import GSEA\n",
    "from maayanlab_bioinformatics.dge import limma_voom\n",
    "from geode import chdir\n",
    "\n",
    "#creating visualizations\n",
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "from IPython.display import display, HTML, IFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3213bac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide\n",
    "{% do SectionField(\n",
    "    name='gsea', \n",
    "    title='Augmented Gene Set Enrichment Analysis', \n",
    "    subtitle='Perform AGSEA by inputting the following files. Please make sure that the files are loaded properly before submitting, or else there will be an error. Follow the [link](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats) to learn about the proper data formats.',\n",
    "    img='plot_icon.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='data', \n",
    "    title='Data Submission', \n",
    "    subtitle='Option 1: Upload a preranked gene list (.rnk). Option 2: Upload an expression dataset with gene counts (.gct) and a phenotype labels file (.cls), and then choose a ranking method. Please make sure that the gene symbols used in the expression file match the types used in the library file for the appyter to work properly.',\n",
    "    img='data_icon.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='library', \n",
    "    title='Library', \n",
    "    subtitle='Choose an Enrichr gene set library or upload your own (.gmt). ',\n",
    "    img='library_icon.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='result', \n",
    "    title='Result Options', \n",
    "    subtitle='Choose how to get the top gene sets and whether you want co-expressed genes. NOTE: co-expressed genes will only be retrieved for certain Enrichr libraries and are not available for your own library files.',\n",
    "    img='results_icon.png'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55eb9865",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide\n",
    "{% set data_tab = TabField(\n",
    "    name='data_tab',\n",
    "    label='Submit Your Data',\n",
    "    default='Option 1',\n",
    "    description='',\n",
    "    choices={\n",
    "        'Option 1': [FileField(\n",
    "                name='ranked_filename',\n",
    "                label='Preranked Gene List',\n",
    "                default='logFC_ranked_GSE70466.rnk',\n",
    "                examples={'logFC_ranked_GSE70466.rnk': 'https://appyters.maayanlab.cloud/storage/GSEA_Appyter/logFC_ranked_GSE70466.rnk'},\n",
    "                description='RNK file containing preranked genelist.',\n",
    "                required=True,\n",
    "                section='data',\n",
    "            )],\n",
    "        'Option 2': [FileField(\n",
    "                name='expression_filename',\n",
    "                label='Expression Dataset',\n",
    "                description='GCT file containing sample expression dataset.',\n",
    "                default='GSE70466.gct',\n",
    "                examples={\n",
    "        'GSE70466.gct': 'https://appyters.maayanlab.cloud/storage/GSEA_Appyter/GSE70466.gct'},\n",
    "                required=True,\n",
    "                section='data'),\n",
    "            FileField(\n",
    "                name='phenotype_filename',\n",
    "                label='Phenotype Labels',\n",
    "                description='CLS file containing sample expression dataset.',\n",
    "                default='GSE70466.cls',\n",
    "                examples={\n",
    "        'GSE70466.cls': 'https://appyters.maayanlab.cloud/storage/GSEA_Appyter/GSE70466.cls'},\n",
    "                required=True,\n",
    "                section='data'),\n",
    "            ChoiceField(\n",
    "                name='ranking_method',\n",
    "                label='Method for Ranking Genes',\n",
    "                choices={'Characteristic Direction': 'cd',  \n",
    "                'Difference of Class Means': 'diff_of_classes',\n",
    "                'Limma': 'limma',\n",
    "                'Log2 Ratio of Class Means': 'log2_ratio_of_classes',\n",
    "                'Ratio of Class Means (Fold Change)': 'ratio_of_classes',\n",
    "                'Signal-to-Noise': 'signal_to_noise', \n",
    "                'T-test': 't_test'},\n",
    "                default='Characteristic Direction',\n",
    "                description='The method used to calculate a correlation or ranking.',\n",
    "                section='data')]\n",
    "            },\n",
    "        section='data')\n",
    "    %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4c59f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide\n",
    "{% set library_tab = TabField(\n",
    "    name='library_tab',\n",
    "    label='Submit Your Library',\n",
    "    default='Upload',\n",
    "    description='',\n",
    "    choices={'Upload':[FileField(\n",
    "            name='geneset_filename',\n",
    "            label='Library',\n",
    "            description='GMT file containing sample gene set.',\n",
    "            default='Data_Acquisition_Method.gmt',\n",
    "            required=True,\n",
    "            examples={'Data_Acquisition_Method.gmt': 'https://appyters.maayanlab.cloud/storage/GSEA_Appyter/Data_Acquisition_Method_Most_Popular_Genes.gmt'}\n",
    "            )],\n",
    "        'Select an Enrichr Library': [ChoiceField(\n",
    "            name='enrichr_choice',\n",
    "            label='Enrichr Library',\n",
    "            default='MSigDB_Hallmark_2020',\n",
    "            description='',\n",
    "            choices=['ChEA_2016',\n",
    "            'ARCHS4_TFs_Coexp',\n",
    "            'COVID-19_Related_Gene_Sets', \n",
    "            'ENCODE_TF_ChIP-seq_2015', \n",
    "            'KEGG_2021_Human', \n",
    "            'WikiPathway_2021_Human', \n",
    "            'Reactome_2016', \n",
    "            'GO_Biological_Process_2021', \n",
    "            'GO_Cellular_Component_2021', \n",
    "            'GO_Molecular_Function_2021', \n",
    "            'MGI_Mammalian_Phenotype_Level_4_2021', \n",
    "            'MSigDB_Hallmark_2020',\n",
    "            'Human_Phenotype_Ontology', \n",
    "            'GWAS_Catalog_2019', \n",
    "            'UK_Biobank_GWAS_v1', \n",
    "            'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "            'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019', \n",
    "            'DisGeNET', \n",
    "            'CellMarker_Augmented_2021',  \n",
    "            'HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression', \n",
    "            'Allen_Brain_Atlas_10x_scRNA_2021', \n",
    "            'Pfam_InterPro_Domains'])]\n",
    "        },\n",
    "    section='library')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56e72733",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide\n",
    "{% set result_number = IntField(\n",
    "    name='result_number',\n",
    "    label='Number of Top Gene Sets',\n",
    "    min=1,\n",
    "    max=50,\n",
    "    default=5,\n",
    "    description='The number of gene sets that will be displayed in the enrichment results table and plots.',\n",
    "    section='result')\n",
    "%}\n",
    "{% set result_criteria = ChoiceField(\n",
    "    name='result_criteria',\n",
    "    label='Criteria for Top Gene Sets',\n",
    "    choices={'Enrichment Score': 'es', 'Normalized Enrichment Score': 'nes', 'P-Value': 'pval', 'False Discovery Rate': 'fdr'},\n",
    "    default='P-Value',\n",
    "    description='The top gene sets will be calculated with the chosen criteria.',\n",
    "    section='result')\n",
    "%}\n",
    "{% set augmentation = BoolField(\n",
    "    name='augmentation', \n",
    "    label='Apply Augmented GSEA?', \n",
    "    default='false',\n",
    "    description='Choose whether you want to get co-expressed genes.',\n",
    "    section='result')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dad89d10",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "result_number = {{ result_number.value }}\n",
    "result_criteria = '{{ result_criteria.value }}'\n",
    "augmentation = {{augmentation.value}}\n",
    "\n",
    "{%- if library_tab.raw_value == 'Upload'%}\n",
    "library_tab = 'Upload'\n",
    "library_filename = {{ library_tab.value[0] }}\n",
    "library_name = library_filename.replace('_', ' ').replace('.gmt', '')\n",
    "\n",
    "{%- else %}\n",
    "library_tab = 'Select an Enrichr Library'\n",
    "library_filename = '{{ library_tab.value[0] }}'\n",
    "library_name = '{{ library_tab.value[0] }}'\n",
    "{%- endif %}\n",
    "\n",
    "{%- if data_tab.raw_value == 'Option 1'%}\n",
    "data_tab = 'Option 1'\n",
    "ranked_filename = {{ data_tab.value[0] }}\n",
    "\n",
    "{%- else %}\n",
    "data_tab = 'Option 2'\n",
    "expression_filename = {{ data_tab.value[0] }}\n",
    "phenotype_filename = {{ data_tab.value[1] }}\n",
    "ranking_method = '{{ data_tab.value[2] }}'\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7de89db",
   "metadata": {},
   "source": [
    "## Loading Data\n",
    "\n",
    "Loading the data may take a while... Please be patient. It may take up to five minutes for the data to load, for inputs similar in size to the default examples and input choices. Once the data is done loading, a message will be printed underneath this section. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "412b5ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#checks if inputs are valid\n",
    "def checkInputs():\n",
    "    if data_tab == 'Option 1':\n",
    "        if library_filename == '' or ranked_filename == '':\n",
    "            raise Exception('Please submit necessary materials for Option 1 to continue.')\n",
    "        if ranked_filename.endswith('.rnk') == False:\n",
    "            raise Exception('Please upload a RNK file (ends in .rnk).')\n",
    "    else:\n",
    "        if (library_filename == '' or expression_filename == '' or\n",
    "        phenotype_filename == ''):\n",
    "            raise Exception('Please submit necessary materials for Option 2 to continue.')\n",
    "        if expression_filename.endswith('.gct') == False:\n",
    "            raise Exception('Please upload a GCT file (ends in .gct).')\n",
    "        if phenotype_filename.endswith('.cls') == False:\n",
    "            raise Exception('Please upload a CLS file (ends in .cls).')\n",
    "    if (library_tab == 'Upload' and library_filename.endswith('.gmt') \n",
    "    == False):\n",
    "        raise Exception('Please upload a GMT file (ends in .gmt).')\n",
    "\n",
    "#performs GSEA\n",
    "def gsea():\n",
    "    results = gp.prerank(rnk=ranked, gene_sets=library_filename, max_size=500)\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beb3a14b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#loads a preranked list and converts it to a dataframe for Option 1\n",
    "#creates a ranked list with chosen ranking method and converts to dataframe for Option 2\n",
    "def loadRanked():\n",
    "    if data_tab == 'Option 1':\n",
    "        values = pd.read_csv(ranked_filename, header=None, sep=\"\\t\")\n",
    "        if len(values.index) < 5000:\n",
    "            raise Exception('Current ranked gene list has less than 5000 genes. We expect all human coding genes to be in the list, which is around 20000.')\n",
    "        values.sort_values(by=values.columns[1], ascending=False, inplace=True)\n",
    "        values.reset_index(drop=True, inplace=True)\n",
    "        values.columns = ['Gene', 'Rank']\n",
    "        pos, neg = None, None\n",
    "    elif data_tab == 'Option 2':\n",
    "        neg, pos, classes = gp.parser.gsea_cls_parser(phenotype_filename)\n",
    "        exp = pd.read_csv(expression_filename, skiprows=1, header=None, index_col=0, \n",
    "            sep=\"\\t\")\n",
    "        obj = GSEA(data=exp, gene_sets=library_filename, classes=classes)\n",
    "        exp_df, classesDict = obj.load_data(classes)\n",
    "        if ranking_method == 'cd':\n",
    "            data = exp_df.to_numpy()\n",
    "            sampleclass = [1 if s == neg else 2 if s == pos else 0 for s in classes]\n",
    "            genes = exp_df.index.tolist()\n",
    "            chdir_res = chdir(data=data, sampleclass=sampleclass,\n",
    "                genes=genes)\n",
    "            chdir_res = sorted(chdir_res, key=lambda x: x[0], reverse=True)\n",
    "            chdir_res = {k:v for v,k in chdir_res}\n",
    "            values = pd.Series(data=chdir_res)\n",
    "        elif ranking_method == 'limma':\n",
    "            neg_columns = [i for i in range(len(classes)) if classes[i] == neg]\n",
    "            neg_df = exp_df.iloc[:,neg_columns]\n",
    "            pos_columns = [i for i in range(len(classes)) if classes[i] == pos]\n",
    "            pos_df = exp_df.iloc[:,pos_columns]\n",
    "            limma_res = limma_voom.limma_voom_differential_expression(\n",
    "                controls_mat=neg_df, cases_mat=pos_df, all_data_mat=exp_df)\n",
    "            values = pd.Series(data=limma_res['logFC'], index=limma_res.index)\n",
    "            values.sort_values(axis=0, ascending=False, inplace=True)\n",
    "        else:\n",
    "            values = gp.algorithm.ranking_metric(df=exp_df, pos=pos, neg=neg, \n",
    "            method=ranking_method, classes=classesDict, ascending=False)\n",
    "        values = pd.DataFrame({'Gene':values.index, 'Rank':values.values})\n",
    "    return values, pos, neg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#loads libraries\n",
    "def downloadLibrary(name):\n",
    "    urllib.request.urlretrieve(f'https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName={name}', filename=f'{name}')\n",
    "    if library_tab != 'Upload' and augmentation == True:\n",
    "        urllib.request.urlretrieve(f'https://appyters.maayanlab.cloud/storage/GSEA_Appyter/{name}_augmented.gmt', filename=f'{name}_augmented')\n",
    "\n",
    "def getLibraryData(library):\n",
    "    data = dict()\n",
    "    with open(library, 'r') as f:\n",
    "        for line in f:\n",
    "            splited = line.strip().split(\"\\t\")\n",
    "            elements = splited[2:]\n",
    "            if len(elements) > 0:\n",
    "                # to upper case\n",
    "                data[splited[0]] = [x.upper() for x in elements]\n",
    "    return data\n",
    "\n",
    "def loadLibrary(library):\n",
    "    if library_tab == 'Select an Enrichr Library':\n",
    "        downloadLibrary(library)\n",
    "    library_data = getLibraryData(library)\n",
    "    if len(library_data) < result_number:\n",
    "        raise Exception('There cannot be less gene sets in the library than the number of gene sets displayed for the results.')\n",
    "    if library_tab != 'Upload' and augmentation == True:\n",
    "        augmented_data = getLibraryData(f'{library}_augmented')\n",
    "    else:\n",
    "        augmented_data = None\n",
    "    return library_data, augmented_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f57239d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter_code_exec\n",
    "#loading all needed data\n",
    "print('Checking inputs...')\n",
    "checkInputs()\n",
    "print('Loading data for plots...')\n",
    "ranked, pheno1, pheno2 = loadRanked()\n",
    "library_data, augmented_data = loadLibrary(library_filename)\n",
    "print('Performing GSEA...')\n",
    "result = gsea().results\n",
    "print('Data loaded successfully!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acc3830b",
   "metadata": {},
   "source": [
    "## Enrichment Results Table\n",
    "\n",
    "The following table displays the enrichment analysis result statistics of the top ranked most enriched annotated gene sets with the chosen criteria. The columns can be re-arranged by dragging.  \n",
    "\n",
    "The table is available for download as a PNG file by clicking the camera icon on the upper right. The CSV file containing the full version of the results sorted by the input criteria is available for download by clicking the link below the table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f3fd466",
   "metadata": {},
   "outputs": [],
   "source": [
    "#makes values appropriate for table viewing\n",
    "def validateNumber(num):\n",
    "    num = round(num, 4)\n",
    "    if abs(num) <= 0.001:\n",
    "        num = '<0.001'\n",
    "    elif abs(num) <= 0.01:\n",
    "        num = '{:.2E}'.format(num)\n",
    "    return num\n",
    "\n",
    "def geneStringToList(string):\n",
    "    return string.split(';')\n",
    "\n",
    "#create a downloadable csv file of all the results\n",
    "def create_download_link(df, title = \"Download CSV file of all results\", filename = \"GSEA_Enrichment_Results_data.csv\"):  \n",
    "    csv = df.to_csv(filename, index = True)\n",
    "    html = f'<a href=\"{filename}\" target=_blank>{title}</a>'\n",
    "    return HTML(f\"<div style='font-size:1.25rem;'>{html}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61409c9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#organizes gsea results into sorted dataframe (for download)\n",
    "result_df = pd.DataFrame.from_dict(result, orient='index')\n",
    "result_df.index.name = 'gene set'\n",
    "if result_criteria == 'pval':\n",
    "    sorted_result_df = result_df.sort_values(by=['pval', 'fdr'], ascending=True)\n",
    "elif result_criteria == 'fdr':\n",
    "     sorted_result_df = result_df.sort_values(by=['fdr', 'pval'], ascending=True)\n",
    "else:\n",
    "    sorted_result_df = result_df.sort_values(by=result_criteria, ascending=False)\n",
    "sorted_result_df = sorted_result_df.drop(['RES', 'hit_indices'], axis=1)\n",
    "sorted_result_df['genes'] = sorted_result_df['genes'].map(geneStringToList)\n",
    "sorted_result_df['ledge_genes'] = sorted_result_df['ledge_genes'].map(geneStringToList)\n",
    "\n",
    "#gets top gene sets' data (calculated statistics only) needed for the table display\n",
    "top_result_df = sorted_result_df.copy(deep=True)\n",
    "top_result_df.reset_index(inplace=True)\n",
    "top_result_df = top_result_df.truncate(before=0, after=result_number-1)\n",
    "top_result_df.set_index(keys='gene set', inplace=True)\n",
    "top_result_df = top_result_df.drop(['genes', 'ledge_genes'], axis=1)\n",
    "top_result_df = top_result_df.applymap(validateNumber)\n",
    "\n",
    "#converts necessary dataframes to list\n",
    "rankedGenes = ranked['Gene'].tolist()\n",
    "rankedCorrelation = ranked['Rank'].tolist()\n",
    "top = top_result_df.index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91f2d382",
   "metadata": {},
   "outputs": [],
   "source": [
    "#displays this section\n",
    "fig1 = go.Figure(data=[go.Table(columnwidth=[200, 75, 75, 75, 75, 75, 75],\n",
    "    header=dict(values=['Gene Set', 'ES', 'NES', 'P-value', 'FDR', 'Set Size', 'Genes Matched'],\n",
    "    height=40), \n",
    "    cells=dict(values=[top_result_df.index, top_result_df.es, top_result_df.nes, \n",
    "    top_result_df.pval, top_result_df.fdr, top_result_df.geneset_size, \n",
    "    top_result_df.matched_size], height=30))])\n",
    "fig1.update_layout(width=900, font_size=14)\n",
    "fig1.show()\n",
    "display(create_download_link(sorted_result_df))\n",
    "display(HTML(f\"<i><b>Table 1: Enrichment Results Table for the top {len(top)} gene sets in the {library_name} library. The Enrichment Score (ES) is the random walk's farthest deviation from 0. NES is the Normalized Enrichment Score, the P-value is a measure of significance of the NES, and the FDR is the False Discovery Rate corrected p-value. Set Size is the number of genes in the set, and Genes Matched is how many genes in the gene set are also in the ranked list of genes.</i></b>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8877daf8",
   "metadata": {},
   "source": [
    "## Generating GSEA Plots\n",
    "\n",
    "For each gene set library, the top gene sets with the top enrichment scores are plotted. You can choose which gene set to view from the dropdown menu. Note that the hit indices plot will only be present for individual gene sets. \n",
    "\n",
    "Various interactive options are available from the toolbar on the upper right. Lines can be toggled on and off by clicking the name in the legend. Hovering over the running sum line on the Enrichment Plot will display the gene, its ranking, and the running sum score at that point. Also be aware that zooming in and out of the GSEA plot will simultaneously zoom in or out of the Hit Indices plot. \n",
    "\n",
    "The GSEA plots are available for download as a PNG file by clicking the camera icon on the toolbar. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a6709e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#breaks strings that are too long into new lines\n",
    "def lineBreak(string):\n",
    "    words = string.split(\" \")\n",
    "    name = \"\"\n",
    "    check = \"\"\n",
    "    for i in range(len(words)):\n",
    "        check += words[i] + \" \"\n",
    "        if len(check) > 20 and i < len(words)-1:\n",
    "            name += check + \"<br>\"\n",
    "            check = \"\"\n",
    "    name += check\n",
    "    return name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "904e4361",
   "metadata": {},
   "outputs": [],
   "source": [
    "#initialize figure basics\n",
    "fig2 = make_subplots(rows=2, cols=1, row_heights = [0.8, 0.2],\n",
    "        shared_xaxes=True, vertical_spacing = 0.10,\n",
    "        subplot_titles=(\"Enrichment Plot\", \"Hit Indices\"))\n",
    "arbitrary = [dict(type=\"line\", xref=\"x2\", yref=\"y2\", x0=0, y0=0,\n",
    "            x1=len(rankedGenes), y1=0, line=dict(color=\"White\", width=0.5)), \n",
    "            dict(type=\"line\", xref=\"x2\", yref=\"y2\", x0=0, y0=-1,\n",
    "            x1=0, y1=1, line=dict(color=\"White\", width=0.5))]\n",
    "            #pair of arbitrary lines that maintain scale of plot\n",
    "fig2.update_layout(shapes=arbitrary, height=600, width=750, font_size=16)\n",
    "buttons = [dict(label='All Top Gene Sets', method=\"update\",\n",
    "            args=[{\"visible\": [True for i in range(len(top))]}, {\"shapes\": arbitrary}])]\n",
    "\n",
    "xvals = [i for i in range(len(rankedGenes))]\n",
    "for geneSet in top:\n",
    "#plots values for each top gene set \n",
    "    #gets hit indices and running sum vector\n",
    "    hits, yvals = result_df.at[geneSet, 'hit_indices'], result_df.at[geneSet, 'RES']\n",
    "    index = top.index(geneSet)\n",
    "    pval = top_result_df.at[geneSet, 'pval']\n",
    "    #creates enrichment plot\n",
    "    setName = lineBreak(geneSet)\n",
    "    fig2.add_trace(go.Scatter(x=xvals, y=yvals, mode='lines', showlegend=True,\n",
    "        name=setName, meta=[pval],\n",
    "        text = ['Gene: {}'.format(gene) for gene in rankedGenes],\n",
    "        hovertemplate=('%{text}' + \n",
    "        '<br>Ranking: %{x}' + \n",
    "        '<br>Running Sum: %{y}' +\n",
    "        '<br>P-value: %{meta[0]}')),\n",
    "        row=1, col=1)\n",
    "    #creates hit indices\n",
    "    lines = []\n",
    "    for x in hits:\n",
    "        line = dict(type=\"line\", x0=x, y0=-1, x1=x, y1=1, xref = \"x2\", yref = \"y2\",\n",
    "        line=dict(color=\"Blue\", width=0.5))\n",
    "        lines.append(line)\n",
    "    #creates drop down menu option\n",
    "    visible = [True if i==index else False for i in range(len(top))]\n",
    "    button = dict(label=f'{geneSet}', method=\"update\",\n",
    "        args=[{\"visible\": visible}, {\"shapes\": lines}])\n",
    "    buttons.append(button)\n",
    "\n",
    "#make drop down menu\n",
    "fig2.update_layout(updatemenus=[\n",
    "    dict(buttons=buttons, pad = {\"r\": 10, \"t\": 10}, showactive = True,\n",
    "    x = 0, xanchor = \"left\", y = 1.2, yanchor = \"top\")])\n",
    "fig2.update_xaxes(title_text=\"Gene Rankings\", row=2, col=1)\n",
    "fig2.update_yaxes(title_text=\"Enrichment Score (ES)\", row=1, col=1)\n",
    "fig2.update_yaxes(visible=False, showticklabels=False, row=2, col=1)\n",
    "\n",
    "fig2.show()\n",
    "\n",
    "if pheno1 != None and pheno2 != None:\n",
    "    display(HTML(f'<b><i>Figure 1: GSEA plot for the top {len(top)} gene sets in the {library_name} library. Those with a positive Enrichment Score and more hits on the left are correlated with {pheno1}, while those with a negative Enrichment Score and more hits on the right are correlated with {pheno2}.</i></b>'))\n",
    "else:\n",
    "    display(HTML(f'<b><i>Figure 1: GSEA plot for the top {len(top)} gene sets in the {library_name} library. Those with a positive Enrichment Score and more hits on the left are correlated with genes at the top of the ranked list, while those with a negative Enrichment Score and more hits on the right are correlated with genes at the bottom.</i></b>'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab09bcd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{%- if library_tab.raw_value == 'Select an Enrichr Library' and augmentation.value == True %}\n",
    "## Highlighting Co-expressed Genes\n",
    "\n",
    "The augmented gene set library was retrieved using the GMT Augmentation Appyter (Clarke et al., 2021) and the ARCHS4 Co-expression matrix (Lachmann et al., 2018). The augmented gene sets are plotted against the original gene sets and the combination of the two sets below. The blue hit indices are hit indices from the original gene set and the red ones represent hits from the augmented gene set. \n",
    "\n",
    "The table below the plots displays each gene set and the list of the top ten co-expressed genes from the augmented set, which is derived from the plots.\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%- if library_tab.raw_value == 'Select an Enrichr Library' and augmentation.value == True %}\n",
    "#retrieves leading edge genes\n",
    "def ledgeGenes(results):\n",
    "    es, hit_indices, res = results\n",
    "    res = res.tolist()\n",
    "    rank = res.index(es)\n",
    "    ledge = []\n",
    "    hits = []\n",
    "    if es > 0:\n",
    "        i = 0\n",
    "        while rank >= hit_indices[i]:\n",
    "            ledge.append(rankedGenes[hit_indices[i]])\n",
    "            hits.append(hit_indices[i])\n",
    "            i += 1\n",
    "    else:\n",
    "        i = len(hit_indices)-1\n",
    "        while rank <= hit_indices[i]:\n",
    "            ledge.append(rankedGenes[hit_indices[i]])\n",
    "            hits.append(hit_indices[i])\n",
    "            i -= 1\n",
    "    if len(ledge) > 10:\n",
    "        ledge = ledge[:10]\n",
    "        hits = hits[:10]\n",
    "    return {'genes': ledge, 'ranks': hits}\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51bffb1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%- if library_tab.raw_value == 'Select an Enrichr Library' and augmentation.value == True %}\n",
    "fig3 = make_subplots(rows=2, cols=1, row_heights = [0.8, 0.2],\n",
    "        shared_xaxes=True, vertical_spacing = 0.10,\n",
    "        subplot_titles=(\"Enrichment Plot\", \"Hit Indices\"))\n",
    "fig3.update_layout(shapes=arbitrary, height=600, width=750, font_size=16)\n",
    "plot_buttons = [dict(label='All', method=\"update\",\n",
    "            args=[{\"visible\": [True for i in range(len(top*3))]}, {\"shapes\": arbitrary}])]\n",
    "fig4 = go.Figure()\n",
    "table_buttons = list()\n",
    "\n",
    "for geneSet in top:\n",
    "    #gets values to plot\n",
    "    hits0, yvals0 = result_df.at[geneSet, 'hit_indices'], result_df.at[geneSet, 'RES']\n",
    "    augmented = gp.algorithm.enrichment_score(rankedGenes, rankedCorrelation, augmented_data[geneSet])\n",
    "    hits1, yvals1 = augmented[2], augmented[3]\n",
    "    both = gp.algorithm.enrichment_score(rankedGenes, rankedCorrelation, augmented_data[geneSet]+library_data[geneSet])\n",
    "    yvals2 = both[3]\n",
    "    yvals = [yvals0, yvals1, yvals2]\n",
    "    #gets top co-expressed genes\n",
    "    results = both[0], augmented[2], both[3]\n",
    "    coexp = ledgeGenes(results)\n",
    "    coexp_df = pd.DataFrame.from_dict(data=coexp, orient='columns')\n",
    "    fig4.add_trace(go.Table(columnwidth=[200, 200],\n",
    "    header=dict(values=['Genes', 'Rank'], height=40), \n",
    "    cells=dict(values=[coexp_df.genes, coexp_df.ranks], height=30)))\n",
    "    #plots the three random walks\n",
    "    for i in range(len(yvals)):\n",
    "        if i == 0:\n",
    "            newName = geneSet + ' original'\n",
    "        elif i == 1:\n",
    "            newName = geneSet + ' augmented'\n",
    "        elif i == 2:\n",
    "            newName = geneSet + ' both'\n",
    "        newName = lineBreak(newName)\n",
    "        fig3.add_trace(go.Scatter(x=xvals, y=yvals[i], mode='lines', showlegend=True,\n",
    "        name=newName, text = ['Gene: {}'.format(gene) for gene in rankedGenes],\n",
    "        hovertemplate=('%{text}' + \n",
    "        '<br>Ranking: %{x}' + \n",
    "        '<br>Running Sum: %{y}')),\n",
    "        row=1, col=1)\n",
    "    #creates hit indices\n",
    "    lines = []\n",
    "    for x in hits0:\n",
    "        line = dict(type='line', x0=x, y0=-1, x1=x, y1=1, xref = 'x2', yref = 'y2',\n",
    "        line=dict(color=\"Blue\", width=0.5))\n",
    "        lines.append(line)\n",
    "    for x in hits1:\n",
    "        line = dict(type='line', x0=x, y0=-1, x1=x, y1=1, xref = 'x2', yref = 'y2',\n",
    "        line=dict(color=\"Red\", width=0.5))\n",
    "        lines.append(line)\n",
    "    #creates drop down menu option\n",
    "    index = top.index(geneSet)\n",
    "    plot_visible = [True if i >= index*3 and i < (index+1)*3 else False for i in range(len(top*3))]\n",
    "    plot_button = dict(label=f'{geneSet}', method=\"update\",\n",
    "        args=[{\"visible\": plot_visible}, {\"shapes\": lines}])\n",
    "    plot_buttons.append(plot_button)\n",
    "    table_visible = [True if i == index else False for i in range(len(top))]\n",
    "    table_button = dict(label=f'{geneSet}', method=\"update\",\n",
    "        args=[{\"visible\": table_visible}])\n",
    "    table_buttons.append(table_button)\n",
    "\n",
    "fig3.update_layout(updatemenus=[\n",
    "    dict(buttons=plot_buttons, pad = {\"r\": 10, \"t\": 10}, showactive = True,\n",
    "    x = 0, xanchor = \"left\", y = 1.2, yanchor = \"top\")])\n",
    "fig3.update_xaxes(title_text=\"Gene Rankings\", row=2, col=1)\n",
    "fig3.update_yaxes(title_text=\"Enrichment Score (ES)\", row=1, col=1)\n",
    "fig3.update_yaxes(visible=False, showticklabels=False, row=2, col=1)\n",
    "fig3.show()\n",
    "display(HTML(f\"<i><b>Figure 2: Random walk plots of augmented gene sets for the top {len(top)} gene sets in the {library_name} library. Each gene set has three plots: one with the original set, one with the augmented genes only, and one with them combined.</i></b>\"))\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%- if library_tab.raw_value == 'Select an Enrichr Library' and augmentation.value == True %}\n",
    "#displays top co-expressed genes in table\n",
    "fig4.update_layout(updatemenus=[\n",
    "    dict(buttons=table_buttons, pad = {\"r\": 10, \"t\": 10}, showactive = True,\n",
    "    x = 0, xanchor = \"left\", y = 1.2, yanchor = \"top\")])\n",
    "fig4.update_layout(width=400, font_size=14)\n",
    "fig4.show()\n",
    "display(HTML(f\"<i><b>Table 2: The Top Co-Expressed Genes Table for the top {len(top)} gene sets in the {library_name} library. The table displays the top ten co-expressed genes for each set, as well as their ranking within the input signature.</i></b>\"))\n",
    "{%- endif %}"
   ]
  },
  {
   "source": [
    "## References\n",
    "Blanco-Melo D, Nilsson-Payant BE, Liu WC, Uhl S et al. Imbalanced Host Response to SARS-CoV-2 Drives Development of COVID-19. Cell 2020 May 28;181(5):1036-1045.e9. PMID: 32416070. \n",
    "\n",
    "Clark, N.R., Hu, K.S., Feldmann, A.S. et al. The characteristic direction: a geometrical approach to identify differentially expressed genes. BMC Bioinformatics 15, 79 (2014). https://doi.org/10.1186/1471-2105-15-79 \n",
    "\n",
    "Clarke, D. J. B., Jeon, M., Stein, D. J., Moiseyev, N., Kropiwnicki, E., Dai, C., Xie, Z., Wojciechowicz, M. L., Litz, S., Hom, J., Evangelista, J. E., Goldman, L., Zhang, S., Yoon, C., Ahamed, T., Bhuiyan, S., Cheng, M., Karam, J., Jagodnik, K. M., … Ma'ayan, A. (2021). Appyters: Turning Jupyter notebooks into data-driven web apps. Patterns, 2(3), 100213. https://doi.org/10.1016/j.patter.2021.100213. \n",
    "\n",
    "Daamen AR, Bachali P, Owen KA, Kingsmore KM et al. Comprehensive transcriptomic analysis of COVID-19 blood, lung, and airway. Sci Rep 2021 Mar 29;11(1):7052. PMID: 33782412. \n",
    "\n",
    "Kuleshov, M. V., Jones, M. R., Rouillard, A. D., Fernandez, N. F., Duan, Q., Wang, Z., Koplev, S., Jenkins, S. L., Jagodnik, K. M., Lachmann, A., McDermott, M. G., Monteiro, C. D., Gundersen, G. W., & Ma'ayan, A. (2016). Enrichr: A Comprehensive Gene set enrichment analysis web Server 2016 update. Nucleic Acids Research, 44(W1). https://doi.org/10.1093/nar/gkw377. \n",
    "\n",
    "Lachmann, A., Torre, D., Keenan, A.B. et al. Massive mining of publicly available RNA-seq data from human and mouse. Nat Commun 9, 1366 (2018). https://doi.org/10.1038/s41467-018-03751-6. \n",
    "\n",
    "Law, CW, Chen, Y, Shi, W, and Smyth, GK (2014). Voom: precision weights unlock linear model analysis tools for RNA-seq read counts. Genome Biology 15, R29. \n",
    "\n",
    "Subramanian, A., Tamayo, P., Mootha, V. K., Mukherjee, S., Ebert, B. L., Gillette, M. A., Paulovich, A., Pomeroy, S. L., Golub, T. R., Lander, E. S., & Mesirov, J. P. (2005). Gene set enrichment analysis: A knowledge-based approach for interpreting genome-wide expression profiles. Proceedings of the National Academy of Sciences, 102(43), 15545–15550. https://doi.org/10.1073/pnas.0506580102 "
   ],
   "cell_type": "markdown",
   "metadata": {}
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "GSEA",
   "language": "python",
   "name": "gsea"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}