{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# scRNA-seq Analysis Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This pipeline enables you to analyze and visualize your single cell RNA sequencing datasets with an array of algorithms and data visualization methods. The pipeline includes quality control, library size analysis, normalization, plotting most highly expressed genes, plotting samples, MAGIC normalization, clustering, enrichment analysis, trajectory inference, and cell type prediction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic libraries\n",
    "import pandas as pd\n",
    "import random\n",
    "import time\n",
    "import numpy as np\n",
    "import warnings\n",
    "\n",
    "# Visualization\n",
    "import seaborn as sns\n",
    "import scipy.stats as ss\n",
    "import plotly\n",
    "from plotly import tools\n",
    "import plotly.express as px\n",
    "import plotly.graph_objs as go\n",
    "import matplotlib.pyplot as plt; plt.rcdefaults()\n",
    "from matplotlib import rcParams\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib_venn import venn2, venn3\n",
    "import IPython\n",
    "from IPython.display import HTML, display, Markdown, IFrame, FileLink, Image\n",
    "from itertools import combinations\n",
    "from scipy import stats\n",
    "import chart_studio\n",
    "import chart_studio.plotly as py\n",
    "from PIL import Image\n",
    "\n",
    "# Data analysis\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import quantile_transform\n",
    "from sklearn import cluster\n",
    "from sklearn.metrics import silhouette_score\n",
    "from sklearn.manifold import TSNE\n",
    "import umap\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "from magic import MAGIC\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import DigitalCellSorter\n",
    "from maayanlab_bioinformatics.enrichment.crisp import enrich_crisp, fisher_overlap\n",
    "\n",
    "# Bokeh\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool, CustomJS, ColumnDataSource, Span, Select, Legend, PreText, Paragraph, LinearColorMapper, ColorBar, CategoricalColorMapper\n",
    "from bokeh.layouts import layout, row, column, gridplot\n",
    "from bokeh.palettes import all_palettes\n",
    "import colorcet as cc\n",
    "from bokeh.palettes import Category20\n",
    "\n",
    "# External Code\n",
    "from utils import *\n",
    "from plotly.offline import init_notebook_mode\n",
    "init_notebook_mode(connected = False)\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% do SectionField(\n",
    "    name='Data_Section',\n",
    "    title='Load your Data',\n",
    "    subtitle='Load your expression data. In comma/tab separated formats, genes should be in rows and samples shoud be in columns. You can also upload Cell Ranger files (matrix, genes, and barcords files). ',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Normalization_Section',\n",
    "    title='Select Normalization Methods',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Visualization_Section',\n",
    "    title='Select Visualization Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='DEG_Section',\n",
    "    title='Select Differentially Exprssed Gene Analysis Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='TI_Section',\n",
    "    title='Select Trajectory Inference Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Time_TI_Section',\n",
    "    title='Select Time-series Trajectory Inference Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='CT_Section',\n",
    "    title='Select Cell Type Prediction Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set file_kind = TabField(\n",
    "    name='file_kind',\n",
    "    label='Data file',\n",
    "    default='Plain text',\n",
    "    description='Upload your expression files',\n",
    "    choices={\n",
    "        'Plain text': [             \n",
    "            FileField(\n",
    "                name='rnaseq_data_filename', \n",
    "                label='RNA-seq data file (.csv, .txt or .tsv)', \n",
    "                default='GSE110499_example_expression.txt',\n",
    "                examples={'GSE110499_example_expression.txt': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/GSE110499-expression.txt'}, \n",
    "                description='Upload RNA-seq expression data as comma seperated or tab seperated format. The index of the dataset are genes, the columns are samples.', \n",
    "                section='Data_Section'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='meta_data_filename', \n",
    "                label='(Optional) Meta data file (.csv, .txt or .tsv)', \n",
    "                default='GSE110499_example_metadata.txt',\n",
    "                examples={'GSE110499_example_metadata.txt': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/GSE110499-metadata.txt'}, \n",
    "                description='Upload metadata as two-column comma seperated or tab seperated format. The first column contains sample IDs and the other column contains sample labels',\n",
    "                section='Data_Section'\n",
    "            ),         \n",
    "            StringField(\n",
    "                name='meta_class_column_name', \n",
    "                label='(Optional) Class column name in metadata', \n",
    "                default='prep-site', \n",
    "                description='class column name of metadata', \n",
    "                section='Data_Section'\n",
    "            )\n",
    "\n",
    "\n",
    "        ],\n",
    "        \n",
    "        '.mtx from 10x Genomics': [\n",
    "            FileField(\n",
    "                name='mtx_data_filename', \n",
    "                label='RNA-seq data file (.mtx)', \n",
    "                default='example_matrix.mtx',\n",
    "                examples={'example_matrix.mtx': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/example_matrix.mtx'}, \n",
    "                description='Expression data file from 10x Genomics need to be in .mtx format which store the expression data in sparse matrix.',\n",
    "                section='Data_Section'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='gene_data_filename', \n",
    "                label='Gene infomation file (.tsv)', \n",
    "                default='example_genes.tsv',\n",
    "                examples={'example_genes.tsv': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/example_genes.tsv'}, \n",
    "                description='A tab delimited file of the corresponding genes in the .mtx expression matrix file. The first column should be the gene IDs whereas the secound column corresponds to gene symbols.', \n",
    "                section='Data_Section'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='barcode_data_filename', \n",
    "                label='Barcode information file (.tsv)', \n",
    "                default='example_barcodes.tsv',\n",
    "                examples={'example_barcodes.tsv': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/example_barcodes.tsv'}, \n",
    "                description='A tab delimited file of the corresponding barcodes in the .mtx expression matrix file. The first column should be the unique barcodes for cells.', \n",
    "                section='Data_Section'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='meta_data_filename_mtx', \n",
    "                label='(Optional) Meta data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                examples={'example_metadata.tsv': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/example_metadata.tsv'}, \n",
    "                description='Upload metadata as two-column comma seperated or tab seperated format. The first column contains sample IDs and the other column contains sample labels',\n",
    "                section='Data_Section'\n",
    "            ),         \n",
    "            StringField(\n",
    "                name='meta_class_column_name_mtx', \n",
    "                label='(Optional) Class column name in metadata', \n",
    "                default='', \n",
    "                description='class column name of metadata', \n",
    "                section='Data_Section'\n",
    "            )\n",
    "        ]\n",
    "        \n",
    "    },\n",
    "    section = 'Data_Section',\n",
    ") %}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set qc_filter_genes = BoolField(\n",
    "    name='qc_filter_genes', \n",
    "    label='Filter genes by quality control?', \n",
    "    default='true', \n",
    "    description='Check if you want cells to be filtered by quality control', \n",
    "    section='Normalization_Section')\n",
    "%}\n",
    "\n",
    "{% set qc_threshold = FloatField(\n",
    "    name='qc_threshold', \n",
    "    label='Mitochondria Quality Control threshold', \n",
    "    default='0.05', \n",
    "    description='Remove cells that have too many mitochondrial genes expressed.', \n",
    "    section='Normalization_Section')\n",
    "%} \n",
    "\n",
    "{% set log_normalization = BoolField(\n",
    "    name='log_normalization', \n",
    "    label='Log normalization?', \n",
    "    default='true', \n",
    "    description='Check if you want the dataset to be log-transformed', \n",
    "    section='Normalization_Section')\n",
    "%}\n",
    "{% set normalization_method = ChoiceField(\n",
    "    name='normalization_method',\n",
    "    label='Normalization',\n",
    "    choices={'None': 'None', 'Satija et al. (2015, Nature Biotechnology)': 'Seurat', 'Zheng et al. (2017, Nature Communications)': 'Zheng17','Weinreb et al. (2018, PNAS)': 'Weignreb17'},\n",
    "    default='Satija et al. (2015, Nature Biotechnology)', \n",
    "    description='Standard normlization recipe for scRNA-seq datasets', \n",
    "    section='Normalization_Section')\n",
    "%}\n",
    "{% set magic = BoolField(\n",
    "    name='magic', \n",
    "    label='MAGIC?', \n",
    "    default='true', \n",
    "    description='Check if you want impute missing data with Markov Affinity-based Graph Imputation of Cells (MAGIC)', \n",
    "    section='Normalization_Section')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set highest_expr_n_genes = IntField(\n",
    "    name='highest_expr_n_genes', \n",
    "    label='Genes for highest expression', \n",
    "    min=0, \n",
    "    max=100, \n",
    "    default=10, \n",
    "    description='The number of genes with highest expression values', \n",
    "    section='Visualization_Section')\n",
    "%}\n",
    "\n",
    "{% set nr_genes = IntField(\n",
    "    name='nr_genes', \n",
    "    label='Genes for dimension reduction', \n",
    "    min=0, \n",
    "    max=30000, \n",
    "    default=500, \n",
    "    description='The maximum number of genes for dimension reduction analysis', \n",
    "    section='Visualization_Section')\n",
    "%}\n",
    "gene_list_for_clustergrammer = {{TextField(\n",
    "    name='gene_list_for_clustergrammer', \n",
    "    label='Gene List for Clustergrammer (Optional)', \n",
    "    default='', \n",
    "    description='Paste your gene list (One gene per row) for Clustergrammer heatmap plots.', \n",
    "    section = 'Visualization_Section')}}\n",
    "{% set clustering_topk = IntField(\n",
    "    name='clustering_topk', \n",
    "    label='Genes for clustergrammer', \n",
    "    min=0, \n",
    "    max=1000, \n",
    "    default=800, \n",
    "    description='The number of genes with largest variance for Clustergrammer', \n",
    "    section='Visualization_Section')\n",
    "%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set deg = BoolField(\n",
    "    name='deg',\n",
    "    label='Differentially expressed gene analysis?',\n",
    "    default='true',\n",
    "    description='Check if you want differentially expressed gene analysis', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "{% set bar_chart = BoolField(\n",
    "    name='bar_chart', \n",
    "    label='Bar chart?', \n",
    "    default='false', \n",
    "    description='Check if you want to see bar charts. It may take a while.', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "{% set enrichment_groupby = ChoiceField(\n",
    "    name='enrichment_groupby',\n",
    "    label='Group for differentially expressed gene analysis',\n",
    "    choices={'Cluster': 'Cluster', 'User-defined class': 'user_defined_class'},\n",
    "    default='User-defined class', \n",
    "    description='Specify groups for enrichment analysis. Clusters will be automatically generated by a clustering method', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "{% set diff_gex_method = ChoiceField(\n",
    "    name='diff_gex_method',\n",
    "    label='Differential expression analysis method',\n",
    "    choices={'limma': 'limma', 'wilcoxon': 'wilcoxon', 'characteristic direction': 'characteristic_direction', 'edgeR': 'edgeR', 'DESeq2': 'DESeq2'},\n",
    "    default='limma', \n",
    "    description='Set a method to get differentially expressed genes', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "\n",
    "\n",
    "{% set gene_topk = IntField(\n",
    "    name='gene_topk', \n",
    "    label='Maximum genes for Enrichr', \n",
    "    min=0, \n",
    "    max=1000, \n",
    "    default=200, \n",
    "    description='The maximum number of genes discovered by the DEG method', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "\n",
    "{% set enrichr_libraries = MultiChoiceField(\n",
    "    name='enrichr_libraries',\n",
    "    label='Enrichr Libraries (upto 2)',\n",
    "    descriptions='Enrichr libraries to be visualized. Select one or two libraries',\n",
    "    choices=['Gene Ontology',\n",
    "            'Pathway',\n",
    "            'Kinase',\n",
    "            'Transcription Factor',\n",
    "            'miRNA',\n",
    "            'Cell Type',\n",
    "             'Disease'],\n",
    "    default=['Gene Ontology'],\n",
    "    section='DEG_Section'\n",
    "    )\n",
    "%}\n",
    "\n",
    "{% set nr_genesets = IntField(\n",
    "    name='nr_genesets', \n",
    "    label='Top ranked gene sets', \n",
    "    min=0, \n",
    "    max=100,\n",
    "    default=15, \n",
    "    description='the number of result gene sets', \n",
    "    section='DEG_Section')\n",
    "%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set trajectory = BoolField(\n",
    "    name='trajectory',\n",
    "    label='Trajectory Inference?',\n",
    "    default='true',\n",
    "    description='Check if you want trajectory inference analysis', \n",
    "    section='TI_Section')\n",
    "%}\n",
    "\n",
    "{% set trajectory_method = ChoiceField(\n",
    "    name='trajectory_method',\n",
    "    label='Trajectory inference method',\n",
    "    choices={'DPT(diffusion pseudotime)': 'dpt', 'monocle': 'monocle'},\n",
    "    default='DPT(diffusion pseudotime)', \n",
    "    description='Trajectory inference algorithm', \n",
    "    section='TI_Section')\n",
    "%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set time_series_trajectory = BoolField(\n",
    "    name='time_series_trajectory', \n",
    "    label='Time series trajectory analysis?', \n",
    "    default='false', \n",
    "    description='Check if you want time-series trajectory analysis', \n",
    "    section='Time_TI_Section')\n",
    "%}\n",
    "{% set timepoint_labels_column_name = StringField(\n",
    "    name='timepoint_labels_column_name', \n",
    "    label='Timepoint column name in metadata', \n",
    "    default='Timepoints', \n",
    "    description='Timepoint column name of metadata', \n",
    "    section='Time_TI_Section')\n",
    "%}\n",
    "timepoint_labels = {{ TextField(\n",
    "    name='timepoint_labels', \n",
    "    label='Ordered timepoint labels in the timepoint label column', \n",
    "    default='', \n",
    "    description='Paste your timepoint labels in order (One timepoint per row). e.g., 0H, 12H, 24H', \n",
    "    section = 'Time_TI_Section') }}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set cell_type_prediction = BoolField(\n",
    "    name='cell_type_prediction', \n",
    "    label='Cell type prediction?', \n",
    "    default='false', \n",
    "    description='Check if you want cell type prediction', \n",
    "    section='CT_Section')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if file_kind.raw_value == 'Plain text' %}\n",
    "rnaseq_data_filename = {{ file_kind.value[0] }}\n",
    "meta_data_filename = {{ file_kind.value[1] }}\n",
    "meta_class_column_name = {{ file_kind.value[2] }}\n",
    "\n",
    "{%- elif file_kind.raw_value == '.mtx from 10x Genomics' %}\n",
    "mtx_data_filename = {{ file_kind.value[0] }}\n",
    "gene_data_filename = {{ file_kind.value[1] }}\n",
    "barcode_data_filename = {{ file_kind.value[2] }}\n",
    "meta_data_filename = {{ file_kind.value[3] }}\n",
    "meta_class_column_name = {{ file_kind.value[4] }}\n",
    "\n",
    "{%- elif file_kind.raw_value == '.h5 from 10x Genomics' %}\n",
    "h5_data_filename = {{ file_kind.value[0] }}\n",
    "key = {{ file_kind.value[1] }}\n",
    "{%- endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "qc_filter_genes = {{qc_filter_genes.value}}\n",
    "qc_threshold = {{qc_threshold.value}}\n",
    "log_normalization = {{log_normalization.value}}\n",
    "normalization_method = \"{{normalization_method.value}}\"\n",
    "magic = {{magic.value}}\n",
    "\n",
    "highest_expr_n_genes = {{highest_expr_n_genes.value}}\n",
    "nr_genes = {{nr_genes.value}}\n",
    "clustering_topk = {{clustering_topk.value}}\n",
    "\n",
    "deg = {{deg.value}}\n",
    "bar_chart = {{bar_chart.value}}\n",
    "enrichment_groupby = \"{{enrichment_groupby.value}}\"\n",
    "diff_gex_method = \"{{diff_gex_method.value}}\"\n",
    "gene_topk = {{gene_topk.value}}\n",
    "enrichr_libraries = {{enrichr_libraries.value}}\n",
    "nr_genesets = {{nr_genesets.value}}\n",
    "\n",
    "trajectory = {{trajectory.value}}\n",
    "trajectory_method = \"{{trajectory_method.value}}\"\n",
    "\n",
    "time_series_trajectory = {{time_series_trajectory.value}}\n",
    "timepoint_labels_column_name = \"{{timepoint_labels_column_name.value}}\"\n",
    "\n",
    "cell_type_prediction = {{cell_type_prediction.value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings('ignore')\n",
    "random.seed(0)\n",
    "chart_studio.tools.set_credentials_file(username='mjjeon', api_key='v0rpMa6lhST28Sq7XqtM')\n",
    "pandas2ri.activate()\n",
    "plot_type='interactive'\n",
    "results = {}\n",
    "table_counter = 1\n",
    "figure_counter = 1\n",
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.settings.set_figure_params(dpi=80)\n",
    "sc.settings.figdir=\"./\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_metadata(adata, meta_data_filename, meta_class_column_name):\n",
    "    if meta_data_filename != \"\":\n",
    "        if meta_data_filename.endswith(\".csv\"):\n",
    "            meta_df = pd.read_csv(meta_data_filename, index_col=0)\n",
    "        else:\n",
    "            meta_df = pd.read_csv(meta_data_filename, sep=\"\\t\", index_col=0)\n",
    "        if meta_class_column_name == \"\":\n",
    "            raise Exception (\"Run time error: Please provide a proper column name for sample classes in metadata\")\n",
    "        try:\n",
    "            check_df(meta_df, meta_class_column_name)\n",
    "        except:\n",
    "            raise Exception (f\"Error! Column '{meta_class_column_name}' is not in metadata\")\n",
    "        adata.obs[meta_class_column_name] = meta_df.loc[:, meta_class_column_name]\n",
    "        adata.var_names_make_unique()\n",
    "\n",
    "    else:\n",
    "        meta_class_column_name = \"Class\"\n",
    "        adata.obs[meta_class_column_name] = [\"Class0\"]*adata.n_obs\n",
    "        adata.var_names_make_unique()\n",
    "    \n",
    "    return adata, meta_class_column_name\n",
    "def display_statistics(data, description=\"\"):\n",
    "    print(description)\n",
    "    print(\"Sample size:\", data.n_obs)\n",
    "    print(\"Feature size:\", data.n_vars)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{%- if file_kind.raw_value == 'Plain text' %}\n",
    "check_files(rnaseq_data_filename)\n",
    "    \n",
    "try:\n",
    "    if rnaseq_data_filename.endswith(\".csv\"):\n",
    "        expr_df = pd.read_csv(rnaseq_data_filename, index_col=0).sort_index()\n",
    "    else:\n",
    "        expr_df = pd.read_csv(rnaseq_data_filename, index_col=0, sep=\"\\t\").sort_index()\n",
    "    \n",
    "    # convert df into anndata\n",
    "    # adata matrix: sample x gene\n",
    "    adata = anndata.AnnData(expr_df.T)\n",
    "    adata.X = adata.X.astype('float64')    \n",
    "    \n",
    "except:\n",
    "    print(\"Error! Input files are in a wrong format. \\\n",
    "    Please check if the index of the expression data are genes and the columns are sample IDs. \\\n",
    "    Sample IDs in the expression data and the metadata should be matched\")\n",
    "\n",
    "del expr_df\n",
    "\n",
    "{%- elif file_kind.raw_value == '.mtx from 10x Genomics' %}\n",
    "adata = load_seurat_files(mtx_data_filename, gene_data_filename, barcode_data_filename)\n",
    "\n",
    "{%- endif -%}\n",
    "# load meta data\n",
    "adata, meta_class_column_name = load_metadata(adata, meta_data_filename, meta_class_column_name)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_counter = display_object(table_counter, \"Raw data. The table displays the first 5 rows of the quantified RNA-seq expression dataset. Rows represent genes, columns represent samples, and values show the number of mapped reads.\", adata.to_df().iloc[:10,:5].T.head(), istable=True)\n",
    "table_counter = display_object(table_counter, \"Metadata. The table displays the metadata associated with the samples in the RNA-seq dataset. Rows represent RNA-seq samples, columns represent metadata categories.\", adata.obs.head(), istable=True)\n",
    "table_counter = display_object(table_counter, \"Sample size for each class. The table displays the number of samples in each class.\", adata.obs.reset_index().groupby(meta_class_column_name).count(), istable=True)\n",
    "display_statistics(adata, \"### Statistics of data ###\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{%if qc_filter_genes.value == True %}\n",
    "# Filter Out Cells Based On Mitochondrial Genes\n",
    "High expression levels of mitochondrial genes could be an indicator of poor quality cells (Islam, Saiful, et al. 2014, Ilicic, Tomislav, et al. 2016). In a situation where the cell membrane is broken, cytoplasmic RNA will be lost, but RNAs enclosed in the mitochondria will be retained. This analysis removes single cells that are likely having the cell membrane broken.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%if qc_filter_genes.value == True %}\n",
    "if adata.var_names[0].startswith(\"ENSG\") == False:\n",
    "    mito_genes = adata.var_names.str.startswith('MT-')    \n",
    "else:\n",
    "    gene_symbol_var_names = adata.var_names\n",
    "    mito_genes = [True if x in gene_id_map_dict and str(gene_id_map_dict[x]).startswith('MT-') else False for x in gene_symbol_var_names ]\n",
    "\n",
    "# for each cell compute fraction of counts in mito genes vs. all genes\n",
    "adata.obs['percent_mito'] = np.sum(\n",
    "    adata[:, mito_genes].X, axis=1) / np.sum(adata.X, axis=1)\n",
    "\n",
    "sc.pl.violin(adata, ['percent_mito'],\n",
    "         jitter=0.4, multi_panel=True, show=True, save=True)\n",
    "figure_counter = display_object(figure_counter, \"Violin plot of the percentage of mitochondrial gene expression counts in each cell\", istable=False)\n",
    "display_link(\"violin.pdf\", \"Download figure\")\n",
    "adata = adata[adata.obs.percent_mito < qc_threshold, :]\n",
    "display_statistics(adata, \"### Statistics of data after QC ###\")\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.raw = adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Library Size Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to quantify gene expression in an RNA-seq dataset, reads generated from the sequencing step are mapped to a reference genome and subsequently aggregated into numeric gene counts. Due to experimental variations and random technical noise, samples in an RNA-seq dataset often have variable amounts of the total RNA. Library size analysis calculates and displays the total number of reads mapped for each sample in the RNA-seq dataset, facilitating the identification of outlying samples and the assessment of the overall quality of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['reads'] = adata.X.sum(axis=1)\n",
    "library_sum_df = pd.DataFrame(adata.obs['reads'])\n",
    "if len(library_sum_df.index) > 200:\n",
    "    library_sum_df = library_sum_df.sample(200)\n",
    "fig = px.histogram(library_sum_df, x='reads')\n",
    "fig.update_yaxes(title=\"samples/cells\")\n",
    "fig.show()\n",
    "figure_counter = display_object(figure_counter, \"Histogram of the total number of reads mapped for each sample. The figure contains an interactive bar chart which displays the number of samples according to the total number of reads mapped to each RNA-seq sample in the dataset. Additional information for each sample is available by hovering over the bars.\", istable=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if normalization_method.value != \"None\" %}\n",
    "# Normalization\n",
    "Various normalization methods can be applied based on the selection made by the user (Zheng, Grace XY, et al. 2017, Weinreb, Caleb, et al. 2018, Butler, Andrew, et al. 2018). These normalization methods convert the raw read counts into standardized measures of gene expression by removing factors that may negatively affect the analysis.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if normalization_method.value != \"None\" %}\n",
    "if normalization_method == \"Zheng17\":\n",
    "    sc.pp.recipe_zheng17(adata, log=log_normalization, plot=False)\n",
    "elif normalization_method == \"Weinreb17\":\n",
    "    sc.pp.recipe_weinreb17(adata, log=log_normalization)\n",
    "elif normalization_method == \"Seurat\":\n",
    "    sc.pp.filter_cells(adata, min_genes=200)\n",
    "    sc.pp.filter_genes(adata, min_cells=3)\n",
    "    sc.pp.normalize_total(adata, target_sum=1e4)        \n",
    "    if log_normalization:\n",
    "        sc.pp.log1p(adata)\n",
    "    sc.pp.scale(adata, max_value=10)\n",
    "{% endif %}\n",
    "\n",
    "\n",
    "{% if magic.value == True %}    \n",
    "adata = run_magic(dataset=adata)\n",
    "table_counter = display_object(table_counter, \"Normalized data. The table displays the expression values after normalization.\", adata.uns[\"magic\"].T.head(), istable=True)\n",
    "display_statistics(adata, \"### Statistics of data after normalization ###\")\n",
    "display(create_download_link(adata.to_df().T, filename=\"Normalized_data.csv\"))\n",
    "\n",
    "{% elif normalization_method.value != \"None\" %}\n",
    "table_counter = display_object(table_counter, \"Normalized data. The table displays the expression values after normalization.\", adata.to_df().T.head(), istable=True)\n",
    "display_statistics(adata, \"### Statistics of data after normalization ###\")\n",
    "display(create_download_link(adata.to_df().T, filename=\"Normalized_data.csv\"))\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Most Highly Expressed Genes Across All samples/Cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This analysis displays the genes that yield the highest fraction of counts in each single cell, across all cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = adata.to_df()\n",
    "df = df.divide(df.sum(axis=1), axis=0)\n",
    "mean_df = df.mean().sort_values(ascending=False)\n",
    "yaxis_gene_names = list(mean_df.index)[:highest_expr_n_genes]\n",
    "median_df = df.median().sort_values(ascending=False)\n",
    "selected_median_df = median_df.loc[yaxis_gene_names]\n",
    "sorted_yaxis_gene_names = list(selected_median_df.sort_values(ascending=False).index)\n",
    "sc.pl.highest_expr_genes(adata, log=True, n_top=highest_expr_n_genes, order=sorted_yaxis_gene_names, show=True, save=True)\n",
    "figure_counter = display_object(figure_counter, \"The {} genes with the highest mean fraction over all cells sorted by median.\".format(highest_expr_n_genes), istable=False)\n",
    "display_link(\"highest_expr_genes.pdf\", \"Download figure\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dimensionality Reduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Principal Component Analysis (PCA ) (Clark, N.R. and Maâ€™ayan, A., 2011), T-distributed Stochastic Neighbor Embedding (t-SNE) (Maaten, Laurens van der, and Geoffrey Hinton, 2008), Uniform Manifold Approximation and Projection (UMAP) (McInnes, Leland, et al. 2018) are statistical techniques used to identify global patterns in high-dimensional datasets. These techniques are commonly used to explore the similarity of biological samples in RNA-seq datasets. To achieve this, gene expression values are transformed into 2- or 3-dimensional latent vectors and subsequently visualized using a scatter plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run analysis\n",
    "results['dimension_reduction_pca'] = run_dimension_reduction(dim_reduction_method=\"PCA\", dataset=adata, meta_class_column_name=meta_class_column_name, magic_normalization=False, nr_genes=nr_genes, plot_type=plot_type)\n",
    "results['dimension_reduction_tsne'] = run_dimension_reduction(dim_reduction_method=\"t-SNE\", dataset=adata, meta_class_column_name=meta_class_column_name, magic_normalization=False, nr_genes=nr_genes, plot_type=plot_type)\n",
    "results['dimension_reduction_umap'] = run_dimension_reduction(dim_reduction_method=\"UMAP\", dataset=adata, meta_class_column_name=meta_class_column_name, magic_normalization=False, nr_genes=nr_genes, plot_type=plot_type)\n",
    "\n",
    "# Display results\n",
    "# PCA\n",
    "plot_dimension_reduction(results['dimension_reduction_pca'])\n",
    "figure_counter = display_object(figure_counter, subcounter=\"A\", caption='3D PCA plot for samples using {} genes having largest variance. \\\n",
    "    The figure displays an interactive, three-dimensional scatter plot of the data. \\\n",
    "    Each point represents an RNA-seq sample. \\\n",
    "    Samples with similar gene expression profiles are closer in the three-dimensional space. \\\n",
    "    If provided, sample groups are indicated using different colors, allowing for easier interpretation of the results.'.format(nr_genes), istable=False)\n",
    "# t-SNE\n",
    "plot_dimension_reduction(results['dimension_reduction_tsne'])\n",
    "figure_counter = display_object(figure_counter, subcounter=\"B\", caption='3D t-SNE plot for samples using {} genes having largest variance. \\\n",
    "    The figure displays an interactive, three-dimensional scatter plot of the data. \\\n",
    "    Each point represents an RNA-seq sample. \\\n",
    "    Samples with similar gene expression profiles are closer in the three-dimensional space. \\\n",
    "    If provided, sample groups are indicated using different colors, allowing for easier interpretation of the results.'.format(nr_genes), istable=False)\n",
    "# UMAP\n",
    "plot_dimension_reduction(results['dimension_reduction_umap'])\n",
    "figure_counter = display_object(figure_counter, subcounter=\"C\", caption='3D UMAP plot for samples using {} genes having largest variance. \\\n",
    "    The figure displays an interactive, three-dimensional scatter plot of the data. \\\n",
    "    Each point represents an RNA-seq sample. \\\n",
    "    Samples with similar gene expression profiles are closer in the three-dimensional space. \\\n",
    "    If provided, sample groups are indicated using different colors, allowing for easier interpretation of the results.'.format(nr_genes), istable=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if magic.value == True %}\n",
    "# Imputation of Dropouts with MAGIC\n",
    "Markov Affinity-based Graph Imputation of Cells (MAGIC) (van Dijk, D., et al. 2017) is an algorithm for denoising high-dimensional data. It is most commonly applied to impute single-cell RNA sequencing data. MAGIC learns the manifold data, using the resultant graph to smooth the features and restore the original structure of the data.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if magic.value == True %}\n",
    "# Run analysis\n",
    "results['magic_dimension_reduction_pca'] = run_dimension_reduction(dim_reduction_method=\"PCA\", dataset=adata, meta_class_column_name=meta_class_column_name, magic_normalization=magic, nr_genes=nr_genes, color_by='auto')\n",
    "results['magic_dimension_reduction_tsne'] = run_dimension_reduction(dim_reduction_method=\"t-SNE\", dataset=adata, meta_class_column_name=meta_class_column_name, magic_normalization=magic, nr_genes=nr_genes, color_by='auto')\n",
    "results['magic_dimension_reduction_umap'] = run_dimension_reduction(dim_reduction_method=\"UMAP\", dataset=adata, meta_class_column_name=meta_class_column_name, magic_normalization=magic, nr_genes=nr_genes, color_by='auto')\n",
    "\n",
    "# Display results\n",
    "plot_dimension_reduction(results['magic_dimension_reduction_pca'])\n",
    "figure_counter = display_object(figure_counter, subcounter=\"A\", caption=\"3D PCA plot for samples using {} genes having largest variance after MAGIC imputation. The figure displays an interactive, three-dimensional scatter plot of the data. Each point represents an RNA-seq sample. Samples with similar gene expression profiles are closer in the three-dimensional space. If provided, sample groups are indicated using different colors, allowing for easier interpretation of the results.\".format(nr_genes), istable=False)\n",
    "\n",
    "plot_dimension_reduction(results['magic_dimension_reduction_tsne'])\n",
    "figure_counter = display_object(figure_counter, subcounter=\"B\", caption=\"3D t-SNE plot for samples using {} genes having largest variance after MAGIC imputation. The figure displays an interactive, three-dimensional scatter plot of the data. Each point represents an RNA-seq sample. Samples with similar gene expression profiles are closer in the three-dimensional space. If provided, sample groups are indicated using different colors, allowing for easier interpretation of the results.\".format(nr_genes), istable=False)\n",
    "\n",
    "plot_dimension_reduction(results['magic_dimension_reduction_umap'])\n",
    "figure_counter = display_object(figure_counter, subcounter=\"C\", caption=\"3D UMAP plot for samples using {} genes having largest variance after MAGIC imputation. The figure displays an interactive, three-dimensional scatter plot of the data. Each point represents an RNA-seq sample. Samples with similar gene expression profiles are closer in the three-dimensional space. If provided, sample groups are indicated using different colors, allowing for easier interpretation of the results.\".format(nr_genes), istable=False)\n",
    "\n",
    "adata.X = adata.uns['magic']\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clustergrammer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clustergrammer (Fernandez, Nicolas F., et al. 2017) is a web-based tool for visualizing and analyzing high-dimensional data as interactive and hierarchically clustered heatmaps. It is commonly used to explore the similarity between samples in an RNA-seq dataset. In addition to identifying clusters of samples, it also allows to identify the genes which contribute to the clustering. It visualizes the top 800 genes with largest variance by default or you can upload your gene list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run analysis\n",
    "results['clustergrammer'] = run_clustergrammer(dataset=adata, meta_class_column_name=meta_class_column_name, gene_list=gene_list_for_clustergrammer)\n",
    "\n",
    "# Display results\n",
    "plot_clustergrammar(results['clustergrammer'])\n",
    "figure_counter = display_object(figure_counter, \"Clustered heatmap plot. The figure contains an interactive heatmap displaying gene expression for each sample in the RNA-seq dataset. Every row of the heatmap represents a gene, every column represents a sample, and every cell displays normalized gene expression values. The heatmap additionally features color bars beside each column which represent prior knowledge of each sample, such as the tissue of origin or experimental treatment.\".format(highest_expr_n_genes), istable=False)\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Automatic Identification of Clusters\n",
    "The Leiden algorithm (Traag, Vincent A., et al. 2019) is a method to identify well-connected clusters in networks. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata, n_neighbors=15)\n",
    "sc.tl.leiden(adata, resolution=1.0)\n",
    "sc.tl.umap(adata, min_dist=0.1)\n",
    "display(create_download_link(adata.obs[\"leiden\"], filename=\"clustering.csv\"))\n",
    "\n",
    "# sort by clusters\n",
    "new_order = adata.obs.sort_values(by='leiden').index.tolist()\n",
    "adata = adata[new_order, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# umap info into dataframe \n",
    "umap_df = pd.DataFrame(adata.obsm['X_umap'])\n",
    "umap_df.columns = ['x', 'y']\n",
    "\n",
    "values_dict = dict()\n",
    "values_dict[\"Clusters\"] = adata.obs[\"leiden\"].values\n",
    "\n",
    "figure_counter = plot_scatter(umap_df, values_dict, [\"Clusters\"], adata.obs.index.tolist(), \"Scatter plot of the samples. Each dot represents a sample and it is colored by \", category=True, dropdown=False, figure_counter=figure_counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{%if deg.value == True %}\n",
    "# Differential Gene Expression Analysis\n",
    "Gene expression signatures are alterations in the patterns of gene expression that occur as a result of cellular perturbations such as drug treatments, gene knock-downs or diseases. They can be quantified using differential gene expression (DGE) methods (Ritchie, Matthew E., et al. 2015, Clark, Neil R., et al. 2014), which compare gene expression between two groups of samples to identify genes whose expression is significantly altered in the perturbation.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if deg.value == True %}\n",
    "\n",
    "{% if enrichment_groupby.value == \"user_defined_class\" %}\n",
    "classes = adata.obs[meta_class_column_name].unique().tolist()\n",
    "cluster = False\n",
    "if len(classes) < 2:\n",
    "    print(\"Warning: Please provide at least 2 classes in the metadata\")\n",
    "{% else %}\n",
    "meta_class_column_name = \"leiden\"\n",
    "classes = sorted(adata.obs[\"leiden\"].unique().tolist())\n",
    "classes.sort(key=int)\n",
    "cluster=True\n",
    "\n",
    "{% endif %}\n",
    "if len(classes) > 5 and adata.n_obs > 5000:\n",
    "    if diff_gex_method == \"wilcoxon\":\n",
    "        print('Warning: There are too many cells/clusters. It cannot execute the analysis code for the data. The appyter randomly select 1000 samples. If you want to execute it with the whole data, please run it locally.')\n",
    "    else:\n",
    "        print('Warning: There are too many cells/clusters. It cannot execute the analysis code for the data. The appyter switched to Wilcoxon rank-sum method and randomly select 1000 samples. If you want to execute it with the whole data, please run it locally.')\n",
    "        diff_gex_method = \"wilcoxon\"\n",
    "    # randomly select 5K samples\n",
    "    random_selected_samples = random.sample(adata.obs.index.tolist(), 5000)\n",
    "    adata_random_sampled = adata[random_selected_samples, :]\n",
    "    signatures = get_signatures(classes, adata_random_sampled, method=diff_gex_method, meta_class_column_name=meta_class_column_name, cluster=cluster)\n",
    "else:\n",
    "    signatures = get_signatures(classes, adata, method=diff_gex_method, meta_class_column_name=meta_class_column_name, cluster=cluster)\n",
    "    \n",
    "    \n",
    "if len(classes) > 1:\n",
    "    \n",
    "    for label, signature in signatures.items():\n",
    "        table_counter = display_object(table_counter, \"Differentially expressed genes between {} using {}. Every row of the table represents a gene; the columns display the estimated measures of differential expression.\".format(label, diff_gex_method), signature, istable=True)\n",
    "        display(create_download_link(signature, filename=\"DEG_{}.csv\".format(label)))\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{%if deg.value == True %}\n",
    "# Enrichment Analysis\n",
    "Enrichment analysis is a bioinformatics method used to identify prior knowledge terms which are over-represented in a given gene set by comparing the gene set to many annotated gene sets. The prior-knowledge gene sets can represent cell signaling pathways, molecular functions, diseases, and a wide variety of other terms obtained by processing data from multiple resources. Enrichr (Kuleshov, M.V., et al. 2016) is a web-based application that  performs enrichment analysis against a large collection of gene-set libraries. It provides various interactive visualizations to display enrichment results.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%if deg.value == True %}\n",
    "if diff_gex_method == \"characteristic_direction\":\n",
    "    fc_colname = \"CD-coefficient\"\n",
    "    sort_genes_by = \"CD-coefficient\"\n",
    "    ascending = False\n",
    "elif diff_gex_method == \"limma\":\n",
    "    fc_colname = \"logFC\"\n",
    "    sort_genes_by = \"t\"\n",
    "    ascending = False\n",
    "elif diff_gex_method == \"edgeR\":\n",
    "    fc_colname = \"logFC\"\n",
    "    sort_genes_by = \"PValue\"\n",
    "    ascending = True\n",
    "elif diff_gex_method == \"DESeq2\":\n",
    "    fc_colname = \"log2FoldChange\"\n",
    "    sort_genes_by = \"padj\"\n",
    "    ascending = True\n",
    "elif diff_gex_method == \"wilcoxon\":\n",
    "    fc_colname = \"logfoldchanges\"\n",
    "    sort_genes_by = \"pvals\"\n",
    "    ascending = True\n",
    "\n",
    "results['enrichr'] = {}\n",
    "{% if enrichment_groupby.value != \"user_defined_class\" %}\n",
    "default_node_size = 120000 / len(adata.obs.index)\n",
    "node_size = min(100, 120000 / len(adata.obs.index))\n",
    "sc.pl.umap(adata, color=['leiden'], size=node_size)\n",
    "{% else %}\n",
    "default_node_size = 120000 / len(adata.obs.index)\n",
    "node_size = min(100, 120000 / len(adata.obs.index))\n",
    "sc.pl.umap(adata, color=[meta_class_column_name], size=node_size)\n",
    "\n",
    "{% endif %}\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    {% if enrichment_groupby.value == \"user_defined_class\" %}\n",
    "    case_name = label.split(\" vs. \")[1]\n",
    "    {% else %}\n",
    "    case_name = label.split(\" vs. \")[0]\n",
    "    {% endif %}\n",
    "    results['enrichr'][label] = run_enrichr(signature=signature, signature_label=label, fc_colname=fc_colname,geneset_size=gene_topk, sort_genes_by = sort_genes_by,ascending=ascending)\n",
    "    display(Markdown(\"*Enrichment Analysis Result: {} (Up-regulated in {})*\".format(label, case_name)))\n",
    "    display_link(\"https://amp.pharm.mssm.edu/Enrichr/enrich?dataset={}\".format(results['enrichr'][label][\"upregulated\"][\"shortId\"]))\n",
    "    display(Markdown(\"*Enrichment Analysis Result: {} (Down-regulated in {})*\".format(label, case_name)))\n",
    "    display_link(\"https://amp.pharm.mssm.edu/Enrichr/enrich?dataset={}\".format(results['enrichr'][label][\"downregulated\"][\"shortId\"]))\n",
    "table_counter = display_object(table_counter, \"The table displays links to Enrichr containing the results of enrichment analyses generated by analyzing the up-regulated and down-regulated genes from a differential expression analysis. By clicking on these links, users can interactively explore and download the enrichment results from the Enrichr website.\", istable=True)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if deg.value == True %}\n",
    "# Bar Charts for Selected Libraries\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if deg.value == True and \"Gene Ontology\" in enrichr_libraries.value %}\n",
    "### Gene Ontology Enrichment Analysis\n",
    "Gene Ontology (GO) (Ashburner, M., et al. 2000) is a major initiative aimed at unifying the representation of gene attributes across all species. It contains a large collection of experimentally validated and predicted associations between genes and biological terms. This information can be leveraged by Enrichr to identify enriched biological processes, molecular functions and cellular components which are over-represented in the up-regulated genes from each cluster.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if deg.value == True and \"Gene Ontology\" in enrichr_libraries.value %}\n",
    "\n",
    "results['go_enrichment'] = {}\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    results['go_enrichment'][label] = get_enrichr_results_by_library(results['enrichr'][label], label, library_type='go', version='2018')\n",
    "{% endif %}\n",
    "\n",
    "{% if deg.value == True and \"Gene Ontology\" in enrichr_libraries.value and bar_chart.value == True %}    \n",
    "for label, signature in signatures.items():\n",
    "    {% if enrichment_groupby.value == \"user_defined_class\" %}\n",
    "\n",
    "    case_name = label.split(\" vs. \")[1]\n",
    "    {% else %}\n",
    "    case_name = label.split(\" vs. \")[0]\n",
    "    {% endif %}\n",
    "\n",
    "    # Create dataframe\n",
    "    enrichment_results = results['go_enrichment'][label]\n",
    "    enrichment_dataframe = pd.concat([enrichment_results['upregulated'], enrichment_results['downregulated']])\n",
    "\n",
    "    # Plot barcharts\n",
    "    libraries = enrichment_dataframe['gene_set_library'].unique()   \n",
    "    for gene_set_library in libraries:\n",
    "        plot_library_barchart(enrichment_results, gene_set_library, enrichment_results['signature_label'], case_name, enrichment_results['sort_results_by'], nr_genesets=nr_genesets, plot_type=plot_type) # 10 300\n",
    "    figure_counter = display_object(figure_counter, \"Enrichment Analysis Results for {} in Gene Onotology. The figure contains interactive bar charts displaying the results of the Gene Ontology enrichment analysis generated using Enrichr. The x axis indicates the -log10(P-value) for each term. Significant terms are highlighted in bold. Additional information about enrichment results is available by hovering over each bar.\".format(label), istable=False)\n",
    "{% endif %}    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if deg.value == True and \"Pathway\" in enrichr_libraries.value %}\n",
    "### Pathway Enrichment Analysis\n",
    "Biological pathways are sequences of interactions between biochemical compounds which play a key role in determining cellular behavior. Databases such as KEGG (Kanehisa et al. 2000), Reactome (Croft et al. 2014) and WikiPathways (Kelder et al. 2012) contain a large number of associations between such pathways and genes. This information can be leveraged by Enrichr to identify the biological pathways which are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if deg.value == True and \"Pathway\" in enrichr_libraries.value %}\n",
    "# Initialize results\n",
    "results['pathway_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    results['pathway_enrichment'][label] = get_enrichr_results_by_library(results['enrichr'][label], label, library_type='pathway')\n",
    "{% endif %}\n",
    "{% if deg.value == True and \"Pathway\" in enrichr_libraries.value and bar_chart.value == True %}    \n",
    "\n",
    "    for label, signature in signatures.items():\n",
    "        {% if enrichment_groupby.value == \"user_defined_class\" %}\n",
    "        case_name = label.split(\" vs. \")[1]\n",
    "        {% else %}\n",
    "        case_name = label.split(\" vs. \")[0]\n",
    "        {% endif %}\n",
    "\n",
    "        # Create dataframe\n",
    "        enrichment_results = results['pathway_enrichment'][label]\n",
    "        enrichment_dataframe = pd.concat([enrichment_results['upregulated'], enrichment_results['downregulated']])\n",
    "\n",
    "        # Plot barcharts\n",
    "        libraries = enrichment_dataframe['gene_set_library'].unique()   \n",
    "        for gene_set_library in libraries:\n",
    "            # Display results\n",
    "            plot_library_barchart(enrichment_results, gene_set_library, enrichment_results['signature_label'], case_name, enrichment_results['sort_results_by'], nr_genesets=nr_genesets, plot_type=plot_type)\n",
    "        figure_counter = display_object(figure_counter, \"Enrichment Analysis Results for {} in KEGG Pathways, WikiPathways and Reactome Pathways. The figure contains interactive bar charts displaying the results of the Gene Ontology enrichment analysis generated using Enrichr. The x axis indicates the -log10(P-value) for each term. Significant terms are highlighted in bold. Additional information about enrichment results is available by hovering over each bar.\".format(label), istable=False)\n",
    "    \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if deg.value == True and \"Transcription Factor\" in enrichr_libraries.value %}\n",
    "### Transcription Factor Enrichment Analysis\n",
    "Transcription Factors (TFs) are proteins involved in the transcriptional regulation of gene expression. Databases such as ChEA (Lachmann et al. 2010) and ENCODE (Consortium, 2014) contain a large number of associations between TFs and their transcriptional targets. This information can be leveraged by Enrichr to identify the transcription factors whose targets are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if deg.value == True and \"Transcription Factor\" in enrichr_libraries.value %}\n",
    "# Initialize results\n",
    "results['tf_enrichment'] = {}\n",
    "# Loop through results\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    results['tf_enrichment'][label] = get_enrichr_result_tables_by_library(enrichr_results=results['enrichr'][label], signature_label=label, library_type='tf')\n",
    "    {% if bar_chart.value == True%}\n",
    "        table_counter = display_table(results['tf_enrichment'][label], \"Transcription Factor\", table_counter)\n",
    "    {% endif %} \n",
    "{% endif %} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if deg.value == True and \"Kinase\" in enrichr_libraries.value %}\n",
    "### Kinase Enrichment Analysis\n",
    "Protein kinases are enzymes that modify other proteins by chemically adding phosphate groups. Databases such as KEA (Lachmann et al. 2009) contain a large number of associations between kinases and their substrates. This information can be leveraged by Enrichr to identify the protein kinases whose substrates are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if deg.value == True and \"Kinase\" in enrichr_libraries.value %}\n",
    "# Initialize results\n",
    "results['kinase_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, enrichr_results in results['enrichr'].items():\n",
    "    # Run analysis\n",
    "    results['kinase_enrichment'][label] = get_enrichr_result_tables_by_library(enrichr_results=enrichr_results, signature_label=label, library_type=\"ke\")\n",
    "\n",
    "    # Display results\n",
    "    {% if bar_chart.value == True%}\n",
    "        table_counter = display_table(results['kinase_enrichment'][label], \"Kinase\", table_counter)\n",
    "    {% endif %}\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if deg.value == True and \"miRNA\" in enrichr_libraries.value %}\n",
    "### miRNA Enrichment Analysis\n",
    "microRNAs (miRNAs) are small non-coding RNA molecules which play a key role in the post-transcriptional regulation of gene expression. Databases such as TargetScan (Agarwal et al. 2015) and MiRTarBase (Chou et al. 2016) contain a large number of associations between miRNAs and their targets. This information can be leveraged by Enrichr to identify the miRNAs whose targets are over-represented in the up-regulated and down-regulated genes identified by comparing two groups of samples.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if deg.value == True and \"miRNA\" in enrichr_libraries.value %}\n",
    "\n",
    "results['mirna_enrichment'] = {}\n",
    "\n",
    "# Loop through results\n",
    "for label, enrichr_results in results['enrichr'].items():\n",
    "    # Run analysis\n",
    "    results['mirna_enrichment'][label] = get_enrichr_result_tables_by_library(enrichr_results=enrichr_results, signature_label=label, library_type=\"mirna\")\n",
    "\n",
    "    # Display results\n",
    "    {% if bar_chart.value == True%}\n",
    "        table_counter = display_table(results['mirna_enrichment'][label], \"miRNA\", table_counter)\n",
    "    {% endif %}\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if deg.value == True and \"Cell Type\" in enrichr_libraries.value %}\n",
    "### Cell Type Enrichment Analysis\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if deg.value == True and \"Cell Type\" in enrichr_libraries.value %}\n",
    "\n",
    "results['celltype_enrichment'] = {}\n",
    "\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    results['celltype_enrichment'][label] = get_enrichr_results_by_library(results['enrichr'][label], label, library_type='celltype')\n",
    "\n",
    "{% endif %}\n",
    "{% if deg.value == True and \"Cell Type\" in enrichr_libraries.value and bar_chart.value == True %}    \n",
    "\n",
    "    for label, signature in signatures.items():\n",
    "        {% if enrichment_groupby.value == \"user_defined_class\" %}\n",
    "        case_name = label.split(\" vs. \")[1]\n",
    "        {% else %}\n",
    "        case_name = label.split(\" vs. \")[0]\n",
    "        {% endif %}\n",
    "\n",
    "        # Create dataframe\n",
    "        enrichment_results = results['celltype_enrichment'][label]\n",
    "        enrichment_dataframe = pd.concat([enrichment_results['upregulated'], enrichment_results['downregulated']])\n",
    "\n",
    "        # Plot barcharts\n",
    "        libraries = enrichment_dataframe['gene_set_library'].unique()   \n",
    "        for gene_set_library in libraries:\n",
    "            plot_library_barchart(enrichment_results, gene_set_library, enrichment_results['signature_label'], case_name, enrichment_results['sort_results_by'], nr_genesets=nr_genesets, plot_type=plot_type) # 10 300\n",
    "        figure_counter = display_object(figure_counter, \"Enrichment Analysis Results for {} in Cell Type. The figure contains interactive bar charts displaying the results of the Gene Ontology enrichment analysis generated using Enrichr. The x axis indicates the -log10(P-value) for each term. Significant terms are highlighted in bold. Additional information about enrichment results is available by hovering over each bar.\".format(label), istable=False)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if deg.value == True and \"Disease\" in enrichr_libraries.value %}\n",
    "### Disease Enrichment Analysis\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if deg.value == True and \"Disease\" in enrichr_libraries.value %}\n",
    "\n",
    "results['disease_enrichment'] = {}\n",
    "for label, signature in signatures.items():\n",
    "    # Run analysis\n",
    "    results['disease_enrichment'][label] = get_enrichr_results_by_library(results['enrichr'][label], label, library_type='disease', version='2018')\n",
    "{% endif %}\n",
    "{% if deg.value == True and \"Disease\" in enrichr_libraries.value and bar_chart.value == True %}    \n",
    "    \n",
    "    for label, signature in signatures.items():\n",
    "        {% if enrichment_groupby.value == \"user_defined_class\" %}\n",
    "        case_name = label.split(\" vs. \")[1]\n",
    "        {% else %}\n",
    "        case_name = label.split(\" vs. \")[0]\n",
    "        {% endif %}\n",
    "\n",
    "        # Create dataframe\n",
    "        enrichment_results = results['disease_enrichment'][label]\n",
    "        enrichment_dataframe = pd.concat([enrichment_results['upregulated'], enrichment_results['downregulated']])\n",
    "\n",
    "        # Plot barcharts\n",
    "        libraries = enrichment_dataframe['gene_set_library'].unique()   \n",
    "        for gene_set_library in libraries:\n",
    "            plot_library_barchart(enrichment_results, gene_set_library, enrichment_results['signature_label'], case_name, enrichment_results['sort_results_by'], nr_genesets=nr_genesets, plot_type=plot_type) # 10 300\n",
    "        figure_counter = display_object(figure_counter, \"Enrichment Analysis Results for {} in Disease. The figure contains interactive bar charts displaying the results of the Gene Ontology enrichment analysis generated using Enrichr. The x axis indicates the -log10(P-value) for each term. Significant terms are highlighted in bold. Additional information about enrichment results is available by hovering over each bar.\".format(label), istable=False)\n",
    "    \n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if deg.value == True %}\n",
    "# Interactive Scatter Plots of Enriched Terms\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if deg.value == True%}\n",
    "    \n",
    "library_option_list = set()\n",
    "for label, signature in signatures.items():\n",
    "    if cluster == True:\n",
    "        cluster_names = [label.split(\" vs. \")[0].replace(\"Cluster \", \"\")]\n",
    "    else:\n",
    "        cluster_names = label.split(\" vs. \")\n",
    "    \n",
    "    for key in results.keys():\n",
    "        if key.endswith(\"enrichment\") == False:\n",
    "            continue\n",
    "        enrichment_results = results[key][label]\n",
    "        meta_df = adata.obs\n",
    "        if cluster == True:\n",
    "            for cluster_name in cluster_names:            \n",
    "                for direction in ['upregulated']:\n",
    "                    if direction in enrichment_results:\n",
    "                        enrichment_dataframe = enrichment_results[direction]\n",
    "                    else:\n",
    "                        enrichment_dataframe = enrichment_results[\"enrichment_dataframe\"]\n",
    "                    libraries = enrichment_dataframe['gene_set_library'].unique()  \n",
    "                    for library in libraries:\n",
    "                        enrichment_dataframe_library = enrichment_dataframe[enrichment_dataframe['gene_set_library']==library]\n",
    "                        top_term = enrichment_dataframe_library.iloc[0]['term_name']\n",
    "                        if cluster == True:\n",
    "                            meta_df.loc[meta_df['leiden']==cluster_name, library] = top_term\n",
    "                        else:\n",
    "                            meta_df.loc[meta_df[meta_class_column_name]==cluster_name, library] = top_term\n",
    "                        library_option_list.add(library)\n",
    "        else: # cluster == False\n",
    "            for direction in ['upregulated', 'downregulated']:\n",
    "                if direction in enrichment_results:\n",
    "                    enrichment_dataframe = enrichment_results[direction]\n",
    "                    libraries = enrichment_dataframe['gene_set_library'].unique()  \n",
    "                    for library in libraries:\n",
    "                        enrichment_dataframe_library = enrichment_dataframe[enrichment_dataframe['gene_set_library']==library]\n",
    "                        top_term = enrichment_dataframe_library.iloc[0]['term_name']\n",
    "                        if direction == \"upregulated\":\n",
    "                            meta_df.loc[meta_df[meta_class_column_name]==cluster_names[0], library] = top_term\n",
    "                        else:\n",
    "                            meta_df.loc[meta_df[meta_class_column_name]==cluster_names[1], library] = top_term\n",
    "                        library_option_list.add(library)\n",
    "                else:\n",
    "                    enrichment_dataframe = enrichment_results[\"enrichment_dataframe\"]\n",
    "                    libraries = enrichment_dataframe['gene_set_library'].unique()  \n",
    "                    for library in libraries:\n",
    "                        enrichment_dataframe_library = enrichment_dataframe[enrichment_dataframe['gene_set_library']==library]\n",
    "                        top_term = enrichment_dataframe_library.iloc[0]['term_name']\n",
    "                        bottom_term = enrichment_dataframe_library.iloc[-1]['term_name']\n",
    "                        meta_df.loc[meta_df[meta_class_column_name]==cluster_names[0], library] = top_term\n",
    "                        meta_df.loc[meta_df[meta_class_column_name]==cluster_names[1], library] = bottom_term\n",
    "                        library_option_list.add(library)\n",
    "                    \n",
    "library_option_list = list(library_option_list)\n",
    "\n",
    "# umap info into dataframe \n",
    "umap_df = pd.DataFrame(adata.obsm['X_umap'])\n",
    "umap_df.columns = ['x', 'y']\n",
    "\n",
    "option_list = library_option_list  \n",
    "adata_selected = adata.obs[option_list].fillna(\"NaN\")\n",
    "values_dict = dict(zip(adata_selected.T.index.tolist(), adata_selected.T.values))\n",
    "\n",
    "figure_counter = plot_scatter(umap_df, values_dict, option_list, adata.obs.index.tolist(), \"Scatter plot of the samples. Each dot represents a sample and it is colored by enriched terms in library \", location='below', category=True, dropdown=True, figure_counter=figure_counter)\n",
    "\n",
    "{% endif %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if trajectory.value == True %}\n",
    "# Trajectory Inference\n",
    "Trajectory inference is a computational technique used in single-cell transcriptomics to arrange cells based on their progression through the differentiation process. It orders single cells in pseudotime, placing them along a trajectory corresponding to a biological process such as cell differentiation by taking advantage of single cellsâ€™ asynchronous progression through those processes. Diffusion pseudotime (DPT) (Haghverdi, Laleh, et al. 2016), which measures transitions between cells using diffusion-like random walks is one of the methods used to identify such trajectories. Monocle (Trapnell, Cole, et al. 2014) is a method for ordering cells by learning an explicit principal graph from the single cell expression data with advanced machine learning techniques (Reversed Graph Embedding), which robustly and accurately resolves complicated biological processes.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if trajectory.value == True %}\n",
    "if trajectory_method == \"monocle\":\n",
    "    # Run analysis\n",
    "    results['monocle'] = run_monocle(dataset=adata, plot_type=plot_type, color_by='Pseudotime')\n",
    "\n",
    "    # Display results\n",
    "    plot_monocle(results['monocle'])\n",
    "    \n",
    "elif trajectory_method == \"dpt\":\n",
    "    adata.uns['iroot'] = 0\n",
    "    sc.pl.umap(adata, color=['leiden'], size=node_size)\n",
    "    sc.tl.dpt(adata)\n",
    "    sc.pl.umap(adata, color=['dpt_pseudotime'], size=node_size)\n",
    "    display_link(\"draw_graph_fa.pdf\", \"Download figure\")\n",
    "figure_counter = display_object(figure_counter, \"Trajectory inference result using {}. Each point represents an RNA-seq sample. Sample colors are based on pseudotime.\".format(trajectory_method), istable=False)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if time_series_trajectory.value == True %}\n",
    "# Time-series Trajectory Analysis\n",
    "When time-series data is available, it is possible to order cells by their progression through a dynamic\n",
    "process using time information. Tempora (Tran, Thinh N., and Gary Bader 2016), a pathway-based cell trajectory inference method that orders cells using time information from the data, infers developmental linages based on biological pathway information. This time-series based analysis can generate insights into dynamic processes and their biological regulation. \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if time_series_trajectory.value == True %}\n",
    "run_tempora(adata, timepoint_labels_column_name, timepoint_labels)\n",
    "display(Image.open(\"Tempora_plot.jpg\"))\n",
    "figure_counter = display_object(figure_counter, \"Time-series trajectory inference result using Tempora. Tempora visualizes the result as a network, with the piechart at each node representing the composition of cells collected at different time points in the experiment and the arrow connecting each pair of nodes representing lineage relationship between them.\", istable=False)\n",
    "display(FileLink(\"Tempora_plot.jpg\", result_html_prefix=\"Download figure\"))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if cell_type_prediction.value == True %}\n",
    "# Cell Type Prediction with DigitalCellSorter \n",
    "DigitalCellSorter (Domanskyi, Sergii, et al. 2019) is an unbiased cell type recognition algorithm for scRNA-seq. It leverages a manual assessment using a few selected markers of cell types and infers the cell of origin for each of the single cells independently.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if cell_type_prediction.value == True %}\n",
    "DCS = DigitalCellSorter.DigitalCellSorter(nComponentsPCA=3)\n",
    "DCS.annotationMethod='pDCS'\n",
    "DCS.makePlots = False\n",
    "df_expr = adata.raw.to_adata().to_df().T\n",
    "DCS.prepare(df_expr)\n",
    "DCS.process()\n",
    "\n",
    "ann= DCS.annotate()\n",
    "\n",
    "# Run analysis\n",
    "results['dimension_reduction'] = run_dimension_reduction(dim_reduction_method, dataset=adata, meta_class_column_name=meta_class_column_name, nr_genes=nr_genes, plot_type=plot_type)\n",
    "ids = [x[1] for x in DCS.df_expr.columns]\n",
    "clusters = [ann[x[2]] for x in DCS.df_expr.columns] \n",
    "new_cluster_df = pd.DataFrame([ids, clusters]).T.set_index(0)\n",
    "new_cluster_df.columns = [meta_class_column_name]\n",
    "results['dimension_reduction']['sample_metadata'] = new_cluster_df\n",
    "\n",
    "# # Display results\n",
    "plot_dimension_reduction(results['dimension_reduction'])\n",
    "figure_counter = display_object(figure_counter, \"Cell type prediction result using DigitalCellSorter. Each point represents an RNA-seq sample. Predicted sample cell types are indicated using different colors.\", istable=False)\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# References "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Ashburner, M., Ball, C.A., Blake, J.A., Botstein, D., Butler, H., Cherry, J.M., Davis, A.P., Dolinski, K., Dwight, S.S. and Eppig, J.T. (2000) Gene Ontology: tool for the unification of biology. Nature genetics, 25, 25.\n",
    "<br>\n",
    "Butler, Andrew, et al. \"Integrating single-cell transcriptomic data across different conditions, technologies, and species.\" Nature biotechnology 36.5 (2018): 411-420.\n",
    "<br>\n",
    "Clark, N.R. and Maâ€™ayan, A. (2011) Introduction to statistical methods to analyze large data sets: principal components analysis. Sci. Signal., 4, tr3-tr3.\n",
    "<br>\n",
    "Clark, Neil R., et al. \"The characteristic direction: a geometrical approach to identify differentially expressed genes.\" BMC bioinformatics 15.1 (2014): 79.\n",
    "<br>\n",
    "Consortium, E.P. (2004) The ENCODE (ENCyclopedia of DNA elements) project. Science, 306, 636-640.\n",
    "<br>\n",
    "Croft, David, et al. \"The Reactome pathway knowledgebase.\" Nucleic acids research 42.D1 (2014): D472-D477.\n",
    "<br>\n",
    "Domanskyi, Sergii, et al. \"Polled Digital Cell Sorter (p-DCS): Automatic identification of hematological cell types from single cell RNA-sequencing clusters.\" BMC bioinformatics 20.1 (2019): 369.\n",
    "<br>\n",
    "Fernandez, Nicolas F., et al. \"Clustergrammer, a web-based heatmap visualization and analysis tool for high-dimensional biological data.\" Scientific data 4 (2017): 170151.\n",
    "<br>\n",
    "Haghverdi, Laleh, et al. \"Diffusion pseudotime robustly reconstructs lineage branching.\" Nature methods 13.10 (2016): 845.\n",
    "<br>\n",
    "Ilicic, Tomislav, et al. \"Classification of low quality cells from single-cell RNA-seq data.\" Genome biology 17.1 (2016): 29.\n",
    "<br>\n",
    "Islam, Saiful, et al. \"Quantitative single-cell RNA-seq with unique molecular identifiers.\" Nature methods 11.2 (2014): 163.\n",
    "<br>\n",
    "Kanehisa, M. and Goto, S. (2000) KEGG: kyoto encyclopedia of genes and genomes. Nucleic acids research, 28, 27-30.\n",
    "<br>\n",
    "Kuleshov, M.V., Jones, M.R., Rouillard, A.D., Fernandez, N.F., Duan, Q., Wang, Z., Koplev, S., Jenkins, S.L., Jagodnik, K.M. and Lachmann, A. (2016) Enrichr: a comprehensive gene set enrichment analysis web server 2016 update. Nucleic acids research, 44, W90-W97.\n",
    "<br>\n",
    "Lachmann, A., Xu, H., Krishnan, J., Berger, S.I., Mazloom, A.R. and Ma'ayan, A. (2010) ChEA: transcription factor regulation inferred from integrating genome-wide ChIP-X experiments. Bioinformatics, 26, 2438-2444.\n",
    "<br>\n",
    "Lachmann, Alexander, and Avi Ma'ayan. \"KEA: kinase enrichment analysis.\" Bioinformatics 25.5 (2009): 684-686.\n",
    "<br>\n",
    "Maaten, Laurens van der, and Geoffrey Hinton. \"Visualizing data using t-SNE.\" Journal of machine learning research 9.Nov (2008): 2579-2605.\n",
    "<br>\n",
    "McInnes, Leland, John Healy, and James Melville. \"Umap: Uniform manifold approximation and projection for dimension reduction.\" arXiv preprint arXiv:1802.03426 (2018).\n",
    "<br>\n",
    "Ritchie, Matthew E., et al. \"limma powers differential expression analyses for RNA-sequencing and microarray studies.\" Nucleic acids research 43.7 (2015): e47-e47.\n",
    "<br>\n",
    "Traag, Vincent A., Ludo Waltman, and Nees Jan van Eck. \"From Louvain to Leiden: guaranteeing well-connected communities.\" Scientific reports 9.1 (2019): 1-12.\n",
    "<br>\n",
    "Tran, Thinh N., and Gary Bader. \"Tempora: cell trajectory inference using time-series single-cell RNA sequencing data.\" bioRxiv (2019): 846907.\n",
    "<br>\n",
    "Trapnell, Cole, et al. \"The dynamics and regulators of cell fate decisions are revealed by pseudotemporal ordering of single cells.\" Nature biotechnology 32.4 (2014): 381.\n",
    "<br>\n",
    "van Dijk, D., Nainys, J., Sharma, R., Kathail, P., Carr, A.J., Moon, K.R., Mazutis, L., Wolf, G., Krishnaswamy, S. and Pe'er, D. (2017) MAGIC: A diffusion-based imputation method reveals gene-gene interactions in single-cell RNA-sequencing data. BioRxiv, 111591.\n",
    "<br>\n",
    "Weinreb, Caleb, et al. \"Fundamental limits on dynamic inference from single-cell snapshots.\" Proceedings of the National Academy of Sciences 115.10 (2018): E2467-E2476.\n",
    "<br>\n",
    "Zheng, Grace XY, et al. \"Massively parallel digital transcriptional profiling of single cells.\" Nature communications 8.1 (2017): 1-12."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "appyter_new",
   "language": "python",
   "name": "appyter_new"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "179.429px"
   },
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 122.857142,
   "position": {
    "height": "40px",
    "left": "627.143px",
    "right": "20px",
    "top": "120px",
    "width": "250px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "none",
   "window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
