{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# scRNA-seq Analysis Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This pipeline enables you to analyze and visualize your single cell RNA sequencing datasets with an array of algorithms and data visualization methods. The pipeline includes quality control, library size analysis, normalization, plotting most highly expressed genes, plotting samples, MAGIC normalization, clustering, enrichment analysis, trajectory inference, and cell type prediction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic libraries\n",
    "import pandas as pd\n",
    "import random\n",
    "import time\n",
    "import numpy as np\n",
    "import warnings\n",
    "\n",
    "\n",
    "# Visualization\n",
    "import scipy.stats as ss\n",
    "import IPython\n",
    "from IPython.display import HTML, display, Markdown, IFrame, FileLink, Image\n",
    "from itertools import combinations\n",
    "\n",
    "# Data analysis\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import quantile_transform\n",
    "from sklearn import cluster\n",
    "from sklearn.metrics import silhouette_score\n",
    "from sklearn.manifold import TSNE\n",
    "import umap\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "from magic import MAGIC as MG\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "from maayanlab_bioinformatics.enrichment.crisp import enrich_crisp, fisher_overlap\n",
    "\n",
    "# Bokeh\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.models import HoverTool, CustomJS, ColumnDataSource, Span, Select, Legend, PreText, Paragraph, LinearColorMapper, ColorBar, CategoricalColorMapper\n",
    "from bokeh.layouts import layout, row, column, gridplot\n",
    "from bokeh.palettes import all_palettes\n",
    "import colorcet as cc\n",
    "from bokeh.palettes import Category20\n",
    "\n",
    "# External Code\n",
    "from utils import *\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% do SectionField(\n",
    "    name='Data_Section1',\n",
    "    title='Load your Dataset 1',\n",
    "    subtitle='Load your expression data. In comma/tab separated formats, genes should be in rows and samples shoud be in columns. You can also upload Cell Ranger files (matrix, genes, and barcords files). ',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Data_Section2',\n",
    "    title='Load your Dataset 2',\n",
    "    subtitle='Load your expression data. In comma/tab separated formats, genes should be in rows and samples shoud be in columns. You can also upload Cell Ranger files (matrix, genes, and barcords files). ',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Data_Section3',\n",
    "    title='Load your Dataset 3',\n",
    "    subtitle='Load your expression data. In comma/tab separated formats, genes should be in rows and samples shoud be in columns. You can also upload Cell Ranger files (matrix, genes, and barcords files). ',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Data_Section4',\n",
    "    title='Load your Dataset 4',\n",
    "    subtitle='Load your expression data. In comma/tab separated formats, genes should be in rows and samples shoud be in columns. You can also upload Cell Ranger files (matrix, genes, and barcords files). ',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Normalization_Section',\n",
    "    title='Select Normalization Methods',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Visualization_Section',\n",
    "    title='Select Visualization Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='DEG_Section',\n",
    "    title='Select Differentially Exprssed Gene Analysis Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='TI_Section',\n",
    "    title='Select Trajectory Inference Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Time_TI_Section',\n",
    "    title='Select Time-series Trajectory Inference Parameters',\n",
    "    subtitle='',\n",
    "    img='analysis.png'\n",
    "    \n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set file_kind1 = TabField(\n",
    "    name='file_kind1',\n",
    "    label='Data file',\n",
    "    default='.mtx from 10x Genomics',\n",
    "    description='Upload your expression files',\n",
    "    choices={\n",
    "        'Plain text': [             \n",
    "            FileField(\n",
    "                name='rnaseq_data_filename1', \n",
    "                label='RNA-seq data file (.csv, .txt or .tsv)', \n",
    "                default='GSE110499_example_expression.txt',\n",
    "                examples={'GSE110499_example_expression.txt': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/GSE110499-expression.txt'}, \n",
    "                description='Upload RNA-seq expression data as comma seperated or tab seperated format. The index of the dataset are genes, the columns are samples.', \n",
    "                section='Data_Section1'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='meta_data_filename1', \n",
    "                label='(Optional) Meta data file (.csv, .txt or .tsv)', \n",
    "                default='GSE110499_example_metadata.txt',\n",
    "                examples={'GSE110499_example_metadata.txt': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/GSE110499-metadata.txt'}, \n",
    "                description='Upload metadata as two-column comma seperated or tab seperated format. The first column contains sample IDs and the other column contains sample labels',\n",
    "                section='Data_Section1'\n",
    "            ),         \n",
    "            StringField(\n",
    "                name='meta_class_column_name1', \n",
    "                label='(Optional) Class column name in metadata', \n",
    "                default='prep-site', \n",
    "                description='class column name of metadata', \n",
    "                section='Data_Section1'\n",
    "            )\n",
    "\n",
    "        ],\n",
    "        \n",
    "        '.mtx from 10x Genomics': [\n",
    "            FileField(\n",
    "                name='mtx_data_filename1', \n",
    "                label='RNA-seq data file (.mtx)', \n",
    "                default='example_matrix_GDC_C3N-00662.mtx',\n",
    "                examples={'example_matrix_GDC_C3N-00662.mtx': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/example_matrix_GDC_C3N-00662.mtx'}, \n",
    "                description='Expression data file from 10x Genomics need to be in .mtx format which store the expression data in sparse matrix.',\n",
    "                section='Data_Section1'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='gene_data_filename1', \n",
    "                label='Gene infomation file (.tsv)', \n",
    "                default='example_features_GDC_C3N-00662.tsv',\n",
    "                examples={'example_features_GDC_C3N-00662.tsv': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/example_features_GDC_C3N-00662.tsv'}, \n",
    "                description='A tab delimited file of the corresponding genes in the .mtx expression matrix file. The first column should be the gene IDs whereas the secound column corresponds to gene symbols.', \n",
    "                section='Data_Section1'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='barcode_data_filename1', \n",
    "                label='Barcode information file (.tsv)', \n",
    "                default='example_barcodes_GDC_C3N-00662.tsv',\n",
    "                examples={'example_barcodes_GDC_C3N-00662.tsv': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/example_barcodes_GDC_C3N-00662.tsv'}, \n",
    "                description='A tab delimited file of the corresponding barcodes in the .mtx expression matrix file. The first column should be the unique barcodes for cells.', \n",
    "                section='Data_Section1'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='meta_data_filename_mtx1', \n",
    "                label='(Optional) Meta data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                description='Upload metadata as two-column comma seperated or tab seperated format. The first column contains sample IDs and the other column contains sample labels',\n",
    "                section='Data_Section1'\n",
    "            ),         \n",
    "            StringField(\n",
    "                name='meta_class_column_name_mtx1', \n",
    "                label='(Optional) Class column name in metadata', \n",
    "                default='', \n",
    "                description='class column name of metadata', \n",
    "                section='Data_Section1'\n",
    "            )\n",
    "        ]\n",
    "        \n",
    "    },\n",
    "    section = 'Data_Section1',\n",
    ") %}\n",
    "{% set dataset_name1 = StringField(\n",
    "    name='dataset_name1', \n",
    "    label='Dataset Name', \n",
    "    default='Dataset1', \n",
    "    description='', \n",
    "    section='Data_Section1'\n",
    "    )\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set file_kind2 = TabField(\n",
    "    name='file_kind2',\n",
    "    label='Data file',\n",
    "    default='Plain text',\n",
    "    description='Upload your expression files',\n",
    "    choices={\n",
    "        'Plain text': [             \n",
    "            FileField(\n",
    "                name='rnaseq_data_filename2', \n",
    "                label='RNA-seq data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                description='Upload RNA-seq expression data as comma seperated or tab seperated format. The index of the dataset are genes, the columns are samples.', \n",
    "                section='Data_Section2'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='meta_data_filename2', \n",
    "                label='(Optional) Meta data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                description='Upload metadata as two-column comma seperated or tab seperated format. The first column contains sample IDs and the other column contains sample labels',\n",
    "                section='Data_Section2'\n",
    "            ),         \n",
    "            StringField(\n",
    "                name='meta_class_column_name2', \n",
    "                label='(Optional) Class column name in metadata', \n",
    "                default='', \n",
    "                description='class column name of metadata', \n",
    "                section='Data_Section2'\n",
    "            )\n",
    "\n",
    "\n",
    "        ],\n",
    "        \n",
    "        '.mtx from 10x Genomics': [\n",
    "            FileField(\n",
    "                name='mtx_data_filename2', \n",
    "                label='RNA-seq data file (.mtx)', \n",
    "                default='',\n",
    "                description='Expression data file from 10x Genomics need to be in .mtx format which store the expression data in sparse matrix.',\n",
    "                section='Data_Section2'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='gene_data_filename2', \n",
    "                label='Gene infomation file (.tsv)', \n",
    "                default='',\n",
    "                description='A tab delimited file of the corresponding genes in the .mtx expression matrix file. The first column should be the gene IDs whereas the secound column corresponds to gene symbols.', \n",
    "                section='Data_Section2'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='barcode_data_filename2', \n",
    "                label='Barcode information file (.tsv)', \n",
    "                default='',\n",
    "                description='A tab delimited file of the corresponding barcodes in the .mtx expression matrix file. The first column should be the unique barcodes for cells.', \n",
    "                section='Data_Section2'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='meta_data_filename_mtx2', \n",
    "                label='(Optional) Meta data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                description='Upload metadata as two-column comma seperated or tab seperated format. The first column contains sample IDs and the other column contains sample labels',\n",
    "                section='Data_Section2'\n",
    "            ),         \n",
    "            StringField(\n",
    "                name='meta_class_column_name_mtx2', \n",
    "                label='(Optional) Class column name in metadata', \n",
    "                default='', \n",
    "                description='class column name of metadata', \n",
    "                section='Data_Section2'\n",
    "            )\n",
    "        ]\n",
    "        \n",
    "    },\n",
    "    section = 'Data_Section2',\n",
    ") %}\n",
    "{% set dataset_name2 = StringField(\n",
    "    name='dataset_name2', \n",
    "    label='Dataset Name', \n",
    "    default='Dataset2', \n",
    "    description='', \n",
    "    section='Data_Section2'\n",
    "    )\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set file_kind3 = TabField(\n",
    "    name='file_kind3',\n",
    "    label='Data file',\n",
    "    default='Plain text',\n",
    "    description='Upload your expression files',\n",
    "    choices={\n",
    "        'Plain text': [             \n",
    "            FileField(\n",
    "                name='rnaseq_data_filename3', \n",
    "                label='RNA-seq data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                description='Upload RNA-seq expression data as comma seperated or tab seperated format. The index of the dataset are genes, the columns are samples.', \n",
    "                section='Data_Section3'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='meta_data_filename3', \n",
    "                label='(Optional) Meta data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                description='Upload metadata as two-column comma seperated or tab seperated format. The first column contains sample IDs and the other column contains sample labels',\n",
    "                section='Data_Section3'\n",
    "            ),         \n",
    "            StringField(\n",
    "                name='meta_class_column_name3', \n",
    "                label='(Optional) Class column name in metadata', \n",
    "                default='', \n",
    "                description='class column name of metadata', \n",
    "                section='Data_Section3'\n",
    "            )\n",
    "\n",
    "\n",
    "        ],\n",
    "        \n",
    "        '.mtx from 10x Genomics': [\n",
    "            FileField(\n",
    "                name='mtx_data_filename3', \n",
    "                label='RNA-seq data file (.mtx)', \n",
    "                default='',\n",
    "                description='Expression data file from 10x Genomics need to be in .mtx format which store the expression data in sparse matrix.',\n",
    "                section='Data_Section3'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='gene_data_filename3', \n",
    "                label='Gene infomation file (.tsv)', \n",
    "                default='',\n",
    "                description='A tab delimited file of the corresponding genes in the .mtx expression matrix file. The first column should be the gene IDs whereas the secound column corresponds to gene symbols.', \n",
    "                section='Data_Section3'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='barcode_data_filename3', \n",
    "                label='Barcode information file (.tsv)', \n",
    "                default='',\n",
    "                description='A tab delimited file of the corresponding barcodes in the .mtx expression matrix file. The first column should be the unique barcodes for cells.', \n",
    "                section='Data_Section3'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='meta_data_filename_mtx3', \n",
    "                label='(Optional) Meta data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                description='Upload metadata as two-column comma seperated or tab seperated format. The first column contains sample IDs and the other column contains sample labels',\n",
    "                section='Data_Section3'\n",
    "            ),         \n",
    "            StringField(\n",
    "                name='meta_class_column_name_mtx3', \n",
    "                label='(Optional) Class column name in metadata', \n",
    "                default='', \n",
    "                description='class column name of metadata', \n",
    "                section='Data_Section3'\n",
    "            )\n",
    "        ]\n",
    "        \n",
    "    },\n",
    "    section = 'Data_Section3',\n",
    ") %}\n",
    "{% set dataset_name3 = StringField(\n",
    "    name='dataset_name3', \n",
    "    label='Dataset Name', \n",
    "    default='Dataset3', \n",
    "    description='', \n",
    "    section='Data_Section3'\n",
    "    )\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set file_kind4 = TabField(\n",
    "    name='file_kind4',\n",
    "    label='Data file',\n",
    "    default='Plain text',\n",
    "    description='Upload your expression files',\n",
    "    choices={\n",
    "        'Plain text': [             \n",
    "            FileField(\n",
    "                name='rnaseq_data_filename4', \n",
    "                label='RNA-seq data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                description='Upload RNA-seq expression data as comma seperated or tab seperated format. The index of the dataset are genes, the columns are samples.', \n",
    "                section='Data_Section4'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='meta_data_filename4', \n",
    "                label='(Optional) Meta data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                description='Upload metadata as two-column comma seperated or tab seperated format. The first column contains sample IDs and the other column contains sample labels',\n",
    "                section='Data_Section4'\n",
    "            ),         \n",
    "            StringField(\n",
    "                name='meta_class_column_name4', \n",
    "                label='(Optional) Class column name in metadata', \n",
    "                default='', \n",
    "                description='class column name of metadata', \n",
    "                section='Data_Section4'\n",
    "            )\n",
    "\n",
    "\n",
    "        ],\n",
    "        \n",
    "        '.mtx from 10x Genomics': [\n",
    "            FileField(\n",
    "                name='mtx_data_filename4', \n",
    "                label='RNA-seq data file (.mtx)', \n",
    "                default='',\n",
    "                description='Expression data file from 10x Genomics need to be in .mtx format which store the expression data in sparse matrix.',\n",
    "                section='Data_Section4'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='gene_data_filename4', \n",
    "                label='Gene infomation file (.tsv)', \n",
    "                default='',\n",
    "                description='A tab delimited file of the corresponding genes in the .mtx expression matrix file. The first column should be the gene IDs whereas the secound column corresponds to gene symbols.', \n",
    "                section='Data_Section4'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='barcode_data_filename4', \n",
    "                label='Barcode information file (.tsv)', \n",
    "                default='',\n",
    "                description='A tab delimited file of the corresponding barcodes in the .mtx expression matrix file. The first column should be the unique barcodes for cells.', \n",
    "                section='Data_Section4'\n",
    "            ),\n",
    "            FileField(\n",
    "                name='meta_data_filename_mtx4', \n",
    "                label='(Optional) Meta data file (.csv, .txt or .tsv)', \n",
    "                default='',\n",
    "                description='Upload metadata as two-column comma seperated or tab seperated format. The first column contains sample IDs and the other column contains sample labels',\n",
    "                section='Data_Section4'\n",
    "            ),         \n",
    "            StringField(\n",
    "                name='meta_class_column_name_mtx4', \n",
    "                label='(Optional) Class column name in metadata', \n",
    "                default='', \n",
    "                description='class column name of metadata', \n",
    "                section='Data_Section4'\n",
    "            )\n",
    "        ]\n",
    "        \n",
    "    },\n",
    "    section = 'Data_Section4',\n",
    ") %}\n",
    "{% set dataset_name4 = StringField(\n",
    "    name='dataset_name4', \n",
    "    label='Dataset Name', \n",
    "    default='Dataset4', \n",
    "    description='', \n",
    "    section='Data_Section4'\n",
    "    )\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set qc_filter_genes = BoolField(\n",
    "    name='qc_filter_genes', \n",
    "    label='Filter cells by quality control?', \n",
    "    default='true', \n",
    "    description='Check if you want cells to be filtered by mitochondrial gene expressions', \n",
    "    section='Normalization_Section')\n",
    "%}\n",
    "{% set qc_threshold = FloatField(\n",
    "    name='qc_threshold', \n",
    "    label='Mitochondria Quality Control threshold', \n",
    "    default=0.05,\n",
    "    min=0.0,\n",
    "    max=1.0,\n",
    "    step=0.001,\n",
    "    description='Remove cells that have too many mitochondrial genes expressed.', \n",
    "    section='Normalization_Section')\n",
    "%} \n",
    "{% set qc_filter_doublets = BoolField(\n",
    "    name='qc_filter_doublets', \n",
    "    label='Filter cells by doublet prediction?', \n",
    "    default='false', \n",
    "    description='Check if you want cells to be filtered by doublet prediction. If you analyze large data, it may take a long time. ', \n",
    "    section='Normalization_Section')\n",
    "%}\n",
    "{% set log_normalization = BoolField(\n",
    "    name='log_normalization', \n",
    "    label='Log normalization?', \n",
    "    default='true', \n",
    "    description='Check if you want the dataset to be log-transformed', \n",
    "    section='Normalization_Section')\n",
    "%}\n",
    "{% set normalization_method = ChoiceField(\n",
    "    name='normalization_method',\n",
    "    label='Normalization',\n",
    "    choices={'None': 'None', 'Satija et al. (2015, Nature Biotechnology)': 'Seurat', 'Zheng et al. (2017, Nature Communications)': 'Zheng17','Weinreb et al. (2018, PNAS)': 'Weignreb17'},\n",
    "    default='Satija et al. (2015, Nature Biotechnology)', \n",
    "    description='Standard normlization recipe for scRNA-seq datasets', \n",
    "    section='Normalization_Section')\n",
    "%}\n",
    "{% set magic = BoolField(\n",
    "    name='magic', \n",
    "    label='MAGIC?', \n",
    "    default='false', \n",
    "    description='Check if you want impute missing data with Markov Affinity-based Graph Imputation of Cells (MAGIC)', \n",
    "    section='Normalization_Section')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set interactive_plot = BoolField(\n",
    "    name='interactive_plot', \n",
    "    label='Interactive plot?', \n",
    "    default='false', \n",
    "    description='Check if you want interactive plots. It may require high computational resources and it may take time to render.', \n",
    "    section='Visualization_Section')\n",
    "%}\n",
    "{% set highest_expr_n_genes = IntField(\n",
    "    name='highest_expr_n_genes', \n",
    "    label='Genes for highest expression', \n",
    "    min=0, \n",
    "    max=100, \n",
    "    default=10, \n",
    "    description='The number of genes with highest expression values', \n",
    "    section='Visualization_Section')\n",
    "%}\n",
    "\n",
    "gene_list_for_clustergrammer = {{TextField(\n",
    "    name='gene_list_for_clustergrammer', \n",
    "    label='Gene List for Clustergrammer (Optional)', \n",
    "    default='', \n",
    "    description='Paste your gene list (One gene per row) for Clustergrammer heatmap plots.', \n",
    "    section = 'Visualization_Section')}}\n",
    "{% set clustering_topk = IntField(\n",
    "    name='clustering_topk', \n",
    "    label='Genes for clustergrammer', \n",
    "    min=0, \n",
    "    max=1000, \n",
    "    default=800, \n",
    "    description='The number of genes with largest variance for Clustergrammer', \n",
    "    section='Visualization_Section')\n",
    "%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set deg = BoolField(\n",
    "    name='deg',\n",
    "    label='Differentially expressed gene analysis?',\n",
    "    default='true',\n",
    "    description='Check if you want differentially expressed gene analysis', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "{% set integration_option = MultiChoiceField(\n",
    "    name='integration_option',\n",
    "    label='Analysis option for multiple datasets',\n",
    "    choices=['intra-dataset','integrated-dataset', 'inter-dataset'],\n",
    "    default=['intra-dataset'], \n",
    "    description='If you upload multiple datasets, select an analysis option.', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "\n",
    "{% set enrichment_groupby = ChoiceField(\n",
    "    name='enrichment_groupby',\n",
    "    label='Group for differentially expressed gene analysis',\n",
    "    choices={'Cluster': 'Cluster', 'User-defined class': 'user_defined_class'},\n",
    "    default='Cluster', \n",
    "    description='Specify groups for enrichment analysis. Clusters will be automatically generated by a clustering method', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "{% set diff_gex_method = ChoiceField(\n",
    "    name='diff_gex_method',\n",
    "    label='Differential expression analysis method',\n",
    "    choices={'limma': 'limma', 'wilcoxon': 'wilcoxon', 'characteristic direction': 'characteristic_direction', 'edgeR': 'edgeR', 'DESeq2': 'DESeq2'},\n",
    "    default='wilcoxon', \n",
    "    description='Set a method to get differentially expressed genes', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "\n",
    "\n",
    "{% set gene_topk = IntField(\n",
    "    name='gene_topk', \n",
    "    label='Maximum genes for Enrichr', \n",
    "    min=0, \n",
    "    max=1000, \n",
    "    default=200, \n",
    "    description='The maximum number of genes discovered by the DEG method', \n",
    "    section='DEG_Section')\n",
    "%}\n",
    "\n",
    "\n",
    "{% set libraries_tab = TabField(\n",
    "    name='libraries_tab',\n",
    "    label='Enrichr libraries?',\n",
    "    default='Yes',\n",
    "    description='',\n",
    "    choices={\n",
    "        'Yes': [             \n",
    "            MultiChoiceField(\n",
    "                name='enrichr_libraries',\n",
    "                label='Select Enrichr Libraries (upto 2)',\n",
    "                descriptions='Enrichr libraries to be visualized. Select one or two libraries',\n",
    "                choices=['Gene Ontology',\n",
    "                        'Pathway',\n",
    "                        'Kinase',\n",
    "                        'Transcription Factor',\n",
    "                        'miRNA',\n",
    "                        'Cell Type',\n",
    "                         'Disease'],\n",
    "                default=['Cell Type',\n",
    "                         'Transcription Factor',\n",
    "                        ],\n",
    "                section='DEG_Section'\n",
    "                ),\n",
    "            \n",
    "\n",
    "        ],\n",
    "        \n",
    "        'No': [\n",
    "            FileField(\n",
    "                name='library_filename', \n",
    "                label='Upload your library file (.gmt)', \n",
    "                default='HubMap_ASCT_plus_B_augmented_w_RNAseq_Coexpression.gmt',\n",
    "                examples={'PBMC_cell_type_biomarkers.gmt': 'https://appyters.maayanlab.cloud/storage/SC_RNA_seq/PBMC_cell_type_biomarkers.gmt'}, \n",
    "                description='',\n",
    "                section='Data_Section'\n",
    "            ),\n",
    "            \n",
    "        ]\n",
    "        \n",
    "    },\n",
    "    section = 'DEG_Section',\n",
    ") %}\n",
    "\n",
    "\n",
    "\n",
    "{% set nr_genesets = IntField(\n",
    "    name='nr_genesets', \n",
    "    label='Top ranked gene sets', \n",
    "    min=0, \n",
    "    max=100,\n",
    "    default=15, \n",
    "    description='the number of result gene sets', \n",
    "    section='DEG_Section')\n",
    "%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set trajectory = BoolField(\n",
    "    name='trajectory',\n",
    "    label='Trajectory Inference?',\n",
    "    default='true',\n",
    "    description='Check if you want trajectory inference analysis', \n",
    "    section='TI_Section')\n",
    "%}\n",
    "\n",
    "{% set trajectory_method = ChoiceField(\n",
    "    name='trajectory_method',\n",
    "    label='Trajectory inference method',\n",
    "    choices={'DPT(diffusion pseudotime)': 'dpt', 'monocle': 'monocle'},\n",
    "    default='DPT(diffusion pseudotime)', \n",
    "    description='Trajectory inference algorithm', \n",
    "    section='TI_Section')\n",
    "%}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% set time_series_trajectory = BoolField(\n",
    "    name='time_series_trajectory', \n",
    "    label='Time series trajectory analysis?', \n",
    "    default='false', \n",
    "    description='Check if you want time-series trajectory analysis', \n",
    "    section='Time_TI_Section')\n",
    "%}\n",
    "{% set timepoint_labels_column_name = StringField(\n",
    "    name='timepoint_labels_column_name', \n",
    "    label='Timepoint column name in metadata', \n",
    "    default='Timepoints', \n",
    "    description='Timepoint column name of metadata', \n",
    "    section='Time_TI_Section')\n",
    "%}\n",
    "timepoint_labels = {{ TextField(\n",
    "    name='timepoint_labels', \n",
    "    label='Ordered timepoint labels in the timepoint label column', \n",
    "    default='', \n",
    "    description='Paste your timepoint labels in order (One timepoint per row). e.g., 0H, 12H, 24H', \n",
    "    section = 'Time_TI_Section') }}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "rnaseq_data_filename1 = \"\"\n",
    "mtx_data_filename1 = \"\"\n",
    "gene_data_filename1 = \"\"\n",
    "barcode_data_filename1 = \"\"\n",
    "\n",
    "{%- if file_kind1.raw_value == 'Plain text' %}\n",
    "rnaseq_data_filename1 = {{ file_kind1.value[0] }}\n",
    "meta_data_filename1 = {{ file_kind1.value[1] }}\n",
    "meta_class_column_name1 = {{ file_kind1.value[2] }}\n",
    "\n",
    "{%- elif file_kind1.raw_value == '.mtx from 10x Genomics' %}\n",
    "mtx_data_filename1 = {{ file_kind1.value[0] }}\n",
    "gene_data_filename1 = {{ file_kind1.value[1] }}\n",
    "barcode_data_filename1 = {{ file_kind1.value[2] }}\n",
    "\n",
    "meta_data_filename1 = {{ file_kind1.value[3] }}\n",
    "meta_class_column_name1 = {{ file_kind1.value[4] }}\n",
    "\n",
    "{%- endif %}\n",
    "dataset_name1 = \"{{dataset_name1.value}}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "rnaseq_data_filename2 = \"\"\n",
    "mtx_data_filename2 = \"\"\n",
    "gene_data_filename2 = \"\"\n",
    "barcode_data_filename2 = \"\"\n",
    "\n",
    "{%- if file_kind2.raw_value == 'Plain text' %}\n",
    "rnaseq_data_filename2 = {{ file_kind2.value[0] }}\n",
    "meta_data_filename2 = {{ file_kind2.value[1] }}\n",
    "meta_class_column_name2 = {{ file_kind2.value[2] }}\n",
    "\n",
    "{%- elif file_kind2.raw_value == '.mtx from 10x Genomics' %}\n",
    "mtx_data_filename2 = {{ file_kind2.value[0] }}\n",
    "gene_data_filename2 = {{ file_kind2.value[1] }}\n",
    "barcode_data_filename2 = {{ file_kind2.value[2] }}\n",
    "\n",
    "meta_data_filename2 = {{ file_kind2.value[3] }}\n",
    "meta_class_column_name2 = {{ file_kind2.value[4] }}\n",
    "\n",
    "{%- endif %}\n",
    "dataset_name2 = \"{{dataset_name2.value}}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "rnaseq_data_filename3 = \"\"\n",
    "mtx_data_filename3 = \"\"\n",
    "gene_data_filename3 = \"\"\n",
    "barcode_data_filename3 = \"\"\n",
    "\n",
    "{%- if file_kind3.raw_value == 'Plain text' %}\n",
    "rnaseq_data_filename3 = {{ file_kind3.value[0] }}\n",
    "meta_data_filename3 = {{ file_kind3.value[1] }}\n",
    "meta_class_column_name3 = {{ file_kind3.value[2] }}\n",
    "\n",
    "{%- elif file_kind3.raw_value == '.mtx from 10x Genomics' %}\n",
    "mtx_data_filename3 = {{ file_kind3.value[0] }}\n",
    "gene_data_filename3 = {{ file_kind3.value[1] }}\n",
    "barcode_data_filename3 = {{ file_kind3.value[2] }}\n",
    "\n",
    "meta_data_filename3 = {{ file_kind3.value[3] }}\n",
    "meta_class_column_name3 = {{ file_kind3.value[4] }}\n",
    "\n",
    "{%- endif %}\n",
    "dataset_name3 = \"{{dataset_name3.value}}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "rnaseq_data_filename4 = \"\"\n",
    "mtx_data_filename4 = \"\"\n",
    "gene_data_filename4 = \"\"\n",
    "barcode_data_filename4 = \"\"\n",
    "\n",
    "{%- if file_kind4.raw_value == 'Plain text' %}\n",
    "rnaseq_data_filename4 = {{ file_kind4.value[0] }}\n",
    "meta_data_filename4 = {{ file_kind4.value[1] }}\n",
    "meta_class_column_name4 = {{ file_kind4.value[2] }}\n",
    "\n",
    "{%- elif file_kind4.raw_value == '.mtx from 10x Genomics' %}\n",
    "mtx_data_filename4 = {{ file_kind4.value[0] }}\n",
    "gene_data_filename4 = {{ file_kind4.value[1] }}\n",
    "barcode_data_filename4 = {{ file_kind4.value[2] }}\n",
    "\n",
    "meta_data_filename4 = {{ file_kind4.value[3] }}\n",
    "meta_class_column_name4 = {{ file_kind4.value[4] }}\n",
    "\n",
    "{%- endif %}\n",
    "dataset_name4 = \"{{dataset_name4.value}}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%- if libraries_tab.raw_value == 'Yes' %}\n",
    "enrichr_libraries_filename = []\n",
    "enrichr_libraries = {{ libraries_tab.value[0] }}\n",
    "\n",
    "{%- else %}\n",
    "enrichr_libraries_filename = {{ libraries_tab.value[0] }}\n",
    "enrichr_libraries = []\n",
    "{%- endif %}\n",
    "libraries_tab = \"{{libraries_tab.raw_value}}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "interactive_plot = {{interactive_plot.value}}\n",
    "qc_filter_genes = {{qc_filter_genes.value}}\n",
    "qc_threshold = {{qc_threshold.value}}\n",
    "qc_filter_doublets = {{qc_filter_doublets.value}}\n",
    "log_normalization = {{log_normalization.value}}\n",
    "normalization_method = \"{{normalization_method.value}}\"\n",
    "magic = {{magic.value}}\n",
    "\n",
    "highest_expr_n_genes = {{highest_expr_n_genes.value}}\n",
    "clustering_topk = {{clustering_topk.value}}\n",
    "\n",
    "deg = {{deg.value}}\n",
    "integration_option = {{integration_option.value}}\n",
    "enrichment_groupby = \"{{enrichment_groupby.value}}\"\n",
    "diff_gex_method = \"{{diff_gex_method.value}}\"\n",
    "gene_topk = {{gene_topk.value}}\n",
    "nr_genesets = {{nr_genesets.value}}\n",
    "\n",
    "trajectory = {{trajectory.value}}\n",
    "trajectory_method = \"{{trajectory_method.value}}\"\n",
    "\n",
    "time_series_trajectory = {{time_series_trajectory.value}}\n",
    "timepoint_labels_column_name = \"{{timepoint_labels_column_name.value}}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings('ignore')\n",
    "random.seed(0)\n",
    "pandas2ri.activate()\n",
    "if interactive_plot == True:\n",
    "    plot_type='interactive'\n",
    "else:\n",
    "    plot_type = 'static'\n",
    "    \n",
    "results = {}\n",
    "table_counter = 1\n",
    "figure_counter = 1\n",
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.settings.set_figure_params(dpi=80)\n",
    "sc.settings.figdir=\"./\"\n",
    "pd.options.plotting.backend = \"plotly\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_list = list()\n",
    "adata_adt_list = list()\n",
    "\n",
    "adata1, table_counter, meta_class_column_name = load_data(dataset_name1, rnaseq_data_filename1, mtx_data_filename1, gene_data_filename1, barcode_data_filename1, meta_data_filename1, meta_class_column_name1, table_counter)\n",
    "adata2, table_counter, _ = load_data(dataset_name2, rnaseq_data_filename2, mtx_data_filename2, gene_data_filename2, barcode_data_filename2, meta_data_filename2, meta_class_column_name2, table_counter)\n",
    "adata3, table_counter, _ = load_data(dataset_name3, rnaseq_data_filename3, mtx_data_filename3, gene_data_filename3, barcode_data_filename3, meta_data_filename3, meta_class_column_name3, table_counter)\n",
    "adata4, table_counter, _ = load_data(dataset_name4, rnaseq_data_filename4, mtx_data_filename4, gene_data_filename4, barcode_data_filename4, meta_data_filename4, meta_class_column_name4, table_counter)\n",
    "\n",
    "# check options are valid\n",
    "if deg == True and enrichment_groupby == 'user_defined_class':\n",
    "    for meta_data_filename, adata in zip([meta_data_filename1, meta_data_filename2, meta_data_filename3, meta_data_filename4], [adata1, adata2, adata3, adata4]):\n",
    "        if adata is not None and meta_data_filename == \"\":\n",
    "            raise Exception(\"Error: If you want to perform enrichment analysis between classes, please upload metadata with class annotations for samples. Please refer to the example file. If not, please set the 'Group for differentially expressed gene analysis' option as Cluster.\")\n",
    "\n",
    "# common genes\n",
    "gene_set = set()\n",
    "for adata in [adata1, adata2, adata3, adata4]:\n",
    "    if adata is not None:\n",
    "        if len(gene_set) == 0:\n",
    "            gene_set.update(set(adata.var.index.values))\n",
    "        else:\n",
    "            gene_set = gene_set.intersection(set(adata.var.index.values))\n",
    "\n",
    "for adata in [adata1, adata2, adata3, adata4]:\n",
    "    if adata is not None:\n",
    "        adata = adata[:, list(gene_set)]\n",
    "        adata_list.append(adata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%if qc_filter_genes.value == True %}\n",
    "for adata in adata_list:\n",
    "    sc.pp.filter_genes(adata, min_cells=3)\n",
    "    sc.pp.filter_cells(adata, min_genes=200)\n",
    "    display_statistics(adata, \"### Statistics of data ###\") \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if qc_filter_doublets == True:\n",
    "    display(Markdown(\"# Predict Doublet Scores\"))\n",
    "    display(Markdown(\"Single cell RNA-seq often generates technical artifacts known as doublets where multiple cells receive the same barcode. The appyter identifies problemtic doublets by using Scrublet (Wolock, Samuel L., et al. 2019) and filters out predicted doublets. Scrublet predicts doublets using a nearest-neighbor classifier trained on observed transcriptome and simulated doublets. \"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if qc_filter_doublets == True:\n",
    "    new_adata_list = list()\n",
    "    for adata in adata_list:\n",
    "        sc.external.pp.scrublet(adata)\n",
    "        selected_pred = adata.obs[adata.obs[\"predicted_doublet\"]==False]\n",
    "        adata = adata[selected_pred.index]\n",
    "        ds_name = adata.obs[\"batch\"].unique()[0]\n",
    "        display_statistics(adata, f\"### Statistics of RNA data in {ds_name}###\")\n",
    "        new_adata_list.append(adata)\n",
    "    adata_list = new_adata_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for adata in adata_list:\n",
    "    if adata.n_obs < 20000:\n",
    "        table_counter = display_object(table_counter, \"Raw data. The table displays the first 5 rows of the quantified RNA-seq expression dataset. Rows represent genes, columns represent samples, and values show the number of mapped reads.\", adata.to_df().iloc[:10,:5].T.head(), istable=True)\n",
    "        table_counter = display_object(table_counter, \"Metadata. The table displays the metadata associated with the samples in the RNA-seq dataset. Rows represent RNA-seq samples, columns represent metadata categories.\", adata.obs.head(), istable=True)\n",
    "        table_counter = display_object(table_counter, \"Sample size for each class. The table displays the number of samples in each class.\", adata.obs.reset_index().groupby(meta_class_column_name).count(), istable=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_merged = anndata.concat(adata_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{%if qc_filter_genes.value == True %}\n",
    "# Filter Out Cells Based On Mitochondrial Genes\n",
    "High expression levels of mitochondrial genes could be an indicator of poor quality cells (Islam, Saiful, et al. 2014, Ilicic, Tomislav, et al. 2016). In a situation where the cell membrane is broken, cytoplasmic RNA will be lost, but RNAs enclosed in the mitochondria will be retained. This analysis removes single cells that are likely having the cell membrane broken.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{%if qc_filter_genes.value == True %}\n",
    "if adata_merged.var_names[0].startswith(\"ENSG\") == False:\n",
    "    mito_genes = adata_merged.var_names.str.startswith('MT-')    \n",
    "else:\n",
    "    gene_symbol_var_names = adata_merged.var_names\n",
    "    mito_genes = [True if x in gene_id_map_dict and str(gene_id_map_dict[x]).startswith('MT-') else False for x in gene_symbol_var_names ]\n",
    "\n",
    "# for each cell compute fraction of counts in mito genes vs. all genes\n",
    "adata_merged.obs['percent_mito'] = np.sum(\n",
    "    adata_merged[:, mito_genes].X, axis=1) / np.sum(adata_merged.X, axis=1)\n",
    "\n",
    "sc.pl.violin(adata_merged, ['percent_mito'],\n",
    "         jitter=0.4, multi_panel=True, show=True, save=True)\n",
    "figure_counter = display_object(figure_counter, \"Violin plot of the percentage of mitochondrial gene expression counts in each cell\", istable=False)\n",
    "display_link(\"violin.pdf\", \"Download figure\")\n",
    "adata_merged = adata_merged[adata_merged.obs.percent_mito < qc_threshold, :]\n",
    "display_statistics(adata_merged, \"### Statistics of data after QC ###\")\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_merged.raw = adata_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Library Size Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to quantify gene expression in an RNA-seq dataset, reads generated from the sequencing step are mapped to a reference genome and subsequently aggregated into numeric gene counts. Due to experimental variations and random technical noise, samples in an RNA-seq dataset often have variable amounts of the total RNA. Library size analysis calculates and displays the total number of reads mapped for each sample in the RNA-seq dataset, facilitating the identification of outlying samples and the assessment of the overall quality of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_merged.obs['reads'] = adata_merged.X.sum(axis=1)\n",
    "library_sum_df = pd.DataFrame(adata_merged.obs['reads'])\n",
    "\n",
    "fig = px.histogram(library_sum_df, x='reads')\n",
    "fig.update_yaxes(title=\"samples/cells\")\n",
    "if plot_type == \"static\":\n",
    "    fig.show(renderer=\"png\")\n",
    "else:\n",
    "    fig.show()\n",
    "    \n",
    "figure_counter = display_object(figure_counter, \"Histogram of the total number of reads mapped for each sample. The figure contains an interactive bar chart which displays the number of samples according to the total number of reads mapped to each RNA-seq sample in the dataset. Additional information for each sample is available by hovering over the bars.\", istable=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if normalization_method.value != \"None\" %}\n",
    "# Normalization\n",
    "Various normalization methods can be applied based on the selection made by the user (Zheng, Grace XY, et al. 2017, Weinreb, Caleb, et al. 2018, Butler, Andrew, et al. 2018). These normalization methods convert the raw read counts into standardized measures of gene expression by removing factors that may negatively affect the analysis.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if normalization_method.value != \"None\" %}\n",
    "adata_norm = normalize(adata_merged, normalization_method, log_normalization)\n",
    "{% endif %}\n",
    "\n",
    "{% if magic.value == True %}    \n",
    "adata_norm = run_magic(dataset=adata_norm)\n",
    "table_counter = display_object(table_counter, \"Normalized data. The table displays the expression values after normalization.\", adata_norm.uns[\"magic\"].T.head(), istable=True)\n",
    "display_statistics(adata_norm, \"### Statistics of data after normalization ###\")\n",
    "\n",
    "{% elif normalization_method.value != \"None\" %}\n",
    "table_counter = display_object(table_counter, \"Normalized data. The table displays the expression values after normalization.\", adata_norm.to_df().T.head(), istable=True)\n",
    "display_statistics(adata_norm, \"### Statistics of data after normalization ###\")\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Most Highly Expressed Genes Across All samples/Cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This analysis displays the genes that yield the highest fraction of counts in each single cell, across all cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = adata_norm.to_df()\n",
    "df = df.divide(df.sum(axis=1), axis=0)\n",
    "mean_df = df.mean().sort_values(ascending=False)\n",
    "yaxis_gene_names = list(mean_df.index)[:highest_expr_n_genes]\n",
    "median_df = df.median().sort_values(ascending=False)\n",
    "selected_median_df = median_df.loc[yaxis_gene_names]\n",
    "sorted_yaxis_gene_names = list(selected_median_df.sort_values(ascending=False).index)\n",
    "sc.pl.highest_expr_genes(adata_norm, log=True, n_top=highest_expr_n_genes, order=sorted_yaxis_gene_names, show=True, save=True)\n",
    "figure_counter = display_object(figure_counter, \"The {} genes with the highest mean fraction over all cells sorted by median.\".format(highest_expr_n_genes), istable=False)\n",
    "display_link(\"highest_expr_genes.pdf\", \"Download figure\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clustergrammer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clustergrammer (Fernandez, Nicolas F., et al. 2017) is a web-based tool for visualizing and analyzing high-dimensional data as interactive and hierarchically clustered heatmaps. It is commonly used to explore the similarity between samples in an RNA-seq dataset. In addition to identifying clusters of samples, it also allows to identify the genes which contribute to the clustering. It visualizes the top 800 genes with largest variance by default or you can upload your gene list. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run analysis\n",
    "results['clustergrammer'] = run_clustergrammer(dataset=adata_norm, meta_class_column_name=meta_class_column_name, nr_genes=clustering_topk, gene_list=gene_list_for_clustergrammer)\n",
    "\n",
    "# Display results\n",
    "plot_clustergrammar(results['clustergrammer'])\n",
    "figure_counter = display_object(figure_counter, \"Clustered heatmap plot. The figure contains an interactive heatmap displaying gene expression for each sample in the RNA-seq dataset. Every row of the heatmap represents a gene, every column represents a sample, and every cell displays normalized gene expression values. The heatmap additionally features color bars beside each column which represent prior knowledge of each sample, such as the tissue of origin or experimental treatment.\".format(highest_expr_n_genes), istable=False)\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential Gene Expression Analysis\n",
    "In the following section, the appyter computes differentially expressed genes for each cluster and performs enrichment analysis using the top genes. You can select analysis options such as intra dataset, integrated dataset, and inter dataset. Option “intra dataset” will analyze the data within each dataset. This will be useful when you want to analyze datasets from different conditions separately and see the difference between them. Option “integrated dataset” will analyze the integrated data. It will be useful when you have multiple datasets from the same condition. Option “inter dataset” will analyze the difference between datasets. It will be similar with bulk RNA-seq analysis.<br>\n",
    "- The appyter clusters samples using the Leiden algorithm (Traag, Vincent A., et al. 2019) to identify well-connected clusters in networks. In this procedure, the appyter takes into account both modalities of the data by integrating connectivity graphs generated from each modality (ref https://scanpy-tutorials.readthedocs.io/en/multiomics/cite-seq/pbmc5k.html#Clustering). It visualizes the samples colored by clusters. <br>\n",
    "- Gene expression signatures are alterations in the patterns of gene expression that occur as a result of cellular perturbations such as drug treatments, gene knock-downs or diseases. They can be quantified using differential gene expression (DGE) methods (Ritchie, Matthew E., et al. 2015, Clark, Neil R., et al. 2014), which compare gene expression between two groups of samples to identify genes whose expression is significantly altered in the perturbation.\n",
    "<br>\n",
    "- Enrichment analysis is a bioinformatics method used to identify prior knowledge terms which are over-represented in a given gene set by comparing the gene set to many annotated gene sets. The prior-knowledge gene sets can represent cell signaling pathways, molecular functions, diseases, and a wide variety of other terms obtained by processing data from multiple resources. The appyter allows you to input user-created gene set libraries or to select Enrichr libraries (Kuleshov, M.V., et al. 2016) which is a web-based application that performs enrichment analysis against a large collection of gene-set libraries. The appyter performs enrichment analysis for each cluster and visualizes the top enriched term for each cluster. <br>\n",
    "- The appyter summarizes the enrichment analysis results by showing the number of samples for each enriched term. <br>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_object_dict = dict()\n",
    "bool_plot = True\n",
    "if \"intra-dataset\" in integration_option or len(adata_list) == 1:\n",
    "    datasets = adata_norm.obs[\"batch\"].unique()\n",
    "    for dataset in datasets:\n",
    "        display(Markdown(f\"### Analysis results of {dataset}\"))\n",
    "        tmp_adata_norm = adata_norm[adata_norm.obs[\"batch\"]==dataset].copy()\n",
    "        sc.pp.neighbors(tmp_adata_norm, n_neighbors=30)\n",
    "        \n",
    "        # clustering\n",
    "        display(Markdown(f\"#### Clustering of {dataset}\"))\n",
    "        tmp_adata, figure_counter = clustering(tmp_adata_norm, dataset, bool_plot=bool_plot, figure_counter=figure_counter, batch_correction=False)\n",
    "        \n",
    "        \n",
    "        # differential_gene_expression_analysis\n",
    "        display(Markdown(f\"#### Differential Gene Expression Analysis of {dataset}\"))\n",
    "        signatures, bool_cluster, table_counter = differential_gene_expression_analysis(tmp_adata, diff_gex_method, enrichment_groupby, meta_class_column_name, table_counter)\n",
    "\n",
    "        # enrichment analysis\n",
    "        display(Markdown(f\"#### Enrichment Analysis of {dataset}\"))\n",
    "        tmp_adata, option_list, figure_counter, table_counter = visualize_enrichment_analysis(tmp_adata, signatures, meta_class_column_name, diff_gex_method, enrichr_libraries_filename, enrichr_libraries, enrichment_groupby, libraries_tab, gene_topk, bool_cluster, bool_plot, figure_counter, table_counter)\n",
    "        \n",
    "        # summary        \n",
    "        display(Markdown(f\"#### Summary of Enrichment Analysis of {dataset}\"))\n",
    "        option_list.append('leiden')\n",
    "        table_counter = summary(tmp_adata, option_list, table_counter)\n",
    "        \n",
    "        adata_object_dict[dataset] = tmp_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"integrated-dataset\" in integration_option and len(adata_list) > 1:\n",
    "    display(Markdown(f\"### Analysis Results of Integrated Dataset\"))\n",
    "    \n",
    "    sc.pp.neighbors(adata_norm, n_neighbors=30)\n",
    "    # clustering\n",
    "    display(Markdown(f\"#### Clustering\"))\n",
    "    tmp_adata, figure_counter = clustering(adata_norm, dataset=\"integrated\", bool_plot=bool_plot, figure_counter=figure_counter, batch_correction=True)\n",
    "\n",
    "    # differential_gene_expression_analysis\n",
    "    display(Markdown(f\"#### Differential Gene Expression Analysis\"))\n",
    "    signatures, bool_cluster, table_counter = differential_gene_expression_analysis(tmp_adata, diff_gex_method, enrichment_groupby, meta_class_column_name, table_counter)\n",
    "\n",
    "    # enrichment analysis\n",
    "    display(Markdown(f\"#### Enrichment Analysis\"))\n",
    "    tmp_adata, option_list, figure_counter, table_counter = visualize_enrichment_analysis(tmp_adata, signatures, meta_class_column_name, diff_gex_method, enrichr_libraries_filename, enrichr_libraries, enrichment_groupby, libraries_tab, gene_topk, bool_cluster, bool_plot, figure_counter, table_counter)\n",
    "\n",
    "    # summary\n",
    "    display(Markdown(f\"#### Summary of Enrichment Analysis\"))\n",
    "    option_list.append('leiden')\n",
    "    table_counter = summary(tmp_adata, option_list, table_counter)\n",
    "        \n",
    "    adata_object_dict[\"Integrated Dataset\"] = tmp_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"inter-dataset\" in integration_option and len(adata_list) > 1:\n",
    "    display(Markdown(f\"### Analysis Results of Inter Dataset\"))\n",
    "    bool_plot=True\n",
    "    # differential_gene_expression_analysis\n",
    "    display(Markdown(f\"#### Differential Gene Expression Analysis\"))\n",
    "    signatures, bool_cluster, table_counter = differential_gene_expression_analysis(adata_norm, diff_gex_method, \"batch\", table_counter)\n",
    "\n",
    "    # enrichment analysis\n",
    "    display(Markdown(f\"#### Enrichment Analysis\"))\n",
    "    joint, option_list, figure_counter, table_counter = visualize_enrichment_analysis(adata_norm, signatures, meta_class_column_name, diff_gex_method, enrichr_libraries_filename, enrichr_libraries, \"batch\", libraries_tab, gene_topk, bool_cluster, bool_plot, figure_counter, table_counter)\n",
    "    \n",
    "    # summary\n",
    "    display(Markdown(f\"#### Summary of Enrichment Analysis\"))\n",
    "    option_list.append('batch')\n",
    "    table_counter = summary(adata_norm, option_list, table_counter)\n",
    "    adata_object_dict[\"Inter_Dataset\"] = adata_norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if trajectory.value == True %}\n",
    "# Trajectory Inference\n",
    "Trajectory inference is a computational technique used in single-cell transcriptomics to arrange cells based on their progression through the differentiation process. It orders single cells in pseudotime, placing them along a trajectory corresponding to a biological process such as cell differentiation by taking advantage of single cells’ asynchronous progression through those processes. Diffusion pseudotime (DPT) (Haghverdi, Laleh, et al. 2016), which measures transitions between cells using diffusion-like random walks is one of the methods used to identify such trajectories. Monocle (Trapnell, Cole, et al. 2014) is a method for ordering cells by learning an explicit principal graph from the single cell expression data with advanced machine learning techniques (Reversed Graph Embedding), which robustly and accurately resolves complicated biological processes.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# trajectory\n",
    "if trajectory == True:\n",
    "    for dataset, tmp_adata in adata_object_dict.items():\n",
    "        if dataset != \"Inter-dataset\":\n",
    "            tmp_adata, figure_counter = trajectory_inference(tmp_adata, trajectory_method, figure_counter)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if time_series_trajectory.value == True %}\n",
    "# Time-series Trajectory Analysis\n",
    "When time-series data is available, it is possible to order cells by their progression through a dynamic\n",
    "process using time information. Tempora (Tran, Thinh N., and Gary Bader 2016), a pathway-based cell trajectory inference method that orders cells using time information from the data, infers developmental linages based on biological pathway information. This time-series based analysis can generate insights into dynamic processes and their biological regulation. \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if time_series_trajectory == True:\n",
    "    for dataset, tmp_adata in adata_object_dict.items():\n",
    "        if dataset != \"Inter-dataset\":\n",
    "            figure_counter = time_series_trajectory_inference(tmp_adata, timepoint_labels_column_name, timepoint_labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download Analysis Results Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for name, obj in adata_object_dict.items():\n",
    "    try: \n",
    "        obj.uns.pop(\"magic\")\n",
    "    except:\n",
    "        pass\n",
    "    display(create_download_link(obj.obs, filename=f\"analysis_results_{name}.csv\"))\n",
    "    display(create_download_link(obj, title=\"Download h5ad file: {}\", filename=f\"data_{name}.h5ad\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# References "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "Ashburner, M., Ball, C.A., Blake, J.A., Botstein, D., Butler, H., Cherry, J.M., Davis, A.P., Dolinski, K., Dwight, S.S. and Eppig, J.T. (2000) Gene Ontology: tool for the unification of biology. Nature genetics, 25, 25.\n",
    "<br>\n",
    "Butler, Andrew, et al. \"Integrating single-cell transcriptomic data across different conditions, technologies, and species.\" Nature biotechnology 36.5 (2018): 411-420.\n",
    "<br>\n",
    "Clark, N.R. and Ma’ayan, A. (2011) Introduction to statistical methods to analyze large data sets: principal components analysis. Sci. Signal., 4, tr3-tr3.\n",
    "<br>\n",
    "Clark, Neil R., et al. \"The characteristic direction: a geometrical approach to identify differentially expressed genes.\" BMC bioinformatics 15.1 (2014): 79.\n",
    "<br>\n",
    "Consortium, E.P. (2004) The ENCODE (ENCyclopedia of DNA elements) project. Science, 306, 636-640.\n",
    "<br>\n",
    "Croft, David, et al. \"The Reactome pathway knowledgebase.\" Nucleic acids research 42.D1 (2014): D472-D477.\n",
    "<br>\n",
    "Fernandez, Nicolas F., et al. \"Clustergrammer, a web-based heatmap visualization and analysis tool for high-dimensional biological data.\" Scientific data 4 (2017): 170151.\n",
    "<br>\n",
    "Haghverdi, Laleh, et al. \"Diffusion pseudotime robustly reconstructs lineage branching.\" Nature methods 13.10 (2016): 845.\n",
    "<br>\n",
    "Ilicic, Tomislav, et al. \"Classification of low quality cells from single-cell RNA-seq data.\" Genome biology 17.1 (2016): 29.\n",
    "<br>\n",
    "Islam, Saiful, et al. \"Quantitative single-cell RNA-seq with unique molecular identifiers.\" Nature methods 11.2 (2014): 163.\n",
    "<br>\n",
    "Kanehisa, M. and Goto, S. (2000) KEGG: kyoto encyclopedia of genes and genomes. Nucleic acids research, 28, 27-30.\n",
    "<br>\n",
    "Kuleshov, M.V., Jones, M.R., Rouillard, A.D., Fernandez, N.F., Duan, Q., Wang, Z., Koplev, S., Jenkins, S.L., Jagodnik, K.M. and Lachmann, A. (2016) Enrichr: a comprehensive gene set enrichment analysis web server 2016 update. Nucleic acids research, 44, W90-W97.\n",
    "<br>\n",
    "Lachmann, A., Xu, H., Krishnan, J., Berger, S.I., Mazloom, A.R. and Ma'ayan, A. (2010) ChEA: transcription factor regulation inferred from integrating genome-wide ChIP-X experiments. Bioinformatics, 26, 2438-2444.\n",
    "<br>\n",
    "Lachmann, Alexander, and Avi Ma'ayan. \"KEA: kinase enrichment analysis.\" Bioinformatics 25.5 (2009): 684-686.\n",
    "<br>\n",
    "Maaten, Laurens van der, and Geoffrey Hinton. \"Visualizing data using t-SNE.\" Journal of machine learning research 9.Nov (2008): 2579-2605.\n",
    "<br>\n",
    "McInnes, Leland, John Healy, and James Melville. \"Umap: Uniform manifold approximation and projection for dimension reduction.\" arXiv preprint arXiv:1802.03426 (2018).\n",
    "<br>\n",
    "Ritchie, Matthew E., et al. \"limma powers differential expression analyses for RNA-sequencing and microarray studies.\" Nucleic acids research 43.7 (2015): e47-e47.\n",
    "<br>\n",
    "Traag, Vincent A., Ludo Waltman, and Nees Jan van Eck. \"From Louvain to Leiden: guaranteeing well-connected communities.\" Scientific reports 9.1 (2019): 1-12.\n",
    "<br>\n",
    "Tran, Thinh N., and Gary Bader. \"Tempora: cell trajectory inference using time-series single-cell RNA sequencing data.\" bioRxiv (2019): 846907.\n",
    "<br>\n",
    "Trapnell, Cole, et al. \"The dynamics and regulators of cell fate decisions are revealed by pseudotemporal ordering of single cells.\" Nature biotechnology 32.4 (2014): 381.\n",
    "<br>\n",
    "van Dijk, D., Nainys, J., Sharma, R., Kathail, P., Carr, A.J., Moon, K.R., Mazutis, L., Wolf, G., Krishnaswamy, S. and Pe'er, D. (2017) MAGIC: A diffusion-based imputation method reveals gene-gene interactions in single-cell RNA-sequencing data. BioRxiv, 111591.\n",
    "<br>\n",
    "Weinreb, Caleb, et al. \"Fundamental limits on dynamic inference from single-cell snapshots.\" Proceedings of the National Academy of Sciences 115.10 (2018): E2467-E2476.\n",
    "<br>\n",
    "Zheng, Grace XY, et al. \"Massively parallel digital transcriptional profiling of single cells.\" Nature communications 8.1 (2017): 1-12."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "appyter_new",
   "language": "python",
   "name": "appyter_new"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "179.429px"
   },
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 122.857142,
   "position": {
    "height": "40px",
    "left": "627.143px",
    "right": "20px",
    "top": "120px",
    "width": "250px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "none",
   "window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
