{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide\n",
    "\n",
    "{% do SectionField(\n",
    "    title='FASTQ Data',\n",
    "    subtitle='Choose FASTQ files for alignment.',\n",
    "    name='FASTQ',\n",
    "    img='fastq.png',\n",
    ") %}\n",
    "\n",
    "{% set fastq_mode = TabField(\n",
    "    name='fastq_select',\n",
    "    label='FASTQ mode',\n",
    "    default='Single-read',\n",
    "    description='Choose between single-read or paired-end FASTQ data. Paired end requires the input of two FASTQ files.',\n",
    "    section='FASTQ',\n",
    "    choices={\n",
    "        'Single-read': [\n",
    "            FileField(\n",
    "                name='fastq_file',\n",
    "                label='FASTQ',\n",
    "                default=None,\n",
    "                description='Enter path to FASTQ file.',\n",
    "                section = 'FASTQ')\n",
    "        ],\n",
    "        'Paired-end': [\n",
    "            FileField(\n",
    "                name='fastq_paired_1',\n",
    "                label='FASTQ 1',\n",
    "                default=None,\n",
    "                description='Enter path to FASTQ file.',\n",
    "                section = 'FASTQ'),\n",
    "            FileField(\n",
    "                name='fastq_paired_2',\n",
    "                label='FASTQ 2',\n",
    "                default=None,\n",
    "                description='Enter path to FASTQ file.',\n",
    "                section = 'FASTQ')\n",
    "        ]\n",
    "    }\n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    title='Gene Identifier',\n",
    "    subtitle='Select identifier to aggregate gene level counts.',\n",
    "    name='IDENTIFIER',\n",
    "    img='genemap.png',\n",
    ") %}\n",
    "\n",
    "{% set identifier = ChoiceField(\n",
    "    name='identifier',\n",
    "    label='Gene identifier',\n",
    "    description='Select identifier to aggregate gene level counts.',\n",
    "    default='symbol',\n",
    "    choices=['symbol','ensembl_id','entrezgene_id'],\n",
    "    section='IDENTIFIER') %}\n",
    "\n",
    "{% do SectionField(\n",
    "    title='Alignment Algorithm Selection',\n",
    "    subtitle='Choose an alignment algorithm.',\n",
    "    name='ALIGNER',\n",
    "    img='algorithm.png',\n",
    ") %}\n",
    "\n",
    "{% set aligner_field = ChoiceField(\n",
    "    name='aligner_select',\n",
    "    label='Aligner',\n",
    "    description='Select alignment algorithm.',\n",
    "    default='kallisto',\n",
    "    choices=[\"kallisto\", \"salmon\"],\n",
    "    section='ALIGNER') %}\n",
    "\n",
    "\n",
    "{% do SectionField(\n",
    "    title='Species Selection',\n",
    "    subtitle='Choose a reference species.',\n",
    "    name='SPECIES',\n",
    "    img='species.png'\n",
    ") %}\n",
    "\n",
    "\n",
    "{% set species_field = TabField(\n",
    "    name='species_select',\n",
    "    label='Species',\n",
    "    default='Popular Species',\n",
    "    description='Select a reference species for alignment. Index will be built from Ensembl cDNA.',\n",
    "    section='SPECIES',\n",
    "    choices={\n",
    "        'Popular Species': [\n",
    "            ChoiceField(\n",
    "                name='pop_species_select',\n",
    "                label='Species',\n",
    "                description='Select species.',\n",
    "                default='Human',\n",
    "                choices=['Human', 'Mouse'],\n",
    "                section='SPECIES')\n",
    "        ],\n",
    "        'All species': [\n",
    "            ChoiceField(\n",
    "                name='all_species_select',\n",
    "                label='Species',\n",
    "                description='Select species.',\n",
    "                default='Human',\n",
    "                choices=['Abingdon island giant tortoise', 'African ostrich', \"Agassiz's desert tortoise\", 'Algerian mouse', 'Alpaca', 'Alpine marmot', 'Amazon molly', 'American beaver', 'American bison', 'American black bear', 'American mink', 'Angola colobus', 'Arabian camel', 'Arctic ground squirrel', 'Argentine black and white tegu', 'Armadillo', 'Asian bonytongue', 'Asiatic black bear', 'Atlantic cod', 'Atlantic herring', 'Atlantic salmon', 'Australian saltwater crocodile', 'Ballan wrasse', 'Barramundi perch', 'Beluga whale', 'Bengalese finch', 'Bicolor damselfish', 'Black snub-nosed monkey', 'Blind barbel', 'Blue tilapia', 'Blue tit', 'Blue whale', 'Blue-crowned manakin', 'Blue-ringed sea krait', 'Blunt-snouted clingfish', 'Bolivian squirrel monkey', 'Bonobo', 'Brazilian guinea pig', 'Brown trout', 'Budgerigar', 'Burrowing owl', \"Burton's mouthbrooder\", 'Bushbaby', 'C.intestinalis', 'C.savignyi', 'Caenorhabditis elegans', 'California sea lion', 'Canada lynx', 'Capuchin', 'Cat', 'Central bearded dragon', 'Chacoan peccary', 'Channel bull blenny', 'Channel catfish', 'Chicken', 'Chilean tinamou', 'Chimpanzee', 'Chinese hamster CHOK1GS', 'Chinese hamster CriGri', 'Chinese hamster PICR', 'Chinese medaka', 'Chinese softshell turtle', 'Chinook salmon', 'Climbing perch', 'Clown anemonefish', 'Coelacanth', 'Coho salmon', 'Collared flycatcher', 'Common canary', 'Common carp', 'Common carp german mirror', 'Common carp hebao red', 'Common carp huanghe', 'Common kestrel', 'Common snapping turtle', 'Common wall lizard', 'Common wombat', \"Coquerel's sifaka\", 'Cow', 'Crab-eating macaque', 'Damara mole rat', 'Dark-eyed junco', 'Daurian ground squirrel', 'Degu', 'Denticle herring', 'Dingo', 'Dog', 'Dog - Basenji', 'Dog - Great Dane', 'Dolphin', 'Domestic yak', 'Donkey', 'Drill', 'Drosophila melanogaster', 'Duck', 'Eastern brown snake', 'Eastern buzzard', 'Eastern happy', 'Eastern spot-billed duck', 'Electric eel', 'Elephant', 'Elephant shark', 'Emu', 'Eurasian eagle-owl', 'Eurasian red squirrel', 'Eurasian sparrowhawk', 'European seabass', 'Ferret', 'Fugu', 'Gelada', 'Giant panda', 'Gibbon', 'Gilthead seabream', 'Goat', 'Goat (black bengal)', 'Golden Hamster', 'Golden eagle', 'Golden pheasant', 'Golden snub-nosed monkey', 'Golden-collared manakin', 'Golden-line barbel', 'Goldfish', 'Goodes thornscrub tortoise', 'Gorilla', 'Gouldian finch', 'Great Tit', 'Great spotted kiwi', 'Greater amberjack', 'Greater bamboo lemur', 'Greater horseshoe bat', 'Green anole', 'Guinea Pig', 'Guppy', 'Hagfish', 'Hedgehog', 'Helmeted guineafowl', 'Horned golden-line barbel', 'Horse', 'Huchen', 'Human', 'Hybrid - Bos Indicus', 'Hybrid - Bos Taurus', 'Hyrax', 'Indian cobra', 'Indian glassy fish', 'Indian medaka', 'Indian peafowl', 'Japanese medaka HNI', 'Japanese medaka HSOK', 'Japanese medaka HdrR', 'Japanese quail', 'Javanese ricefish', 'Jewelled blenny', 'Kakapo', 'Kangaroo rat', 'Koala', 'Komodo dragon', 'Lamprey', 'Large yellow croaker', 'Leishan spiny toad', 'Leopard', 'Lesser Egyptian jerboa', 'Lesser hedgehog tenrec', 'Lion', 'Little spotted kiwi', 'Live sharksucker', 'Long-tailed chinchilla', 'Lumpfish', 'Lyretail cichlid', \"Ma's night monkey\", 'Macaque', 'Mainland tiger snake', 'Makobe Island cichlid', 'Mallard', 'Mangrove rivulus', 'Marmoset', 'Medium ground-finch', 'Meerkat', 'Megabat', 'Mexican tetra', 'Microbat', 'Midas cichlid', 'Mongolian gerbil', 'Monterrey platyfish', 'Mouse', 'Mouse 129S1/SvImJ', 'Mouse A/J', 'Mouse AKR/J', 'Mouse BALB/cJ', 'Mouse C3H/HeJ', 'Mouse C57BL/6NJ', 'Mouse CAST/EiJ', 'Mouse CBA/J', 'Mouse DBA/2J', 'Mouse FVB/NJ', 'Mouse LP/J', 'Mouse Lemur', 'Mouse NOD/ShiLtJ', 'Mouse NZO/HlLtJ', 'Mouse PWK/PhJ', 'Mouse WSB/EiJ', 'Mummichog', 'Muscovy Duck (domestic type)', 'Naked mole-rat female', 'Naked mole-rat male', 'Narwhal', 'New Caledonian crow', 'Nile tilapia', 'Northern American deer mouse', 'Northern pike', 'Northern spotted owl', 'Ocean sunfish', 'Okarito brown kiwi', 'Olive baboon', 'Opossum', 'Orange clownfish', 'Orangutan', 'Orbiculate cardinalfish', 'Oriental scops-owl', 'Pachon cavefish', 'Painted turtle', 'Paramormyrops kingsleyae', 'Periophthalmus magnuspinnatus', 'Pig', 'Pig - Bamei', 'Pig - Berkshire', 'Pig - Hampshire', 'Pig - Jinhua', 'Pig - Landrace', 'Pig - Largewhite', 'Pig - Meishan', 'Pig - Pietrain', 'Pig - Rongchang', 'Pig - Tibetan', 'Pig - Wuzhishan', 'Pig USMARC', 'Pig-tailed macaque', 'Pika', 'Pike-perch', 'Pinecone soldierfish', 'Pink-footed goose', 'Platyfish', 'Platypus', 'Polar bear', 'Prairie vole', 'Rabbit', 'Rainbow trout', 'Rat', 'Red fox', 'Red-bellied piranha', 'Reedfish', 'Ring-necked pheasant', 'Round goby', 'Ruff', 'Rufous-capped babbler', 'Ryukyu mouse', 'Saccharomyces cerevisiae', 'Sailfin molly', 'Sheep', 'Sheep (texel)', 'Sheepshead minnow', 'Shortfin molly', 'Shrew', 'Shrew mouse', 'Siamese fighting fish', 'Siberian musk deer', 'Silver-eye', 'Sloth', 'Small tree finch', 'Sooty mangabey', 'Sperm whale', 'Spiny chromis', 'Spoon-billed sandpiper', 'Spotted gar', 'Squirrel', 'Steppe mouse', 'Stickleback', 'Superb fairywren', \"Swainson's thrush\", 'Swamp eel', 'Swan goose', 'Tarsier', 'Tasmanian devil', 'Tetraodon', 'Three-toed box turtle', 'Tiger', 'Tiger tail seahorse', 'Tongue sole', 'Tree Shrew', 'Tropical clawed frog', 'Tuatara', 'Turbot', 'Turkey', 'Turquoise killifish', 'Ugandan red Colobus', 'Upper Galilee mountains blind mole rat', 'Vaquita', 'Vervet-AGM', 'Wallaby', 'West African mud turtle', 'Western mosquitofish', 'White-throated sparrow', 'Wild yak', 'Yarkand deer', 'Yellow-billed parrot', 'Yellowtail amberjack', 'Zebra finch', 'Zebra mbuna', 'Zebrafish', 'Zig-zag eel'],\n",
    "                section='SPECIES')\n",
    "        ]\n",
    "    }\n",
    ") %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "<h1> <img width=200 src=\"{{ url_for('static', filename='xalign.png') }}\" /></h1>\n",
    "<br>\n",
    "<div> This Appyter builds an alignment index for the specified species using the Ensembl database cDNA entries. Once the index is built the appyter will commence to generate transcript and gene level counts. </div>\n",
    "<br><br>\n",
    "<div style=\"font-size: 20px;\"> <img style=\"height: 40px; float: left; margin-top:-7px; margin-right: 10px;\" src=\"{{ url_for('static', filename='algorithm.png') }}\" /> {{ aligner_field }}</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "#!pip3 install --force git+https://github.com/MaayanLab/xalign.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xalign\n",
    "import time\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import display, FileLink, Markdown, HTML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "# Use the selected user input\n",
    "ALIGNER = '{{ aligner_field }}'\n",
    "\n",
    "{% if fastq_mode.raw_value == 'Align folder' %}\n",
    "FASTQ_DIR = {{ fastq_mode.value[0] }}\n",
    "{% elif fastq_mode.raw_value == 'Single-read' %}\n",
    "FASTQ = {{ fastq_mode.value[0] }}\n",
    "{% elif fastq_mode.raw_value == 'Paired-end' %}\n",
    "FASTQ_1 = '{{ fastq_mode.value[0].value }}'\n",
    "FASTQ_2 = '{{ fastq_mode.value[1].value }}'\n",
    "{% endif %}\n",
    "\n",
    "IDENTIFIER = '{{ identifier }}'\n",
    "SPECIES = xalign.organism_display_to_name('{{ species_field.value[0] }}')\n",
    "THREADS = 8"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align FASTQ data\n",
    "\n",
    "Use the xalign package to either align the contents of a complete folder, or an individual sample. In case a whole folder was selected, xalign identifies whether samples are single or paired-end and match the correct sample pairs automatically."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'FASTQ_DIR' in locals():\n",
    "    print(\"Align all samples in folder\")\n",
    "    gene_counts, transcript_counts = xalign.align_folder(SPECIES, FASTQ_DIR, aligner=ALIGNER, identifier=IDENTIFIER, t=THREADS, overwrite=False)\n",
    "    gene_counts.to_csv(\"gene_counts.tsv\", sep=\"\\t\")\n",
    "    transcript_counts.to_csv(\"transcript_counts.tsv\", sep=\"\\t\")\n",
    "elif 'FASTQ' in locals():\n",
    "    print(\"Align single-read FASTQ file\")\n",
    "    transcript_counts = xalign.align_fastq(SPECIES, FASTQ, aligner=ALIGNER, t=THREADS, overwrite=False)\n",
    "    transcript_counts.to_csv(\"transcript_counts.tsv\", sep=\"\\t\")\n",
    "    gene_counts = xalign.ensembl.agg_gene_counts(transcript_counts, \"homo_sapiens\", identifier=IDENTIFIER)\n",
    "    gene_counts.to_csv(\"gene_counts.tsv\", sep=\"\\t\")\n",
    "else:\n",
    "    print(\"Align paired-end FASTQ file\")\n",
    "    transcript_counts = xalign.align_fastq(SPECIES, [FASTQ_1, FASTQ_2], aligner=ALIGNER, t=THREADS, overwrite=False)\n",
    "    transcript_counts.to_csv(\"transcript_counts.tsv\", sep=\"\\t\")\n",
    "    gene_counts = xalign.ensembl.agg_gene_counts(transcript_counts, \"homo_sapiens\", identifier=IDENTIFIER)\n",
    "    gene_counts.to_csv(\"gene_counts.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download Files\n",
    "\n",
    "The analysis generated two download files. One contains transcript counts and the other gene level counts. For gene level counts all transcript counts matching a gene are summed. If a transcript cannot be matched to a gene the counts will be discarded at the gene level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(FileLink(\"gene_counts.tsv\", result_html_prefix=str('Download gene counts: ')))\n",
    "display(FileLink(\"transcript_counts.tsv\", result_html_prefix=str('Download transcript counts: ')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting count distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'FASTQ_DIR' in locals():\n",
    "    fig = plt.figure(figsize=(10, 6))\n",
    "    ax = fig.add_axes([0,0,1,1])\n",
    "    ax.bar(gene_counts.columns, gene_counts.sum(axis=0))\n",
    "    plt.title('Gene Counts')\n",
    "    plt.xlabel('samples')\n",
    "    plt.ylabel('read counts')\n",
    "    plt.show()\n",
    "else:\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    n, bins, patches = plt.hist(np.log2(1+gene_counts.loc[:,\"counts\"]), 100, facecolor='blue', alpha=0.5)\n",
    "    plt.title('Gene Counts')\n",
    "    plt.xlabel('log2(1 + gene counts)')\n",
    "    plt.ylabel('counts')\n",
    "    plt.show()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.8.5 64-bit",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
